---
output: html_document
editor_options: 
  chunk_output_type: console
---
# Retrieving multiple sequences in R

```{r eval = T}
library(compbio4all)
```


**By**: Avril Coghlan.  
Multiple Alignment and Phylogenetic trees
https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter5.html

**Adapted, edited and expanded**: Nathan Brouwer under the Creative Commons 3.0 Attribution License [(CC BY 3.0)](https://creativecommons.org/licenses/by/3.0/).


## Prelminaries

<!-- TODO:  add entrez_fetch_list to combio4all -->
<!-- Add output to combio4all -->
<!-- add motivation -->

```{r}
## package
library(compbio4all)
library(rentrez)    # still needed?
```


## Retrieving a set of sequences from UniProt

Using websites or *R* you can search fort DNA or protein sequences in sequence databases such as the **NCBI** database and **UniProt**.  Oftentimes, it is useful to retrieve several sequences at once. The *R* function `entrez_fetch()` from the *rentrez* package is useful for this purpose.  Other packages can also, such as `sequinr` this but *rentrez* has the cleanest interface.

<!-- have I ever mention ACNUC before -->

<!-- IS this necessary?: -->
<!-- As its input, you need to give `entrez_fetch()` a vector containing the **accessions** for the sequences you wish to retrieve, as well as the name of the **ACNUC sub-database** that the sequences should be retrieved from. In this case, we want to retrieve sequences from **UniProt**, so the sequences should be in the **swissprot** ACNUC sub-database.  (It can be tricky to know exactly where to get sequences from sometimes so I will usually give you the code to do this). -->

<!-- This isn't actually how it works -->
<!-- The `entrez_fetch()` function returns a **list** variable, in which each element is a separate **vector** containing one of the sequences of interest.  **Lists* are a common data structure in *R* and can take some getting used to, but by working with them we will become familiar with how they work.  Unfortunately there are some quirks in how *R* likes you to write code referring to lists, so I will frequently provide the necessary code for this. -->

We'll retrieve the protein sequences for these UniProt accessions 

1. P06747: rabies virus phosphoprotein
1. P0C569: Mokola virus phosphoprotein
1. O56773: Lagos bat virus phosphoprotein
1. Q5VKP1: Western Caucasian bat virus phosphoprotein

Rabies virus is the virus responsible for rabies, which is classified by the WHO as a **neglected tropical disease**. Rabies is not a major human pathogen in the USA and Europe, but is problem in Africa.   [Mokola virus](https://en.wikipedia.org/wiki/Mokola_lyssavirus() and rabies virus are closely related viruses that both belong to a group of viruses called the [Lyssaviruses](https://en.wikipedia.org/wiki/Lyssavirus). Mokola virus causes a rabies-like infection in mammals including humans.

You can type make a vector containing the names of the sequences.   Note that the accessions aren't numbers but are **quoted character strings**: 

```{r}
seqnames <- c("P06747", 
              "P0C569", 
              "O56773", 
              "Q5VKP1")  
```


Confirm that we are working with character data using `is.character()`
```{r}
is.character(seqnames)
```


We can access the first element of the vector, P06747, using **bracket notation** like this:
```{r}
seqnames[1]
```


The code to access the second and third elements of the vector of accessions is:
```{r}
# 2nd accession
seqnames[2]

# 3rd accession
seqnames[3]

```

Now let's use this vector of accessions to download sequence data.  To make sure we understand what we're doing, first we'll download just the sequences one by one.  This code retrieves the first sequence and store them in vector variable `seqs`.
```{r}
seq1 <- rentrez::entrez_fetch(db = "protein", 
                          id = seqnames[1], 
                          rettype = "fasta")      

```

We can do the next two sequences by change `id = ...` to ` seqnames[2]` and ` seqnames[3]`.
```{r}
# sequence two using seqnames[2]
seq2 <- rentrez::entrez_fetch(db = "protein", 
                          id = seqnames[2], 
                          rettype = "fasta") 

# sequence two using seqnames[3]
seq3 <- rentrez::entrez_fetch(db = "protein", 
                          id = seqnames[3], 
                          rettype = "fasta")

# sequence two using seqnames[4]
seq4 <- rentrez::entrez_fetch(db = "protein", 
                          id = seqnames[4], 
                          rettype = "fasta")
```


Each of these is in raw FASTA format
```{r}
seq1
```

To do an analysis of these we need to clean each of these vectors using `fasta_cleaner()`.
```{r}
seq1 <- fasta_cleaner(seq1, parse = T)  
seq2 <- fasta_cleaner(seq2, parse = T)  
seq3 <- fasta_cleaner(seq3, parse = T)  
seq4 <- fasta_cleaner(seq4, parse = T)  
```

Print out the first 20 letters of the first sequence
```{r}
seq1[1:20]                                   
```

Print out the first 20 letters of the second sequence
```{r}
seq2[1:20]                                   
```

We have these four sequences, each cleaned and in its own vector.  We can therefore do dotplots, alignments, and other analyses.  Creating all these separate vectors is a bit laborious.  Luckily `entrez_fetch()` can download multiple sequences for us.

## Downloading sequences in bulk

Previously we were giving `entrez_fetch` the name of just one sequence at a time by using square brackets on our `seqnames` vector, e.g. `seqnames[1]`, `seqnames[2]` etc.  If don't include square brackets, `entrez_fetch()` will download *all* of the sequences in succession and package them up into a single long, formatted string.  This may take a second or two, depending on your internet connection and how busy the NCBI servers are.

```{r}
seq_1_2_3_4 <- rentrez::entrez_fetch(db = "protein", 
                          id = seqnames, 
                          rettype = "fasta")
```

We can view what we have in a nice format using the `cat()` function.
```{r}
cat(seq_1_2_3_4)
```

This is a good way to store FASTA files, but we can't work with them in this format - they need to be in vectors.  To give us FASTA data in a usable form the `compbio4all` package has a function called `entrez_fetch_list()`, which is a **wrapper** for `entrez_fetch()` which returns each sequence in its own separate slot in a list.

```{r}
seq_1_2_3_4 <- entrez_fetch_list(db = "protein", 
                          id = seqnames, 
                          rettype = "fasta")
```

Let's look at the output
```{r}
seq_1_2_3_4
```

Note that before each sequence is its name, preceded by a dollar sign, e.g. `$P06747`.

This is the name of each element of our list.  We can confirm that we have a list using `is.list()`.

```{r}
is.list(seq_1_2_3_4)
```


The size of the list is the number of elements it contains, not the amount of data in each element.  There are 4 sequences, so 4 elements, so length = 4.
```{r}
length(seq_1_2_3_4)
```

We can access each element of the list by name, like this:

```{r}
seq_1_2_3_4$P06747
```

We can also access it by its index number, like this, using **double-bracket notation.**

```{r}
seq_1_2_3_4[[1]] #NOTE: double brackets
```


Each element of the list is a vector.  We can check this using `is.vector()` like this 
```{r}
is.vector(seq_1_2_3_4$P06747)
```

or using double brackets like this

```{r, eval = F}
is.vector(seq_1_2_3_4[[1]])
```

Its character data, which we can confirm with `class()`
```{r, eval = F}
class(seq_1_2_3_4$P06747) # dollar sign notation
class(seq_1_2_3_4[[1]])   # double-bracket notation
```




<!-- Once you have retrieved the sequences using `entrez_fetch()`, you can then use the function write.fasta() from the SeqinR package to write the sequences to a FASTA-format file. As its arguments (inputs), the write.fasta() function takes the list variable containing the sequences, and a vector containing the names of the sequences, and the name that you want to give to the FASTA-format file. For example: -->

<!-- TODO: make this a data object in package -->
<!-- TODO: where should students write this to? -->
<!-- ```{r. eval = F} -->
<!-- seqinr::write.fasta(seq_1_2_3_4, seqnames,file="data-raw/phosphoproteins.fasta") -->
<!-- ``` -->

<!-- The command above will write the sequences in list variable seqs to a FASTA-format file called “phosphoproteins.fasta” in the “My Documents” folder on your computer. -->

