---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Multiple sequence alignment in R



**By**: Nathan Brouwer, with some content adapted Coghlan (2011) [Multiple Alignment and Phylogenetic trees](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter5.html) and under the Creative Commons 3.0 Attribution License [(CC BY 3.0)](https://creativecommons.org/licenses/by/3.0/).  Functions print_msa() and clean_alignment() adapted from ([Coglan 2011](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/#)).


## Preliminaries 

<!-- all needed? -->

### Packages 

We'll be using the package `ggmsa` for the first time and you will have to install it using `install.packages("ggmsa")`.  You may be asked to re-restart R more then once during the installation process.

```{r , warning = F, message = F}
# new packages
## Only install once
# install.packages("ggmsa")
library(ggmsa)       # visualize MSA

# other packages
library(compbio4all)
library(Biostrings)  # convert FASTA to AAStringSet
library(msa)         # multiple sequence alignment
```

### Functions

The following key functions from compbio4all are used in this lesson

* fasta_cleaner()
* entrez_fetch_list()
* print_msa()
* clean_alignment()


## Multiple sequence alignment (MSA)

<!-- This paragraph adapated from Coghlan -->
A common task in bioinformatics is to download a set of related sequences from a database, and then to align those sequences using multiple alignment software. This is the first step in almost all phylogenetic analyses using sequence data.
<!-- This paragraph adapated from Coghlan -->


## Make MSA with msa()

WE'll use a package called `msa` ([Bodenhofer et al. 2015](https://academic.oup.com/bioinformatics/article/31/24/3997/197486)).  There are several packages that can do multiple sequence alignment in R, but they all require loading an external piece of alignment software that is just accessed via R.  The  `msa` package actually runs the alignment algorithms entirely in R, making workflows simpler.

### Data preparation 

The data is stored in object in combio4all
```{r}
data(seq_1_2_3_4)
```


We'll need need to convert our set of sequences to a particular format in preparation for alignment.  This is done with the `AAStringSet()` function from `Biostrings`.

```{r}
seq_1_2_3_4_stringset <- Biostrings::AAStringSet(seq_1_2_3_4)
```

This just puts things in a format that makes the software happy.  Doing this is a theme of bioinformatics work!

```{r}
seq_1_2_3_4_stringset
```

Next, we can run the alignment algorithm with the `msa()` function.  There are many algorithms and pieces software for building alignments. The `msa` packages implements three major ones: 

1. ClustalW
1. ClustalOmega
1. Muscle

We'll use ClustalW.  Depending on the size and number of sequences this may take a little bit of time.  
```{r}
virusaln <- msa(inputSeqs = seq_1_2_3_4_stringset,
                     method = "ClustalW")
```

We can view a snapshot of the alignment.
```{r}
virusaln
```


Each sequence is on a line.  The **consensus sequence** indicates something similar to the average of all the sequences and is on the bottom and labeled `Con`.  Question marks indicate that the software could not determine a consensus.  Dashes indicate either **indels** (**insertions** or **deletions**), or are added at the begining and end of sequences of unequal length so that they line up.


The output from `msa()` is a particular class of R object, a `MsaAAMultipleAlignment`.

```{r}
is(virusaln)
class(virusaln)
```

Next we'll want to visualize our alignment.   In order to do further work with the MSA we're going to - I bet you can guess what happens next - make a conversion to the object.  In this case we're going to make a subtle change by calling up the `class()` of the alignment and changing it from `MsaAAMultipleAlignment` (with "Msa" at the beginning) to `AAMultipleAlignment` (no "Msa").  (This is an annoying step and is needed because the folks who wrote the `msa` package have yet to collaobrate with the folks who wrote another package we'll use in a little bit).

Don't worry if you don't understand what's going on here - just run the code.
```{r}
class(virusaln) <- "AAMultipleAlignment"
```

## Viewing your MSA

There are several ways to view and explore your MSA

1. Within the R console using `compbio4all::print_msa()`
1. As an R plot using `ggmsa::ggmas()`
1. OPTIONAL: As a PDF using `msa::msaPrettyPrint()`

### Viewing a long multiple alignment in the R console.

If you want to view a long multiple alignment within the R console, it is convenient to view the multiple alignment in blocks.

The function `print_msa()` ([Coglan 2011](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/#))below will do this for you.  As its inputs, the function `print_msa()` takes the two things

1. `alignment`: input alignment 
1. `chunksize`: the number of columns to print out in each block.

To use `print_msa()` we first need to do a little format conversion:
```{r}
virusaln_seqinr <- msaConvert(virusaln, type = "seqinr::alignment")
```

Then we can print it out like this, making the alignment 60 bases wide:
```{r}
print_msa(alignment = virusaln_seqinr, 
          chunksize = 60)
```


<!-- TODO: do auto converion of format, print consensus -->

### Visualizing alignments as an R plot

A powerful tool for visualizing focal parts of an alignment is `ggmsa`.  If you haven't already, download it with `install.packages("ggmsa")` and load it with `library(ggmsa)`.


`ggmsa` prints a sequence alignment out within RStudio.  Alignments can be large, so its important to select a subset of the alignment for visualzation.

First, let's look at the first 20 bases of our alignment. Note that we are using `virusaln`, NOT `virusaln_seqinr` (sorry for the back and forth between objects.)
```{r}
ggmsa(virusaln,   # virusaln, NOT virusaln_seqinr
      start = 1, 
      end = 20) 
```

#### OPTIONAL: File types used by `ggmsa`

The `ggmsa` packages currently only works with certain types of alignment output.  We can see what these are with `available_msa()`.

```{r}
available_msa()
```

As you can see there are a number of ways multiple sequence alignments can be represented in *R*.  This has to do with the facts that i) There are many pieces of software / algorithms for making MSAs, and many bioinformatics packages that work with them.,

You can see that `AAMultipleAlignment` is listed, which the the format we set previously using the `class()` command.

The `msa` packages has a function `msaConvert()` which can change formats between different ways of representing MSAs which may be useful.


### OPTIONAL: Print MSA to PDF

The `msa` package has a fabulous function, `msaPrettyPrint()` for rendering an MSA to PDF.  It can take a little bit to run, and in order to vioew the PDF you need to locate the output.  (Again, we'll use `virusaln`, not `virusaln_seqinr`).

```{r, eval = F}
msaPrettyPrint(virusaln,     # virusaln, NOT virusaln_seqinr
               file = "my_msa.pdf",
               askForOverwrite = F)
```

On a Mac usually searching in Finder will locate the file even after is just created.  You can ask R where it is saving thing using `getwd()`.

```{r}
getwd()
```

You can change where R is saving things using the RStudio menu by clicking on Session -> Set Working Directory -> Choose directory...


## Discarding very poorly conserved regions from an alignment


<!-- By Coglan -->

It is often a good idea to discard very **poorly conserved** regions from a multiple sequence alignment before visualizing it or building a phylogenetic tree, as the very poorly conserved regions are likely to be regions that are either **non-homologous** between the sequences being considered (and so do not have any phylogenetic signal), or are homologous but are so **diverged** that they are very difficult to align accurately (and so may add noise to the phylogenetic analysis, and decrease the accuracy of the inferred tree).
<!-- By Coglan -->

<!-- By Coglan -->
To discard very poorly conserved regions from a multiple alignment, you can use the following R function, `clean_alignment()` (([Coglan 2011](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/#)))
<!-- By Coglan -->

<!-- By Coglan -->
The function `clean_alignment()` takes three arguments (inputs): 

1. the input alignment; 
1. minpcnongap: the minimum percent of letters in an alignment column that must be non-gap characters for the column to be kept; and 
1. minpcid: the minimum percent of pairs of letters in an alignment column that must be identical for the column to be kept.
<!-- By Coglan -->

<!-- By Coglan, with some formatting by me -->
For example, if we have a single column (**locus**) with letters “T”, “A”, “T”, “-” (in four sequences), then 75% of the letters are non-gap characters; and the pairs of letters between the three non-gap sequences are

* 1 versus 2: “T,A”, 
* 1 versus 3: “T,T”, 
* 2 versus 3: “A,T”, 

Therefore 33% of the pairs of letters are identical (**PID**) for that position in the alignment.
<!-- By Coglan -->

<!-- By Coglan -->
If you look at the multiple alignment for the virus phosphoprotein sequences (which we printed out using function `print_msa()`, see above), we can see that the last few columns are poorly aligned (contain many gaps and mismatches), and probably add noise to the phylogenetic analysis.
<!-- By Coglan -->

Let's cleave off anything with less than 30% non-gap and less than 30% PID.
NOTE: we're bac to using `virusaln_seqinr`, not `virusaln`.

```{r, eval = T}
virusaln_seqinr_clean <- clean_alignment(alignment = virusaln_seqinr,  # virusaln_seqinr
                                   minpcnongap = 30, 
                                   minpcid = 30)
```

<!-- By Coglan -->
In this case, we required that at least 30% of letters in a column are not gap characters for that column to be kept, and that at least 30% of pairs of letters in an alignment column must be identical for the column to be kept.
<!-- By Coglan -->

We can print out the filtered alignment by typing:

```{r, eval = F}
print_msa(virusaln_seqinr_clean)
```

<!-- By Coglan -->
The filtered alignment is shorter, and is missing some of the poorly conserved regions of the original alignment.
<!-- By Coglan -->

<!-- By Coglan -->
Note that it is not a good idea to filter out too much of your alignment, as if you are left with few columns in your filtered alignment, you will be basing your phylogenetic tree upon a very short alignment (little data), and so the tree may be unreliable. Therefore, you need to achieve a balance between discarding the dodgy (poorly aligned) parts of your alignment, and retaining enough columns of the alignment that you will have enough data to based your tree upon.
<!-- By Coglan -->



