---
output: html_document
editor_options: 
  chunk_output_type: console
---
# Calculating genetic distances between sequences

**By**: Nathan Brouwer, with some content adapted Coghlan (2011) [Multiple Alignment and Phylogenetic trees](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter5.html) and under the Creative Commons 3.0 Attribution License [(CC BY 3.0)](https://creativecommons.org/licenses/by/3.0/).  Functions print_msa() and clean_alignment() adapted from ([Coglan 2011](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/#)).

## Preliminaries

```{r, message = F, warning = F}
library(compbio4all)
library(msa)
library(seqinr)
library(ape)
```


## Introduction


A common first step in performing a **phylogenetic analysis** is to calculate the **pairwise genetic distances** between sequences. The **genetic distance** is an estimate of the evolutionary **divergence** between two sequences, and is usually measured in quantity of evolutionary change, e.g., an estimate of the number of mutations that have occurred since the two sequences shared a **common ancestor**.

We can calculate the genetic distances between protein sequences using the `dist.alignment()` function in the `seqinr` package. The `dist.alignment()` function takes a multiple sequence alignment (MSA) as input. Based on the MSA that you give it, `dist.alignment()` calculates the genetic distance between each ***pair*** of proteins in the multiple alignment, yielding pairwise distances. For example, to calculate genetic distances between the virus phosphoproteins based on the multiple sequence alignment stored in virusaln, we type:

```{r}
data(virusaln)
virusdist <- seqinr::dist.alignment(virusaln_seqinr)  # Calculate the genetic distances

virusdist # Print out the genetic distance matrix
```

**NOTE** My result are different from the original results, shown here:
I need to check the settings used for the MSA                                      
                P0C569     O56773     P06747
  O56773      0.4142670
  P06747      0.4678196  0.4714045
  Q5VKP1      0.4828127  0.5067117  0.5034130
  
  
The genetic distance matrix above shows the genetic distance between each pair of proteins. The sequences are referred to by their **UniProt accessions**. Recall that 

* P06747 = rabies virus phosphoprotein
* P0C569 is Mokola virus phosphoprotein
* O56773 is Lagos bat virus phosphoprotein 
* Q5VKP1 is Western Caucasian bat virus phosphoprotein.

Based on the genetic **distance matrix** above, we can see that the genetic distance between Lagos bat virus phosphoprotein (O56773) and Mokola virus phosphoprotein (P0C569) is smallest (about 0.414).  Similarly, the genetic distance between Western Caucasian bat virus phosphoprotein (Q5VKP1) and Lagos bat virus phosphoprotein (O56773) is the biggest (about 0.507).

The larger the genetic distance between two sequences, the more amino acid changes (such as change from Asp to Met) or **indels** that have occurred since they shared a common ancestor, and the longer ago their **common ancestor** probably lived.  (The relationship between number of mutations and time, however, depends on the mutation rate and generation time of the organism).

## Calculating genetic distances between DNA/mRNA sequences

Just like for protein sequences, you can calculate genetic distances between DNA (or mRNA) sequences based on an alignment of the sequences. The RefSeq DNA accession numbers for the proteins we've been using are:

* AF049118 = mRNA sequence for Mokola virus phosphoprotein, 
* AF049114 = mRNA sequence for Mokola virus phosphoprotein, 
* AF049119 = mRNA sequence for Lagos bat virus phosphoprotein, 
* AF049115 = mRNA sequence for Duvenhage virus phosphoprotein.

<!-- mRNA or DNA? cDNA? -->

We can retrieve these DNA sequences using `entrez_fetch_list()`.  Some notes about how we'll use this function works:

1. `db` is short for "database"
1. the database is called `nuccore` (not `genebank` or `gene`)
1. the argument `rettype` is "tt"; I think it stands for "REtURN type" (I also forget the second "t")


```{r}
# put accessions in vector
accessions_mrna <- c("AF049118","AF049114","AF049119","AF049115")

# get sequences
virus_mrna_list <- entrez_fetch_list(db = "nuccore",  # "nuccore" db for DNA 
                                   id = accessions_mrna, 
                                   rettype = "FASTA") # rettype has two t
```

We can clean these three sequences using a simple `for()` loop.  We set `parse = F` so we get things back as single character string.

```{r}
for(i in 1:length(virus_mrna_list)){
  virus_mrna_list[[i]] <- fasta_cleaner(virus_mrna_list[i], 
                                       parse = F)
}
```

Now we need to convert each of these into a named vector of sequences.
<!-- make function to do this -->
```{r}
mra_seq_vector <- c(AF049118 =  virus_mrna_list[[1]] ,
                 AF049114 = virus_mrna_list[[2]] ,
                 AF049119 = virus_mrna_list[[3]] ,
                 AF049115 = virus_mrna_list[[4]])
```


Finally convert this to a "stringset" using `Biostrings::DNAStringSet()`.

```{r}
dna_seq_stringset <- Biostrings::DNAStringSet(mra_seq_vector)
```

Let's see what we got
```{r}
dna_seq_stringset
```

Now we can make an alignment use `msa()`.

```{r}
virus_mrna_aln <- msa(inputSeqs = dna_seq_stringset,
                     method = "ClustalW")
```

The output looks like this
```{r}
virus_mrna_aln
```

This looks a LOT different than an amino acid alignment, which looked like this:
```{r}
data(virusaln)
virusaln
```

Why might the be different?  First, examine the output above and determine how long the DNA alignment is versus the amino acid alignment?  Why are they different, and why is one longer than the other?

The DNA alignment is 1097 columns, while the amino acid alignment is only 306 rows.  Note that 306*3 = 918.  1097 is pretty close to 1097.  What's the relevance of multiplying by 3?


## Calculationg genetic distance

You can calculate a genetic distance for DNA or mRNA sequences using the `dist.dna()` function in the `ape` package. `dist.dna()` takes a MSA of DNA or mRNA sequences as its input, and calculates the genetic distance between each pair of DNA sequences in the multiple alignment.

The `dist.dna()` function requires the input alignment to be in a special format known as `DNAbin` format, so we must use the `as.DNAbin()` function to convert our DNA alignment into this format before using the `dist.dna()` function.

```{r}
 # Convert the alignment to "DNAbin" format
virus_mrna_aln_bin <- ape::as.DNAbin(virus_mrna_aln)
```

The output of `as.DNAbin()` gives us a short summery of the alignment
```{r}
virus_mrna_aln_bin
```


Now to make and view the alignment:
```{r}
 # Calculate the genetic distance matrix
virus_mrna_dist <- ape::dist.dna(virus_mrna_aln_bin)

 # Print out the genetic distance matrix
virus_mrna_dist                             
```

NOTE: my results for this alignment are the same as the original by Coghlan.  I'm not sure why my amino acid alignment produces divergent results but the DNA is the same.


