---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Rooted neighbor-joining phylogenetic trees

```{r eval = T, message = F, warning = F}
library(compbio4all)
library(rentrez)
library(msa)
library(ape)
library(seqinr)
```


**By**: Avril Coghlan.  [Multiple Alignment and Phylogenetic trees](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter5.html)



**Adapted, edited and expanded**: Nathan Brouwer under the Creative Commons 3.0 Attribution License [(CC BY 3.0)](https://creativecommons.org/licenses/by/3.0/).


## Building a rooted phylogenetic tree for protein sequences

In order to convert an **unrooted phylogenetic tree** into a **rooted tree**, we need to add or define an **outgroup sequence**. Normally, the outgroup sequence is a sequence that we know from some prior knowledge to be more **distantly related** to the other sequences under study than they are to each other.  For example, if we were interested in the relationship among the great apes (humans, chimps, bonobos, gorillas, and orangutans) we might use a lemur or a monkey as an outgroup.  

For example, the protein Fox-1 is involved in determining the sex of an embryo in the nematode worm *Caenorhabditis elegans* (UniProt accession Q10572). Related proteins are found in other nematodes, including *Caenorhabditis remanei* (NCBI accession: XP_003109757.1; UniProt: E3M2K8), *Caenorhabditis briggsae* (A8WS01), *Loa loa* (E1FUV2), and *Brugia malayi* (UniProt A8NSK3).

Note that *Caenorhabditis elegans* is a **model organism** commonly studied in molecular biology. The nematodes *Loa loa*, and *Brugia malayi* are parasitic nematodes that cause filariasis, which is classified by the WHO as a **neglected tropical disease**.

The UniProt database contains a distantly related sequence from the fruit fly *Drosophila melanogaster* (UniProt accession Q9VT99). If we were to build a phylogenetic tree of the nematode worm Fox-1 **homologs**, the distantly related sequence from fruitfly would probably be a good choice of outgroup, since the protein is from a different animal group (insects) than the nematode worms. Thus, it is safe to assume that the fruitfly protein is more distantly related to all the nematode proteins than they are to each other.

To retrieve the sequences from UniProt we can use the `entrez_fetch_list()`.

First, a vector of accession numbers:

<!-- # ?*Caenorhabditis briggsae* (A8WS01)? not used in code below? -->
<!-- # ?"Q8WS01", # "C. briggsae* Fox-1 homologue" used below, but not mentioned til later in tutorial -->


```{r eval = T}

worm.seqsnames <- c("Q10572",   # C. elegans
              "XP_003109757.1", #C. remanei; Uniprot = 
              "Q8WS01", # "C. briggsae* Fox-1 homologue"
              "XP_020305433.1", # Loa loa;                       download issues
              "CRZ25068.1", # Brugia malayi #                dowload issues
              "Q9VT99") # Drosophila melanogaster
```


<!-- <!-- Assignment: adapt this for shroom, previous virus assignment --> -->
<!-- Improve function - issue error if doens't download, re-try if doesn't downlaod -->

Now, download the sequences:
```{r eval = T}
worm_seqs_list <- compbio4all::entrez_fetch_list(db = "protein",
                                       id = worm.seqsnames,
                                       rettype = "fasta")
```


Clean up the sequences
```{r}
worm_seqs_list_clean <- worm_seqs_list
for(i in 1:length(worm_seqs_list_clean)){
  worm_seqs_list_clean[[i]] <- fasta_cleaner(worm_seqs_list_clean[i],parse = F)
}
```

Format into single vector

```{r}
# storage vector
worms_seqs_vector <- rep(NA, length(worm_seqs_list_clean))

# extract sequence
for(i in 1:length(worms_seqs_vector)){
  worms_seqs_vector[i] <- worm_seqs_list_clean[[i]]
}

# create named vector
names(worms_seqs_vector) <- worm.seqsnames
```

Turn into a stringset
```{r}
worms_seq_stringset <- Biostrings::AAStringSet(worms_seqs_vector)
```


<!-- todo: save this data in package -->



Run alignment
```{r}
fox1aln <- msa(inputSeqs = worms_seq_stringset,
                     method = "ClustalW")
```

Update class of alignment object
```{r}
class(fox1aln) <- "AAMultipleAlignment"
```

Convert to seqinr format 
```{r}
fox1aln_seqinr <- msaConvert(fox1aln, type = "seqinr::alignment")
```

Look at alignment
```{r}
print_msa(alignment = fox1aln_seqinr, 
          chunksize = 60)
```


## Build phylogenetic tree

The next step is to build a phylogenetic tree of the proteins, which again we can do using the neighbor-joining algorithm.

This time we have an outgroup in our set of sequences, so we can build a rooted tree. The function `rooted_NJ_tree()` can be used to build a rooted tree. It returns the phylogenetic tree, and also makes a picture of the tree:

The function takes the alignment and the name of the outgroup as its inputs. For example, to use it to make a phylogenetic tree of the *C. elegans* Fox-1 protein and its homologous, using the fruit fly protein (UniProt Q9VT99) as the outgroup, we type:

```{r eval = T, error = T, warning = F, message = F}
#debugonce(rooted_NJ_tree)
fox1alntree <- compbio4all::rooted_NJ_tree(alignment = fox1aln_seqinr, 
                                           og = "Q9VT99",
                                           type="protein")
```


Here we can see that E3M2K8 (*C. remanei* Fox-1 homologue) and Q10572 (*C. elegans* Fox-1) have been grouped together with a bootstrap value of 100%, and A8NSK3 (*Brugia malayi* Fox-1 homologue) and E1FUV2 (*Loa loa* Fox-1 homologue) have been grouped together with bootstrap = 100%. These four proteins have also been grouped together in a larger clade with bootstrap =  100%.

Compared to these four proteins, the proteins Q8WS01 (*C. briggsae* Fox-1 homologue) and Q9VT99 (fruit fly outgroup) seem to be relatively distantly related.

As this is a rooted tree, we know the direction that evolutionary time ran. Say we call the ancestor of the four sequences (E3M2K8, Q10572, A8NSK3, E1FUV2) ancestor1, the ancestor of the two sequences (E3M2K8, Q10572) ancestor2, and the ancestor of the two sequences (A8NSK3, E1FUV2) ancestor3.

Because it is a rooted tree, we know that time ran from left to right along the branches of the tree, so that ancestor1 was the ancestor of ancestor2, and ancestor1 was also the ancestor of ancestor3. In other words, ancestor1 lived before ancestor2 or ancestor3; ancestor2 and ancestor3 were descendants of ancestor1.

Another way of saying this is that E3M2K8 and Q10572 shared a common ancestor with each other more recently than they did with A8NSK3 and E1FUV2.

The lengths of branches in this tree are proportional to the amount of **evolutionary change** (estimated number of mutations) that occurred along the branches. The branches leading back from E3M2K8 and Q10572 to their last common ancestor are slightly longer than the branches leading back from A8NSK3 and E1FUV2 to their last common ancestor.

This indicates that there has been more evolutionary change in E3M2K8 (*C. remanei* Fox-1 homologue) and Q10572 (*C. elegans* Fox-1) proteins since they diverged, than there has been in A8NSK3 (*Brugia malayi* Fox-1 homologue) and E1FUV2 (*Loa loa* Fox-1 homologue) since they diverged.

## Building a phylogenetic tree for DNA or mRNA sequences

In the example above, a phylogenetic tree was built for protein sequences. The genomes of distantly related organisms such as vertebrates will have accumulated many mutations since they diverged. Sometimes, so many mutations have occurred since the organisms diverged that their DNA sequences are hard to align correctly and it is also hard to accurately estimate evolutionary distances from alignments of those DNA sequences.

In contrast, as many mutations at the DNA level are synonymous at the protein level, protein sequences diverge at a slower rate than DNA sequences. This is why for reasonably distantly related organisms such as vertebrates, it is usually preferable to use protein sequences for phylogenetic analyses.

If you are studying closely related organisms such as primates, few mutations will have occurred since they diverged. As a result, if you use protein sequences for a phylogenetic analysis, there may be too few amino acid substitutions to provide enough ‘signal’ to use for the phylogenetic analysis. Therefore, it is often preferable to use DNA sequences for a phylogenetic analysis of closely related organisms such as primates.

We can use the functions `unrooted_NJ_tree()` and `rooted_NJ_tree()` described above to build unrooted or rooted neighbour-joining phylogenetic trees based on an alignment of DNA or mRNA sequences. In this case, we need to use `type=DNA` as an argument in these functions, to tell them that we are making a tree of nucleotide sequences, not protein sequences.

For example, to build an unrooted phylogenetic tree based on the alignment of the virus phosphoprotein mRNA sequences, we type in R:

<!-- STOPPED HERE -->

```{r eval = T}
virusmRNAaln  <- read.alignment(file = "virusmRNA.phy", format = "phylip")
```

```{r eval = T}
virusmRNAalntree <- unrootedNJtree(virusmRNAaln, type="DNA")
```


## Saving a phylogenetic tree as a Newick-format tree file

A commonly used format for representing phylogenetic trees is the **Newick format.** Once you have built a phylogenetic tree using *R*, it is convenient to store it as a Newick-format tree file. This can be done using the `write.tree()` function in the `ape`package.

For example, to save the unrooted phylogenetic tree of virus phosphoprotein mRNA sequences as a Newick-format tree file called `virusmRNA.tre`, we type:

```{r eval = T}
write.tree(virusmRNAalntree, 
           "virusmRNA.tre")
```

The Newick-format file `virusmRNA.tre` should now appear in your “My Documents” folder.

## Summary

In this practical, you have learned the following *R* functionss:


<!-- Check these - these list written before chapters split up -->

* `read.alignment()` from the `seqinr` package for reading in a multiple alignment
* `dist.alignment()` from the `seqinr` package for calculating genetic distances between protein sequences
* `dist.dna()` from the `ape` package for calculating genetic distances between DNA or mRNA sequences

## Links and Further Reading
Some links are included here for 
further reading.

For background reading on phylogenetic trees, it is recommended to read Chapter 7 of Introduction to Computational Genomics: a case studies approach by Cristianini and Hahn (Cambridge University Press; www.computational-genomics.net/book/).

Another more in-depth (but very accesssible) book on phylogenetics is Molecular Evolution: A Phylogenetic Approach by Roderic DM Page and Edward C Holmes.

For more in-depth information and more examples on using the SeqinR package for sequence analysis, look at the SeqinR documentation, http://pbil.univ-lyon1.fr/software/seqinr/doc.php?lang=eng.

For more in-depth information and more examples on the Ape package for phylogenetic analysis, look at the Ape documentation, ape.mpl.ird.fr/.

If you are using the Ape package for a phylogenetic analysis project, it would be worthwhile to obtain a copy of the book Analysis of Phylogenetics and Evolution with R by Emmanuel Paradis, published by Springer, which has many nice examples of using R for phylogenetic analyses.

For a more in-depth introduction to R, a good online tutorial is available on the “Kickstarting R” website, cran.r-project.org/doc/contrib/Lemon-kickstart.

There is another nice (slightly more in-depth) tutorial to R available on the “Introduction to R” website, cran.r-project.org/doc/manuals/R-intro.html.

##  Exercises
Answer the following questions, using the R package. For each question, please record your answer, and what you typed into R to get this answer.

Model answers to the exercises are given in Answers to the exercises on Multiple Alignment and Phylogenetic Trees.

1. Calculate the genetic distances between the  NS1 proteins from different Dengue virus strains shown below.   a) Which are the most closely related proteins, and b) which are the *least* closely related, based on the genetic distances?  Note: Dengue virus causes Dengue fever, which is classified by the WHO as a neglected tropical disease. There are four main types of Dengue virus, Dengue virus 1, Dengue virus 2, Dengue virus 3, and Dengue virus 4.
  + Dengue virus 1 NS1 protein (Uniprot Q9YRR4)
  + Dengue virus 2 NS1 protein (UniProt Q9YP96), 
  + Dengue virus 3 NS1 protein (UniProt B0LSS3), and 
  + Dengue virus 4 NS1 protein (UniProt Q6TFL5). 
1. Build an unrooted phylogenetic tree of the NS1 proteins from the Dengue virus 1, Dengue virus 2, Dengue virus 3 and Dengue virus 4, using the neighbour-joining algorithm. 
1. Build an unrooted phylogenetic tree of the NS1 proteins from Dengue viruses 1-4, based on a **filtered alignment** of the four proteins (keeping alignment columns in which at least 30% of letters are not gaps, and in which at least 30% of pairs of letters are identical). Does this differ from the tree based on the unfiltered alignment (in Q2)? Can you explain why?
1. The Zika virus is related to Dengue viruses, but is not a Dengue virus, and so therefore can be used as an **outgroup** in phylogenetic trees of Dengue virus sequences. UniProt accession Q32ZE1 consists of a sequence with similarity to the Dengue NS1 protein, so seems to be a related protein from Zika virus. Build a rooted phylogenetic tree of the Dengue NS1 proteins based on a filtered alignment (keeping alignment columns in which at least 30% of letters are not gaps, and in which at least 30% of pairs of letters are identical), using the Zika virus protein as the outgroup. 
  + Which are the most closely related Dengue virus proteins, based on the tree?
  + What extra information does this tree tell you, compared to the unrooted tree in Q2?

```{r}

```
  
  
