--- 
title: "A Little Book of R for Bioinformatics 2.0"
author: "Avirl Coghlan, with contributions by Nathan L. Brouwer"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes
description: "An entry-level text for bioinformatics and computational biology"
---

# Preface to version 2.0 {-}

Placeholder



<!--chapter:end:index.Rmd-->


# How to review this book

Placeholder


## File format
## What can you do?
## Adding comments
### HTML-tagged comments 
### Code chunk comments
### Keys to exercise
### Problem - fix this code

<!--chapter:end:00a-how_to_review_this_book.Rmd-->

# Test audio link

Here is an example of an audio link  embedded in a bookdown book.


```{r}
knitr::include_url("https://www.genome.gov/sites/default/files/tg/en/narration/single_nucleotide_polymorphisms_snps.mp3",
                   height = "50px")
```


<!--chapter:end:00b_test_audio_link.Rmd-->

# Test shiny app

Here is an example of a Shiny app embedded in a bookdown book.


```{r}
knitr::include_app(url = "https://wrightaprilm.shinyapps.io/treesiftr_app/")
```


<!--chapter:end:00b_test_shiny_app.Rmd-->

# Test YouTube vidoe

Here is an example of a YouTube video  embedded in a bookdown book.



via knitr chunk
```{r}
knitr::include_url("https://www.youtube.com/watch?v=zNzZ1PfUDNk&feature=youtu.be")
```

<!--chapter:end:00b_test_youtube.Rmd-->

# Test webpage insertion

```{r}
knitr::include_url(url ="https://docs.google.com/spreadsheets/d/18dt7cYHAaszV5Iq8Y-7YOEFnv6hIXbD39Vn75-c3fQo/edit?usp=sharing")
```


<!--chapter:end:00c_test_webpage.Rmd-->

---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Datacamp test

<script src=https://cdn.datacamp.com/datacamp-light-latest.min.js></script>

```{r, include=FALSE}
tutorial::go_interactive()
```

By default, `tutorial` will convert all R chunks.

```{r}
a <- 2
b <- 3

a + b
```

<!--chapter:end:00d_test_datacamp_tutorial.Rmd-->


# Downloading R

Placeholder


## Preface
## Introduction to R
## Installing R
### Installing *R* on a Windows PC
### How to install *R* on non-Windows computers (eg. Macintosh or Linux computers)
## Starting *R* 

<!--chapter:end:01-downloading_R_AC01-1.Rmd-->

# Installing the RStudio IDE

**By:** Nathan Brouwer

<!-- TODO: UPDATE with comp bio 4 all -->

The name "R" refers both to the programming language and the program that runs that language.  When you download *it*R* there is also a basic **GUI** (graphical user interface) that you can access via the *R* icon.

Other GUIs are available, and the most popular currently is **RStudio.**  RStudio a for-profit company that is a main driver of development of R.  Much of what they produce has free basic versions or is entirely free.  They produce software (RStudio), cloud-based applications (**RStudio Cloud**), and web server infrastructure for business applications of R. 

A brief overview of installing RStudio can be found here ["Getting RStudio on to your computer"](https://brouwern.github.io/BOOK_R_Ecological_Data_Science/getting-rstudio-on-to-your-computer.html)

## Getting to know RStudio

For a brief overview of RStudio see ["Getting started with RStudio"](https://brouwern.github.io/BOOK_R_Ecological_Data_Science/getting-started-with-rstudio.html)

A good overview of what the different parts of RStudio can be seen in the image in this tweet: https://twitter.com/RLadiesNCL/status/1138812826917724160?s=20


## RStudio versus RStudio Cloud

RStudio and RStudio cloud work almost identically, so anything you read about RStudio will apply to RStudio Cloud.  RStudio is easy to download an use, but RStudio Cloud eliminates even the minor hiccups that occur.  Free accounts with RStudio Cloud allow up to 15 hours per month, which is enough for you to get a taste for using R. 

<!--chapter:end:02-installing_RStudio_AC01-2.Rmd-->


# Installing *R* packages

Placeholder


## Downloading packages with the RStudio IDE
## Downloading packages with the function `install.packages()`
## Using packages after they are downloaded

<!--chapter:end:03-installing_packages_in_RStudio_AC01-3.Rmd-->


# Installing Bioconductor

Placeholder


## Bioconductor
## Installing BiocManager
## The ins and outs of package installation
### Updating other packages when downloading a package
### Packages "from source"
### More on angry red text
## Actually loading a package

<!--chapter:end:04-installing_bioconductor_AC01-4.Rmd-->


# A Brief introduction to R

Placeholder


## Vocabulary
## R functions 
## Interacting with R
## Variables in R
### Vectors
### Vector indexing
### Character vectors
### Lists 
### Tables 
#### Tables - The basics
#### Tables - further details
## Arguments 
## Help files with `help()` and `?`
## Searching for functions with `help.search()` and `RSiteSearch()`
## More on functions 
### Writing your own functions
## Quiting R
## Links and Further Reading

<!--chapter:end:05-brief_intro_to_R_AC01-5.Rmd-->


# DNA descriptive statics - Part 1

Placeholder


## Preface
## Writing TODO:
## Introduction
## Vocabulary
## Functions
## Preliminaries
## Converting DNA from FASTA format
## Length of a DNA sequence
### Base composition of a DNA sequence
### GC Content of DNA
### DNA words
### Summary
## Acknowledgements
### License
### Exercises

<!--chapter:end:06-DNA_sequence_descriptive_stats-AC03-02-.Rmd-->


# Programming in R: for loops

Placeholder


## Preface
## Vocab
## Functions
## Basic for loops in in R
## Challenge: complicated vectors of values

<!--chapter:end:07-programming_in_R-forloops_AC04-02.Rmd-->


# Mini tutorial: Vectors in R

Placeholder


## Preface 
## Vocab 
## Vectors in R
## Math on vectors
## Functions on vectors
## Operations with two vectors
## Subsetting vectors
## Sequences of numbers
## Vectors can hold numeric or character data
## Regular expressions can modify character data

<!--chapter:end:08-vectors-AC04-01.Rmd-->


# Plotting vectors in base R

Placeholder


## Preface 
## Plotting numeric data
## Other plotting packages

<!--chapter:end:09-vectors-plotting-AC04-01.Rmd-->


# Programming in R: functions

Placeholder


## Preface
## Vocab
## Functions
## Functions in R
## Comments in R

<!--chapter:end:10-programming_in_R-functions-AC04-02.Rmd-->


# FASTA Files

Placeholder


## Example FASTA file
## Multiple sequences in a single FASTA file
## Multiple sequence alignments can be stored in FASTA format
## FASTQ Format

<!--chapter:end:11a-FASTA_files.Rmd-->


# Downloading DNA sequences as FASTA files in R

Placeholder


### Functions
### Software/websites
### R vocabulary
### File types
### Bioinformatics vocabulary
### Organisms and Sequence accessions
### Preliminaries
## DNA Sequence Statistics: Part 1
### Using R for Bioinformatics
### R packages for bioinformatics: Bioconductor and SeqinR
### FASTA file format
### The NCBI sequence database
### Retrieving genome sequence data using rentrez
## OPTIONAL: Saving FASTA files
## Next steps

<!--chapter:end:11b-download_FASTA_in_R.Rmd-->


# Downloading DNA sequences as FASTA files in R

Placeholder


## Preliminaries
## Convert FASTA sequence to an R variable
### Removing unwanted characters
### Spliting unbroken strings in character vectors

<!--chapter:end:12-clean_FASTA_in_R-AC03-01.Rmd-->


# Downloading protein sequences in *R*

Placeholder


## Preliminaries
## Retrieving a UniProt protein sequence using rentrez

<!--chapter:end:14-download_protein_sequences.Rmd-->


# Sequence dotplots in *R*

Placeholder


## Preliminaries
### Download sequences
## Visualzing two identical sequences
## Visualizing repeats
## Inversions
## Translocations
## Random sequence
## Comparing two real sequences using a dotplot

<!--chapter:end:15-dotplots.Rmd-->


# Global proteins aligments in *R*

Placeholder


## Preliminaries
### Download sequences
## Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm
## Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm
### Protein score matrices
### Example protein alignment
## Aligning UniProt sequences
## Viewing a long pairwise alignment

<!--chapter:end:16_alignment_global_protein.Rmd-->


# Local protein aligments in *R*

Placeholder


## Preliminaries
### Download sequences
## Pairwise local alignment of protein sequences using the Smith-Waterman algorithm

<!--chapter:end:17-alignment_local-protein_in_R-AC07-02.Rmd-->


# Matrix vocabulary

Placeholder


## Introduction 
## Preliminaries
### Packages 
### Vocab
#### Math / R vocab
#### Bioinformatics vocab 
### Functions used
#### Base R functions
#### Specific function
## Load packages
## Matrix structure
## Diagonal of a matrix

<!--chapter:end:18_matrix_vocab.Rmd-->


# Alignment in R

Placeholder


## Preliminaries
### Packages
### Data

<!--chapter:end:18_pariwise_alignment_in_R_active_learning.Rmd-->


# Manipulating matrices and vectors: worked example

Placeholder


## Introduction 
## Preliminaries
### Packages 
### Vocab
#### Math / R vocab
#### Bioinformatics vocab 
### Functions used
#### Base R functions
### Packages
### Data
## Matrix elements can be accessed with square bracket notation
## R object names
### Accessing items in names vectors
### Accessing items in names matrices

<!--chapter:end:18_working_with_matrices_and_vectors_worked_example.Rmd-->


# Alignment by eye in R

Placeholder


## Preliminaries
### Packages
### Data

<!--chapter:end:18-alignment_by_hand_shrm_active_learning.Rmd-->


# Retrieving multiple sequences in R

Placeholder


## Prelminaries
## Retrieving a set of sequences from UniProt
## Downloading sequences in bulk

<!--chapter:end:18-retrieving_mult_seqs-AC08-01.Rmd-->


# Multiple sequence alignment in R

Placeholder


## Preliminaries 
### Packages 
### Functions
## Multiple sequence alignment (MSA)
## Make MSA with msa()
### Data preparation 
## Viewing your MSA
### Viewing a long multiple alignment in the R console.
### Visualizing alignments as an R plot
#### OPTIONAL: File types used by `ggmsa`
### OPTIONAL: Print MSA to PDF
## Discarding very poorly conserved regions from an alignment

<!--chapter:end:19-MSA_in_R_.Rmd-->


# The BLOSUM scoring matrix in R

Placeholder


## Introduction 
## Preliminaries
### Packages 
### Vocab
#### Math / R vocab
#### Bioinformatics vocab 
### Functions used
#### Base R functions
#### Specific function
## Load packages
## The BLOSUM matrix
## Visualizing the BLOSUM matrix

<!--chapter:end:20_BLOSUM_matrix.Rmd-->


# Calculating genetic distances between sequences

Placeholder


## Preliminaries
## Introduction
## Calculating genetic distances between DNA/mRNA sequences
## Calculationg genetic distance

<!--chapter:end:20a-genetic_distance.Rmd-->


# Unrooted neighbor-joining phylogenetic trees

Placeholder


## Preliminaries
### Key functions 
### Key vocab
## Building an unrooted phylogenetic tree for protein sequences
### Build tree with `unrooted_NJ_tree()`
## Boostrap values indicate support for clades
## Branch lengths indicate divergence between sequences
## Unrooted trees lack an outgroup

<!--chapter:end:21-unrooted_tree_NJ_tree.Rmd-->


# A complete bioinformatics workflow in R

Placeholder


## Introduction
## Software Preliminaires
### Vocab 
### R functions
### Download necessary packages
### Load packages into memory
## Downloading macro-molecular sequences
## Prepping macromolecular sequences
## Aligning sequences
## The shroom family of genes
## Downloading multiple sequences 
## Multiple sequence alignment
### Building an Multiple Sequence Alignment (MSA)
### Viewing an MSA
#### Viewing an MSA in R
#### Displaying an MSA as an R plot
#### Saving an MSA as PDF
## Genetic distance.
## Phylognetic trees (finally!)
### Plotting phylogenetic trees 

<!--chapter:end:23-alignment-walkthrough-shroom.Rmd-->


# Logarithms in R

Placeholder



<!--chapter:end:88_logarithms.Rmd-->


# (PART) Appendices {-}
# Appendix 01: Getting access to R {-}

Placeholder


### {-}
## Getting Started With R and RStudio
### RStudio Cloud
### Getting R onto your own computer
### Getting RStudio onto your computer
### Keep R and RStudio current

<!--chapter:end:99-A01_getting_access_to_R.Rmd-->


# Getting started with R itself (or not) {-}

Placeholder


## Vocabulary {-}
## R commands {-}
#### R's console as a scientific calculator
#### Practice: math in the console
#### The script editor
#### Comments 
## Help!
### Getting "help" from R
### Getting help from the internet
### Getting help from online forums
### Getting help from twitter
## Other features of RStudio
### Ajusting pane the layout
### Adjusting size of windows
## Practice (OPTIONAL)

<!--chapter:end:99-A02_getting_started_with_R.Rmd-->


# BLAST Alignment score distribution

Placeholder


## Packages
## Distributions of alignment scores
## Simulating data
## Lots of data in biology is normal
## Simulating the normal distribution
## The extreme value distribution
## Replicating Figure 4.14
## Why do we care?

<!--chapter:end:BLAST-alignment_score_distribution.Rmd-->


# Calculating E values, Bit scores, and P-values from BLAST output

Placeholder


## Introduction
## Case Study
## BLAST Statistics
## Dotplot
### Dotplot in R - optional, experimental code
## BLAST "Search Summary" 
## BLAST Alignments
## Equations
### E value (eq 4.5, pg 142)
## Bit Scores (S')
## E from Bit scores
## P values

<!--chapter:end:dayoff-E_values.Rmd-->


# Calculating the number of possible phylogenetic trees

Placeholder


## Repliminaries
## Concepts
#### Biology Concepts / Vocab
## Functions
## Introduction
## Number of rooted trees
## Number of rooted trees in R
### Unique topologies versus rotations
## Bilography

<!--chapter:end:tree-think_number_of_trees-theory.Rmd-->



## UPGMA the hard way
## Introudction to UPGMA
## Preliminaries
### R Libraries
#### Phylogenetic libraries
#### Plotting libraries
### Custom function we'll use
## Data matrix
### Building the matrix
## What does this distance matrix represent?
### Genetic distances conceptualized as pairwise linear distances
### Genetic distances conceptualized as 2D map of linear distances
## From linear distances to "tree distances"
## The UPGMA algorithm
### Step 1: find the 2 taxa that are currently closest together
### Step 2: calculate the branch lengths connect the taxa to their shared node
### Step 3a: determine if the algorithm is done
### Step 3a: Continue on if needed and combine the previous cluster into cluster "u"
### Step 4: Determine the distance from u to the other remaining clusters.
### Step 5: Create a new matrix 
### Key step: calculating the distance from a clade to the remaining taxa
## Calculate 2D representation
## Algorithm - round 1
### Determine minimum value in current matrix
### Combine species into clade
### Calculate branch length
### Calculate distance from clade to all other points
#### ab to c
#### ab to d
#### ab to e
### Update matrix with new distance
#### Illustration
#### Estimate 2 D
## Algorithm - next round
### Combine species into clade
### Calculate branch length
### Recalculate distance from abe to all remainign taxa
#### abe to c
#### abe to c
### Update the matrix
## Next iteration
#### Estimate 2 D
### Calculate branch lengths
### Combine species into clade
### Calculate branch length
### Distance of c to abe and d to abe
## Finish up
### Combine species into clade
### Calculate branch length
## Finalizing branch lengths

<!--chapter:end:tree-think_UPGMA_wikidat_compare_to_ordination_calcs_by_hand-UPDATED.Rmd-->


# Representing phylogenetic trees in Newick format

Placeholder


## Vocab
## Introduction
## Trees with three taxa
## Trees with four taxa

<!--chapter:end:tree-think-03-newick_format.Rmd-->


# Plotting trees from Newick notation in R

Placeholder


## Preliminaries 
## Note
## 3 taxa tree in Newick
## 4 taxa tree in Newick
## 5 taxa in Newick Format
## Many-taxa tree
## Newick and Branch lengths
## Additional Reading

<!--chapter:end:tree-think-04-newick_in_R.Rmd-->


# Phylogenetic trees using the UPGMA clustering algoirthm

Placeholder


## Introduction
## Preliminaries
### Vocab
### General R functions
### Specific functions
## Sequence comparisons
## Distance matrices
## From a distance matrix to phylogenetic tree
## Distance matrix in R from raw sequences
## Distance matrix in R from normal matrix
## UPGMA in R
## How UPGMA clustering works
### Challenge
## Further information
### General info
### Worked examples: 
### Additional notes: 

<!--chapter:end:tree-think-05-UPGMA_in_R.Rmd-->


# Tutorial: Writing functions to calculate the number of phylogenetic trees

Placeholder


## Repliminaries
### Functions
## Concepts
#### Biology Concepts:
#### Programming concepts
### Vocab:
#### Programming vocab
### Packages:
## Introduction
## Number of rooted trees
## Number of rooted trees in R
## Functions in R
## Function to calculate the number of possible phylogenetic trees
## Vectorized inputs to functions
## Adding conditional statements
## Adding multiple conditional statements
## Adding additional arguments
## Assignment: _un_rooted trees
## Assignment part 1
### Part 1 answer
## Assignment part 2
### Answer

<!--chapter:end:tree-think-number_of_phylogenetic_trees.Rmd-->



## Libraries
## Function we'll use
## Data matrix
## What does this matrix represent?
### Genetic distances conceptualized as pairwise linear distances
### Genetic distances conceptualized as 2D linear distances
## From linear distances to "tree distances"
## The UPGMA algorithm
## Calculate 2D representation
## Algorithm - round 1
### Determine minimum value in current matrix
### Combine species into clade
### Calcualte distance from clade to all other points
#### ab to c
#### ab to d
#### ab to e
### Update matrix with new distance
#### Illustration
#### Estimate 2 D
## Algorithm - round 2
#### abe to c
#### abe to c
### Update the matrix
#### Estimate 2 D
## Iteration 3
### Distance of c to abe and d to abe

<!--chapter:end:book07-tree_thinking/x_UPGMA_wikipedia_compare_to_ordination_calcs_by_hand-deprecated.Rmd-->



## Data matrix
## Data frame to store stuff
## Algorithm - First step
### Determine minimum value in current matrix
### Combine species into clade
### First branch length estimation
### First distance matrix update - WPGMA version
#### ab to c
#### ab to d
#### ab to e
### Update matrix with new distance
#### Illustration
## Algorithm - Second step
### Combine species into clade
### Calculate branch length
### Recalculate distance from abe to all remainign taxa
#### abe to c
#### abe to c
### Update the matrix
## Third step
### Calculate branch lengths
### Combine species into clade
### Calculate branch length
### Distance of c to abe and d to abe
## Finish up
### Combine species into clade
### Calculate branch length
## Plotting

<!--chapter:end:book07-tree_thinking/xQ_UPGMA-WPGMA_wikipedia-by_hand-deprecatedQ.Rmd-->

