--- 
title: "A Little Book of R for Bioinformatics 2.0"
author: "Avirl Coghlan, with contributions by Nathan L. Brouwer"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes
description: "An entry-level text for bioinformatics and computational biology"
---

# Preface to version 2.0 {-}

Placeholder



<!--chapter:end:index.Rmd-->


# How to review this book

Placeholder


## File format
## What can you do?
## Adding comments
### HTML-tagged comments 
### Code chunk comments
### Keys to exercise
### Problem - fix this code

<!--chapter:end:00a-how_to_review_this_book.Rmd-->

# Test shiny app

Here is an example of a Shiny app embedded in a bookdown book.


```{r}
knitr::include_app(url = "https://wrightaprilm.shinyapps.io/treesiftr_app/")
```


<!--chapter:end:00b_test_shiny_app.Rmd-->

# Test webpage insertion

```{r}
knitr::include_url(url ="https://docs.google.com/spreadsheets/d/18dt7cYHAaszV5Iq8Y-7YOEFnv6hIXbD39Vn75-c3fQo/edit?usp=sharing")
```


<!--chapter:end:00c_test_webpage.Rmd-->

---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Datacamp test

<script src=https://cdn.datacamp.com/datacamp-light-latest.min.js></script>

```{r, include=FALSE}
tutorial::go_interactive()
```

By default, `tutorial` will convert all R chunks.

```{r}
a <- 2
b <- 3

a + b
```

<!--chapter:end:00d_test_datacamp_tutorial.Rmd-->


# Downloading R

Placeholder


## Preface
## Introduction to R
## Installing R
### Installing *R* on a Windows PC
### How to install *R* on non-Windows computers (eg. Macintosh or Linux computers)
## Starting *R* 

<!--chapter:end:01-downloading_R_AC01-1.Rmd-->

# Installing the RStudio IDE

**By:** Nathan Brouwer

<!-- TODO: UPDATE with comp bio 4 all -->

The name "R" refers both to the programming language and the program that runs that language.  When you download *it*R* there is also a basic **GUI** (graphical user interface) that you can access via the *R* icon.

Other GUIs are available, and the most popular currently is **RStudio.**  RStudio a for-profit company that is a main driver of development of R.  Much of what they produce has free basic versions or is entirely free.  They produce software (RStudio), cloud-based applications (**RStudio Cloud**), and web server infrastructure for business applications of R. 

A brief overview of installing RStudio can be found here ["Getting RStudio on to your computer"](https://brouwern.github.io/BOOK_R_Ecological_Data_Science/getting-rstudio-on-to-your-computer.html)

## Getting to know RStudio

For a brief overview of RStudio see ["Getting started with RStudio"](https://brouwern.github.io/BOOK_R_Ecological_Data_Science/getting-started-with-rstudio.html)

A good overview of what the different parts of RStudio can be seen in the image in this tweet: https://twitter.com/RLadiesNCL/status/1138812826917724160?s=20


## RStudio versus RStudio Cloud

RStudio and RStudio cloud work almost identically, so anything you read about RStudio will apply to RStudio Cloud.  RStudio is easy to download an use, but RStudio Cloud eliminates even the minor hiccups that occur.  Free accounts with RStudio Cloud allow up to 15 hours per month, which is enough for you to get a taste for using R. 

<!--chapter:end:02-installing_RStudio_AC01-2.Rmd-->


# Installing *R* packages

Placeholder


## Downloading packages with the RStudio IDE
## Downloading packages with the function `install.packages()`
## Using packages after they are downloaded

<!--chapter:end:03-installing_packages_in_RStudio_AC01-3.Rmd-->


# Installing Bioconductor

Placeholder


## Bioconductor
## Installing BiocManager
## The ins and outs of package installation
### Updating other packages when downloading a package
### Packages "from source"
### More on angry red text
## Actually loading a package

<!--chapter:end:04-installing_bioconductor_AC01-4.Rmd-->


# A Brief introduction to R

Placeholder


## Vocabulary
## R functions 
## Interacting with R
## Variables in R
### Vectors
### Vector indexing
### Character vectors
### Lists 
### Tables 
#### Tables - The basics
#### Tables - further details
## Arguments 
## Help files with `help()` and `?`
## Searching for functions with `help.search()` and `RSiteSearch()`
## More on functions 
### Writing your own functions
## Quiting R
## Links and Further Reading

<!--chapter:end:05-brief_intro_to_R_AC01-5.Rmd-->


# DNA descriptive statics - Part 1

Placeholder


## Preface
## Writing TODO:
## Introduction
## Vocabulary
## Functions
## Preliminaries
## Converting DNA from FASTA format
## Length of a DNA sequence
### Base composition of a DNA sequence
### GC Content of DNA
### DNA words
### Summary
## Acknowledgements
### License
### Exercises

<!--chapter:end:06-DNA_sequence_descriptive_stats-AC03-02-.Rmd-->


# Programming in R: for loops

Placeholder


## Preface
## Vocab
## Functions
## Basic for loops in in R
## Challenge: complicated vectors of values

<!--chapter:end:07-programming_in_R-forloops_AC04-02.Rmd-->


# Mini tutorial: Vectors in R

Placeholder


## Preface 
## Vocab 
## Vectors in R
## Math on vectors
## Functions on vectors
## Operations with two vectors
## Subsetting vectors
## Sequences of numbers
## Vectors can hold numeric or character data
## Regular expressions can modify character data

<!--chapter:end:08-vectors-AC04-01.Rmd-->


# Plotting vectors in base R

Placeholder


## Preface 
## Plotting numeric data
## Other plotting packages

<!--chapter:end:09-vectors-plotting-AC04-01.Rmd-->


# Programming in R: functions

Placeholder


## Preface
## Vocab
## Functions
## Functions in R
## Comments in R

<!--chapter:end:10-programming_in_R-functions-AC04-02.Rmd-->


# FASTA Files

Placeholder


## Example FASTA file
## Multiple sequences in a single FASTA file
## Multiple sequence alignments can be stored in FASTA format
## FASTQ Format

<!--chapter:end:11a-FASTA_files.Rmd-->


# Downloading DNA sequences as FASTA files in R

Placeholder


### Functions
### Software/websites
### R vocabulary
### File types
### Bioinformatics vocabulary
### Organisms and Sequence accessions
### Preliminaries
## DNA Sequence Statistics: Part 1
### Using R for Bioinformatics
### R packages for bioinformatics: Bioconductor and SeqinR
### FASTA file format
### The NCBI sequence database
### Retrieving genome sequence data using rentrez
## OPTIONAL: Saving FASTA files
## Next steps

<!--chapter:end:11b-download_FASTA_in_R.Rmd-->


# Downloading DNA sequences as FASTA files in R

Placeholder


## Preliminaries
## Convert FASTA sequence to an R variable
### Removing unwanted characters
### Spliting unbroken strings in character vectors

<!--chapter:end:12-clean_FASTA_in_R-AC03-01.Rmd-->


# Downloading protein sequences in *R*

Placeholder


## Preliminaries
## Retrieving a UniProt protein sequence using rentrez

<!--chapter:end:14-download_protein_sequences.Rmd-->


# Sequence dotplots in *R*

Placeholder


## Preliminaries
### Download sequences
## Visualzing two identical sequences
## Visualizing repeats
## Inversions
## Translocations
## Random sequence
## Comparing two real sequences using a dotplot

<!--chapter:end:15-dotplots.Rmd-->


# Global proteins aligments in *R*

Placeholder


## Preliminaries
### Download sequences
## Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm
## Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm
### Protein score matrices
### Example protein alignment
## Aligning UniProt sequences
## Viewing a long pairwise alignment

<!--chapter:end:16_alignment_global_protein.Rmd-->


# Local protein aligments in *R*

Placeholder


## Preliminaries
### Download sequences
## Pairwise local alignment of protein sequences using the Smith-Waterman algorithm

<!--chapter:end:17-alignment_local-protein_in_R-AC07-02.Rmd-->


# Retrieving multiple sequences in R

Placeholder


## Prelminaries
## Retrieving a set of sequences from UniProt
## Downloading sequences in bulk

<!--chapter:end:18-retrieving_mult_seqs-AC08-01.Rmd-->


# Multiple sequence alignment in R

Placeholder


## Preliminaries 
### Packages 
### Functions
## Multiple sequence alignment (MSA)
## Make MSA with msa()
### Data preparation 
## Viewing your MSA
### Viewing a long multiple alignment in the R console.
### Visualizing alignments as an R plot
#### OPTIONAL: File types used by `ggmsa`
### OPTIONAL: Print MSA to PDF
## Discarding very poorly conserved regions from an alignment

<!--chapter:end:19-MSA_in_R_.Rmd-->

---
output: html_document
editor_options: 
  chunk_output_type: console
---
# Calculating genetic distances between sequences

**By**: Nathan Brouwer, with some content adapted Coghlan (2011) [Multiple Alignment and Phylogenetic trees](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter5.html) and under the Creative Commons 3.0 Attribution License [(CC BY 3.0)](https://creativecommons.org/licenses/by/3.0/).  Functions print_msa() and clean_alignment() adapted from ([Coglan 2011](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/#)).

## Preliminaries

```{r, message = F, warning = F}
library(compbio4all)
library(msa)
library(seqinr)
library(ape)
```


## Introduction


A common first step in performing a **phylogenetic analysis** is to calculate the **pairwise genetic distances** between sequences. The **genetic distance** is an estimate of the evolutionary **divergence** between two sequences, and is usually measured in quantity of evolutionary change, e.g., an estimate of the number of mutations that have occurred since the two sequences shared a **common ancestor**.

We can calculate the genetic distances between protein sequences using the `dist.alignment()` function in the `seqinr` package. The `dist.alignment()` function takes a multiple sequence alignment (MSA) as input. Based on the MSA that you give it, `dist.alignment()` calculates the genetic distance between each ***pair*** of proteins in the multiple alignment, yielding pairwise distances. For example, to calculate genetic distances between the virus phosphoproteins based on the multiple sequence alignment stored in virusaln, we type:

```{r}
data(virusaln)
virusdist <- seqinr::dist.alignment(virusaln_seqinr)  # Calculate the genetic distances

virusdist # Print out the genetic distance matrix
```

**NOTE** My result are different from the original results, shown here:
I need to check the settings used for the MSA                                      
                P0C569     O56773     P06747
  O56773      0.4142670
  P06747      0.4678196  0.4714045
  Q5VKP1      0.4828127  0.5067117  0.5034130
  
  
The genetic distance matrix above shows the genetic distance between each pair of proteins. The sequences are referred to by their **UniProt accessions**. Recall that 

* P06747 = rabies virus phosphoprotein
* P0C569 is Mokola virus phosphoprotein
* O56773 is Lagos bat virus phosphoprotein 
* Q5VKP1 is Western Caucasian bat virus phosphoprotein.

Based on the genetic **distance matrix** above, we can see that the genetic distance between Lagos bat virus phosphoprotein (O56773) and Mokola virus phosphoprotein (P0C569) is smallest (about 0.414).  Similarly, the genetic distance between Western Caucasian bat virus phosphoprotein (Q5VKP1) and Lagos bat virus phosphoprotein (O56773) is the biggest (about 0.507).

The larger the genetic distance between two sequences, the more amino acid changes (such as change from Asp to Met) or **indels** that have occurred since they shared a common ancestor, and the longer ago their **common ancestor** probably lived.  (The relationship between number of mutations and time, however, depends on the mutation rate and generation time of the organism).

## Calculating genetic distances between DNA/mRNA sequences

Just like for protein sequences, you can calculate genetic distances between DNA (or mRNA) sequences based on an alignment of the sequences. The RefSeq DNA accession numbers for the proteins we've been using are:

* AF049118 = mRNA sequence for Mokola virus phosphoprotein, 
* AF049114 = mRNA sequence for Mokola virus phosphoprotein, 
* AF049119 = mRNA sequence for Lagos bat virus phosphoprotein, 
* AF049115 = mRNA sequence for Duvenhage virus phosphoprotein.

<!-- mRNA or DNA? cDNA? -->

We can retrieve these DNA sequences using `entrez_fetch_list()`.  Some notes about how we'll use this function works:

1. `db` is short for "database"
1. the database is called `nuccore` (not `genebank` or `gene`)
1. the argument `rettype` is "tt"; I think it stands for "REtURN type" (I also forget the second "t")


```{r}
# put accessions in vector
accessions_mrna <- c("AF049118","AF049114","AF049119","AF049115")

# get sequences
virus_mrna_list <- entrez_fetch_list(db = "nuccore",  # "nuccore" db for DNA 
                                   id = accessions_mrna, 
                                   rettype = "FASTA") # rettype has two t
```

We can clean these three sequences using a simple `for()` loop.  We set `parse = F` so we get things back as single character string.

```{r}
for(i in 1:length(virus_mrna_list)){
  virus_mrna_list[[i]] <- fasta_cleaner(virus_mrna_list[i], 
                                       parse = F)
}
```

Now we need to convert each of these into a named vector of sequences.
<!-- make function to do this -->
```{r}
mra_seq_vector <- c(AF049118 =  virus_mrna_list[[1]] ,
                 AF049114 = virus_mrna_list[[2]] ,
                 AF049119 = virus_mrna_list[[3]] ,
                 AF049115 = virus_mrna_list[[4]])
```


Finally convert this to a "stringset" using `Biostrings::DNAStringSet()`.

```{r}
dna_seq_stringset <- Biostrings::DNAStringSet(mra_seq_vector)
```

Let's see what we got
```{r}
dna_seq_stringset
```

Now we can make an alignment use `msa()`.

```{r}
virus_mrna_aln <- msa(inputSeqs = dna_seq_stringset,
                     method = "ClustalW")
```

The output looks like this
```{r}
virus_mrna_aln
```

This looks a LOT different than an amino acid alignment, which looked like this:
```{r}
data(virusaln)
virusaln
```

Why might the be different?  First, examine the output above and determine how long the DNA alignment is versus the amino acid alignment?  Why are they different, and why is one longer than the other?

The DNA alignment is 1097 columns, while the amino acid alignment is only 306 rows.  Note that 306*3 = 918.  1097 is pretty close to 1097.  What's the relevance of multiplying by 3?


## Calculationg genetic distance

You can calculate a genetic distance for DNA or mRNA sequences using the `dist.dna()` function in the `ape` package. `dist.dna()` takes a MSA of DNA or mRNA sequences as its input, and calculates the genetic distance between each pair of DNA sequences in the multiple alignment.

The `dist.dna()` function requires the input alignment to be in a special format known as `DNAbin` format, so we must use the `as.DNAbin()` function to convert our DNA alignment into this format before using the `dist.dna()` function.

```{r}
 # Convert the alignment to "DNAbin" format
virus_mrna_aln_bin <- ape::as.DNAbin(virus_mrna_aln)
```

The output of `as.DNAbin()` gives us a short summery of the alignment
```{r}
virus_mrna_aln_bin
```


Now to make and view the alignment:
```{r}
 # Calculate the genetic distance matrix
virus_mrna_dist <- ape::dist.dna(virus_mrna_aln_bin)

 # Print out the genetic distance matrix
virus_mrna_dist                             
```

NOTE: my results for this alignment are the same as the original by Coghlan.  I'm not sure why my amino acid alignment produces divergent results but the DNA is the same.



<!--chapter:end:20_genetic_distance.Rmd-->


# Unrooted neighbor-joining phylogenetic trees

Placeholder


## Preliminaries
### Key functions 
### Key vocab
## Building an unrooted phylogenetic tree for protein sequences
## Boostrap values indicate support for clades
## Branch lengths indicate divergence between sequences
## Unrooted trees lack an outgroup

<!--chapter:end:21-unrooted_tree_NJ_tree.Rmd-->


# (PART) Appendices {-}
# Appendix 01: Getting access to R {-}

Placeholder


### {-}
## Getting Started With R and RStudio
### RStudio Cloud
### Getting R onto your own computer
### Getting RStudio onto your computer
### Keep R and RStudio current

<!--chapter:end:99-A01_getting_access_to_R.Rmd-->


# Getting started with R itself (or not) {-}

Placeholder


## Vocabulary {-}
## R commands {-}
#### R's console as a scientific calculator
#### Practice: math in the console
#### The script editor
#### Comments 
## Help!
### Getting "help" from R
### Getting help from the internet
### Getting help from online forums
### Getting help from twitter
## Other features of RStudio
### Ajusting pane the layout
### Adjusting size of windows
## Practice (OPTIONAL)

<!--chapter:end:99-A02_getting_started_with_R.Rmd-->

