<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 9 A Brief introduction to R | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 9 A Brief introduction to R | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 9 A Brief introduction to R | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="installing-bioconductor.html"/>
<link rel="next" href="dna-descriptive-statics-part-1.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>1</b> How to review this book</a>
<ul>
<li class="chapter" data-level="1.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>1.1</b> File format</a></li>
<li class="chapter" data-level="1.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>1.2</b> What can you do?</a></li>
<li class="chapter" data-level="1.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>1.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>1.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="1.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>1.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="1.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>1.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="1.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>1.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>2</b> Test shiny app</a></li>
<li class="chapter" data-level="3" data-path="test-webpage-insertion.html"><a href="test-webpage-insertion.html"><i class="fa fa-check"></i><b>3</b> Test webpage insertion</a></li>
<li class="chapter" data-level="4" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>4</b> Datacamp test</a></li>
<li class="chapter" data-level="5" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>5</b> Downloading R</a>
<ul>
<li class="chapter" data-level="5.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>5.1</b> Preface</a></li>
<li class="chapter" data-level="5.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>5.2</b> Introduction to R</a></li>
<li class="chapter" data-level="5.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>5.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>5.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="5.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>5.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>5.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>6</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="6.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>6.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="6.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>6.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>7</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="7.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>7.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="7.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>7.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="7.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>7.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>8</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="8.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>8.1</b> Bioconductor</a></li>
<li class="chapter" data-level="8.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>8.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="8.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>8.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>8.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="8.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>8.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="8.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>8.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>8.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>9</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="9.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>9.1</b> Vocabulary</a></li>
<li class="chapter" data-level="9.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>9.2</b> R functions</a></li>
<li class="chapter" data-level="9.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>9.3</b> Interacting with R</a></li>
<li class="chapter" data-level="9.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>9.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>9.4.1</b> Vectors</a></li>
<li class="chapter" data-level="9.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>9.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="9.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>9.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="9.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>9.4.4</b> Lists</a></li>
<li class="chapter" data-level="9.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>9.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>9.5</b> Arguments</a></li>
<li class="chapter" data-level="9.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>9.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="9.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>9.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="9.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>9.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>9.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>9.9</b> Quiting R</a></li>
<li class="chapter" data-level="9.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>9.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>10</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="10.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-1"><i class="fa fa-check"></i><b>10.1</b> Preface</a></li>
<li class="chapter" data-level="10.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>10.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="10.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>10.3</b> Introduction</a></li>
<li class="chapter" data-level="10.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>10.4</b> Vocabulary</a></li>
<li class="chapter" data-level="10.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions"><i class="fa fa-check"></i><b>10.5</b> Functions</a></li>
<li class="chapter" data-level="10.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>10.6</b> Preliminaries</a></li>
<li class="chapter" data-level="10.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>10.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="10.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>10.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="10.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>10.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="10.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>10.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="10.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>10.8.3</b> DNA words</a></li>
<li class="chapter" data-level="10.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>10.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="10.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>10.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="10.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>10.9.1</b> License</a></li>
<li class="chapter" data-level="10.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>10.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>11</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="11.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-2"><i class="fa fa-check"></i><b>11.1</b> Preface</a></li>
<li class="chapter" data-level="11.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>11.2</b> Vocab</a></li>
<li class="chapter" data-level="11.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-1"><i class="fa fa-check"></i><b>11.3</b> Functions</a></li>
<li class="chapter" data-level="11.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>11.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="11.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>11.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>12</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="12.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-3"><i class="fa fa-check"></i><b>12.1</b> Preface</a></li>
<li class="chapter" data-level="12.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>12.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="functions-2.html"><a href="functions-2.html"><i class="fa fa-check"></i><b>13</b> Functions</a>
<ul>
<li class="chapter" data-level="13.1" data-path="functions-2.html"><a href="functions-2.html#vectors-in-r"><i class="fa fa-check"></i><b>13.1</b> Vectors in R</a></li>
<li class="chapter" data-level="13.2" data-path="functions-2.html"><a href="functions-2.html#math-on-vectors"><i class="fa fa-check"></i><b>13.2</b> Math on vectors</a></li>
<li class="chapter" data-level="13.3" data-path="functions-2.html"><a href="functions-2.html#functions-on-vectors"><i class="fa fa-check"></i><b>13.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="13.4" data-path="functions-2.html"><a href="functions-2.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>13.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="13.5" data-path="functions-2.html"><a href="functions-2.html#subsetting-vectors"><i class="fa fa-check"></i><b>13.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="13.6" data-path="functions-2.html"><a href="functions-2.html#sequences-of-numbers"><i class="fa fa-check"></i><b>13.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="13.7" data-path="functions-2.html"><a href="functions-2.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>13.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="13.8" data-path="functions-2.html"><a href="functions-2.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>13.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>14</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="14.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-4"><i class="fa fa-check"></i><b>14.1</b> Preface</a></li>
<li class="chapter" data-level="14.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>14.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="14.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>14.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>15</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="15.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-5"><i class="fa fa-check"></i><b>15.1</b> Preface</a></li>
<li class="chapter" data-level="15.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>15.2</b> Vocab</a></li>
<li class="chapter" data-level="15.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-3"><i class="fa fa-check"></i><b>15.3</b> Functions</a></li>
<li class="chapter" data-level="15.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>15.4</b> Functions in R</a></li>
<li class="chapter" data-level="15.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>15.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>16</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="16.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-4"><i class="fa fa-check"></i><b>16.0.1</b> Functions</a></li>
<li class="chapter" data-level="16.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>16.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="16.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>16.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="16.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>16.0.4</b> File types</a></li>
<li class="chapter" data-level="16.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>16.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="16.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>16.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="16.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>16.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="16.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>16.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>16.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="16.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>16.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="16.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>16.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="16.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>16.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="16.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>16.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>16.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="16.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>16.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>17</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="17.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-2"><i class="fa fa-check"></i><b>17.1</b> Preliminaries</a></li>
<li class="chapter" data-level="17.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>17.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="17.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>17.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="17.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>17.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="18" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>18</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="18.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>18.1</b> Preliminaries</a></li>
<li class="chapter" data-level="18.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>18.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>19</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="19.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>19.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="19.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>19.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="19.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>19.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="19.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>19.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="19.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>19.4</b> Inversions</a></li>
<li class="chapter" data-level="19.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>19.5</b> Translocations</a></li>
<li class="chapter" data-level="19.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>19.6</b> Random sequence</a></li>
<li class="chapter" data-level="19.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>19.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>20</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="20.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>20.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="20.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>20.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="20.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>20.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="20.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>20.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="20.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>20.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="20.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>20.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="20.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>20.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="20.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>20.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>21</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="21.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>21.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="21.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>21.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>21.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>22</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="22.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>22.1</b> Prelminaries</a></li>
<li class="chapter" data-level="22.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>22.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="22.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>22.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>23</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="23.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>23.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="23.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages"><i class="fa fa-check"></i><b>23.1.1</b> Packages</a></li>
<li class="chapter" data-level="23.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-5"><i class="fa fa-check"></i><b>23.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="23.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>23.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="23.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>23.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="23.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>23.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="23.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>23.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="23.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>23.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="23.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>23.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="23.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>23.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="23.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>23.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="23.6" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>23.6</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="23.6.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>23.6.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="23.6.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>23.6.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="23.6.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>23.6.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="23.6.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>23.6.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-2"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="23.7" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>23.7</b> Help!</a>
<ul>
<li class="chapter" data-level="23.7.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>23.7.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="23.7.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>23.7.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="23.7.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>23.7.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="23.7.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>23.7.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="23.8" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>23.8</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="23.8.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>23.8.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="23.8.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>23.8.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="23.9" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>23.9</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="a-brief-introduction-to-r" class="section level1" number="9">
<h1><span class="header-section-number">Chapter 9</span> A Brief introduction to R</h1>
<p><strong>By</strong>: Avril Coghlan.</p>
<p><strong>Adapted, edited and expanded</strong>: Nathan Brouwer under the Creative Commons 3.0 Attribution License <a href="https://creativecommons.org/licenses/by/3.0/">(CC BY 3.0)</a>.</p>
<p>This chapter provides a brief introduction to R. At the end of are links to additional resources for getting started with R.</p>
<!-- Add link to compbio4all -->
<div id="vocabulary" class="section level2" number="9.1">
<h2><span class="header-section-number">9.1</span> Vocabulary</h2>
<ul>
<li>scalar</li>
<li>vector</li>
<li>list</li>
<li>class</li>
<li>numeric</li>
<li>character</li>
<li>assignment</li>
<li>elements of an object</li>
<li>indices</li>
<li>attributes of an object</li>
<li>argument of a function</li>
</ul>
</div>
<div id="r-functions" class="section level2" number="9.2">
<h2><span class="header-section-number">9.2</span> R functions</h2>
<ul>
<li>&lt;-</li>
<li>[ ]</li>
<li>$</li>
<li>table()</li>
<li>function</li>
<li>c()</li>
<li>log10()</li>
<li>help(), ?</li>
<li>help.search()</li>
<li>RSiteSearch()</li>
<li>mean()</li>
<li>return()</li>
<li>q()</li>
</ul>
</div>
<div id="interacting-with-r" class="section level2" number="9.3">
<h2><span class="header-section-number">9.3</span> Interacting with R</h2>
<p>You will type <em>R</em> commands into the RStudio <strong>console</strong> in order to carry out analyses in <em>R</em>. In the RStudio console you will see the R prompt starting with the symbol “&gt;”. “&gt;” will always be there at the beginning of each new command - don’t try to delete it! Moreover, you never need to type it.</p>
<p><img src="images/R_cursor.png" width="51" /></p>
<p>We type the <strong>commands</strong> needed for a particular task after this prompt. The command is carried out by <em>R</em> after you hit the Return key.</p>
<p>Once you have started R, you can start typing commands into the RStudio console, and the results will be calculated immediately, for example:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiIyKjMifQ==
</div>
<p>Note that prior to the output of “6” it shows “[1]”.</p>
<p>Now subtraction:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiIxMC0zIn0=
</div>
<p>Again, prior to the output of “7” it shows “[1]”.</p>
<p><em>R</em> can act like a basic calculator that you type commands in to. You can also use it like a more advanced scientific calculator and create <strong>variables</strong> that store information. All variables created by R are called <strong>objects</strong>. In R, we assign values to variables using an arrow-looking function <code>&lt;-</code> the <strong>assignment operator</strong>. For example, we can <strong>assign</strong> the value 2*3 to the variable x using the command:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJ4IDwtIDIqMyJ9
</div>
To view the contents of any R object, just type its name, press enter, and the contents of that R object will be displayed:
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJ4In0=
</div>
</div>
<div id="variables-in-r" class="section level2" number="9.4">
<h2><span class="header-section-number">9.4</span> Variables in R</h2>
<p>There are several different types of objects in R with fancy math names, including <strong>scalars</strong>, <strong>vectors</strong>, <strong>matrices</strong> (singular: <strong>matrix), </strong>arrays<strong>, </strong>dataframes<strong>, </strong>tables<strong>, and </strong>lists<strong>. The </strong>scalar** variable x above is one example of an R object. While a scalar variable such as x has just one element, a <strong>vector</strong> consists of several elements. The elements in a vector are all of the same <strong>type</strong> (e.g.. numbers or alphabetic characters), while <strong>lists</strong> may include elements such as characters as well as numeric quantities. Vectors and dataframes are the most common variables you’ll use. You’ll also encounter matrices often, and lists are ubiquitous in R but beginning users often don’t encounter them because they remain behind the scenes.</p>
<div id="vectors" class="section level3" number="9.4.1">
<h3><span class="header-section-number">9.4.1</span> Vectors</h3>
<p>To create a vector, we can use the <code>c()</code> (combine) function. For example, to create a vector called <code>myvector</code> that has elements with values 8, 6, 9, 10, and 5, we type:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteXZlY3RvciA8LSBjKDgsIDYsIDksIDEwLCA1KSAjIG5vdGU6IGNvbW1hcyBiZXR3ZWVuIGVhY2ggbnVtYmVyISJ9
</div>
<p>To see the contents of the variable <code>myvector</code>, we can just type its name and press enter:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteXZlY3RvciJ9
</div>
</div>
<div id="vector-indexing" class="section level3" number="9.4.2">
<h3><span class="header-section-number">9.4.2</span> Vector indexing</h3>
<p>The <code>[1]</code> is the <strong>index</strong> of the first <strong>element</strong> in the vector. We can <strong>extract</strong> any element of the vector by typing the vector name with the index of that element given in <strong>square brackets</strong> <code>[...]</code>.</p>
<p>For example, to get the value of the 4th element in the vector <code>myvector</code>, we type:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteXZlY3Rvcls0XSJ9
</div>
</div>
<div id="character-vectors" class="section level3" number="9.4.3">
<h3><span class="header-section-number">9.4.3</span> Character vectors</h3>
<p>Vectors can contain letters, such as those designating nucleic acids</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteS5zZXEgPC0gYyhcIkFcIixcIlRcIixcIkNcIixcIkdcIikifQ==
</div>
<p>They can also contain multi-letter <strong>strings</strong>:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteS5vbGlnb3MgPC0gYyhcIkFUQ0dDXCIsXCJUVFRDR0NcIixcIkNDQ0dDR1wiLFwiR0dHQ0dDXCIpIn0=
</div>
</div>
<div id="lists" class="section level3" number="9.4.4">
<h3><span class="header-section-number">9.4.4</span> Lists</h3>
<p><strong>NOTE</strong>: <em>below is a discussion of lists in R. This is excellent information, but not necessary if this is your very very first time using R.</em></p>
<p>In contrast to a vector, a <strong>list</strong> can contain elements of different types, for example, both numbers and letters. A list can even include other variables such as a vector. The <code>list()</code> function is used to create a list. For example, we could create a list <code>mylist</code> by typing:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteWxpc3QgPC0gbGlzdChuYW1lPVwiQ2hhcmxlcyBEYXJ3aW5cIiwgXG4gICAgICAgICAgICAgICB3aWZlPVwiRW1tYSBEYXJ3aW5cIiwgXG4gICAgICAgICAgICAgICBteXZlY3RvcikifQ==
</div>
<p>We can then print out the contents of the list mylist by typing its name:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteWxpc3QifQ==
</div>
<p>The <strong>elements</strong> in a list are numbered, and can be referred to using <strong>indices</strong>. We can extract an element of a list by typing the list name with the index of the element given in double <strong>square brackets</strong> (in contrast to a vector, where we only use single square brackets).</p>
<p>We can extract the second element from <code>mylist</code> by typing:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteWxpc3RbWzJdXSAgIyBub3RlIHRoZSBkb3VibGUgc3F1YXJlIGJyYWNrZXRzIFtbLi4uXV0ifQ==
</div>
<p>As a baby step towards our next task, we can wrap index values as in the <code>c()</code> command like this:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteWxpc3RbW2MoMildXSAgIyBub3RlIHRoZSBkb3VibGUgc3F1YXJlIGJyYWNrZXRzIFtbLi4uXV0ifQ==
</div>
<p>The number <code>2</code> and <code>c(2)</code> mean the same thing.</p>
<p>Now, we can extract the second AND third elements from <code>mylist</code>. First, we put the indices 2 and 3 into a vector <code>c(2,3)</code>, then wrap that vector in double square brackets: <code>[c(2,3)]</code>. All together it looks like this.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteWxpc3RbYygyLDMpXSAjIG5vdGUgdGhlIGRvdWJsZSBicmFja2V0cyJ9
</div>
<p>Elements of lists may also be named, resulting in a <strong>named lists</strong>. The elements may then be referred to by giving the list name, followed by “$”, followed by the element name. For example, mylist$name is the same as mylist[[1]] and mylist$wife is the same as mylist[[2]]:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteWxpc3Qkd2lmZSJ9
</div>
<p>We can find out the names of the named elements in a list by using the <code>attributes()</code> function, for example:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJhdHRyaWJ1dGVzKG15bGlzdCkifQ==
</div>
<p>When you use the <code>attributes()</code> function to find the named elements of a list variable, the named elements are always listed under a heading “$names”. Therefore, we see that the named elements of the list variable <code>mylist</code> are called “name” and “wife”, and we can retrieve their values by typing mylist$name and mylist$wife, respectively.</p>
</div>
<div id="tables" class="section level3" number="9.4.5">
<h3><span class="header-section-number">9.4.5</span> Tables</h3>
<p>Another type of object that you will encounter in R is a <strong>table</strong>. The <code>table()</code> function allows you to total up or tabulate the number of times a value occurs within a vector. Tables are typically used on vectors containing <strong>character data</strong>, such as letters, words, or names, but can work on numeric data data.</p>
<div id="tables---the-basics" class="section level4" number="9.4.5.1">
<h4><span class="header-section-number">9.4.5.1</span> Tables - The basics</h4>
<p>If we made a vector variable “nucleotides” containing the of a DNA molecule, we can use the <code>table()</code> function to produce a <strong>table variable</strong> that contains the number of bases with each possible nucleotides:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJiYXNlcyA8LSBjKFwiQVwiLCBcIlRcIiwgXCJBXCIsIFwiQVwiLCBcIlRcIiwgXCJDXCIsIFwiR1wiLCBcIkNcIiwgXCJHXCIpIn0=
</div>
Now make the table
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJ0YWJsZShiYXNlcykifQ==
</div>
We can store the table variable produced by the function <code>table()</code>, and call the stored table “bases.table”, by typing:
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJiYXNlcy50YWJsZSA8LSB0YWJsZShiYXNlcykifQ==
</div>
<p>Tables also work on vectors containing numbers. First, a vector of numbers.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJudW1lcmljLnZlY3RlciA8LSBjKDEsMSwxLDEsMyw0LDQsNCw0KSJ9
</div>
<p>Second, a table, showing how many times each number occurs.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJ0YWJsZShudW1lcmljLnZlY3RlcikifQ==
</div>
</div>
<div id="tables---further-details" class="section level4" number="9.4.5.2">
<h4><span class="header-section-number">9.4.5.2</span> Tables - further details</h4>
<p>To access elements in a table variable, you need to use double square brackets, just like accessing elements in a list. For example, to access the fourth element in the table bases.table (the number of Ts in the sequence), we type:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJiYXNlcy50YWJsZVtbNF1dICAjIGRvdWJsZSBicmFja2V0cyEifQ==
</div>
<p>Alternatively, you can use the name of the fourth element in the table (“John”) to find the value of that table element:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJiYXNlcy50YWJsZVtbXCJUXCJdXSJ9
</div>
</div>
</div>
</div>
<div id="arguments" class="section level2" number="9.5">
<h2><span class="header-section-number">9.5</span> Arguments</h2>
<p>Functions in R usually require <strong>arguments</strong>, which are input variables (i.e.. objects) that are <strong>passed</strong> to them, which they then carry out some operation on. For example, the <code>log10()</code> function is passed a number, and it then calculates the log to the base 10 of that number:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJsb2cxMCgxMDApIn0=
</div>
<p>There’s a more generic function, <code>log()</code>, where we pass it not only a number to take the log of, but also the specific <strong>base</strong> of the logarithm. To take the log base 10 with the <code>log()</code> function we do this</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJsb2coMTAwLCBiYXNlID0gMTApIn0=
</div>
<p>We can also take logs with other bases, such as 2:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJsb2coMTAwLCBiYXNlID0gMikifQ==
</div>
</div>
<div id="help-files-with-help-and" class="section level2" number="9.6">
<h2><span class="header-section-number">9.6</span> Help files with <code>help()</code> and <code>?</code></h2>
<p>In <em>R</em>, you can get help about a particular function by using the <code>help()</code> function. For example, if you want help about the <code>log10()</code> function, you can type:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJoZWxwKFwibG9nMTBcIikifQ==
</div>
<p>When you use the <code>help()</code> function, a box or web pag will show up in one of the panes of RStudio with information about the function that you asked for help with. You can also use the <code>?</code> next to the function like this</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiI/bG9nMTAifQ==
</div>
<p>Help files are a mixed bag in R, and it can take some getting used to them. An excellent overview of this is Kieran Healy’s <a href="https://socviz.co/appendix.html">“How to read an R help page.”</a></p>
</div>
<div id="searching-for-functions-with-help.search-and-rsitesearch" class="section level2" number="9.7">
<h2><span class="header-section-number">9.7</span> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></h2>
<p>If you are not sure of the name of a function, but think you know part of its name, you can search for the function name using the <code>help.search()</code> and <code>RSiteSearch()</code> functions. The <code>help.search()</code> function searches to see if you already have a function installed (from one of the R packages that you have installed) that may be related to some topic you’re interested in. <code>RSiteSearch()</code> searches <em>all</em> R functions (including those in packages that you haven’t yet installed) for functions related to the topic you are interested in.</p>
<p>For example, if you want to know if there is a function to calculate the standard deviation (SD) of a set of numbers, you can search for the names of all installed functions containing the word “deviation” in their description by typing:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJoZWxwLnNlYXJjaChcImRldmlhdGlvblwiKSJ9
</div>
<p>Among the functions that were found, is the function <code>sd()</code> in the <code>stats</code> package (an R package that comes with the base R installation), which is used for calculating the standard deviation.</p>
<p>Now, instead of searching just the packages we’ve have on our computer let’s search all R packages on CRAN. Let’s look for things related to DNA. Note that <code>RSiteSearch()</code> doesn’t provide output within RStudio, but rather opens up your web browser for you to display the results.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJSU2l0ZVNlYXJjaChcIkROQVwiKSJ9
</div>
<p>The results of the <code>RSiteSearch()</code> function will be hits to descriptions of R functions, as well as to R mailing list discussions of those functions.</p>
</div>
<div id="more-on-functions" class="section level2" number="9.8">
<h2><span class="header-section-number">9.8</span> More on functions</h2>
<p>We can perform computations with R using objects such as scalars and vectors. For example, to calculate the average of the values in the vector <code>myvector</code> (i.e.. the average of 8, 6, 9, 10 and 5), we can use the <code>mean()</code> function:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJtZWFuKG15dmVjdG9yKSAjIG5vdGU6IG5vIFwiIFwiIn0=
</div>
<p>We have been using built-in R functions such as mean(), length(), print(), plot(), etc.</p>
<div id="writing-your-own-functions" class="section level3" number="9.8.1">
<h3><span class="header-section-number">9.8.1</span> Writing your own functions</h3>
<p><strong>NOTE</strong>: *Writing your own functions is an advanced skills. New users can skip this section.</p>
<p>We can also create our own functions in R to do calculations that you want to carry out very often on different input data sets. For example, we can create a function to calculate the value of 20 plus square of some input number:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteWZ1bmN0aW9uIDwtIGZ1bmN0aW9uKHgpIHsgcmV0dXJuKDIwICsgKHgqeCkpIH0ifQ==
</div>
<p>This function will calculate the square of a number (x), and then add 20 to that value. The <code>return()</code> statement returns the calculated value. Once you have typed in this function, the function is then available for use. For example, we can use the function for different input numbers (e.g.. 10, 25):</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJteWZ1bmN0aW9uKDEwKSJ9
</div>
</div>
</div>
<div id="quiting-r" class="section level2" number="9.9">
<h2><span class="header-section-number">9.9</span> Quiting R</h2>
<p>To quit R either close the program, or type:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJxKCkifQ==
</div>
</div>
<div id="links-and-further-reading" class="section level2" number="9.10">
<h2><span class="header-section-number">9.10</span> Links and Further Reading</h2>
<p>Some links are included here for further reading.</p>
<p>For a more in-depth introduction to R, a good online tutorial is available on the “Kickstarting R” website, cran.r-project.org/doc/contrib/Lemon-kickstart.</p>
<p>There is another nice (slightly more in-depth) tutorial to R available on the “Introduction to R” website, cran.r-project.org/doc/manuals/R-intro.html.</p>
<p><a href="https://learningstatisticswithr.com/book/introR.html">Chapter 3</a> of Danielle Navarro’s book is an excellent intro to the basics of R.</p>
<!-- Add compbio4all -->

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="installing-bioconductor.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="dna-descriptive-statics-part-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/05-brief_intro_to_R_AC01-5.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/05-brief_intro_to_R_AC01-5.Rmd",
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
