<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 35 BLAST Alignment score distribution | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 35 BLAST Alignment score distribution | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 35 BLAST Alignment score distribution | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-15" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="getting-started-with-r-itself-or-not.html"/>
<link rel="next" href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>1</b> How to review this book</a>
<ul>
<li class="chapter" data-level="1.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>1.1</b> File format</a></li>
<li class="chapter" data-level="1.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>1.2</b> What can you do?</a></li>
<li class="chapter" data-level="1.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>1.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>1.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="1.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>1.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="1.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>1.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="1.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>1.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="test-audio-link.html"><a href="test-audio-link.html"><i class="fa fa-check"></i><b>2</b> Test audio link</a></li>
<li class="chapter" data-level="3" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>3</b> Test shiny app</a></li>
<li class="chapter" data-level="4" data-path="test-youtube-vidoe.html"><a href="test-youtube-vidoe.html"><i class="fa fa-check"></i><b>4</b> Test YouTube vidoe</a></li>
<li class="chapter" data-level="5" data-path="test-webpage-insertion.html"><a href="test-webpage-insertion.html"><i class="fa fa-check"></i><b>5</b> Test webpage insertion</a></li>
<li class="chapter" data-level="6" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>6</b> Datacamp test</a></li>
<li class="chapter" data-level="7" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>7</b> Downloading R</a>
<ul>
<li class="chapter" data-level="7.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>7.1</b> Preface</a></li>
<li class="chapter" data-level="7.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>7.2</b> Introduction to R</a></li>
<li class="chapter" data-level="7.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>7.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>7.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="7.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>7.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>7.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>8</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="8.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>8.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="8.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>8.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>9</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="9.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>9.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="9.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>9.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="9.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>9.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>10</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="10.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>10.1</b> Bioconductor</a></li>
<li class="chapter" data-level="10.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>10.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="10.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>10.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>10.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="10.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>10.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="10.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>10.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>10.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>11</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>11.1</b> Vocabulary</a></li>
<li class="chapter" data-level="11.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>11.2</b> R functions</a></li>
<li class="chapter" data-level="11.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>11.3</b> Interacting with R</a></li>
<li class="chapter" data-level="11.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>11.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>11.4.1</b> Vectors</a></li>
<li class="chapter" data-level="11.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>11.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="11.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>11.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="11.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>11.4.4</b> Lists</a></li>
<li class="chapter" data-level="11.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>11.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>11.5</b> Arguments</a></li>
<li class="chapter" data-level="11.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>11.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="11.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>11.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="11.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>11.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="11.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>11.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="11.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>11.9</b> Quiting R</a></li>
<li class="chapter" data-level="11.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>11.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>12</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="12.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-1"><i class="fa fa-check"></i><b>12.1</b> Preface</a></li>
<li class="chapter" data-level="12.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>12.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="12.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>12.3</b> Introduction</a></li>
<li class="chapter" data-level="12.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>12.4</b> Vocabulary</a></li>
<li class="chapter" data-level="12.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions"><i class="fa fa-check"></i><b>12.5</b> Functions</a></li>
<li class="chapter" data-level="12.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>12.6</b> Preliminaries</a></li>
<li class="chapter" data-level="12.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>12.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="12.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>12.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="12.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>12.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="12.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>12.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="12.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>12.8.3</b> DNA words</a></li>
<li class="chapter" data-level="12.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>12.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="12.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>12.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="12.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>12.9.1</b> License</a></li>
<li class="chapter" data-level="12.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>12.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>13</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="13.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-2"><i class="fa fa-check"></i><b>13.1</b> Preface</a></li>
<li class="chapter" data-level="13.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>13.2</b> Vocab</a></li>
<li class="chapter" data-level="13.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-1"><i class="fa fa-check"></i><b>13.3</b> Functions</a></li>
<li class="chapter" data-level="13.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>13.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="13.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>13.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>14</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="14.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-3"><i class="fa fa-check"></i><b>14.1</b> Preface</a></li>
<li class="chapter" data-level="14.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>14.2</b> Vocab</a></li>
<li class="chapter" data-level="14.3" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vectors-in-r"><i class="fa fa-check"></i><b>14.3</b> Vectors in R</a></li>
<li class="chapter" data-level="14.4" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#math-on-vectors"><i class="fa fa-check"></i><b>14.4</b> Math on vectors</a></li>
<li class="chapter" data-level="14.5" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#functions-on-vectors"><i class="fa fa-check"></i><b>14.5</b> Functions on vectors</a></li>
<li class="chapter" data-level="14.6" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>14.6</b> Operations with two vectors</a></li>
<li class="chapter" data-level="14.7" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#subsetting-vectors"><i class="fa fa-check"></i><b>14.7</b> Subsetting vectors</a></li>
<li class="chapter" data-level="14.8" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#sequences-of-numbers"><i class="fa fa-check"></i><b>14.8</b> Sequences of numbers</a></li>
<li class="chapter" data-level="14.9" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>14.9</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="14.10" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>14.10</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>15</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="15.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-4"><i class="fa fa-check"></i><b>15.1</b> Preface</a></li>
<li class="chapter" data-level="15.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>15.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="15.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>15.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>16</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="16.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-5"><i class="fa fa-check"></i><b>16.1</b> Preface</a></li>
<li class="chapter" data-level="16.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>16.2</b> Vocab</a></li>
<li class="chapter" data-level="16.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-2"><i class="fa fa-check"></i><b>16.3</b> Functions</a></li>
<li class="chapter" data-level="16.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>16.4</b> Functions in R</a></li>
<li class="chapter" data-level="16.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>16.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="fasta-files.html"><a href="fasta-files.html"><i class="fa fa-check"></i><b>17</b> FASTA Files</a>
<ul>
<li class="chapter" data-level="17.1" data-path="fasta-files.html"><a href="fasta-files.html#example-fasta-file"><i class="fa fa-check"></i><b>17.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="17.2" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>17.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="17.3" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>17.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="17.4" data-path="fasta-files.html"><a href="fasta-files.html#fastq-format"><i class="fa fa-check"></i><b>17.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>18</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="18.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-3"><i class="fa fa-check"></i><b>18.0.1</b> Functions</a></li>
<li class="chapter" data-level="18.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>18.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="18.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>18.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="18.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>18.0.4</b> File types</a></li>
<li class="chapter" data-level="18.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>18.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="18.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>18.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="18.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>18.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="18.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>18.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>18.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="18.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>18.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="18.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>18.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="18.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>18.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="18.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>18.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>18.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="18.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>18.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>19</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="19.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-2"><i class="fa fa-check"></i><b>19.1</b> Preliminaries</a></li>
<li class="chapter" data-level="19.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>19.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>19.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="19.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>19.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="20" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>20</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="20.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>20.1</b> Preliminaries</a></li>
<li class="chapter" data-level="20.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>20.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>21</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="21.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>21.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="21.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>21.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>21.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="21.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>21.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="21.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>21.4</b> Inversions</a></li>
<li class="chapter" data-level="21.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>21.5</b> Translocations</a></li>
<li class="chapter" data-level="21.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>21.6</b> Random sequence</a></li>
<li class="chapter" data-level="21.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>21.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>22</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="22.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>22.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="22.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>22.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="22.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>22.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="22.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>22.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="22.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>22.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="22.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>22.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="22.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>22.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="22.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>22.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>23</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="23.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>23.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="23.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>23.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="23.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>23.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html"><i class="fa fa-check"></i><b>24</b> Matrix vocabulary</a>
<ul>
<li class="chapter" data-level="24.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#introduction-1"><i class="fa fa-check"></i><b>24.1</b> Introduction</a></li>
<li class="chapter" data-level="24.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#preliminaries-7"><i class="fa fa-check"></i><b>24.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="24.2.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#packages"><i class="fa fa-check"></i><b>24.2.1</b> Packages</a></li>
<li class="chapter" data-level="24.2.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#vocab-3"><i class="fa fa-check"></i><b>24.2.2</b> Vocab</a></li>
<li class="chapter" data-level="24.2.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#functions-used"><i class="fa fa-check"></i><b>24.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="24.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#load-packages"><i class="fa fa-check"></i><b>24.3</b> Load packages</a></li>
<li class="chapter" data-level="24.4" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#matrix-structure"><i class="fa fa-check"></i><b>24.4</b> Matrix structure</a></li>
<li class="chapter" data-level="24.5" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#diagonal-of-a-matrix"><i class="fa fa-check"></i><b>24.5</b> Diagonal of a matrix</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="alignment-in-r.html"><a href="alignment-in-r.html"><i class="fa fa-check"></i><b>25</b> Alignment in R</a>
<ul>
<li class="chapter" data-level="25.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#preliminaries-8"><i class="fa fa-check"></i><b>25.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="25.1.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#packages-1"><i class="fa fa-check"></i><b>25.1.1</b> Packages</a></li>
<li class="chapter" data-level="25.1.2" data-path="alignment-in-r.html"><a href="alignment-in-r.html#data"><i class="fa fa-check"></i><b>25.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="26" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html"><i class="fa fa-check"></i><b>26</b> Manipulating matrices and vectors: worked example</a>
<ul>
<li class="chapter" data-level="26.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#introduction-2"><i class="fa fa-check"></i><b>26.1</b> Introduction</a></li>
<li class="chapter" data-level="26.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#preliminaries-9"><i class="fa fa-check"></i><b>26.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="26.2.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-2"><i class="fa fa-check"></i><b>26.2.1</b> Packages</a></li>
<li class="chapter" data-level="26.2.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#vocab-4"><i class="fa fa-check"></i><b>26.2.2</b> Vocab</a></li>
<li class="chapter" data-level="26.2.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#functions-used-1"><i class="fa fa-check"></i><b>26.2.3</b> Functions used</a></li>
<li class="chapter" data-level="26.2.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-3"><i class="fa fa-check"></i><b>26.2.4</b> Packages</a></li>
<li class="chapter" data-level="26.2.5" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#data-1"><i class="fa fa-check"></i><b>26.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="26.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#matrix-elements-can-be-accessed-with-square-bracket-notation"><i class="fa fa-check"></i><b>26.3</b> Matrix elements can be accessed with square bracket notation</a></li>
<li class="chapter" data-level="26.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#r-object-names"><i class="fa fa-check"></i><b>26.4</b> R object names</a>
<ul>
<li class="chapter" data-level="26.4.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-vectors"><i class="fa fa-check"></i><b>26.4.1</b> Accessing items in names vectors</a></li>
<li class="chapter" data-level="26.4.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-matrices"><i class="fa fa-check"></i><b>26.4.2</b> Accessing items in names matrices</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="27" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html"><i class="fa fa-check"></i><b>27</b> Alignment by eye in R</a>
<ul>
<li class="chapter" data-level="27.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#preliminaries-10"><i class="fa fa-check"></i><b>27.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="27.1.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#packages-4"><i class="fa fa-check"></i><b>27.1.1</b> Packages</a></li>
<li class="chapter" data-level="27.1.2" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#data-2"><i class="fa fa-check"></i><b>27.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="28" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>28</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="28.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>28.1</b> Prelminaries</a></li>
<li class="chapter" data-level="28.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>28.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="28.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>28.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>29</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="29.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-11"><i class="fa fa-check"></i><b>29.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="29.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages-5"><i class="fa fa-check"></i><b>29.1.1</b> Packages</a></li>
<li class="chapter" data-level="29.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-4"><i class="fa fa-check"></i><b>29.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="29.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>29.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="29.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>29.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="29.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>29.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="29.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>29.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="29.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>29.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="29.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>29.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="29.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>29.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="29.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>29.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html"><i class="fa fa-check"></i><b>30</b> The BLOSUM scoring matrix in R</a>
<ul>
<li class="chapter" data-level="30.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#introduction-3"><i class="fa fa-check"></i><b>30.1</b> Introduction</a></li>
<li class="chapter" data-level="30.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#preliminaries-12"><i class="fa fa-check"></i><b>30.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="30.2.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#packages-6"><i class="fa fa-check"></i><b>30.2.1</b> Packages</a></li>
<li class="chapter" data-level="30.2.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#vocab-5"><i class="fa fa-check"></i><b>30.2.2</b> Vocab</a></li>
<li class="chapter" data-level="30.2.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#functions-used-2"><i class="fa fa-check"></i><b>30.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="30.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#load-packages-1"><i class="fa fa-check"></i><b>30.3</b> Load packages</a></li>
<li class="chapter" data-level="30.4" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#the-blosum-matrix"><i class="fa fa-check"></i><b>30.4</b> The BLOSUM matrix</a></li>
<li class="chapter" data-level="30.5" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#visualizing-the-blosum-matrix"><i class="fa fa-check"></i><b>30.5</b> Visualizing the BLOSUM matrix</a></li>
</ul></li>
<li class="chapter" data-level="31" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>31</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="31.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-13"><i class="fa fa-check"></i><b>31.1</b> Preliminaries</a></li>
<li class="chapter" data-level="31.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-4"><i class="fa fa-check"></i><b>31.2</b> Introduction</a></li>
<li class="chapter" data-level="31.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>31.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="31.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>31.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="32" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>32</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="32.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-14"><i class="fa fa-check"></i><b>32.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="32.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>32.1.1</b> Key functions</a></li>
<li class="chapter" data-level="32.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>32.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="32.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>32.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="32.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>32.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="32.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>32.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="32.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>32.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="32.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>32.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="33" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>33</b> A complete bioinformatics workflow in R</a>
<ul>
<li class="chapter" data-level="33.1" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#introduction-5"><i class="fa fa-check"></i><b>33.1</b> Introduction</a></li>
<li class="chapter" data-level="33.2" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>33.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="33.2.1" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#vocab-6"><i class="fa fa-check"></i><b>33.2.1</b> Vocab</a></li>
<li class="chapter" data-level="33.2.2" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>33.2.2</b> R functions</a></li>
<li class="chapter" data-level="33.2.3" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>33.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="33.2.4" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>33.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="33.3" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>33.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="33.4" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>33.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="33.5" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>33.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="33.6" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>33.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="33.7" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>33.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="33.8" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>33.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="33.8.1" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>33.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="33.8.2" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>33.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="33.9" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>33.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="33.10" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>33.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="33.10.1" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>33.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="34" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>34</b> Logarithms in R</a></li>
<li class="part"><span><b>I Appendices</b></span></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li><a href="appendix-01-getting-access-to-r.html#section"></a></li>
<li class="chapter" data-level="34.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>34.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="34.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>34.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="34.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>34.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="34.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>34.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="34.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>34.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-2"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="34.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>34.2</b> Help!</a>
<ul>
<li class="chapter" data-level="34.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>34.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="34.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>34.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="34.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>34.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="34.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>34.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="34.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>34.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="34.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>34.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="34.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>34.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="34.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>34.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html"><i class="fa fa-check"></i><b>35</b> BLAST Alignment score distribution</a>
<ul>
<li class="chapter" data-level="35.1" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#packages-7"><i class="fa fa-check"></i><b>35.1</b> Packages</a></li>
<li class="chapter" data-level="35.2" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#distributions-of-alignment-scores"><i class="fa fa-check"></i><b>35.2</b> Distributions of alignment scores</a></li>
<li class="chapter" data-level="35.3" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-data"><i class="fa fa-check"></i><b>35.3</b> Simulating data</a></li>
<li class="chapter" data-level="35.4" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#lots-of-data-in-biology-is-normal"><i class="fa fa-check"></i><b>35.4</b> Lots of data in biology is normal</a></li>
<li class="chapter" data-level="35.5" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-the-normal-distribution"><i class="fa fa-check"></i><b>35.5</b> Simulating the normal distribution</a></li>
<li class="chapter" data-level="35.6" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#the-extreme-value-distribution"><i class="fa fa-check"></i><b>35.6</b> The extreme value distribution</a></li>
<li class="chapter" data-level="35.7" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#replicating-figure-4.14"><i class="fa fa-check"></i><b>35.7</b> Replicating Figure 4.14</a></li>
<li class="chapter" data-level="35.8" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#why-do-we-care"><i class="fa fa-check"></i><b>35.8</b> Why do we care?</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><i class="fa fa-check"></i><b>36</b> Calculating E values, Bit scores, and P-values from BLAST output</a>
<ul>
<li class="chapter" data-level="36.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-6"><i class="fa fa-check"></i><b>36.1</b> Introduction</a></li>
<li class="chapter" data-level="36.2" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#case-study"><i class="fa fa-check"></i><b>36.2</b> Case Study</a></li>
<li class="chapter" data-level="36.3" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-statistics"><i class="fa fa-check"></i><b>36.3</b> BLAST Statistics</a></li>
<li class="chapter" data-level="36.4" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot"><i class="fa fa-check"></i><b>36.4</b> Dotplot</a>
<ul>
<li class="chapter" data-level="36.4.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot-in-r---optional-experimental-code"><i class="fa fa-check"></i><b>36.4.1</b> Dotplot in R - optional, experimental code</a></li>
</ul></li>
<li class="chapter" data-level="36.5" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-search-summary"><i class="fa fa-check"></i><b>36.5</b> BLAST “Search Summary”</a></li>
<li class="chapter" data-level="36.6" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-alignments"><i class="fa fa-check"></i><b>36.6</b> BLAST Alignments</a></li>
<li class="chapter" data-level="36.7" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#equations"><i class="fa fa-check"></i><b>36.7</b> Equations</a>
<ul>
<li class="chapter" data-level="36.7.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-value-eq-4.5-pg-142"><i class="fa fa-check"></i><b>36.7.1</b> E value (eq 4.5, pg 142)</a></li>
</ul></li>
<li class="chapter" data-level="36.8" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bit-scores-s"><i class="fa fa-check"></i><b>36.8</b> Bit Scores (S’)</a></li>
<li class="chapter" data-level="36.9" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-from-bit-scores"><i class="fa fa-check"></i><b>36.9</b> E from Bit scores</a></li>
<li class="chapter" data-level="36.10" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#p-values"><i class="fa fa-check"></i><b>36.10</b> P values</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html"><i class="fa fa-check"></i><b>37</b> Calculating the number of possible phylogenetic trees</a>
<ul>
<li class="chapter" data-level="37.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#repliminaries"><i class="fa fa-check"></i><b>37.1</b> Repliminaries</a></li>
<li class="chapter" data-level="37.2" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#concepts"><i class="fa fa-check"></i><b>37.2</b> Concepts</a></li>
<li class="chapter" data-level="37.3" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#functions-5"><i class="fa fa-check"></i><b>37.3</b> Functions</a></li>
<li class="chapter" data-level="37.4" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#introduction-7"><i class="fa fa-check"></i><b>37.4</b> Introduction</a></li>
<li class="chapter" data-level="37.5" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#number-of-rooted-trees"><i class="fa fa-check"></i><b>37.5</b> Number of rooted trees</a></li>
<li class="chapter" data-level="37.6" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#number-of-rooted-trees-in-r"><i class="fa fa-check"></i><b>37.6</b> Number of rooted trees in R</a>
<ul>
<li class="chapter" data-level="37.6.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#unique-topologies-versus-rotations"><i class="fa fa-check"></i><b>37.6.1</b> Unique topologies versus rotations</a></li>
</ul></li>
<li class="chapter" data-level="37.7" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#bilography"><i class="fa fa-check"></i><b>37.7</b> Bilography</a></li>
<li class="chapter" data-level="37.8" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#upgma-the-hard-way"><i class="fa fa-check"></i><b>37.8</b> UPGMA the hard way</a></li>
<li class="chapter" data-level="37.9" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#introudction-to-upgma"><i class="fa fa-check"></i><b>37.9</b> Introudction to UPGMA</a></li>
<li class="chapter" data-level="37.10" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#preliminaries-15"><i class="fa fa-check"></i><b>37.10</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="37.10.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#r-libraries"><i class="fa fa-check"></i><b>37.10.1</b> R Libraries</a></li>
<li class="chapter" data-level="37.10.2" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#custom-function-well-use"><i class="fa fa-check"></i><b>37.10.2</b> Custom function we’ll use</a></li>
</ul></li>
<li class="chapter" data-level="37.11" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#data-matrix"><i class="fa fa-check"></i><b>37.11</b> Data matrix</a>
<ul>
<li class="chapter" data-level="37.11.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#building-the-matrix"><i class="fa fa-check"></i><b>37.11.1</b> Building the matrix</a></li>
</ul></li>
<li class="chapter" data-level="37.12" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#what-does-this-distance-matrix-represent"><i class="fa fa-check"></i><b>37.12</b> What does this distance matrix represent?</a>
<ul>
<li class="chapter" data-level="37.12.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#genetic-distances-conceptualized-as-pairwise-linear-distances"><i class="fa fa-check"></i><b>37.12.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="37.12.2" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#genetic-distances-conceptualized-as-2d-map-of-linear-distances"><i class="fa fa-check"></i><b>37.12.2</b> Genetic distances conceptualized as 2D map of linear distances</a></li>
</ul></li>
<li class="chapter" data-level="37.13" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#from-linear-distances-to-tree-distances"><i class="fa fa-check"></i><b>37.13</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="37.14" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#the-upgma-algorithm"><i class="fa fa-check"></i><b>37.14</b> The UPGMA algorithm</a>
<ul>
<li class="chapter" data-level="37.14.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#step-1-find-the-2-taxa-that-are-currently-closest-together"><i class="fa fa-check"></i><b>37.14.1</b> Step 1: find the 2 taxa that are currently closest together</a></li>
<li class="chapter" data-level="37.14.2" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node"><i class="fa fa-check"></i><b>37.14.2</b> Step 2: calculate the branch lengths connect the taxa to their shared node</a></li>
<li class="chapter" data-level="37.14.3" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#step-3a-determine-if-the-algorithm-is-done"><i class="fa fa-check"></i><b>37.14.3</b> Step 3a: determine if the algorithm is done</a></li>
<li class="chapter" data-level="37.14.4" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u"><i class="fa fa-check"></i><b>37.14.4</b> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</a></li>
<li class="chapter" data-level="37.14.5" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#step-4-determine-the-distance-from-u-to-the-other-remaining-clusters."><i class="fa fa-check"></i><b>37.14.5</b> Step 4: Determine the distance from u to the other remaining clusters.</a></li>
<li class="chapter" data-level="37.14.6" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#step-5-create-a-new-matrix"><i class="fa fa-check"></i><b>37.14.6</b> Step 5: Create a new matrix</a></li>
<li class="chapter" data-level="37.14.7" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa"><i class="fa fa-check"></i><b>37.14.7</b> Key step: calculating the distance from a clade to the remaining taxa</a></li>
</ul></li>
<li class="chapter" data-level="37.15" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#calculate-2d-representation"><i class="fa fa-check"></i><b>37.15</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="37.16" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#algorithm---round-1"><i class="fa fa-check"></i><b>37.16</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="37.16.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#determine-minimum-value-in-current-matrix"><i class="fa fa-check"></i><b>37.16.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="37.16.2" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#combine-species-into-clade"><i class="fa fa-check"></i><b>37.16.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="37.16.3" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#calculate-branch-length"><i class="fa fa-check"></i><b>37.16.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="37.16.4" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#calculate-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>37.16.4</b> Calculate distance from clade to all other points</a></li>
<li class="chapter" data-level="37.16.5" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#update-matrix-with-new-distance"><i class="fa fa-check"></i><b>37.16.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="37.17" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#algorithm---next-round"><i class="fa fa-check"></i><b>37.17</b> Algorithm - next round</a>
<ul>
<li class="chapter" data-level="37.17.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#combine-species-into-clade-1"><i class="fa fa-check"></i><b>37.17.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="37.17.2" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#calculate-branch-length-1"><i class="fa fa-check"></i><b>37.17.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="37.17.3" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#recalculate-distance-from-abe-to-all-remainign-taxa"><i class="fa fa-check"></i><b>37.17.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="37.17.4" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#update-the-matrix"><i class="fa fa-check"></i><b>37.17.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="37.18" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#next-iteration"><i class="fa fa-check"></i><b>37.18</b> Next iteration</a>
<ul>
<li class="chapter" data-level="37.18.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#calculate-branch-lengths"><i class="fa fa-check"></i><b>37.18.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="37.18.2" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#combine-species-into-clade-2"><i class="fa fa-check"></i><b>37.18.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="37.18.3" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#calculate-branch-length-2"><i class="fa fa-check"></i><b>37.18.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="37.18.4" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#distance-of-c-to-abe-and-d-to-abe"><i class="fa fa-check"></i><b>37.18.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="37.19" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#finish-up"><i class="fa fa-check"></i><b>37.19</b> Finish up</a>
<ul>
<li class="chapter" data-level="37.19.1" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#combine-species-into-clade-3"><i class="fa fa-check"></i><b>37.19.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="37.19.2" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#calculate-branch-length-3"><i class="fa fa-check"></i><b>37.19.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="37.20" data-path="calculating-the-number-of-possible-phylogenetic-trees.html"><a href="calculating-the-number-of-possible-phylogenetic-trees.html#finalizing-branch-lengths"><i class="fa fa-check"></i><b>37.20</b> Finalizing branch lengths</a></li>
</ul></li>
<li class="chapter" data-level="38" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html"><i class="fa fa-check"></i><b>38</b> Representing phylogenetic trees in Newick format</a>
<ul>
<li class="chapter" data-level="38.1" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#vocab-7"><i class="fa fa-check"></i><b>38.1</b> Vocab</a></li>
<li class="chapter" data-level="38.2" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#introduction-8"><i class="fa fa-check"></i><b>38.2</b> Introduction</a></li>
<li class="chapter" data-level="38.3" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-three-taxa"><i class="fa fa-check"></i><b>38.3</b> Trees with three taxa</a></li>
<li class="chapter" data-level="38.4" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-four-taxa"><i class="fa fa-check"></i><b>38.4</b> Trees with four taxa</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html"><i class="fa fa-check"></i><b>39</b> Plotting trees from Newick notation in R</a>
<ul>
<li class="chapter" data-level="39.1" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#preliminaries-16"><i class="fa fa-check"></i><b>39.1</b> Preliminaries</a></li>
<li class="chapter" data-level="39.2" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#note"><i class="fa fa-check"></i><b>39.2</b> Note</a></li>
<li class="chapter" data-level="39.3" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick"><i class="fa fa-check"></i><b>39.3</b> 3 taxa tree in Newick</a></li>
<li class="chapter" data-level="39.4" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick-1"><i class="fa fa-check"></i><b>39.4</b> 4 taxa tree in Newick</a></li>
<li class="chapter" data-level="39.5" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-in-newick-format"><i class="fa fa-check"></i><b>39.5</b> 5 taxa in Newick Format</a></li>
<li class="chapter" data-level="39.6" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#many-taxa-tree"><i class="fa fa-check"></i><b>39.6</b> Many-taxa tree</a></li>
<li class="chapter" data-level="39.7" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#newick-and-branch-lengths"><i class="fa fa-check"></i><b>39.7</b> Newick and Branch lengths</a></li>
<li class="chapter" data-level="39.8" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#additional-reading"><i class="fa fa-check"></i><b>39.8</b> Additional Reading</a></li>
</ul></li>
<li class="chapter" data-level="40" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><i class="fa fa-check"></i><b>40</b> Phylogenetic trees using the UPGMA clustering algoirthm</a>
<ul>
<li class="chapter" data-level="40.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#introduction-9"><i class="fa fa-check"></i><b>40.1</b> Introduction</a></li>
<li class="chapter" data-level="40.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#preliminaries-17"><i class="fa fa-check"></i><b>40.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="40.2.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#vocab-8"><i class="fa fa-check"></i><b>40.2.1</b> Vocab</a></li>
<li class="chapter" data-level="40.2.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-r-functions"><i class="fa fa-check"></i><b>40.2.2</b> General R functions</a></li>
<li class="chapter" data-level="40.2.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#specific-functions"><i class="fa fa-check"></i><b>40.2.3</b> Specific functions</a></li>
</ul></li>
<li class="chapter" data-level="40.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#sequence-comparisons"><i class="fa fa-check"></i><b>40.3</b> Sequence comparisons</a></li>
<li class="chapter" data-level="40.4" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrices"><i class="fa fa-check"></i><b>40.4</b> Distance matrices</a></li>
<li class="chapter" data-level="40.5" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#from-a-distance-matrix-to-phylogenetic-tree"><i class="fa fa-check"></i><b>40.5</b> From a distance matrix to phylogenetic tree</a></li>
<li class="chapter" data-level="40.6" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-raw-sequences"><i class="fa fa-check"></i><b>40.6</b> Distance matrix in R from raw sequences</a></li>
<li class="chapter" data-level="40.7" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-normal-matrix"><i class="fa fa-check"></i><b>40.7</b> Distance matrix in R from normal matrix</a></li>
<li class="chapter" data-level="40.8" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#upgma-in-r"><i class="fa fa-check"></i><b>40.8</b> UPGMA in R</a></li>
<li class="chapter" data-level="40.9" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#how-upgma-clustering-works"><i class="fa fa-check"></i><b>40.9</b> How UPGMA clustering works</a>
<ul>
<li class="chapter" data-level="40.9.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#challenge"><i class="fa fa-check"></i><b>40.9.1</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="40.10" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#further-information"><i class="fa fa-check"></i><b>40.10</b> Further information</a>
<ul>
<li class="chapter" data-level="40.10.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-info"><i class="fa fa-check"></i><b>40.10.1</b> General info</a></li>
<li class="chapter" data-level="40.10.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#worked-examples"><i class="fa fa-check"></i><b>40.10.2</b> Worked examples:</a></li>
<li class="chapter" data-level="40.10.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#additional-notes"><i class="fa fa-check"></i><b>40.10.3</b> Additional notes:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="41" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><i class="fa fa-check"></i><b>41</b> Tutorial: Writing functions to calculate the number of phylogenetic trees</a>
<ul>
<li class="chapter" data-level="41.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#repliminaries-1"><i class="fa fa-check"></i><b>41.1</b> Repliminaries</a>
<ul>
<li class="chapter" data-level="41.1.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-6"><i class="fa fa-check"></i><b>41.1.1</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="41.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#concepts-1"><i class="fa fa-check"></i><b>41.2</b> Concepts</a>
<ul>
<li class="chapter" data-level="41.2.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vocab-9"><i class="fa fa-check"></i><b>41.2.1</b> Vocab:</a></li>
<li class="chapter" data-level="41.2.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#packages-8"><i class="fa fa-check"></i><b>41.2.2</b> Packages:</a></li>
</ul></li>
<li class="chapter" data-level="41.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#introduction-10"><i class="fa fa-check"></i><b>41.3</b> Introduction</a></li>
<li class="chapter" data-level="41.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-1"><i class="fa fa-check"></i><b>41.4</b> Number of rooted trees</a></li>
<li class="chapter" data-level="41.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-in-r-1"><i class="fa fa-check"></i><b>41.5</b> Number of rooted trees in R</a></li>
<li class="chapter" data-level="41.6" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-in-r-1"><i class="fa fa-check"></i><b>41.6</b> Functions in R</a></li>
<li class="chapter" data-level="41.7" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-to-calculate-the-number-of-possible-phylogenetic-trees"><i class="fa fa-check"></i><b>41.7</b> Function to calculate the number of possible phylogenetic trees</a></li>
<li class="chapter" data-level="41.8" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vectorized-inputs-to-functions"><i class="fa fa-check"></i><b>41.8</b> Vectorized inputs to functions</a></li>
<li class="chapter" data-level="41.9" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-conditional-statements"><i class="fa fa-check"></i><b>41.9</b> Adding conditional statements</a></li>
<li class="chapter" data-level="41.10" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-multiple-conditional-statements"><i class="fa fa-check"></i><b>41.10</b> Adding multiple conditional statements</a></li>
<li class="chapter" data-level="41.11" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-additional-arguments"><i class="fa fa-check"></i><b>41.11</b> Adding additional arguments</a></li>
<li class="chapter" data-level="41.12" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-_un_rooted-trees"><i class="fa fa-check"></i><b>41.12</b> Assignment: _un_rooted trees</a></li>
<li class="chapter" data-level="41.13" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-1"><i class="fa fa-check"></i><b>41.13</b> Assignment part 1</a>
<ul>
<li class="chapter" data-level="41.13.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#part-1-answer"><i class="fa fa-check"></i><b>41.13.1</b> Part 1 answer</a></li>
</ul></li>
<li class="chapter" data-level="41.14" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-2"><i class="fa fa-check"></i><b>41.14</b> Assignment part 2</a>
<ul>
<li class="chapter" data-level="41.14.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#answer"><i class="fa fa-check"></i><b>41.14.1</b> Answer</a></li>
</ul></li>
<li class="chapter" data-level="41.15" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#libraries"><i class="fa fa-check"></i><b>41.15</b> Libraries</a></li>
<li class="chapter" data-level="41.16" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-well-use"><i class="fa fa-check"></i><b>41.16</b> Function we’ll use</a></li>
<li class="chapter" data-level="41.17" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#data-matrix-1"><i class="fa fa-check"></i><b>41.17</b> Data matrix</a></li>
<li class="chapter" data-level="41.18" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#what-does-this-matrix-represent"><i class="fa fa-check"></i><b>41.18</b> What does this matrix represent?</a>
<ul>
<li class="chapter" data-level="41.18.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#genetic-distances-conceptualized-as-pairwise-linear-distances-1"><i class="fa fa-check"></i><b>41.18.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="41.18.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#genetic-distances-conceptualized-as-2d-linear-distances"><i class="fa fa-check"></i><b>41.18.2</b> Genetic distances conceptualized as 2D linear distances</a></li>
</ul></li>
<li class="chapter" data-level="41.19" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#from-linear-distances-to-tree-distances-1"><i class="fa fa-check"></i><b>41.19</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="41.20" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#the-upgma-algorithm-1"><i class="fa fa-check"></i><b>41.20</b> The UPGMA algorithm</a></li>
<li class="chapter" data-level="41.21" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#calculate-2d-representation-1"><i class="fa fa-check"></i><b>41.21</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="41.22" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#algorithm---round-1-1"><i class="fa fa-check"></i><b>41.22</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="41.22.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#determine-minimum-value-in-current-matrix-1"><i class="fa fa-check"></i><b>41.22.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="41.22.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#combine-species-into-clade-4"><i class="fa fa-check"></i><b>41.22.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="41.22.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#calcualte-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>41.22.3</b> Calcualte distance from clade to all other points</a></li>
<li class="chapter" data-level="41.22.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#update-matrix-with-new-distance-1"><i class="fa fa-check"></i><b>41.22.4</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="41.23" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#algorithm---round-2"><i class="fa fa-check"></i><b>41.23</b> Algorithm - round 2</a>
<ul>
<li class="chapter" data-level="41.23.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#update-the-matrix-1"><i class="fa fa-check"></i><b>41.23.1</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="41.24" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#iteration-3"><i class="fa fa-check"></i><b>41.24</b> Iteration 3</a>
<ul>
<li class="chapter" data-level="41.24.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#distance-of-c-to-abe-and-d-to-abe-1"><i class="fa fa-check"></i><b>41.24.1</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="41.25" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#data-matrix-2"><i class="fa fa-check"></i><b>41.25</b> Data matrix</a></li>
<li class="chapter" data-level="41.26" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#data-frame-to-store-stuff"><i class="fa fa-check"></i><b>41.26</b> Data frame to store stuff</a></li>
<li class="chapter" data-level="41.27" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#algorithm---first-step"><i class="fa fa-check"></i><b>41.27</b> Algorithm - First step</a>
<ul>
<li class="chapter" data-level="41.27.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#determine-minimum-value-in-current-matrix-2"><i class="fa fa-check"></i><b>41.27.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="41.27.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#combine-species-into-clade-5"><i class="fa fa-check"></i><b>41.27.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="41.27.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#first-branch-length-estimation"><i class="fa fa-check"></i><b>41.27.3</b> First branch length estimation</a></li>
<li class="chapter" data-level="41.27.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#first-distance-matrix-update---wpgma-version"><i class="fa fa-check"></i><b>41.27.4</b> First distance matrix update - WPGMA version</a></li>
<li class="chapter" data-level="41.27.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#update-matrix-with-new-distance-2"><i class="fa fa-check"></i><b>41.27.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="41.28" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#algorithm---second-step"><i class="fa fa-check"></i><b>41.28</b> Algorithm - Second step</a>
<ul>
<li class="chapter" data-level="41.28.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#combine-species-into-clade-6"><i class="fa fa-check"></i><b>41.28.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="41.28.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#calculate-branch-length-4"><i class="fa fa-check"></i><b>41.28.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="41.28.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#recalculate-distance-from-abe-to-all-remainign-taxa-1"><i class="fa fa-check"></i><b>41.28.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="41.28.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#update-the-matrix-2"><i class="fa fa-check"></i><b>41.28.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="41.29" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#third-step"><i class="fa fa-check"></i><b>41.29</b> Third step</a>
<ul>
<li class="chapter" data-level="41.29.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#calculate-branch-lengths-1"><i class="fa fa-check"></i><b>41.29.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="41.29.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#combine-species-into-clade-7"><i class="fa fa-check"></i><b>41.29.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="41.29.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#calculate-branch-length-5"><i class="fa fa-check"></i><b>41.29.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="41.29.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#distance-of-c-to-abe-and-d-to-abe-2"><i class="fa fa-check"></i><b>41.29.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="41.30" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#finish-up-1"><i class="fa fa-check"></i><b>41.30</b> Finish up</a>
<ul>
<li class="chapter" data-level="41.30.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#combine-species-into-clade-8"><i class="fa fa-check"></i><b>41.30.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="41.30.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#calculate-branch-length-6"><i class="fa fa-check"></i><b>41.30.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="41.31" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#plotting"><i class="fa fa-check"></i><b>41.31</b> Plotting</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="blast-alignment-score-distribution" class="section level1" number="35">
<h1><span class="header-section-number">Chapter 35</span> BLAST Alignment score distribution</h1>
<p><strong>By</strong>: Nathan Brouwer</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJsaWJyYXJ5KGNvbXBiaW80YWxsKSJ9
</div>
<!-- The following code creates a figure similar to Figure 4.14 in Chapter 4 of Pevsner (page 141). -->
<div id="packages-7" class="section level2" number="35.1">
<h2><span class="header-section-number">35.1</span> Packages</h2>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiIjaW5zdGFsbC5wYWNrYWdlcyhcIkhpc3REYXRhXCIpXG5saWJyYXJ5KEhpc3REYXRhKVxubGlicmFyeShnZ3B1YnIpXG5cbiNpbnN0YWxsLnBhY2thZ2VzKFwiZXZkXCIpXG5saWJyYXJ5KGV2ZCkifQ==
</div>
</div>
<div id="distributions-of-alignment-scores" class="section level2" number="35.2">
<h2><span class="header-section-number">35.2</span> Distributions of alignment scores</h2>
<p>Imagine we could keep track of all of the steps of a BLAST search. At a certain point, BLAST will have made many <strong>local alignments</strong> of a few dozen to a couple hundred bases between the <strong>query sequence</strong> you submitted and sequences in the database. Each of these local, pairwise alignments met all the criteria for being worthy of consideration and so are recored as <strong>hits</strong>. Now we want to judge objectively how good these hits are and which ones are most worthy of further consideration. We can do this by working of an <strong>E-value</strong>.</p>
<p>With our collection of hits, we could take their scores and plot them. Many things in biology when plotted will take on a bell-curve or normal distribution. Scores for alignments, however, take on an <strong>extreme value distribution</strong> (EVD). That is, compared to a <strong>symmetric</strong> normal distribution an EVD has too few low numbers (lower alignment scores) and extra high scores.</p>
<p>Now imagine that instead of BLASTing our sequence against a database of real sequence, we made up a bunch of <strong>random sequences</strong>. We could invent a random sequene by writing all the letters representing amino acids, tossing them into a hat, and pulling out an letter and writiing it down. We would then toss the letter back in, mix up the letters, and pull out another one. Repeat this 100 times and we’d have a perfectly random sequence the would be about the length of a protein, but wouldn’t code for anything (or be very very very very very unlikely to code for anything). If we invent, say 1000 sequence we could make a database of random sequences and would could aline our focal query sequence against our made up database. When could then score the local alignment between the query and each random sequence. Because these sequences are made up, the alignments will be pretty bad and teh scores low. However, occassionally there might be a fairly high score due to chance. So, just due to luck we could accidently invent a sequence that is a pretty good alignment with our query sequence. The question the <strong>E-values</strong> are trying to get at is how often this could happen, and how much better are our the scoures of our BLAST hits than the scores we’d get from BLASTing again random sequences.</p>
<p>The following code will illustrate these concepts.</p>
</div>
<div id="simulating-data" class="section level2" number="35.3">
<h2><span class="header-section-number">35.3</span> Simulating data</h2>
<p>The easiest way to think about simulating data is to roll a dice. Six-sided die allows you to draw numbers from a <strong>uniform distribution</strong>. With a uniform distribution, all values have the same likelihood of occuring. On a die, you have a 1/6 change of getting a 1, a 1/6 change of a 2, etc.</p>
<p>We can simulate a 6-sided die in R using the runif() function, which doesn’t mean “run-if” but rather “r-unif”, for random-uniform.</p>
<p>We can simulate rolling a die once with this code. n is the number of rolls. Min is the minimum possible value, max is the max possible value. Note that the min is set to 0.5 and the max to 6. This will be explained in a second.</p>
So, this code generates a number between 0.5 and 6.5
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJydW5pZihuID0gMSwgbWluID0gMC41LCBtYXggPSA2LjUpIn0=
</div>
This code generates 10 numbers from 0.5 to 6.5.
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJydW5pZihuID0gMTAsIG1pbiA9IDAuNSwgbWF4ID0gNi41KSJ9
</div>
runif() allows for non-integers, so we can make this like a real dice using round()
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJyb3VuZChydW5pZihuID0gMSwgbWluID0gMC41MDAwMDAwMCwgbWF4ID0gNi41MDAwMDAwMCkpIn0=
</div>
<p>If we want to simulate what would happen if 50 people in class all rolled a dice we set n to 50</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJyb3VuZChydW5pZihuID0gNTAsIG1pbiA9IDAuNSwgbWF4ID0gNi41KSkifQ==
</div>
<p>If we had 10000 people in our class:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJyb3VuZChydW5pZihuID0gMTAwMDAsIG1pbiA9IDAuNSwgbWF4ID0gNi41KSkifQ==
</div>
<p>We can save this and plot it:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiIjIHNhdmUgaXRcbnJhbmRvbS51bmlmb3JtIDwtIHJvdW5kKHJ1bmlmKG4gPSAxMDAwMCwgbWluID0gMC41LCBtYXggPSA2LjQ5OTk5OTkpKVxuXG4jIHBsb3QgaXRcbmhpc3QocmFuZG9tLnVuaWZvcm0pIn0=
</div>
<p>The height of the histogram is almost level, indicating that this is a uniform distribution.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiIjIHNhdmUgaXRcbnJhbmRvbS51bmlmb3JtIDwtIHJvdW5kKHJ1bmlmKG4gPSAxMDAwLCBtaW4gPSAwLjUsIG1heCA9IDYuNSkpXG5cbiMgcGxvdCBpdFxuaGlzdChyYW5kb20udW5pZm9ybSkifQ==
</div>
<p>The code above works if the probability of each number (1, 2, 3…) is the same for all values, as for a fair dice.</p>
<p>We can get a similar result more dirctly using the function sample(), which acts like a virtual “pulling numbers from a hat” or rolling a dice.</p>
First I make a vector to represent my die
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJkaWNlIDwtIGMoMSwyLDMsNCw1LDYpIn0=
</div>
I can roll it once like this
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJzYW1wbGUoZGljZSwgc2l6ZSA9IDEpIn0=
</div>
I can roll it twice like this; replace = T means that the same value can occur (its like pulling numbers from a hat and putting each number back in to make it available again).
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJzYW1wbGUoZGljZSwgc2l6ZSA9IDEsIHJlcGxhY2UgPSBUKSJ9
</div>
I can simulate 50 dice like this:
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJzYW1wbGUoZGljZSwgc2l6ZSA9IDUwLCByZXBsYWNlID0gVCkifQ==
</div>
<p>If I want to represent an unfair dice I can add a new arguement which allows me to define the probabilites.</p>
A fair dice would be this
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJwcm9icy5mYWlyIDwtIGMoMS82LCAxLzYsMS82LDEvNiwxLzYsMS82KVxuc2FtcGxlKGRpY2UsIHNpemUgPSAxLCByZXBsYWNlID0gVCwgIHByb2IgPSBwcm9icy5mYWlyKSJ9
</div>
A dice biased for 1could look like this
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJwcm9icy5iaWFzZWQgPC0gYygyLzYsIDEvNiwgMS82LCAxLzYsIDEvNiwgMS82KVxuc2FtcGxlKGRpY2UsIHNpemUgPSAxLCByZXBsYWNlID0gVCwgIHByb2IgPSBwcm9icy5iaWFzZWQpIn0=
</div>
<p>Let me see if its biased. I simulate 1000 dice</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJ4IDwtIHNhbXBsZShkaWNlLCBzaXplID0gMTAwMCwgcmVwbGFjZSA9IFQsICBwcm9iID0gcHJvYnMuYmlhc2VkKVxuaGlzdCh4KSJ9
</div>
</div>
<div id="lots-of-data-in-biology-is-normal" class="section level2" number="35.4">
<h2><span class="header-section-number">35.4</span> Lots of data in biology is normal</h2>
<p>Data from a dice are discrete: you can only have certain numbers (1, 2, 3…) and can’t have values in between (0.6). Also dice produce uniform striguation.</p>
<p>In contrast, the normal distribution is common in the world. Many things are normal or normal-ish. The follow code plots human heights (note: this doesn’t distinquish XY from XX but is still fairly normal)</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJkYXRhKEdhbHRvbilcbmxpYnJhcnkoZ2dwdWJyKVxuZ2doaXN0b2dyYW0oZGF0YSA9IEdhbHRvbiwgeCA9IFwiY2hpbGRcIikifQ==
</div>
The code below plots data on the size of flower petals by species. The middle panel is fairly normal. The left panel is normalish, but squished becuase the values are small and you can’t have flower sizes &lt;0.
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJkYXRhKGlyaXMpXG5nZ2hpc3RvZ3JhbShkYXRhID0gaXJpcywgeCA9IFwiUGV0YWwuTGVuZ3RoXCIsIGZhY2V0LmJ5ID0gXCJTcGVjaWVzXCIpIn0=
</div>
</div>
<div id="simulating-the-normal-distribution" class="section level2" number="35.5">
<h2><span class="header-section-number">35.5</span> Simulating the normal distribution</h2>
<p>We can simulate normal data in R using the rnorm function. Normal data is described by a mean (set here to 0) and a standard deviation (set here to 1). It has a typical bell-shaped curve. It is symmetrical - left of the mean has the same general shape as the data right of the mean. There is a mathematical formula that describes the shape of the curve, and you can generate random numbers from this. When you look at a normal distribution, the higher a given part of a curve is, the more likely a value under that part is to be randomly select. So values near the mean occur most commonly, while values far from the mean become increasinly less likely to occur.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJuIDwtIDEwMDBcbm1lYW4wIDwtIDBcbnNkMSA8LSAxXG5cbm5vcm1hbC5kYXRhIDwtIHJub3JtKG4gPSBuLCBcbiAgICAgICAgICAgICAgICAgICAgIG1lYW4gPSBtZWFuMCwgXG4gICAgICAgICAgICAgICAgICAgICBzZCA9IHNkMSkifQ==
</div>
<p>We can quickly plot it with hist()</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJoaXN0KG5vcm1hbC5kYXRhKVxuYWJsaW5lKHYgPSBtZWFuMCwgY29sID0gMikifQ==
</div>
</div>
<div id="the-extreme-value-distribution" class="section level2" number="35.6">
<h2><span class="header-section-number">35.6</span> The extreme value distribution</h2>
<p>Extreme value distributions are <strong>skewed</strong> and have a <strong>long tail</strong>.</p>
The evd package allows us to simulate extreme value data
s
BLAST theory uses a particular extreme value distribution, the Gumbel distribution. We can simulate data from this distribution with rgumbel()
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJleHRyZW1lLmRhdGEgPC0gcmd1bWJlbChuID0gbiwgbG9jPW1lYW4wLCBzY2FsZT0xKSJ9
</div>
We can plot the data like this
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJoaXN0KGV4dHJlbWUuZGF0YSlcbmFibGluZSh2ID0gbWVhbjAsIGNvbCA9IDIpIn0=
</div>
If you’ve taken a stats class you may have taken the log of a data set to make it more normal-like. If you do this with data that follows the extreme value distribution it doens’t help - the data still don’t look normal. Instead of being skewed one way, the skew is just flipped the other.
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJoaXN0KGxvZyhleHRyZW1lLmRhdGEpKVxuYWJsaW5lKHYgPSBtZWFuMCwgY29sID0gMikifQ==
</div>
</div>
<div id="replicating-figure-4.14" class="section level2" number="35.7">
<h2><span class="header-section-number">35.7</span> Replicating Figure 4.14</h2>
<p>This code evokes Pevsner Figrure 4.14. You don’t need to know these functions.</p>
Note that the main differences correspond to values around -2 and 2.
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJyYW5nZS54LmF4aXMgPC0gc2VxKGZyb20gPSAtNSx0byA9IDUsbGVuZ3RoLm91dCA9IDEwMDApXG5ub3JtLmN1cnZlICAgIDwtIGRub3JtKCAgeCA9IHJhbmdlLnguYXhpcywgbWVhbiA9IG1lYW4wLCBzZCA9IHNkMSlcbmV4dHJlbWUuY3VydmUgPC0gZGd1bWJlbCh4ID0gcmFuZ2UueC5heGlzLCBsb2MgPSBtZWFuMCwgc2NhbGUgPSAxKVxuXG5wbG90KG5vcm0uY3VydmV+cmFuZ2UueC5heGlzLHR5cGUgPSBcImxcIilcbnBvaW50cyhleHRyZW1lLmN1cnZlfnJhbmdlLnguYXhpcyx0eXBlID0gXCJsXCIsIGx0eSA9IDIsIGNvbCA9IDIpIn0=
</div>
</div>
<div id="why-do-we-care" class="section level2" number="35.8">
<h2><span class="header-section-number">35.8</span> Why do we care?</h2>
<p>The starting point for most statistics is the normal distribution. This is where we get z-scores, most p-values, standard errors, and most 95% confidence intervals. When we don’t use normal distributions, we use other common distributions (eg the binomial distribution for binary data, the poisson for count data). None of these common distributions work if we want to do any statistics with DNA alginment; we need tow rok with the extreme value distribution. (Indeed, in 10 years of spending lots of time doing statiscs, this is the only application I’ve run in to of the EVD)</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="getting-started-with-r-itself-or-not.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/BLAST-alignment_score_distribution.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/BLAST-alignment_score_distribution.Rmd",
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
