<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 53 Calculating E values, Bit scores, and P-values from BLAST output | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 53 Calculating E values, Bit scores, and P-values from BLAST output | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 53 Calculating E values, Bit scores, and P-values from BLAST output | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="blast-alignment-score-distribution.html"/>
<link rel="next" href="upgma-the-hard-way.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.5.1/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>1</b> How to review this book</a>
<ul>
<li class="chapter" data-level="1.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>1.1</b> File format</a></li>
<li class="chapter" data-level="1.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>1.2</b> What can you do?</a></li>
<li class="chapter" data-level="1.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>1.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>1.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="1.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>1.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="1.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>1.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="1.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>1.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="test-audio-link.html"><a href="test-audio-link.html"><i class="fa fa-check"></i><b>2</b> Test audio link</a></li>
<li class="chapter" data-level="3" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>3</b> Test shiny app</a></li>
<li class="chapter" data-level="4" data-path="test-youtube-vidoe.html"><a href="test-youtube-vidoe.html"><i class="fa fa-check"></i><b>4</b> Test YouTube vidoe</a></li>
<li class="chapter" data-level="5" data-path="test-google-doc-insertion.html"><a href="test-google-doc-insertion.html"><i class="fa fa-check"></i><b>5</b> Test google doc insertion</a></li>
<li class="chapter" data-level="6" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>6</b> Datacamp test</a></li>
<li class="chapter" data-level="7" data-path="an-example-of-a-plotly-graph.html"><a href="an-example-of-a-plotly-graph.html"><i class="fa fa-check"></i><b>7</b> An example of a plotly graph</a></li>
<li class="chapter" data-level="8" data-path="directly-embedding-of-a-libretext-page.html"><a href="directly-embedding-of-a-libretext-page.html"><i class="fa fa-check"></i><b>8</b> Directly embedding of a libretext page</a></li>
<li class="chapter" data-level="9" data-path="libretext-image.html"><a href="libretext-image.html"><i class="fa fa-check"></i><b>9</b> Libretext image</a>
<ul>
<li class="chapter" data-level="9.1" data-path="libretext-image.html"><a href="libretext-image.html#prelude-to-population-and-community-ecology"><i class="fa fa-check"></i><b>9.1</b> 45.0: Prelude to Population and Community Ecology</a></li>
<li class="chapter" data-level="9.2" data-path="libretext-image.html"><a href="libretext-image.html#note"><i class="fa fa-check"></i><b>9.2</b> Note</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="nucleic-acids.html"><a href="nucleic-acids.html"><i class="fa fa-check"></i><b>10</b> Nucleic Acids</a>
<ul>
<li class="chapter" data-level="10.1" data-path="nucleic-acids.html"><a href="nucleic-acids.html#dna-and-rna"><i class="fa fa-check"></i><b>10.1</b> DNA and RNA</a></li>
<li class="chapter" data-level="10.2" data-path="nucleic-acids.html"><a href="nucleic-acids.html#dna-double-helix-structure"><i class="fa fa-check"></i><b>10.2</b> DNA Double-Helix Structure</a></li>
<li class="chapter" data-level="10.3" data-path="nucleic-acids.html"><a href="nucleic-acids.html#rna"><i class="fa fa-check"></i><b>10.3</b> RNA</a></li>
<li class="chapter" data-level="10.4" data-path="nucleic-acids.html"><a href="nucleic-acids.html#summary"><i class="fa fa-check"></i><b>10.4</b> Summary</a></li>
<li class="chapter" data-level="10.5" data-path="nucleic-acids.html"><a href="nucleic-acids.html#analysis-questions"><i class="fa fa-check"></i><b>10.5</b> Analysis Questions</a></li>
<li class="chapter" data-level="10.6" data-path="nucleic-acids.html"><a href="nucleic-acids.html#glossary"><i class="fa fa-check"></i><b>10.6</b> Glossary</a></li>
<li class="chapter" data-level="10.7" data-path="nucleic-acids.html"><a href="nucleic-acids.html#contributors-and-attributions"><i class="fa fa-check"></i><b>10.7</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>11</b> Downloading R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>11.1</b> Preface</a></li>
<li class="chapter" data-level="11.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>11.2</b> Introduction to R</a></li>
<li class="chapter" data-level="11.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>11.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>11.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="11.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>11.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>11.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>12</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="12.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>12.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="12.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>12.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>13</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="13.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>13.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="13.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>13.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="13.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>13.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>14</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="14.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>14.1</b> Bioconductor</a></li>
<li class="chapter" data-level="14.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>14.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="14.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>14.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>14.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="14.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>14.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="14.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>14.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>14.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>15</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="15.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>15.1</b> Vocabulary</a></li>
<li class="chapter" data-level="15.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>15.2</b> R functions</a></li>
<li class="chapter" data-level="15.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>15.3</b> Interacting with R</a></li>
<li class="chapter" data-level="15.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>15.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="15.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>15.4.1</b> Vectors</a></li>
<li class="chapter" data-level="15.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>15.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="15.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>15.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="15.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>15.4.4</b> Lists</a></li>
<li class="chapter" data-level="15.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>15.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="15.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>15.5</b> Arguments</a></li>
<li class="chapter" data-level="15.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>15.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="15.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>15.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="15.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>15.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="15.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>15.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="15.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>15.9</b> Quiting R</a></li>
<li class="chapter" data-level="15.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>15.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>16</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="16.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-1"><i class="fa fa-check"></i><b>16.1</b> Preface</a></li>
<li class="chapter" data-level="16.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>16.2</b> Vocab</a></li>
<li class="chapter" data-level="16.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions"><i class="fa fa-check"></i><b>16.3</b> Functions</a></li>
<li class="chapter" data-level="16.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>16.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="16.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>16.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>17</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="17.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-2"><i class="fa fa-check"></i><b>17.1</b> Preface</a></li>
<li class="chapter" data-level="17.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>17.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="functions-1.html"><a href="functions-1.html"><i class="fa fa-check"></i><b>18</b> Functions</a>
<ul>
<li class="chapter" data-level="18.1" data-path="functions-1.html"><a href="functions-1.html#vectors-in-r"><i class="fa fa-check"></i><b>18.1</b> Vectors in R</a></li>
<li class="chapter" data-level="18.2" data-path="functions-1.html"><a href="functions-1.html#math-on-vectors"><i class="fa fa-check"></i><b>18.2</b> Math on vectors</a></li>
<li class="chapter" data-level="18.3" data-path="functions-1.html"><a href="functions-1.html#functions-on-vectors"><i class="fa fa-check"></i><b>18.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="18.4" data-path="functions-1.html"><a href="functions-1.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>18.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="18.5" data-path="functions-1.html"><a href="functions-1.html#subsetting-vectors"><i class="fa fa-check"></i><b>18.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="18.6" data-path="functions-1.html"><a href="functions-1.html#sequences-of-numbers"><i class="fa fa-check"></i><b>18.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="18.7" data-path="functions-1.html"><a href="functions-1.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>18.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="18.8" data-path="functions-1.html"><a href="functions-1.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>18.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>19</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="19.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-3"><i class="fa fa-check"></i><b>19.1</b> Preface</a></li>
<li class="chapter" data-level="19.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>19.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="19.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>19.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>20</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="20.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-4"><i class="fa fa-check"></i><b>20.1</b> Preface</a></li>
<li class="chapter" data-level="20.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>20.2</b> Vocab</a></li>
<li class="chapter" data-level="20.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-2"><i class="fa fa-check"></i><b>20.3</b> Functions</a></li>
<li class="chapter" data-level="20.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>20.4</b> Functions in R</a></li>
<li class="chapter" data-level="20.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>20.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html"><i class="fa fa-check"></i><b>21</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="21.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#commands-used"><i class="fa fa-check"></i><b>21.1</b> Commands used</a></li>
<li class="chapter" data-level="21.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#r-objects"><i class="fa fa-check"></i><b>21.2</b> R Objects</a></li>
<li class="chapter" data-level="21.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>21.3</b> Differences between objects</a></li>
<li class="chapter" data-level="21.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-data"><i class="fa fa-check"></i><b>21.4</b> The Data</a></li>
<li class="chapter" data-level="21.5" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>21.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="21.5.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#is"><i class="fa fa-check"></i><b>21.5.1</b> is()</a></li>
<li class="chapter" data-level="21.5.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#length"><i class="fa fa-check"></i><b>21.5.2</b> length()</a></li>
<li class="chapter" data-level="21.5.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#str"><i class="fa fa-check"></i><b>21.5.3</b> str()</a></li>
<li class="chapter" data-level="21.5.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#c"><i class="fa fa-check"></i><b>21.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="21.6" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#debrief"><i class="fa fa-check"></i><b>21.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="fasta-files.html"><a href="fasta-files.html"><i class="fa fa-check"></i><b>22</b> FASTA Files</a>
<ul>
<li class="chapter" data-level="22.1" data-path="fasta-files.html"><a href="fasta-files.html#example-fasta-file"><i class="fa fa-check"></i><b>22.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="22.2" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>22.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="22.3" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>22.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="22.4" data-path="fasta-files.html"><a href="fasta-files.html#fastq-format"><i class="fa fa-check"></i><b>22.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>23</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="23.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-3"><i class="fa fa-check"></i><b>23.0.1</b> Functions</a></li>
<li class="chapter" data-level="23.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>23.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="23.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>23.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="23.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>23.0.4</b> File types</a></li>
<li class="chapter" data-level="23.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>23.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="23.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>23.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="23.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries"><i class="fa fa-check"></i><b>23.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="23.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>23.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="23.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>23.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="23.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>23.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="23.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>23.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="23.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>23.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="23.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>23.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="23.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>23.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="23.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>23.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>24</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="24.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-1"><i class="fa fa-check"></i><b>24.1</b> Preliminaries</a></li>
<li class="chapter" data-level="24.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>24.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="24.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>24.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="24.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>24.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="25" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>25</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="25.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-5"><i class="fa fa-check"></i><b>25.1</b> Preface</a></li>
<li class="chapter" data-level="25.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>25.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="25.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>25.3</b> Introduction</a></li>
<li class="chapter" data-level="25.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>25.4</b> Vocabulary</a></li>
<li class="chapter" data-level="25.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions-4"><i class="fa fa-check"></i><b>25.5</b> Functions</a></li>
<li class="chapter" data-level="25.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries-2"><i class="fa fa-check"></i><b>25.6</b> Preliminaries</a></li>
<li class="chapter" data-level="25.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>25.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="25.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>25.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="25.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>25.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="25.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>25.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="25.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>25.8.3</b> DNA words</a></li>
<li class="chapter" data-level="25.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary-1"><i class="fa fa-check"></i><b>25.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="25.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>25.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="25.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>25.9.1</b> License</a></li>
<li class="chapter" data-level="25.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>25.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="26" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>26</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="26.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>26.1</b> Preliminaries</a></li>
<li class="chapter" data-level="26.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>26.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="changes-to-database-entries.html"><a href="changes-to-database-entries.html"><i class="fa fa-check"></i><b>27</b> Changes to database entries</a></li>
<li class="chapter" data-level="28" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>28</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="28.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>28.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="28.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>28.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="28.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>28.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="28.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>28.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="28.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>28.4</b> Inversions</a></li>
<li class="chapter" data-level="28.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>28.5</b> Translocations</a></li>
<li class="chapter" data-level="28.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>28.6</b> Random sequence</a></li>
<li class="chapter" data-level="28.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>28.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>29</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="29.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>29.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="29.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>29.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="29.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>29.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="29.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>29.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="29.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>29.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="29.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>29.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="29.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>29.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="29.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>29.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>30</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="30.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>30.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="30.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>30.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="30.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>30.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="31" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html"><i class="fa fa-check"></i><b>31</b> Alignment by eye in R</a>
<ul>
<li class="chapter" data-level="31.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>31.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="31.1.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#packages"><i class="fa fa-check"></i><b>31.1.1</b> Packages</a></li>
<li class="chapter" data-level="31.1.2" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#data"><i class="fa fa-check"></i><b>31.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="32" data-path="alignment-in-r.html"><a href="alignment-in-r.html"><i class="fa fa-check"></i><b>32</b> Alignment in R</a>
<ul>
<li class="chapter" data-level="32.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#preliminaries-8"><i class="fa fa-check"></i><b>32.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="32.1.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#packages-1"><i class="fa fa-check"></i><b>32.1.1</b> Packages</a></li>
<li class="chapter" data-level="32.1.2" data-path="alignment-in-r.html"><a href="alignment-in-r.html#data-1"><i class="fa fa-check"></i><b>32.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="33" data-path="working-with-vectors-and-matrices.html"><a href="working-with-vectors-and-matrices.html"><i class="fa fa-check"></i><b>33</b> Working with vectors and matrices</a></li>
<li class="chapter" data-level="34" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html"><i class="fa fa-check"></i><b>34</b> Matrix vocabulary</a>
<ul>
<li class="chapter" data-level="34.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#introduction-1"><i class="fa fa-check"></i><b>34.1</b> Introduction</a></li>
<li class="chapter" data-level="34.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#preliminaries-9"><i class="fa fa-check"></i><b>34.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="34.2.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#packages-2"><i class="fa fa-check"></i><b>34.2.1</b> Packages</a></li>
<li class="chapter" data-level="34.2.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#vocab-3"><i class="fa fa-check"></i><b>34.2.2</b> Vocab</a></li>
<li class="chapter" data-level="34.2.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#functions-used"><i class="fa fa-check"></i><b>34.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="34.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#load-packages"><i class="fa fa-check"></i><b>34.3</b> Load packages</a></li>
<li class="chapter" data-level="34.4" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#matrix-structure"><i class="fa fa-check"></i><b>34.4</b> Matrix structure</a></li>
<li class="chapter" data-level="34.5" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#diagonal-of-a-matrix"><i class="fa fa-check"></i><b>34.5</b> Diagonal of a matrix</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html"><i class="fa fa-check"></i><b>35</b> Testing the significance of an alignment</a>
<ul>
<li class="chapter" data-level="35.1" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#calculating-the-statistical-significance-of-a-pairwise-global-alignment"><i class="fa fa-check"></i><b>35.1</b> Calculating the statistical significance of a pairwise global alignment</a></li>
<li class="chapter" data-level="35.2" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#summary-2"><i class="fa fa-check"></i><b>35.2</b> Summary</a></li>
<li class="chapter" data-level="35.3" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#links-and-further-reading-1"><i class="fa fa-check"></i><b>35.3</b> Links and Further Reading</a></li>
<li class="chapter" data-level="35.4" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#exercises-1"><i class="fa fa-check"></i><b>35.4</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>36</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="36.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>36.1</b> Prelminaries</a></li>
<li class="chapter" data-level="36.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>36.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="36.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>36.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html"><i class="fa fa-check"></i><b>37</b> Manipulating matrices and vectors: worked example</a>
<ul>
<li class="chapter" data-level="37.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#introduction-2"><i class="fa fa-check"></i><b>37.1</b> Introduction</a></li>
<li class="chapter" data-level="37.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#preliminaries-10"><i class="fa fa-check"></i><b>37.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="37.2.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-3"><i class="fa fa-check"></i><b>37.2.1</b> Packages</a></li>
<li class="chapter" data-level="37.2.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#vocab-4"><i class="fa fa-check"></i><b>37.2.2</b> Vocab</a></li>
<li class="chapter" data-level="37.2.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#functions-used-1"><i class="fa fa-check"></i><b>37.2.3</b> Functions used</a></li>
<li class="chapter" data-level="37.2.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-4"><i class="fa fa-check"></i><b>37.2.4</b> Packages</a></li>
<li class="chapter" data-level="37.2.5" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#data-2"><i class="fa fa-check"></i><b>37.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="37.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#matrix-elements-can-be-accessed-with-square-bracket-notation"><i class="fa fa-check"></i><b>37.3</b> Matrix elements can be accessed with square bracket notation</a></li>
<li class="chapter" data-level="37.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#r-object-names"><i class="fa fa-check"></i><b>37.4</b> R object names</a>
<ul>
<li class="chapter" data-level="37.4.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-vectors"><i class="fa fa-check"></i><b>37.4.1</b> Accessing items in names vectors</a></li>
<li class="chapter" data-level="37.4.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-matrices"><i class="fa fa-check"></i><b>37.4.2</b> Accessing items in names matrices</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="38" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>38</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="38.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-11"><i class="fa fa-check"></i><b>38.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="38.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages-5"><i class="fa fa-check"></i><b>38.1.1</b> Packages</a></li>
<li class="chapter" data-level="38.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-5"><i class="fa fa-check"></i><b>38.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="38.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>38.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="38.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>38.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="38.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>38.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="38.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>38.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="38.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>38.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="38.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>38.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="38.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>38.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="38.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>38.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html"><i class="fa fa-check"></i><b>39</b> The BLOSUM scoring matrix in R</a>
<ul>
<li class="chapter" data-level="39.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#introduction-3"><i class="fa fa-check"></i><b>39.1</b> Introduction</a></li>
<li class="chapter" data-level="39.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#preliminaries-12"><i class="fa fa-check"></i><b>39.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="39.2.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#packages-6"><i class="fa fa-check"></i><b>39.2.1</b> Packages</a></li>
<li class="chapter" data-level="39.2.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#vocab-5"><i class="fa fa-check"></i><b>39.2.2</b> Vocab</a></li>
<li class="chapter" data-level="39.2.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#functions-used-2"><i class="fa fa-check"></i><b>39.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="39.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#load-packages-1"><i class="fa fa-check"></i><b>39.3</b> Load packages</a></li>
<li class="chapter" data-level="39.4" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#the-blosum-matrix"><i class="fa fa-check"></i><b>39.4</b> The BLOSUM matrix</a></li>
<li class="chapter" data-level="39.5" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#visualizing-the-blosum-matrix"><i class="fa fa-check"></i><b>39.5</b> Visualizing the BLOSUM matrix</a></li>
</ul></li>
<li class="chapter" data-level="40" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>40</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="41" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>41</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="41.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-4"><i class="fa fa-check"></i><b>41.1</b> Introduction</a></li>
<li class="chapter" data-level="41.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>41.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="41.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-6"><i class="fa fa-check"></i><b>41.2.1</b> Vocab</a></li>
<li class="chapter" data-level="41.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>41.2.2</b> R functions</a></li>
<li class="chapter" data-level="41.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>41.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="41.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>41.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="41.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>41.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="41.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>41.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="41.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>41.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="41.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>41.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="41.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>41.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="41.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>41.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="41.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>41.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="41.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>41.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="41.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>41.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="41.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>41.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="41.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>41.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="42" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>42</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="42.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-13"><i class="fa fa-check"></i><b>42.1</b> Preliminaries</a></li>
<li class="chapter" data-level="42.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-5"><i class="fa fa-check"></i><b>42.2</b> Introduction</a></li>
<li class="chapter" data-level="42.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>42.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="42.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>42.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="43" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>43</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="43.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-14"><i class="fa fa-check"></i><b>43.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="43.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>43.1.1</b> Key functions</a></li>
<li class="chapter" data-level="43.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>43.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="43.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>43.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="43.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>43.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="43.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>43.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="43.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>43.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="43.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>43.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="44" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html"><i class="fa fa-check"></i><b>44</b> Local variation in GC content</a>
<ul>
<li class="chapter" data-level="44.0.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#note-on-the-biology-in-this-section"><i class="fa fa-check"></i><b>44.0.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="44.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#vocabulary-2"><i class="fa fa-check"></i><b>44.1</b> Vocabulary</a></li>
<li class="chapter" data-level="44.2" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#reading-sequence-data-with-rentrezentrez_fetch"><i class="fa fa-check"></i><b>44.2</b> Reading sequence data with rentrez::entrez_fetch</a></li>
<li class="chapter" data-level="44.3" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#local-variation-in-gc-content-1"><i class="fa fa-check"></i><b>44.3</b> Local variation in GC content</a></li>
<li class="chapter" data-level="44.4" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-analysis-of-gc-content"><i class="fa fa-check"></i><b>44.4</b> A sliding window analysis of GC content</a></li>
<li class="chapter" data-level="44.5" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-plot-of-gc-content"><i class="fa fa-check"></i><b>44.5</b> A sliding window plot of GC content</a>
<ul>
<li class="chapter" data-level="44.5.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#over-represented-and-under-represented-dna-words"><i class="fa fa-check"></i><b>44.5.1</b> Over-represented and under-represented DNA words</a></li>
</ul></li>
<li class="chapter" data-level="44.6" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#summary-3"><i class="fa fa-check"></i><b>44.6</b> Summary</a></li>
<li class="chapter" data-level="44.7" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#further-reading"><i class="fa fa-check"></i><b>44.7</b> Further Reading</a></li>
<li class="chapter" data-level="44.8" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#acknowledgements-1"><i class="fa fa-check"></i><b>44.8</b> Acknowledgements</a></li>
<li class="chapter" data-level="44.9" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#exercises-2"><i class="fa fa-check"></i><b>44.9</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="45" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html"><i class="fa fa-check"></i><b>45</b> Computational gene finding</a>
<ul>
<li class="chapter" data-level="45.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#preliminaries-15"><i class="fa fa-check"></i><b>45.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="45.1.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#packages-7"><i class="fa fa-check"></i><b>45.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="45.2" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#the-genetic-code"><i class="fa fa-check"></i><b>45.2</b> The genetic code</a></li>
<li class="chapter" data-level="45.3" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-start-and-stop-codons-in-a-dna-sequence"><i class="fa fa-check"></i><b>45.3</b> Finding start and stop codons in a DNA sequence</a></li>
<li class="chapter" data-level="45.4" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#reading-frames"><i class="fa fa-check"></i><b>45.4</b> Reading frames</a></li>
<li class="chapter" data-level="45.5" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-forward-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>45.5</b> Finding open reading frames on the forward strand of a DNA sequence</a></li>
<li class="chapter" data-level="45.6" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#predicting-the-protein-sequence-for-an-orf"><i class="fa fa-check"></i><b>45.6</b> Predicting the protein sequence for an ORF</a></li>
<li class="chapter" data-level="45.7" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-reverse-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>45.7</b> Finding open reading frames on the reverse strand of a DNA sequence</a></li>
<li class="chapter" data-level="45.8" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#lengths-of-open-reading-frames"><i class="fa fa-check"></i><b>45.8</b> Lengths of open reading frames</a></li>
</ul></li>
<li class="chapter" data-level="46" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html"><i class="fa fa-check"></i><b>46</b> Significance of ORFs</a>
<ul>
<li class="chapter" data-level="46.1" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#preliminaries-16"><i class="fa fa-check"></i><b>46.1</b> Preliminaries</a></li>
<li class="chapter" data-level="46.2" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#identifying-significant-open-reading-frames"><i class="fa fa-check"></i><b>46.2</b> Identifying significant open reading frames</a></li>
<li class="chapter" data-level="46.3" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#summary-4"><i class="fa fa-check"></i><b>46.3</b> Summary</a></li>
<li class="chapter" data-level="46.4" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#links-and-further-reading-2"><i class="fa fa-check"></i><b>46.4</b> Links and Further Reading</a></li>
<li class="chapter" data-level="46.5" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#acknowledgements-2"><i class="fa fa-check"></i><b>46.5</b> Acknowledgements</a></li>
<li class="chapter" data-level="46.6" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#exercises-3"><i class="fa fa-check"></i><b>46.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="47" data-path="comparative-genomics.html"><a href="comparative-genomics.html"><i class="fa fa-check"></i><b>47</b> Comparative Genomics</a>
<ul>
<li class="chapter" data-level="47.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#preliminaries-17"><i class="fa fa-check"></i><b>47.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="47.1.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#packages-8"><i class="fa fa-check"></i><b>47.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="47.2" data-path="comparative-genomics.html"><a href="comparative-genomics.html#introduction-6"><i class="fa fa-check"></i><b>47.2</b> Introduction</a></li>
<li class="chapter" data-level="47.3" data-path="comparative-genomics.html"><a href="comparative-genomics.html#using-the-biomart-r-library-to-query-the-ensembl-database"><i class="fa fa-check"></i><b>47.3</b> Using the biomaRt R Library to Query the Ensembl Database</a></li>
<li class="chapter" data-level="47.4" data-path="comparative-genomics.html"><a href="comparative-genomics.html#comparing-the-number-of-genes-in-two-species"><i class="fa fa-check"></i><b>47.4</b> Comparing the number of genes in two species</a></li>
<li class="chapter" data-level="47.5" data-path="comparative-genomics.html"><a href="comparative-genomics.html#identifying-homologous-genes-between-two-species"><i class="fa fa-check"></i><b>47.5</b> Identifying homologous genes between two species</a></li>
</ul></li>
<li class="chapter" data-level="48" data-path="summary-5.html"><a href="summary-5.html"><i class="fa fa-check"></i><b>48</b> Summary</a></li>
<li class="chapter" data-level="49" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html"><i class="fa fa-check"></i><b>49</b> #Links and Further Reading</a>
<ul>
<li class="chapter" data-level="49.1" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#acknowledgements-3"><i class="fa fa-check"></i><b>49.1</b> Acknowledgements</a></li>
<li class="chapter" data-level="49.2" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#exercises-4"><i class="fa fa-check"></i><b>49.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="50" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html"><i class="fa fa-check"></i><b>50</b> Hidden Markov Models</a>
<ul>
<li class="chapter" data-level="50.1" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-multinomial-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>50.1</b> A multinomial model of DNA sequence evolution</a></li>
<li class="chapter" data-level="50.2" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-multinomial-model"><i class="fa fa-check"></i><b>50.2</b> Generating a DNA sequence using a multinomial model</a></li>
<li class="chapter" data-level="50.3" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>50.3</b> A Markov model of DNA sequence evolution</a></li>
<li class="chapter" data-level="50.4" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-for-a-markov-model"><i class="fa fa-check"></i><b>50.4</b> The transition matrix for a Markov model</a></li>
<li class="chapter" data-level="50.5" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-markov-model"><i class="fa fa-check"></i><b>50.5</b> Generating a DNA sequence using a Markov model</a></li>
<li class="chapter" data-level="50.6" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>50.6</b> A Hidden Markov Model of DNA sequence evolution</a></li>
<li class="chapter" data-level="50.7" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-and-emission-matrix-for-a-hmm"><i class="fa fa-check"></i><b>50.7</b> The transition matrix and emission matrix for a HMM</a></li>
<li class="chapter" data-level="50.8" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-hmm"><i class="fa fa-check"></i><b>50.8</b> Generating a DNA sequence using a HMM</a></li>
<li class="chapter" data-level="50.9" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#inferring-the-states-of-a-hmm-that-generated-a-dna-sequence"><i class="fa fa-check"></i><b>50.9</b> Inferring the states of a HMM that generated a DNA sequence</a></li>
<li class="chapter" data-level="50.10" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-protein-sequence-evolution"><i class="fa fa-check"></i><b>50.10</b> A Hidden Markov Model of protein sequence evolution</a></li>
<li class="chapter" data-level="50.11" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#summary-6"><i class="fa fa-check"></i><b>50.11</b> Summary</a></li>
<li class="chapter" data-level="50.12" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#links-and-further-reading-4"><i class="fa fa-check"></i><b>50.12</b> Links and Further Reading</a></li>
<li class="chapter" data-level="50.13" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#exercises-5"><i class="fa fa-check"></i><b>50.13</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="51" data-path="blast-summary.html"><a href="blast-summary.html"><i class="fa fa-check"></i><b>51</b> BLAST Summary</a>
<ul>
<li class="chapter" data-level="51.1" data-path="blast-summary.html"><a href="blast-summary.html#blast-overview"><i class="fa fa-check"></i><b>51.1</b> BLAST Overview</a>
<ul>
<li class="chapter" data-level="51.1.1" data-path="blast-summary.html"><a href="blast-summary.html#why-do-a-blast-search"><i class="fa fa-check"></i><b>51.1.1</b> Why do a BLAST search?</a></li>
<li class="chapter" data-level="51.1.2" data-path="blast-summary.html"><a href="blast-summary.html#what-does-blast-let-you-do"><i class="fa fa-check"></i><b>51.1.2</b> What does BLAST let you do?</a></li>
</ul></li>
<li class="chapter" data-level="51.2" data-path="blast-summary.html"><a href="blast-summary.html#blast-recipe"><i class="fa fa-check"></i><b>51.2</b> BLAST recipe</a></li>
<li class="chapter" data-level="51.3" data-path="blast-summary.html"><a href="blast-summary.html#modules-in-the-blast-workflow-and-algorithm"><i class="fa fa-check"></i><b>51.3</b> Modules in the BLAST Workflow and Algorithm</a>
<ul>
<li class="chapter" data-level="51.3.1" data-path="blast-summary.html"><a href="blast-summary.html#module-1-setup-query-sequence-protein-blast-oriented"><i class="fa fa-check"></i><b>51.3.1</b> Module 1: Setup query sequence (Protein BLAST oriented)</a></li>
<li class="chapter" data-level="51.3.2" data-path="blast-summary.html"><a href="blast-summary.html#module-2-scanning-database-for-high-scoring-words"><i class="fa fa-check"></i><b>51.3.2</b> Module 2: Scanning database for high-scoring words</a></li>
<li class="chapter" data-level="51.3.3" data-path="blast-summary.html"><a href="blast-summary.html#module-3-traceback"><i class="fa fa-check"></i><b>51.3.3</b> Module 3: Traceback</a></li>
</ul></li>
<li class="chapter" data-level="51.4" data-path="blast-summary.html"><a href="blast-summary.html#blast-alignment-statistics"><i class="fa fa-check"></i><b>51.4</b> BLAST alignment statistics</a></li>
</ul></li>
<li class="chapter" data-level="52" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html"><i class="fa fa-check"></i><b>52</b> BLAST Alignment score distribution</a>
<ul>
<li class="chapter" data-level="52.1" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#packages-9"><i class="fa fa-check"></i><b>52.1</b> Packages</a></li>
<li class="chapter" data-level="52.2" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#distributions-of-alignment-scores"><i class="fa fa-check"></i><b>52.2</b> Distributions of alignment scores</a></li>
<li class="chapter" data-level="52.3" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-data"><i class="fa fa-check"></i><b>52.3</b> Simulating data</a></li>
<li class="chapter" data-level="52.4" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#lots-of-data-in-biology-is-normal"><i class="fa fa-check"></i><b>52.4</b> Lots of data in biology is normal</a></li>
<li class="chapter" data-level="52.5" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-the-normal-distribution"><i class="fa fa-check"></i><b>52.5</b> Simulating the normal distribution</a></li>
<li class="chapter" data-level="52.6" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#the-extreme-value-distribution"><i class="fa fa-check"></i><b>52.6</b> The extreme value distribution</a></li>
<li class="chapter" data-level="52.7" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#replicating-figure-4.14"><i class="fa fa-check"></i><b>52.7</b> Replicating Figure 4.14</a></li>
<li class="chapter" data-level="52.8" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#why-do-we-care"><i class="fa fa-check"></i><b>52.8</b> Why do we care?</a></li>
</ul></li>
<li class="chapter" data-level="53" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><i class="fa fa-check"></i><b>53</b> Calculating E values, Bit scores, and P-values from BLAST output</a>
<ul>
<li class="chapter" data-level="53.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-7"><i class="fa fa-check"></i><b>53.1</b> Introduction</a></li>
<li class="chapter" data-level="53.2" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#case-study"><i class="fa fa-check"></i><b>53.2</b> Case Study</a></li>
<li class="chapter" data-level="53.3" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-statistics"><i class="fa fa-check"></i><b>53.3</b> BLAST Statistics</a></li>
<li class="chapter" data-level="53.4" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot"><i class="fa fa-check"></i><b>53.4</b> Dotplot</a>
<ul>
<li class="chapter" data-level="53.4.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot-in-r---optional-experimental-code"><i class="fa fa-check"></i><b>53.4.1</b> Dotplot in R - optional, experimental code</a></li>
</ul></li>
<li class="chapter" data-level="53.5" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-search-summary"><i class="fa fa-check"></i><b>53.5</b> BLAST “Search Summary”</a></li>
<li class="chapter" data-level="53.6" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-alignments"><i class="fa fa-check"></i><b>53.6</b> BLAST Alignments</a></li>
<li class="chapter" data-level="53.7" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#equations"><i class="fa fa-check"></i><b>53.7</b> Equations</a>
<ul>
<li class="chapter" data-level="53.7.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-value-eq-4.5-pg-142"><i class="fa fa-check"></i><b>53.7.1</b> E value (eq 4.5, pg 142)</a></li>
</ul></li>
<li class="chapter" data-level="53.8" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bit-scores-s"><i class="fa fa-check"></i><b>53.8</b> Bit Scores (S’)</a></li>
<li class="chapter" data-level="53.9" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-from-bit-scores"><i class="fa fa-check"></i><b>53.9</b> E from Bit scores</a></li>
<li class="chapter" data-level="53.10" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#p-values"><i class="fa fa-check"></i><b>53.10</b> P values</a></li>
<li class="chapter" data-level="53.11" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#repliminaries"><i class="fa fa-check"></i><b>53.11</b> Repliminaries</a></li>
<li class="chapter" data-level="53.12" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#concepts"><i class="fa fa-check"></i><b>53.12</b> Concepts</a></li>
<li class="chapter" data-level="53.13" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#functions-6"><i class="fa fa-check"></i><b>53.13</b> Functions</a></li>
<li class="chapter" data-level="53.14" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-8"><i class="fa fa-check"></i><b>53.14</b> Introduction</a></li>
<li class="chapter" data-level="53.15" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees"><i class="fa fa-check"></i><b>53.15</b> Number of rooted trees</a></li>
<li class="chapter" data-level="53.16" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees-in-r"><i class="fa fa-check"></i><b>53.16</b> Number of rooted trees in R</a>
<ul>
<li class="chapter" data-level="53.16.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#unique-topologies-versus-rotations"><i class="fa fa-check"></i><b>53.16.1</b> Unique topologies versus rotations</a></li>
</ul></li>
<li class="chapter" data-level="53.17" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bilography"><i class="fa fa-check"></i><b>53.17</b> Bilography</a></li>
</ul></li>
<li class="chapter" data-level="54" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html"><i class="fa fa-check"></i><b>54</b> UPGMA the hard way</a>
<ul>
<li class="chapter" data-level="54.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#introudction-to-upgma"><i class="fa fa-check"></i><b>54.1</b> Introudction to UPGMA</a></li>
<li class="chapter" data-level="54.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#preliminaries-18"><i class="fa fa-check"></i><b>54.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="54.2.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#r-libraries"><i class="fa fa-check"></i><b>54.2.1</b> R Libraries</a></li>
<li class="chapter" data-level="54.2.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#custom-function-well-use"><i class="fa fa-check"></i><b>54.2.2</b> Custom function we’ll use</a></li>
</ul></li>
<li class="chapter" data-level="54.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#data-matrix"><i class="fa fa-check"></i><b>54.3</b> Data matrix</a>
<ul>
<li class="chapter" data-level="54.3.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#building-the-matrix"><i class="fa fa-check"></i><b>54.3.1</b> Building the matrix</a></li>
</ul></li>
<li class="chapter" data-level="54.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#what-does-this-distance-matrix-represent"><i class="fa fa-check"></i><b>54.4</b> What does this distance matrix represent?</a>
<ul>
<li class="chapter" data-level="54.4.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-pairwise-linear-distances"><i class="fa fa-check"></i><b>54.4.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="54.4.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-2d-map-of-linear-distances"><i class="fa fa-check"></i><b>54.4.2</b> Genetic distances conceptualized as 2D map of linear distances</a></li>
</ul></li>
<li class="chapter" data-level="54.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#from-linear-distances-to-tree-distances"><i class="fa fa-check"></i><b>54.5</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="54.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#the-upgma-algorithm"><i class="fa fa-check"></i><b>54.6</b> The UPGMA algorithm</a>
<ul>
<li class="chapter" data-level="54.6.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-1-find-the-2-taxa-that-are-currently-closest-together"><i class="fa fa-check"></i><b>54.6.1</b> Step 1: find the 2 taxa that are currently closest together</a></li>
<li class="chapter" data-level="54.6.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node"><i class="fa fa-check"></i><b>54.6.2</b> Step 2: calculate the branch lengths connect the taxa to their shared node</a></li>
<li class="chapter" data-level="54.6.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-determine-if-the-algorithm-is-done"><i class="fa fa-check"></i><b>54.6.3</b> Step 3a: determine if the algorithm is done</a></li>
<li class="chapter" data-level="54.6.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u"><i class="fa fa-check"></i><b>54.6.4</b> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</a></li>
<li class="chapter" data-level="54.6.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-4-determine-the-distance-from-u-to-the-other-remaining-clusters."><i class="fa fa-check"></i><b>54.6.5</b> Step 4: Determine the distance from u to the other remaining clusters.</a></li>
<li class="chapter" data-level="54.6.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-5-create-a-new-matrix"><i class="fa fa-check"></i><b>54.6.6</b> Step 5: Create a new matrix</a></li>
<li class="chapter" data-level="54.6.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa"><i class="fa fa-check"></i><b>54.6.7</b> Key step: calculating the distance from a clade to the remaining taxa</a></li>
</ul></li>
<li class="chapter" data-level="54.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-2d-representation"><i class="fa fa-check"></i><b>54.7</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="54.8" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---round-1"><i class="fa fa-check"></i><b>54.8</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="54.8.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#determine-minimum-value-in-current-matrix"><i class="fa fa-check"></i><b>54.8.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="54.8.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade"><i class="fa fa-check"></i><b>54.8.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.8.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length"><i class="fa fa-check"></i><b>54.8.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="54.8.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>54.8.4</b> Calculate distance from clade to all other points</a></li>
<li class="chapter" data-level="54.8.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-matrix-with-new-distance"><i class="fa fa-check"></i><b>54.8.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="54.9" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---next-round"><i class="fa fa-check"></i><b>54.9</b> Algorithm - next round</a>
<ul>
<li class="chapter" data-level="54.9.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-1"><i class="fa fa-check"></i><b>54.9.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.9.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-1"><i class="fa fa-check"></i><b>54.9.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="54.9.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#recalculate-distance-from-abe-to-all-remainign-taxa"><i class="fa fa-check"></i><b>54.9.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="54.9.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-the-matrix"><i class="fa fa-check"></i><b>54.9.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="54.10" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#next-iteration"><i class="fa fa-check"></i><b>54.10</b> Next iteration</a>
<ul>
<li class="chapter" data-level="54.10.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-lengths"><i class="fa fa-check"></i><b>54.10.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="54.10.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-2"><i class="fa fa-check"></i><b>54.10.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.10.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-2"><i class="fa fa-check"></i><b>54.10.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="54.10.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#distance-of-c-to-abe-and-d-to-abe"><i class="fa fa-check"></i><b>54.10.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="54.11" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finish-up"><i class="fa fa-check"></i><b>54.11</b> Finish up</a>
<ul>
<li class="chapter" data-level="54.11.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-3"><i class="fa fa-check"></i><b>54.11.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.11.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-3"><i class="fa fa-check"></i><b>54.11.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="54.12" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finalizing-branch-lengths"><i class="fa fa-check"></i><b>54.12</b> Finalizing branch lengths</a></li>
</ul></li>
<li class="chapter" data-level="55" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html"><i class="fa fa-check"></i><b>55</b> Representing phylogenetic trees in Newick format</a>
<ul>
<li class="chapter" data-level="55.1" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#vocab-7"><i class="fa fa-check"></i><b>55.1</b> Vocab</a></li>
<li class="chapter" data-level="55.2" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#introduction-9"><i class="fa fa-check"></i><b>55.2</b> Introduction</a></li>
<li class="chapter" data-level="55.3" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-three-taxa"><i class="fa fa-check"></i><b>55.3</b> Trees with three taxa</a></li>
<li class="chapter" data-level="55.4" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-four-taxa"><i class="fa fa-check"></i><b>55.4</b> Trees with four taxa</a></li>
</ul></li>
<li class="chapter" data-level="56" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html"><i class="fa fa-check"></i><b>56</b> Plotting trees from Newick notation in R</a>
<ul>
<li class="chapter" data-level="56.1" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#preliminaries-19"><i class="fa fa-check"></i><b>56.1</b> Preliminaries</a></li>
<li class="chapter" data-level="56.2" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#note-1"><i class="fa fa-check"></i><b>56.2</b> Note</a></li>
<li class="chapter" data-level="56.3" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick"><i class="fa fa-check"></i><b>56.3</b> 3 taxa tree in Newick</a></li>
<li class="chapter" data-level="56.4" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick-1"><i class="fa fa-check"></i><b>56.4</b> 4 taxa tree in Newick</a></li>
<li class="chapter" data-level="56.5" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-in-newick-format"><i class="fa fa-check"></i><b>56.5</b> 5 taxa in Newick Format</a></li>
<li class="chapter" data-level="56.6" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#many-taxa-tree"><i class="fa fa-check"></i><b>56.6</b> Many-taxa tree</a></li>
<li class="chapter" data-level="56.7" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#newick-and-branch-lengths"><i class="fa fa-check"></i><b>56.7</b> Newick and Branch lengths</a></li>
<li class="chapter" data-level="56.8" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#additional-reading"><i class="fa fa-check"></i><b>56.8</b> Additional Reading</a></li>
</ul></li>
<li class="chapter" data-level="57" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><i class="fa fa-check"></i><b>57</b> Phylogenetic trees using the UPGMA clustering algoirthm</a>
<ul>
<li class="chapter" data-level="57.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#introduction-10"><i class="fa fa-check"></i><b>57.1</b> Introduction</a></li>
<li class="chapter" data-level="57.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#preliminaries-20"><i class="fa fa-check"></i><b>57.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="57.2.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#vocab-8"><i class="fa fa-check"></i><b>57.2.1</b> Vocab</a></li>
<li class="chapter" data-level="57.2.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-r-functions"><i class="fa fa-check"></i><b>57.2.2</b> General R functions</a></li>
<li class="chapter" data-level="57.2.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#specific-functions"><i class="fa fa-check"></i><b>57.2.3</b> Specific functions</a></li>
</ul></li>
<li class="chapter" data-level="57.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#sequence-comparisons"><i class="fa fa-check"></i><b>57.3</b> Sequence comparisons</a></li>
<li class="chapter" data-level="57.4" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrices"><i class="fa fa-check"></i><b>57.4</b> Distance matrices</a></li>
<li class="chapter" data-level="57.5" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#from-a-distance-matrix-to-phylogenetic-tree"><i class="fa fa-check"></i><b>57.5</b> From a distance matrix to phylogenetic tree</a></li>
<li class="chapter" data-level="57.6" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-raw-sequences"><i class="fa fa-check"></i><b>57.6</b> Distance matrix in R from raw sequences</a></li>
<li class="chapter" data-level="57.7" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-normal-matrix"><i class="fa fa-check"></i><b>57.7</b> Distance matrix in R from normal matrix</a></li>
<li class="chapter" data-level="57.8" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#upgma-in-r"><i class="fa fa-check"></i><b>57.8</b> UPGMA in R</a></li>
<li class="chapter" data-level="57.9" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#how-upgma-clustering-works"><i class="fa fa-check"></i><b>57.9</b> How UPGMA clustering works</a>
<ul>
<li class="chapter" data-level="57.9.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#challenge"><i class="fa fa-check"></i><b>57.9.1</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="57.10" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#further-information"><i class="fa fa-check"></i><b>57.10</b> Further information</a>
<ul>
<li class="chapter" data-level="57.10.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-info"><i class="fa fa-check"></i><b>57.10.1</b> General info</a></li>
<li class="chapter" data-level="57.10.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#worked-examples"><i class="fa fa-check"></i><b>57.10.2</b> Worked examples:</a></li>
<li class="chapter" data-level="57.10.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#additional-notes"><i class="fa fa-check"></i><b>57.10.3</b> Additional notes:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="58" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><i class="fa fa-check"></i><b>58</b> Tutorial: Writing functions to calculate the number of phylogenetic trees</a>
<ul>
<li class="chapter" data-level="58.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#repliminaries-1"><i class="fa fa-check"></i><b>58.1</b> Repliminaries</a>
<ul>
<li class="chapter" data-level="58.1.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-7"><i class="fa fa-check"></i><b>58.1.1</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="58.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#concepts-1"><i class="fa fa-check"></i><b>58.2</b> Concepts</a>
<ul>
<li class="chapter" data-level="58.2.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vocab-9"><i class="fa fa-check"></i><b>58.2.1</b> Vocab:</a></li>
<li class="chapter" data-level="58.2.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#packages-10"><i class="fa fa-check"></i><b>58.2.2</b> Packages:</a></li>
</ul></li>
<li class="chapter" data-level="58.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#introduction-11"><i class="fa fa-check"></i><b>58.3</b> Introduction</a></li>
<li class="chapter" data-level="58.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-1"><i class="fa fa-check"></i><b>58.4</b> Number of rooted trees</a></li>
<li class="chapter" data-level="58.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-in-r-1"><i class="fa fa-check"></i><b>58.5</b> Number of rooted trees in R</a></li>
<li class="chapter" data-level="58.6" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-in-r-1"><i class="fa fa-check"></i><b>58.6</b> Functions in R</a></li>
<li class="chapter" data-level="58.7" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-to-calculate-the-number-of-possible-phylogenetic-trees"><i class="fa fa-check"></i><b>58.7</b> Function to calculate the number of possible phylogenetic trees</a></li>
<li class="chapter" data-level="58.8" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vectorized-inputs-to-functions"><i class="fa fa-check"></i><b>58.8</b> Vectorized inputs to functions</a></li>
<li class="chapter" data-level="58.9" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-conditional-statements"><i class="fa fa-check"></i><b>58.9</b> Adding conditional statements</a></li>
<li class="chapter" data-level="58.10" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-multiple-conditional-statements"><i class="fa fa-check"></i><b>58.10</b> Adding multiple conditional statements</a></li>
<li class="chapter" data-level="58.11" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-additional-arguments"><i class="fa fa-check"></i><b>58.11</b> Adding additional arguments</a></li>
<li class="chapter" data-level="58.12" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-_un_rooted-trees"><i class="fa fa-check"></i><b>58.12</b> Assignment: _un_rooted trees</a></li>
<li class="chapter" data-level="58.13" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-1"><i class="fa fa-check"></i><b>58.13</b> Assignment part 1</a>
<ul>
<li class="chapter" data-level="58.13.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#part-1-answer"><i class="fa fa-check"></i><b>58.13.1</b> Part 1 answer</a></li>
</ul></li>
<li class="chapter" data-level="58.14" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-2"><i class="fa fa-check"></i><b>58.14</b> Assignment part 2</a>
<ul>
<li class="chapter" data-level="58.14.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#answer"><i class="fa fa-check"></i><b>58.14.1</b> Answer</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="59" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html"><i class="fa fa-check"></i><b>59</b> Writing functions - practice problems</a>
<ul>
<li class="chapter" data-level="59.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#introduction-12"><i class="fa fa-check"></i><b>59.1</b> Introduction</a></li>
<li class="chapter" data-level="59.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#part-1-pure-practice"><i class="fa fa-check"></i><b>59.2</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="59.2.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-1-hello-world"><i class="fa fa-check"></i><b>59.2.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="59.2.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______"><i class="fa fa-check"></i><b>59.2.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="59.3" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function"><i class="fa fa-check"></i><b>59.3</b> Practice Function 3: Make you own natural log function</a></li>
<li class="chapter" data-level="59.4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-4-make-function-that-converts-dna-directly-to-rna"><i class="fa fa-check"></i><b>59.4</b> Practice Function 4: Make function that converts DNA directly to RNA</a></li>
</ul></li>
<li class="chapter" data-level="60" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html"><i class="fa fa-check"></i><b>60</b> Writing user-friendly functions</a>
<ul>
<li class="chapter" data-level="60.1" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-1"><i class="fa fa-check"></i><b>60.1</b> Version 1</a></li>
<li class="chapter" data-level="60.2" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-2"><i class="fa fa-check"></i><b>60.2</b> Version 2</a></li>
<li class="chapter" data-level="60.3" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-3"><i class="fa fa-check"></i><b>60.3</b> Version 3</a></li>
<li class="chapter" data-level="60.4" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-4-adding-conditions-and-warnigns"><i class="fa fa-check"></i><b>60.4</b> Version 4: Adding conditions and warnigns</a></li>
<li class="chapter" data-level="60.5" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-5"><i class="fa fa-check"></i><b>60.5</b> Version 5</a></li>
</ul></li>
<li class="chapter" data-level="61" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html"><i class="fa fa-check"></i><b>61</b> Function practice exercises</a>
<ul>
<li class="chapter" data-level="61.1" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#part-1-pure-practice-1"><i class="fa fa-check"></i><b>61.1</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="61.1.1" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#practice-function-1-hello-world-1"><i class="fa fa-check"></i><b>61.1.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="61.1.2" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#practice-function-2-hello-_______-1"><i class="fa fa-check"></i><b>61.1.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="61.2" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#practice-function-3-make-you-own-natural-log-function-1"><i class="fa fa-check"></i><b>61.2</b> Practice Function 3: Make you own natural log function</a></li>
</ul></li>
<li class="chapter" data-level="62" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html"><i class="fa fa-check"></i><b>62</b> for() loops in R</a>
<ul>
<li class="chapter" data-level="62.1" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#introduction-13"><i class="fa fa-check"></i><b>62.1</b> Introduction</a></li>
<li class="chapter" data-level="62.2" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#reminders"><i class="fa fa-check"></i><b>62.2</b> Reminders</a></li>
<li class="chapter" data-level="62.3" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop"><i class="fa fa-check"></i><b>62.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="62.4" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-classic-for-loop"><i class="fa fa-check"></i><b>62.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="62.5" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-power-user-for-loop"><i class="fa fa-check"></i><b>62.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="62.6" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-4-a-for-loop-for-all-purposes"><i class="fa fa-check"></i><b>62.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="62.7" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#advanced-aside-if-else-and-next-statements"><i class="fa fa-check"></i><b>62.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="62.8" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#on-avoiding-for-loops"><i class="fa fa-check"></i><b>62.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="62.9" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#challenge-1"><i class="fa fa-check"></i><b>62.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="63" data-path="random-number-generation.html"><a href="random-number-generation.html"><i class="fa fa-check"></i><b>63</b> Random number generation</a>
<ul>
<li class="chapter" data-level="63.1" data-path="random-number-generation.html"><a href="random-number-generation.html#introduction-14"><i class="fa fa-check"></i><b>63.1</b> Introduction</a></li>
<li class="chapter" data-level="63.2" data-path="random-number-generation.html"><a href="random-number-generation.html#using-set.seed-to-make-simulations-reproducible"><i class="fa fa-check"></i><b>63.2</b> Using set.seed() to make simulations reproducible</a></li>
<li class="chapter" data-level="63.3" data-path="random-number-generation.html"><a href="random-number-generation.html#types-of-random-number-generators"><i class="fa fa-check"></i><b>63.3</b> Types of random number generators</a></li>
<li class="chapter" data-level="63.4" data-path="random-number-generation.html"><a href="random-number-generation.html#reporting-simulations"><i class="fa fa-check"></i><b>63.4</b> Reporting simulations</a></li>
<li class="chapter" data-level="63.5" data-path="random-number-generation.html"><a href="random-number-generation.html#notes"><i class="fa fa-check"></i><b>63.5</b> Notes:</a></li>
</ul></li>
<li class="chapter" data-level="64" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html"><i class="fa fa-check"></i><b>64</b> Randomization practice question</a>
<ul>
<li class="chapter" data-level="64.1" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#practice-question"><i class="fa fa-check"></i><b>64.1</b> Practice question:</a></li>
<li class="chapter" data-level="64.2" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#your-answer"><i class="fa fa-check"></i><b>64.2</b> Your answer:</a></li>
<li class="chapter" data-level="64.3" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-0"><i class="fa fa-check"></i><b>64.3</b> Hint 0</a></li>
<li class="chapter" data-level="64.4" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-1"><i class="fa fa-check"></i><b>64.4</b> Hint 1:</a></li>
<li class="chapter" data-level="64.5" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-2-follow-up-to-hint-0"><i class="fa fa-check"></i><b>64.5</b> Hint 2 (follow up to Hint 0)</a></li>
<li class="chapter" data-level="64.6" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-3"><i class="fa fa-check"></i><b>64.6</b> Hint 3:</a></li>
<li class="chapter" data-level="64.7" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#answer-1"><i class="fa fa-check"></i><b>64.7</b> Answer</a></li>
</ul></li>
<li class="chapter" data-level="65" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html"><i class="fa fa-check"></i><b>65</b> Simulating random amino acid sequences</a>
<ul>
<li class="chapter" data-level="65.1" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#selecting-random-letters-from-a-vector"><i class="fa fa-check"></i><b>65.1</b> Selecting random letters from a vector</a></li>
<li class="chapter" data-level="65.2" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-letters-from-the-whole-alphabet"><i class="fa fa-check"></i><b>65.2</b> Select random letters from the whole alphabet</a></li>
<li class="chapter" data-level="65.3" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-amino-acids-to-build-polypeptide"><i class="fa fa-check"></i><b>65.3</b> Select random amino acids to build polypeptide</a></li>
</ul></li>
<li class="chapter" data-level="66" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>66</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="66.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>66.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="66.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>66.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="66.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>66.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="66.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>66.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="66.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>66.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-3"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="66.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>66.2</b> Help!</a>
<ul>
<li class="chapter" data-level="66.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>66.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="66.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>66.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="66.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>66.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="66.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>66.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="66.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>66.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="66.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>66.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="66.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>66.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="66.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>66.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="calculating-e-values-bit-scores-and-p-values-from-blast-output" class="section level1" number="53">
<h1><span class="header-section-number">Chapter 53</span> Calculating E values, Bit scores, and P-values from BLAST output</h1>
<p><strong>By:</strong> Nathan Brouwer</p>
<p><strong>NOTE</strong>: This has NOT be edited and NOT been spell checked</p>
<div id="introduction-7" class="section level2" number="53.1">
<h2><span class="header-section-number">53.1</span> Introduction</h2>
<p>BLAST gives you a number of statistics for understanding the output of a search. Below I R code showing how this output relates to the underlying equations used to derive them.</p>
<div class="sourceCode" id="cb989"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb989-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb989-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb989-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb989-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(seqinr)</span>
<span id="cb989-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb989-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(BLOSUM62)</span></code></pre></div>
</div>
<div id="case-study" class="section level2" number="53.2">
<h2><span class="header-section-number">53.2</span> Case Study</h2>
<p>To explore BLAST we’ll do a protein “BLAST 2 sequences” search comparing an isoform of Drosophila enabled (NP_725859.2) against its human homolog (NP_060682.2). We’ll use the default BLOSUM 62 scoring matrix, but modify a few things</p>
<ul>
<li>Use Gap costs of existence = 12 and extension = 1</li>
<li>No compositional adjustment</li>
<li>No masking low complexity sequences</li>
</ul>
<p>Here are the sequences:</p>
<div class="sourceCode" id="cb990"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb990-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-1" aria-hidden="true" tabindex="-1"></a>d.ena <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;MAMKKLYAKTSFTSKKPSSAANSTPILAYHQQQHQQPGNGICEFQVVAPGHSGELMIRRSQSMHHKMSPPVGGLGSKSEYYSIEELQELDLLDYRHPMYHHYQQQELRQRYHEHEQLVLQLPKATSPKAGPIYEAPQRSQQQQDQMFEQSIIGARASVMVYDDNQKKWVPSGSSSGLSKVQIYHHQQNNTFRVVGRKLQDHEVVINCSIL</span></span>
<span id="cb990-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-2" aria-hidden="true" tabindex="-1"></a><span class="st">KGLKYNQATATFHQWRDSKFVYGLNFSSQNDAENFARAMMHALEVLSGRVANNPGGPPTNGNGYEEDMGY</span></span>
<span id="cb990-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-3" aria-hidden="true" tabindex="-1"></a><span class="st">RTMTSEDAAILRQNNSIGGHVTPSAQTPTSQTNQNNIPQSPPTPQGHHRTSRNSLSAPPAPQPQQQQQQQ</span></span>
<span id="cb990-4"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-4" aria-hidden="true" tabindex="-1"></a><span class="st">QAQQMGQPGSHYGPTGNGPTSNGLPQQVNSQIPPAPQQQPQQQQFQQQQQQQQYQQMVQAGYAPSQQYQQ</span></span>
<span id="cb990-5"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-5" aria-hidden="true" tabindex="-1"></a><span class="st">PHYVLSNSNPNLTVHQYPTQQAQQQPPQAPQPPLQNGGMYMVGHGHLPSSASANSVVYASQQQMLPQAHP</span></span>
<span id="cb990-6"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-6" aria-hidden="true" tabindex="-1"></a><span class="st">QAPQAPTMPGPGYGGPPVPPPQQQAENPYGQVPMPPPMNPSQQQQPGQVPLNRMSSQGGPGGPPAPAPPP</span></span>
<span id="cb990-7"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-7" aria-hidden="true" tabindex="-1"></a><span class="st">PPPSFGGAAGGGPPPPAPPQMFNGAPPPPAMGGGPPPAPPAPPAMGGGPPPAPGGPGAPPPPPPPPGLGG</span></span>
<span id="cb990-8"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-8" aria-hidden="true" tabindex="-1"></a><span class="st">APKKEDPQADLMGSLASQLQQIKLKKNKVTTSAPENSGSSTSSGGSGNYGTIGRSSNGMASMMDEMAKTL</span></span>
<span id="cb990-9"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-9" aria-hidden="true" tabindex="-1"></a><span class="st">ARRRAQAEKKDPDPEAEVKKRPWEKSNTLPHKLSGGAGSGSAGSGHEGANGNSGGAGSNTTNSGGESPRP</span></span>
<span id="cb990-10"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-10" aria-hidden="true" tabindex="-1"></a><span class="st">MRKRFGSASEETILKQVNGDGLSLALSNGDLDTLKAEIVREMRLEIQKVKNEIIDAIKSEFNRR</span></span>
<span id="cb990-11"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-11" aria-hidden="true" tabindex="-1"></a><span class="st">&quot;</span>)</span>
<span id="cb990-12"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb990-13"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-13" aria-hidden="true" tabindex="-1"></a>h.ena <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;MSEQSICQARAAVMVYDDANKKWVPAGGSTGFSRVHIYHHTGNNTFRVVGRKIQDHQVVINCAIPKGLKYNQATQTFHQWRDARQVYGLNFGSKEDANVFASAMMHALEVLNSQETGPTLPRQNSQLPAQVQNGPSQEELEIQRRQLQEQQRQKELERERLERERMERERLERERLERERLERERLEQEQLERERQERERQERLERQERL</span></span>
<span id="cb990-14"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-14" aria-hidden="true" tabindex="-1"></a><span class="st">ERQERLERQERLDRERQERQERERLERLERERQERERQEQLEREQLEWERERRISSAAAPASVETPLNSV</span></span>
<span id="cb990-15"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-15" aria-hidden="true" tabindex="-1"></a><span class="st">LGDSSASEPGLQAASQPAETPSQQGIVLGPLAPPPPPPLPPGPAQASVALPPPPGPPPPPPLPSTGPPPP</span></span>
<span id="cb990-16"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-16" aria-hidden="true" tabindex="-1"></a><span class="st">PPPPPLPNQVPPPPPPPPAPPLPASGFFLASMSEDNRPLTGLAAAIAGAKLRKVSRMEDTSFPSGGNAIG</span></span>
<span id="cb990-17"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-17" aria-hidden="true" tabindex="-1"></a><span class="st">VNSASSKTDTGRGNGPLPLGGSGLMEEMSALLARRRRIAEKGSTIETEQKEDKGEDSEPVTSKASSTSTP</span></span>
<span id="cb990-18"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-18" aria-hidden="true" tabindex="-1"></a><span class="st">EPTRKPWERTNTMNGSKSPVISRPKSTPLSQPSANGVQTEGLDYDRLKQDILDEMRKELTKLKEELIDAI</span></span>
<span id="cb990-19"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb990-19" aria-hidden="true" tabindex="-1"></a><span class="st">RQELSKSNTA&quot;</span>)</span></code></pre></div>
<p>We have to remove the newline characters:</p>
<div class="sourceCode" id="cb991"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb991-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb991-1" aria-hidden="true" tabindex="-1"></a>d.ena <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,<span class="st">&quot;&quot;</span>,d.ena)</span>
<span id="cb991-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb991-2" aria-hidden="true" tabindex="-1"></a>h.ena <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,<span class="st">&quot;&quot;</span>,h.ena)</span></code></pre></div>
</div>
<div id="blast-statistics" class="section level2" number="53.3">
<h2><span class="header-section-number">53.3</span> BLAST Statistics</h2>
<ul>
<li>E values = “Expected values”</li>
<li>Bit scores = Scores which can be compared between searches with different parameters. BLAST output just refers to these as “scores” though sometimes indicates that the units is “bits”. In equations this is S’ (“S-prime”).</li>
<li>Scores (S) = raw scores derived from scoring matrix and gap penalties. Available when looking at the individual local alignments.</li>
</ul>
<p>On the main output screen BLAST gives the “max score”, “total score”, and “E-value.” This should be read as “max bit score” and “total bit score.” The total bit score is the sum of the bit scores of all the high-scoring local alignments that were created within each database sequence.</p>
<p>For our case study, E is returned as 8e-70, the max bit score is 236, and the total bit score is 256 (again, BLAST is showing BIT scores, not raw scores). We can save these in R object</p>
<div class="sourceCode" id="cb992"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb992-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb992-1" aria-hidden="true" tabindex="-1"></a>E <span class="ot">&lt;-</span> <span class="fl">8e-70</span></span>
<span id="cb992-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb992-2" aria-hidden="true" tabindex="-1"></a>S.bit.max <span class="ot">&lt;-</span> <span class="dv">236</span></span>
<span id="cb992-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb992-3" aria-hidden="true" tabindex="-1"></a>S.bit.total <span class="ot">&lt;-</span> <span class="dv">256</span></span></code></pre></div>
<p>The main page also tells use the size of our query sequence, which we’ll call m. This is the length of our Drosophila enabled protein</p>
<div class="sourceCode" id="cb993"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb993-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb993-1" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">834</span></span></code></pre></div>
<p>Because we did a “BLAST two sequences” search we are also shown “Subject length”; this is the length of our human enabled homolog.</p>
<div class="sourceCode" id="cb994"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb994-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb994-1" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">570</span></span></code></pre></div>
<p>Since we specified a single sequence for comparison, this second sequence is the totality of our database for this search.</p>
<p>m x n is our <strong>search space</strong>. As will be discussed below, this is our raw search space; BLAST will use a small effective search space for actual calculations.</p>
<div class="sourceCode" id="cb995"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb995-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb995-1" aria-hidden="true" tabindex="-1"></a>mn <span class="ot">&lt;-</span> m<span class="sc">*</span>n</span>
<span id="cb995-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb995-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb995-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb995-3" aria-hidden="true" tabindex="-1"></a>mn</span></code></pre></div>
<pre><code>## [1] 475380</code></pre>
<p>m x n = 475380; this is the total number of amino acids compared by the search.</p>
</div>
<div id="dotplot" class="section level2" number="53.4">
<h2><span class="header-section-number">53.4</span> Dotplot</h2>
<p>BLAST produces dotplots which are quick visual representations of alignments. When only a two sequences are compare with BLAST 2 sequences, the dotplot is a visual represent representation of the search space.</p>
<div id="dotplot-in-r---optional-experimental-code" class="section level3" number="53.4.1">
<h3><span class="header-section-number">53.4.1</span> Dotplot in R - optional, experimental code</h3>
<p>We can make a dotplot in R, but its kind of clunky and I have yet to be able to really reproduce BLASTs output.</p>
<p>First set up the data (don’t worry about the details)</p>
<div class="sourceCode" id="cb997"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb997-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb997-1" aria-hidden="true" tabindex="-1"></a>d.ena.vect <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(d.ena,<span class="st">&quot;&quot;</span>))</span>
<span id="cb997-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb997-2" aria-hidden="true" tabindex="-1"></a>h.ena.vect <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(h.ena,<span class="st">&quot;&quot;</span>))</span></code></pre></div>
<p>Then make the dotplot (this is slooooooow, and I’m not happy with the results. I don’t recommend running it.).</p>
<div class="sourceCode" id="cb998"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb998-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb998-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dotPlot</span>(<span class="at">seq1 =</span> d.ena.vect,</span>
<span id="cb998-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb998-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">seq2 =</span> h.ena.vect,</span>
<span id="cb998-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb998-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">wsize =</span> <span class="dv">3</span>,</span>
<span id="cb998-4"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb998-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">wstep =</span> <span class="dv">1</span>,</span>
<span id="cb998-5"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb998-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">nmatch =</span> <span class="dv">3</span>)</span></code></pre></div>
<p>I started writing my own dotplot code but haven’t gotten very far.</p>
<div class="sourceCode" id="cb999"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb999-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data =</span> <span class="dv">0</span>,</span>
<span id="cb999-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">nrow =</span> <span class="fu">length</span>(h.ena.vect),</span>
<span id="cb999-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">ncol =</span> <span class="fu">length</span>(h.ena.vect))</span>
<span id="cb999-4"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb999-5"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(h.ena.vect)){</span>
<span id="cb999-6"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(h.ena.vect)){</span>
<span id="cb999-7"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-7" aria-hidden="true" tabindex="-1"></a>    n.i <span class="ot">&lt;-</span> d.ena.vect[i]</span>
<span id="cb999-8"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-8" aria-hidden="true" tabindex="-1"></a>    m.j <span class="ot">&lt;-</span> h.ena.vect[j]</span>
<span id="cb999-9"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-9" aria-hidden="true" tabindex="-1"></a>    score.ij <span class="ot">&lt;-</span> BLOSUM62[n.i, m.j]</span>
<span id="cb999-10"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-10" aria-hidden="true" tabindex="-1"></a>    df[i,j] <span class="ot">&lt;-</span> <span class="fu">ifelse</span>( score.ij<span class="sc">&gt;</span> <span class="dv">0</span>,score.ij,<span class="dv">0</span>)</span>
<span id="cb999-11"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb999-12"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb999-13"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb999-14"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-14" aria-hidden="true" tabindex="-1"></a><span class="fu">image</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(h.ena.vect),</span>
<span id="cb999-15"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">y =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(h.ena.vect),</span>
<span id="cb999-16"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb999-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">z =</span> df)</span></code></pre></div>
</div>
</div>
<div id="blast-search-summary" class="section level2" number="53.5">
<h2><span class="header-section-number">53.5</span> BLAST “Search Summary”</h2>
<p>BLAST provides a “Search summary” tab. At the top of the results three tabs are listed</p>
<p>Edit Search | Save Search | Search Summary</p>
<p>The search summary summarizes key aspects of how you defined the search (matrix, gap penalties, etc) and also the underlying parameters (lambda, K) defined in the equations below.</p>
<p>The “effective search space” is smaller than just m*n; after the test we’ll talk about the difference between the search space and the effective search space.</p>
<p>I’ll save the effective search space as an object</p>
<div class="sourceCode" id="cb1000"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1000-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1000-1" aria-hidden="true" tabindex="-1"></a>mn.effective <span class="ot">&lt;-</span> <span class="dv">428800</span></span></code></pre></div>
<p>Lambda can be obtained from the “search summary” tab. We want the value in the right-hand column.</p>
<div class="sourceCode" id="cb1001"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1001-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1001-1" aria-hidden="true" tabindex="-1"></a>lambda <span class="ot">&lt;-</span> <span class="fl">0.283</span></span></code></pre></div>
<p>K is also in the search summary tab.</p>
<div class="sourceCode" id="cb1002"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1002-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1002-1" aria-hidden="true" tabindex="-1"></a>K <span class="ot">&lt;-</span> <span class="fl">0.059</span></span></code></pre></div>
</div>
<div id="blast-alignments" class="section level2" number="53.6">
<h2><span class="header-section-number">53.6</span> BLAST Alignments</h2>
<p>On its main results page BLAST report the bit scores. BLAST reports raw scores derived directly from the scoring matrix on the “Alignments” tab.</p>
<p>When two sequences are compared BLAST can find more than one high-scoring, non-overlapping alignment. The Alignment tab shows us the two alignments for this search which passed BLASTs thresholds.</p>
<p>The first alignment has the annotation under “Score” as “236 bits(570)”. This is “bit score = 236 (raw score = 570)”. Bit score is S’ (S prime) and raw score is just S in equations. To be clear we’ll call this S.bit.max. S we’ll call S.max. We made S.bit.max above.</p>
<div class="sourceCode" id="cb1003"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1003-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1003-1" aria-hidden="true" tabindex="-1"></a>S.max <span class="ot">&lt;-</span> <span class="dv">570</span></span></code></pre></div>
<p>There’s a second alignment which has the annotation “19.6 bits(38).</p>
<div class="sourceCode" id="cb1004"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1004-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1004-1" aria-hidden="true" tabindex="-1"></a>S.bit.other <span class="ot">&lt;-</span> <span class="fl">19.6</span></span>
<span id="cb1004-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1004-2" aria-hidden="true" tabindex="-1"></a>S.other <span class="ot">&lt;-</span> <span class="dv">38</span></span></code></pre></div>
<!-- Check these equations -->
<p>Note that</p>
<div class="sourceCode" id="cb1005"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1005-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1005-1" aria-hidden="true" tabindex="-1"></a>S.max<span class="sc">+</span>S.bit.other</span></code></pre></div>
<pre><code>## [1] 589.6</code></pre>
<div class="sourceCode" id="cb1007"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1007-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1007-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(S.max<span class="sc">+</span>S.bit.other)</span></code></pre></div>
<pre><code>## [1] 590</code></pre>
<p>This is the total bit score (rounded up) that BLAST reports on the main results page</p>
<div class="sourceCode" id="cb1009"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1009-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1009-1" aria-hidden="true" tabindex="-1"></a>S.bit.total <span class="sc">==</span> <span class="fu">round</span>(S.max <span class="sc">+</span> S.bit.other)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<p>The local alignments also have an E value listed. The E-value for the best alignment is identical to the E-value from the main results page.</p>
</div>
<div id="equations" class="section level2" number="53.7">
<h2><span class="header-section-number">53.7</span> Equations</h2>
<p>Equation numbers refer to Pevsner</p>
<div id="e-value-eq-4.5-pg-142" class="section level3" number="53.7.1">
<h3><span class="header-section-number">53.7.1</span> E value (eq 4.5, pg 142)</h3>
<p>The E value is calculate as:
E = Kmne^(-lambda*S)</p>
<ul>
<li>K = a constant dependent on the scoring matrix (eg K for PAM200 is different than BLOSUM62)</li>
<li>m and n are the size of the query sequence and the size of the database</li>
<li>m*n = the search space: all possible aligned positions under consideration; in practice this gets tweaked to be the effective search space.<br />
</li>
<li>e = e^…; the exp() function in R</li>
<li>lambda = A constant; depends on scoring matrix (eg lambda for PAM200 is different than BLOSUM62)</li>
<li>S = raw score from local BLAST alignment (raw score, NOT bit score)</li>
</ul>
<p>BLAST uses an <strong>effective search space</strong> which can be obtained from the Search Summary tab.</p>
<p>We can try to replicate BLASTs calculation with this equation. Instead of m*n we’ll use mn.effective, but ignore why for now. We’ll use S.max instead of S to be clear which of the 2 searches we’re looking at.</p>
<p>I’ll do the calculation and save it into an object called “E from K” to distinguish it from the official E from BLAST</p>
<div class="sourceCode" id="cb1011"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1011-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1011-1" aria-hidden="true" tabindex="-1"></a>E.from.K <span class="ot">&lt;-</span> K<span class="sc">*</span>mn.effective<span class="sc">*</span><span class="fu">exp</span>(<span class="sc">-</span>lambda<span class="sc">*</span>S.max)</span>
<span id="cb1011-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1011-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1011-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1011-3" aria-hidden="true" tabindex="-1"></a>E.from.K</span></code></pre></div>
<pre><code>## [1] 2.223637e-66</code></pre>
<p>This result is fairly close but not exactly to what BLAST reports as the E value.</p>
<div class="sourceCode" id="cb1013"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1013-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1013-1" aria-hidden="true" tabindex="-1"></a>E</span></code></pre></div>
<pre><code>## [1] 8e-70</code></pre>
<p>This could be within rounding error. For example, perhaps lambda or K have been rounded off. It could also potentially be due to R’s limitations on working with a value like e-66.</p>
<p>Let’s check the other S value to see what we get</p>
<div class="sourceCode" id="cb1015"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1015-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1015-1" aria-hidden="true" tabindex="-1"></a>K<span class="sc">*</span>mn.effective<span class="sc">*</span><span class="fu">exp</span>(<span class="sc">-</span>lambda<span class="sc">*</span>S.other)</span></code></pre></div>
<pre><code>## [1] 0.5403858</code></pre>
<p>This is very close the the value reported in BLAST.</p>
</div>
</div>
<div id="bit-scores-s" class="section level2" number="53.8">
<h2><span class="header-section-number">53.8</span> Bit Scores (S’)</h2>
<p>Bit scores can be compared between BLAST searches.</p>
<p>S’ = (lambda*S - ln K)/(ln 2)</p>
<p>Recall that the natural log in R is just log()</p>
<p>We can try to reproduce BLASTS bit score. This doesn’t require any exceptional calculations on R’s part so there’s should be numeric problems.</p>
<div class="sourceCode" id="cb1017"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1017-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1017-1" aria-hidden="true" tabindex="-1"></a>(lambda<span class="sc">*</span>S.max <span class="sc">-</span> <span class="fu">log</span>(K))<span class="sc">/</span><span class="fu">log</span>(<span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [1] 236.8043</code></pre>
<div class="sourceCode" id="cb1019"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1019-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1019-1" aria-hidden="true" tabindex="-1"></a>S.bit.max</span></code></pre></div>
<pre><code>## [1] 236</code></pre>
<p>This is very very close to what BLAST reports, but is just a bit off. The alignment score (S.max) should be reported exactly because scoring matrices use integer values. For example</p>
<div class="sourceCode" id="cb1021"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1021-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1021-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;BLOSUM62&quot;</span>)</span>
<span id="cb1021-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1021-2" aria-hidden="true" tabindex="-1"></a>BLOSUM62[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##    A  R  N  D  C  Q  E  G  H  I
## A  4 -1 -2 -2  0 -1 -1  0 -2 -1
## R -1  5  0 -2 -3  1  0 -2  0 -3
## N -2  0  6  1 -3  0  0  0  1 -3
## D -2 -2  1  6 -3  0  2 -1 -1 -3
## C  0 -3 -3 -3  9 -3 -4 -3 -3 -1
## Q -1  1  0  0 -3  5  2 -2  0 -3
## E -1  0  0  2 -4  2  5 -2  0 -3
## G  0 -2  0 -1 -3 -2 -2  6 -2 -4
## H -2  0  1 -1 -3  0  0 -2  8 -3
## I -1 -3 -3 -3 -1 -3 -3 -4 -3  4</code></pre>
<p>Lambda and/or K might be reported or rounded just a bit off.</p>
</div>
<div id="e-from-bit-scores" class="section level2" number="53.9">
<h2><span class="header-section-number">53.9</span> E from Bit scores</h2>
<p>E = mn*2^-S’</p>
<p>For our best score</p>
<div class="sourceCode" id="cb1023"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1023-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1023-1" aria-hidden="true" tabindex="-1"></a>E.from.bit <span class="ot">&lt;-</span> mn.effective<span class="sc">*</span><span class="dv">2</span><span class="sc">^</span>(<span class="sc">-</span>S.bit.max)</span>
<span id="cb1023-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1023-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1023-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1023-3" aria-hidden="true" tabindex="-1"></a>E</span></code></pre></div>
<pre><code>## [1] 8e-70</code></pre>
<div class="sourceCode" id="cb1025"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1025-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1025-1" aria-hidden="true" tabindex="-1"></a>E.from.bit</span></code></pre></div>
<pre><code>## [1] 3.883075e-66</code></pre>
<div class="sourceCode" id="cb1027"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1027-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1027-1" aria-hidden="true" tabindex="-1"></a>E.from.K</span></code></pre></div>
<pre><code>## [1] 2.223637e-66</code></pre>
<p>The official E from BLAST is larger by a bit, while E from the bit score and E from the K (and Lambda) are the same order of magnitude. Not sure what’s going on here.</p>
<p>We can see what happens if we calculate “E from K” using the total bit score</p>
<div class="sourceCode" id="cb1029"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1029-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1029-1" aria-hidden="true" tabindex="-1"></a>K<span class="sc">*</span>mn.effective<span class="sc">*</span><span class="fu">exp</span>(<span class="sc">-</span>lambda<span class="sc">*</span>(S.max<span class="sc">+</span>S.other))</span></code></pre></div>
<pre><code>## [1] 4.749644e-71</code></pre>
<p>From our E value we can back calculate our (effective) search space</p>
<p>From</p>
<p>E = mn*2^-S’</p>
<p>we divide by both sides by 2^-S’ to get.
E/2^-S’ = mn</p>
<p>So mn is</p>
<div class="sourceCode" id="cb1031"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1031-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1031-1" aria-hidden="true" tabindex="-1"></a>E<span class="sc">/</span>(<span class="dv">2</span><span class="sc">^-</span>S.bit.max)</span></code></pre></div>
<pre><code>## [1] 88.34235</code></pre>
<p>Which is way off.</p>
<p>But</p>
<div class="sourceCode" id="cb1033"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1033-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1033-1" aria-hidden="true" tabindex="-1"></a>E.from.K<span class="sc">/</span>(<span class="dv">2</span><span class="sc">^-</span>S.bit.max)</span></code></pre></div>
<pre><code>## [1] 245551.7</code></pre>
<div class="sourceCode" id="cb1035"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1035-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1035-1" aria-hidden="true" tabindex="-1"></a>mn</span></code></pre></div>
<pre><code>## [1] 475380</code></pre>
<p>is kinda close.</p>
</div>
<div id="p-values" class="section level2" number="53.10">
<h2><span class="header-section-number">53.10</span> P values</h2>
<p>P = 1 - e^-E</p>
<p>For the second of the two alignments</p>
<div class="sourceCode" id="cb1037"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1037-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1037-1" aria-hidden="true" tabindex="-1"></a>E.other <span class="ot">&lt;-</span> <span class="fl">0.53</span></span>
<span id="cb1037-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1037-2" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="sc">-</span> <span class="fu">exp</span>(<span class="sc">-</span>E.other)</span></code></pre></div>
<pre><code>## [1] 0.411395</code></pre>
<p>You can convert from a p-value if you had it to an E-value</p>
<p>P = 1 - e^-E
-1+P = -e^-E
1-P = e^-E
log(1-P) = -E
-1*log(1-P) = E</p>

<p>output: html_document
editor_options:
chunk_output_type: console
—
# Calculating the number of possible phylogenetic trees</p>
<div class="sourceCode" id="cb1039"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1039-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1039-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compbio4all)</span></code></pre></div>
<!-- remove all R code? move to other section -->
</div>
<div id="repliminaries" class="section level2" number="53.11">
<h2><span class="header-section-number">53.11</span> Repliminaries</h2>
</div>
<div id="concepts" class="section level2" number="53.12">
<h2><span class="header-section-number">53.12</span> Concepts</h2>
<div id="biology-concepts-vocab" class="section level4" number="53.12.0.1">
<h4><span class="header-section-number">53.12.0.1</span> Biology Concepts / Vocab</h4>
<ul>
<li>toplogy</li>
<li>rotation</li>
<li>Number of possible phylogenetics trees</li>
</ul>
</div>
</div>
<div id="functions-6" class="section level2" number="53.13">
<h2><span class="header-section-number">53.13</span> Functions</h2>
<ul>
<li>factorial</li>
</ul>
</div>
<div id="introduction-8" class="section level2" number="53.14">
<h2><span class="header-section-number">53.14</span> Introduction</h2>
<p><strong>Phylogenetics</strong> is one of the original fields of computational biology. This is because as the number of <strong>taxa</strong> that need to be organized on a tree increases, the number of <em>unique</em> phylogenetic trees describing possible relationships among those taxa increases rapidly.</p>
<!-- Add example -->
<p>The following refers to <strong>bifurcating</strong> tree which always branch into two branches at each <strong>node</strong>. There are the type of branches built for phylogenies.</p>
</div>
<div id="number-of-rooted-trees" class="section level2" number="53.15">
<h2><span class="header-section-number">53.15</span> Number of rooted trees</h2>
<p>The number of possible rooted phylogenetic trees is calculated using the equation below, where n is the number of taxa:</p>
<p>In text, the equation looks like this:
(2n-3)!/(2^(n-2))(n-2)!</p>
<p>Written in a math book it would look like this:</p>
<p><span class="math inline">\(\frac{(2*n-3)!}{2^{n-2} * (n-2)!}\)</span></p>
<p>This equation is occasionally mis-printed (as I did in the part of this original assignment) so its worth being explicit:</p>
<p>The numerator is: (2n-3)!, where “!” is factorial
The denominator is: (2^n-2)*(n-2)!; the factorial is evaluated first.</p>
<p>So, for n = 3 taxa</p>
<p>(2n-3)!/(2^(n-2))(n-2)!
= (2x3-3)!/(2^(3-2))x(3-2)!
= (6 -3)!/(2^(1))x(1)!
= (3)!/(2)x(1)!
= (1 x 2 x 3)/(2 x 1)
= 6 / 2
= 3</p>
<p>One of our first tasks will be to translate this equation into R code.</p>
</div>
<div id="number-of-rooted-trees-in-r" class="section level2" number="53.16">
<h2><span class="header-section-number">53.16</span> Number of rooted trees in R</h2>
<p>In R we take factorials using the <code>factorial()</code> function</p>
<div class="sourceCode" id="cb1040"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1040-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1040-1" aria-hidden="true" tabindex="-1"></a><span class="fu">factorial</span>(<span class="dv">3</span>)</span></code></pre></div>
<pre><code>## [1] 6</code></pre>
<div class="sourceCode" id="cb1042"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1042-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1042-1" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span><span class="sc">*</span><span class="dv">2</span><span class="sc">*</span><span class="dv">1</span></span></code></pre></div>
<pre><code>## [1] 6</code></pre>
<div class="sourceCode" id="cb1044"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1044-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1044-1" aria-hidden="true" tabindex="-1"></a><span class="fu">factorial</span>(<span class="dv">4</span>)</span></code></pre></div>
<pre><code>## [1] 24</code></pre>
<div class="sourceCode" id="cb1046"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1046-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1046-1" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span><span class="sc">*</span><span class="dv">3</span><span class="sc">*</span><span class="dv">2</span><span class="sc">*</span><span class="dv">1</span></span></code></pre></div>
<pre><code>## [1] 24</code></pre>
<p>We can re-write our text equation above as:
factorial(2n-3)/(2^(n-2))*(factorial(n-2))</p>
<p>In this would be</p>
<div class="sourceCode" id="cb1048"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1048-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1048-1" aria-hidden="true" tabindex="-1"></a>n<span class="ot">&lt;-</span><span class="dv">3</span></span>
<span id="cb1048-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1048-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1048-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1048-3" aria-hidden="true" tabindex="-1"></a><span class="fu">factorial</span>(<span class="dv">2</span><span class="sc">*</span>n<span class="dv">-3</span>)<span class="sc">/</span>((<span class="dv">2</span><span class="sc">^</span>(n<span class="dv">-2</span>))<span class="sc">*</span><span class="fu">factorial</span>(n<span class="dv">-2</span>))</span></code></pre></div>
<pre><code>## [1] 3</code></pre>
<p>All of the parentheses make this a bit nutty. Let me write this out as a separate numerator and denominator</p>
<div class="sourceCode" id="cb1050"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1050-1"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1050-1" aria-hidden="true" tabindex="-1"></a><span class="co">#numerator</span></span>
<span id="cb1050-2"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1050-2" aria-hidden="true" tabindex="-1"></a>numerator   <span class="ot">&lt;-</span> <span class="fu">factorial</span>(<span class="dv">2</span><span class="sc">*</span>n<span class="dv">-3</span>)</span>
<span id="cb1050-3"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1050-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1050-4"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1050-4" aria-hidden="true" tabindex="-1"></a><span class="co">#denominator</span></span>
<span id="cb1050-5"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1050-5" aria-hidden="true" tabindex="-1"></a>denominator <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">^</span>(n<span class="dv">-2</span>)<span class="sc">*</span><span class="fu">factorial</span>(n<span class="dv">-2</span>)</span>
<span id="cb1050-6"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1050-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1050-7"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1050-7" aria-hidden="true" tabindex="-1"></a><span class="co">#division</span></span>
<span id="cb1050-8"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#cb1050-8" aria-hidden="true" tabindex="-1"></a>numerator<span class="sc">/</span>denominator</span></code></pre></div>
<pre><code>## [1] 3</code></pre>
<div id="unique-topologies-versus-rotations" class="section level3" number="53.16.1">
<h3><span class="header-section-number">53.16.1</span> Unique topologies versus rotations</h3>
<p>Note that this applies to all tree typologies. For n = 3 species the three typologies are</p>
<ol style="list-style-type: decimal">
<li>((Human, Chimp), Gorilla)</li>
<li>((Human, Gorilla), Chimp)</li>
<li>((Chimp, Gorilla), Human)</li>
</ol>
<p>In this discussion Since we are referring to evolutionary <em>unique</em> typologies. Therefore order within a clade does <em>not</em> matter. Therefore, these two are identical typologies and don’t get counted among the three unique typologies the equation is telling use exist:.</p>
<ol style="list-style-type: decimal">
<li>((Human, Chimp), Gorilla)</li>
<li>((Chimp, Human), Gorilla)</li>
</ol>
<!-- plot examples -->
<!-- number of unrooted trees -->
</div>
</div>
<div id="bilography" class="section level2" number="53.17">
<h2><span class="header-section-number">53.17</span> Bilography</h2>
<p>This topic is discussed on page 187-189 of Baum &amp; Smith Tree Thinking. They discuss it in the concept of <strong>maximum parsimony</strong> estimation of phylogenetic trees.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="blast-alignment-score-distribution.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="upgma-the-hard-way.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/555-BLAST/BLAST-dayoff-E_values.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/555-BLAST/BLAST-dayoff-E_values.Rmd",
"text": null
},
"download": ["lbrb.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
