<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 55 Comparative Genomics | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 55 Comparative Genomics | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 55 Comparative Genomics | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="significance-of-orfs.html"/>
<link rel="next" href="summary-4.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>1</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="1.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>1.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="1.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>1.3</b> Introduction</a></li>
<li class="chapter" data-level="1.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary"><i class="fa fa-check"></i><b>1.4</b> Vocabulary</a></li>
<li class="chapter" data-level="1.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions"><i class="fa fa-check"></i><b>1.5</b> Functions</a></li>
<li class="chapter" data-level="1.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>1.6</b> Preliminaries</a></li>
<li class="chapter" data-level="1.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>1.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="1.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>1.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="1.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>1.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="1.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>1.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="1.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>1.8.3</b> DNA words</a></li>
<li class="chapter" data-level="1.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>1.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="1.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>1.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="1.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>1.9.1</b> License</a></li>
<li class="chapter" data-level="1.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>1.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="working-with-vectors-and-matrices.html"><a href="working-with-vectors-and-matrices.html"><i class="fa fa-check"></i><b>2</b> Working with vectors and matrices</a></li>
<li class="chapter" data-level="3" data-path="a-note-on-changes-to-database-entries.html"><a href="a-note-on-changes-to-database-entries.html"><i class="fa fa-check"></i><b>3</b> A note on changes to database entries</a></li>
<li class="chapter" data-level="4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html"><i class="fa fa-check"></i><b>4</b> Writing functions - practice problems</a>
<ul>
<li class="chapter" data-level="4.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#introduction-1"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#part-1-pure-practice"><i class="fa fa-check"></i><b>4.2</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-1-hello-world"><i class="fa fa-check"></i><b>4.2.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="4.2.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______"><i class="fa fa-check"></i><b>4.2.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function"><i class="fa fa-check"></i><b>4.3</b> Practice Function 3: Make you own natural log function</a></li>
<li class="chapter" data-level="4.4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-4-make-function-that-converts-dna-directly-to-rna"><i class="fa fa-check"></i><b>4.4</b> Practice Function 4: Make function that converts DNA directly to RNA</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html"><i class="fa fa-check"></i><b>5</b> Computational Biology Portfolio: Practice Working with Strings</a>
<ul>
<li class="chapter" data-level="5.1" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#introduction-2"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#purpose"><i class="fa fa-check"></i><b>5.2</b> Purpose</a></li>
<li class="chapter" data-level="5.3" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#gonder-et-al-2007-data"><i class="fa fa-check"></i><b>5.3</b> Gonder et al 2007 data</a></li>
<li class="chapter" data-level="5.4" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#gonder-et-al-2007-new-sequences"><i class="fa fa-check"></i><b>5.4</b> Gonder et al 2007 new sequences</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#creating-the-codes"><i class="fa fa-check"></i><b>5.4.1</b> Creating the codes</a></li>
<li class="chapter" data-level="5.4.2" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#selecting-a-code-for-demonstration"><i class="fa fa-check"></i><b>5.4.2</b> Selecting a code for demonstration</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#searching-genbank"><i class="fa fa-check"></i><b>5.5</b> Searching Genbank</a></li>
<li class="chapter" data-level="5.6" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#gonder-et-al-2007-previously-published-sequences"><i class="fa fa-check"></i><b>5.6</b> Gonder et al 2007 previously published sequences</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#creating-the-codes-1"><i class="fa fa-check"></i><b>5.6.1</b> Creating the codes</a></li>
<li class="chapter" data-level="5.6.2" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#selecting-a-code-for-demonstration-1"><i class="fa fa-check"></i><b>5.6.2</b> Selecting a code for demonstration</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#searching-genbank-1"><i class="fa fa-check"></i><b>5.7</b> Searching Genbank</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html"><i class="fa fa-check"></i><b>6</b> Writing user-friendly functions</a>
<ul>
<li class="chapter" data-level="6.1" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-1"><i class="fa fa-check"></i><b>6.1</b> Version 1</a></li>
<li class="chapter" data-level="6.2" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-2"><i class="fa fa-check"></i><b>6.2</b> Version 2</a></li>
<li class="chapter" data-level="6.3" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-3"><i class="fa fa-check"></i><b>6.3</b> Version 3</a></li>
<li class="chapter" data-level="6.4" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-4-adding-conditions-and-warnigns"><i class="fa fa-check"></i><b>6.4</b> Version 4: Adding conditions and warnigns</a></li>
<li class="chapter" data-level="6.5" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-5"><i class="fa fa-check"></i><b>6.5</b> Version 5</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html"><i class="fa fa-check"></i><b>7</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="7.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#commands-used"><i class="fa fa-check"></i><b>7.1</b> Commands used</a></li>
<li class="chapter" data-level="7.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#r-objects"><i class="fa fa-check"></i><b>7.2</b> R Objects</a></li>
<li class="chapter" data-level="7.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>7.3</b> Differences between objects</a></li>
<li class="chapter" data-level="7.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-data"><i class="fa fa-check"></i><b>7.4</b> The Data</a></li>
<li class="chapter" data-level="7.5" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>7.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#is"><i class="fa fa-check"></i><b>7.5.1</b> is()</a></li>
<li class="chapter" data-level="7.5.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#length"><i class="fa fa-check"></i><b>7.5.2</b> length()</a></li>
<li class="chapter" data-level="7.5.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#str"><i class="fa fa-check"></i><b>7.5.3</b> str()</a></li>
<li class="chapter" data-level="7.5.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#c"><i class="fa fa-check"></i><b>7.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#you-can-save-statistical-output-as-an-object"><i class="fa fa-check"></i><b>7.6</b> You can save statistical output as an object</a></li>
<li class="chapter" data-level="7.7" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#how-does-r-save-statistical-output"><i class="fa fa-check"></i><b>7.7</b> How does R save statistical output</a></li>
<li class="chapter" data-level="7.8" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#lists-in-r"><i class="fa fa-check"></i><b>7.8</b> Lists in R</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="an-aside-about-t-tests.html"><a href="an-aside-about-t-tests.html"><i class="fa fa-check"></i><b>8</b> An aside about t-tests</a></li>
<li class="chapter" data-level="9" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html"><i class="fa fa-check"></i><b>9</b> Working with square brackets</a>
<ul>
<li class="chapter" data-level="9.1" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#square-brackets-for-real"><i class="fa fa-check"></i><b>9.1</b> Square brackets for real</a></li>
<li class="chapter" data-level="9.2" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#difference-between-means-using-square-brackets"><i class="fa fa-check"></i><b>9.2</b> Difference between means using square brackets</a></li>
<li class="chapter" data-level="9.3" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#upshot-plotting-output-of-an-r-statistical-test"><i class="fa fa-check"></i><b>9.3</b> Upshot: plotting output of an R statistical test</a></li>
<li class="chapter" data-level="9.4" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#the-not-bad-way-typing-directly-into-r"><i class="fa fa-check"></i><b>9.4</b> The “not bad way”: typing directly into R</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#not-bad-way-with-a-bar-plot"><i class="fa fa-check"></i><b>9.4.1</b> “NOt bad way” with a bar plot</a></li>
<li class="chapter" data-level="9.4.2" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#not-bad-way-with-using-a-single-dot"><i class="fa fa-check"></i><b>9.4.2</b> “NOt bad way” with using a single dot</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#the-best-way"><i class="fa fa-check"></i><b>9.5</b> The best way</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#a-final-fancy-touch"><i class="fa fa-check"></i><b>9.5.1</b> A final fancy touch</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#debrief"><i class="fa fa-check"></i><b>9.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html"><i class="fa fa-check"></i><b>10</b> or loops in R - copy? alt?</a>
<ul>
<li class="chapter" data-level="10.1" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#introduction-3"><i class="fa fa-check"></i><b>10.1</b> Introduction</a></li>
<li class="chapter" data-level="10.2" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#reminders"><i class="fa fa-check"></i><b>10.2</b> Reminders</a></li>
<li class="chapter" data-level="10.3" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop"><i class="fa fa-check"></i><b>10.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="10.4" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-3-the-classic-for-loop"><i class="fa fa-check"></i><b>10.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="10.5" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-3-the-power-user-for-loop"><i class="fa fa-check"></i><b>10.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="10.6" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-4-a-for-loop-for-all-purposes"><i class="fa fa-check"></i><b>10.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="10.7" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#advanced-aside-if-else-and-next-statements"><i class="fa fa-check"></i><b>10.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="10.8" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#on-avoiding-for-loops"><i class="fa fa-check"></i><b>10.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="10.9" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#challenge"><i class="fa fa-check"></i><b>10.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="function-practice.html"><a href="function-practice.html"><i class="fa fa-check"></i><b>11</b> Function practice</a>
<ul>
<li class="chapter" data-level="11.1" data-path="function-practice.html"><a href="function-practice.html#part-1-pure-practice-1"><i class="fa fa-check"></i><b>11.1</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="function-practice.html"><a href="function-practice.html#practice-function-1-hello-world-1"><i class="fa fa-check"></i><b>11.1.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="11.1.2" data-path="function-practice.html"><a href="function-practice.html#practice-function-2-hello-_______-1"><i class="fa fa-check"></i><b>11.1.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="function-practice.html"><a href="function-practice.html#practice-function-3-make-you-own-natural-log-function-1"><i class="fa fa-check"></i><b>11.2</b> Practice Function 3: Make you own natural log function</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html"><i class="fa fa-check"></i><b>12</b> for() loops in R</a>
<ul>
<li class="chapter" data-level="12.1" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#introduction-4"><i class="fa fa-check"></i><b>12.1</b> Introduction</a></li>
<li class="chapter" data-level="12.2" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#reminders-1"><i class="fa fa-check"></i><b>12.2</b> Reminders</a></li>
<li class="chapter" data-level="12.3" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop-1"><i class="fa fa-check"></i><b>12.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="12.4" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-classic-for-loop-1"><i class="fa fa-check"></i><b>12.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="12.5" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-power-user-for-loop-1"><i class="fa fa-check"></i><b>12.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="12.6" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-4-a-for-loop-for-all-purposes-1"><i class="fa fa-check"></i><b>12.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="12.7" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#advanced-aside-if-else-and-next-statements-1"><i class="fa fa-check"></i><b>12.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="12.8" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#on-avoiding-for-loops-1"><i class="fa fa-check"></i><b>12.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="12.9" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#challenge-1"><i class="fa fa-check"></i><b>12.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="random-number-generation.html"><a href="random-number-generation.html"><i class="fa fa-check"></i><b>13</b> Random number generation</a>
<ul>
<li class="chapter" data-level="13.1" data-path="random-number-generation.html"><a href="random-number-generation.html#introduction-5"><i class="fa fa-check"></i><b>13.1</b> Introduction</a></li>
<li class="chapter" data-level="13.2" data-path="random-number-generation.html"><a href="random-number-generation.html#using-set.seed-to-make-simulations-reproducible"><i class="fa fa-check"></i><b>13.2</b> Using set.seed() to make simulations reproducible</a></li>
<li class="chapter" data-level="13.3" data-path="random-number-generation.html"><a href="random-number-generation.html#types-of-random-number-generators"><i class="fa fa-check"></i><b>13.3</b> Types of random number generators</a></li>
<li class="chapter" data-level="13.4" data-path="random-number-generation.html"><a href="random-number-generation.html#reporting-simulations"><i class="fa fa-check"></i><b>13.4</b> Reporting simulations</a></li>
<li class="chapter" data-level="13.5" data-path="random-number-generation.html"><a href="random-number-generation.html#notes"><i class="fa fa-check"></i><b>13.5</b> Notes:</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html"><i class="fa fa-check"></i><b>14</b> Sample test question-Randomization</a>
<ul>
<li class="chapter" data-level="14.1" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#sample-test-question"><i class="fa fa-check"></i><b>14.1</b> Sample test question:</a></li>
<li class="chapter" data-level="14.2" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#your-answer"><i class="fa fa-check"></i><b>14.2</b> Your answer:</a></li>
<li class="chapter" data-level="14.3" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-0"><i class="fa fa-check"></i><b>14.3</b> Hint 0</a></li>
<li class="chapter" data-level="14.4" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-1"><i class="fa fa-check"></i><b>14.4</b> Hint 1:</a></li>
<li class="chapter" data-level="14.5" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-2-follow-up-to-hint-0"><i class="fa fa-check"></i><b>14.5</b> Hint 2 (follow up to Hint 0)</a></li>
<li class="chapter" data-level="14.6" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-3"><i class="fa fa-check"></i><b>14.6</b> Hint 3:</a></li>
<li class="chapter" data-level="14.7" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#answer"><i class="fa fa-check"></i><b>14.7</b> Answer</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html"><i class="fa fa-check"></i><b>15</b> Simulating random amino acid sequences</a>
<ul>
<li class="chapter" data-level="15.1" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#selecting-random-letters-from-a-vector"><i class="fa fa-check"></i><b>15.1</b> Selecting random letters from a vector</a></li>
<li class="chapter" data-level="15.2" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-letters-from-the-whole-alphabet"><i class="fa fa-check"></i><b>15.2</b> Select random letters from the whole alphabet</a></li>
<li class="chapter" data-level="15.3" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-amino-acids-to-build-polypeptide"><i class="fa fa-check"></i><b>15.3</b> Select random amino acids to build polypeptide</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>16</b> How to review this book</a>
<ul>
<li class="chapter" data-level="16.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>16.1</b> File format</a></li>
<li class="chapter" data-level="16.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>16.2</b> What can you do?</a></li>
<li class="chapter" data-level="16.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>16.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>16.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="16.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>16.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="16.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>16.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="16.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>16.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="test-audio-link.html"><a href="test-audio-link.html"><i class="fa fa-check"></i><b>17</b> Test audio link</a></li>
<li class="chapter" data-level="18" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>18</b> Test shiny app</a></li>
<li class="chapter" data-level="19" data-path="test-youtube-vidoe.html"><a href="test-youtube-vidoe.html"><i class="fa fa-check"></i><b>19</b> Test YouTube vidoe</a></li>
<li class="chapter" data-level="20" data-path="test-webpage-insertion.html"><a href="test-webpage-insertion.html"><i class="fa fa-check"></i><b>20</b> Test webpage insertion</a></li>
<li class="chapter" data-level="21" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>21</b> Datacamp test</a></li>
<li class="chapter" data-level="22" data-path="an-example-of-a-plotly-graph.html"><a href="an-example-of-a-plotly-graph.html"><i class="fa fa-check"></i><b>22</b> An example of a plotly graph</a></li>
<li class="chapter" data-level="23" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>23</b> Downloading R</a>
<ul>
<li class="chapter" data-level="23.1" data-path="downloading-r.html"><a href="downloading-r.html#preface-1"><i class="fa fa-check"></i><b>23.1</b> Preface</a></li>
<li class="chapter" data-level="23.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>23.2</b> Introduction to R</a></li>
<li class="chapter" data-level="23.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>23.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="23.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>23.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="23.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>23.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="23.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>23.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>24</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="24.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>24.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="24.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>24.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>25</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="25.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>25.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="25.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>25.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="25.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>25.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>26</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="26.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>26.1</b> Bioconductor</a></li>
<li class="chapter" data-level="26.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>26.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="26.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>26.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="26.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>26.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="26.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>26.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="26.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>26.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="26.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>26.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>27</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="27.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary-1"><i class="fa fa-check"></i><b>27.1</b> Vocabulary</a></li>
<li class="chapter" data-level="27.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>27.2</b> R functions</a></li>
<li class="chapter" data-level="27.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>27.3</b> Interacting with R</a></li>
<li class="chapter" data-level="27.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>27.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="27.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>27.4.1</b> Vectors</a></li>
<li class="chapter" data-level="27.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>27.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="27.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>27.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="27.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>27.4.4</b> Lists</a></li>
<li class="chapter" data-level="27.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>27.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="27.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>27.5</b> Arguments</a></li>
<li class="chapter" data-level="27.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>27.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="27.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>27.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="27.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>27.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="27.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>27.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="27.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>27.9</b> Quiting R</a></li>
<li class="chapter" data-level="27.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>27.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="28" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>28</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="28.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-2"><i class="fa fa-check"></i><b>28.1</b> Preface</a></li>
<li class="chapter" data-level="28.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>28.2</b> Vocab</a></li>
<li class="chapter" data-level="28.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-1"><i class="fa fa-check"></i><b>28.3</b> Functions</a></li>
<li class="chapter" data-level="28.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>28.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="28.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>28.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>29</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="29.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-3"><i class="fa fa-check"></i><b>29.1</b> Preface</a></li>
<li class="chapter" data-level="29.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>29.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="functions-2.html"><a href="functions-2.html"><i class="fa fa-check"></i><b>30</b> Functions</a>
<ul>
<li class="chapter" data-level="30.1" data-path="functions-2.html"><a href="functions-2.html#vectors-in-r"><i class="fa fa-check"></i><b>30.1</b> Vectors in R</a></li>
<li class="chapter" data-level="30.2" data-path="functions-2.html"><a href="functions-2.html#math-on-vectors"><i class="fa fa-check"></i><b>30.2</b> Math on vectors</a></li>
<li class="chapter" data-level="30.3" data-path="functions-2.html"><a href="functions-2.html#functions-on-vectors"><i class="fa fa-check"></i><b>30.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="30.4" data-path="functions-2.html"><a href="functions-2.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>30.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="30.5" data-path="functions-2.html"><a href="functions-2.html#subsetting-vectors"><i class="fa fa-check"></i><b>30.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="30.6" data-path="functions-2.html"><a href="functions-2.html#sequences-of-numbers"><i class="fa fa-check"></i><b>30.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="30.7" data-path="functions-2.html"><a href="functions-2.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>30.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="30.8" data-path="functions-2.html"><a href="functions-2.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>30.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="31" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>31</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="31.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-4"><i class="fa fa-check"></i><b>31.1</b> Preface</a></li>
<li class="chapter" data-level="31.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>31.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="31.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>31.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="32" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>32</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="32.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-5"><i class="fa fa-check"></i><b>32.1</b> Preface</a></li>
<li class="chapter" data-level="32.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>32.2</b> Vocab</a></li>
<li class="chapter" data-level="32.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-3"><i class="fa fa-check"></i><b>32.3</b> Functions</a></li>
<li class="chapter" data-level="32.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>32.4</b> Functions in R</a></li>
<li class="chapter" data-level="32.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>32.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="33" data-path="fasta-files.html"><a href="fasta-files.html"><i class="fa fa-check"></i><b>33</b> FASTA Files</a>
<ul>
<li class="chapter" data-level="33.1" data-path="fasta-files.html"><a href="fasta-files.html#example-fasta-file"><i class="fa fa-check"></i><b>33.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="33.2" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>33.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="33.3" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>33.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="33.4" data-path="fasta-files.html"><a href="fasta-files.html#fastq-format"><i class="fa fa-check"></i><b>33.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="34" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>34</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="34.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-4"><i class="fa fa-check"></i><b>34.0.1</b> Functions</a></li>
<li class="chapter" data-level="34.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>34.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="34.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>34.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="34.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>34.0.4</b> File types</a></li>
<li class="chapter" data-level="34.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>34.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="34.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>34.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="34.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>34.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="34.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>34.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="34.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>34.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="34.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>34.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="34.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>34.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="34.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>34.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="34.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>34.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="34.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>34.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="34.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>34.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>35</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="35.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-2"><i class="fa fa-check"></i><b>35.1</b> Preliminaries</a></li>
<li class="chapter" data-level="35.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>35.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="35.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>35.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="35.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>35.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="36" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>36</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="36.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>36.1</b> Preliminaries</a></li>
<li class="chapter" data-level="36.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>36.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>37</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="37.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>37.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="37.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>37.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="37.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>37.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="37.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>37.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="37.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>37.4</b> Inversions</a></li>
<li class="chapter" data-level="37.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>37.5</b> Translocations</a></li>
<li class="chapter" data-level="37.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>37.6</b> Random sequence</a></li>
<li class="chapter" data-level="37.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>37.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="38" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>38</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="38.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>38.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="38.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>38.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="38.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>38.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="38.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>38.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="38.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>38.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="38.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>38.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="38.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>38.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="38.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>38.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>39</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="39.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>39.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="39.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>39.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="39.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>39.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="40" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html"><i class="fa fa-check"></i><b>40</b> Alignment by eye in R</a>
<ul>
<li class="chapter" data-level="40.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>40.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="40.1.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#packages"><i class="fa fa-check"></i><b>40.1.1</b> Packages</a></li>
<li class="chapter" data-level="40.1.2" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#data"><i class="fa fa-check"></i><b>40.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="41" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html"><i class="fa fa-check"></i><b>41</b> Matrix vocabulary</a>
<ul>
<li class="chapter" data-level="41.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#introduction-6"><i class="fa fa-check"></i><b>41.1</b> Introduction</a></li>
<li class="chapter" data-level="41.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#preliminaries-8"><i class="fa fa-check"></i><b>41.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="41.2.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#packages-1"><i class="fa fa-check"></i><b>41.2.1</b> Packages</a></li>
<li class="chapter" data-level="41.2.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#vocab-3"><i class="fa fa-check"></i><b>41.2.2</b> Vocab</a></li>
<li class="chapter" data-level="41.2.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#functions-used"><i class="fa fa-check"></i><b>41.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="41.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#load-packages"><i class="fa fa-check"></i><b>41.3</b> Load packages</a></li>
<li class="chapter" data-level="41.4" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#matrix-structure"><i class="fa fa-check"></i><b>41.4</b> Matrix structure</a></li>
<li class="chapter" data-level="41.5" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#diagonal-of-a-matrix"><i class="fa fa-check"></i><b>41.5</b> Diagonal of a matrix</a></li>
</ul></li>
<li class="chapter" data-level="42" data-path="alignment-in-r.html"><a href="alignment-in-r.html"><i class="fa fa-check"></i><b>42</b> Alignment in R</a>
<ul>
<li class="chapter" data-level="42.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#preliminaries-9"><i class="fa fa-check"></i><b>42.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="42.1.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#packages-2"><i class="fa fa-check"></i><b>42.1.1</b> Packages</a></li>
<li class="chapter" data-level="42.1.2" data-path="alignment-in-r.html"><a href="alignment-in-r.html#data-1"><i class="fa fa-check"></i><b>42.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="43" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html"><i class="fa fa-check"></i><b>43</b> Testing the significance of an alignment</a>
<ul>
<li class="chapter" data-level="43.1" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#calculating-the-statistical-significance-of-a-pairwise-global-alignment"><i class="fa fa-check"></i><b>43.1</b> Calculating the statistical significance of a pairwise global alignment</a></li>
<li class="chapter" data-level="43.2" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#summary-1"><i class="fa fa-check"></i><b>43.2</b> Summary</a></li>
<li class="chapter" data-level="43.3" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#links-and-further-reading-1"><i class="fa fa-check"></i><b>43.3</b> Links and Further Reading</a></li>
<li class="chapter" data-level="43.4" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#exercises-1"><i class="fa fa-check"></i><b>43.4</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="44" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>44</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="44.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>44.1</b> Prelminaries</a></li>
<li class="chapter" data-level="44.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>44.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="44.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>44.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="45" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html"><i class="fa fa-check"></i><b>45</b> Manipulating matrices and vectors: worked example</a>
<ul>
<li class="chapter" data-level="45.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#introduction-7"><i class="fa fa-check"></i><b>45.1</b> Introduction</a></li>
<li class="chapter" data-level="45.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#preliminaries-10"><i class="fa fa-check"></i><b>45.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="45.2.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-3"><i class="fa fa-check"></i><b>45.2.1</b> Packages</a></li>
<li class="chapter" data-level="45.2.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#vocab-4"><i class="fa fa-check"></i><b>45.2.2</b> Vocab</a></li>
<li class="chapter" data-level="45.2.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#functions-used-1"><i class="fa fa-check"></i><b>45.2.3</b> Functions used</a></li>
<li class="chapter" data-level="45.2.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-4"><i class="fa fa-check"></i><b>45.2.4</b> Packages</a></li>
<li class="chapter" data-level="45.2.5" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#data-2"><i class="fa fa-check"></i><b>45.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="45.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#matrix-elements-can-be-accessed-with-square-bracket-notation"><i class="fa fa-check"></i><b>45.3</b> Matrix elements can be accessed with square bracket notation</a></li>
<li class="chapter" data-level="45.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#r-object-names"><i class="fa fa-check"></i><b>45.4</b> R object names</a>
<ul>
<li class="chapter" data-level="45.4.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-vectors"><i class="fa fa-check"></i><b>45.4.1</b> Accessing items in names vectors</a></li>
<li class="chapter" data-level="45.4.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-matrices"><i class="fa fa-check"></i><b>45.4.2</b> Accessing items in names matrices</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="46" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>46</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="46.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-11"><i class="fa fa-check"></i><b>46.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="46.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages-5"><i class="fa fa-check"></i><b>46.1.1</b> Packages</a></li>
<li class="chapter" data-level="46.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-5"><i class="fa fa-check"></i><b>46.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="46.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>46.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="46.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>46.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="46.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>46.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="46.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>46.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="46.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>46.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="46.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>46.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="46.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>46.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="46.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>46.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="47" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html"><i class="fa fa-check"></i><b>47</b> The BLOSUM scoring matrix in R</a>
<ul>
<li class="chapter" data-level="47.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#introduction-8"><i class="fa fa-check"></i><b>47.1</b> Introduction</a></li>
<li class="chapter" data-level="47.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#preliminaries-12"><i class="fa fa-check"></i><b>47.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="47.2.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#packages-6"><i class="fa fa-check"></i><b>47.2.1</b> Packages</a></li>
<li class="chapter" data-level="47.2.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#vocab-5"><i class="fa fa-check"></i><b>47.2.2</b> Vocab</a></li>
<li class="chapter" data-level="47.2.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#functions-used-2"><i class="fa fa-check"></i><b>47.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="47.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#load-packages-1"><i class="fa fa-check"></i><b>47.3</b> Load packages</a></li>
<li class="chapter" data-level="47.4" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#the-blosum-matrix"><i class="fa fa-check"></i><b>47.4</b> The BLOSUM matrix</a></li>
<li class="chapter" data-level="47.5" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#visualizing-the-blosum-matrix"><i class="fa fa-check"></i><b>47.5</b> Visualizing the BLOSUM matrix</a></li>
</ul></li>
<li class="chapter" data-level="48" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>48</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="49" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>49</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="49.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-9"><i class="fa fa-check"></i><b>49.1</b> Introduction</a></li>
<li class="chapter" data-level="49.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>49.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="49.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-6"><i class="fa fa-check"></i><b>49.2.1</b> Vocab</a></li>
<li class="chapter" data-level="49.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>49.2.2</b> R functions</a></li>
<li class="chapter" data-level="49.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>49.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="49.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>49.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="49.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>49.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="49.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>49.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="49.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>49.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="49.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>49.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="49.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>49.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="49.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>49.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="49.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>49.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="49.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>49.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="49.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>49.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="49.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>49.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="49.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>49.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="50" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>50</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="50.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-13"><i class="fa fa-check"></i><b>50.1</b> Preliminaries</a></li>
<li class="chapter" data-level="50.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-10"><i class="fa fa-check"></i><b>50.2</b> Introduction</a></li>
<li class="chapter" data-level="50.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>50.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="50.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>50.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="51" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>51</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="51.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-14"><i class="fa fa-check"></i><b>51.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="51.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>51.1.1</b> Key functions</a></li>
<li class="chapter" data-level="51.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>51.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="51.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>51.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="51.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>51.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="51.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>51.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="51.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>51.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="51.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>51.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="52" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html"><i class="fa fa-check"></i><b>52</b> Local variation in GC content</a>
<ul>
<li class="chapter" data-level="52.0.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#note-on-the-biology-in-this-section"><i class="fa fa-check"></i><b>52.0.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="52.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#vocabulary-2"><i class="fa fa-check"></i><b>52.1</b> Vocabulary</a></li>
<li class="chapter" data-level="52.2" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#reading-sequence-data-with-rentrezentrez_fetch"><i class="fa fa-check"></i><b>52.2</b> Reading sequence data with rentrez::entrez_fetch</a></li>
<li class="chapter" data-level="52.3" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#local-variation-in-gc-content-1"><i class="fa fa-check"></i><b>52.3</b> Local variation in GC content</a></li>
<li class="chapter" data-level="52.4" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-analysis-of-gc-content"><i class="fa fa-check"></i><b>52.4</b> A sliding window analysis of GC content</a></li>
<li class="chapter" data-level="52.5" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-plot-of-gc-content"><i class="fa fa-check"></i><b>52.5</b> A sliding window plot of GC content</a>
<ul>
<li class="chapter" data-level="52.5.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#over-represented-and-under-represented-dna-words"><i class="fa fa-check"></i><b>52.5.1</b> Over-represented and under-represented DNA words</a></li>
</ul></li>
<li class="chapter" data-level="52.6" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#summary-2"><i class="fa fa-check"></i><b>52.6</b> Summary</a></li>
<li class="chapter" data-level="52.7" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#further-reading"><i class="fa fa-check"></i><b>52.7</b> Further Reading</a></li>
<li class="chapter" data-level="52.8" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#acknowledgements-1"><i class="fa fa-check"></i><b>52.8</b> Acknowledgements</a></li>
<li class="chapter" data-level="52.9" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#exercises-2"><i class="fa fa-check"></i><b>52.9</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="53" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html"><i class="fa fa-check"></i><b>53</b> Computational gene finding</a>
<ul>
<li class="chapter" data-level="53.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#preliminaries-15"><i class="fa fa-check"></i><b>53.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="53.1.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#packages-7"><i class="fa fa-check"></i><b>53.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="53.2" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#the-genetic-code"><i class="fa fa-check"></i><b>53.2</b> The genetic code</a></li>
<li class="chapter" data-level="53.3" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-start-and-stop-codons-in-a-dna-sequence"><i class="fa fa-check"></i><b>53.3</b> Finding start and stop codons in a DNA sequence</a></li>
<li class="chapter" data-level="53.4" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#reading-frames"><i class="fa fa-check"></i><b>53.4</b> Reading frames</a></li>
<li class="chapter" data-level="53.5" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-forward-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>53.5</b> Finding open reading frames on the forward strand of a DNA sequence</a></li>
<li class="chapter" data-level="53.6" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#predicting-the-protein-sequence-for-an-orf"><i class="fa fa-check"></i><b>53.6</b> Predicting the protein sequence for an ORF</a></li>
<li class="chapter" data-level="53.7" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-reverse-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>53.7</b> Finding open reading frames on the reverse strand of a DNA sequence</a></li>
<li class="chapter" data-level="53.8" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#lengths-of-open-reading-frames"><i class="fa fa-check"></i><b>53.8</b> Lengths of open reading frames</a></li>
</ul></li>
<li class="chapter" data-level="54" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html"><i class="fa fa-check"></i><b>54</b> Significance of ORFs</a>
<ul>
<li class="chapter" data-level="54.1" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#preliminaries-16"><i class="fa fa-check"></i><b>54.1</b> Preliminaries</a></li>
<li class="chapter" data-level="54.2" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#identifying-significant-open-reading-frames"><i class="fa fa-check"></i><b>54.2</b> Identifying significant open reading frames</a></li>
<li class="chapter" data-level="54.3" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#summary-3"><i class="fa fa-check"></i><b>54.3</b> Summary</a></li>
<li class="chapter" data-level="54.4" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#links-and-further-reading-2"><i class="fa fa-check"></i><b>54.4</b> Links and Further Reading</a></li>
<li class="chapter" data-level="54.5" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#acknowledgements-2"><i class="fa fa-check"></i><b>54.5</b> Acknowledgements</a></li>
<li class="chapter" data-level="54.6" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#exercises-3"><i class="fa fa-check"></i><b>54.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="55" data-path="comparative-genomics.html"><a href="comparative-genomics.html"><i class="fa fa-check"></i><b>55</b> Comparative Genomics</a>
<ul>
<li class="chapter" data-level="55.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#preliminaries-17"><i class="fa fa-check"></i><b>55.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="55.1.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#packages-8"><i class="fa fa-check"></i><b>55.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="55.2" data-path="comparative-genomics.html"><a href="comparative-genomics.html#introduction-11"><i class="fa fa-check"></i><b>55.2</b> Introduction</a></li>
<li class="chapter" data-level="55.3" data-path="comparative-genomics.html"><a href="comparative-genomics.html#using-the-biomart-r-library-to-query-the-ensembl-database"><i class="fa fa-check"></i><b>55.3</b> Using the biomaRt R Library to Query the Ensembl Database</a></li>
<li class="chapter" data-level="55.4" data-path="comparative-genomics.html"><a href="comparative-genomics.html#comparing-the-number-of-genes-in-two-species"><i class="fa fa-check"></i><b>55.4</b> Comparing the number of genes in two species</a></li>
<li class="chapter" data-level="55.5" data-path="comparative-genomics.html"><a href="comparative-genomics.html#identifying-homologous-genes-between-two-species"><i class="fa fa-check"></i><b>55.5</b> Identifying homologous genes between two species</a></li>
</ul></li>
<li class="chapter" data-level="56" data-path="summary-4.html"><a href="summary-4.html"><i class="fa fa-check"></i><b>56</b> Summary</a></li>
<li class="chapter" data-level="57" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html"><i class="fa fa-check"></i><b>57</b> #Links and Further Reading</a>
<ul>
<li class="chapter" data-level="57.1" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#acknowledgements-3"><i class="fa fa-check"></i><b>57.1</b> Acknowledgements</a></li>
<li class="chapter" data-level="57.2" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#exercises-4"><i class="fa fa-check"></i><b>57.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="58" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html"><i class="fa fa-check"></i><b>58</b> Hidden Markov Models</a>
<ul>
<li class="chapter" data-level="58.1" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-multinomial-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>58.1</b> A multinomial model of DNA sequence evolution</a></li>
<li class="chapter" data-level="58.2" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-multinomial-model"><i class="fa fa-check"></i><b>58.2</b> Generating a DNA sequence using a multinomial model</a></li>
<li class="chapter" data-level="58.3" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>58.3</b> A Markov model of DNA sequence evolution</a></li>
<li class="chapter" data-level="58.4" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-for-a-markov-model"><i class="fa fa-check"></i><b>58.4</b> The transition matrix for a Markov model</a></li>
<li class="chapter" data-level="58.5" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-markov-model"><i class="fa fa-check"></i><b>58.5</b> Generating a DNA sequence using a Markov model</a></li>
<li class="chapter" data-level="58.6" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>58.6</b> A Hidden Markov Model of DNA sequence evolution</a></li>
<li class="chapter" data-level="58.7" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-and-emission-matrix-for-a-hmm"><i class="fa fa-check"></i><b>58.7</b> The transition matrix and emission matrix for a HMM</a></li>
<li class="chapter" data-level="58.8" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-hmm"><i class="fa fa-check"></i><b>58.8</b> Generating a DNA sequence using a HMM</a></li>
<li class="chapter" data-level="58.9" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#inferring-the-states-of-a-hmm-that-generated-a-dna-sequence"><i class="fa fa-check"></i><b>58.9</b> Inferring the states of a HMM that generated a DNA sequence</a></li>
<li class="chapter" data-level="58.10" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-protein-sequence-evolution"><i class="fa fa-check"></i><b>58.10</b> A Hidden Markov Model of protein sequence evolution</a></li>
<li class="chapter" data-level="58.11" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#summary-5"><i class="fa fa-check"></i><b>58.11</b> Summary</a></li>
<li class="chapter" data-level="58.12" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#links-and-further-reading-4"><i class="fa fa-check"></i><b>58.12</b> Links and Further Reading</a></li>
<li class="chapter" data-level="58.13" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#exercises-5"><i class="fa fa-check"></i><b>58.13</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="59" data-path="blast-summary.html"><a href="blast-summary.html"><i class="fa fa-check"></i><b>59</b> BLAST Summary</a>
<ul>
<li class="chapter" data-level="59.1" data-path="blast-summary.html"><a href="blast-summary.html#blast-overview"><i class="fa fa-check"></i><b>59.1</b> BLAST Overview</a>
<ul>
<li class="chapter" data-level="59.1.1" data-path="blast-summary.html"><a href="blast-summary.html#why-do-a-blast-search"><i class="fa fa-check"></i><b>59.1.1</b> Why do a BLAST search?</a></li>
<li class="chapter" data-level="59.1.2" data-path="blast-summary.html"><a href="blast-summary.html#what-does-blast-let-you-do"><i class="fa fa-check"></i><b>59.1.2</b> What does BLAST let you do?</a></li>
</ul></li>
<li class="chapter" data-level="59.2" data-path="blast-summary.html"><a href="blast-summary.html#blast-recipe"><i class="fa fa-check"></i><b>59.2</b> BLAST recipe</a></li>
<li class="chapter" data-level="59.3" data-path="blast-summary.html"><a href="blast-summary.html#modules-in-the-blast-workflow-and-algorithm"><i class="fa fa-check"></i><b>59.3</b> Modules in the BLAST Workflow and Algorithm</a>
<ul>
<li class="chapter" data-level="59.3.1" data-path="blast-summary.html"><a href="blast-summary.html#module-1-setup-query-sequence-protein-blast-oriented"><i class="fa fa-check"></i><b>59.3.1</b> Module 1: Setup query sequence (Protein BLAST oriented)</a></li>
<li class="chapter" data-level="59.3.2" data-path="blast-summary.html"><a href="blast-summary.html#module-2-scanning-database-for-high-scoring-words"><i class="fa fa-check"></i><b>59.3.2</b> Module 2: Scanning database for high-scoring words</a></li>
<li class="chapter" data-level="59.3.3" data-path="blast-summary.html"><a href="blast-summary.html#module-3-traceback"><i class="fa fa-check"></i><b>59.3.3</b> Module 3: Traceback</a></li>
</ul></li>
<li class="chapter" data-level="59.4" data-path="blast-summary.html"><a href="blast-summary.html#blast-alignment-statistics"><i class="fa fa-check"></i><b>59.4</b> BLAST alignment statistics</a></li>
</ul></li>
<li class="chapter" data-level="60" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html"><i class="fa fa-check"></i><b>60</b> BLAST Alignment score distribution</a>
<ul>
<li class="chapter" data-level="60.1" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#packages-9"><i class="fa fa-check"></i><b>60.1</b> Packages</a></li>
<li class="chapter" data-level="60.2" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#distributions-of-alignment-scores"><i class="fa fa-check"></i><b>60.2</b> Distributions of alignment scores</a></li>
<li class="chapter" data-level="60.3" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-data"><i class="fa fa-check"></i><b>60.3</b> Simulating data</a></li>
<li class="chapter" data-level="60.4" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#lots-of-data-in-biology-is-normal"><i class="fa fa-check"></i><b>60.4</b> Lots of data in biology is normal</a></li>
<li class="chapter" data-level="60.5" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-the-normal-distribution"><i class="fa fa-check"></i><b>60.5</b> Simulating the normal distribution</a></li>
<li class="chapter" data-level="60.6" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#the-extreme-value-distribution"><i class="fa fa-check"></i><b>60.6</b> The extreme value distribution</a></li>
<li class="chapter" data-level="60.7" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#replicating-figure-4.14"><i class="fa fa-check"></i><b>60.7</b> Replicating Figure 4.14</a></li>
<li class="chapter" data-level="60.8" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#why-do-we-care"><i class="fa fa-check"></i><b>60.8</b> Why do we care?</a></li>
</ul></li>
<li class="chapter" data-level="61" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><i class="fa fa-check"></i><b>61</b> Calculating E values, Bit scores, and P-values from BLAST output</a>
<ul>
<li class="chapter" data-level="61.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-12"><i class="fa fa-check"></i><b>61.1</b> Introduction</a></li>
<li class="chapter" data-level="61.2" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#case-study"><i class="fa fa-check"></i><b>61.2</b> Case Study</a></li>
<li class="chapter" data-level="61.3" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-statistics"><i class="fa fa-check"></i><b>61.3</b> BLAST Statistics</a></li>
<li class="chapter" data-level="61.4" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot"><i class="fa fa-check"></i><b>61.4</b> Dotplot</a>
<ul>
<li class="chapter" data-level="61.4.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot-in-r---optional-experimental-code"><i class="fa fa-check"></i><b>61.4.1</b> Dotplot in R - optional, experimental code</a></li>
</ul></li>
<li class="chapter" data-level="61.5" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-search-summary"><i class="fa fa-check"></i><b>61.5</b> BLAST “Search Summary”</a></li>
<li class="chapter" data-level="61.6" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-alignments"><i class="fa fa-check"></i><b>61.6</b> BLAST Alignments</a></li>
<li class="chapter" data-level="61.7" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#equations"><i class="fa fa-check"></i><b>61.7</b> Equations</a>
<ul>
<li class="chapter" data-level="61.7.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-value-eq-4.5-pg-142"><i class="fa fa-check"></i><b>61.7.1</b> E value (eq 4.5, pg 142)</a></li>
</ul></li>
<li class="chapter" data-level="61.8" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bit-scores-s"><i class="fa fa-check"></i><b>61.8</b> Bit Scores (S’)</a></li>
<li class="chapter" data-level="61.9" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-from-bit-scores"><i class="fa fa-check"></i><b>61.9</b> E from Bit scores</a></li>
<li class="chapter" data-level="61.10" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#p-values"><i class="fa fa-check"></i><b>61.10</b> P values</a></li>
<li class="chapter" data-level="61.11" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#repliminaries"><i class="fa fa-check"></i><b>61.11</b> Repliminaries</a></li>
<li class="chapter" data-level="61.12" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#concepts"><i class="fa fa-check"></i><b>61.12</b> Concepts</a></li>
<li class="chapter" data-level="61.13" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#functions-6"><i class="fa fa-check"></i><b>61.13</b> Functions</a></li>
<li class="chapter" data-level="61.14" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-13"><i class="fa fa-check"></i><b>61.14</b> Introduction</a></li>
<li class="chapter" data-level="61.15" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees"><i class="fa fa-check"></i><b>61.15</b> Number of rooted trees</a></li>
<li class="chapter" data-level="61.16" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees-in-r"><i class="fa fa-check"></i><b>61.16</b> Number of rooted trees in R</a>
<ul>
<li class="chapter" data-level="61.16.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#unique-topologies-versus-rotations"><i class="fa fa-check"></i><b>61.16.1</b> Unique topologies versus rotations</a></li>
</ul></li>
<li class="chapter" data-level="61.17" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bilography"><i class="fa fa-check"></i><b>61.17</b> Bilography</a></li>
</ul></li>
<li class="chapter" data-level="62" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html"><i class="fa fa-check"></i><b>62</b> UPGMA the hard way</a>
<ul>
<li class="chapter" data-level="62.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#introudction-to-upgma"><i class="fa fa-check"></i><b>62.1</b> Introudction to UPGMA</a></li>
<li class="chapter" data-level="62.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#preliminaries-18"><i class="fa fa-check"></i><b>62.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="62.2.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#r-libraries"><i class="fa fa-check"></i><b>62.2.1</b> R Libraries</a></li>
<li class="chapter" data-level="62.2.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#custom-function-well-use"><i class="fa fa-check"></i><b>62.2.2</b> Custom function we’ll use</a></li>
</ul></li>
<li class="chapter" data-level="62.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#data-matrix"><i class="fa fa-check"></i><b>62.3</b> Data matrix</a>
<ul>
<li class="chapter" data-level="62.3.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#building-the-matrix"><i class="fa fa-check"></i><b>62.3.1</b> Building the matrix</a></li>
</ul></li>
<li class="chapter" data-level="62.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#what-does-this-distance-matrix-represent"><i class="fa fa-check"></i><b>62.4</b> What does this distance matrix represent?</a>
<ul>
<li class="chapter" data-level="62.4.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-pairwise-linear-distances"><i class="fa fa-check"></i><b>62.4.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="62.4.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-2d-map-of-linear-distances"><i class="fa fa-check"></i><b>62.4.2</b> Genetic distances conceptualized as 2D map of linear distances</a></li>
</ul></li>
<li class="chapter" data-level="62.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#from-linear-distances-to-tree-distances"><i class="fa fa-check"></i><b>62.5</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="62.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#the-upgma-algorithm"><i class="fa fa-check"></i><b>62.6</b> The UPGMA algorithm</a>
<ul>
<li class="chapter" data-level="62.6.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-1-find-the-2-taxa-that-are-currently-closest-together"><i class="fa fa-check"></i><b>62.6.1</b> Step 1: find the 2 taxa that are currently closest together</a></li>
<li class="chapter" data-level="62.6.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node"><i class="fa fa-check"></i><b>62.6.2</b> Step 2: calculate the branch lengths connect the taxa to their shared node</a></li>
<li class="chapter" data-level="62.6.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-determine-if-the-algorithm-is-done"><i class="fa fa-check"></i><b>62.6.3</b> Step 3a: determine if the algorithm is done</a></li>
<li class="chapter" data-level="62.6.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u"><i class="fa fa-check"></i><b>62.6.4</b> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</a></li>
<li class="chapter" data-level="62.6.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-4-determine-the-distance-from-u-to-the-other-remaining-clusters."><i class="fa fa-check"></i><b>62.6.5</b> Step 4: Determine the distance from u to the other remaining clusters.</a></li>
<li class="chapter" data-level="62.6.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-5-create-a-new-matrix"><i class="fa fa-check"></i><b>62.6.6</b> Step 5: Create a new matrix</a></li>
<li class="chapter" data-level="62.6.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa"><i class="fa fa-check"></i><b>62.6.7</b> Key step: calculating the distance from a clade to the remaining taxa</a></li>
</ul></li>
<li class="chapter" data-level="62.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-2d-representation"><i class="fa fa-check"></i><b>62.7</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="62.8" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---round-1"><i class="fa fa-check"></i><b>62.8</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="62.8.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#determine-minimum-value-in-current-matrix"><i class="fa fa-check"></i><b>62.8.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="62.8.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade"><i class="fa fa-check"></i><b>62.8.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="62.8.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length"><i class="fa fa-check"></i><b>62.8.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="62.8.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>62.8.4</b> Calculate distance from clade to all other points</a></li>
<li class="chapter" data-level="62.8.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-matrix-with-new-distance"><i class="fa fa-check"></i><b>62.8.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="62.9" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---next-round"><i class="fa fa-check"></i><b>62.9</b> Algorithm - next round</a>
<ul>
<li class="chapter" data-level="62.9.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-1"><i class="fa fa-check"></i><b>62.9.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="62.9.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-1"><i class="fa fa-check"></i><b>62.9.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="62.9.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#recalculate-distance-from-abe-to-all-remainign-taxa"><i class="fa fa-check"></i><b>62.9.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="62.9.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-the-matrix"><i class="fa fa-check"></i><b>62.9.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="62.10" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#next-iteration"><i class="fa fa-check"></i><b>62.10</b> Next iteration</a>
<ul>
<li class="chapter" data-level="62.10.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-lengths"><i class="fa fa-check"></i><b>62.10.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="62.10.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-2"><i class="fa fa-check"></i><b>62.10.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="62.10.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-2"><i class="fa fa-check"></i><b>62.10.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="62.10.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#distance-of-c-to-abe-and-d-to-abe"><i class="fa fa-check"></i><b>62.10.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="62.11" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finish-up"><i class="fa fa-check"></i><b>62.11</b> Finish up</a>
<ul>
<li class="chapter" data-level="62.11.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-3"><i class="fa fa-check"></i><b>62.11.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="62.11.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-3"><i class="fa fa-check"></i><b>62.11.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="62.12" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finalizing-branch-lengths"><i class="fa fa-check"></i><b>62.12</b> Finalizing branch lengths</a></li>
</ul></li>
<li class="chapter" data-level="63" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html"><i class="fa fa-check"></i><b>63</b> Representing phylogenetic trees in Newick format</a>
<ul>
<li class="chapter" data-level="63.1" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#vocab-7"><i class="fa fa-check"></i><b>63.1</b> Vocab</a></li>
<li class="chapter" data-level="63.2" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#introduction-14"><i class="fa fa-check"></i><b>63.2</b> Introduction</a></li>
<li class="chapter" data-level="63.3" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-three-taxa"><i class="fa fa-check"></i><b>63.3</b> Trees with three taxa</a></li>
<li class="chapter" data-level="63.4" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-four-taxa"><i class="fa fa-check"></i><b>63.4</b> Trees with four taxa</a></li>
</ul></li>
<li class="chapter" data-level="64" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html"><i class="fa fa-check"></i><b>64</b> Plotting trees from Newick notation in R</a>
<ul>
<li class="chapter" data-level="64.1" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#preliminaries-19"><i class="fa fa-check"></i><b>64.1</b> Preliminaries</a></li>
<li class="chapter" data-level="64.2" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#note"><i class="fa fa-check"></i><b>64.2</b> Note</a></li>
<li class="chapter" data-level="64.3" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick"><i class="fa fa-check"></i><b>64.3</b> 3 taxa tree in Newick</a></li>
<li class="chapter" data-level="64.4" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick-1"><i class="fa fa-check"></i><b>64.4</b> 4 taxa tree in Newick</a></li>
<li class="chapter" data-level="64.5" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-in-newick-format"><i class="fa fa-check"></i><b>64.5</b> 5 taxa in Newick Format</a></li>
<li class="chapter" data-level="64.6" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#many-taxa-tree"><i class="fa fa-check"></i><b>64.6</b> Many-taxa tree</a></li>
<li class="chapter" data-level="64.7" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#newick-and-branch-lengths"><i class="fa fa-check"></i><b>64.7</b> Newick and Branch lengths</a></li>
<li class="chapter" data-level="64.8" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#additional-reading"><i class="fa fa-check"></i><b>64.8</b> Additional Reading</a></li>
</ul></li>
<li class="chapter" data-level="65" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><i class="fa fa-check"></i><b>65</b> Phylogenetic trees using the UPGMA clustering algoirthm</a>
<ul>
<li class="chapter" data-level="65.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#introduction-15"><i class="fa fa-check"></i><b>65.1</b> Introduction</a></li>
<li class="chapter" data-level="65.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#preliminaries-20"><i class="fa fa-check"></i><b>65.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="65.2.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#vocab-8"><i class="fa fa-check"></i><b>65.2.1</b> Vocab</a></li>
<li class="chapter" data-level="65.2.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-r-functions"><i class="fa fa-check"></i><b>65.2.2</b> General R functions</a></li>
<li class="chapter" data-level="65.2.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#specific-functions"><i class="fa fa-check"></i><b>65.2.3</b> Specific functions</a></li>
</ul></li>
<li class="chapter" data-level="65.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#sequence-comparisons"><i class="fa fa-check"></i><b>65.3</b> Sequence comparisons</a></li>
<li class="chapter" data-level="65.4" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrices"><i class="fa fa-check"></i><b>65.4</b> Distance matrices</a></li>
<li class="chapter" data-level="65.5" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#from-a-distance-matrix-to-phylogenetic-tree"><i class="fa fa-check"></i><b>65.5</b> From a distance matrix to phylogenetic tree</a></li>
<li class="chapter" data-level="65.6" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-raw-sequences"><i class="fa fa-check"></i><b>65.6</b> Distance matrix in R from raw sequences</a></li>
<li class="chapter" data-level="65.7" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-normal-matrix"><i class="fa fa-check"></i><b>65.7</b> Distance matrix in R from normal matrix</a></li>
<li class="chapter" data-level="65.8" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#upgma-in-r"><i class="fa fa-check"></i><b>65.8</b> UPGMA in R</a></li>
<li class="chapter" data-level="65.9" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#how-upgma-clustering-works"><i class="fa fa-check"></i><b>65.9</b> How UPGMA clustering works</a>
<ul>
<li class="chapter" data-level="65.9.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#challenge-2"><i class="fa fa-check"></i><b>65.9.1</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="65.10" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#further-information"><i class="fa fa-check"></i><b>65.10</b> Further information</a>
<ul>
<li class="chapter" data-level="65.10.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-info"><i class="fa fa-check"></i><b>65.10.1</b> General info</a></li>
<li class="chapter" data-level="65.10.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#worked-examples"><i class="fa fa-check"></i><b>65.10.2</b> Worked examples:</a></li>
<li class="chapter" data-level="65.10.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#additional-notes"><i class="fa fa-check"></i><b>65.10.3</b> Additional notes:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="66" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><i class="fa fa-check"></i><b>66</b> Tutorial: Writing functions to calculate the number of phylogenetic trees</a>
<ul>
<li class="chapter" data-level="66.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#repliminaries-1"><i class="fa fa-check"></i><b>66.1</b> Repliminaries</a>
<ul>
<li class="chapter" data-level="66.1.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-7"><i class="fa fa-check"></i><b>66.1.1</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="66.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#concepts-1"><i class="fa fa-check"></i><b>66.2</b> Concepts</a>
<ul>
<li class="chapter" data-level="66.2.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vocab-9"><i class="fa fa-check"></i><b>66.2.1</b> Vocab:</a></li>
<li class="chapter" data-level="66.2.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#packages-10"><i class="fa fa-check"></i><b>66.2.2</b> Packages:</a></li>
</ul></li>
<li class="chapter" data-level="66.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#introduction-16"><i class="fa fa-check"></i><b>66.3</b> Introduction</a></li>
<li class="chapter" data-level="66.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-1"><i class="fa fa-check"></i><b>66.4</b> Number of rooted trees</a></li>
<li class="chapter" data-level="66.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-in-r-1"><i class="fa fa-check"></i><b>66.5</b> Number of rooted trees in R</a></li>
<li class="chapter" data-level="66.6" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-in-r-1"><i class="fa fa-check"></i><b>66.6</b> Functions in R</a></li>
<li class="chapter" data-level="66.7" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-to-calculate-the-number-of-possible-phylogenetic-trees"><i class="fa fa-check"></i><b>66.7</b> Function to calculate the number of possible phylogenetic trees</a></li>
<li class="chapter" data-level="66.8" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vectorized-inputs-to-functions"><i class="fa fa-check"></i><b>66.8</b> Vectorized inputs to functions</a></li>
<li class="chapter" data-level="66.9" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-conditional-statements"><i class="fa fa-check"></i><b>66.9</b> Adding conditional statements</a></li>
<li class="chapter" data-level="66.10" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-multiple-conditional-statements"><i class="fa fa-check"></i><b>66.10</b> Adding multiple conditional statements</a></li>
<li class="chapter" data-level="66.11" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-additional-arguments"><i class="fa fa-check"></i><b>66.11</b> Adding additional arguments</a></li>
<li class="chapter" data-level="66.12" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-_un_rooted-trees"><i class="fa fa-check"></i><b>66.12</b> Assignment: _un_rooted trees</a></li>
<li class="chapter" data-level="66.13" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-1"><i class="fa fa-check"></i><b>66.13</b> Assignment part 1</a>
<ul>
<li class="chapter" data-level="66.13.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#part-1-answer"><i class="fa fa-check"></i><b>66.13.1</b> Part 1 answer</a></li>
</ul></li>
<li class="chapter" data-level="66.14" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-2"><i class="fa fa-check"></i><b>66.14</b> Assignment part 2</a>
<ul>
<li class="chapter" data-level="66.14.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#answer-1"><i class="fa fa-check"></i><b>66.14.1</b> Answer</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="67" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>67</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="67.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>67.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="67.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>67.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="67.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>67.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="67.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>67.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="67.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>67.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-3"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="67.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>67.2</b> Help!</a>
<ul>
<li class="chapter" data-level="67.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>67.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="67.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>67.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="67.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>67.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="67.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>67.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="67.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>67.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="67.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>67.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="67.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>67.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="67.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>67.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="comparative-genomics" class="section level1" number="55">
<h1><span class="header-section-number">Chapter 55</span> Comparative Genomics</h1>
<p><strong>By</strong>: Avril Coghlan.</p>
<p><strong>Adapted, edited and expanded</strong>: Nathan Brouwer under the Creative Commons 3.0 Attribution License <a href="https://creativecommons.org/licenses/by/3.0/">(CC BY 3.0)</a>, with assistance from Aman Virmani.</p>
<p><strong>NOTE: this chapter has not yet been significaly revised</strong>.</p>
<div id="preliminaries-17" class="section level2" number="55.1">
<h2><span class="header-section-number">55.1</span> Preliminaries</h2>
<div id="packages-8" class="section level3" number="55.1.1">
<h3><span class="header-section-number">55.1.1</span> Packages</h3>
<div class="sourceCode" id="cb1216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1216-1"><a href="comparative-genomics.html#cb1216-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compbio4all)</span>
<span id="cb1216-2"><a href="comparative-genomics.html#cb1216-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1216-3"><a href="comparative-genomics.html#cb1216-3" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install(&quot;biomaRt&quot;)</span></span>
<span id="cb1216-4"><a href="comparative-genomics.html#cb1216-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;biomaRt&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:seqinr&#39;:
## 
##     getSequence</code></pre>
</div>
</div>
<div id="introduction-11" class="section level2" number="55.2">
<h2><span class="header-section-number">55.2</span> Introduction</h2>
<p><strong>Comparative genomics</strong> is the field of bioinformatics that involves comparing the genomes of two different species, or of two different strains of the same species.</p>
<p>One of the first questions to ask when comparing the genomes of two species is: do the two species have the same number of genes (i.e.. the same gene content)? Since all life on earth shared a common ancestor at some point, any two species, for example, human and a fruit fly, must have descended from a common ancestor species.</p>
<p>Since the time of the common ancestor of two species (e.g. of human and mouse), some of the genes that were present in the common ancestor species may have been lost from either of the two descendant lineages. Furthermore, the two descendant lineages may have gained genes that were not present in the common ancestor species.</p>
</div>
<div id="using-the-biomart-r-library-to-query-the-ensembl-database" class="section level2" number="55.3">
<h2><span class="header-section-number">55.3</span> Using the biomaRt R Library to Query the Ensembl Database</h2>
<p>To carry out comparative genomic analyses of two animal species whose genomes have been fully sequenced (e.g. human and mouse), it is useful to analyze the data in the Ensembl database (www.ensembl.org).</p>
<p>The main Ensembl database which you can browse on the main Ensembl webpage contains genes from fully sequenced vertebrates, as well as <em>Saccharomyces cerevisiae</em> (yeast) and a small number of additional model organism animals (e.g. the nematode worm <em>Caenorhabditis elegans</em> and the fruit fly <em>Drosophila melanogaster</em>).</p>
<p>There are also Ensembl databases for other groups of organisms, for example Ensembl Protists for Protists, Ensembl Metazoa for Metazoans, Ensembl Bacteria for Bacteria, Ensembl Plants for Plants, and Ensembl Fungi for Fungi.</p>
<p>It is possible to carry out analyses of the Ensembl database using R, with the <code>biomaRt</code> R package. The <code>biomaRt</code> package can connect to the Ensembl database, and perform queries on the data.</p>
<p>The <code>biomaRt</code> <em>R</em> package is part of the Bioconductor set of <em>R</em> packages, and so can be installed as explained here.</p>
<p>Once you have installed the <code>biomaRt</code> package, you can get a list of databases that can be queried using this package by typing:</p>
<div class="sourceCode" id="cb1219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1219-1"><a href="comparative-genomics.html#cb1219-1" aria-hidden="true" tabindex="-1"></a><span class="fu">listEnsemblGenomes</span>()        <span class="co"># List all Ensembl genomes that can be queried</span></span></code></pre></div>
<pre><code>##               biomart                        version
## 1       protists_mart      Ensembl Protists Genes 51
## 2 protists_variations Ensembl Protists Variations 51
## 3          fungi_mart         Ensembl Fungi Genes 51
## 4    fungi_variations    Ensembl Fungi Variations 51
## 5        metazoa_mart       Ensembl Metazoa Genes 51
## 6  metazoa_variations  Ensembl Metazoa Variations 51
## 7         plants_mart        Ensembl Plants Genes 51
## 8   plants_variations   Ensembl Plants Variations 51</code></pre>
<p>The names of the databases are listed, and the version of the database.</p>
<p>The <code>biomaRt</code> <em>R</em> package can actually be used to query many different databases including WormBase, UniProt, Ensembl, etc.</p>
<p>Here, we will discuss using the <code>biomaRt</code> package to query the Ensembl database, but it is worth remembering that it also be used to perform queries on other databases such as UniProt.</p>
<p>If you want to perform a query on the Ensembl database using <code>biomaRt</code>, you first need to specify that this is the database that you want to query. You can do this using the <code>useEnsemblGenomes()</code> function from the <code>biomaRt</code> package:</p>
<div class="sourceCode" id="cb1221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1221-1"><a href="comparative-genomics.html#cb1221-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Specify that we want to query the Ensembl database</span></span>
<span id="cb1221-2"><a href="comparative-genomics.html#cb1221-2" aria-hidden="true" tabindex="-1"></a>ensemblprotists <span class="ot">&lt;-</span> <span class="fu">useEnsemblGenomes</span>(<span class="st">&quot;protists_mart&quot;</span>) </span></code></pre></div>
<p>This tells <code>biomaRt</code> that you want to query the Ensembl Protists database. The Ensembl Protists database contains data sets of genomic information for different protist species whose genomes have been fully sequenced.</p>
<p>To see which data sets you can query in the database that you have selected (using <code>useDataset()</code>), you can type:</p>
<div class="sourceCode" id="cb1222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1222-1"><a href="comparative-genomics.html#cb1222-1" aria-hidden="true" tabindex="-1"></a><span class="fu">listDatasets</span>(ensemblprotists)</span></code></pre></div>
<pre><code>##                    dataset
## 1       alaibachii_eg_gene
## 2          bnatans_eg_gene
## 3      ddiscoideum_eg_gene
## 4     ehistolytica_eg_gene
## 5         ehuxleyi_eg_gene
## 6         glamblia_eg_gene
## 7           gtheta_eg_gene
## 8   harabidopsidis_eg_gene
## 9           lmajor_eg_gene
## 10 paphanidermatum_eg_gene
## 11   parrhenomanes_eg_gene
## 12        pberghei_eg_gene
## 13       pchabaudi_eg_gene
## 14     pfalciparum_eg_gene
## 15      pinfestans_eg_gene
## 16     pirregulare_eg_gene
## 17       piwayamai_eg_gene
## 18      pkernoviae_eg_gene
## 19       pknowlesi_eg_gene
## 20      plateralis_eg_gene
## 21   pmultistriata_eg_gene
## 22     pparasitica_eg_gene
## 23        pramorum_eg_gene
## 24          psojae_eg_gene
## 25    ptetraurelia_eg_gene
## 26    ptricornutum_eg_gene
## 27        pultimum_eg_gene
## 28         pvexans_eg_gene
## 29          pvivax_eg_gene
## 30         tbrucei_eg_gene
## 31         tgondii_eg_gene
## 32     tpseudonana_eg_gene
## 33    tthermophila_eg_gene
##                                                                      description
## 1                                                 Albugo laibachii genes (ENA 1)
## 2                                             Bigelowiella natans genes (Bigna1)
## 3                                     Dictyostelium discoideum genes (dicty_2.7)
## 4                                    Entamoeba histolytica genes (JCVI-ESG2-1.0)
## 5  Emiliania huxleyi genes (Emiliana huxleyi CCMP1516 main genome assembly v1.0)
## 6                                                    Giardia lamblia genes (GL2)
## 7                                       Guillardia theta CCMP2712 genes (Guith1)
## 8                         Hyaloperonospora arabidopsidis genes (HyaAraEmoy2_2.0)
## 9                                              Leishmania major genes (ASM272v2)
## 10                              Pythium aphanidermatum genes (pag1_scaffolds_v1)
## 11                                 Pythium arrhenomanes genes (par_scaffolds_v1)
## 12                                           Plasmodium berghei genes (PBANKA01)
## 13                                           Plasmodium chabaudi genes (PCHAS01)
## 14                                    Plasmodium falciparum 3D7 genes (ASM276v2)
## 15                                     Phytophthora infestans genes (ASM14294v1)
## 16                                   Pythium irregulare genes (pir_scaffolds_v1)
## 17                                     Pythium iwayamai genes (piw_scaffolds_v1)
## 18                                Phytophthora kernoviae genes (PhyKer238/432v1)
## 19                                          Plasmodium knowlesi genes (ASM635v1)
## 20                                      Phytophthora lateralis genes (MPF4_v1.0)
## 21                           Pseudo-nitzschia multistriata genes (ASM90066040v1)
## 22                            Phytophthora parasitica genes (Phyt_para_P1569_V1)
## 23                                       Phytophthora ramorum genes (ASM14973v1)
## 24                                       Phytophthora sojae genes (P.sojae V3.0)
## 25                                     Paramecium tetraurelia genes (ASM16542v1)
## 26                                  Phaeodactylum tricornutum genes (ASM15095v2)
## 27                                                   Pythium ultimum genes (pug)
## 28                                       Pythium vexans genes (pve_scaffolds_v1)
## 29                                             Plasmodium vivax genes (ASM241v2)
## 30                               Trypanosoma brucei genes (TryBru_Apr2005_chr11)
## 31                                           Toxoplasma gondii ME49 genes (TGA4)
## 32                                   Thalassiosira pseudonana genes (ASM14940v2)
## 33                                 Tetrahymena thermophila genes (JCVI-TTA1-2.2)
##                                                version
## 1                                                ENA 1
## 2                                               Bigna1
## 3                                            dicty_2.7
## 4                                        JCVI-ESG2-1.0
## 5  Emiliana huxleyi CCMP1516 main genome assembly v1.0
## 6                                                  GL2
## 7                                               Guith1
## 8                                      HyaAraEmoy2_2.0
## 9                                             ASM272v2
## 10                                   pag1_scaffolds_v1
## 11                                    par_scaffolds_v1
## 12                                            PBANKA01
## 13                                             PCHAS01
## 14                                            ASM276v2
## 15                                          ASM14294v1
## 16                                    pir_scaffolds_v1
## 17                                    piw_scaffolds_v1
## 18                                     PhyKer238/432v1
## 19                                            ASM635v1
## 20                                           MPF4_v1.0
## 21                                       ASM90066040v1
## 22                                  Phyt_para_P1569_V1
## 23                                          ASM14973v1
## 24                                        P.sojae V3.0
## 25                                          ASM16542v1
## 26                                          ASM15095v2
## 27                                                 pug
## 28                                    pve_scaffolds_v1
## 29                                            ASM241v2
## 30                                TryBru_Apr2005_chr11
## 31                                                TGA4
## 32                                          ASM14940v2
## 33                                       JCVI-TTA1-2.2</code></pre>
<p>You will see a long list of the organisms for which the Ensembl Protists database has genome data, including <em>Plasmodium falciparium</em> (which causes malaria), and <em>Leishmania major</em>, which causes leishmaniasis, which is classified by the WHO as a neglected tropical disease.</p>
<p>To perform a query on the Ensembl database using the <code>biomaRt</code> R package, you first need to specify which Ensembl data set your query relates to. You can do this using the <code>useDataset()</code> function from the <code>biomaRt</code> package. For example, to specify that you want to perform a query on the Ensembl <em>Leishmania major</em> data set, you would type:</p>
<div class="sourceCode" id="cb1224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1224-1"><a href="comparative-genomics.html#cb1224-1" aria-hidden="true" tabindex="-1"></a>ensemblleishmania <span class="ot">&lt;-</span> <span class="fu">useDataset</span>(<span class="st">&quot;lmajor_eg_gene&quot;</span>,</span>
<span id="cb1224-2"><a href="comparative-genomics.html#cb1224-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">mart=</span>ensemblprotists)</span></code></pre></div>
<p>Note that the name of the <em>Leishmania major</em> Ensembl data set is “lmajor_eg_gene”; this is the data set listed for <em>Leishmania major</em> genomic information when we typed <code>listDatasets(ensemblprotists)</code> above.</p>
<p>Once you have specified the particular Ensembl data set that you want to perform a query on, you can perform the query using the <code>getBM()</code> function from the <code>biomaRt</code> package.</p>
<p>Usually, you will want to perform a query to a particular set of features from the <em>Leishmania major</em> Ensembl data set. What types of features can you search for? You can find this out by using the <code>listAttributes()</code> function from the <code>biomaRt</code> package:</p>
<div class="sourceCode" id="cb1225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1225-1"><a href="comparative-genomics.html#cb1225-1" aria-hidden="true" tabindex="-1"></a>leishmaniaattributes <span class="ot">&lt;-</span> <span class="fu">listAttributes</span>(ensemblleishmania)</span></code></pre></div>
<p>The <code>l</code>istAttributes()` function returns a list object, the first element of which is a vector of all possible features that you can select, and the second element of which is a vector containing explanations of all those features:</p>
<div class="sourceCode" id="cb1226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1226-1"><a href="comparative-genomics.html#cb1226-1" aria-hidden="true" tabindex="-1"></a>attributenames <span class="ot">&lt;-</span> leishmaniaattributes[[<span class="dv">1</span>]]</span>
<span id="cb1226-2"><a href="comparative-genomics.html#cb1226-2" aria-hidden="true" tabindex="-1"></a>attributedescriptions <span class="ot">&lt;-</span> leishmaniaattributes[[<span class="dv">2</span>]]</span>
<span id="cb1226-3"><a href="comparative-genomics.html#cb1226-3" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(attributenames)  <span class="co"># Find the length of vector &quot;attributenames&quot;</span></span></code></pre></div>
<pre><code>## [1] 630</code></pre>
<div class="sourceCode" id="cb1228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1228-1"><a href="comparative-genomics.html#cb1228-1" aria-hidden="true" tabindex="-1"></a>attributenames[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]    <span class="co"># Print out the first 10 entries </span></span></code></pre></div>
<pre><code>##  [1] &quot;ensembl_gene_id&quot;       &quot;ensembl_transcript_id&quot; &quot;ensembl_peptide_id&quot;   
##  [4] &quot;ensembl_exon_id&quot;       &quot;description&quot;           &quot;chromosome_name&quot;      
##  [7] &quot;start_position&quot;        &quot;end_position&quot;          &quot;strand&quot;               
## [10] &quot;band&quot;</code></pre>
<div class="sourceCode" id="cb1230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1230-1"><a href="comparative-genomics.html#cb1230-1" aria-hidden="true" tabindex="-1"></a>                        <span class="co"># in vector &quot;attributenames&quot;</span></span>
<span id="cb1230-2"><a href="comparative-genomics.html#cb1230-2" aria-hidden="true" tabindex="-1"></a>attributedescriptions[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]   <span class="co"># Print out the first 10 entries </span></span></code></pre></div>
<pre><code>##  [1] &quot;Gene stable ID&quot;           &quot;Transcript stable ID&quot;    
##  [3] &quot;Protein stable ID&quot;        &quot;Exon stable ID&quot;          
##  [5] &quot;Gene description&quot;         &quot;Chromosome/scaffold name&quot;
##  [7] &quot;Gene start (bp)&quot;          &quot;Gene end (bp)&quot;           
##  [9] &quot;Strand&quot;                   &quot;Karyotype band&quot;</code></pre>
<div class="sourceCode" id="cb1232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1232-1"><a href="comparative-genomics.html#cb1232-1" aria-hidden="true" tabindex="-1"></a>                              <span class="co"># in vector &quot;attributedescriptions&quot;</span></span></code></pre></div>
<p>This gives us a very long list of 630 features in the <em>Leishmania major</em> Ensembl data set that we can search for by querying the database, such as genes, transcripts (mRNAs), peptides (proteins), chromosomes, GO (Gene Ontology) terms, and so on.</p>
<p>When you are performing a query on the Ensembl <em>Leishmania major</em> data set using <code>getBM()</code>, you have to specify which of these features you want to retrieve. For example, you can see from the output of <code>listAttributes()</code> (see above) that one possible type of feature we can search for are <em>Leishmania major</em> genes. To retrieve a list of all <em>Leishmania major</em> genes from the <em>Leishmania major</em> Ensembl data set, we just need to type:</p>
<div class="sourceCode" id="cb1233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1233-1"><a href="comparative-genomics.html#cb1233-1" aria-hidden="true" tabindex="-1"></a>leishmaniagenes <span class="ot">&lt;-</span> <span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>), </span>
<span id="cb1233-2"><a href="comparative-genomics.html#cb1233-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">mart=</span>ensemblleishmania)</span></code></pre></div>
<p>This returns a list variable leishmaniagenes, the first element of which is a vector containing the names of all <em>Leishmania major</em> genes. Thus, to find the number of genes, and print out the names of the first ten genes stored in the vector, we can type:</p>
<div class="sourceCode" id="cb1234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1234-1"><a href="comparative-genomics.html#cb1234-1" aria-hidden="true" tabindex="-1"></a> <span class="co"># Get the vector of the names of all L. major genes</span></span>
<span id="cb1234-2"><a href="comparative-genomics.html#cb1234-2" aria-hidden="true" tabindex="-1"></a>leishmaniagenenames <span class="ot">&lt;-</span> leishmaniagenes[[<span class="dv">1</span>]]</span>
<span id="cb1234-3"><a href="comparative-genomics.html#cb1234-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1234-4"><a href="comparative-genomics.html#cb1234-4" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(leishmaniagenenames)</span></code></pre></div>
<pre><code>## [1] 10030</code></pre>
<div class="sourceCode" id="cb1236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1236-1"><a href="comparative-genomics.html#cb1236-1" aria-hidden="true" tabindex="-1"></a>leishmaniagenenames[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;ENSRNAG00049765313&quot; &quot;ENSRNAG00049765315&quot; &quot;ENSRNAG00049765317&quot;
##  [4] &quot;ENSRNAG00049765319&quot; &quot;ENSRNAG00049765321&quot; &quot;ENSRNAG00049765323&quot;
##  [7] &quot;ENSRNAG00049765325&quot; &quot;ENSRNAG00049765327&quot; &quot;ENSRNAG00049765329&quot;
## [10] &quot;ENSRNAG00049765333&quot;</code></pre>
<p>This tells us that there are 10300 different <em>Leishmania major</em> genes in the L. major Ensembl data set. Note that this includes various types of genes including protein-coding genes (both “known” and “novel” genes, where the “novel” genes are gene predictions that don’t have sequence similarity to any sequences in sequence databases), RNA genes, and pseudogenes.</p>
<p>What if we are only interested in protein-coding genes? If you look at the output of <code>listAttributes(ensemblleishmania)</code>, you will see that one of the features is “gene_biotype”, which is tells us what sort of gene each gene is (e.g. protein-coding, pseudogene, etc.):</p>
<div class="sourceCode" id="cb1238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1238-1"><a href="comparative-genomics.html#cb1238-1" aria-hidden="true" tabindex="-1"></a>leishmaniagenes2 <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;gene_biotype&quot;</span>),</span>
<span id="cb1238-2"><a href="comparative-genomics.html#cb1238-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">mart=</span>ensemblleishmania)</span></code></pre></div>
<p>In this case, the <code>getBM()</code> function will return a list variable leishmaniagenes2, the first element of which is a vector containing the names of all <em>Leishmania major</em> genes, and the second of which is a vector containing the types of those genes:</p>
<div class="sourceCode" id="cb1239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1239-1"><a href="comparative-genomics.html#cb1239-1" aria-hidden="true" tabindex="-1"></a>leishmaniagenenames2 <span class="ot">&lt;-</span> leishmaniagenes2[[<span class="dv">1</span>]] <span class="co"># Get the vector of the names of all L. major genes</span></span>
<span id="cb1239-2"><a href="comparative-genomics.html#cb1239-2" aria-hidden="true" tabindex="-1"></a>leishmaniagenebiotypes2 <span class="ot">&lt;-</span> leishmaniagenes2[[<span class="dv">2</span>]] <span class="co"># Get the vector of the biotypes of all genes</span></span></code></pre></div>
<p>We can make a table of all the different types of genes using the table() function:</p>
<div class="sourceCode" id="cb1240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1240-1"><a href="comparative-genomics.html#cb1240-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(leishmaniagenebiotypes2)</span></code></pre></div>
<pre><code>## leishmaniagenebiotypes2
##          ncRNA protein_coding     pseudogene           rRNA          snRNA 
##           1130           8315             94             92            233 
##           tRNA 
##            166</code></pre>
<p>This tells us that there are 8315 protein-coding genes, 94 pseudogenes, and various types of RNA genes (tRNA genes, rRNA genes, snRNA genes, etc.). Thus, there are 8315 human protein-coding genes.</p>
</div>
<div id="comparing-the-number-of-genes-in-two-species" class="section level2" number="55.4">
<h2><span class="header-section-number">55.4</span> Comparing the number of genes in two species</h2>
<p>Ensembl is a very useful resource for comparing the gene content of different species. For example, one simple question that we can ask by analyzing the Ensembl data is: how many protein-coding genes are there in <em>Leishmania major</em>, and how many in <em>Plasmodium falciparum</em>?</p>
<p>We know how many protein-coding genes are in <em>Leishmania major</em> (8315; see above), but what about <em>Plasmodium falciparum</em>? To answer this question, we first need to tell the <code>biomaRt</code> package that we want to make a query on the Ensembl <em>Plasmodium falciparum</em> data set.</p>
<p>We can do this using the <code>useDataset()</code> function to select the <em>Plasmodium falciparum</em> Ensembl data set.</p>
<div class="sourceCode" id="cb1242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1242-1"><a href="comparative-genomics.html#cb1242-1" aria-hidden="true" tabindex="-1"></a>ensemblpfalciparum <span class="ot">&lt;-</span> <span class="fu">useDataset</span>(<span class="st">&quot;pfalciparum_eg_gene&quot;</span>,</span>
<span id="cb1242-2"><a href="comparative-genomics.html#cb1242-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">mart=</span>ensemblprotists)</span></code></pre></div>
<p>Note that the name of the <em>Plasmodium falciparum</em> Ensembl data set is “pfalciparum_eg_gene”; this is the data set listed for <em>Plasmodium falciparum</em> genomic information when we typed listDatasets(ensemblprotists) above.</p>
<p>We can then use <code>getBM()</code> as above to retrieve the names of all <em>Plasmodium falciparum</em> protein-coding genes. This time we have to set the “mart” option in the <code>getBM()</code> function to “ensemblpfalciparum”, to specify that we want to query the <em>Plasmodium falciparum</em> Ensembl data set rather than the <em>Leishmania major</em> Ensembl data set:</p>
<div class="sourceCode" id="cb1243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1243-1"><a href="comparative-genomics.html#cb1243-1" aria-hidden="true" tabindex="-1"></a>pfalciparumgenes <span class="ot">&lt;-</span> <span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;gene_biotype&quot;</span>),</span>
<span id="cb1243-2"><a href="comparative-genomics.html#cb1243-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">mart=</span>ensemblpfalciparum)</span>
<span id="cb1243-3"><a href="comparative-genomics.html#cb1243-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1243-4"><a href="comparative-genomics.html#cb1243-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the names of the P. falciparum genes</span></span>
<span id="cb1243-5"><a href="comparative-genomics.html#cb1243-5" aria-hidden="true" tabindex="-1"></a>pfalciparumgenenames <span class="ot">&lt;-</span> pfalciparumgenes[[<span class="dv">1</span>]] </span>
<span id="cb1243-6"><a href="comparative-genomics.html#cb1243-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1243-7"><a href="comparative-genomics.html#cb1243-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the number of P. falciparum genes</span></span>
<span id="cb1243-8"><a href="comparative-genomics.html#cb1243-8" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(pfalciparumgenenames)                  </span></code></pre></div>
<pre><code>## [1] 5767</code></pre>
<div class="sourceCode" id="cb1245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1245-1"><a href="comparative-genomics.html#cb1245-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the types of the P. falciparum genes</span></span>
<span id="cb1245-2"><a href="comparative-genomics.html#cb1245-2" aria-hidden="true" tabindex="-1"></a>pfalciparumgenebiotypes <span class="ot">&lt;-</span> pfalciparumgenes[[<span class="dv">2</span>]] </span>
<span id="cb1245-3"><a href="comparative-genomics.html#cb1245-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1245-4"><a href="comparative-genomics.html#cb1245-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(pfalciparumgenebiotypes)</span></code></pre></div>
<pre><code>## pfalciparumgenebiotypes
##              ncRNA nontranslating_CDS     protein_coding         pseudogene 
##                102                  4               5358                153 
##               rRNA              snRNA               sRNA               tRNA 
##                 44                 10                 17                 79</code></pre>
<p>This tells us that there are 5767 <em>Plasmodium falciparum</em> protein-coding genes in Ensembl. That is, <em>Plasmodium falciparum</em> seems to have less protein-coding genes than <em>Leishmania major</em> (8315 protein-coding genes; see above).</p>
<p>It is interesting to ask: why does <em>Plasmodium falciparum</em> have less protein-coding genes than <em>Leishmania major</em>? There are several possible explanations: (i) that there have been gene duplications in the <em>Leishmania major</em> lineage since Leishmania and Plasmodium shared a common ancestor, which gave rise to new <em>Leishmania major</em> genes; (ii) that completely new genes (that are not related to any other <em>Leishmania major</em> gene) have arisen in the <em>Leishmania major</em> lineage since Leishmania and Plasmodium shared a common ancestor; or (iii) that there have been genes lost from the <em>Plasmodium falciparum</em> genome since Leishmania and Plasmodium shared a common ancestor.</p>
<p>To investigate which of these explanations is most likely to be correct, we need to figure out how the <em>Leishmania major</em> protein-coding genes are related to <em>Plasmodium falciparum</em> protein-coding genes.</p>
</div>
<div id="identifying-homologous-genes-between-two-species" class="section level2" number="55.5">
<h2><span class="header-section-number">55.5</span> Identifying homologous genes between two species</h2>
<p>The Ensembl database groups homologous (related) genes together into gene families. If a gene from <em>Leishmania major</em> and a gene from <em>Plasmodium falciparum</em> are related, they should be placed together into the same Ensembl gene family in the Ensembl Protists database. In fact, if a <em>Leishmania major</em> gene has any homologs (related genes) in other protists, it should be placed into some Ensembl gene family in the Ensembl Protists database.</p>
<p>For all <em>Leishmania major</em> and <em>Plasmodium falciparum</em> genes that are placed together in a gene family, Ensembl classifies the relationship between each pair of <em>Leishmania major</em> and <em>Plasmodium falciparum</em> genes as orthologs (related genes that shared a common ancestor in the ancestor of Leishmania and Plasmodium, and arose due to the <em>Leishmania</em> - <em>Plasmodium</em> speciation event) or paralogs (related genes that arose due to a duplication event within a species, for example, due to a duplication event in <em>Leishmania major</em>, or a duplication event in the Leishmania - <em>Plasmodium</em> ancestor).</p>
<p>If you type <code>listAttributes(ensemblleishmania)</code> again, you will see that one possible feature that you can search for is “pfalciparum_eg_homolog_ensembl_gene”, which is the <em>Plasmodium falciparum</em> ortholog of a <em>Leishmania major</em> gene.</p>
<p>Another possible feature that you can search for is “pfalciparum_eg_homology_orthology_type”, which describes the type of orthology relationship between a particular <em>Leishmania major</em> gene and its <em>Plasmodium falciparum</em> ortholog. For example, if a particular <em>Leishmania major</em> gene has two <em>Plasmodium falciparum</em> orthologs, the relationship between the <em>Leishmania major</em> gene and each of the <em>Plasmodium falciparum</em> orthologs will be “ortholog_one2many” (one-to-many orthology).</p>
<p>This can arise in the case where there was a duplication in the <em>Plasmodium falciparum</em> lineage after Plasmodium and Leishmania diverged, which means that two different <em>Plasmodium falciparum</em> genes (which are paralogs of each other) are both orthologs of the same <em>Leishmania major</em> gene.</p>
<p>Therefore, we can retrieve the Ensembl identifiers of the <em>Plasmodium falciparum</em> orthologs of all <em>Leishmania major</em> genes by typing:</p>
<div class="sourceCode" id="cb1247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1247-1"><a href="comparative-genomics.html#cb1247-1" aria-hidden="true" tabindex="-1"></a>leishmaniagenes <span class="ot">&lt;-</span> <span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>,</span>
<span id="cb1247-2"><a href="comparative-genomics.html#cb1247-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">&quot;pfalciparum_eg_homolog_ensembl_gene&quot;</span>,</span>
<span id="cb1247-3"><a href="comparative-genomics.html#cb1247-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="st">&quot;pfalciparum_eg_homolog_orthology_type&quot;</span>),</span>
<span id="cb1247-4"><a href="comparative-genomics.html#cb1247-4" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">mart=</span>ensemblleishmania)</span></code></pre></div>
<p>This will return an <em>R</em> list variable leishmaniagenes, the first element of which is a vector of Ensembl identifiers for all <em>Leishmania major</em> coding genes, and the second element of which is a vector of Ensembl identifiers for their <em>Plasmodium falciparum</em> orthologs, and the third element of which is a vector with information on the orthology types.</p>
<p>We can print out the names of the first 10 <em>Leishmania major</em> genes and their <em>Plasmodium falciparum</em> orthologs, and their orthology types, by typing:</p>
<div class="sourceCode" id="cb1248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1248-1"><a href="comparative-genomics.html#cb1248-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the names of all *Leishmania major* genes</span></span>
<span id="cb1248-2"><a href="comparative-genomics.html#cb1248-2" aria-hidden="true" tabindex="-1"></a>leishmaniagenenames <span class="ot">&lt;-</span> leishmaniagenes[[<span class="dv">1</span>]]  </span>
<span id="cb1248-3"><a href="comparative-genomics.html#cb1248-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1248-4"><a href="comparative-genomics.html#cb1248-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the P. falciparum orthologs of all L. major genes</span></span>
<span id="cb1248-5"><a href="comparative-genomics.html#cb1248-5" aria-hidden="true" tabindex="-1"></a>leishmaniaPforthologues <span class="ot">&lt;-</span> leishmaniagenes[[<span class="dv">2</span>]]  </span>
<span id="cb1248-6"><a href="comparative-genomics.html#cb1248-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1248-7"><a href="comparative-genomics.html#cb1248-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the orthology relationship type</span></span>
<span id="cb1248-8"><a href="comparative-genomics.html#cb1248-8" aria-hidden="true" tabindex="-1"></a>leishmaniaPforthologuetypes <span class="ot">&lt;-</span> leishmaniagenes[[<span class="dv">3</span>]] </span>
<span id="cb1248-9"><a href="comparative-genomics.html#cb1248-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1248-10"><a href="comparative-genomics.html#cb1248-10" aria-hidden="true" tabindex="-1"></a>leishmaniagenenames[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;LMJF_01_0780&quot; &quot;LMJF_01_0350&quot; &quot;LMJF_01_0280&quot; &quot;LMJF_01_0290&quot; &quot;LMJF_01_0335&quot;
##  [6] &quot;LMJF_01_0600&quot; &quot;LMJF_01_0220&quot; &quot;LMJF_01_0220&quot; &quot;LMJF_01_0220&quot; &quot;LMJF_01_0220&quot;</code></pre>
<div class="sourceCode" id="cb1250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1250-1"><a href="comparative-genomics.html#cb1250-1" aria-hidden="true" tabindex="-1"></a>leishmaniaPforthologues[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;&quot;              &quot;&quot;              &quot;&quot;              &quot;&quot;             
##  [5] &quot;&quot;              &quot;&quot;              &quot;PF3D7_0528700&quot; &quot;PF3D7_1116300&quot;
##  [9] &quot;PF3D7_0804800&quot; &quot;PF3D7_1423200&quot;</code></pre>
<div class="sourceCode" id="cb1252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1252-1"><a href="comparative-genomics.html#cb1252-1" aria-hidden="true" tabindex="-1"></a>leishmaniaPforthologuetypes[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;&quot;                   &quot;&quot;                   &quot;&quot;                  
##  [4] &quot;&quot;                   &quot;&quot;                   &quot;&quot;                  
##  [7] &quot;ortholog_many2many&quot; &quot;ortholog_many2many&quot; &quot;ortholog_many2many&quot;
## [10] &quot;ortholog_many2many&quot;</code></pre>
<p>Not all <em>Leishmania major</em> genes have <em>Plasmodium falciparum</em> orthologs; this is why when we print out the first 10 elements of the vector leishmaniaPforthologues, some of the elements are empty.</p>
<p>To find out how many <em>Leishmania major</em> genes have orthologs in <em>Plasmodium falciparum</em>, we can first find the indices of the elements of the vector leishmaniaPforthologues that are empty:</p>
<div class="sourceCode" id="cb1254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1254-1"><a href="comparative-genomics.html#cb1254-1" aria-hidden="true" tabindex="-1"></a>myindex <span class="ot">&lt;-</span> leishmaniaPforthologues <span class="sc">==</span> <span class="st">&quot;&quot;</span></span></code></pre></div>
<p>We can then find out the names of the Leishmania gene genes corresponding to those indices:</p>
<div class="sourceCode" id="cb1255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1255-1"><a href="comparative-genomics.html#cb1255-1" aria-hidden="true" tabindex="-1"></a>leishmaniagenenames2 <span class="ot">&lt;-</span> leishmaniagenenames[myindex]</span>
<span id="cb1255-2"><a href="comparative-genomics.html#cb1255-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(leishmaniagenenames2)</span></code></pre></div>
<pre><code>## [1] 8191</code></pre>
<p>This tells us that 8191 <em>Leishmania major</em> genes do not have <em>Plasmodium falciparum</em> orthologs.</p>
<p>How many of the 8191 <em>Leishmania major</em> genes that do not have <em>Plasmodium falaciparum</em> orthologs are protein-coding genes? To answer this question, we can merge together the information in the R list variable leishmaniagenes2 (which contains information on the name of each <em>Leishmania major</em> gene and its type; see above), and the R list variable leishmaniagenes (which contains information on the name of each L. major gene and its <em>Plasmodium falciparum</em> orthologs).</p>
<p>Remember that leishmaniagenes2 was created by typing:</p>
<div class="sourceCode" id="cb1257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1257-1"><a href="comparative-genomics.html#cb1257-1" aria-hidden="true" tabindex="-1"></a>leishmaniagenes2 <span class="ot">&lt;-</span> <span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;gene_biotype&quot;</span>), </span>
<span id="cb1257-2"><a href="comparative-genomics.html#cb1257-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">mart=</span>ensemblleishmania)</span></code></pre></div>
<p>To combine leishmaniagenes and leishmaniagenes2, we can use the merge() function in R, which can merge together two list variables that contain some named elements in common (in this case, both list variables contain a vector that has the names of <em>Leishmania major</em> genes):</p>
<div class="sourceCode" id="cb1258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1258-1"><a href="comparative-genomics.html#cb1258-1" aria-hidden="true" tabindex="-1"></a>leishmaniagenes3 <span class="ot">&lt;-</span> <span class="fu">merge</span>(leishmaniagenes2, leishmaniagenes)</span></code></pre></div>
<p>The first element of the merged list variable leishmaniagenes3 contains a vector of the <em>Leishmania major</em> gene names, the second has a vector of the types of those genes (e.g. protein-coding, pseudogene etc.), and the third element has a vector of the <em>Plasmodium falciparum</em> orthologs’ names. We can therefore find out how many protein-coding <em>Leishmania major</em> genes lack <em>Plasmodium falciparum</em> orthologs by typing:</p>
<div class="sourceCode" id="cb1259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1259-1"><a href="comparative-genomics.html#cb1259-1" aria-hidden="true" tabindex="-1"></a>leishmaniagenenames <span class="ot">&lt;-</span> leishmaniagenes3[[<span class="dv">1</span>]]</span>
<span id="cb1259-2"><a href="comparative-genomics.html#cb1259-2" aria-hidden="true" tabindex="-1"></a>leishmaniagenebiotypes <span class="ot">&lt;-</span> leishmaniagenes3[[<span class="dv">2</span>]]</span>
<span id="cb1259-3"><a href="comparative-genomics.html#cb1259-3" aria-hidden="true" tabindex="-1"></a>leishmaniaPforthologues <span class="ot">&lt;-</span> leishmaniagenes3[[<span class="dv">3</span>]]</span>
<span id="cb1259-4"><a href="comparative-genomics.html#cb1259-4" aria-hidden="true" tabindex="-1"></a>myindex <span class="ot">&lt;-</span> leishmaniaPforthologues<span class="sc">==</span><span class="st">&quot;&quot;</span> <span class="sc">&amp;</span> leishmaniagenebiotypes<span class="sc">==</span><span class="st">&quot;protein_coding&quot;</span></span>
<span id="cb1259-5"><a href="comparative-genomics.html#cb1259-5" aria-hidden="true" tabindex="-1"></a>leishmaniagenenames2 <span class="ot">&lt;-</span> leishmaniagenenames[myindex]</span>
<span id="cb1259-6"><a href="comparative-genomics.html#cb1259-6" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(leishmaniagenenames2)</span></code></pre></div>
<pre><code>## [1] 6476</code></pre>
<p>This tells us that there are 6476 <em>Leishmania major</em> protein-coding genes that lack <em>Plasmodium falciparum</em> orthologs.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="significance-of-orfs.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="summary-4.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/010-comparative_genomics/30-comparative_genomics.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/010-comparative_genomics/30-comparative_genomics.Rmd",
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
