<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 14 Downloading DNA sequences as FASTA files in R | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 14 Downloading DNA sequences as FASTA files in R | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 14 Downloading DNA sequences as FASTA files in R | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="programming-in-r-functions.html"/>
<link rel="next" href="downloading-dna-sequences-as-fasta-files-in-r-1.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>1</b> Test shiny app</a></li>
<li class="chapter" data-level="2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>2</b> How to review this book</a>
<ul>
<li class="chapter" data-level="2.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>2.1</b> File format</a></li>
<li class="chapter" data-level="2.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>2.2</b> What can you do?</a></li>
<li class="chapter" data-level="2.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>2.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>2.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="2.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>2.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="2.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>2.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="2.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>2.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>3</b> Downloading R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>3.1</b> Preface</a></li>
<li class="chapter" data-level="3.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>3.2</b> Introduction to R</a></li>
<li class="chapter" data-level="3.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>3.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>3.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="3.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>3.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>3.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>4</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="4.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>4.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="4.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>4.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>5</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="5.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>5.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="5.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>5.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="5.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>5.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>6</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="6.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>6.1</b> Bioconductor</a></li>
<li class="chapter" data-level="6.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>6.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="6.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>6.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>6.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="6.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>6.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="6.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>6.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>6.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>7</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="7.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>7.1</b> Vocabulary</a></li>
<li class="chapter" data-level="7.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>7.2</b> R functions</a></li>
<li class="chapter" data-level="7.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>7.3</b> Interacting with R</a></li>
<li class="chapter" data-level="7.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>7.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>7.4.1</b> Vectors</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>7.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>7.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="7.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>7.4.4</b> Lists</a></li>
<li class="chapter" data-level="7.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>7.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>7.5</b> Arguments</a></li>
<li class="chapter" data-level="7.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>7.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="7.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>7.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="7.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>7.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="7.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>7.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>7.9</b> Quiting R</a></li>
<li class="chapter" data-level="7.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>7.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>8</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-1"><i class="fa fa-check"></i><b>8.1</b> Preface</a></li>
<li class="chapter" data-level="8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>8.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>8.3</b> Introduction</a></li>
<li class="chapter" data-level="8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>8.4</b> Vocabulary</a></li>
<li class="chapter" data-level="8.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions"><i class="fa fa-check"></i><b>8.5</b> Functions</a></li>
<li class="chapter" data-level="8.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>8.6</b> Preliminaries</a></li>
<li class="chapter" data-level="8.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>8.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="8.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>8.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>8.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="8.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>8.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="8.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>8.8.3</b> DNA words</a></li>
<li class="chapter" data-level="8.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>8.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>8.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="8.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>8.9.1</b> License</a></li>
<li class="chapter" data-level="8.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>8.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>9</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="9.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-2"><i class="fa fa-check"></i><b>9.1</b> Preface</a></li>
<li class="chapter" data-level="9.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>9.2</b> Vocab</a></li>
<li class="chapter" data-level="9.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-1"><i class="fa fa-check"></i><b>9.3</b> Functions</a></li>
<li class="chapter" data-level="9.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>9.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="9.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>9.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>10</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="10.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-3"><i class="fa fa-check"></i><b>10.1</b> Preface</a></li>
<li class="chapter" data-level="10.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>10.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="functions-2.html"><a href="functions-2.html"><i class="fa fa-check"></i><b>11</b> Functions</a>
<ul>
<li class="chapter" data-level="11.1" data-path="functions-2.html"><a href="functions-2.html#vectors-in-r"><i class="fa fa-check"></i><b>11.1</b> Vectors in R</a></li>
<li class="chapter" data-level="11.2" data-path="functions-2.html"><a href="functions-2.html#math-on-vectors"><i class="fa fa-check"></i><b>11.2</b> Math on vectors</a></li>
<li class="chapter" data-level="11.3" data-path="functions-2.html"><a href="functions-2.html#functions-on-vectors"><i class="fa fa-check"></i><b>11.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="11.4" data-path="functions-2.html"><a href="functions-2.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>11.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="11.5" data-path="functions-2.html"><a href="functions-2.html#subsetting-vectors"><i class="fa fa-check"></i><b>11.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="11.6" data-path="functions-2.html"><a href="functions-2.html#sequences-of-numbers"><i class="fa fa-check"></i><b>11.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="11.7" data-path="functions-2.html"><a href="functions-2.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>11.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="11.8" data-path="functions-2.html"><a href="functions-2.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>11.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>12</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="12.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-4"><i class="fa fa-check"></i><b>12.1</b> Preface</a></li>
<li class="chapter" data-level="12.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>12.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="12.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>12.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>13</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="13.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-5"><i class="fa fa-check"></i><b>13.1</b> Preface</a></li>
<li class="chapter" data-level="13.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>13.2</b> Vocab</a></li>
<li class="chapter" data-level="13.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-3"><i class="fa fa-check"></i><b>13.3</b> Functions</a></li>
<li class="chapter" data-level="13.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>13.4</b> Functions in R</a></li>
<li class="chapter" data-level="13.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>13.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>14</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="14.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-4"><i class="fa fa-check"></i><b>14.0.1</b> Functions</a></li>
<li class="chapter" data-level="14.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>14.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="14.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>14.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="14.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>14.0.4</b> File types</a></li>
<li class="chapter" data-level="14.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>14.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="14.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>14.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="14.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>14.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="14.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>14.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>14.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="14.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>14.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="14.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>14.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="14.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>14.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="14.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>14.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>14.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="14.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>14.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>15</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="15.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-2"><i class="fa fa-check"></i><b>15.1</b> Preliminaries</a></li>
<li class="chapter" data-level="15.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>15.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>15.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="15.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>15.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>16</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="16.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>16.1</b> Preliminaries</a></li>
<li class="chapter" data-level="16.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>16.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>17</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="17.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>17.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="17.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>17.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="17.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>17.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="17.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>17.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="17.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>17.4</b> Inversions</a></li>
<li class="chapter" data-level="17.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>17.5</b> Translocations</a></li>
<li class="chapter" data-level="17.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>17.6</b> Random sequence</a></li>
<li class="chapter" data-level="17.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>17.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>18</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="18.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>18.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>18.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>18.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="18.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>18.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>18.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="18.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>18.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="18.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>18.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="18.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>18.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>19</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="19.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>19.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="19.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>19.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="19.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>19.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>20</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="20.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>20.1</b> Prelminaries</a></li>
<li class="chapter" data-level="20.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>20.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="20.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>20.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="20.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>20.4</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="20.4.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>20.4.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="20.4.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>20.4.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="20.4.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>20.4.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="20.4.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>20.4.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-2"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="20.5" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>20.5</b> Help!</a>
<ul>
<li class="chapter" data-level="20.5.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>20.5.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="20.5.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>20.5.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="20.5.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>20.5.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="20.5.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>20.5.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="20.6" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>20.6</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="20.6.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>20.6.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="20.6.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>20.6.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="20.7" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>20.7</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="downloading-dna-sequences-as-fasta-files-in-r" class="section level1" number="14">
<h1><span class="header-section-number">Chapter 14</span> Downloading DNA sequences as FASTA files in R</h1>
<p>This is a modification of <a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter1.html">“DNA Sequence Statistics”</a> from Avril Coghlan’s <a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/index.html"><em>A little book of R for bioinformatics.</em></a>. Most of the text and code was originally written by Dr. Coghlan and distributed under the <a href="https://creativecommons.org/licenses/by/3.0/us/">Creative Commons 3.0</a> license.</p>
<!-- how much is sequinr actually useD? -->
<p><strong>NOTE</strong>: There is some reduncnacy in this current draft that needs to be eliminated.</p>
<div id="functions-4" class="section level3" number="14.0.1">
<h3><span class="header-section-number">14.0.1</span> Functions</h3>
<ul>
<li>library()</li>
<li>help()</li>
<li>length</li>
<li>table</li>
<li>seqinr::GC()</li>
<li>seqinr::count()</li>
<li>seqinr::write.fasta()</li>
</ul>
</div>
<div id="softwarewebsites" class="section level3" number="14.0.2">
<h3><span class="header-section-number">14.0.2</span> Software/websites</h3>
<ul>
<li>www.ncbi.nlm.nih.gov</li>
<li>Text editors (e.g. Notepad++, TextWrangler)</li>
</ul>
</div>
<div id="r-vocabulary" class="section level3" number="14.0.3">
<h3><span class="header-section-number">14.0.3</span> R vocabulary</h3>
<ul>
<li>list</li>
<li>library</li>
<li>package</li>
<li>CRAN</li>
<li>wrapper</li>
</ul>
</div>
<div id="file-types" class="section level3" number="14.0.4">
<h3><span class="header-section-number">14.0.4</span> File types</h3>
<ul>
<li>FASTA</li>
</ul>
</div>
<div id="bioinformatics-vocabulary" class="section level3" number="14.0.5">
<h3><span class="header-section-number">14.0.5</span> Bioinformatics vocabulary</h3>
<ul>
<li>accession, accession number</li>
<li>NCBI</li>
<li>NCBI Sequence Database</li>
<li>EMBL Sequence Database</li>
<li>FASTA file</li>
</ul>
</div>
<div id="organisms-and-sequence-accessions" class="section level3" number="14.0.6">
<h3><span class="header-section-number">14.0.6</span> Organisms and Sequence accessions</h3>
<ul>
<li>Dengue virus: DEN-1, DEN-2, DEN-3, and DEN-4.</li>
</ul>
<p>The NCBI accessions for the DNA sequences of the DEN-1, DEN-2, DEN-3, and DEN-4 Dengue viruses are NC_001477, NC_001474, NC_001475 and NC_002640, respectively.</p>
<p>According to Wikipedia</p>
<blockquote>
<p>“Dengue virus (DENV) is the cause of dengue fever. It is a mosquito-borne, single positive-stranded RNA virus … Five serotypes of the virus have been found, all of which can cause the full spectrum of disease. Nevertheless, scientists’ understanding of dengue virus may be simplistic, as rather than distinct … groups, a continuum appears to exist.” <a href="https://en.wikipedia.org/wiki/Dengue_virus" class="uri">https://en.wikipedia.org/wiki/Dengue_virus</a></p>
</blockquote>
</div>
<div id="preliminaries-1" class="section level3" number="14.0.7">
<h3><span class="header-section-number">14.0.7</span> Preliminaries</h3>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb198-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rentrez)</span>
<span id="cb198-2"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb198-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compbio4all)  </span></code></pre></div>
<!-- is compbio4all called? -->
</div>
<div id="dna-sequence-statistics-part-1" class="section level2" number="14.1">
<h2><span class="header-section-number">14.1</span> DNA Sequence Statistics: Part 1</h2>
<div id="using-r-for-bioinformatics" class="section level3" number="14.1.1">
<h3><span class="header-section-number">14.1.1</span> Using R for Bioinformatics</h3>
<p>The chapter will guide you through the process of using R to carry out simple analyses that are common in bioinformatics and computational biology. In particular, the focus is on computational analysis of biological sequence data such as genome sequences and protein sequences. The programming approaches, however, are broadly generalizable to statistics and data science.</p>
<p>The tutorials assume that the reader has some basic knowledge of biology, but not necessarily of bioinformatics. The focus is to explain simple bioinformatics analysis, and to explain how to carry out these analyses using <em>R</em>.</p>
</div>
<div id="r-packages-for-bioinformatics-bioconductor-and-seqinr" class="section level3" number="14.1.2">
<h3><span class="header-section-number">14.1.2</span> R packages for bioinformatics: Bioconductor and SeqinR</h3>
<p>Many authors have written <em>R</em> packages for performing a wide variety of analyses. These do not come with the standard <em>R</em> installation, but must be installed and loaded as “add-ons.”</p>
<!-- old Loading the dayoff package will automatically load most of the packages we need. -->
<!-- have these as dependencies of combio4all? -->
<p>Bioinformaticians have written numerous specialized packages for <em>R</em>. In this tutorial, you will learn to use some of the function in the <a href="https://cran.r-project.org/web/packages/seqinr/index.html"><code>SeqinR</code></a> package to to carry out simple analyses of DNA sequences. (<code>SeqinR</code> can retrieve sequences from a DNA sequence database, but this has largely been replaced by the functions in the package <code>rentrez</code>)</p>
<p>Many well-known bioinformatics packages for <em>R</em> are in the Bioconductor set of <em>R</em> packages (www.bioconductor.org), which contains packages with many <em>R</em> functions for analyzing biological data sets such as microarray data. The <a href="https://cran.r-project.org/web/packages/seqinr/index.html"><code>SeqinR</code></a> package is from CRAN, which contains R functions for obtaining sequences from DNA and protein sequence databases, and for analyzing DNA and protein sequences.</p>
<!-- SequinR will automatically load when you load dayoff.   For instructions on how to install an R package on your own see [How to install an R package](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/installr.html). -->
<p>We will also use functions from the <code>rentrez</code> and <code>ape</code> packages.</p>
<!-- Both of these also are loaded automatically by the dayoff package. -->
<!-- ape still used?. -->
<p>Remember that you can ask for more information about a particular <em>R</em> command by using the <code>help()</code> function. For example, to ask for more information about the <code>library()</code>, you can type:</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb199-1" aria-hidden="true" tabindex="-1"></a><span class="fu">help</span>(<span class="st">&quot;library&quot;</span>)</span></code></pre></div>
<p>You can also do this</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb200-1" aria-hidden="true" tabindex="-1"></a>?library</span></code></pre></div>
</div>
<div id="fasta-file-format" class="section level3" number="14.1.3">
<h3><span class="header-section-number">14.1.3</span> FASTA file format</h3>
<p>The FASTA format is a simple and widely used format for storing biological (e.g. DNA or protein) sequences. It was first used by the <a href="https://en.wikipedia.org/wiki/FASTA">FASTA program</a> for sequence alignment in the 1980s and has been adopted as standard by many other programs.</p>
<p>FASTA files begin with a single-line description starting with a greater-than sign <code>&gt;</code> character, followed on the next line by the sequences. Here is an example of a FASTA file. (If you’re looking at the source script for this lesson you’ll see the <code>cat()</code> command, which is just a text display function used format the text when you run the code).</p>
<pre><code>## &gt;A06852 183 residues MPRLFSYLLGVWLLLSQLPREIPGQSTNDFIKACGRELVRLWVEICGSVSWGRTALSLEEPQLETGPPAETMPSSITKDAEILKMMLEFVPNLPQELKATLSERQPSLRELQQSASKDSNLNFEEFKKIILNRQNEAEDKSLLELKNLGLDKHSRKKRLFRMTLSEKCCQVGCIRKDIARLC</code></pre>
</div>
<div id="the-ncbi-sequence-database" class="section level3" number="14.1.4">
<h3><span class="header-section-number">14.1.4</span> The NCBI sequence database</h3>
<p>The US <a href="www.ncbi.nlm.nih.gov">National Centre for Biotechnology Information (NCBI)</a> maintains the <strong>NCBI Sequence Database</strong>, a huge database of all the DNA and protein sequence data that has been collected. There are also similar databases in Europe, the <a href="www.ebi.ac.uk/embl">European Molecular Biology Laboratory (EMBL) Sequence Database</a>, and Japan, the <a href="www.ddbj.nig.ac.jp">DNA Data Bank of Japan (DDBJ)</a>. These three databases exchange data every night, so at any one point in time, they contain almost identical data.</p>
<p>Each sequence in the NCBI Sequence Database is stored in a separate <strong>record</strong>, and is assigned a unique identifier that can be used to refer to that record. The identifier is known as an <strong>accession</strong>, and consists of a mixture of numbers and letters.</p>
<p>For example, Dengue virus causes Dengue fever, which is classified as a <strong>neglected tropical disease</strong> by the World Health Organization (WHO), is classified by any one of four types of Dengue virus: DEN-1, DEN-2, DEN-3, and DEN-4. The NCBI accessions for the DNA sequences of the DEN-1, DEN-2, DEN-3, and DEN-4 Dengue viruses are</p>
<ul>
<li>NC_001477</li>
<li>NC_001474</li>
<li>NC_001475<br />
</li>
<li>NC_002640</li>
</ul>
<p>Note that because the NCBI Sequence Database, the EMBL Sequence Database, and DDBJ exchange data every night, the DEN-1 (and DEN-2, DEN-3, DEN-4) Dengue virus sequence are present in all three databases, but they have different accessions in each database, as they each use their own numbering systems for referring to their own sequence records.</p>
</div>
<div id="retrieving-genome-sequence-data-using-rentrez" class="section level3" number="14.1.5">
<h3><span class="header-section-number">14.1.5</span> Retrieving genome sequence data using rentrez</h3>
<p>You can retrieve sequence data from NCBI directly from <em>R</em> using the <code>rentrez</code> package. The DEN-1 Dengue virus genome sequence has NCBI accession NC_001477. To retrieve a sequence with a particular NCBI accession, you can use the function <code>entrez_fetch()</code> from the <code>rentrez</code> package. Note that to be specific where the function comes from I write it as <code>package::function()</code>.</p>
<!-- QUESTION: How many arguments does entrez_fetch take in this call? -->
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb202-1" aria-hidden="true" tabindex="-1"></a>dengueseq_fasta <span class="ot">&lt;-</span> rentrez<span class="sc">::</span><span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;nucleotide&quot;</span>, </span>
<span id="cb202-2"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb202-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;NC_001477&quot;</span>, </span>
<span id="cb202-3"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb202-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>Note that the “<em>" in the name is just an arbitrary way to separate two words. Another common format would be <code>dengueseq.fasta</code>. Some people like <code>dengueseqFasta</code>, called <strong>camel case</strong> because the capital letter makes a hump in the middle of the word. Underscores are becoming most common and are favored by developers associated with RStudio and the <strong>tidyverse</strong> of packages that many data scientists use. I switch between "." and "</em>” as separators, usually favoring "_" for function names and “.” for objects; I personally find camel case harder to read and to type.</p>
<p>Ok, so what exactly have we done when we made <code>dengueseq_fasta</code>? We have an R object <code>dengueseq_fasta</code> which has the sequence linked to the accession number “NC_001477.” So where is the sequence, and what is it?</p>
<p>First, what is it?</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb203-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>## [1] &quot;character&quot;           &quot;vector&quot;              &quot;data.frameRowLabels&quot;
## [4] &quot;SuperClassMethod&quot;</code></pre>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb205-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>## [1] &quot;character&quot;</code></pre>
<p>How big is it? Try the <code>dim()</code> and <code>length()</code> commands and see which one works. Do you know why one works and the other doesn’t?
<!-- question:l which one works --></p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb207-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>## NULL</code></pre>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb209-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<p>The size of the object is 1. Why is this? This is the genomics sequence of a virus, so you’d expect it to be fairly large. We’ll use another function below to explore that issue. Think about this first: how many pieces of unique information are in the <code>dengueseq</code> object? In what sense is there only <em>one</em> piece of information?</p>
<p>If we want to actually see the sequence we can type just type <code>dengueseq_fasta</code> and press enter. This will print the WHOLE genomic sequence out but it will probably run of your screen.</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb211-1" aria-hidden="true" tabindex="-1"></a>dengueseq_fasta</span></code></pre></div>
<p>This is a whole genome sequence, but its stored as single entry in a vector, so the <code>length()</code> command just tells us how many entries there are in the vector, which is just one! What this means is that the entire genomic sequence is stored in a single entry of the vector <code>dengueseq_fasta</code>. (If you’re not following along with this, no worries - its not essential to actually working with the data)</p>
<p>If we want to actually know how long the sequence is, we need to use the function <code>nchar()</code>.</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb212-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>## [1] 10935</code></pre>
<p>The sequence is 10935 bases long. All of these bases are stored as a single <strong>character string</strong> with no spaces in a single entry of our <code>dengueseq_fasta</code> vector. This isn’t actually a useful format for us, so below were’ going to convert it to something more useful.</p>
<p>If we want to see just part of the sequence we can use the <code>strtrim()</code> function. Before you run the code below, predict what the 100 means.</p>
<pre><code>## [1] &quot;&gt;NC_001477.1 Dengue virus 1, complete genome\nAGTTGTTAGTCTACGTGGACCGACAAGAACAGTTTCGAATCGGAAGCTTGCTTAA&quot;</code></pre>
<p>Note that at the end of the name is a slash followed by an n, which indicates to the computer that this is a <strong>newline</strong>; this is read by text editor, but is ignored by R in this context.</p>
<pre><code>## [1] &quot;&gt;NC_001477.1 Dengue virus 1, complete genome\n&quot;</code></pre>
<p>After the <code>\\n</code> begins the sequence, which will continue on for a LOOOOOONG way. Let’s just print a little bit.</p>
<pre><code>## [1] &quot;&gt;NC_001477.1 Dengue virus 1, complete genome\nAGTTGTT&quot;</code></pre>
<p>Let’s print some more. Do you notice anything beside A, T, C and G in the sequence?</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb217-1" aria-hidden="true" tabindex="-1"></a><span class="fu">strtrim</span>(dengueseq_fasta, <span class="dv">200</span>)</span></code></pre></div>
<pre><code>## [1] &quot;&gt;NC_001477.1 Dengue virus 1, complete genome\nAGTTGTTAGTCTACGTGGACCGACAAGAACAGTTTCGAATCGGAAGCTTGCTTAACGTAGTTCTAACAGT\nTTTTTATTAGAGAGCAGATCTCTGATGAACAACCAACGGAAAAAGACGGGTCGACCGTCTTTCAATATGC\nTGAAACGCGCGAG&quot;</code></pre>
<p>Again, there are the <code>\\n</code> newline characters, which tell text editors and wordprocessors how to display the file.</p>
<p>Now that we a sense of what we’re looking at let’s explore the <code>dengueseq_fasta</code> a bit more.</p>
<p>We can find out more information about what it is using the <code>class()</code>command.</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb219-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>## [1] &quot;character&quot;</code></pre>
<p>As noted before, this is character data.</p>
<p>Many things in R are vectors so we can ask <em>R </em><code>is.vector()</code></p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb221-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is.vector</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Yup, that’s true.</p>
<p>Ok, let’s see what else. A handy though often verbose command is <code>is()</code>, which tells us what an object, well, what it is:</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb223-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>## [1] &quot;character&quot;           &quot;vector&quot;              &quot;data.frameRowLabels&quot;
## [4] &quot;SuperClassMethod&quot;</code></pre>
<p>There is a lot here but if you scan for some key words you will see “character” and “vector” at the top. The other stuff you can ignore. The first two things, though, tell us the dengueseq_fasta is a <strong>vector</strong> of the class <strong>character</strong>: that is, a <strong>character vector</strong>.</p>
<p>Another handy function is <code>str()</code>, which gives us a peak at the context and structure of an <em>R</em> object. This is most useful when you are working in the R console or with dataframes, but is a useful function to run on all <em>R</em> objects. How does this output differ from other ways we’ve displayed dengueseq_fasta?</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb225-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>##  chr &quot;&gt;NC_001477.1 Dengue virus 1, complete genome\nAGTTGTTAGTCTACGTGGACCGACAAGAACAGTTTCGAATCGGAAGCTTGCTTAACGTAGTTCTA&quot;| __truncated__</code></pre>
<p>We know it contains character data - how many characters? <code>nchar()</code> for “number of characters” answers that:</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(dengueseq_fasta)</span></code></pre></div>
<pre><code>## [1] 10935</code></pre>
</div>
</div>
<div id="optional-saving-fasta-files" class="section level2" number="14.2">
<h2><span class="header-section-number">14.2</span> OPTIONAL: Saving FASTA files</h2>
<p>We can save our data as .fasta file for safe keeping. The <code>write()</code> function will save the data we downloaded as a plain text file.</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb229-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write</span>(dengueseq_fasta, </span>
<span id="cb229-2"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb229-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">file=</span><span class="st">&quot;dengueseq.fasta&quot;</span>)</span></code></pre></div>
<p>If you do this, you’ll need to figure out where <em>R</em> is saving things, which requires and understanding <em>R’s</em> <strong>file system</strong>, which can take some getting used to, especially if you’re new to programming. As a start, you can see where <em>R</em> saves things by using the <code>getwd()</code> command, which tells you where on your harddrive R currently is using as its home base for files.</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#cb230-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code></pre></div>
<pre><code>## [1] &quot;/Users/nlb24/google_backup_sync_nlb24/lbrb&quot;</code></pre>
</div>
<div id="next-steps" class="section level2" number="14.3">
<h2><span class="header-section-number">14.3</span> Next steps</h2>
<p>On their own, FASTA files in R are not directly useful. In the next lesson we’ll process our <code>dengueseq_fasta</code> file so that we can use it in analyses.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="programming-in-r-functions.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="downloading-dna-sequences-as-fasta-files-in-r-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/11-download_FASTA_in_R.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/11-download_FASTA_in_R.Rmd",
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
