<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 18 Global proteins aligments in R | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 18 Global proteins aligments in R | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 18 Global proteins aligments in R | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="sequence-dotplots-in-r.html"/>
<link rel="next" href="local-protein-aligments-in-r.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>1</b> Test shiny app</a></li>
<li class="chapter" data-level="2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>2</b> How to review this book</a>
<ul>
<li class="chapter" data-level="2.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>2.1</b> File format</a></li>
<li class="chapter" data-level="2.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>2.2</b> What can you do?</a></li>
<li class="chapter" data-level="2.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>2.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>2.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="2.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>2.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="2.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>2.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="2.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>2.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>3</b> Downloading R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>3.1</b> Preface</a></li>
<li class="chapter" data-level="3.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>3.2</b> Introduction to R</a></li>
<li class="chapter" data-level="3.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>3.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>3.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="3.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>3.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>3.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>4</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="4.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>4.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="4.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>4.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>5</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="5.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>5.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="5.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>5.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="5.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>5.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>6</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="6.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>6.1</b> Bioconductor</a></li>
<li class="chapter" data-level="6.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>6.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="6.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>6.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>6.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="6.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>6.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="6.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>6.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>6.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>7</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="7.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>7.1</b> Vocabulary</a></li>
<li class="chapter" data-level="7.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>7.2</b> R functions</a></li>
<li class="chapter" data-level="7.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>7.3</b> Interacting with R</a></li>
<li class="chapter" data-level="7.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>7.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>7.4.1</b> Vectors</a></li>
<li class="chapter" data-level="7.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>7.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="7.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>7.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="7.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>7.4.4</b> Lists</a></li>
<li class="chapter" data-level="7.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>7.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>7.5</b> Arguments</a></li>
<li class="chapter" data-level="7.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>7.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="7.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>7.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="7.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>7.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="7.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>7.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>7.9</b> Quiting R</a></li>
<li class="chapter" data-level="7.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>7.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>8</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-1"><i class="fa fa-check"></i><b>8.1</b> Preface</a></li>
<li class="chapter" data-level="8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>8.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>8.3</b> Introduction</a></li>
<li class="chapter" data-level="8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>8.4</b> Vocabulary</a></li>
<li class="chapter" data-level="8.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions"><i class="fa fa-check"></i><b>8.5</b> Functions</a></li>
<li class="chapter" data-level="8.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>8.6</b> Preliminaries</a></li>
<li class="chapter" data-level="8.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>8.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="8.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>8.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>8.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="8.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>8.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="8.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>8.8.3</b> DNA words</a></li>
<li class="chapter" data-level="8.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>8.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>8.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="8.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>8.9.1</b> License</a></li>
<li class="chapter" data-level="8.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>8.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>9</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="9.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-2"><i class="fa fa-check"></i><b>9.1</b> Preface</a></li>
<li class="chapter" data-level="9.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>9.2</b> Vocab</a></li>
<li class="chapter" data-level="9.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-1"><i class="fa fa-check"></i><b>9.3</b> Functions</a></li>
<li class="chapter" data-level="9.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>9.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="9.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>9.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>10</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="10.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-3"><i class="fa fa-check"></i><b>10.1</b> Preface</a></li>
<li class="chapter" data-level="10.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>10.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="functions-2.html"><a href="functions-2.html"><i class="fa fa-check"></i><b>11</b> Functions</a>
<ul>
<li class="chapter" data-level="11.1" data-path="functions-2.html"><a href="functions-2.html#vectors-in-r"><i class="fa fa-check"></i><b>11.1</b> Vectors in R</a></li>
<li class="chapter" data-level="11.2" data-path="functions-2.html"><a href="functions-2.html#math-on-vectors"><i class="fa fa-check"></i><b>11.2</b> Math on vectors</a></li>
<li class="chapter" data-level="11.3" data-path="functions-2.html"><a href="functions-2.html#functions-on-vectors"><i class="fa fa-check"></i><b>11.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="11.4" data-path="functions-2.html"><a href="functions-2.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>11.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="11.5" data-path="functions-2.html"><a href="functions-2.html#subsetting-vectors"><i class="fa fa-check"></i><b>11.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="11.6" data-path="functions-2.html"><a href="functions-2.html#sequences-of-numbers"><i class="fa fa-check"></i><b>11.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="11.7" data-path="functions-2.html"><a href="functions-2.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>11.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="11.8" data-path="functions-2.html"><a href="functions-2.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>11.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>12</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="12.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-4"><i class="fa fa-check"></i><b>12.1</b> Preface</a></li>
<li class="chapter" data-level="12.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>12.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="12.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>12.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>13</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="13.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-5"><i class="fa fa-check"></i><b>13.1</b> Preface</a></li>
<li class="chapter" data-level="13.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>13.2</b> Vocab</a></li>
<li class="chapter" data-level="13.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-3"><i class="fa fa-check"></i><b>13.3</b> Functions</a></li>
<li class="chapter" data-level="13.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>13.4</b> Functions in R</a></li>
<li class="chapter" data-level="13.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>13.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>14</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="14.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-4"><i class="fa fa-check"></i><b>14.0.1</b> Functions</a></li>
<li class="chapter" data-level="14.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>14.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="14.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>14.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="14.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>14.0.4</b> File types</a></li>
<li class="chapter" data-level="14.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>14.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="14.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>14.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="14.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>14.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="14.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>14.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>14.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="14.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>14.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="14.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>14.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="14.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>14.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="14.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>14.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>14.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="14.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>14.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>15</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="15.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-2"><i class="fa fa-check"></i><b>15.1</b> Preliminaries</a></li>
<li class="chapter" data-level="15.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>15.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>15.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="15.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>15.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>16</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="16.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>16.1</b> Preliminaries</a></li>
<li class="chapter" data-level="16.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>16.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>17</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="17.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>17.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="17.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>17.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="17.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>17.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="17.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>17.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="17.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>17.4</b> Inversions</a></li>
<li class="chapter" data-level="17.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>17.5</b> Translocations</a></li>
<li class="chapter" data-level="17.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>17.6</b> Random sequence</a></li>
<li class="chapter" data-level="17.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>17.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>18</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="18.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>18.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>18.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>18.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="18.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>18.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>18.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="18.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>18.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="18.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>18.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="18.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>18.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>19</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="19.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>19.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="19.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>19.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="19.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>19.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>20</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="20.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>20.1</b> Prelminaries</a></li>
<li class="chapter" data-level="20.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>20.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="20.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>20.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="20.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>20.4</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="20.4.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>20.4.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="20.4.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>20.4.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="20.4.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>20.4.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="20.4.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>20.4.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-2"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="20.5" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>20.5</b> Help!</a>
<ul>
<li class="chapter" data-level="20.5.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>20.5.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="20.5.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>20.5.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="20.5.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>20.5.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="20.5.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>20.5.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="20.6" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>20.6</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="20.6.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>20.6.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="20.6.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>20.6.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="20.7" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>20.7</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="global-proteins-aligments-in-r" class="section level1" number="18">
<h1><span class="header-section-number">Chapter 18</span> Global proteins aligments in <em>R</em></h1>
<p><strong>By</strong>: Avril Coghlan.</p>
<p><strong>Adapted, edited and expanded</strong>: Nathan Brouwer under the Creative Commons 3.0 Attribution License <a href="https://creativecommons.org/licenses/by/3.0/">(CC BY 3.0)</a>.</p>
<div id="preliminaries-5" class="section level2" number="18.1">
<h2><span class="header-section-number">18.1</span> Preliminaries</h2>
<!-- is compbio4all needed? -->
<!-- have on gone over biodoncdutor and biostrings? -->
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="global-proteins-aligments-in-r.html#cb269-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compbio4all)</span>
<span id="cb269-2"><a href="global-proteins-aligments-in-r.html#cb269-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb269-3"><a href="global-proteins-aligments-in-r.html#cb269-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span></code></pre></div>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: XVector</code></pre>
<pre><code>## 
## Attaching package: &#39;Biostrings&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:seqinr&#39;:
## 
##     translate</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     strsplit</code></pre>
<div id="download-sequences-1" class="section level3" number="18.1.1">
<h3><span class="header-section-number">18.1.1</span> Download sequences</h3>
<p>As we did in the previous lesson on dotplots, we’ll look at two sequences.</p>
<div class="sourceCode" id="cb285"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb285-1"><a href="global-proteins-aligments-in-r.html#cb285-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download</span></span>
<span id="cb285-2"><a href="global-proteins-aligments-in-r.html#cb285-2" aria-hidden="true" tabindex="-1"></a><span class="do">## sequence 1: Q9CD83</span></span>
<span id="cb285-3"><a href="global-proteins-aligments-in-r.html#cb285-3" aria-hidden="true" tabindex="-1"></a>leprae_fasta <span class="ot">&lt;-</span> rentrez<span class="sc">::</span><span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>,</span>
<span id="cb285-4"><a href="global-proteins-aligments-in-r.html#cb285-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">id =</span> <span class="st">&quot;Q9CD83&quot;</span>,</span>
<span id="cb285-5"><a href="global-proteins-aligments-in-r.html#cb285-5" aria-hidden="true" tabindex="-1"></a>                         <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb285-6"><a href="global-proteins-aligments-in-r.html#cb285-6" aria-hidden="true" tabindex="-1"></a><span class="do">## sequence 2: OIN17619.1</span></span>
<span id="cb285-7"><a href="global-proteins-aligments-in-r.html#cb285-7" aria-hidden="true" tabindex="-1"></a>ulcerans_fasta <span class="ot">&lt;-</span> rentrez<span class="sc">::</span><span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>,</span>
<span id="cb285-8"><a href="global-proteins-aligments-in-r.html#cb285-8" aria-hidden="true" tabindex="-1"></a>                         <span class="at">id =</span> <span class="st">&quot;OIN17619.1&quot;</span>,</span>
<span id="cb285-9"><a href="global-proteins-aligments-in-r.html#cb285-9" aria-hidden="true" tabindex="-1"></a>                         <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb285-10"><a href="global-proteins-aligments-in-r.html#cb285-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-11"><a href="global-proteins-aligments-in-r.html#cb285-11" aria-hidden="true" tabindex="-1"></a><span class="co"># clean</span></span>
<span id="cb285-12"><a href="global-proteins-aligments-in-r.html#cb285-12" aria-hidden="true" tabindex="-1"></a>leprae_vector   <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(leprae_fasta)</span>
<span id="cb285-13"><a href="global-proteins-aligments-in-r.html#cb285-13" aria-hidden="true" tabindex="-1"></a>ulcerans_vector <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(ulcerans_fasta)</span></code></pre></div>
</div>
</div>
<div id="pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm" class="section level2" number="18.2">
<h2><span class="header-section-number">18.2</span> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</h2>
<p>If you are studying a particular pair of genes or proteins, an important question is to what extent the two sequences are similar.</p>
<!-- Could precede this by an R exercise to calculate compositional similarity -->
<p>To quantify similarity, it is necessary to <strong>align</strong> the two sequences, and then you can calculate a similarity score based on the alignment.</p>
<p>There are two types of alignment in general. A <strong>global alignment</strong> is an alignment of the <em>full</em> length of two sequences from beginning to end, for example, of two protein sequences or of two DNA sequences. A <strong>local alignment</strong> is an alignment of part of one sequence to part of another sequence; the parts the end up getting aligned are the most similar, and determined by the alignment algorithm.</p>
<!-- review mutations? -->
<!-- Introduction PID -->
<p>The first step in computing a alignment (global or local) is to decide on a <strong>scoring system</strong>. For example, we may decide to give a score of +2 to a match and a penalty of -1 to a mismatch, and a penalty of -2 to a <strong>gap</strong> due to an <strong>indexl</strong>. Thus, for the alignment:</p>
<pre><code>## [1] &quot;G A A T T C&quot;</code></pre>
<pre><code>## [1] &quot;G A T T - A&quot;</code></pre>
<p>we would compute a score of</p>
<ol style="list-style-type: decimal">
<li>G vs G = matchb = 2</li>
<li>A vs A = match = 2</li>
<li>A vs T = mismatch = -1</li>
<li>T vs T = match = 2</li>
<li>T vs - = gap = -2</li>
<li>C vs A = mismatch = 2</li>
</ol>
<p>So, the scores is 2 + 2 -1 + 2 -2 - 1 = 2.</p>
<p>Similarly, the score for the following alignment is 2 + 2 -2 + 2 + 2 -1 = 5:</p>
<pre><code>## [1] &quot;G A A T T C&quot;</code></pre>
<pre><code>## [1] &quot;G A - T T A&quot;</code></pre>
<p>The <strong>scoring system</strong> above can be represented by a <strong>scoring matrix</strong> (also known as a <strong>substitution matrix</strong>). The scoring matrix has one row and one column for each possible letter in our alphabet of letters (e.g. 4 rows and 4 columns for DNA and RNA sequences, 20 x 20 for amino acids). The (i,j) element of the matrix has a value of +2 in case of a match and -1 in case of a mismatch.</p>
<p>We can make a scoring matrix in R by using the <code>nucleotideSubstitutionMatrix()</code> function in the <code>Biostrings</code> package. <code>Biostrings</code> is part of a set of R packages for bioinformatics analysis known as <a href="www.bioconductor.org/">Bioconductor</a> (www.bioconductor.org/).</p>
<p>The arguments (inputs) for the <code>nucleotideSubstitutionMatrix()</code> function are the score that we want to assign to a match and the score that we want to assign to a mismatch. We can also specify that we want to use only the four letters representing the four nucleotides (ie. A, C, G, T) by setting <code>baseOnly=TRUE</code>, or whether we also want to use the letters that represent <strong>ambiguous cases</strong> where we are not sure what the nucleotide is (e.g. ‘N’ = A/C/G/T; ambiguous cases occur in some sequences due to sequencing errors or ambiguities).</p>
<p>To make a scoring matrix which assigns a score of +2 to a match and -1 to a mismatch, and store it in the variable <code>sigma</code>, we type:</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="global-proteins-aligments-in-r.html#cb290-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make the matrix</span></span>
<span id="cb290-2"><a href="global-proteins-aligments-in-r.html#cb290-2" aria-hidden="true" tabindex="-1"></a>sigma <span class="ot">&lt;-</span> <span class="fu">nucleotideSubstitutionMatrix</span>(<span class="at">match =</span> <span class="dv">2</span>, </span>
<span id="cb290-3"><a href="global-proteins-aligments-in-r.html#cb290-3" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">mismatch =</span> <span class="sc">-</span><span class="dv">1</span>, </span>
<span id="cb290-4"><a href="global-proteins-aligments-in-r.html#cb290-4" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">baseOnly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb290-5"><a href="global-proteins-aligments-in-r.html#cb290-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out the matrix</span></span>
<span id="cb290-6"><a href="global-proteins-aligments-in-r.html#cb290-6" aria-hidden="true" tabindex="-1"></a>sigma </span></code></pre></div>
<pre><code>##    A  C  G  T
## A  2 -1 -1 -1
## C -1  2 -1 -1
## G -1 -1  2 -1
## T -1 -1 -1  2</code></pre>
<p>Instead of assigning the same penalty (e.g. -8) to every gap position, it is common instead to assign a <strong>gap opening penalty</strong> to the first position in a gap (e.g. -8), and a smaller <strong>gap extension penalty</strong> to every subsequent position in the same gap.</p>
<p>The reason for doing this is that it is likely that adjacent gap positions were created by the same insertion or deletion event, rather than by several independent insertion or deletion events. Therefore, we don’t want to penalize a 3-letter gap (AAA—AAA) as much as we would penalize three separate 1-letter gaps (AA-A-A-AA), as the 3-letter gap may have arisen due to just one insertion or deletion event, while the 3 separate 1-letter gaps probably arose due to three independent insertion or deletion events.</p>
<p>For example, if we want to compute the score for a global alignment of two short DNA sequences ‘GAATTC’ and ‘GATTA,’ we can use the <strong>Needleman-Wunsch</strong> algorithm to calculate the highest-scoring alignment using a particular scoring function.</p>
<p>The <code>pairwiseAlignment()</code> function in the Biostrings package finds the score for the optimal global alignment between two sequences using the Needleman-Wunsch algorithm, given a particular scoring system.</p>
<p>As arguments (inputs), <code>pairwiseAlignment()</code> takes</p>
<ol style="list-style-type: decimal">
<li>the two sequences that you want to align,</li>
<li>the scoring matrix,</li>
<li>the gap opening penalty, and</li>
<li>the gap extension penalty.</li>
</ol>
<p>You can also tell the function that you want to just have the optimal global alignment’s score by setting <code>scoreOnly = TRUE</code>, or that you want to have both the optimal global alignment and its score by setting <code>scoreOnly = FALSE</code>.</p>
<p>For example, let’s find the score for the optimal global alignment between the sequences ‘GAATTC’ and ‘GATTA.’</p>
<p>First, we’ll store the sequences as <strong>character vectors</strong>:</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="global-proteins-aligments-in-r.html#cb292-1" aria-hidden="true" tabindex="-1"></a>s1 <span class="ot">&lt;-</span> <span class="st">&quot;GAATTC&quot;</span></span>
<span id="cb292-2"><a href="global-proteins-aligments-in-r.html#cb292-2" aria-hidden="true" tabindex="-1"></a>s2 <span class="ot">&lt;-</span> <span class="st">&quot;GATTA&quot;</span></span></code></pre></div>
<p>Now we’ll align them:</p>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="global-proteins-aligments-in-r.html#cb293-1" aria-hidden="true" tabindex="-1"></a>globalAligns1s2 <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">pairwiseAlignment</span>(s1, s2, </span>
<span id="cb293-2"><a href="global-proteins-aligments-in-r.html#cb293-2" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">substitutionMatrix =</span> sigma, </span>
<span id="cb293-3"><a href="global-proteins-aligments-in-r.html#cb293-3" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">gapOpening =</span> <span class="sc">-</span><span class="dv">2</span>,</span>
<span id="cb293-4"><a href="global-proteins-aligments-in-r.html#cb293-4" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">gapExtension =</span> <span class="sc">-</span><span class="dv">8</span>, </span>
<span id="cb293-5"><a href="global-proteins-aligments-in-r.html#cb293-5" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">scoreOnly =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>The output:</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="global-proteins-aligments-in-r.html#cb294-1" aria-hidden="true" tabindex="-1"></a>globalAligns1s2 </span></code></pre></div>
<pre><code>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: GAATTC
## subject: GA-TTA
## score: -3</code></pre>
<p>The above commands print out the optimal global alignment for the two sequences and its score.</p>
<p><strong>Note</strong> we set <code>gapOpening</code> to be -2 and <code>gapExtension</code> to be -8, which means that the first position of a gap is assigned a score of -8 - 2= -10, and every subsequent position in a gap is given a score of -8. Here the alignment contains four matches, one mismatch, and one gap of length 1, so its score is (4<em>2)+(1</em>-1)+(1*-10) = -3.</p>
</div>
<div id="pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm" class="section level2" number="18.3">
<h2><span class="header-section-number">18.3</span> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</h2>
<p>As well as DNA alignments, it is also possible to make alignments of protein sequences. In this case it is necessary to use a scoring matrix for amino acids rather than for nucleotides.</p>
<div id="protein-score-matrices" class="section level3" number="18.3.1">
<h3><span class="header-section-number">18.3.1</span> Protein score matrices</h3>
<p>There are several well known scoring matrices that come with <em>R</em>, such as the <strong>BLOSUM</strong> series of matrices. Different BLOSUM matrices exist, named with different numbers. BLOSUM with high numbers are designed for comparing closely related sequences, while BLOSUM with low numbers are designed for comparing evolutionarily distantly related sequences. For example, <strong>BLOSUM62</strong> is used for <strong>less divergent alignments</strong> (alignments of sequences that differ little), and <strong>BLOSUM30</strong> is used for more divergent alignments (alignments of sequences that differ a lot).</p>
<p>Many <em>R</em> packages come with example data sets or data files and you use the <code>data()</code> function is used to load these data files. You can use the <code>data()</code> function to load a data set of BLOSUM matrices that comes with <code>Biostrings</code></p>
<p>To load the BLOSUM50 matrix, we type:</p>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="global-proteins-aligments-in-r.html#cb296-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(BLOSUM50)</span>
<span id="cb296-2"><a href="global-proteins-aligments-in-r.html#cb296-2" aria-hidden="true" tabindex="-1"></a>BLOSUM50 <span class="co"># Print out the data</span></span></code></pre></div>
<pre><code>##    A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  B  Z  X  *
## A  5 -2 -1 -2 -1 -1 -1  0 -2 -1 -2 -1 -1 -3 -1  1  0 -3 -2  0 -2 -1 -1 -5
## R -2  7 -1 -2 -4  1  0 -3  0 -4 -3  3 -2 -3 -3 -1 -1 -3 -1 -3 -1  0 -1 -5
## N -1 -1  7  2 -2  0  0  0  1 -3 -4  0 -2 -4 -2  1  0 -4 -2 -3  4  0 -1 -5
## D -2 -2  2  8 -4  0  2 -1 -1 -4 -4 -1 -4 -5 -1  0 -1 -5 -3 -4  5  1 -1 -5
## C -1 -4 -2 -4 13 -3 -3 -3 -3 -2 -2 -3 -2 -2 -4 -1 -1 -5 -3 -1 -3 -3 -2 -5
## Q -1  1  0  0 -3  7  2 -2  1 -3 -2  2  0 -4 -1  0 -1 -1 -1 -3  0  4 -1 -5
## E -1  0  0  2 -3  2  6 -3  0 -4 -3  1 -2 -3 -1 -1 -1 -3 -2 -3  1  5 -1 -5
## G  0 -3  0 -1 -3 -2 -3  8 -2 -4 -4 -2 -3 -4 -2  0 -2 -3 -3 -4 -1 -2 -2 -5
## H -2  0  1 -1 -3  1  0 -2 10 -4 -3  0 -1 -1 -2 -1 -2 -3  2 -4  0  0 -1 -5
## I -1 -4 -3 -4 -2 -3 -4 -4 -4  5  2 -3  2  0 -3 -3 -1 -3 -1  4 -4 -3 -1 -5
## L -2 -3 -4 -4 -2 -2 -3 -4 -3  2  5 -3  3  1 -4 -3 -1 -2 -1  1 -4 -3 -1 -5
## K -1  3  0 -1 -3  2  1 -2  0 -3 -3  6 -2 -4 -1  0 -1 -3 -2 -3  0  1 -1 -5
## M -1 -2 -2 -4 -2  0 -2 -3 -1  2  3 -2  7  0 -3 -2 -1 -1  0  1 -3 -1 -1 -5
## F -3 -3 -4 -5 -2 -4 -3 -4 -1  0  1 -4  0  8 -4 -3 -2  1  4 -1 -4 -4 -2 -5
## P -1 -3 -2 -1 -4 -1 -1 -2 -2 -3 -4 -1 -3 -4 10 -1 -1 -4 -3 -3 -2 -1 -2 -5
## S  1 -1  1  0 -1  0 -1  0 -1 -3 -3  0 -2 -3 -1  5  2 -4 -2 -2  0  0 -1 -5
## T  0 -1  0 -1 -1 -1 -1 -2 -2 -1 -1 -1 -1 -2 -1  2  5 -3 -2  0  0 -1  0 -5
## W -3 -3 -4 -5 -5 -1 -3 -3 -3 -3 -2 -3 -1  1 -4 -4 -3 15  2 -3 -5 -2 -3 -5
## Y -2 -1 -2 -3 -3 -1 -2 -3  2 -1 -1 -2  0  4 -3 -2 -2  2  8 -1 -3 -2 -1 -5
## V  0 -3 -3 -4 -1 -3 -3 -4 -4  4  1 -3  1 -1 -3 -2  0 -3 -1  5 -4 -3 -1 -5
## B -2 -1  4  5 -3  0  1 -1  0 -4 -4  0 -3 -4 -2  0  0 -5 -3 -4  5  2 -1 -5
## Z -1  0  0  1 -3  4  5 -2  0 -3 -3  1 -1 -4 -1  0 -1 -2 -2 -3  2  5 -1 -5
## X -1 -1 -1 -1 -2 -1 -1 -2 -1 -1 -1 -1 -1 -2 -2 -1  0 -3 -1 -1 -1 -1 -1 -5
## * -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5 -5  1</code></pre>
<p>You can get a list of the available scoring matrices that come with the Biostrings package by using the data() function, which takes as an argument the name of the package for which you want to know the data sets that come with it:</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="global-proteins-aligments-in-r.html#cb298-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="at">package=</span><span class="st">&quot;Biostrings&quot;</span>)</span></code></pre></div>
<p>Another well-known series of scoring matrices are the <strong>PAM</strong> matrices developed by <a href="https://en.wikipedia.org/wiki/Margaret_Oakley_Dayhoff">Margaret Dayhoff</a> and her team. These have largely been replaced by BLOSUM but are important for historical reasons because they represent one of the first major bioinformatics, computational biology, and phyolgenetics projects ever.</p>
</div>
<div id="example-protein-alignment" class="section level3" number="18.3.2">
<h3><span class="header-section-number">18.3.2</span> Example protein alignment</h3>
<p>Let’s find the optimal global alignment between the protein sequences “PAWHEAE” and “HEAGAWGHEE” using the Needleman-Wunsch algorithm using the BLOSUM50 matrix.</p>
<p>First, load the scoring matrix <code>BLOSUM50</code> and make vectors for the sequence</p>
<div class="sourceCode" id="cb299"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb299-1"><a href="global-proteins-aligments-in-r.html#cb299-1" aria-hidden="true" tabindex="-1"></a><span class="co"># matrix</span></span>
<span id="cb299-2"><a href="global-proteins-aligments-in-r.html#cb299-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(BLOSUM50)</span>
<span id="cb299-3"><a href="global-proteins-aligments-in-r.html#cb299-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb299-4"><a href="global-proteins-aligments-in-r.html#cb299-4" aria-hidden="true" tabindex="-1"></a><span class="co"># sequences</span></span>
<span id="cb299-5"><a href="global-proteins-aligments-in-r.html#cb299-5" aria-hidden="true" tabindex="-1"></a>s3 <span class="ot">&lt;-</span> <span class="st">&quot;PAWHEAE&quot;</span></span>
<span id="cb299-6"><a href="global-proteins-aligments-in-r.html#cb299-6" aria-hidden="true" tabindex="-1"></a>s4 <span class="ot">&lt;-</span> <span class="st">&quot;HEAGAWGHEE&quot;</span></span></code></pre></div>
<p>Now do the alignments.</p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="global-proteins-aligments-in-r.html#cb300-1" aria-hidden="true" tabindex="-1"></a>globalAligns3s4 <span class="ot">&lt;-</span> <span class="fu">pairwiseAlignment</span>(s3, s4,</span>
<span id="cb300-2"><a href="global-proteins-aligments-in-r.html#cb300-2" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">substitutionMatrix =</span> <span class="st">&quot;BLOSUM50&quot;</span>, </span>
<span id="cb300-3"><a href="global-proteins-aligments-in-r.html#cb300-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">gapOpening =</span> <span class="sc">-</span><span class="dv">2</span>,</span>
<span id="cb300-4"><a href="global-proteins-aligments-in-r.html#cb300-4" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">gapExtension =</span> <span class="sc">-</span><span class="dv">8</span>, </span>
<span id="cb300-5"><a href="global-proteins-aligments-in-r.html#cb300-5" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">scoreOnly =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Look at the results:</p>
<div class="sourceCode" id="cb301"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb301-1"><a href="global-proteins-aligments-in-r.html#cb301-1" aria-hidden="true" tabindex="-1"></a>globalAligns3s4 <span class="co"># Print out the optimal global alignment and its score</span></span></code></pre></div>
<pre><code>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: P---AWHEAE
## subject: HEAGAWGHEE
## score: -5</code></pre>
<p>We set <code>gapOpening</code> to be -2 and <code>gapExtension</code> to be -8, which means that the first position of a gap is assigned a score of -8-2=-10, and every subsequent position in a gap is given a score of -8. This means that the gap will be given a score of -10-8-8 = -26.</p>
</div>
</div>
<div id="aligning-uniprot-sequences" class="section level2" number="18.4">
<h2><span class="header-section-number">18.4</span> Aligning UniProt sequences</h2>
<p>We discussed previously how you can search for UniProt accessions and retrieve the corresponding protein sequences, either via the UniProt website or using the <code>rentrez</code> package.</p>
<p>In the examples given above, we learned how to retrieve the sequences for the chorismate lyase proteins from <em>Mycobacterium leprae</em> (UniProt Q9CD83) and <em>Mycobacterium ulcerans</em> (UniProt A0PQ23), and read them into R, and store them as vectors lepraeseq and ulceransseq.</p>
<p>You can align these sequences using <code>pairwiseAlignment()</code> from the Biostrings package.</p>
<p>As its input, the <code>pairwiseAlignment()</code>function requires that the sequences be in the form of a single string (e.g. “ACGTA”), rather than as a vector of characters (e.g. a vector with the first element as “A,” the second element as “C,” etc.). Therefore, to align the <em>M. leprae</em> and <em>M. ulcerans</em> chorismate lyase proteins, we first need to convert the vectors lepraeeq and ulceransseq into strings. We can do this using the <code>paste()</code> function:</p>
<div class="sourceCode" id="cb303"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb303-1"><a href="global-proteins-aligments-in-r.html#cb303-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert leprae_vector to an object lepraeseq_string</span></span>
<span id="cb303-2"><a href="global-proteins-aligments-in-r.html#cb303-2" aria-hidden="true" tabindex="-1"></a>lepraeseq_string <span class="ot">&lt;-</span><span class="fu">paste</span>(leprae_vector,<span class="at">collapse =</span> <span class="st">&quot;&quot;</span>)    </span>
<span id="cb303-3"><a href="global-proteins-aligments-in-r.html#cb303-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb303-4"><a href="global-proteins-aligments-in-r.html#cb303-4" aria-hidden="true" tabindex="-1"></a><span class="co"># convert ulcerans_vector to an object ulceransseq_string</span></span>
<span id="cb303-5"><a href="global-proteins-aligments-in-r.html#cb303-5" aria-hidden="true" tabindex="-1"></a>ulceransseq_string <span class="ot">&lt;-</span><span class="fu">paste</span>(ulcerans_vector,<span class="at">collapse =</span> <span class="st">&quot;&quot;</span>) </span></code></pre></div>
<p>Furthermore, pairwiseAlignment() requires that the sequences be stored as uppercase characters. Therefore, if they are not already in uppercase, we need to use the <code>toupper()</code> function to convert <code>lepraeseq_string</code> and <code>ulceransseq_string</code> to uppercase:</p>
<div class="sourceCode" id="cb304"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb304-1"><a href="global-proteins-aligments-in-r.html#cb304-1" aria-hidden="true" tabindex="-1"></a>lepraeseq_string   <span class="ot">&lt;-</span> <span class="fu">toupper</span>(lepraeseq_string)</span>
<span id="cb304-2"><a href="global-proteins-aligments-in-r.html#cb304-2" aria-hidden="true" tabindex="-1"></a>ulceransseq_string <span class="ot">&lt;-</span> <span class="fu">toupper</span>(ulceransseq_string)</span></code></pre></div>
<p>Check the output</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="global-proteins-aligments-in-r.html#cb305-1" aria-hidden="true" tabindex="-1"></a>lepraeseq_string <span class="co"># Print out the content of &quot;lepraeseq_string&quot;</span></span></code></pre></div>
<pre><code>## [1] &quot;MTNRTLSREEIRKLDRDLRILVATNGTLTRVLNVVANEEIVVDIINQQLLDVAPKIPELENLKIGRILQRDILLKGQKSGILFVAAESLIVIDLLPTAITTYLTKTHHPIGEIMAASRIETYKEDAQVWIGDLPCWLADYGYWDLPKRAVGRRYRIIAGGQPVIITTEYFLRSVFQDTPREELDRCQYSNDIDTRSGDRFVLHGRVFKNL&quot;</code></pre>
<p>We can now align the the <em>M. leprae</em> and <em>M. ulcerans</em> chorismate lyase protein sequences using the <code>pairwiseAlignment()</code> function:</p>
<div class="sourceCode" id="cb307"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb307-1"><a href="global-proteins-aligments-in-r.html#cb307-1" aria-hidden="true" tabindex="-1"></a>globalAlignLepraeUlcerans <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">pairwiseAlignment</span>(lepraeseq_string, </span>
<span id="cb307-2"><a href="global-proteins-aligments-in-r.html#cb307-2" aria-hidden="true" tabindex="-1"></a>                                               ulceransseq_string,</span>
<span id="cb307-3"><a href="global-proteins-aligments-in-r.html#cb307-3" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">substitutionMatrix =</span> BLOSUM50, </span>
<span id="cb307-4"><a href="global-proteins-aligments-in-r.html#cb307-4" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">gapOpening =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb307-5"><a href="global-proteins-aligments-in-r.html#cb307-5" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">gapExtension =</span> <span class="sc">-</span><span class="dv">8</span>, </span>
<span id="cb307-6"><a href="global-proteins-aligments-in-r.html#cb307-6" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">scoreOnly =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>The output:</p>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb308-1"><a href="global-proteins-aligments-in-r.html#cb308-1" aria-hidden="true" tabindex="-1"></a>globalAlignLepraeUlcerans <span class="co"># Print out the optimal global alignment and its score</span></span></code></pre></div>
<pre><code>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: MT-----NR--T---LSREEIRKLDRDLRILVAT...DTPREELDRCQYSNDIDTRSGDRFVLHGRVFKNL
## subject: MLAVLPEKREMTECHLSDEEIRKLNRDLRILIAT...DNSREEPIRHQRS--VGT-SA-R---SGRSICT-
## score: 627</code></pre>
<p>As the alignment is very long, when you type <code>globalAlignLepraeUlcerans</code>, you only see the start and the end of the alignment. Therefore, we need to have a function to print out the whole alignment (see below).</p>
</div>
<div id="viewing-a-long-pairwise-alignment" class="section level2" number="18.5">
<h2><span class="header-section-number">18.5</span> Viewing a long pairwise alignment</h2>
<p>If you want to view a long pairwise alignment such as that between the <em>M. leprae</em> and <em>M. ulerans</em> chorismate lyase proteins, it is convenient to print out the alignment in blocks.</p>
<p>The R function <code>printPairwiseAlignment()</code> below will do this for you:</p>
<div class="sourceCode" id="cb310"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb310-1"><a href="global-proteins-aligments-in-r.html#cb310-1" aria-hidden="true" tabindex="-1"></a><span class="fu">printPairwiseAlignment</span>(globalAlignLepraeUlcerans, <span class="dv">60</span>)</span></code></pre></div>
<pre><code>## [1] &quot;MT-----NR--T---LSREEIRKLDRDLRILVATNGTLTRVLNVVANEEIVVDIINQQLL 50&quot;
## [1] &quot;MLAVLPEKREMTECHLSDEEIRKLNRDLRILIATNGTLTRILNVLANDEIVVEIVKQQIQ 60&quot;
## [1] &quot; &quot;
## [1] &quot;DVAPKIPELENLKIGRILQRDILLKGQKSGILFVAAESLIVIDLLPTAITTYLTKTHHPI 110&quot;
## [1] &quot;DAAPEMDGCDHSSIGRVLRRDIVLKGRRSGIPFVAAESFIAIDLLPPEIVASLLETHRPI 120&quot;
## [1] &quot; &quot;
## [1] &quot;GEIMAASRIETYKEDAQVWIGDLPCWLADYGYWDLPKRAVGRRYRIIAGGQPVIITTEYF 170&quot;
## [1] &quot;GEVMAASCIETFKEEAKVWAGESPAWLELDRRRNLPPKVVGRQYRVIAEGRPVIIITEYF 180&quot;
## [1] &quot; &quot;
## [1] &quot;LRSVFQDTPREELDRCQYSNDIDTRSGDRFVLHGRVFKN 230&quot;
## [1] &quot;LRSVFEDNSREEPIRHQRS--VGT-SA-R---SGRSICT 233&quot;
## [1] &quot; &quot;</code></pre>
<p>The position in the protein of the amino acid that is at the end of each line of the printed alignment is shown after the end of the line. For example, the first line of the alignment above finishes at amino acid position 50 in the <em>M. leprae</em> protein and also at amino acid position 60 in the <em>M. ulcerans</em> protein. Becaues there as a difference of 60-50 = 10 bases, there must be 10 insertions in the <em>M. leprae</em> to get it to line up. Count the number of dashes in the sequence to see how many there are.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="sequence-dotplots-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="local-protein-aligments-in-r.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/16_alignment_global_protein.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/16_alignment_global_protein.Rmd",
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
