<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 44 Local variation in GC content | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 44 Local variation in GC content | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 44 Local variation in GC content | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="unrooted-neighbor-joining-phylogenetic-trees.html"/>
<link rel="next" href="computational-gene-finding.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.5.1/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>1</b> How to review this book</a>
<ul>
<li class="chapter" data-level="1.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>1.1</b> File format</a></li>
<li class="chapter" data-level="1.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>1.2</b> What can you do?</a></li>
<li class="chapter" data-level="1.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>1.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>1.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="1.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>1.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="1.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>1.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="1.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>1.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="test-audio-link.html"><a href="test-audio-link.html"><i class="fa fa-check"></i><b>2</b> Test audio link</a></li>
<li class="chapter" data-level="3" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>3</b> Test shiny app</a></li>
<li class="chapter" data-level="4" data-path="test-youtube-vidoe.html"><a href="test-youtube-vidoe.html"><i class="fa fa-check"></i><b>4</b> Test YouTube vidoe</a></li>
<li class="chapter" data-level="5" data-path="test-google-doc-insertion.html"><a href="test-google-doc-insertion.html"><i class="fa fa-check"></i><b>5</b> Test google doc insertion</a></li>
<li class="chapter" data-level="6" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>6</b> Datacamp test</a></li>
<li class="chapter" data-level="7" data-path="an-example-of-a-plotly-graph.html"><a href="an-example-of-a-plotly-graph.html"><i class="fa fa-check"></i><b>7</b> An example of a plotly graph</a></li>
<li class="chapter" data-level="8" data-path="directly-embedding-of-a-libretext-page.html"><a href="directly-embedding-of-a-libretext-page.html"><i class="fa fa-check"></i><b>8</b> Directly embedding of a libretext page</a></li>
<li class="chapter" data-level="9" data-path="libretext-image.html"><a href="libretext-image.html"><i class="fa fa-check"></i><b>9</b> Libretext image</a>
<ul>
<li class="chapter" data-level="9.1" data-path="libretext-image.html"><a href="libretext-image.html#prelude-to-population-and-community-ecology"><i class="fa fa-check"></i><b>9.1</b> 45.0: Prelude to Population and Community Ecology</a></li>
<li class="chapter" data-level="9.2" data-path="libretext-image.html"><a href="libretext-image.html#note"><i class="fa fa-check"></i><b>9.2</b> Note</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="nucleic-acids.html"><a href="nucleic-acids.html"><i class="fa fa-check"></i><b>10</b> Nucleic Acids</a>
<ul>
<li class="chapter" data-level="10.1" data-path="nucleic-acids.html"><a href="nucleic-acids.html#dna-and-rna"><i class="fa fa-check"></i><b>10.1</b> DNA and RNA</a></li>
<li class="chapter" data-level="10.2" data-path="nucleic-acids.html"><a href="nucleic-acids.html#dna-double-helix-structure"><i class="fa fa-check"></i><b>10.2</b> DNA Double-Helix Structure</a></li>
<li class="chapter" data-level="10.3" data-path="nucleic-acids.html"><a href="nucleic-acids.html#rna"><i class="fa fa-check"></i><b>10.3</b> RNA</a></li>
<li class="chapter" data-level="10.4" data-path="nucleic-acids.html"><a href="nucleic-acids.html#summary"><i class="fa fa-check"></i><b>10.4</b> Summary</a></li>
<li class="chapter" data-level="10.5" data-path="nucleic-acids.html"><a href="nucleic-acids.html#analysis-questions"><i class="fa fa-check"></i><b>10.5</b> Analysis Questions</a></li>
<li class="chapter" data-level="10.6" data-path="nucleic-acids.html"><a href="nucleic-acids.html#glossary"><i class="fa fa-check"></i><b>10.6</b> Glossary</a></li>
<li class="chapter" data-level="10.7" data-path="nucleic-acids.html"><a href="nucleic-acids.html#contributors-and-attributions"><i class="fa fa-check"></i><b>10.7</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>11</b> Downloading R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>11.1</b> Preface</a></li>
<li class="chapter" data-level="11.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>11.2</b> Introduction to R</a></li>
<li class="chapter" data-level="11.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>11.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>11.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="11.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>11.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>11.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>12</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="12.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>12.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="12.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>12.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>13</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="13.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>13.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="13.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>13.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="13.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>13.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>14</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="14.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>14.1</b> Bioconductor</a></li>
<li class="chapter" data-level="14.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>14.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="14.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>14.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>14.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="14.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>14.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="14.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>14.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>14.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>15</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="15.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>15.1</b> Vocabulary</a></li>
<li class="chapter" data-level="15.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>15.2</b> R functions</a></li>
<li class="chapter" data-level="15.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>15.3</b> Interacting with R</a></li>
<li class="chapter" data-level="15.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>15.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="15.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>15.4.1</b> Vectors</a></li>
<li class="chapter" data-level="15.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>15.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="15.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>15.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="15.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>15.4.4</b> Lists</a></li>
<li class="chapter" data-level="15.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>15.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="15.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>15.5</b> Arguments</a></li>
<li class="chapter" data-level="15.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>15.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="15.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>15.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="15.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>15.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="15.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>15.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="15.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>15.9</b> Quiting R</a></li>
<li class="chapter" data-level="15.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>15.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>16</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="16.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-1"><i class="fa fa-check"></i><b>16.1</b> Preface</a></li>
<li class="chapter" data-level="16.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>16.2</b> Vocab</a></li>
<li class="chapter" data-level="16.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions"><i class="fa fa-check"></i><b>16.3</b> Functions</a></li>
<li class="chapter" data-level="16.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>16.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="16.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>16.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>17</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="17.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-2"><i class="fa fa-check"></i><b>17.1</b> Preface</a></li>
<li class="chapter" data-level="17.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>17.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="functions-1.html"><a href="functions-1.html"><i class="fa fa-check"></i><b>18</b> Functions</a>
<ul>
<li class="chapter" data-level="18.1" data-path="functions-1.html"><a href="functions-1.html#vectors-in-r"><i class="fa fa-check"></i><b>18.1</b> Vectors in R</a></li>
<li class="chapter" data-level="18.2" data-path="functions-1.html"><a href="functions-1.html#math-on-vectors"><i class="fa fa-check"></i><b>18.2</b> Math on vectors</a></li>
<li class="chapter" data-level="18.3" data-path="functions-1.html"><a href="functions-1.html#functions-on-vectors"><i class="fa fa-check"></i><b>18.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="18.4" data-path="functions-1.html"><a href="functions-1.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>18.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="18.5" data-path="functions-1.html"><a href="functions-1.html#subsetting-vectors"><i class="fa fa-check"></i><b>18.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="18.6" data-path="functions-1.html"><a href="functions-1.html#sequences-of-numbers"><i class="fa fa-check"></i><b>18.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="18.7" data-path="functions-1.html"><a href="functions-1.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>18.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="18.8" data-path="functions-1.html"><a href="functions-1.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>18.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>19</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="19.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-3"><i class="fa fa-check"></i><b>19.1</b> Preface</a></li>
<li class="chapter" data-level="19.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>19.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="19.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>19.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>20</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="20.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-4"><i class="fa fa-check"></i><b>20.1</b> Preface</a></li>
<li class="chapter" data-level="20.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>20.2</b> Vocab</a></li>
<li class="chapter" data-level="20.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-2"><i class="fa fa-check"></i><b>20.3</b> Functions</a></li>
<li class="chapter" data-level="20.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>20.4</b> Functions in R</a></li>
<li class="chapter" data-level="20.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>20.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html"><i class="fa fa-check"></i><b>21</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="21.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#commands-used"><i class="fa fa-check"></i><b>21.1</b> Commands used</a></li>
<li class="chapter" data-level="21.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#r-objects"><i class="fa fa-check"></i><b>21.2</b> R Objects</a></li>
<li class="chapter" data-level="21.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>21.3</b> Differences between objects</a></li>
<li class="chapter" data-level="21.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-data"><i class="fa fa-check"></i><b>21.4</b> The Data</a></li>
<li class="chapter" data-level="21.5" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>21.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="21.5.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#is"><i class="fa fa-check"></i><b>21.5.1</b> is()</a></li>
<li class="chapter" data-level="21.5.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#length"><i class="fa fa-check"></i><b>21.5.2</b> length()</a></li>
<li class="chapter" data-level="21.5.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#str"><i class="fa fa-check"></i><b>21.5.3</b> str()</a></li>
<li class="chapter" data-level="21.5.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#c"><i class="fa fa-check"></i><b>21.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="21.6" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#debrief"><i class="fa fa-check"></i><b>21.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="fasta-files.html"><a href="fasta-files.html"><i class="fa fa-check"></i><b>22</b> FASTA Files</a>
<ul>
<li class="chapter" data-level="22.1" data-path="fasta-files.html"><a href="fasta-files.html#example-fasta-file"><i class="fa fa-check"></i><b>22.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="22.2" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>22.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="22.3" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>22.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="22.4" data-path="fasta-files.html"><a href="fasta-files.html#fastq-format"><i class="fa fa-check"></i><b>22.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>23</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="23.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-3"><i class="fa fa-check"></i><b>23.0.1</b> Functions</a></li>
<li class="chapter" data-level="23.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>23.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="23.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>23.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="23.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>23.0.4</b> File types</a></li>
<li class="chapter" data-level="23.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>23.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="23.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>23.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="23.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries"><i class="fa fa-check"></i><b>23.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="23.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>23.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="23.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>23.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="23.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>23.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="23.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>23.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="23.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>23.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="23.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>23.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="23.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>23.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="23.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>23.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>24</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="24.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-1"><i class="fa fa-check"></i><b>24.1</b> Preliminaries</a></li>
<li class="chapter" data-level="24.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>24.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="24.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>24.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="24.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>24.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="25" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>25</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="25.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-5"><i class="fa fa-check"></i><b>25.1</b> Preface</a></li>
<li class="chapter" data-level="25.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>25.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="25.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>25.3</b> Introduction</a></li>
<li class="chapter" data-level="25.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>25.4</b> Vocabulary</a></li>
<li class="chapter" data-level="25.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions-4"><i class="fa fa-check"></i><b>25.5</b> Functions</a></li>
<li class="chapter" data-level="25.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries-2"><i class="fa fa-check"></i><b>25.6</b> Preliminaries</a></li>
<li class="chapter" data-level="25.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>25.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="25.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>25.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="25.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>25.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="25.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>25.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="25.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>25.8.3</b> DNA words</a></li>
<li class="chapter" data-level="25.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary-1"><i class="fa fa-check"></i><b>25.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="25.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>25.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="25.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>25.9.1</b> License</a></li>
<li class="chapter" data-level="25.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>25.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="26" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>26</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="26.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>26.1</b> Preliminaries</a></li>
<li class="chapter" data-level="26.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>26.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="changes-to-database-entries.html"><a href="changes-to-database-entries.html"><i class="fa fa-check"></i><b>27</b> Changes to database entries</a></li>
<li class="chapter" data-level="28" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>28</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="28.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>28.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="28.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>28.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="28.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>28.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="28.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>28.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="28.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>28.4</b> Inversions</a></li>
<li class="chapter" data-level="28.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>28.5</b> Translocations</a></li>
<li class="chapter" data-level="28.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>28.6</b> Random sequence</a></li>
<li class="chapter" data-level="28.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>28.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>29</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="29.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>29.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="29.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>29.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="29.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>29.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="29.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>29.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="29.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>29.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="29.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>29.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="29.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>29.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="29.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>29.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>30</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="30.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>30.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="30.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>30.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="30.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>30.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="31" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html"><i class="fa fa-check"></i><b>31</b> Alignment by eye in R</a>
<ul>
<li class="chapter" data-level="31.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>31.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="31.1.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#packages"><i class="fa fa-check"></i><b>31.1.1</b> Packages</a></li>
<li class="chapter" data-level="31.1.2" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#data"><i class="fa fa-check"></i><b>31.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="32" data-path="alignment-in-r.html"><a href="alignment-in-r.html"><i class="fa fa-check"></i><b>32</b> Alignment in R</a>
<ul>
<li class="chapter" data-level="32.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#preliminaries-8"><i class="fa fa-check"></i><b>32.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="32.1.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#packages-1"><i class="fa fa-check"></i><b>32.1.1</b> Packages</a></li>
<li class="chapter" data-level="32.1.2" data-path="alignment-in-r.html"><a href="alignment-in-r.html#data-1"><i class="fa fa-check"></i><b>32.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="33" data-path="working-with-vectors-and-matrices.html"><a href="working-with-vectors-and-matrices.html"><i class="fa fa-check"></i><b>33</b> Working with vectors and matrices</a></li>
<li class="chapter" data-level="34" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html"><i class="fa fa-check"></i><b>34</b> Matrix vocabulary</a>
<ul>
<li class="chapter" data-level="34.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#introduction-1"><i class="fa fa-check"></i><b>34.1</b> Introduction</a></li>
<li class="chapter" data-level="34.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#preliminaries-9"><i class="fa fa-check"></i><b>34.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="34.2.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#packages-2"><i class="fa fa-check"></i><b>34.2.1</b> Packages</a></li>
<li class="chapter" data-level="34.2.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#vocab-3"><i class="fa fa-check"></i><b>34.2.2</b> Vocab</a></li>
<li class="chapter" data-level="34.2.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#functions-used"><i class="fa fa-check"></i><b>34.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="34.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#load-packages"><i class="fa fa-check"></i><b>34.3</b> Load packages</a></li>
<li class="chapter" data-level="34.4" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#matrix-structure"><i class="fa fa-check"></i><b>34.4</b> Matrix structure</a></li>
<li class="chapter" data-level="34.5" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#diagonal-of-a-matrix"><i class="fa fa-check"></i><b>34.5</b> Diagonal of a matrix</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html"><i class="fa fa-check"></i><b>35</b> Testing the significance of an alignment</a>
<ul>
<li class="chapter" data-level="35.1" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#calculating-the-statistical-significance-of-a-pairwise-global-alignment"><i class="fa fa-check"></i><b>35.1</b> Calculating the statistical significance of a pairwise global alignment</a></li>
<li class="chapter" data-level="35.2" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#summary-2"><i class="fa fa-check"></i><b>35.2</b> Summary</a></li>
<li class="chapter" data-level="35.3" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#links-and-further-reading-1"><i class="fa fa-check"></i><b>35.3</b> Links and Further Reading</a></li>
<li class="chapter" data-level="35.4" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#exercises-1"><i class="fa fa-check"></i><b>35.4</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>36</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="36.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>36.1</b> Prelminaries</a></li>
<li class="chapter" data-level="36.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>36.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="36.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>36.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html"><i class="fa fa-check"></i><b>37</b> Manipulating matrices and vectors: worked example</a>
<ul>
<li class="chapter" data-level="37.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#introduction-2"><i class="fa fa-check"></i><b>37.1</b> Introduction</a></li>
<li class="chapter" data-level="37.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#preliminaries-10"><i class="fa fa-check"></i><b>37.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="37.2.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-3"><i class="fa fa-check"></i><b>37.2.1</b> Packages</a></li>
<li class="chapter" data-level="37.2.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#vocab-4"><i class="fa fa-check"></i><b>37.2.2</b> Vocab</a></li>
<li class="chapter" data-level="37.2.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#functions-used-1"><i class="fa fa-check"></i><b>37.2.3</b> Functions used</a></li>
<li class="chapter" data-level="37.2.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-4"><i class="fa fa-check"></i><b>37.2.4</b> Packages</a></li>
<li class="chapter" data-level="37.2.5" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#data-2"><i class="fa fa-check"></i><b>37.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="37.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#matrix-elements-can-be-accessed-with-square-bracket-notation"><i class="fa fa-check"></i><b>37.3</b> Matrix elements can be accessed with square bracket notation</a></li>
<li class="chapter" data-level="37.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#r-object-names"><i class="fa fa-check"></i><b>37.4</b> R object names</a>
<ul>
<li class="chapter" data-level="37.4.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-vectors"><i class="fa fa-check"></i><b>37.4.1</b> Accessing items in names vectors</a></li>
<li class="chapter" data-level="37.4.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-matrices"><i class="fa fa-check"></i><b>37.4.2</b> Accessing items in names matrices</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="38" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>38</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="38.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-11"><i class="fa fa-check"></i><b>38.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="38.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages-5"><i class="fa fa-check"></i><b>38.1.1</b> Packages</a></li>
<li class="chapter" data-level="38.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-5"><i class="fa fa-check"></i><b>38.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="38.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>38.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="38.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>38.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="38.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>38.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="38.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>38.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="38.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>38.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="38.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>38.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="38.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>38.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="38.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>38.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html"><i class="fa fa-check"></i><b>39</b> The BLOSUM scoring matrix in R</a>
<ul>
<li class="chapter" data-level="39.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#introduction-3"><i class="fa fa-check"></i><b>39.1</b> Introduction</a></li>
<li class="chapter" data-level="39.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#preliminaries-12"><i class="fa fa-check"></i><b>39.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="39.2.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#packages-6"><i class="fa fa-check"></i><b>39.2.1</b> Packages</a></li>
<li class="chapter" data-level="39.2.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#vocab-5"><i class="fa fa-check"></i><b>39.2.2</b> Vocab</a></li>
<li class="chapter" data-level="39.2.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#functions-used-2"><i class="fa fa-check"></i><b>39.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="39.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#load-packages-1"><i class="fa fa-check"></i><b>39.3</b> Load packages</a></li>
<li class="chapter" data-level="39.4" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#the-blosum-matrix"><i class="fa fa-check"></i><b>39.4</b> The BLOSUM matrix</a></li>
<li class="chapter" data-level="39.5" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#visualizing-the-blosum-matrix"><i class="fa fa-check"></i><b>39.5</b> Visualizing the BLOSUM matrix</a></li>
</ul></li>
<li class="chapter" data-level="40" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>40</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="41" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>41</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="41.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-4"><i class="fa fa-check"></i><b>41.1</b> Introduction</a></li>
<li class="chapter" data-level="41.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>41.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="41.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-6"><i class="fa fa-check"></i><b>41.2.1</b> Vocab</a></li>
<li class="chapter" data-level="41.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>41.2.2</b> R functions</a></li>
<li class="chapter" data-level="41.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>41.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="41.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>41.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="41.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>41.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="41.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>41.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="41.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>41.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="41.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>41.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="41.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>41.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="41.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>41.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="41.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>41.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="41.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>41.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="41.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>41.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="41.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>41.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="41.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>41.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="42" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>42</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="42.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-13"><i class="fa fa-check"></i><b>42.1</b> Preliminaries</a></li>
<li class="chapter" data-level="42.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-5"><i class="fa fa-check"></i><b>42.2</b> Introduction</a></li>
<li class="chapter" data-level="42.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>42.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="42.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>42.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="43" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>43</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="43.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-14"><i class="fa fa-check"></i><b>43.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="43.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>43.1.1</b> Key functions</a></li>
<li class="chapter" data-level="43.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>43.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="43.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>43.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="43.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>43.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="43.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>43.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="43.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>43.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="43.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>43.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="44" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html"><i class="fa fa-check"></i><b>44</b> Local variation in GC content</a>
<ul>
<li class="chapter" data-level="44.0.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#note-on-the-biology-in-this-section"><i class="fa fa-check"></i><b>44.0.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="44.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#vocabulary-2"><i class="fa fa-check"></i><b>44.1</b> Vocabulary</a></li>
<li class="chapter" data-level="44.2" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#reading-sequence-data-with-rentrezentrez_fetch"><i class="fa fa-check"></i><b>44.2</b> Reading sequence data with rentrez::entrez_fetch</a></li>
<li class="chapter" data-level="44.3" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#local-variation-in-gc-content-1"><i class="fa fa-check"></i><b>44.3</b> Local variation in GC content</a></li>
<li class="chapter" data-level="44.4" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-analysis-of-gc-content"><i class="fa fa-check"></i><b>44.4</b> A sliding window analysis of GC content</a></li>
<li class="chapter" data-level="44.5" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-plot-of-gc-content"><i class="fa fa-check"></i><b>44.5</b> A sliding window plot of GC content</a>
<ul>
<li class="chapter" data-level="44.5.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#over-represented-and-under-represented-dna-words"><i class="fa fa-check"></i><b>44.5.1</b> Over-represented and under-represented DNA words</a></li>
</ul></li>
<li class="chapter" data-level="44.6" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#summary-3"><i class="fa fa-check"></i><b>44.6</b> Summary</a></li>
<li class="chapter" data-level="44.7" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#further-reading"><i class="fa fa-check"></i><b>44.7</b> Further Reading</a></li>
<li class="chapter" data-level="44.8" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#acknowledgements-1"><i class="fa fa-check"></i><b>44.8</b> Acknowledgements</a></li>
<li class="chapter" data-level="44.9" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#exercises-2"><i class="fa fa-check"></i><b>44.9</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="45" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html"><i class="fa fa-check"></i><b>45</b> Computational gene finding</a>
<ul>
<li class="chapter" data-level="45.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#preliminaries-15"><i class="fa fa-check"></i><b>45.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="45.1.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#packages-7"><i class="fa fa-check"></i><b>45.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="45.2" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#the-genetic-code"><i class="fa fa-check"></i><b>45.2</b> The genetic code</a></li>
<li class="chapter" data-level="45.3" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-start-and-stop-codons-in-a-dna-sequence"><i class="fa fa-check"></i><b>45.3</b> Finding start and stop codons in a DNA sequence</a></li>
<li class="chapter" data-level="45.4" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#reading-frames"><i class="fa fa-check"></i><b>45.4</b> Reading frames</a></li>
<li class="chapter" data-level="45.5" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-forward-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>45.5</b> Finding open reading frames on the forward strand of a DNA sequence</a></li>
<li class="chapter" data-level="45.6" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#predicting-the-protein-sequence-for-an-orf"><i class="fa fa-check"></i><b>45.6</b> Predicting the protein sequence for an ORF</a></li>
<li class="chapter" data-level="45.7" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-reverse-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>45.7</b> Finding open reading frames on the reverse strand of a DNA sequence</a></li>
<li class="chapter" data-level="45.8" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#lengths-of-open-reading-frames"><i class="fa fa-check"></i><b>45.8</b> Lengths of open reading frames</a></li>
</ul></li>
<li class="chapter" data-level="46" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html"><i class="fa fa-check"></i><b>46</b> Significance of ORFs</a>
<ul>
<li class="chapter" data-level="46.1" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#preliminaries-16"><i class="fa fa-check"></i><b>46.1</b> Preliminaries</a></li>
<li class="chapter" data-level="46.2" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#identifying-significant-open-reading-frames"><i class="fa fa-check"></i><b>46.2</b> Identifying significant open reading frames</a></li>
<li class="chapter" data-level="46.3" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#summary-4"><i class="fa fa-check"></i><b>46.3</b> Summary</a></li>
<li class="chapter" data-level="46.4" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#links-and-further-reading-2"><i class="fa fa-check"></i><b>46.4</b> Links and Further Reading</a></li>
<li class="chapter" data-level="46.5" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#acknowledgements-2"><i class="fa fa-check"></i><b>46.5</b> Acknowledgements</a></li>
<li class="chapter" data-level="46.6" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#exercises-3"><i class="fa fa-check"></i><b>46.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="47" data-path="comparative-genomics.html"><a href="comparative-genomics.html"><i class="fa fa-check"></i><b>47</b> Comparative Genomics</a>
<ul>
<li class="chapter" data-level="47.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#preliminaries-17"><i class="fa fa-check"></i><b>47.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="47.1.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#packages-8"><i class="fa fa-check"></i><b>47.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="47.2" data-path="comparative-genomics.html"><a href="comparative-genomics.html#introduction-6"><i class="fa fa-check"></i><b>47.2</b> Introduction</a></li>
<li class="chapter" data-level="47.3" data-path="comparative-genomics.html"><a href="comparative-genomics.html#using-the-biomart-r-library-to-query-the-ensembl-database"><i class="fa fa-check"></i><b>47.3</b> Using the biomaRt R Library to Query the Ensembl Database</a></li>
<li class="chapter" data-level="47.4" data-path="comparative-genomics.html"><a href="comparative-genomics.html#comparing-the-number-of-genes-in-two-species"><i class="fa fa-check"></i><b>47.4</b> Comparing the number of genes in two species</a></li>
<li class="chapter" data-level="47.5" data-path="comparative-genomics.html"><a href="comparative-genomics.html#identifying-homologous-genes-between-two-species"><i class="fa fa-check"></i><b>47.5</b> Identifying homologous genes between two species</a></li>
</ul></li>
<li class="chapter" data-level="48" data-path="summary-5.html"><a href="summary-5.html"><i class="fa fa-check"></i><b>48</b> Summary</a></li>
<li class="chapter" data-level="49" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html"><i class="fa fa-check"></i><b>49</b> #Links and Further Reading</a>
<ul>
<li class="chapter" data-level="49.1" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#acknowledgements-3"><i class="fa fa-check"></i><b>49.1</b> Acknowledgements</a></li>
<li class="chapter" data-level="49.2" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#exercises-4"><i class="fa fa-check"></i><b>49.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="50" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html"><i class="fa fa-check"></i><b>50</b> Hidden Markov Models</a>
<ul>
<li class="chapter" data-level="50.1" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-multinomial-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>50.1</b> A multinomial model of DNA sequence evolution</a></li>
<li class="chapter" data-level="50.2" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-multinomial-model"><i class="fa fa-check"></i><b>50.2</b> Generating a DNA sequence using a multinomial model</a></li>
<li class="chapter" data-level="50.3" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>50.3</b> A Markov model of DNA sequence evolution</a></li>
<li class="chapter" data-level="50.4" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-for-a-markov-model"><i class="fa fa-check"></i><b>50.4</b> The transition matrix for a Markov model</a></li>
<li class="chapter" data-level="50.5" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-markov-model"><i class="fa fa-check"></i><b>50.5</b> Generating a DNA sequence using a Markov model</a></li>
<li class="chapter" data-level="50.6" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>50.6</b> A Hidden Markov Model of DNA sequence evolution</a></li>
<li class="chapter" data-level="50.7" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-and-emission-matrix-for-a-hmm"><i class="fa fa-check"></i><b>50.7</b> The transition matrix and emission matrix for a HMM</a></li>
<li class="chapter" data-level="50.8" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-hmm"><i class="fa fa-check"></i><b>50.8</b> Generating a DNA sequence using a HMM</a></li>
<li class="chapter" data-level="50.9" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#inferring-the-states-of-a-hmm-that-generated-a-dna-sequence"><i class="fa fa-check"></i><b>50.9</b> Inferring the states of a HMM that generated a DNA sequence</a></li>
<li class="chapter" data-level="50.10" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-protein-sequence-evolution"><i class="fa fa-check"></i><b>50.10</b> A Hidden Markov Model of protein sequence evolution</a></li>
<li class="chapter" data-level="50.11" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#summary-6"><i class="fa fa-check"></i><b>50.11</b> Summary</a></li>
<li class="chapter" data-level="50.12" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#links-and-further-reading-4"><i class="fa fa-check"></i><b>50.12</b> Links and Further Reading</a></li>
<li class="chapter" data-level="50.13" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#exercises-5"><i class="fa fa-check"></i><b>50.13</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="51" data-path="blast-summary.html"><a href="blast-summary.html"><i class="fa fa-check"></i><b>51</b> BLAST Summary</a>
<ul>
<li class="chapter" data-level="51.1" data-path="blast-summary.html"><a href="blast-summary.html#blast-overview"><i class="fa fa-check"></i><b>51.1</b> BLAST Overview</a>
<ul>
<li class="chapter" data-level="51.1.1" data-path="blast-summary.html"><a href="blast-summary.html#why-do-a-blast-search"><i class="fa fa-check"></i><b>51.1.1</b> Why do a BLAST search?</a></li>
<li class="chapter" data-level="51.1.2" data-path="blast-summary.html"><a href="blast-summary.html#what-does-blast-let-you-do"><i class="fa fa-check"></i><b>51.1.2</b> What does BLAST let you do?</a></li>
</ul></li>
<li class="chapter" data-level="51.2" data-path="blast-summary.html"><a href="blast-summary.html#blast-recipe"><i class="fa fa-check"></i><b>51.2</b> BLAST recipe</a></li>
<li class="chapter" data-level="51.3" data-path="blast-summary.html"><a href="blast-summary.html#modules-in-the-blast-workflow-and-algorithm"><i class="fa fa-check"></i><b>51.3</b> Modules in the BLAST Workflow and Algorithm</a>
<ul>
<li class="chapter" data-level="51.3.1" data-path="blast-summary.html"><a href="blast-summary.html#module-1-setup-query-sequence-protein-blast-oriented"><i class="fa fa-check"></i><b>51.3.1</b> Module 1: Setup query sequence (Protein BLAST oriented)</a></li>
<li class="chapter" data-level="51.3.2" data-path="blast-summary.html"><a href="blast-summary.html#module-2-scanning-database-for-high-scoring-words"><i class="fa fa-check"></i><b>51.3.2</b> Module 2: Scanning database for high-scoring words</a></li>
<li class="chapter" data-level="51.3.3" data-path="blast-summary.html"><a href="blast-summary.html#module-3-traceback"><i class="fa fa-check"></i><b>51.3.3</b> Module 3: Traceback</a></li>
</ul></li>
<li class="chapter" data-level="51.4" data-path="blast-summary.html"><a href="blast-summary.html#blast-alignment-statistics"><i class="fa fa-check"></i><b>51.4</b> BLAST alignment statistics</a></li>
</ul></li>
<li class="chapter" data-level="52" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html"><i class="fa fa-check"></i><b>52</b> BLAST Alignment score distribution</a>
<ul>
<li class="chapter" data-level="52.1" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#packages-9"><i class="fa fa-check"></i><b>52.1</b> Packages</a></li>
<li class="chapter" data-level="52.2" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#distributions-of-alignment-scores"><i class="fa fa-check"></i><b>52.2</b> Distributions of alignment scores</a></li>
<li class="chapter" data-level="52.3" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-data"><i class="fa fa-check"></i><b>52.3</b> Simulating data</a></li>
<li class="chapter" data-level="52.4" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#lots-of-data-in-biology-is-normal"><i class="fa fa-check"></i><b>52.4</b> Lots of data in biology is normal</a></li>
<li class="chapter" data-level="52.5" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-the-normal-distribution"><i class="fa fa-check"></i><b>52.5</b> Simulating the normal distribution</a></li>
<li class="chapter" data-level="52.6" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#the-extreme-value-distribution"><i class="fa fa-check"></i><b>52.6</b> The extreme value distribution</a></li>
<li class="chapter" data-level="52.7" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#replicating-figure-4.14"><i class="fa fa-check"></i><b>52.7</b> Replicating Figure 4.14</a></li>
<li class="chapter" data-level="52.8" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#why-do-we-care"><i class="fa fa-check"></i><b>52.8</b> Why do we care?</a></li>
</ul></li>
<li class="chapter" data-level="53" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><i class="fa fa-check"></i><b>53</b> Calculating E values, Bit scores, and P-values from BLAST output</a>
<ul>
<li class="chapter" data-level="53.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-7"><i class="fa fa-check"></i><b>53.1</b> Introduction</a></li>
<li class="chapter" data-level="53.2" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#case-study"><i class="fa fa-check"></i><b>53.2</b> Case Study</a></li>
<li class="chapter" data-level="53.3" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-statistics"><i class="fa fa-check"></i><b>53.3</b> BLAST Statistics</a></li>
<li class="chapter" data-level="53.4" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot"><i class="fa fa-check"></i><b>53.4</b> Dotplot</a>
<ul>
<li class="chapter" data-level="53.4.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot-in-r---optional-experimental-code"><i class="fa fa-check"></i><b>53.4.1</b> Dotplot in R - optional, experimental code</a></li>
</ul></li>
<li class="chapter" data-level="53.5" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-search-summary"><i class="fa fa-check"></i><b>53.5</b> BLAST “Search Summary”</a></li>
<li class="chapter" data-level="53.6" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-alignments"><i class="fa fa-check"></i><b>53.6</b> BLAST Alignments</a></li>
<li class="chapter" data-level="53.7" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#equations"><i class="fa fa-check"></i><b>53.7</b> Equations</a>
<ul>
<li class="chapter" data-level="53.7.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-value-eq-4.5-pg-142"><i class="fa fa-check"></i><b>53.7.1</b> E value (eq 4.5, pg 142)</a></li>
</ul></li>
<li class="chapter" data-level="53.8" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bit-scores-s"><i class="fa fa-check"></i><b>53.8</b> Bit Scores (S’)</a></li>
<li class="chapter" data-level="53.9" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-from-bit-scores"><i class="fa fa-check"></i><b>53.9</b> E from Bit scores</a></li>
<li class="chapter" data-level="53.10" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#p-values"><i class="fa fa-check"></i><b>53.10</b> P values</a></li>
<li class="chapter" data-level="53.11" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#repliminaries"><i class="fa fa-check"></i><b>53.11</b> Repliminaries</a></li>
<li class="chapter" data-level="53.12" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#concepts"><i class="fa fa-check"></i><b>53.12</b> Concepts</a></li>
<li class="chapter" data-level="53.13" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#functions-6"><i class="fa fa-check"></i><b>53.13</b> Functions</a></li>
<li class="chapter" data-level="53.14" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-8"><i class="fa fa-check"></i><b>53.14</b> Introduction</a></li>
<li class="chapter" data-level="53.15" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees"><i class="fa fa-check"></i><b>53.15</b> Number of rooted trees</a></li>
<li class="chapter" data-level="53.16" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees-in-r"><i class="fa fa-check"></i><b>53.16</b> Number of rooted trees in R</a>
<ul>
<li class="chapter" data-level="53.16.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#unique-topologies-versus-rotations"><i class="fa fa-check"></i><b>53.16.1</b> Unique topologies versus rotations</a></li>
</ul></li>
<li class="chapter" data-level="53.17" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bilography"><i class="fa fa-check"></i><b>53.17</b> Bilography</a></li>
</ul></li>
<li class="chapter" data-level="54" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html"><i class="fa fa-check"></i><b>54</b> UPGMA the hard way</a>
<ul>
<li class="chapter" data-level="54.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#introudction-to-upgma"><i class="fa fa-check"></i><b>54.1</b> Introudction to UPGMA</a></li>
<li class="chapter" data-level="54.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#preliminaries-18"><i class="fa fa-check"></i><b>54.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="54.2.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#r-libraries"><i class="fa fa-check"></i><b>54.2.1</b> R Libraries</a></li>
<li class="chapter" data-level="54.2.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#custom-function-well-use"><i class="fa fa-check"></i><b>54.2.2</b> Custom function we’ll use</a></li>
</ul></li>
<li class="chapter" data-level="54.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#data-matrix"><i class="fa fa-check"></i><b>54.3</b> Data matrix</a>
<ul>
<li class="chapter" data-level="54.3.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#building-the-matrix"><i class="fa fa-check"></i><b>54.3.1</b> Building the matrix</a></li>
</ul></li>
<li class="chapter" data-level="54.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#what-does-this-distance-matrix-represent"><i class="fa fa-check"></i><b>54.4</b> What does this distance matrix represent?</a>
<ul>
<li class="chapter" data-level="54.4.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-pairwise-linear-distances"><i class="fa fa-check"></i><b>54.4.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="54.4.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-2d-map-of-linear-distances"><i class="fa fa-check"></i><b>54.4.2</b> Genetic distances conceptualized as 2D map of linear distances</a></li>
</ul></li>
<li class="chapter" data-level="54.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#from-linear-distances-to-tree-distances"><i class="fa fa-check"></i><b>54.5</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="54.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#the-upgma-algorithm"><i class="fa fa-check"></i><b>54.6</b> The UPGMA algorithm</a>
<ul>
<li class="chapter" data-level="54.6.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-1-find-the-2-taxa-that-are-currently-closest-together"><i class="fa fa-check"></i><b>54.6.1</b> Step 1: find the 2 taxa that are currently closest together</a></li>
<li class="chapter" data-level="54.6.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node"><i class="fa fa-check"></i><b>54.6.2</b> Step 2: calculate the branch lengths connect the taxa to their shared node</a></li>
<li class="chapter" data-level="54.6.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-determine-if-the-algorithm-is-done"><i class="fa fa-check"></i><b>54.6.3</b> Step 3a: determine if the algorithm is done</a></li>
<li class="chapter" data-level="54.6.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u"><i class="fa fa-check"></i><b>54.6.4</b> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</a></li>
<li class="chapter" data-level="54.6.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-4-determine-the-distance-from-u-to-the-other-remaining-clusters."><i class="fa fa-check"></i><b>54.6.5</b> Step 4: Determine the distance from u to the other remaining clusters.</a></li>
<li class="chapter" data-level="54.6.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-5-create-a-new-matrix"><i class="fa fa-check"></i><b>54.6.6</b> Step 5: Create a new matrix</a></li>
<li class="chapter" data-level="54.6.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa"><i class="fa fa-check"></i><b>54.6.7</b> Key step: calculating the distance from a clade to the remaining taxa</a></li>
</ul></li>
<li class="chapter" data-level="54.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-2d-representation"><i class="fa fa-check"></i><b>54.7</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="54.8" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---round-1"><i class="fa fa-check"></i><b>54.8</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="54.8.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#determine-minimum-value-in-current-matrix"><i class="fa fa-check"></i><b>54.8.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="54.8.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade"><i class="fa fa-check"></i><b>54.8.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.8.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length"><i class="fa fa-check"></i><b>54.8.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="54.8.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>54.8.4</b> Calculate distance from clade to all other points</a></li>
<li class="chapter" data-level="54.8.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-matrix-with-new-distance"><i class="fa fa-check"></i><b>54.8.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="54.9" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---next-round"><i class="fa fa-check"></i><b>54.9</b> Algorithm - next round</a>
<ul>
<li class="chapter" data-level="54.9.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-1"><i class="fa fa-check"></i><b>54.9.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.9.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-1"><i class="fa fa-check"></i><b>54.9.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="54.9.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#recalculate-distance-from-abe-to-all-remainign-taxa"><i class="fa fa-check"></i><b>54.9.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="54.9.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-the-matrix"><i class="fa fa-check"></i><b>54.9.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="54.10" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#next-iteration"><i class="fa fa-check"></i><b>54.10</b> Next iteration</a>
<ul>
<li class="chapter" data-level="54.10.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-lengths"><i class="fa fa-check"></i><b>54.10.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="54.10.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-2"><i class="fa fa-check"></i><b>54.10.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.10.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-2"><i class="fa fa-check"></i><b>54.10.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="54.10.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#distance-of-c-to-abe-and-d-to-abe"><i class="fa fa-check"></i><b>54.10.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="54.11" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finish-up"><i class="fa fa-check"></i><b>54.11</b> Finish up</a>
<ul>
<li class="chapter" data-level="54.11.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-3"><i class="fa fa-check"></i><b>54.11.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.11.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-3"><i class="fa fa-check"></i><b>54.11.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="54.12" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finalizing-branch-lengths"><i class="fa fa-check"></i><b>54.12</b> Finalizing branch lengths</a></li>
</ul></li>
<li class="chapter" data-level="55" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html"><i class="fa fa-check"></i><b>55</b> Representing phylogenetic trees in Newick format</a>
<ul>
<li class="chapter" data-level="55.1" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#vocab-7"><i class="fa fa-check"></i><b>55.1</b> Vocab</a></li>
<li class="chapter" data-level="55.2" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#introduction-9"><i class="fa fa-check"></i><b>55.2</b> Introduction</a></li>
<li class="chapter" data-level="55.3" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-three-taxa"><i class="fa fa-check"></i><b>55.3</b> Trees with three taxa</a></li>
<li class="chapter" data-level="55.4" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-four-taxa"><i class="fa fa-check"></i><b>55.4</b> Trees with four taxa</a></li>
</ul></li>
<li class="chapter" data-level="56" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html"><i class="fa fa-check"></i><b>56</b> Plotting trees from Newick notation in R</a>
<ul>
<li class="chapter" data-level="56.1" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#preliminaries-19"><i class="fa fa-check"></i><b>56.1</b> Preliminaries</a></li>
<li class="chapter" data-level="56.2" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#note-1"><i class="fa fa-check"></i><b>56.2</b> Note</a></li>
<li class="chapter" data-level="56.3" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick"><i class="fa fa-check"></i><b>56.3</b> 3 taxa tree in Newick</a></li>
<li class="chapter" data-level="56.4" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick-1"><i class="fa fa-check"></i><b>56.4</b> 4 taxa tree in Newick</a></li>
<li class="chapter" data-level="56.5" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-in-newick-format"><i class="fa fa-check"></i><b>56.5</b> 5 taxa in Newick Format</a></li>
<li class="chapter" data-level="56.6" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#many-taxa-tree"><i class="fa fa-check"></i><b>56.6</b> Many-taxa tree</a></li>
<li class="chapter" data-level="56.7" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#newick-and-branch-lengths"><i class="fa fa-check"></i><b>56.7</b> Newick and Branch lengths</a></li>
<li class="chapter" data-level="56.8" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#additional-reading"><i class="fa fa-check"></i><b>56.8</b> Additional Reading</a></li>
</ul></li>
<li class="chapter" data-level="57" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><i class="fa fa-check"></i><b>57</b> Phylogenetic trees using the UPGMA clustering algoirthm</a>
<ul>
<li class="chapter" data-level="57.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#introduction-10"><i class="fa fa-check"></i><b>57.1</b> Introduction</a></li>
<li class="chapter" data-level="57.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#preliminaries-20"><i class="fa fa-check"></i><b>57.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="57.2.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#vocab-8"><i class="fa fa-check"></i><b>57.2.1</b> Vocab</a></li>
<li class="chapter" data-level="57.2.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-r-functions"><i class="fa fa-check"></i><b>57.2.2</b> General R functions</a></li>
<li class="chapter" data-level="57.2.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#specific-functions"><i class="fa fa-check"></i><b>57.2.3</b> Specific functions</a></li>
</ul></li>
<li class="chapter" data-level="57.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#sequence-comparisons"><i class="fa fa-check"></i><b>57.3</b> Sequence comparisons</a></li>
<li class="chapter" data-level="57.4" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrices"><i class="fa fa-check"></i><b>57.4</b> Distance matrices</a></li>
<li class="chapter" data-level="57.5" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#from-a-distance-matrix-to-phylogenetic-tree"><i class="fa fa-check"></i><b>57.5</b> From a distance matrix to phylogenetic tree</a></li>
<li class="chapter" data-level="57.6" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-raw-sequences"><i class="fa fa-check"></i><b>57.6</b> Distance matrix in R from raw sequences</a></li>
<li class="chapter" data-level="57.7" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-normal-matrix"><i class="fa fa-check"></i><b>57.7</b> Distance matrix in R from normal matrix</a></li>
<li class="chapter" data-level="57.8" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#upgma-in-r"><i class="fa fa-check"></i><b>57.8</b> UPGMA in R</a></li>
<li class="chapter" data-level="57.9" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#how-upgma-clustering-works"><i class="fa fa-check"></i><b>57.9</b> How UPGMA clustering works</a>
<ul>
<li class="chapter" data-level="57.9.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#challenge"><i class="fa fa-check"></i><b>57.9.1</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="57.10" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#further-information"><i class="fa fa-check"></i><b>57.10</b> Further information</a>
<ul>
<li class="chapter" data-level="57.10.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-info"><i class="fa fa-check"></i><b>57.10.1</b> General info</a></li>
<li class="chapter" data-level="57.10.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#worked-examples"><i class="fa fa-check"></i><b>57.10.2</b> Worked examples:</a></li>
<li class="chapter" data-level="57.10.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#additional-notes"><i class="fa fa-check"></i><b>57.10.3</b> Additional notes:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="58" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><i class="fa fa-check"></i><b>58</b> Tutorial: Writing functions to calculate the number of phylogenetic trees</a>
<ul>
<li class="chapter" data-level="58.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#repliminaries-1"><i class="fa fa-check"></i><b>58.1</b> Repliminaries</a>
<ul>
<li class="chapter" data-level="58.1.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-7"><i class="fa fa-check"></i><b>58.1.1</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="58.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#concepts-1"><i class="fa fa-check"></i><b>58.2</b> Concepts</a>
<ul>
<li class="chapter" data-level="58.2.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vocab-9"><i class="fa fa-check"></i><b>58.2.1</b> Vocab:</a></li>
<li class="chapter" data-level="58.2.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#packages-10"><i class="fa fa-check"></i><b>58.2.2</b> Packages:</a></li>
</ul></li>
<li class="chapter" data-level="58.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#introduction-11"><i class="fa fa-check"></i><b>58.3</b> Introduction</a></li>
<li class="chapter" data-level="58.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-1"><i class="fa fa-check"></i><b>58.4</b> Number of rooted trees</a></li>
<li class="chapter" data-level="58.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-in-r-1"><i class="fa fa-check"></i><b>58.5</b> Number of rooted trees in R</a></li>
<li class="chapter" data-level="58.6" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-in-r-1"><i class="fa fa-check"></i><b>58.6</b> Functions in R</a></li>
<li class="chapter" data-level="58.7" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-to-calculate-the-number-of-possible-phylogenetic-trees"><i class="fa fa-check"></i><b>58.7</b> Function to calculate the number of possible phylogenetic trees</a></li>
<li class="chapter" data-level="58.8" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vectorized-inputs-to-functions"><i class="fa fa-check"></i><b>58.8</b> Vectorized inputs to functions</a></li>
<li class="chapter" data-level="58.9" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-conditional-statements"><i class="fa fa-check"></i><b>58.9</b> Adding conditional statements</a></li>
<li class="chapter" data-level="58.10" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-multiple-conditional-statements"><i class="fa fa-check"></i><b>58.10</b> Adding multiple conditional statements</a></li>
<li class="chapter" data-level="58.11" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-additional-arguments"><i class="fa fa-check"></i><b>58.11</b> Adding additional arguments</a></li>
<li class="chapter" data-level="58.12" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-_un_rooted-trees"><i class="fa fa-check"></i><b>58.12</b> Assignment: _un_rooted trees</a></li>
<li class="chapter" data-level="58.13" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-1"><i class="fa fa-check"></i><b>58.13</b> Assignment part 1</a>
<ul>
<li class="chapter" data-level="58.13.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#part-1-answer"><i class="fa fa-check"></i><b>58.13.1</b> Part 1 answer</a></li>
</ul></li>
<li class="chapter" data-level="58.14" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-2"><i class="fa fa-check"></i><b>58.14</b> Assignment part 2</a>
<ul>
<li class="chapter" data-level="58.14.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#answer"><i class="fa fa-check"></i><b>58.14.1</b> Answer</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="59" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html"><i class="fa fa-check"></i><b>59</b> Writing functions - practice problems</a>
<ul>
<li class="chapter" data-level="59.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#introduction-12"><i class="fa fa-check"></i><b>59.1</b> Introduction</a></li>
<li class="chapter" data-level="59.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#part-1-pure-practice"><i class="fa fa-check"></i><b>59.2</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="59.2.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-1-hello-world"><i class="fa fa-check"></i><b>59.2.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="59.2.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______"><i class="fa fa-check"></i><b>59.2.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="59.3" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function"><i class="fa fa-check"></i><b>59.3</b> Practice Function 3: Make you own natural log function</a></li>
<li class="chapter" data-level="59.4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-4-make-function-that-converts-dna-directly-to-rna"><i class="fa fa-check"></i><b>59.4</b> Practice Function 4: Make function that converts DNA directly to RNA</a></li>
</ul></li>
<li class="chapter" data-level="60" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html"><i class="fa fa-check"></i><b>60</b> Writing user-friendly functions</a>
<ul>
<li class="chapter" data-level="60.1" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-1"><i class="fa fa-check"></i><b>60.1</b> Version 1</a></li>
<li class="chapter" data-level="60.2" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-2"><i class="fa fa-check"></i><b>60.2</b> Version 2</a></li>
<li class="chapter" data-level="60.3" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-3"><i class="fa fa-check"></i><b>60.3</b> Version 3</a></li>
<li class="chapter" data-level="60.4" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-4-adding-conditions-and-warnigns"><i class="fa fa-check"></i><b>60.4</b> Version 4: Adding conditions and warnigns</a></li>
<li class="chapter" data-level="60.5" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-5"><i class="fa fa-check"></i><b>60.5</b> Version 5</a></li>
</ul></li>
<li class="chapter" data-level="61" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html"><i class="fa fa-check"></i><b>61</b> Function practice exercises</a>
<ul>
<li class="chapter" data-level="61.1" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#part-1-pure-practice-1"><i class="fa fa-check"></i><b>61.1</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="61.1.1" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#practice-function-1-hello-world-1"><i class="fa fa-check"></i><b>61.1.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="61.1.2" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#practice-function-2-hello-_______-1"><i class="fa fa-check"></i><b>61.1.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="61.2" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#practice-function-3-make-you-own-natural-log-function-1"><i class="fa fa-check"></i><b>61.2</b> Practice Function 3: Make you own natural log function</a></li>
</ul></li>
<li class="chapter" data-level="62" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html"><i class="fa fa-check"></i><b>62</b> for() loops in R</a>
<ul>
<li class="chapter" data-level="62.1" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#introduction-13"><i class="fa fa-check"></i><b>62.1</b> Introduction</a></li>
<li class="chapter" data-level="62.2" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#reminders"><i class="fa fa-check"></i><b>62.2</b> Reminders</a></li>
<li class="chapter" data-level="62.3" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop"><i class="fa fa-check"></i><b>62.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="62.4" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-classic-for-loop"><i class="fa fa-check"></i><b>62.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="62.5" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-power-user-for-loop"><i class="fa fa-check"></i><b>62.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="62.6" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-4-a-for-loop-for-all-purposes"><i class="fa fa-check"></i><b>62.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="62.7" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#advanced-aside-if-else-and-next-statements"><i class="fa fa-check"></i><b>62.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="62.8" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#on-avoiding-for-loops"><i class="fa fa-check"></i><b>62.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="62.9" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#challenge-1"><i class="fa fa-check"></i><b>62.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="63" data-path="random-number-generation.html"><a href="random-number-generation.html"><i class="fa fa-check"></i><b>63</b> Random number generation</a>
<ul>
<li class="chapter" data-level="63.1" data-path="random-number-generation.html"><a href="random-number-generation.html#introduction-14"><i class="fa fa-check"></i><b>63.1</b> Introduction</a></li>
<li class="chapter" data-level="63.2" data-path="random-number-generation.html"><a href="random-number-generation.html#using-set.seed-to-make-simulations-reproducible"><i class="fa fa-check"></i><b>63.2</b> Using set.seed() to make simulations reproducible</a></li>
<li class="chapter" data-level="63.3" data-path="random-number-generation.html"><a href="random-number-generation.html#types-of-random-number-generators"><i class="fa fa-check"></i><b>63.3</b> Types of random number generators</a></li>
<li class="chapter" data-level="63.4" data-path="random-number-generation.html"><a href="random-number-generation.html#reporting-simulations"><i class="fa fa-check"></i><b>63.4</b> Reporting simulations</a></li>
<li class="chapter" data-level="63.5" data-path="random-number-generation.html"><a href="random-number-generation.html#notes"><i class="fa fa-check"></i><b>63.5</b> Notes:</a></li>
</ul></li>
<li class="chapter" data-level="64" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html"><i class="fa fa-check"></i><b>64</b> Randomization practice question</a>
<ul>
<li class="chapter" data-level="64.1" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#practice-question"><i class="fa fa-check"></i><b>64.1</b> Practice question:</a></li>
<li class="chapter" data-level="64.2" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#your-answer"><i class="fa fa-check"></i><b>64.2</b> Your answer:</a></li>
<li class="chapter" data-level="64.3" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-0"><i class="fa fa-check"></i><b>64.3</b> Hint 0</a></li>
<li class="chapter" data-level="64.4" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-1"><i class="fa fa-check"></i><b>64.4</b> Hint 1:</a></li>
<li class="chapter" data-level="64.5" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-2-follow-up-to-hint-0"><i class="fa fa-check"></i><b>64.5</b> Hint 2 (follow up to Hint 0)</a></li>
<li class="chapter" data-level="64.6" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-3"><i class="fa fa-check"></i><b>64.6</b> Hint 3:</a></li>
<li class="chapter" data-level="64.7" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#answer-1"><i class="fa fa-check"></i><b>64.7</b> Answer</a></li>
</ul></li>
<li class="chapter" data-level="65" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html"><i class="fa fa-check"></i><b>65</b> Simulating random amino acid sequences</a>
<ul>
<li class="chapter" data-level="65.1" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#selecting-random-letters-from-a-vector"><i class="fa fa-check"></i><b>65.1</b> Selecting random letters from a vector</a></li>
<li class="chapter" data-level="65.2" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-letters-from-the-whole-alphabet"><i class="fa fa-check"></i><b>65.2</b> Select random letters from the whole alphabet</a></li>
<li class="chapter" data-level="65.3" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-amino-acids-to-build-polypeptide"><i class="fa fa-check"></i><b>65.3</b> Select random amino acids to build polypeptide</a></li>
</ul></li>
<li class="chapter" data-level="66" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>66</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="66.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>66.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="66.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>66.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="66.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>66.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="66.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>66.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="66.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>66.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-3"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="66.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>66.2</b> Help!</a>
<ul>
<li class="chapter" data-level="66.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>66.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="66.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>66.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="66.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>66.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="66.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>66.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="66.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>66.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="66.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>66.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="66.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>66.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="66.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>66.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="local-variation-in-gc-content" class="section level1" number="44">
<h1><span class="header-section-number">Chapter 44</span> Local variation in GC content</h1>
<p><strong>By</strong>: Avril Coghlan.</p>
<p><strong>Adapted, edited and expanded</strong>: Nathan Brouwer under the Creative Commons 3.0 Attribution License <a href="https://creativecommons.org/licenses/by/3.0/">(CC BY 3.0)</a>.</p>
<p>This is a modification of <a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter1.html">“DNA Sequence Statistics (1)”</a> from Avril Coghlan’s <a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/index.html"><em>A little book of R for bioinformatics.</em></a>. Almost all of text and code was originally written by Dr. Coghlan and distributed under the <a href="https://creativecommons.org/licenses/by/3.0/us/">Creative Commons 3.0</a> license.</p>
<p>Preliminaries</p>
<div class="sourceCode" id="cb724"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb724-1"><a href="local-variation-in-gc-content.html#cb724-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rentrez)</span>
<span id="cb724-2"><a href="local-variation-in-gc-content.html#cb724-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(seqinr)</span>
<span id="cb724-3"><a href="local-variation-in-gc-content.html#cb724-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compbio4all)</span></code></pre></div>
<div id="note-on-the-biology-in-this-section" class="section level3" number="44.0.1">
<h3><span class="header-section-number">44.0.1</span> Note on the biology in this section</h3>
<p>Some of the biology in this tutorial appears to be out of date. For example, using variation in GC content to ID horizontal gene transfer is currently considered to be biased. The examples are still good for practicing <em>R</em> skills.</p>
</div>
<div id="vocabulary-2" class="section level2" number="44.1">
<h2><span class="header-section-number">44.1</span> Vocabulary</h2>
<ul>
<li><a href="https://en.wikipedia.org/wiki/GC-content"><strong>GC content</strong></a></li>
<li><a href="en.wikipedia.org/wiki/Horizontal_gene_transfer"><strong>horizontal transfer</strong></a></li>
<li><strong>local variation</strong> in GC content …</li>
</ul>
</div>
<div id="reading-sequence-data-with-rentrezentrez_fetch" class="section level2" number="44.2">
<h2><span class="header-section-number">44.2</span> Reading sequence data with rentrez::entrez_fetch</h2>
<p>In a previous section you learned how to use to search for and download the sequence data for a given NCBI accession from the NCBI Sequence Database, either via the NCBI website using <code>entrez_fetch()</code> from the rentrez package.</p>
<p>For example, you could have downloaded the sequence data for a the DEN-1 Dengue virus sequence (NCBI accession NC_001477), and stored it on a file on your computer (eg. <code>dengue_fasta.fasta</code>).</p>
<p>Download direct:</p>
<div class="sourceCode" id="cb725"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb725-1"><a href="local-variation-in-gc-content.html#cb725-1" aria-hidden="true" tabindex="-1"></a>dengueseq_fasta <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;nucleotide&quot;</span>, </span>
<span id="cb725-2"><a href="local-variation-in-gc-content.html#cb725-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;NC_001477&quot;</span>, </span>
<span id="cb725-3"><a href="local-variation-in-gc-content.html#cb725-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>From <code>combio4all</code> package</p>
<div class="sourceCode" id="cb726"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb726-1"><a href="local-variation-in-gc-content.html#cb726-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dengueseq_fasta)</span></code></pre></div>
<p>As noted before, the file gets downloaded in FASTA format, which isn’t directly useable in <em>R</em></p>
<pre><code>## [1] &quot;&gt;NC_001477.1 Dengue virus 1, complete genome\nAGTTGTTAGTCTACGTGGACCGACAAGAACAGTTTCGAATCGGAAGCTTGCTTAACGTAGTTCTAACAGT\nTTTTTATTAGAGAGCAGATCTCTGATGAACAACCAACGGAAAAAGACGGGTCGACCGTCTTTCAATATGC\nTGAAACGCGCGAGAAACCGCGTGTCAACTGTTTCACAGTTGGCGAAGAGATTCTCAAAAGGAT&quot;</code></pre>
<p>We can convert our FASTA object into a vector using the function fasta_cleaner()</p>
<div class="sourceCode" id="cb728"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb728-1"><a href="local-variation-in-gc-content.html#cb728-1" aria-hidden="true" tabindex="-1"></a>header. <span class="ot">&lt;-</span> <span class="st">&quot;&gt;NC_001477.1 Dengue virus 1, complete genome&quot;</span></span>
<span id="cb728-2"><a href="local-variation-in-gc-content.html#cb728-2" aria-hidden="true" tabindex="-1"></a>dengueseq_vector <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(dengueseq_fasta)</span></code></pre></div>
<p>Once you have retrieved a sequence from the NCBI Sequence Database and stored it in a vector variable such as dengueseq_vector in the example above, it is possible to extract <strong>subsequence</strong> of the sequence by typing the name of the vector (eg. dengueseq_vector) followed by the square brackets containing the indices for those nucleotides. For example, to obtain nucleotides 452-535 of the DEN-1 Dengue virus genome, we can type:</p>
<div class="sourceCode" id="cb729"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb729-1"><a href="local-variation-in-gc-content.html#cb729-1" aria-hidden="true" tabindex="-1"></a>dengueseq_vector[<span class="dv">452</span><span class="sc">:</span><span class="dv">535</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;C&quot; &quot;G&quot; &quot;A&quot; &quot;G&quot; &quot;G&quot; &quot;G&quot; &quot;G&quot; &quot;G&quot; &quot;A&quot; &quot;G&quot; &quot;A&quot; &quot;G&quot; &quot;C&quot; &quot;C&quot; &quot;G&quot; &quot;C&quot; &quot;A&quot; &quot;C&quot; &quot;A&quot;
## [20] &quot;T&quot; &quot;G&quot; &quot;A&quot; &quot;T&quot; &quot;A&quot; &quot;G&quot; &quot;T&quot; &quot;T&quot; &quot;A&quot; &quot;G&quot; &quot;C&quot; &quot;A&quot; &quot;A&quot; &quot;G&quot; &quot;C&quot; &quot;A&quot; &quot;G&quot; &quot;G&quot; &quot;A&quot;
## [39] &quot;A&quot; &quot;A&quot; &quot;G&quot; &quot;A&quot; &quot;G&quot; &quot;G&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;T&quot; &quot;C&quot; &quot;A&quot; &quot;C&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; &quot;G&quot;
## [58] &quot;T&quot; &quot;T&quot; &quot;T&quot; &quot;A&quot; &quot;A&quot; &quot;G&quot; &quot;A&quot; &quot;C&quot; &quot;C&quot; &quot;T&quot; &quot;C&quot; &quot;T&quot; &quot;G&quot; &quot;C&quot; &quot;A&quot; &quot;G&quot; &quot;G&quot; &quot;T&quot; &quot;G&quot;
## [77] &quot;T&quot; &quot;C&quot; &quot;A&quot; &quot;A&quot; &quot;C&quot; &quot;A&quot; &quot;T&quot; &quot;G&quot;</code></pre>
</div>
<div id="local-variation-in-gc-content-1" class="section level2" number="44.3">
<h2><span class="header-section-number">44.3</span> Local variation in GC content</h2>
<p>In a previous section, you learned that to find out the <a href="https://en.wikipedia.org/wiki/GC-content"><strong>GC content</strong></a> of a genome sequence (percentage of nucleotides in a genome sequence that are Gs or Cs), you can use the GC() function in the seqinr package. For example, to find the GC content of the DEN-1 Dengue virus sequence that we have stored in the vector dengueseq_vector, we can type:</p>
<div class="sourceCode" id="cb731"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb731-1"><a href="local-variation-in-gc-content.html#cb731-1" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">GC</span>(dengueseq_vector)</span></code></pre></div>
<pre><code>## [1] 0.4666977</code></pre>
<p>The output of the GC() is the fraction of nucleotides in a sequence that are Gs or Cs, so to convert it to a percentage we need to multiply by 100.</p>
<div class="sourceCode" id="cb733"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb733-1"><a href="local-variation-in-gc-content.html#cb733-1" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">GC</span>(dengueseq_vector)<span class="sc">*</span><span class="dv">100</span></span></code></pre></div>
<pre><code>## [1] 46.66977</code></pre>
<p>Thus, the GC content of the DEN-1 Dengue virus genome is about 0.467 or 46.7%.</p>
<p>Although the GC content of the whole DEN-1 Dengue virus genome sequence is about 46.7%, there is probably <strong>local variation</strong> in GC content within the genome. That is, some regions of the genome sequence may have GC contents quite a bit higher than 46.7%, while some regions of the genome sequence may have GC contents that are quite a big lower than 46.7%. Local fluctuations in GC content within the genome sequence can provide different interesting information, for example, they may reveal cases of <a href="en.wikipedia.org/wiki/Horizontal_gene_transfer"><strong>horizontal transfer</strong></a> or reveal biases in mutation.</p>
<p>If a chunk of DNA has moved by horizontal transfer from the genome of a species with low GC content to a species with high GC content, the chunk of horizontally transferred DNA could be detected as a region of unusually low GC content in the high-GC recipient genome.</p>
<p>On the other hand, a region unusually low GC content in an otherwise high-GC content genome could also arise due to biases in mutation in that region of the genome, for example, if mutations from Gs/Cs to Ts/As are more common for some reason in that region of the genome than in the rest of the genome.</p>
</div>
<div id="a-sliding-window-analysis-of-gc-content" class="section level2" number="44.4">
<h2><span class="header-section-number">44.4</span> A sliding window analysis of GC content</h2>
<p>In order to study local variation in GC content within a genome sequence, we could calculate the GC content for small chunks of the genome sequence. The DEN-1 Dengue virus genome sequence is 10735 nucleotides long. To study variation in GC content within the genome sequence, we could calculate the GC content of chunks of the DEN-1 Dengue virus genome, for example, for each 2000-nucleotide chunk of the genome sequence:</p>
<div class="sourceCode" id="cb735"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb735-1"><a href="local-variation-in-gc-content.html#cb735-1" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">GC</span>(dengueseq_vector[<span class="dv">1</span><span class="sc">:</span><span class="dv">2000</span>])      <span class="co"># Calculate the GC content of nucleotides 1-2000 of the Dengue genome</span></span></code></pre></div>
<pre><code>## [1] 0.465</code></pre>
<div class="sourceCode" id="cb737"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb737-1"><a href="local-variation-in-gc-content.html#cb737-1" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">GC</span>(dengueseq_vector[<span class="dv">2001</span><span class="sc">:</span><span class="dv">4000</span>])   <span class="co"># Calculate the GC content of nucleotides 2001-4000 of the Dengue genome</span></span></code></pre></div>
<pre><code>## [1] 0.4525</code></pre>
<p>From the output of the above calculations, we see that the region of the DEN-1 Dengue virus genome from nucleotides 1-2000 has a GC content of 46.5%, while the region of the Dengue genome from nucleotides 2001-4000 has a GC content of about 45.3%. Thus, there seems to be some local variation in GC content within the Dengue genome sequence.</p>
<p>Instead of typing in the commands above to tell R to calculate the GC content for each 2000-nucleotide chunk of the DEN-1 Dengue genome, we can use a for loop to carry out the same calculations, but by typing far fewer commands. That is, we can use a for loop to take each 2000-nucleotide chunk of the DEN-1 Dengue virus genome, and to calculate the GC content of each 2000-nucleotide chunk. Below we will explain the following for loop that has been written for this purpose:</p>
<div class="sourceCode" id="cb739"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb739-1"><a href="local-variation-in-gc-content.html#cb739-1" aria-hidden="true" tabindex="-1"></a>starts <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">length</span>(dengueseq_vector)<span class="sc">-</span><span class="dv">2000</span>, <span class="at">by =</span> <span class="dv">2000</span>)</span>
<span id="cb739-2"><a href="local-variation-in-gc-content.html#cb739-2" aria-hidden="true" tabindex="-1"></a>starts</span></code></pre></div>
<pre><code>## [1]    1 2001 4001 6001 8001</code></pre>
<div class="sourceCode" id="cb741"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb741-1"><a href="local-variation-in-gc-content.html#cb741-1" aria-hidden="true" tabindex="-1"></a> n <span class="ot">&lt;-</span> <span class="fu">length</span>(starts)    <span class="co"># Find the length of the vector &quot;starts&quot;</span></span>
<span id="cb741-2"><a href="local-variation-in-gc-content.html#cb741-2" aria-hidden="true" tabindex="-1"></a> <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>n) {</span>
<span id="cb741-3"><a href="local-variation-in-gc-content.html#cb741-3" aria-hidden="true" tabindex="-1"></a>        chunk <span class="ot">&lt;-</span> dengueseq_vector[starts[i]<span class="sc">:</span>(starts[i]<span class="sc">+</span><span class="dv">1999</span>)]</span>
<span id="cb741-4"><a href="local-variation-in-gc-content.html#cb741-4" aria-hidden="true" tabindex="-1"></a>        chunkGC <span class="ot">&lt;-</span> <span class="fu">GC</span>(chunk)</span>
<span id="cb741-5"><a href="local-variation-in-gc-content.html#cb741-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">print</span> (chunkGC)</span>
<span id="cb741-6"><a href="local-variation-in-gc-content.html#cb741-6" aria-hidden="true" tabindex="-1"></a>     }</span></code></pre></div>
<pre><code>## [1] 0.465
## [1] 0.4525
## [1] 0.4705
## [1] 0.479
## [1] 0.4545</code></pre>
<p>The command <code>starts &lt;- seq(1, length(dengueseq_vector)-2000, by = 2000)</code> stores the result of the <code>seq()</code> command in the vector starts, which contains the values 1, 2001, 4001, 6001, and 8001.</p>
<p>We set the variable n to be equal to the number of elements in the vector starts, so it will be 5 here, since the vector starts contains the five elements 1, 2001, 4001, 6001 and 8001. The line “for (i in 1:n)” means that the counter i will take values of 1-5 in subsequent cycles of the for loop. The for loop above is spread over several lines. However, R will not execute the commands within the for loop until you have typed the final “}” at the end of the for loop and pressed “Return”.</p>
<p>Each of the three commands within the for loop are carried out in each cycle of the loop. In the first cycle of the loop, i is 1, the vector variable chunk is used to store the region from nucleotides 1-2000 of the Dengue virus sequence, the GC content of that region is calculated and stored in the variable chunkGC, and the value of chunkGC is printed out.</p>
<p>In the second cycle of the loop, i is 2, the vector variable chunk is used to store the region from nucleotides 2001-4000 of the Dengue virus sequence, the GC content of that region is calculated and stored in the variable <code>chunkGC</code>, and the value of <code>chunkGC</code> is printed out. The loop continues until the value of i is 5. In the fifth cycle through the loop, the value of i is 5, and so the GC content of the region from nucleotides 8001-10000 is printed out.</p>
<p>Note that we stop the loop when we are looking at the region from nucleotides 8001-10000, instead of continuing to another cycle of the loop where the region under examination would be from nucleotides 10001-12000. The reason for this is because the length of the Dengue virus genome sequence is just 10735 nucleotides, so there is not a full 2000-nucleotide region from nucleotide 10001 to the end of the sequence at nucleotide 10735.</p>
<p>The above analysis of local variation in GC content is what is known as a sliding window analysis of GC content. By calculating the GC content in each 2000-nucleotide chunk of the Dengue virus genome, you are effectively sliding a 2000-nucleotide window along the DNA sequence from start to end, and calculating the GC content in each non-overlapping window (chunk of DNA).</p>
<p>Note that this sliding window analysis of GC content is a slightly simplified version of the method usually carried out by bioinformaticians. In this simplified version, we have calculated the GC content in non-overlapping windows along a DNA sequence. However, it is more usual to calculate GC content in overlapping windows along a sequence, although that makes the code slightly more complicated.</p>
</div>
<div id="a-sliding-window-plot-of-gc-content" class="section level2" number="44.5">
<h2><span class="header-section-number">44.5</span> A sliding window plot of GC content</h2>
<p>It is common to use the data generated from a sliding window analysis to create a sliding window plot of GC content. To create a sliding window plot of GC content, you plot the local GC content in each window of the genome, versus the nucleotide position of the start of each window. We can create a sliding window plot of GC content by typing:</p>
<div class="sourceCode" id="cb743"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb743-1"><a href="local-variation-in-gc-content.html#cb743-1" aria-hidden="true" tabindex="-1"></a> starts <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">length</span>(dengueseq_vector)<span class="sc">-</span><span class="dv">2000</span>, <span class="at">by =</span> <span class="dv">2000</span>)</span>
<span id="cb743-2"><a href="local-variation-in-gc-content.html#cb743-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb743-3"><a href="local-variation-in-gc-content.html#cb743-3" aria-hidden="true" tabindex="-1"></a> n <span class="ot">&lt;-</span> <span class="fu">length</span>(starts)    <span class="co"># Find the length of the vector &quot;starts&quot;</span></span>
<span id="cb743-4"><a href="local-variation-in-gc-content.html#cb743-4" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb743-5"><a href="local-variation-in-gc-content.html#cb743-5" aria-hidden="true" tabindex="-1"></a> chunkGCs <span class="ot">&lt;-</span> <span class="fu">numeric</span>(n) <span class="co"># Make a vector of the same </span></span>
<span id="cb743-6"><a href="local-variation-in-gc-content.html#cb743-6" aria-hidden="true" tabindex="-1"></a>                        <span class="co">#length as vector &quot;starts&quot;, </span></span>
<span id="cb743-7"><a href="local-variation-in-gc-content.html#cb743-7" aria-hidden="true" tabindex="-1"></a>                        <span class="co"># but just containing zeroes</span></span>
<span id="cb743-8"><a href="local-variation-in-gc-content.html#cb743-8" aria-hidden="true" tabindex="-1"></a> <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>n) {</span>
<span id="cb743-9"><a href="local-variation-in-gc-content.html#cb743-9" aria-hidden="true" tabindex="-1"></a>        chunk <span class="ot">&lt;-</span> dengueseq_vector[starts[i]<span class="sc">:</span>(starts[i]<span class="sc">+</span><span class="dv">1999</span>)]</span>
<span id="cb743-10"><a href="local-variation-in-gc-content.html#cb743-10" aria-hidden="true" tabindex="-1"></a>        chunkGC <span class="ot">&lt;-</span> <span class="fu">GC</span>(chunk)</span>
<span id="cb743-11"><a href="local-variation-in-gc-content.html#cb743-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">print</span>(chunkGC)</span>
<span id="cb743-12"><a href="local-variation-in-gc-content.html#cb743-12" aria-hidden="true" tabindex="-1"></a>        chunkGCs[i] <span class="ot">&lt;-</span> chunkGC</span>
<span id="cb743-13"><a href="local-variation-in-gc-content.html#cb743-13" aria-hidden="true" tabindex="-1"></a>     }</span></code></pre></div>
<pre><code>## [1] 0.465
## [1] 0.4525
## [1] 0.4705
## [1] 0.479
## [1] 0.4545</code></pre>
<div class="sourceCode" id="cb745"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb745-1"><a href="local-variation-in-gc-content.html#cb745-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">plot</span>(starts, </span>
<span id="cb745-2"><a href="local-variation-in-gc-content.html#cb745-2" aria-hidden="true" tabindex="-1"></a>      chunkGCs,</span>
<span id="cb745-3"><a href="local-variation-in-gc-content.html#cb745-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">type=</span><span class="st">&quot;b&quot;</span>, </span>
<span id="cb745-4"><a href="local-variation-in-gc-content.html#cb745-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">xlab=</span><span class="st">&quot;Nucleotide start position&quot;</span>,</span>
<span id="cb745-5"><a href="local-variation-in-gc-content.html#cb745-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">ylab=</span><span class="st">&quot;GC content&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-523-1.png" width="672" /></p>
<p>In the code above, the line <code>chunkGCs &lt;- numeric(n)</code> makes a new vector <code>chunkGCs</code> which has the same number of elements as the vector starts (5 elements here). This vector chunkGCs is then used within the for loop for storing the GC content of each chunk of DNA.</p>
<p>After the loop, the vector starts can be plotted against the vector chunkGCs using the <code>plot()</code> function, to get a plot of GC content against nucleotide position in the genome sequence. This is a sliding window plot of GC content.</p>
<p>You may want to use the code above to create sliding window plots of GC content of different species’ genomes, using different window sizes. Therefore, it makes sense to use a function to do the sliding window plot, that can take the windo wsize that the user wants to use and the sequence that the user wants to study as arguments (inputs).</p>
<p>The compbio4all has a function <code>plot_sliding_window()</code> that does this.</p>
<p>This function will make a sliding window plot of GC content for a particular input sequence inputseq specified by the user, using a particular window size window size specified by the user. Once you have typed in this function once, you can use it again and again to make sliding window plots of GC contents for different input DNA sequences, with different window sizes. For example, you could create two different sliding window plots of the DEN-1 Dengue virus genome sequence, using window sizes of 3000 and 300 nucleotides, respectively:</p>
<div class="sourceCode" id="cb746"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb746-1"><a href="local-variation-in-gc-content.html#cb746-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">plot_sliding_window</span>(<span class="dv">3000</span>, dengueseq_vector)</span></code></pre></div>
<pre><code>## [1] 0.4646667
## [1] 0.4606667
## [1] 0.4653333</code></pre>
<p><img src="lbrb_files/figure-html/unnamed-chunk-524-1.png" width="672" /></p>
<div class="sourceCode" id="cb748"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb748-1"><a href="local-variation-in-gc-content.html#cb748-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">plot_sliding_window</span>(<span class="dv">300</span>, dengueseq_vector)</span></code></pre></div>
<pre><code>## [1] 0.4366667
## [1] 0.4766667
## [1] 0.5
## [1] 0.5066667
## [1] 0.4466667
## [1] 0.4333333
## [1] 0.4466667
## [1] 0.4866667
## [1] 0.4633333
## [1] 0.45
## [1] 0.44
## [1] 0.4633333
## [1] 0.4433333
## [1] 0.4
## [1] 0.4766667
## [1] 0.49
## [1] 0.4733333
## [1] 0.4666667
## [1] 0.4666667
## [1] 0.4866667
## [1] 0.4966667
## [1] 0.48
## [1] 0.49
## [1] 0.49
## [1] 0.4633333
## [1] 0.46
## [1] 0.4633333
## [1] 0.42
## [1] 0.4466667
## [1] 0.4433333
## [1] 0.4466667
## [1] 0.4733333
## [1] 0.4733333
## [1] 0.4966667
## [1] 0.4833333</code></pre>
<p><img src="lbrb_files/figure-html/unnamed-chunk-525-1.png" width="672" /></p>
<div id="over-represented-and-under-represented-dna-words" class="section level3" number="44.5.1">
<h3><span class="header-section-number">44.5.1</span> Over-represented and under-represented DNA words</h3>
<p>In the previous chapter, you learned that the <code>count()</code> function in the <code>seqinr</code> R package can calculate the frequency of all DNA words of a certain length in a DNA sequence. For example, if you want to know the frequency of all DNA words that are 2 nucleotides long in the Dengue virus genome sequence, you can type:</p>
<!-- Is this working? -->
<div class="sourceCode" id="cb750"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb750-1"><a href="local-variation-in-gc-content.html#cb750-1" aria-hidden="true" tabindex="-1"></a>dengueseq_vector_lc <span class="ot">&lt;-</span> <span class="fu">tolower</span>(dengueseq_vector)</span>
<span id="cb750-2"><a href="local-variation-in-gc-content.html#cb750-2" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">count</span>(dengueseq_vector_lc, <span class="dv">2</span>)</span></code></pre></div>
<pre><code>## 
##   aa   ac   ag   at   ca   cc   cg   ct   ga   gc   gg   gt   ta   tc   tg   tt 
## 1108  720  890  708  901  523  261  555  976  500  787  507  440  497  832  529</code></pre>
<!-- make ordered plot -->
<p>It is interesting to identify DNA words that are two nucleotides long (<strong>dinucleotides</strong>, i.e.. “AT”, “AC”, etc.) that are over-represented or under-represented in a DNA sequence. If a particular DNA word is over-represented in a sequence, it means that it occurs many more times in the sequence than you would have expected by chance. Similarly, if a particular DNA word is under-represented in a sequence, it means it occurs far fewer times in the sequence than you would have expected.</p>
<p>A statistic called <em>p</em> (<em>Rho</em>) is used to measure how over- or under-represented a particular DNA word is. For a 2-nucleotide (dinucleotide) DNA word p is calculated as:</p>
<!-- formatt latex -->
<p>p(xy) = fxy/(fx*fy),</p>
<p>where fxy and fx are the frequencies of the DNA words xy and x in the DNA sequence under study. For example, the value of for the DNA word “TA” can be calculated as: p(TA) = fTA/(fT* fA), where fTA, fT and fA are the frequencies of the DNA words “TA”, “T” and “A” in the DNA sequence.</p>
<p>The idea behind the p statistic is that, if a DNA sequence had a frequency fx of a 1-nucleotide DNA word x, and a frequency fy of a 1-nucleotide DNA word y, then we expect the frequency of the 2-nucleotide DNA word xy to be fx* fy. That is, the frequencies of the 2-nucleotide DNA words in a sequence are expected to be equal the products of the specific frequencies of the two nucleotides that compose them. If this were true, then p would be equal to 1. If we find that p is much greater than 1 for a particular 2-nucleotide word in a sequence, it indicates that that 2-nucleotide word is much more common in that sequence than expected (i.e.. it is over-represented).</p>
<p>For example, say that your input sequence has only 5% Ts (i.e.. fT = 0.05). In a random DNA sequence with 5% Ts, you would expect to see the word “TT” very infrequently. In fact, we would only expect 0.05 * 0.05=0.0025 (0.25%) of 2-nucleotide words to be TTs (i.e.. we expect fTT = fT* fT). This is because Ts are rare, so they are expected to be adjacent to each other very infrequently if the few Ts are randomly scattered throughout the DNA. Therefore, if you see lots of TT 2-nucleotide words in your real input sequence (eg. fTT = 0.3, so p = 0.3/0.0025 = 120), you would suspect that natural selection has acted to increase the number of occurrences of the TT word in the sequence (presumably because it has some beneficial biological function).</p>
<p>To find over-represented and under-represented DNA words that are 2 nucleotides long in the DEN-1 Dengue virus sequence, we can calculate the p statistic for each 2-nucleotide word in the sequence. For example, given the number of occurrences of the individual nucleotides A, C, G and T in the Dengue sequence, and the number of occurrences of the DNA word GC in the sequence (500, from above), we can calculate the value of p for the 2-nucleotide DNA word “GC”, using the formula p(GC) = fGC/(fG * fC), where fGC, fG and fC are the frequencies of the DNA words “GC”, “G” and “C” in the DNA sequence:</p>
<!-- TODO: is this messed up? -->
<div class="sourceCode" id="cb752"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb752-1"><a href="local-variation-in-gc-content.html#cb752-1" aria-hidden="true" tabindex="-1"></a><span class="fu">count</span>(dengueseq_vector_lc, <span class="at">wordsize =</span> <span class="dv">1</span>) <span class="co"># Get the number of occurrences of 1-nucleotide DNA words</span></span></code></pre></div>
<pre><code>## 
##    a    c    g    t 
## 3426 2240 2770 2299</code></pre>
<div class="sourceCode" id="cb754"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb754-1"><a href="local-variation-in-gc-content.html#cb754-1" aria-hidden="true" tabindex="-1"></a> <span class="dv">2770</span><span class="sc">/</span>(<span class="dv">3426</span><span class="sc">+</span><span class="dv">2240</span><span class="sc">+</span><span class="dv">2770</span><span class="sc">+</span><span class="dv">2299</span>) <span class="co"># Get fG</span></span></code></pre></div>
<pre><code>## [1] 0.2580345</code></pre>
<div class="sourceCode" id="cb756"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb756-1"><a href="local-variation-in-gc-content.html#cb756-1" aria-hidden="true" tabindex="-1"></a> <span class="dv">2240</span><span class="sc">/</span>(<span class="dv">3426</span><span class="sc">+</span><span class="dv">2240</span><span class="sc">+</span><span class="dv">2770</span><span class="sc">+</span><span class="dv">2299</span>) <span class="co"># Get fC</span></span></code></pre></div>
<pre><code>## [1] 0.2086633</code></pre>
<div class="sourceCode" id="cb758"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb758-1"><a href="local-variation-in-gc-content.html#cb758-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">count</span>(dengueseq_vector_lc, <span class="dv">2</span>) <span class="co"># Get the number of occurrences of 2-nucleotide DNA words</span></span></code></pre></div>
<pre><code>## 
##   aa   ac   ag   at   ca   cc   cg   ct   ga   gc   gg   gt   ta   tc   tg   tt 
## 1108  720  890  708  901  523  261  555  976  500  787  507  440  497  832  529</code></pre>
<div class="sourceCode" id="cb760"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb760-1"><a href="local-variation-in-gc-content.html#cb760-1" aria-hidden="true" tabindex="-1"></a><span class="dv">500</span><span class="sc">/</span>(<span class="dv">1108</span><span class="sc">+</span><span class="dv">720</span><span class="sc">+</span><span class="dv">890</span><span class="sc">+</span><span class="dv">708</span><span class="sc">+</span><span class="dv">901</span><span class="sc">+</span><span class="dv">523</span><span class="sc">+</span><span class="dv">261</span><span class="sc">+</span><span class="dv">555</span><span class="sc">+</span><span class="dv">976</span><span class="sc">+</span><span class="dv">500</span><span class="sc">+</span><span class="dv">787</span><span class="sc">+</span><span class="dv">507</span><span class="sc">+</span><span class="dv">440</span><span class="sc">+</span><span class="dv">497</span><span class="sc">+</span><span class="dv">832</span><span class="sc">+</span><span class="dv">529</span>) <span class="co"># Get fGC</span></span></code></pre></div>
<pre><code>## [1] 0.04658096</code></pre>
<div class="sourceCode" id="cb762"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb762-1"><a href="local-variation-in-gc-content.html#cb762-1" aria-hidden="true" tabindex="-1"></a> <span class="fl">0.04658096</span><span class="sc">/</span>(<span class="fl">0.2580345</span><span class="sc">*</span><span class="fl">0.2086633</span>) <span class="co"># Get rho(GC)</span></span></code></pre></div>
<pre><code>## [1] 0.8651364</code></pre>
<p>We calculate a value of p(GC) of approximately 0.865. This means that the DNA word “GC” is about 0.865 times as common in the DEN-1 Dengue virus sequence than expected. That is, it seems to be slightly under-represented.</p>
<p>Note that if the ratio of the observed to expected frequency of a particular DNA word is very low or very high, then we would suspect that there is a statistical under-representation or over-representation of that DNA word. However, to be sure that this over- or under-representation is statistically significant, we would need to do a statistical test. We will not deal with the topic of how to carry out the statistical test here.</p>
</div>
</div>
<div id="summary-3" class="section level2" number="44.6">
<h2><span class="header-section-number">44.6</span> Summary</h2>
<p>In this chapter, you will have learned to use the following functions:</p>
<ul>
<li><code>seq()</code> for creating a sequence of numbers</li>
<li><code>print()</code> for printing out the value of a variable</li>
<li><code>plot()</code> for making a plot (eg. a scatterplot)</li>
<li><code>numeric()</code> for making a numeric vector of a particular length</li>
</ul>
<p>All of these functions belong to the standard installation of R. You also learned how to use for loops to carry out the same operation again and again, each time on different inputs.</p>
</div>
<div id="further-reading" class="section level2" number="44.7">
<h2><span class="header-section-number">44.7</span> Further Reading</h2>
<p>For background reading on DNA sequence statistics, it is recommended to read Chapter 1 of Introduction to Computational Genomics: a case studies approach by Cristianini and Hahn (Cambridge University Press; www.computational-genomics.net/book/).</p>
<p>For more in-depth information and more examples on using the seqinr package for sequence analysis, look at the seqinr documentation, <a href="http://pbil.univ-lyon1.fr/software/seqinr/doc.php?lang=eng" class="uri">http://pbil.univ-lyon1.fr/software/seqinr/doc.php?lang=eng</a>.</p>
<p>There is also a very nice chapter on “Analyzing Sequences”, which includes examples of using seqinr for sequence analysis, in the book Applied statistics for bioinformatics using R by Krijnen (available online at cran.r-project.org/doc/contrib/Krijnen-IntroBioInfStatistics.pdf).</p>
</div>
<div id="acknowledgements-1" class="section level2" number="44.8">
<h2><span class="header-section-number">44.8</span> Acknowledgements</h2>
<p>In “A little book…” Coghlan write “Many of the ideas for the examples and exercises for this chapter were inspired by the Matlab case studies on <em>Haemophilus influenzae</em> (www.computational-genomics.net/case_studies/haemophilus_demo.html) and Bacteriophage lambda (<a href="http://www.computational-genomics.net/case_studies/lambdaphage_demo.html" class="uri">http://www.computational-genomics.net/case_studies/lambdaphage_demo.html</a>) from the website that accompanies the book Introduction to Computational Genomics: a case studies approach by Cristianini and Hahn (Cambridge University Press; www.computational-genomics.net/book/).”</p>
</div>
<div id="exercises-2" class="section level2" number="44.9">
<h2><span class="header-section-number">44.9</span> Exercises</h2>
<p>Answer the following questions, using the R package. For each question, please record your answer, and what you typed into R to get this answer.</p>
<p>Model answers to the exercises are given in Answers to the exercises on DNA Sequence Statistics.</p>
<ol style="list-style-type: decimal">
<li>Draw a sliding window plot of GC content in the DEN-1 Dengue virus genome, using a window size of 200 nucleotides. Do you see any regions of unusual DNA content in the genome (eg. a high peak or low trough)?
Make a sketch of each plot that you draw. At what position (in base-pairs) in the genome is there the largest change in local GC content (approximate position is fine here)? Compare the sliding window plots of GC content created using window sizes of 200 and 2000 nucleotides. How does window size affect your ability to detect differences within the Dengue virus genome?</li>
<li>Draw a sliding window plot of GC content in the genome sequence for the bacterium Mycobacterium leprae strain TN (accession NC_002677) using a window size of 20000 nucleotides. Do you see any regions of unusual DNA content in the genome (eg. a high peak or low trough)?</li>
<li>Make a sketch of each plot that you drew for the previous quetion. Write down the approximate nucleotide position of the highest peak or lowest trough that you see. Why do you think a window size of 20000 nucleotides was chosen? What do you see if you use a much smaller window size (eg. 200 nucleotides) or a much larger window size (eg. 200,000 nucleotides)?</li>
<li>Advanced: Write a function to calculate the AT content of a DNA sequence (i.e.. the fraction of the nucleotides in the sequence that are As or Ts). What is the AT content of the Mycobacterium leprae TN genome? Hint: use the function count() to make a table containing the number of As, Gs, Ts and Cs in the sequence. Remember that function count() produces a table object, and you can access the elements of a table object using double square brackets. Do you notice a relationship between the AT content of the <em>Mycobacterium leprae</em> TN genome, and its GC content?</li>
<li>Advanced: Write a function to draw a sliding window plot of AT content. Use it to make a sliding window plot of AT content along the Mycobacterium leprae TN genome, using a window size of 20000 nucleotides. Do you notice any relationship between the sliding window plot of GC content along the Mycobacterium leprae genome, and the sliding window plot of AT content? Make a sketch of the plot that you draw.</li>
<li>Is the 3-nucleotide word GAC GC over-represented or under-represented in the Mycobacterium leprae TN genome sequence? What is the frequency of this word in the sequence? What is the expected frequency of this word in the sequence? What is the p (Rho) value for this word? How would you figure out whether there is already an R function to calculate p (Rho)? Is there one that you could use?</li>
</ol>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="unrooted-neighbor-joining-phylogenetic-trees.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="computational-gene-finding.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/008-GC_content_local_variation/23-GC_content_local_variation.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/008-GC_content_local_variation/23-GC_content_local_variation.Rmd",
"text": null
},
"download": ["lbrb.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
