<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 11 Mini tutorial: Vectors in R | Computational Biology for All! Data science, bioinformatics, and other life-sciences awesomeness in R</title>
  <meta name="description" content="An entry-level text for computational biology" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 11 Mini tutorial: Vectors in R | Computational Biology for All! Data science, bioinformatics, and other life-sciences awesomeness in R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 11 Mini tutorial: Vectors in R | Computational Biology for All! Data science, bioinformatics, and other life-sciences awesomeness in R" />
  
  <meta name="twitter:description" content="An entry-level text for computational biology" />
  

<meta name="author" content="Nathan L. Brouwer" />


<meta name="date" content="2021-07-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="programming-in-r-for-loops-and-functions.html"/>
<link rel="next" href="note-please-save-this-file-with-a-new-file-name-to-your-documents-folder-or-desktop.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Computational Biology for All!</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>1</b> Downloading R</a>
<ul>
<li class="chapter" data-level="1.1" data-path="downloading-r.html"><a href="downloading-r.html#preface-1"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>1.2</b> Introduction to R</a></li>
<li class="chapter" data-level="1.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>1.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>1.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="1.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>1.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>1.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>2</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="2.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>2.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="2.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>2.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>3</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>3.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="3.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>3.2</b> Downloading packages with the function install.packages()</a></li>
<li class="chapter" data-level="3.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>3.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>4</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="4.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>4.1</b> Bioconductor</a></li>
<li class="chapter" data-level="4.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>4.2</b> Installing BiocManager</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>5</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="5.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>5.1</b> Vocabulary</a></li>
<li class="chapter" data-level="5.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions-etc"><i class="fa fa-check"></i><b>5.2</b> R functions etc</a></li>
<li class="chapter" data-level="5.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#a-brief-introduction-to-r-1"><i class="fa fa-check"></i><b>5.3</b> A brief introduction to R</a></li>
<li class="chapter" data-level="5.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>5.4</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html"><i class="fa fa-check"></i><b>6</b> Downloading NCBI sequence data by hand</a>
<ul>
<li class="chapter" data-level="6.1" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html#retrieving-genome-sequence-data-via-the-ncbi-website"><i class="fa fa-check"></i><b>6.1</b> Retrieving genome sequence data via the NCBI website</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>7</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="7.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>7.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#note-on-the-biology-in-this-section"><i class="fa fa-check"></i><b>7.1.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="7.1.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>7.1.2</b> Vocabulary</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>7.2</b> Preliminaries</a></li>
<li class="chapter" data-level="7.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>7.3</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="7.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>7.4</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>7.4.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="7.4.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>7.4.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="7.4.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>7.4.3</b> DNA words</a></li>
<li class="chapter" data-level="7.4.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>7.4.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>7.5</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>7.5.1</b> License</a></li>
<li class="chapter" data-level="7.5.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>7.5.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html"><i class="fa fa-check"></i><b>8</b> Downloading DNA sequences with R</a>
<ul>
<li class="chapter" data-level="8.0.1" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#functions"><i class="fa fa-check"></i><b>8.0.1</b> Functions</a></li>
<li class="chapter" data-level="8.0.2" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#softwarewebsites"><i class="fa fa-check"></i><b>8.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="8.0.3" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#r-vocabulary"><i class="fa fa-check"></i><b>8.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="8.0.4" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#file-types"><i class="fa fa-check"></i><b>8.0.4</b> File types</a></li>
<li class="chapter" data-level="8.0.5" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>8.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="8.0.6" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>8.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="8.0.7" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#preliminaries-1"><i class="fa fa-check"></i><b>8.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="8.1" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>8.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>8.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="8.1.2" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>8.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="8.1.3" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#fasta-file-format"><i class="fa fa-check"></i><b>8.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="8.1.4" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>8.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="8.1.5" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>8.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>8.2</b> Convert FASTA sequence to an R variable</a></li>
<li class="chapter" data-level="8.3" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#saving-fasta-files"><i class="fa fa-check"></i><b>8.3</b> Saving FASTA files</a></li>
<li class="chapter" data-level="8.4" data-path="downloading-dna-sequences-with-r.html"><a href="downloading-dna-sequences-with-r.html#reading-in-fasta-files"><i class="fa fa-check"></i><b>8.4</b> Reading in FASTA files</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html"><i class="fa fa-check"></i><b>9</b> Introduction to biological sequences databases</a>
<ul>
<li class="chapter" data-level="9.1" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#to-do"><i class="fa fa-check"></i><b>9.1</b> To do</a></li>
<li class="chapter" data-level="9.2" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#biological-sequence-databases"><i class="fa fa-check"></i><b>9.2</b> Biological sequence databases</a></li>
<li class="chapter" data-level="9.3" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-ncbi-sequence-database-1"><i class="fa fa-check"></i><b>9.3</b> The NCBI Sequence Database</a></li>
<li class="chapter" data-level="9.4" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-ncbi-sub-databases"><i class="fa fa-check"></i><b>9.4</b> The NCBI Sub-Databases</a></li>
<li class="chapter" data-level="9.5" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#ncbi-genbank-record-format"><i class="fa fa-check"></i><b>9.5</b> NCBI GenBank Record Format</a></li>
<li class="chapter" data-level="9.6" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-fasta-file-format"><i class="fa fa-check"></i><b>9.6</b> The FASTA file format</a></li>
<li class="chapter" data-level="9.7" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#graphics"><i class="fa fa-check"></i><b>9.7</b> Graphics</a></li>
<li class="chapter" data-level="9.8" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#refseq"><i class="fa fa-check"></i><b>9.8</b> RefSeq</a></li>
<li class="chapter" data-level="9.9" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#querying-the-ncbi-database"><i class="fa fa-check"></i><b>9.9</b> Querying the NCBI Database</a></li>
<li class="chapter" data-level="9.10" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#querying-the-ncbi-database-via-the-ncbi-website-for-reference"><i class="fa fa-check"></i><b>9.10</b> Querying the NCBI Database via the NCBI Website (for reference)</a></li>
<li class="chapter" data-level="9.11" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#example-finding-the-sequences-published-in-nature-460352-358-for-reference"><i class="fa fa-check"></i><b>9.11</b> Example: finding the sequences published in Nature 460:352-358 (for reference)</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="programming-in-r-for-loops-and-functions.html"><a href="programming-in-r-for-loops-and-functions.html"><i class="fa fa-check"></i><b>10</b> Programming in R: for loops and functions</a>
<ul>
<li class="chapter" data-level="10.1" data-path="programming-in-r-for-loops-and-functions.html"><a href="programming-in-r-for-loops-and-functions.html#vocab"><i class="fa fa-check"></i><b>10.1</b> Vocab</a></li>
<li class="chapter" data-level="10.2" data-path="programming-in-r-for-loops-and-functions.html"><a href="programming-in-r-for-loops-and-functions.html#basic-programming-in-r"><i class="fa fa-check"></i><b>10.2</b> Basic programming in R</a></li>
<li class="chapter" data-level="10.3" data-path="programming-in-r-for-loops-and-functions.html"><a href="programming-in-r-for-loops-and-functions.html#functions-in-r"><i class="fa fa-check"></i><b>10.3</b> Functions in R</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>11</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-and-functions"><i class="fa fa-check"></i><b>11.1</b> Vocab and functions</a></li>
<li class="chapter" data-level="11.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vectors-in-r"><i class="fa fa-check"></i><b>11.2</b> Vectors in R</a></li>
<li class="chapter" data-level="11.3" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#math-on-vectors"><i class="fa fa-check"></i><b>11.3</b> Math on vectors</a></li>
<li class="chapter" data-level="11.4" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#functions-on-vectors"><i class="fa fa-check"></i><b>11.4</b> Functions on vectors</a></li>
<li class="chapter" data-level="11.5" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>11.5</b> Operations with two vectors</a></li>
<li class="chapter" data-level="11.6" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#subsetting-vectors"><i class="fa fa-check"></i><b>11.6</b> Subsetting vectors</a></li>
<li class="chapter" data-level="11.7" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#sequences-of-numbers"><i class="fa fa-check"></i><b>11.7</b> Sequences of numbers</a></li>
<li class="chapter" data-level="11.8" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>11.8</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="11.9" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>11.9</b> Regular expressions can modify character data</a></li>
<li class="chapter" data-level="11.10" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#plotting-vectors-in-base-r"><i class="fa fa-check"></i><b>11.10</b> Plotting vectors in base R</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="note-please-save-this-file-with-a-new-file-name-to-your-documents-folder-or-desktop.html"><a href="note-please-save-this-file-with-a-new-file-name-to-your-documents-folder-or-desktop.html"><i class="fa fa-check"></i><b>12</b> NOTE: please save this file with a new file name to your documents folder or desktop</a></li>
<li class="chapter" data-level="13" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html"><i class="fa fa-check"></i><b>13</b> Downloading sequences from UniProt by hand</a>
<ul>
<li class="chapter" data-level="13.1" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#vocab-1"><i class="fa fa-check"></i><b>13.1</b> Vocab</a></li>
<li class="chapter" data-level="13.2" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#uniprot"><i class="fa fa-check"></i><b>13.2</b> UniProt</a></li>
<li class="chapter" data-level="13.3" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#viewing-the-uniprot-webpage-for-a-protein-sequence"><i class="fa fa-check"></i><b>13.3</b> Viewing the UniProt webpage for a protein sequence</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#protein-function"><i class="fa fa-check"></i><b>13.3.1</b> Protein function</a></li>
<li class="chapter" data-level="13.3.2" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#protein-sequence-and-size"><i class="fa fa-check"></i><b>13.3.2</b> Protein sequence and size</a></li>
<li class="chapter" data-level="13.3.3" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#other-information"><i class="fa fa-check"></i><b>13.3.3</b> Other information</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#retrieving-a-uniprot-protein-sequence-via-the-uniprot-website"><i class="fa fa-check"></i><b>13.4</b> Retrieving a UniProt protein sequence via the UniProt website</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="sequence-databases.html"><a href="sequence-databases.html"><i class="fa fa-check"></i><b>14</b> Sequence Databases</a>
<ul>
<li class="chapter" data-level="14.1" data-path="sequence-databases.html"><a href="sequence-databases.html#querying-the-ncbi-database-via-r"><i class="fa fa-check"></i><b>14.1</b> Querying the NCBI Database via R</a></li>
<li class="chapter" data-level="14.2" data-path="sequence-databases.html"><a href="sequence-databases.html#summary-1"><i class="fa fa-check"></i><b>14.2</b> Summary</a></li>
<li class="chapter" data-level="14.3" data-path="sequence-databases.html"><a href="sequence-databases.html#acknowledgements-1"><i class="fa fa-check"></i><b>14.3</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html"><i class="fa fa-check"></i><b>15</b> Local variation in GC content</a>
<ul>
<li class="chapter" data-level="15.0.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#note-on-the-biology-in-this-section-1"><i class="fa fa-check"></i><b>15.0.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="15.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#vocabulary-2"><i class="fa fa-check"></i><b>15.1</b> Vocabulary</a></li>
<li class="chapter" data-level="15.2" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#reading-sequence-data-with-rentrezentrez_fetch"><i class="fa fa-check"></i><b>15.2</b> Reading sequence data with rentrez::entrez_fetch</a></li>
<li class="chapter" data-level="15.3" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#local-variation-in-gc-content-1"><i class="fa fa-check"></i><b>15.3</b> Local variation in GC content</a></li>
<li class="chapter" data-level="15.4" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-analysis-of-gc-content"><i class="fa fa-check"></i><b>15.4</b> A sliding window analysis of GC content</a>
<ul>
<li class="chapter" data-level="15.4.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#over-represented-and-under-represented-dna-words"><i class="fa fa-check"></i><b>15.4.1</b> Over-represented and under-represented DNA words</a></li>
</ul></li>
<li class="chapter" data-level="15.5" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#summary-2"><i class="fa fa-check"></i><b>15.5</b> Summary</a></li>
<li class="chapter" data-level="15.6" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#further-reading"><i class="fa fa-check"></i><b>15.6</b> Further Reading</a></li>
<li class="chapter" data-level="15.7" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#acknowledgements-2"><i class="fa fa-check"></i><b>15.7</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="15.7.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#license-1"><i class="fa fa-check"></i><b>15.7.1</b> License</a></li>
</ul></li>
<li class="chapter" data-level="15.8" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#exercises-1"><i class="fa fa-check"></i><b>15.8</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="aligments-in-r.html"><a href="aligments-in-r.html"><i class="fa fa-check"></i><b>16</b> Aligments in R</a>
<ul>
<li class="chapter" data-level="16.1" data-path="aligments-in-r.html"><a href="aligments-in-r.html#retrieving-a-uniprot-protein-sequence-using-seqinr"><i class="fa fa-check"></i><b>16.1</b> Retrieving a UniProt protein sequence using SeqinR</a></li>
<li class="chapter" data-level="16.2" data-path="aligments-in-r.html"><a href="aligments-in-r.html#comparing-two-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>16.2</b> Comparing two sequences using a dotplot</a></li>
<li class="chapter" data-level="16.3" data-path="aligments-in-r.html"><a href="aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>16.3</b> Aligning UniProt sequences</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>17</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="17.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>17.1</b> Prelminaries</a></li>
<li class="chapter" data-level="17.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-list-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>17.2</b> Retrieving a list of sequences from UniProt</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="multiple-alignment-with-clustal.html"><a href="multiple-alignment-with-clustal.html"><i class="fa fa-check"></i><b>18</b> Multiple alignment with Clustal</a>
<ul>
<li class="chapter" data-level="18.1" data-path="multiple-alignment-with-clustal.html"><a href="multiple-alignment-with-clustal.html#installing-the-clustal-multiple-alignment-software"><i class="fa fa-check"></i><b>18.1</b> Installing the CLUSTAL multiple alignment software</a></li>
<li class="chapter" data-level="18.2" data-path="multiple-alignment-with-clustal.html"><a href="multiple-alignment-with-clustal.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>18.2</b> Discarding very poorly conserved regions from an alignment</a></li>
<li class="chapter" data-level="18.3" data-path="multiple-alignment-with-clustal.html"><a href="multiple-alignment-with-clustal.html#calculating-genetic-distances-between-protein-sequences"><i class="fa fa-check"></i><b>18.3</b> Calculating genetic distances between protein sequences</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="rooted-neighbor-joining-phylogenetic-trees.html"><a href="rooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>19</b> Rooted Neighbor-joining Phylogenetic trees</a>
<ul>
<li class="chapter" data-level="19.1" data-path="rooted-neighbor-joining-phylogenetic-trees.html"><a href="rooted-neighbor-joining-phylogenetic-trees.html#building-a-rooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>19.1</b> Building a rooted phylogenetic tree for protein sequences</a></li>
<li class="chapter" data-level="19.2" data-path="rooted-neighbor-joining-phylogenetic-trees.html"><a href="rooted-neighbor-joining-phylogenetic-trees.html#building-a-phylogenetic-tree-for-dna-or-mrna-sequences"><i class="fa fa-check"></i><b>19.2</b> Building a phylogenetic tree for DNA or mRNA sequences</a></li>
<li class="chapter" data-level="19.3" data-path="rooted-neighbor-joining-phylogenetic-trees.html"><a href="rooted-neighbor-joining-phylogenetic-trees.html#summary-3"><i class="fa fa-check"></i><b>19.3</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>20</b> Unrooted Neighbor-joining Phylogenetic trees</a>
<ul>
<li class="chapter" data-level="20.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>20.1</b> Building an unrooted phylogenetic tree for protein sequences</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html"><i class="fa fa-check"></i><b>21</b> Testing the significance of an alignment</a>
<ul>
<li class="chapter" data-level="21.1" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#calculating-the-statistical-significance-of-a-pairwise-global-alignment"><i class="fa fa-check"></i><b>21.1</b> Calculating the statistical significance of a pairwise global alignment</a></li>
<li class="chapter" data-level="21.2" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#summary-4"><i class="fa fa-check"></i><b>21.2</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html"><i class="fa fa-check"></i><b>22</b> Computational gene finding</a>
<ul>
<li class="chapter" data-level="22.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#the-genetic-code"><i class="fa fa-check"></i><b>22.1</b> The genetic code</a></li>
<li class="chapter" data-level="22.2" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#reading-frames"><i class="fa fa-check"></i><b>22.2</b> Reading frames</a></li>
<li class="chapter" data-level="22.3" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-forward-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>22.3</b> Finding open reading frames on the forward strand of a DNA sequence</a></li>
<li class="chapter" data-level="22.4" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#lengths-of-open-reading-frames"><i class="fa fa-check"></i><b>22.4</b> Lengths of open reading frames</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html"><i class="fa fa-check"></i><b>23</b> Significance of ORFs</a>
<ul>
<li class="chapter" data-level="23.1" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#identifying-significant-open-reading-frames"><i class="fa fa-check"></i><b>23.1</b> Identifying significant open reading frames</a></li>
<li class="chapter" data-level="23.2" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#summary-5"><i class="fa fa-check"></i><b>23.2</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="comparative-genomics.html"><a href="comparative-genomics.html"><i class="fa fa-check"></i><b>24</b> Comparative Genomics</a></li>
<li class="chapter" data-level="25" data-path="introduction-1.html"><a href="introduction-1.html"><i class="fa fa-check"></i><b>25</b> Introduction</a></li>
<li class="chapter" data-level="26" data-path="using-the-biomart-r-library-to-query-the-ensembl-database.html"><a href="using-the-biomart-r-library-to-query-the-ensembl-database.html"><i class="fa fa-check"></i><b>26</b> Using the biomaRt R Library to Query the Ensembl Database</a></li>
<li class="chapter" data-level="27" data-path="comparing-the-number-of-genes-in-two-species.html"><a href="comparing-the-number-of-genes-in-two-species.html"><i class="fa fa-check"></i><b>27</b> Comparing the number of genes in two species</a></li>
<li class="chapter" data-level="28" data-path="identifying-homologous-genes-between-two-species.html"><a href="identifying-homologous-genes-between-two-species.html"><i class="fa fa-check"></i><b>28</b> Identifying homologous genes between two species</a></li>
<li class="chapter" data-level="29" data-path="summary-6.html"><a href="summary-6.html"><i class="fa fa-check"></i><b>29</b> Summary</a></li>
<li class="chapter" data-level="30" data-path="links-and-further-reading-1.html"><a href="links-and-further-reading-1.html"><i class="fa fa-check"></i><b>30</b> Links and Further Reading</a></li>
<li class="chapter" data-level="31" data-path="acknowledgements-3.html"><a href="acknowledgements-3.html"><i class="fa fa-check"></i><b>31</b> Acknowledgements</a></li>
<li class="chapter" data-level="32" data-path="contact.html"><a href="contact.html"><i class="fa fa-check"></i><b>32</b> Contact</a></li>
<li class="chapter" data-level="33" data-path="license-2.html"><a href="license-2.html"><i class="fa fa-check"></i><b>33</b> License</a></li>
<li class="chapter" data-level="34" data-path="exercises-2.html"><a href="exercises-2.html"><i class="fa fa-check"></i><b>34</b> Exercises</a></li>
<li class="chapter" data-level="35" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html"><i class="fa fa-check"></i><b>35</b> Hidden Markov Models</a>
<ul>
<li class="chapter" data-level="35.1" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#r-programming-101-initializing-vectors"><i class="fa fa-check"></i><b>35.1</b> R Programming 101: Initializing vectors</a></li>
<li class="chapter" data-level="35.2" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-multinomial-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>35.2</b> A multinomial model of DNA sequence evolution</a></li>
<li class="chapter" data-level="35.3" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-multinomial-model"><i class="fa fa-check"></i><b>35.3</b> Generating a DNA sequence using a multinomial model</a></li>
<li class="chapter" data-level="35.4" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>35.4</b> A Markov model of DNA sequence evolution</a></li>
<li class="chapter" data-level="35.5" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-for-a-markov-model"><i class="fa fa-check"></i><b>35.5</b> The transition matrix for a Markov model</a></li>
<li class="chapter" data-level="35.6" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-markov-model"><i class="fa fa-check"></i><b>35.6</b> Generating a DNA sequence using a Markov model</a></li>
<li class="chapter" data-level="35.7" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>35.7</b> A Hidden Markov Model of DNA sequence evolution</a></li>
<li class="chapter" data-level="35.8" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-and-emission-matrix-for-a-hmm"><i class="fa fa-check"></i><b>35.8</b> The transition matrix and emission matrix for a HMM</a></li>
<li class="chapter" data-level="35.9" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-hmm"><i class="fa fa-check"></i><b>35.9</b> Generating a DNA sequence using a HMM</a></li>
<li class="chapter" data-level="35.10" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#inferring-the-states-of-a-hmm-that-generated-a-dna-sequence"><i class="fa fa-check"></i><b>35.10</b> Inferring the states of a HMM that generated a DNA sequence</a></li>
<li class="chapter" data-level="35.11" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-protein-sequence-evolution"><i class="fa fa-check"></i><b>35.11</b> A Hidden Markov Model of protein sequence evolution</a></li>
<li class="chapter" data-level="35.12" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#summary-7"><i class="fa fa-check"></i><b>35.12</b> Summary</a></li>
<li class="chapter" data-level="35.13" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#links-and-further-reading-2"><i class="fa fa-check"></i><b>35.13</b> Links and Further Reading</a></li>
<li class="chapter" data-level="35.14" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#exercises-3"><i class="fa fa-check"></i><b>35.14</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="35.15" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>35.15</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="35.15.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>35.15.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="35.15.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>35.15.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="35.15.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>35.15.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="35.15.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>35.15.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-3"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="35.16" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>35.16</b> Help!</a>
<ul>
<li class="chapter" data-level="35.16.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>35.16.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="35.16.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>35.16.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="35.16.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>35.16.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="35.16.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>35.16.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="35.17" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>35.17</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="35.17.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>35.17.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="35.17.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>35.17.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="35.18" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>35.18</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Computational Biology for All! Data science, bioinformatics, and other life-sciences awesomeness in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mini-tutorial-vectors-in-r" class="section level1" number="11">
<h1><span class="header-section-number">Chapter 11</span> Mini tutorial: Vectors in R</h1>
<p>This is a modification of part of<a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter2.html">“DNA Sequence Statistics (2)”</a> from Avril Coghlan’s <a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/index.html"><em>A little book of R for bioinformatics.</em></a>. Most of text and code was originally written by Dr. Coghlan and distributed under the <a href="https://creativecommons.org/licenses/by/3.0/us/">Creative Commons 3.0</a> license.</p>
<div id="vocab-and-functions" class="section level2" number="11.1">
<h2><span class="header-section-number">11.1</span> Vocab and functions</h2>
<ul>
<li>base R</li>
<li>scalar, vector, matrix</li>
<li>regular expressions</li>
<li>seq()</li>
<li>is(), is.vector(), is.matrix()</li>
<li>gsub()</li>
</ul>
</div>
<div id="vectors-in-r" class="section level2" number="11.2">
<h2><span class="header-section-number">11.2</span> Vectors in R</h2>
<p><strong>Variables</strong> in R include <strong>scalars</strong>, <strong>vectors</strong>, and <strong>lists</strong>. <strong>Functions</strong> in R carry out operations on variables, for example, using the log10() function to calculate the log to the base 10 of a scalar variable x, or using the mean() function to calculate the average of the values in a vector variable myvector. For example, we can use log10() on a scalar like this:</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="mini-tutorial-vectors-in-r.html#cb142-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb142-2"><a href="mini-tutorial-vectors-in-r.html#cb142-2" aria-hidden="true" tabindex="-1"></a><span class="fu">log10</span>(x)</span></code></pre></div>
<pre><code>## [1] 2</code></pre>
<p>Note that while mathematically x is a single number, or a scalar, R considers it to be a vector:</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="mini-tutorial-vectors-in-r.html#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is.vector</span>(x)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>There are many “is” commands. What is returned when you run is.matrix() on a vector?</p>
<p>Mathematically this is a bit odd, since often a vector is defined as a one dimenionsal matrix, eg, a single column or single row of a matrix. But in R land, a vector is a vector, and matrix is a matrix, and there are no explicit scalars.</p>
</div>
<div id="math-on-vectors" class="section level2" number="11.3">
<h2><span class="header-section-number">11.3</span> Math on vectors</h2>
<p>Vectors can serve as the input for mathematical operations. When this is done R does the mathematical operation seperately on each element of the vector.</p>
<p>Let’s make a vector of numbers</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="mini-tutorial-vectors-in-r.html#cb146-1" aria-hidden="true" tabindex="-1"></a>myvector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">30</span>,<span class="dv">16</span>,<span class="dv">303</span>,<span class="dv">99</span>,<span class="dv">11</span>,<span class="dv">111</span>)</span></code></pre></div>
<p>What happens when we multiply myvector by 10?</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="mini-tutorial-vectors-in-r.html#cb147-1" aria-hidden="true" tabindex="-1"></a>myvector<span class="sc">*</span><span class="dv">10</span></span></code></pre></div>
<pre><code>## [1]  300  160 3030  990  110 1110</code></pre>
<p>The normal order of operations rules apply to vectors as they do to operations were more used to. So multplying myvector by 10 is the same whether you put he 10 before or after vector. Write the code below to check this:</p>
<p>What happen when you sutract 30 from myvector? Write the code below.</p>
<p>You can also square a vector</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="mini-tutorial-vectors-in-r.html#cb149-1" aria-hidden="true" tabindex="-1"></a>myvector<span class="sc">^</span><span class="dv">2</span></span></code></pre></div>
<pre><code>## [1]   900   256 91809  9801   121 12321</code></pre>
<p>Take the square root of a vector</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="mini-tutorial-vectors-in-r.html#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(myvector)</span></code></pre></div>
<pre><code>## [1]  5.477226  4.000000 17.406895  9.949874  3.316625 10.535654</code></pre>
<p>and take the log of a vector</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="mini-tutorial-vectors-in-r.html#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">log</span>(myvector)</span></code></pre></div>
<pre><code>## [1] 3.401197 2.772589 5.713733 4.595120 2.397895 4.709530</code></pre>
<p>Here we are working on a seperate vector object; all of these rules apply to a column in a matrix or a dataframe.</p>
</div>
<div id="functions-on-vectors" class="section level2" number="11.4">
<h2><span class="header-section-number">11.4</span> Functions on vectors</h2>
<p>We can use functions on vectors. Typically these use the vectors as an input and all the numbers are procssed into an output. Call the mean() function on the vector we made called myvector</p>
<p>The function sd() calcualtes the standard deviation. Apply the sd() to myvector:</p>
</div>
<div id="operations-with-two-vectors" class="section level2" number="11.5">
<h2><span class="header-section-number">11.5</span> Operations with two vectors</h2>
<p>You can also subtract one vector from another vector. Make another vector with the numbers 5, 10, 15, 20, 25, 30. Call this myvector2:</p>
<p>Now subtract myvector2 from myvector. What happens?</p>
</div>
<div id="subsetting-vectors" class="section level2" number="11.6">
<h2><span class="header-section-number">11.6</span> Subsetting vectors</h2>
<p>You can extract an <strong>element</strong> of a vector by typing the vector name with the index of that element given in <strong>square brackets</strong>. For example, to get the value of the 3rd element in the vector myvector, we type:</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="mini-tutorial-vectors-in-r.html#cb155-1" aria-hidden="true" tabindex="-1"></a>myvector[<span class="dv">3</span>]</span></code></pre></div>
<pre><code>## [1] 303</code></pre>
<p>Extract the 4th element of the vector</p>
<p>You extract more than one element by using a vector in the brackets:</p>
<p>First, say I want to extract the 3rd and the 4th element. I can make a vector with 3 and 4 in it:</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="mini-tutorial-vectors-in-r.html#cb157-1" aria-hidden="true" tabindex="-1"></a>nums <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">4</span>)</span></code></pre></div>
<p>Then put that vector in the brackets:</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="mini-tutorial-vectors-in-r.html#cb158-1" aria-hidden="true" tabindex="-1"></a>myvector[nums]</span></code></pre></div>
<pre><code>## [1] 303  99</code></pre>
<p>We can also do it directly like this:</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="mini-tutorial-vectors-in-r.html#cb160-1" aria-hidden="true" tabindex="-1"></a>myvector[<span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">4</span>)]</span></code></pre></div>
<pre><code>## [1] 303  99</code></pre>
<p>When numbers are consecutive you can do this:</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="mini-tutorial-vectors-in-r.html#cb162-1" aria-hidden="true" tabindex="-1"></a>myvector[<span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>]</span></code></pre></div>
<pre><code>## [1] 303  99</code></pre>
<p>In the chunk below extract tghe 1st and 2nd elements</p>
</div>
<div id="sequences-of-numbers" class="section level2" number="11.7">
<h2><span class="header-section-number">11.7</span> Sequences of numbers</h2>
<p>Often we want a vectors of numbers in sequential order. The easiest way to do this is using a colo</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="mini-tutorial-vectors-in-r.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span></span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20</code></pre>
<p>Note that in R 1:20 is equivalent to c(1:20)</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="mini-tutorial-vectors-in-r.html#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20</code></pre>
<p>Usually to empaphasize that a vector is being created I will use c(1:20)</p>
<p>We can do any number to any numbers</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="mini-tutorial-vectors-in-r.html#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="dv">20</span><span class="sc">:</span><span class="dv">40</span>)</span></code></pre></div>
<pre><code>##  [1] 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40</code></pre>
<p>We can also do it in reverse. In the code below put 40 before 20:</p>
<p>A useful function in R is the seq() function, which is an explicit function that can be used to create a vector containing a sequence of numbers that run from a particular number to another particular number.</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="mini-tutorial-vectors-in-r.html#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">10</span>)</span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
<p>Using seq() instead of a : can be useful for readability to make is explicity what is going on. More importantly, seq has an arguement by = For example, if we want to create the sequence of numbers from 1-10 in steps of 1 (ie. 1, 2, 3, 4, … 10), we can type:</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="mini-tutorial-vectors-in-r.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
<p>We can change the <strong>step size</strong> by altering the value of the “by” argument given to the function seq(). For example, if we want to create a sequence of numbers from 1-100 in steps of 20 (ie. 1, 21, 41, … 101), we can type:</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="mini-tutorial-vectors-in-r.html#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">101</span>, <span class="at">by =</span> <span class="dv">20</span>)</span></code></pre></div>
<pre><code>## [1]   1  21  41  61  81 101</code></pre>
</div>
<div id="vectors-can-hold-numeric-or-character-data" class="section level2" number="11.8">
<h2><span class="header-section-number">11.8</span> Vectors can hold numeric or character data</h2>
<p>The vector we created above holds numeric data, as indicated by is()</p>
<p>As we’ve seen, vectors can holder character data, like the genetic code</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="mini-tutorial-vectors-in-r.html#cb176-1" aria-hidden="true" tabindex="-1"></a>myvector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;T&quot;</span>,<span class="st">&quot;G&quot;</span>)</span>
<span id="cb176-2"><a href="mini-tutorial-vectors-in-r.html#cb176-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-3"><a href="mini-tutorial-vectors-in-r.html#cb176-3" aria-hidden="true" tabindex="-1"></a>myvector</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;T&quot; &quot;G&quot;</code></pre>
</div>
<div id="regular-expressions-can-modify-character-data" class="section level2" number="11.9">
<h2><span class="header-section-number">11.9</span> Regular expressions can modify character data</h2>
<p>We can use <strong>regular expressions</strong> to modify character data. For example, change the Ts to Us</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="mini-tutorial-vectors-in-r.html#cb178-1" aria-hidden="true" tabindex="-1"></a>myvector <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;U&quot;</span>, myvector)</span></code></pre></div>
<p>Now check it out</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="mini-tutorial-vectors-in-r.html#cb179-1" aria-hidden="true" tabindex="-1"></a>myvector</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;U&quot; &quot;G&quot;</code></pre>
</div>
<div id="plotting-vectors-in-base-r" class="section level2" number="11.10">
<h2><span class="header-section-number">11.10</span> Plotting vectors in base R</h2>
<p>R allows the production of a variety of plots, including scatterplots, histograms, piecharts, and boxplots. For example, if you have two vectors of numbers myvector1 and myvector2, you can plot a scatterplot of the values in myvector1 against the values in myvector2 using the <strong>base R</strong> plot() function.</p>
<p>First, let’s make up some data in put it in vectors:</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="mini-tutorial-vectors-in-r.html#cb181-1" aria-hidden="true" tabindex="-1"></a>myvector1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">15</span>, <span class="dv">22</span>, <span class="dv">35</span>, <span class="dv">43</span>)</span>
<span id="cb181-2"><a href="mini-tutorial-vectors-in-r.html#cb181-2" aria-hidden="true" tabindex="-1"></a>myvector2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="fl">3.2</span>, <span class="fl">3.9</span>, <span class="fl">4.1</span>, <span class="fl">5.2</span>)</span></code></pre></div>
<p>Not plot with the base R plot() function:</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="mini-tutorial-vectors-in-r.html#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(myvector1, myvector2)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-152-1.png" width="672" /></p>
<p>If you want to label the axes on the plot, you can do this by giving the plot() function values for its optional arguments xlab and ylab:</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="mini-tutorial-vectors-in-r.html#cb183-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(myvector1, </span>
<span id="cb183-2"><a href="mini-tutorial-vectors-in-r.html#cb183-2" aria-hidden="true" tabindex="-1"></a>     myvector2, </span>
<span id="cb183-3"><a href="mini-tutorial-vectors-in-r.html#cb183-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;vector1&quot;</span>, </span>
<span id="cb183-4"><a href="mini-tutorial-vectors-in-r.html#cb183-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;vector2&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-153-1.png" width="672" /></p>
<p>We can store character data in vectors so if we want we could do this to set up our labels:</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="mini-tutorial-vectors-in-r.html#cb184-1" aria-hidden="true" tabindex="-1"></a>mylabels <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="st">&quot;vector1&quot;</span>,<span class="st">&quot;vector2&quot;</span>)</span></code></pre></div>
<p>Then use bracket notation to call the labels from nthe vector</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="mini-tutorial-vectors-in-r.html#cb185-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(myvector1, </span>
<span id="cb185-2"><a href="mini-tutorial-vectors-in-r.html#cb185-2" aria-hidden="true" tabindex="-1"></a>     myvector2, </span>
<span id="cb185-3"><a href="mini-tutorial-vectors-in-r.html#cb185-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span>mylabels[<span class="dv">1</span>],</span>
<span id="cb185-4"><a href="mini-tutorial-vectors-in-r.html#cb185-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span>mylabels[<span class="dv">1</span>])</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-155-1.png" width="672" /></p>
<p>The above plot has an error: I intentionlly made it so that “vector1” is printed on both axes. In the chunk below make it so that “vector2” is printed on the y axis.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="mini-tutorial-vectors-in-r.html#cb186-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(myvector1, </span>
<span id="cb186-2"><a href="mini-tutorial-vectors-in-r.html#cb186-2" aria-hidden="true" tabindex="-1"></a>     myvector2, </span>
<span id="cb186-3"><a href="mini-tutorial-vectors-in-r.html#cb186-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span>mylabels[<span class="dv">1</span>],</span>
<span id="cb186-4"><a href="mini-tutorial-vectors-in-r.html#cb186-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span>mylabels[<span class="dv">1</span>])</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-156-1.png" width="672" /></p>
<p>Base R has lots of plotting functions; additionally, people have written packages to implement new plotting capabilities. ggplot2 is currently the most popular plotting package, and ggpubr is a package which makes ggplot2 easier to use. For quick plots we’ll use base R functions, and when we get to more important things we’ll use ggplot2 and ggpubr.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="programming-in-r-for-loops-and-functions.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="note-please-save-this-file-with-a-new-file-name-to-your-documents-folder-or-desktop.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
