<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 26 Multiple sequence alignment in R | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 26 Multiple sequence alignment in R | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 26 Multiple sequence alignment in R | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="retrieving-multiple-sequences-in-r.html"/>
<link rel="next" href="calculating-genetic-distances-between-sequences.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>1</b> How to review this book</a>
<ul>
<li class="chapter" data-level="1.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>1.1</b> File format</a></li>
<li class="chapter" data-level="1.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>1.2</b> What can you do?</a></li>
<li class="chapter" data-level="1.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>1.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>1.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="1.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>1.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="1.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>1.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="1.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>1.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="test-audio-link.html"><a href="test-audio-link.html"><i class="fa fa-check"></i><b>2</b> Test audio link</a></li>
<li class="chapter" data-level="3" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>3</b> Test shiny app</a></li>
<li class="chapter" data-level="4" data-path="test-youtube-vidoe.html"><a href="test-youtube-vidoe.html"><i class="fa fa-check"></i><b>4</b> Test YouTube vidoe</a></li>
<li class="chapter" data-level="5" data-path="test-webpage-insertion.html"><a href="test-webpage-insertion.html"><i class="fa fa-check"></i><b>5</b> Test webpage insertion</a></li>
<li class="chapter" data-level="6" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>6</b> Datacamp test</a></li>
<li class="chapter" data-level="7" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>7</b> Downloading R</a>
<ul>
<li class="chapter" data-level="7.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>7.1</b> Preface</a></li>
<li class="chapter" data-level="7.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>7.2</b> Introduction to R</a></li>
<li class="chapter" data-level="7.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>7.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>7.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="7.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>7.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>7.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>8</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="8.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>8.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="8.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>8.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>9</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="9.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>9.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="9.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>9.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="9.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>9.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>10</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="10.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>10.1</b> Bioconductor</a></li>
<li class="chapter" data-level="10.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>10.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="10.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>10.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>10.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="10.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>10.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="10.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>10.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>10.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>11</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>11.1</b> Vocabulary</a></li>
<li class="chapter" data-level="11.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>11.2</b> R functions</a></li>
<li class="chapter" data-level="11.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>11.3</b> Interacting with R</a></li>
<li class="chapter" data-level="11.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>11.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>11.4.1</b> Vectors</a></li>
<li class="chapter" data-level="11.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>11.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="11.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>11.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="11.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>11.4.4</b> Lists</a></li>
<li class="chapter" data-level="11.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>11.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>11.5</b> Arguments</a></li>
<li class="chapter" data-level="11.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>11.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="11.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>11.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="11.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>11.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="11.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>11.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="11.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>11.9</b> Quiting R</a></li>
<li class="chapter" data-level="11.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>11.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>12</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="12.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-1"><i class="fa fa-check"></i><b>12.1</b> Preface</a></li>
<li class="chapter" data-level="12.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>12.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="12.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>12.3</b> Introduction</a></li>
<li class="chapter" data-level="12.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>12.4</b> Vocabulary</a></li>
<li class="chapter" data-level="12.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions"><i class="fa fa-check"></i><b>12.5</b> Functions</a></li>
<li class="chapter" data-level="12.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>12.6</b> Preliminaries</a></li>
<li class="chapter" data-level="12.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>12.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="12.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>12.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="12.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>12.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="12.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>12.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="12.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>12.8.3</b> DNA words</a></li>
<li class="chapter" data-level="12.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>12.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="12.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>12.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="12.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>12.9.1</b> License</a></li>
<li class="chapter" data-level="12.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>12.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>13</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="13.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-2"><i class="fa fa-check"></i><b>13.1</b> Preface</a></li>
<li class="chapter" data-level="13.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>13.2</b> Vocab</a></li>
<li class="chapter" data-level="13.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-1"><i class="fa fa-check"></i><b>13.3</b> Functions</a></li>
<li class="chapter" data-level="13.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>13.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="13.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>13.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>14</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="14.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-3"><i class="fa fa-check"></i><b>14.1</b> Preface</a></li>
<li class="chapter" data-level="14.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>14.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="functions-2.html"><a href="functions-2.html"><i class="fa fa-check"></i><b>15</b> Functions</a>
<ul>
<li class="chapter" data-level="15.1" data-path="functions-2.html"><a href="functions-2.html#vectors-in-r"><i class="fa fa-check"></i><b>15.1</b> Vectors in R</a></li>
<li class="chapter" data-level="15.2" data-path="functions-2.html"><a href="functions-2.html#math-on-vectors"><i class="fa fa-check"></i><b>15.2</b> Math on vectors</a></li>
<li class="chapter" data-level="15.3" data-path="functions-2.html"><a href="functions-2.html#functions-on-vectors"><i class="fa fa-check"></i><b>15.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="15.4" data-path="functions-2.html"><a href="functions-2.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>15.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="15.5" data-path="functions-2.html"><a href="functions-2.html#subsetting-vectors"><i class="fa fa-check"></i><b>15.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="15.6" data-path="functions-2.html"><a href="functions-2.html#sequences-of-numbers"><i class="fa fa-check"></i><b>15.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="15.7" data-path="functions-2.html"><a href="functions-2.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>15.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="15.8" data-path="functions-2.html"><a href="functions-2.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>15.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>16</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="16.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-4"><i class="fa fa-check"></i><b>16.1</b> Preface</a></li>
<li class="chapter" data-level="16.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>16.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="16.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>16.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>17</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="17.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-5"><i class="fa fa-check"></i><b>17.1</b> Preface</a></li>
<li class="chapter" data-level="17.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>17.2</b> Vocab</a></li>
<li class="chapter" data-level="17.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-3"><i class="fa fa-check"></i><b>17.3</b> Functions</a></li>
<li class="chapter" data-level="17.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>17.4</b> Functions in R</a></li>
<li class="chapter" data-level="17.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>17.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="fasta-files.html"><a href="fasta-files.html"><i class="fa fa-check"></i><b>18</b> FASTA Files</a>
<ul>
<li class="chapter" data-level="18.1" data-path="fasta-files.html"><a href="fasta-files.html#example-fasta-file"><i class="fa fa-check"></i><b>18.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="18.2" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>18.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="18.3" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>18.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="18.4" data-path="fasta-files.html"><a href="fasta-files.html#fastq-format"><i class="fa fa-check"></i><b>18.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>19</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="19.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-4"><i class="fa fa-check"></i><b>19.0.1</b> Functions</a></li>
<li class="chapter" data-level="19.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>19.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="19.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>19.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="19.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>19.0.4</b> File types</a></li>
<li class="chapter" data-level="19.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>19.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="19.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>19.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="19.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>19.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="19.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>19.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="19.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>19.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="19.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>19.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="19.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>19.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="19.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>19.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="19.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>19.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="19.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>19.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="19.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>19.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>20</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="20.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-2"><i class="fa fa-check"></i><b>20.1</b> Preliminaries</a></li>
<li class="chapter" data-level="20.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>20.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="20.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>20.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="20.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>20.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="21" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>21</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="21.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>21.1</b> Preliminaries</a></li>
<li class="chapter" data-level="21.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>21.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>22</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="22.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>22.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="22.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>22.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="22.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>22.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="22.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>22.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="22.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>22.4</b> Inversions</a></li>
<li class="chapter" data-level="22.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>22.5</b> Translocations</a></li>
<li class="chapter" data-level="22.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>22.6</b> Random sequence</a></li>
<li class="chapter" data-level="22.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>22.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>23</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="23.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>23.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="23.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>23.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="23.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>23.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="23.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>23.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="23.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>23.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="23.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>23.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="23.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>23.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="23.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>23.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>24</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="24.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>24.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="24.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>24.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="24.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>24.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>25</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="25.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>25.1</b> Prelminaries</a></li>
<li class="chapter" data-level="25.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>25.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="25.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>25.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>26</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="26.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>26.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="26.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages"><i class="fa fa-check"></i><b>26.1.1</b> Packages</a></li>
<li class="chapter" data-level="26.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-5"><i class="fa fa-check"></i><b>26.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="26.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>26.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="26.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>26.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="26.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>26.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="26.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>26.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="26.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>26.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="26.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>26.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="26.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>26.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="26.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>26.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>27</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="27.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-8"><i class="fa fa-check"></i><b>27.1</b> Preliminaries</a></li>
<li class="chapter" data-level="27.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-1"><i class="fa fa-check"></i><b>27.2</b> Introduction</a></li>
<li class="chapter" data-level="27.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>27.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="27.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>27.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="28" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>28</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="28.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-9"><i class="fa fa-check"></i><b>28.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="28.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>28.1.1</b> Key functions</a></li>
<li class="chapter" data-level="28.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>28.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="28.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>28.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="28.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>28.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="28.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>28.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="28.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>28.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="28.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>28.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>29</b> A complete bioinformatics workflow in R</a>
<ul>
<li class="chapter" data-level="29.1" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>29.1</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="29.1.1" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#vocab-3"><i class="fa fa-check"></i><b>29.1.1</b> Vocab</a></li>
<li class="chapter" data-level="29.1.2" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>29.1.2</b> R functions</a></li>
<li class="chapter" data-level="29.1.3" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>29.1.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="29.1.4" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>29.1.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="29.2" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>29.2</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="29.3" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>29.3</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="29.4" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>29.4</b> Aligning sequences</a></li>
<li class="chapter" data-level="29.5" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>29.5</b> The shroom family of genes</a></li>
<li class="chapter" data-level="29.6" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>29.6</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="29.7" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>29.7</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="29.7.1" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>29.7.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="29.7.2" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>29.7.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="29.8" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>29.8</b> Genetic distance.</a></li>
<li class="chapter" data-level="29.9" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>29.9</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="29.9.1" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>29.9.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="29.10" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>29.10</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="29.10.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>29.10.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="29.10.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>29.10.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="29.10.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>29.10.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="29.10.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>29.10.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-2"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="29.11" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>29.11</b> Help!</a>
<ul>
<li class="chapter" data-level="29.11.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>29.11.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="29.11.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>29.11.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="29.11.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>29.11.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="29.11.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>29.11.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="29.12" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>29.12</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="29.12.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>29.12.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="29.12.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>29.12.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="29.13" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>29.13</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="multiple-sequence-alignment-in-r" class="section level1" number="26">
<h1><span class="header-section-number">Chapter 26</span> Multiple sequence alignment in R</h1>
<p><strong>By</strong>: Nathan Brouwer, with some content adapted Coghlan (2011) <a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter5.html">Multiple Alignment and Phylogenetic trees</a> and under the Creative Commons 3.0 Attribution License <a href="https://creativecommons.org/licenses/by/3.0/">(CC BY 3.0)</a>. Functions print_msa() and clean_alignment() adapted from (<a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/#">Coglan 2011</a>).</p>
<div id="preliminaries-7" class="section level2" number="26.1">
<h2><span class="header-section-number">26.1</span> Preliminaries</h2>
<!-- all needed? -->
<div id="packages" class="section level3" number="26.1.1">
<h3><span class="header-section-number">26.1.1</span> Packages</h3>
<p>We’ll be using the package <code>ggmsa</code> for the first time and you will have to install it using <code>install.packages("ggmsa")</code>. You may be asked to re-restart R more then once during the installation process.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiIjIG5ldyBwYWNrYWdlc1xuIyMgT25seSBpbnN0YWxsIG9uY2VcbiMgaW5zdGFsbC5wYWNrYWdlcyhcImdnbXNhXCIpXG5saWJyYXJ5KGdnbXNhKSAgICAgICAjIHZpc3VhbGl6ZSBNU0FcblxuIyBvdGhlciBwYWNrYWdlc1xubGlicmFyeShjb21wYmlvNGFsbClcbmxpYnJhcnkoQmlvc3RyaW5ncykgICMgY29udmVydCBGQVNUQSB0byBBQVN0cmluZ1NldFxubGlicmFyeShtc2EpICAgICAgICAgIyBtdWx0aXBsZSBzZXF1ZW5jZSBhbGlnbm1lbnQifQ==
</div>
</div>
<div id="functions-5" class="section level3" number="26.1.2">
<h3><span class="header-section-number">26.1.2</span> Functions</h3>
<p>The following key functions from compbio4all are used in this lesson</p>
<ul>
<li>fasta_cleaner()</li>
<li>entrez_fetch_list()</li>
<li>print_msa()</li>
<li>clean_alignment()</li>
</ul>
</div>
</div>
<div id="multiple-sequence-alignment-msa" class="section level2" number="26.2">
<h2><span class="header-section-number">26.2</span> Multiple sequence alignment (MSA)</h2>
<!-- This paragraph adapated from Coghlan -->
<p>A common task in bioinformatics is to download a set of related sequences from a database, and then to align those sequences using multiple alignment software. This is the first step in almost all phylogenetic analyses using sequence data.
<!-- This paragraph adapated from Coghlan --></p>
</div>
<div id="make-msa-with-msa" class="section level2" number="26.3">
<h2><span class="header-section-number">26.3</span> Make MSA with msa()</h2>
<p>WE’ll use a package called <code>msa</code> (<a href="https://academic.oup.com/bioinformatics/article/31/24/3997/197486">Bodenhofer et al. 2015</a>). There are several packages that can do multiple sequence alignment in R, but they all require loading an external piece of alignment software that is just accessed via R. The <code>msa</code> package actually runs the alignment algorithms entirely in R, making workflows simpler.</p>
<div id="data-preparation" class="section level3" number="26.3.1">
<h3><span class="header-section-number">26.3.1</span> Data preparation</h3>
The data is stored in object in combio4all
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJkYXRhKHNlcV8xXzJfM180KSAifQ==
</div>
This data is in the form of a list
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJpcy5saXN0KHNlcV8xXzJfM180KSJ9
</div>
<p>First we need to clean each one using <code>fasta_cleaner()</code>.</p>
<!-- Update this so parse = F and we can skip the paste() step -->
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJzZXFfMV8yXzNfNFtbMV1dIDwtIGZhc3RhX2NsZWFuZXIoc2VxXzFfMl8zXzRbWzFdXSlcbnNlcV8xXzJfM180W1syXV0gPC0gZmFzdGFfY2xlYW5lcihzZXFfMV8yXzNfNFtbMl1dKVxuc2VxXzFfMl8zXzRbWzNdXSA8LSBmYXN0YV9jbGVhbmVyKHNlcV8xXzJfM180W1szXV0pXG5zZXFfMV8yXzNfNFtbNF1dIDwtIGZhc3RhX2NsZWFuZXIoc2VxXzFfMl8zXzRbWzRdXSkifQ==
</div>
<!-- Skipping some steps here e.g. why I'm using paste(), what a named vector is -->
<!-- should write wrapper for this -->
<p>We need to put it into the form of a vector; in particular a <strong>named vector</strong>.</p>
<!-- Make function to do this -->
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJzZXFfMV8yXzNfNF92ZWN0b3IgPC0gYyhQMDY3NDcgPSBwYXN0ZShzZXFfMV8yXzNfNFtbMV1dLCBjb2xsYXBzZSA9IFwiXCIpLFxuICAgICAgICAgICAgICAgICBQMEM1NjkgPSBwYXN0ZShzZXFfMV8yXzNfNFtbMl1dLCBjb2xsYXBzZSA9IFwiXCIpLFxuICAgICAgICAgICAgICAgICBPNTY3NzMgPSBwYXN0ZShzZXFfMV8yXzNfNFtbM11dLCBjb2xsYXBzZSA9IFwiXCIpLFxuICAgICAgICAgICAgICAgICBRNVZLUDEgPSBwYXN0ZShzZXFfMV8yXzNfNFtbNF1dLCBjb2xsYXBzZSA9IFwiXCIpKSJ9
</div>
<p>We’ll need need to convert our set of sequences to a particular format in preparation for alignment. This is done with the <code>AAStringSet()</code> function from <code>Biostrings</code>.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJzZXFfMV8yXzNfNF9zdHJpbmdzZXQgPC0gQmlvc3RyaW5nczo6QUFTdHJpbmdTZXQoc2VxXzFfMl8zXzRfdmVjdG9yKSJ9
</div>
<p>This just puts things in a format that makes the software happy. Doing this is a theme of bioinformatics work!</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJzZXFfMV8yXzNfNF9zdHJpbmdzZXQifQ==
</div>
<p>Next, we can run the alignment algorithm with the <code>msa()</code> function. There are many algorithms and pieces software for building alignments. The <code>msa</code> packages implements three major ones:</p>
<ol style="list-style-type: decimal">
<li>ClustalW</li>
<li>ClustalOmega</li>
<li>Muscle</li>
</ol>
We’ll use ClustalW. Depending on the size and number of sequences this may take a little bit of time.<br />

<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJ2aXJ1c2FsbiA8LSBtc2EoaW5wdXRTZXFzID0gc2VxXzFfMl8zXzRfc3RyaW5nc2V0LFxuICAgICAgICAgICAgICAgICAgICAgbWV0aG9kID0gXCJDbHVzdGFsV1wiKSJ9
</div>
We can view a snapshot of the alignment.
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJ2aXJ1c2FsbiJ9
</div>
<p>Each sequence is on a line. The <strong>consensus sequence</strong> indicates something similar to the average of all the sequences and is on the bottom and labeled <code>Con</code>. Question marks indicate that the software could not determine a consensus. Dashes indicate either <strong>indels</strong> (<strong>insertions</strong> or <strong>deletions</strong>), or are added at the begining and end of sequences of unequal length so that they line up.</p>
<p>The output from <code>msa()</code> is a particular class of R object, a <code>MsaAAMultipleAlignment</code>.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJpcyh2aXJ1c2FsbilcbmNsYXNzKHZpcnVzYWxuKSJ9
</div>
<p>Next we’ll want to visualize our alignment. In order to do further work with the MSA we’re going to - I bet you can guess what happens next - make a conversion to the object. In this case we’re going to make a subtle change by calling up the <code>class()</code> of the alignment and changing it from <code>MsaAAMultipleAlignment</code> (with “Msa” at the beginning) to <code>AAMultipleAlignment</code> (no “Msa”). (This is an annoying step and is needed because the folks who wrote the <code>msa</code> package have yet to collaobrate with the folks who wrote another package we’ll use in a little bit).</p>
Don’t worry if you don’t understand what’s going on here - just run the code.
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJjbGFzcyh2aXJ1c2FsbikgPC0gXCJBQU11bHRpcGxlQWxpZ25tZW50XCIifQ==
</div>
</div>
</div>
<div id="viewing-your-msa" class="section level2" number="26.4">
<h2><span class="header-section-number">26.4</span> Viewing your MSA</h2>
<p>There are several ways to view and explore your MSA</p>
<ol style="list-style-type: decimal">
<li>Within the R console using <code>compbio4all::print_msa()</code></li>
<li>As an R plot using <code>ggmsa::ggmas()</code></li>
<li>OPTIONAL: As a PDF using <code>msa::msaPrettyPrint()</code></li>
</ol>
<div id="viewing-a-long-multiple-alignment-in-the-r-console." class="section level3" number="26.4.1">
<h3><span class="header-section-number">26.4.1</span> Viewing a long multiple alignment in the R console.</h3>
<p>If you want to view a long multiple alignment within the R console, it is convenient to view the multiple alignment in blocks.</p>
<p>The function <code>print_msa()</code> (<a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/#">Coglan 2011</a>)below will do this for you. As its inputs, the function <code>print_msa()</code> takes the two things</p>
<ol style="list-style-type: decimal">
<li><code>alignment</code>: input alignment</li>
<li><code>chunksize</code>: the number of columns to print out in each block.</li>
</ol>
To use <code>print_msa()</code> we first need to do a little format conversion:
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJ2aXJ1c2Fsbl9zZXFpbnIgPC0gbXNhQ29udmVydCh2aXJ1c2FsbiwgdHlwZSA9IFwic2VxaW5yOjphbGlnbm1lbnRcIikifQ==
</div>
Then we can print it out like this, making the alignment 60 bases wide:
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJwcmludF9tc2EoYWxpZ25tZW50ID0gdmlydXNhbG5fc2VxaW5yLCBcbiAgICAgICAgICBjaHVua3NpemUgPSA2MCkifQ==
</div>
<!-- TODO: do auto converion of format, print consensus -->
</div>
<div id="visualizing-alignments-as-an-r-plot" class="section level3" number="26.4.2">
<h3><span class="header-section-number">26.4.2</span> Visualizing alignments as an R plot</h3>
<p>A powerful tool for visualizing focal parts of an alignment is <code>ggmsa</code>. If you haven’t already, download it with <code>install.packages("ggmsa")</code> and load it with <code>library(ggmsa)</code>.</p>
<p><code>ggmsa</code> prints a sequence alignment out within RStudio. Alignments can be large, so its important to select a subset of the alignment for visualzation.</p>
First, let’s look at the first 20 bases of our alignment. Note that we are using <code>virusaln</code>, NOT <code>virusaln_seqinr</code> (sorry for the back and forth between objects.)
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJnZ21zYSh2aXJ1c2FsbiwgICAjIHZpcnVzYWxuLCBOT1QgdmlydXNhbG5fc2VxaW5yXG4gICAgICBzdGFydCA9IDEsIFxuICAgICAgZW5kID0gMjApICJ9
</div>
<div id="optional-file-types-used-by-ggmsa" class="section level4" number="26.4.2.1">
<h4><span class="header-section-number">26.4.2.1</span> OPTIONAL: File types used by <code>ggmsa</code></h4>
<p>The <code>ggmsa</code> packages currently only works with certain types of alignment output. We can see what these are with <code>available_msa()</code>.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJhdmFpbGFibGVfbXNhKCkifQ==
</div>
<p>As you can see there are a number of ways multiple sequence alignments can be represented in <em>R</em>. This has to do with the facts that i) There are many pieces of software / algorithms for making MSAs, and many bioinformatics packages that work with them.,</p>
<p>You can see that <code>AAMultipleAlignment</code> is listed, which the the format we set previously using the <code>class()</code> command.</p>
<p>The <code>msa</code> packages has a function <code>msaConvert()</code> which can change formats between different ways of representing MSAs which may be useful.</p>
</div>
</div>
<div id="optional-print-msa-to-pdf" class="section level3" number="26.4.3">
<h3><span class="header-section-number">26.4.3</span> OPTIONAL: Print MSA to PDF</h3>
<p>The <code>msa</code> package has a fabulous function, <code>msaPrettyPrint()</code> for rendering an MSA to PDF. It can take a little bit to run, and in order to vioew the PDF you need to locate the output. (Again, we’ll use <code>virusaln</code>, not <code>virusaln_seqinr</code>).</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJtc2FQcmV0dHlQcmludCh2aXJ1c2FsbiwgICAgICMgdmlydXNhbG4sIE5PVCB2aXJ1c2Fsbl9zZXFpbnJcbiAgICAgICAgICAgICAgIGZpbGUgPSBcIm15X21zYS5wZGZcIixcbiAgICAgICAgICAgICAgIGFza0Zvck92ZXJ3cml0ZSA9IEYpIn0=
</div>
<p>On a Mac usually searching in Finder will locate the file even after is just created. You can ask R where it is saving thing using <code>getwd()</code>.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJnZXR3ZCgpIn0=
</div>
<p>You can change where R is saving things using the RStudio menu by clicking on Session -&gt; Set Working Directory -&gt; Choose directory…</p>
</div>
</div>
<div id="discarding-very-poorly-conserved-regions-from-an-alignment" class="section level2" number="26.5">
<h2><span class="header-section-number">26.5</span> Discarding very poorly conserved regions from an alignment</h2>
<!-- By Coglan -->
<p>It is often a good idea to discard very <strong>poorly conserved</strong> regions from a multiple sequence alignment before visualizing it or building a phylogenetic tree, as the very poorly conserved regions are likely to be regions that are either <strong>non-homologous</strong> between the sequences being considered (and so do not have any phylogenetic signal), or are homologous but are so <strong>diverged</strong> that they are very difficult to align accurately (and so may add noise to the phylogenetic analysis, and decrease the accuracy of the inferred tree).
<!-- By Coglan --></p>
<!-- By Coglan -->
<p>To discard very poorly conserved regions from a multiple alignment, you can use the following R function, <code>clean_alignment()</code> ((<a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/#">Coglan 2011</a>))
<!-- By Coglan --></p>
<!-- By Coglan -->
<p>The function <code>clean_alignment()</code> takes three arguments (inputs):</p>
<ol style="list-style-type: decimal">
<li>the input alignment;</li>
<li>minpcnongap: the minimum percent of letters in an alignment column that must be non-gap characters for the column to be kept; and</li>
<li>minpcid: the minimum percent of pairs of letters in an alignment column that must be identical for the column to be kept.
<!-- By Coglan --></li>
</ol>
<!-- By Coglan, with some formatting by me -->
<p>For example, if we have a single column (<strong>locus</strong>) with letters “T”, “A”, “T”, “-” (in four sequences), then 75% of the letters are non-gap characters; and the pairs of letters between the three non-gap sequences are</p>
<ul>
<li>1 versus 2: “T,A”,</li>
<li>1 versus 3: “T,T”,</li>
<li>2 versus 3: “A,T”,</li>
</ul>
<p>Therefore 33% of the pairs of letters are identical (<strong>PID</strong>) for that position in the alignment.
<!-- By Coglan --></p>
<!-- By Coglan -->
<p>If you look at the multiple alignment for the virus phosphoprotein sequences (which we printed out using function <code>print_msa()</code>, see above), we can see that the last few columns are poorly aligned (contain many gaps and mismatches), and probably add noise to the phylogenetic analysis.
<!-- By Coglan --></p>
<p>Let’s cleave off anything with less than 30% non-gap and less than 30% PID.
NOTE: we’re bac to using <code>virusaln_seqinr</code>, not <code>virusaln</code>.</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJ2aXJ1c2Fsbl9zZXFpbnJfY2xlYW4gPC0gY2xlYW5fYWxpZ25tZW50KGFsaWdubWVudCA9IHZpcnVzYWxuX3NlcWluciwgICMgdmlydXNhbG5fc2VxaW5yXG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIG1pbnBjbm9uZ2FwID0gMzAsIFxuICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICBtaW5wY2lkID0gMzApIn0=
</div>
<!-- By Coglan -->
<p>In this case, we required that at least 30% of letters in a column are not gap characters for that column to be kept, and that at least 30% of pairs of letters in an alignment column must be identical for the column to be kept.
<!-- By Coglan --></p>
<p>We can print out the filtered alignment by typing:</p>
<div data-datacamp-exercise="" data-height="300" data-encoded="true">
eyJsYW5ndWFnZSI6InIiLCJzYW1wbGUiOiJwcmludF9tc2EodmlydXNhbG5fc2VxaW5yX2NsZWFuKSJ9
</div>
<!-- By Coglan -->
<p>The filtered alignment is shorter, and is missing some of the poorly conserved regions of the original alignment.
<!-- By Coglan --></p>
<!-- By Coglan -->
<p>Note that it is not a good idea to filter out too much of your alignment, as if you are left with few columns in your filtered alignment, you will be basing your phylogenetic tree upon a very short alignment (little data), and so the tree may be unreliable. Therefore, you need to achieve a balance between discarding the dodgy (poorly aligned) parts of your alignment, and retaining enough columns of the alignment that you will have enough data to based your tree upon.
<!-- By Coglan --></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="retrieving-multiple-sequences-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="calculating-genetic-distances-between-sequences.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/19-MSA_in_R_.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/19-MSA_in_R_.Rmd",
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
