<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 or loops in R - copy? alt? | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 or loops in R - copy? alt? | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 or loops in R - copy? alt? | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="working-with-square-brackets.html"/>
<link rel="next" href="function-practice.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.5.1/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>1</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="1.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>1.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="1.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>1.3</b> Introduction</a></li>
<li class="chapter" data-level="1.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary"><i class="fa fa-check"></i><b>1.4</b> Vocabulary</a></li>
<li class="chapter" data-level="1.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions"><i class="fa fa-check"></i><b>1.5</b> Functions</a></li>
<li class="chapter" data-level="1.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>1.6</b> Preliminaries</a></li>
<li class="chapter" data-level="1.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>1.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="1.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>1.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="1.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>1.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="1.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>1.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="1.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>1.8.3</b> DNA words</a></li>
<li class="chapter" data-level="1.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>1.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="1.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>1.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="1.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>1.9.1</b> License</a></li>
<li class="chapter" data-level="1.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>1.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="working-with-vectors-and-matrices.html"><a href="working-with-vectors-and-matrices.html"><i class="fa fa-check"></i><b>2</b> Working with vectors and matrices</a></li>
<li class="chapter" data-level="3" data-path="a-note-on-changes-to-database-entries.html"><a href="a-note-on-changes-to-database-entries.html"><i class="fa fa-check"></i><b>3</b> A note on changes to database entries</a></li>
<li class="chapter" data-level="4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html"><i class="fa fa-check"></i><b>4</b> Writing functions - practice problems</a>
<ul>
<li class="chapter" data-level="4.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#introduction-1"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#part-1-pure-practice"><i class="fa fa-check"></i><b>4.2</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-1-hello-world"><i class="fa fa-check"></i><b>4.2.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="4.2.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______"><i class="fa fa-check"></i><b>4.2.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function"><i class="fa fa-check"></i><b>4.3</b> Practice Function 3: Make you own natural log function</a></li>
<li class="chapter" data-level="4.4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-4-make-function-that-converts-dna-directly-to-rna"><i class="fa fa-check"></i><b>4.4</b> Practice Function 4: Make function that converts DNA directly to RNA</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html"><i class="fa fa-check"></i><b>5</b> Computational Biology Portfolio: Practice Working with Strings</a>
<ul>
<li class="chapter" data-level="5.1" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#introduction-2"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#purpose"><i class="fa fa-check"></i><b>5.2</b> Purpose</a></li>
<li class="chapter" data-level="5.3" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#gonder-et-al-2007-data"><i class="fa fa-check"></i><b>5.3</b> Gonder et al 2007 data</a></li>
<li class="chapter" data-level="5.4" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#gonder-et-al-2007-new-sequences"><i class="fa fa-check"></i><b>5.4</b> Gonder et al 2007 new sequences</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#creating-the-codes"><i class="fa fa-check"></i><b>5.4.1</b> Creating the codes</a></li>
<li class="chapter" data-level="5.4.2" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#selecting-a-code-for-demonstration"><i class="fa fa-check"></i><b>5.4.2</b> Selecting a code for demonstration</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#searching-genbank"><i class="fa fa-check"></i><b>5.5</b> Searching Genbank</a></li>
<li class="chapter" data-level="5.6" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#gonder-et-al-2007-previously-published-sequences"><i class="fa fa-check"></i><b>5.6</b> Gonder et al 2007 previously published sequences</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#creating-the-codes-1"><i class="fa fa-check"></i><b>5.6.1</b> Creating the codes</a></li>
<li class="chapter" data-level="5.6.2" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#selecting-a-code-for-demonstration-1"><i class="fa fa-check"></i><b>5.6.2</b> Selecting a code for demonstration</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#searching-genbank-1"><i class="fa fa-check"></i><b>5.7</b> Searching Genbank</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html"><i class="fa fa-check"></i><b>6</b> Writing user-friendly functions</a>
<ul>
<li class="chapter" data-level="6.1" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-1"><i class="fa fa-check"></i><b>6.1</b> Version 1</a></li>
<li class="chapter" data-level="6.2" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-2"><i class="fa fa-check"></i><b>6.2</b> Version 2</a></li>
<li class="chapter" data-level="6.3" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-3"><i class="fa fa-check"></i><b>6.3</b> Version 3</a></li>
<li class="chapter" data-level="6.4" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-4-adding-conditions-and-warnigns"><i class="fa fa-check"></i><b>6.4</b> Version 4: Adding conditions and warnigns</a></li>
<li class="chapter" data-level="6.5" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-5"><i class="fa fa-check"></i><b>6.5</b> Version 5</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html"><i class="fa fa-check"></i><b>7</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="7.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#commands-used"><i class="fa fa-check"></i><b>7.1</b> Commands used</a></li>
<li class="chapter" data-level="7.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#r-objects"><i class="fa fa-check"></i><b>7.2</b> R Objects</a></li>
<li class="chapter" data-level="7.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>7.3</b> Differences between objects</a></li>
<li class="chapter" data-level="7.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-data"><i class="fa fa-check"></i><b>7.4</b> The Data</a></li>
<li class="chapter" data-level="7.5" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>7.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#is"><i class="fa fa-check"></i><b>7.5.1</b> is()</a></li>
<li class="chapter" data-level="7.5.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#length"><i class="fa fa-check"></i><b>7.5.2</b> length()</a></li>
<li class="chapter" data-level="7.5.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#str"><i class="fa fa-check"></i><b>7.5.3</b> str()</a></li>
<li class="chapter" data-level="7.5.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#c"><i class="fa fa-check"></i><b>7.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#you-can-save-statistical-output-as-an-object"><i class="fa fa-check"></i><b>7.6</b> You can save statistical output as an object</a></li>
<li class="chapter" data-level="7.7" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#how-does-r-save-statistical-output"><i class="fa fa-check"></i><b>7.7</b> How does R save statistical output</a></li>
<li class="chapter" data-level="7.8" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#lists-in-r"><i class="fa fa-check"></i><b>7.8</b> Lists in R</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="an-aside-about-t-tests.html"><a href="an-aside-about-t-tests.html"><i class="fa fa-check"></i><b>8</b> An aside about t-tests</a></li>
<li class="chapter" data-level="9" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html"><i class="fa fa-check"></i><b>9</b> Working with square brackets</a>
<ul>
<li class="chapter" data-level="9.1" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#square-brackets-for-real"><i class="fa fa-check"></i><b>9.1</b> Square brackets for real</a></li>
<li class="chapter" data-level="9.2" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#difference-between-means-using-square-brackets"><i class="fa fa-check"></i><b>9.2</b> Difference between means using square brackets</a></li>
<li class="chapter" data-level="9.3" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#upshot-plotting-output-of-an-r-statistical-test"><i class="fa fa-check"></i><b>9.3</b> Upshot: plotting output of an R statistical test</a></li>
<li class="chapter" data-level="9.4" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#the-not-bad-way-typing-directly-into-r"><i class="fa fa-check"></i><b>9.4</b> The “not bad way”: typing directly into R</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#not-bad-way-with-a-bar-plot"><i class="fa fa-check"></i><b>9.4.1</b> “NOt bad way” with a bar plot</a></li>
<li class="chapter" data-level="9.4.2" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#not-bad-way-with-using-a-single-dot"><i class="fa fa-check"></i><b>9.4.2</b> “NOt bad way” with using a single dot</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#the-best-way"><i class="fa fa-check"></i><b>9.5</b> The best way</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#a-final-fancy-touch"><i class="fa fa-check"></i><b>9.5.1</b> A final fancy touch</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#debrief"><i class="fa fa-check"></i><b>9.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html"><i class="fa fa-check"></i><b>10</b> or loops in R - copy? alt?</a>
<ul>
<li class="chapter" data-level="10.1" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#introduction-3"><i class="fa fa-check"></i><b>10.1</b> Introduction</a></li>
<li class="chapter" data-level="10.2" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#reminders"><i class="fa fa-check"></i><b>10.2</b> Reminders</a></li>
<li class="chapter" data-level="10.3" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop"><i class="fa fa-check"></i><b>10.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="10.4" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-3-the-classic-for-loop"><i class="fa fa-check"></i><b>10.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="10.5" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-3-the-power-user-for-loop"><i class="fa fa-check"></i><b>10.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="10.6" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-4-a-for-loop-for-all-purposes"><i class="fa fa-check"></i><b>10.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="10.7" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#advanced-aside-if-else-and-next-statements"><i class="fa fa-check"></i><b>10.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="10.8" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#on-avoiding-for-loops"><i class="fa fa-check"></i><b>10.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="10.9" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#challenge"><i class="fa fa-check"></i><b>10.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="function-practice.html"><a href="function-practice.html"><i class="fa fa-check"></i><b>11</b> Function practice</a>
<ul>
<li class="chapter" data-level="11.1" data-path="function-practice.html"><a href="function-practice.html#part-1-pure-practice-1"><i class="fa fa-check"></i><b>11.1</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="function-practice.html"><a href="function-practice.html#practice-function-1-hello-world-1"><i class="fa fa-check"></i><b>11.1.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="11.1.2" data-path="function-practice.html"><a href="function-practice.html#practice-function-2-hello-_______-1"><i class="fa fa-check"></i><b>11.1.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="function-practice.html"><a href="function-practice.html#practice-function-3-make-you-own-natural-log-function-1"><i class="fa fa-check"></i><b>11.2</b> Practice Function 3: Make you own natural log function</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html"><i class="fa fa-check"></i><b>12</b> for() loops in R</a>
<ul>
<li class="chapter" data-level="12.1" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#introduction-4"><i class="fa fa-check"></i><b>12.1</b> Introduction</a></li>
<li class="chapter" data-level="12.2" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#reminders-1"><i class="fa fa-check"></i><b>12.2</b> Reminders</a></li>
<li class="chapter" data-level="12.3" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop-1"><i class="fa fa-check"></i><b>12.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="12.4" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-classic-for-loop-1"><i class="fa fa-check"></i><b>12.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="12.5" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-power-user-for-loop-1"><i class="fa fa-check"></i><b>12.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="12.6" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-4-a-for-loop-for-all-purposes-1"><i class="fa fa-check"></i><b>12.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="12.7" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#advanced-aside-if-else-and-next-statements-1"><i class="fa fa-check"></i><b>12.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="12.8" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#on-avoiding-for-loops-1"><i class="fa fa-check"></i><b>12.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="12.9" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#challenge-1"><i class="fa fa-check"></i><b>12.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="random-number-generation.html"><a href="random-number-generation.html"><i class="fa fa-check"></i><b>13</b> Random number generation</a>
<ul>
<li class="chapter" data-level="13.1" data-path="random-number-generation.html"><a href="random-number-generation.html#introduction-5"><i class="fa fa-check"></i><b>13.1</b> Introduction</a></li>
<li class="chapter" data-level="13.2" data-path="random-number-generation.html"><a href="random-number-generation.html#using-set.seed-to-make-simulations-reproducible"><i class="fa fa-check"></i><b>13.2</b> Using set.seed() to make simulations reproducible</a></li>
<li class="chapter" data-level="13.3" data-path="random-number-generation.html"><a href="random-number-generation.html#types-of-random-number-generators"><i class="fa fa-check"></i><b>13.3</b> Types of random number generators</a></li>
<li class="chapter" data-level="13.4" data-path="random-number-generation.html"><a href="random-number-generation.html#reporting-simulations"><i class="fa fa-check"></i><b>13.4</b> Reporting simulations</a></li>
<li class="chapter" data-level="13.5" data-path="random-number-generation.html"><a href="random-number-generation.html#notes"><i class="fa fa-check"></i><b>13.5</b> Notes:</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html"><i class="fa fa-check"></i><b>14</b> Sample test question-Randomization</a>
<ul>
<li class="chapter" data-level="14.1" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#sample-test-question"><i class="fa fa-check"></i><b>14.1</b> Sample test question:</a></li>
<li class="chapter" data-level="14.2" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#your-answer"><i class="fa fa-check"></i><b>14.2</b> Your answer:</a></li>
<li class="chapter" data-level="14.3" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-0"><i class="fa fa-check"></i><b>14.3</b> Hint 0</a></li>
<li class="chapter" data-level="14.4" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-1"><i class="fa fa-check"></i><b>14.4</b> Hint 1:</a></li>
<li class="chapter" data-level="14.5" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-2-follow-up-to-hint-0"><i class="fa fa-check"></i><b>14.5</b> Hint 2 (follow up to Hint 0)</a></li>
<li class="chapter" data-level="14.6" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-3"><i class="fa fa-check"></i><b>14.6</b> Hint 3:</a></li>
<li class="chapter" data-level="14.7" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#answer"><i class="fa fa-check"></i><b>14.7</b> Answer</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html"><i class="fa fa-check"></i><b>15</b> Simulating random amino acid sequences</a>
<ul>
<li class="chapter" data-level="15.1" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#selecting-random-letters-from-a-vector"><i class="fa fa-check"></i><b>15.1</b> Selecting random letters from a vector</a></li>
<li class="chapter" data-level="15.2" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-letters-from-the-whole-alphabet"><i class="fa fa-check"></i><b>15.2</b> Select random letters from the whole alphabet</a></li>
<li class="chapter" data-level="15.3" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-amino-acids-to-build-polypeptide"><i class="fa fa-check"></i><b>15.3</b> Select random amino acids to build polypeptide</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>16</b> How to review this book</a>
<ul>
<li class="chapter" data-level="16.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>16.1</b> File format</a></li>
<li class="chapter" data-level="16.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>16.2</b> What can you do?</a></li>
<li class="chapter" data-level="16.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>16.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>16.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="16.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>16.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="16.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>16.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="16.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>16.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="test-audio-link.html"><a href="test-audio-link.html"><i class="fa fa-check"></i><b>17</b> Test audio link</a></li>
<li class="chapter" data-level="18" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>18</b> Test shiny app</a></li>
<li class="chapter" data-level="19" data-path="test-youtube-vidoe.html"><a href="test-youtube-vidoe.html"><i class="fa fa-check"></i><b>19</b> Test YouTube vidoe</a></li>
<li class="chapter" data-level="20" data-path="test-google-doc-insertion.html"><a href="test-google-doc-insertion.html"><i class="fa fa-check"></i><b>20</b> Test google doc insertion</a></li>
<li class="chapter" data-level="21" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>21</b> Datacamp test</a></li>
<li class="chapter" data-level="22" data-path="an-example-of-a-plotly-graph.html"><a href="an-example-of-a-plotly-graph.html"><i class="fa fa-check"></i><b>22</b> An example of a plotly graph</a></li>
<li class="chapter" data-level="23" data-path="directly-embedding-of-a-libretext-page.html"><a href="directly-embedding-of-a-libretext-page.html"><i class="fa fa-check"></i><b>23</b> Directly embedding of a libretext page</a></li>
<li class="chapter" data-level="24" data-path="libretext-image.html"><a href="libretext-image.html"><i class="fa fa-check"></i><b>24</b> Libretext image</a>
<ul>
<li class="chapter" data-level="24.1" data-path="libretext-image.html"><a href="libretext-image.html#prelude-to-population-and-community-ecology"><i class="fa fa-check"></i><b>24.1</b> 45.0: Prelude to Population and Community Ecology</a></li>
<li class="chapter" data-level="24.2" data-path="libretext-image.html"><a href="libretext-image.html#note"><i class="fa fa-check"></i><b>24.2</b> Note</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>25</b> Downloading R</a>
<ul>
<li class="chapter" data-level="25.1" data-path="downloading-r.html"><a href="downloading-r.html#preface-1"><i class="fa fa-check"></i><b>25.1</b> Preface</a></li>
<li class="chapter" data-level="25.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>25.2</b> Introduction to R</a></li>
<li class="chapter" data-level="25.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>25.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="25.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>25.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="25.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>25.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="25.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>25.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>26</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="26.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>26.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="26.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>26.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>27</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="27.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>27.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="27.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>27.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="27.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>27.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="28" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>28</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="28.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>28.1</b> Bioconductor</a></li>
<li class="chapter" data-level="28.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>28.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="28.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>28.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="28.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>28.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="28.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>28.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="28.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>28.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="28.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>28.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>29</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="29.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary-1"><i class="fa fa-check"></i><b>29.1</b> Vocabulary</a></li>
<li class="chapter" data-level="29.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>29.2</b> R functions</a></li>
<li class="chapter" data-level="29.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>29.3</b> Interacting with R</a></li>
<li class="chapter" data-level="29.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>29.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="29.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>29.4.1</b> Vectors</a></li>
<li class="chapter" data-level="29.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>29.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="29.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>29.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="29.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>29.4.4</b> Lists</a></li>
<li class="chapter" data-level="29.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>29.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="29.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>29.5</b> Arguments</a></li>
<li class="chapter" data-level="29.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>29.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="29.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>29.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="29.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>29.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="29.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>29.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="29.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>29.9</b> Quiting R</a></li>
<li class="chapter" data-level="29.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>29.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>30</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="30.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-2"><i class="fa fa-check"></i><b>30.1</b> Preface</a></li>
<li class="chapter" data-level="30.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>30.2</b> Vocab</a></li>
<li class="chapter" data-level="30.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-1"><i class="fa fa-check"></i><b>30.3</b> Functions</a></li>
<li class="chapter" data-level="30.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>30.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="30.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>30.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="31" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>31</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="31.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-3"><i class="fa fa-check"></i><b>31.1</b> Preface</a></li>
<li class="chapter" data-level="31.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>31.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="32" data-path="functions-2.html"><a href="functions-2.html"><i class="fa fa-check"></i><b>32</b> Functions</a>
<ul>
<li class="chapter" data-level="32.1" data-path="functions-2.html"><a href="functions-2.html#vectors-in-r"><i class="fa fa-check"></i><b>32.1</b> Vectors in R</a></li>
<li class="chapter" data-level="32.2" data-path="functions-2.html"><a href="functions-2.html#math-on-vectors"><i class="fa fa-check"></i><b>32.2</b> Math on vectors</a></li>
<li class="chapter" data-level="32.3" data-path="functions-2.html"><a href="functions-2.html#functions-on-vectors"><i class="fa fa-check"></i><b>32.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="32.4" data-path="functions-2.html"><a href="functions-2.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>32.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="32.5" data-path="functions-2.html"><a href="functions-2.html#subsetting-vectors"><i class="fa fa-check"></i><b>32.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="32.6" data-path="functions-2.html"><a href="functions-2.html#sequences-of-numbers"><i class="fa fa-check"></i><b>32.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="32.7" data-path="functions-2.html"><a href="functions-2.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>32.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="32.8" data-path="functions-2.html"><a href="functions-2.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>32.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="33" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>33</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="33.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-4"><i class="fa fa-check"></i><b>33.1</b> Preface</a></li>
<li class="chapter" data-level="33.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>33.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="33.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>33.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="34" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>34</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="34.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-5"><i class="fa fa-check"></i><b>34.1</b> Preface</a></li>
<li class="chapter" data-level="34.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>34.2</b> Vocab</a></li>
<li class="chapter" data-level="34.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-3"><i class="fa fa-check"></i><b>34.3</b> Functions</a></li>
<li class="chapter" data-level="34.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>34.4</b> Functions in R</a></li>
<li class="chapter" data-level="34.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>34.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="fasta-files.html"><a href="fasta-files.html"><i class="fa fa-check"></i><b>35</b> FASTA Files</a>
<ul>
<li class="chapter" data-level="35.1" data-path="fasta-files.html"><a href="fasta-files.html#example-fasta-file"><i class="fa fa-check"></i><b>35.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="35.2" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>35.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="35.3" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>35.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="35.4" data-path="fasta-files.html"><a href="fasta-files.html#fastq-format"><i class="fa fa-check"></i><b>35.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>36</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="36.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-4"><i class="fa fa-check"></i><b>36.0.1</b> Functions</a></li>
<li class="chapter" data-level="36.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>36.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="36.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>36.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="36.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>36.0.4</b> File types</a></li>
<li class="chapter" data-level="36.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>36.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="36.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>36.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="36.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>36.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="36.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>36.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="36.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>36.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="36.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>36.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="36.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>36.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="36.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>36.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="36.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>36.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="36.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>36.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="36.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>36.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>37</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="37.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-2"><i class="fa fa-check"></i><b>37.1</b> Preliminaries</a></li>
<li class="chapter" data-level="37.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>37.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="37.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>37.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="37.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>37.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="38" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>38</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="38.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>38.1</b> Preliminaries</a></li>
<li class="chapter" data-level="38.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>38.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>39</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="39.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>39.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="39.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>39.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="39.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>39.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="39.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>39.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="39.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>39.4</b> Inversions</a></li>
<li class="chapter" data-level="39.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>39.5</b> Translocations</a></li>
<li class="chapter" data-level="39.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>39.6</b> Random sequence</a></li>
<li class="chapter" data-level="39.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>39.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="40" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>40</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="40.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>40.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="40.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>40.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="40.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>40.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="40.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>40.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="40.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>40.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="40.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>40.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="40.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>40.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="40.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>40.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="41" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>41</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="41.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>41.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="41.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>41.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="41.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>41.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="42" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html"><i class="fa fa-check"></i><b>42</b> Alignment by eye in R</a>
<ul>
<li class="chapter" data-level="42.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>42.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="42.1.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#packages"><i class="fa fa-check"></i><b>42.1.1</b> Packages</a></li>
<li class="chapter" data-level="42.1.2" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#data"><i class="fa fa-check"></i><b>42.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="43" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html"><i class="fa fa-check"></i><b>43</b> Matrix vocabulary</a>
<ul>
<li class="chapter" data-level="43.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#introduction-6"><i class="fa fa-check"></i><b>43.1</b> Introduction</a></li>
<li class="chapter" data-level="43.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#preliminaries-8"><i class="fa fa-check"></i><b>43.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="43.2.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#packages-1"><i class="fa fa-check"></i><b>43.2.1</b> Packages</a></li>
<li class="chapter" data-level="43.2.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#vocab-3"><i class="fa fa-check"></i><b>43.2.2</b> Vocab</a></li>
<li class="chapter" data-level="43.2.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#functions-used"><i class="fa fa-check"></i><b>43.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="43.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#load-packages"><i class="fa fa-check"></i><b>43.3</b> Load packages</a></li>
<li class="chapter" data-level="43.4" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#matrix-structure"><i class="fa fa-check"></i><b>43.4</b> Matrix structure</a></li>
<li class="chapter" data-level="43.5" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#diagonal-of-a-matrix"><i class="fa fa-check"></i><b>43.5</b> Diagonal of a matrix</a></li>
</ul></li>
<li class="chapter" data-level="44" data-path="alignment-in-r.html"><a href="alignment-in-r.html"><i class="fa fa-check"></i><b>44</b> Alignment in R</a>
<ul>
<li class="chapter" data-level="44.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#preliminaries-9"><i class="fa fa-check"></i><b>44.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="44.1.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#packages-2"><i class="fa fa-check"></i><b>44.1.1</b> Packages</a></li>
<li class="chapter" data-level="44.1.2" data-path="alignment-in-r.html"><a href="alignment-in-r.html#data-1"><i class="fa fa-check"></i><b>44.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="45" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html"><i class="fa fa-check"></i><b>45</b> Testing the significance of an alignment</a>
<ul>
<li class="chapter" data-level="45.1" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#calculating-the-statistical-significance-of-a-pairwise-global-alignment"><i class="fa fa-check"></i><b>45.1</b> Calculating the statistical significance of a pairwise global alignment</a></li>
<li class="chapter" data-level="45.2" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#summary-1"><i class="fa fa-check"></i><b>45.2</b> Summary</a></li>
<li class="chapter" data-level="45.3" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#links-and-further-reading-1"><i class="fa fa-check"></i><b>45.3</b> Links and Further Reading</a></li>
<li class="chapter" data-level="45.4" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#exercises-1"><i class="fa fa-check"></i><b>45.4</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="46" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>46</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="46.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>46.1</b> Prelminaries</a></li>
<li class="chapter" data-level="46.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>46.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="46.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>46.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="47" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html"><i class="fa fa-check"></i><b>47</b> Manipulating matrices and vectors: worked example</a>
<ul>
<li class="chapter" data-level="47.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#introduction-7"><i class="fa fa-check"></i><b>47.1</b> Introduction</a></li>
<li class="chapter" data-level="47.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#preliminaries-10"><i class="fa fa-check"></i><b>47.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="47.2.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-3"><i class="fa fa-check"></i><b>47.2.1</b> Packages</a></li>
<li class="chapter" data-level="47.2.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#vocab-4"><i class="fa fa-check"></i><b>47.2.2</b> Vocab</a></li>
<li class="chapter" data-level="47.2.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#functions-used-1"><i class="fa fa-check"></i><b>47.2.3</b> Functions used</a></li>
<li class="chapter" data-level="47.2.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-4"><i class="fa fa-check"></i><b>47.2.4</b> Packages</a></li>
<li class="chapter" data-level="47.2.5" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#data-2"><i class="fa fa-check"></i><b>47.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="47.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#matrix-elements-can-be-accessed-with-square-bracket-notation"><i class="fa fa-check"></i><b>47.3</b> Matrix elements can be accessed with square bracket notation</a></li>
<li class="chapter" data-level="47.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#r-object-names"><i class="fa fa-check"></i><b>47.4</b> R object names</a>
<ul>
<li class="chapter" data-level="47.4.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-vectors"><i class="fa fa-check"></i><b>47.4.1</b> Accessing items in names vectors</a></li>
<li class="chapter" data-level="47.4.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-matrices"><i class="fa fa-check"></i><b>47.4.2</b> Accessing items in names matrices</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="48" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>48</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="48.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-11"><i class="fa fa-check"></i><b>48.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="48.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages-5"><i class="fa fa-check"></i><b>48.1.1</b> Packages</a></li>
<li class="chapter" data-level="48.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-5"><i class="fa fa-check"></i><b>48.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="48.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>48.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="48.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>48.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="48.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>48.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="48.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>48.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="48.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>48.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="48.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>48.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="48.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>48.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="48.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>48.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="49" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html"><i class="fa fa-check"></i><b>49</b> The BLOSUM scoring matrix in R</a>
<ul>
<li class="chapter" data-level="49.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#introduction-8"><i class="fa fa-check"></i><b>49.1</b> Introduction</a></li>
<li class="chapter" data-level="49.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#preliminaries-12"><i class="fa fa-check"></i><b>49.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="49.2.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#packages-6"><i class="fa fa-check"></i><b>49.2.1</b> Packages</a></li>
<li class="chapter" data-level="49.2.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#vocab-5"><i class="fa fa-check"></i><b>49.2.2</b> Vocab</a></li>
<li class="chapter" data-level="49.2.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#functions-used-2"><i class="fa fa-check"></i><b>49.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="49.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#load-packages-1"><i class="fa fa-check"></i><b>49.3</b> Load packages</a></li>
<li class="chapter" data-level="49.4" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#the-blosum-matrix"><i class="fa fa-check"></i><b>49.4</b> The BLOSUM matrix</a></li>
<li class="chapter" data-level="49.5" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#visualizing-the-blosum-matrix"><i class="fa fa-check"></i><b>49.5</b> Visualizing the BLOSUM matrix</a></li>
</ul></li>
<li class="chapter" data-level="50" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>50</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="51" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>51</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="51.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-9"><i class="fa fa-check"></i><b>51.1</b> Introduction</a></li>
<li class="chapter" data-level="51.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>51.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="51.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-6"><i class="fa fa-check"></i><b>51.2.1</b> Vocab</a></li>
<li class="chapter" data-level="51.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>51.2.2</b> R functions</a></li>
<li class="chapter" data-level="51.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>51.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="51.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>51.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="51.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>51.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="51.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>51.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="51.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>51.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="51.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>51.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="51.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>51.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="51.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>51.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="51.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>51.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="51.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>51.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="51.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>51.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="51.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>51.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="51.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>51.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="52" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>52</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="52.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-13"><i class="fa fa-check"></i><b>52.1</b> Preliminaries</a></li>
<li class="chapter" data-level="52.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-10"><i class="fa fa-check"></i><b>52.2</b> Introduction</a></li>
<li class="chapter" data-level="52.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>52.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="52.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>52.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="53" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>53</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="53.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-14"><i class="fa fa-check"></i><b>53.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="53.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>53.1.1</b> Key functions</a></li>
<li class="chapter" data-level="53.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>53.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="53.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>53.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="53.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>53.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="53.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>53.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="53.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>53.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="53.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>53.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="54" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html"><i class="fa fa-check"></i><b>54</b> Local variation in GC content</a>
<ul>
<li class="chapter" data-level="54.0.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#note-on-the-biology-in-this-section"><i class="fa fa-check"></i><b>54.0.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="54.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#vocabulary-2"><i class="fa fa-check"></i><b>54.1</b> Vocabulary</a></li>
<li class="chapter" data-level="54.2" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#reading-sequence-data-with-rentrezentrez_fetch"><i class="fa fa-check"></i><b>54.2</b> Reading sequence data with rentrez::entrez_fetch</a></li>
<li class="chapter" data-level="54.3" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#local-variation-in-gc-content-1"><i class="fa fa-check"></i><b>54.3</b> Local variation in GC content</a></li>
<li class="chapter" data-level="54.4" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-analysis-of-gc-content"><i class="fa fa-check"></i><b>54.4</b> A sliding window analysis of GC content</a></li>
<li class="chapter" data-level="54.5" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-plot-of-gc-content"><i class="fa fa-check"></i><b>54.5</b> A sliding window plot of GC content</a>
<ul>
<li class="chapter" data-level="54.5.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#over-represented-and-under-represented-dna-words"><i class="fa fa-check"></i><b>54.5.1</b> Over-represented and under-represented DNA words</a></li>
</ul></li>
<li class="chapter" data-level="54.6" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#summary-2"><i class="fa fa-check"></i><b>54.6</b> Summary</a></li>
<li class="chapter" data-level="54.7" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#further-reading"><i class="fa fa-check"></i><b>54.7</b> Further Reading</a></li>
<li class="chapter" data-level="54.8" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#acknowledgements-1"><i class="fa fa-check"></i><b>54.8</b> Acknowledgements</a></li>
<li class="chapter" data-level="54.9" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#exercises-2"><i class="fa fa-check"></i><b>54.9</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="55" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html"><i class="fa fa-check"></i><b>55</b> Computational gene finding</a>
<ul>
<li class="chapter" data-level="55.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#preliminaries-15"><i class="fa fa-check"></i><b>55.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="55.1.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#packages-7"><i class="fa fa-check"></i><b>55.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="55.2" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#the-genetic-code"><i class="fa fa-check"></i><b>55.2</b> The genetic code</a></li>
<li class="chapter" data-level="55.3" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-start-and-stop-codons-in-a-dna-sequence"><i class="fa fa-check"></i><b>55.3</b> Finding start and stop codons in a DNA sequence</a></li>
<li class="chapter" data-level="55.4" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#reading-frames"><i class="fa fa-check"></i><b>55.4</b> Reading frames</a></li>
<li class="chapter" data-level="55.5" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-forward-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>55.5</b> Finding open reading frames on the forward strand of a DNA sequence</a></li>
<li class="chapter" data-level="55.6" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#predicting-the-protein-sequence-for-an-orf"><i class="fa fa-check"></i><b>55.6</b> Predicting the protein sequence for an ORF</a></li>
<li class="chapter" data-level="55.7" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-reverse-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>55.7</b> Finding open reading frames on the reverse strand of a DNA sequence</a></li>
<li class="chapter" data-level="55.8" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#lengths-of-open-reading-frames"><i class="fa fa-check"></i><b>55.8</b> Lengths of open reading frames</a></li>
</ul></li>
<li class="chapter" data-level="56" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html"><i class="fa fa-check"></i><b>56</b> Significance of ORFs</a>
<ul>
<li class="chapter" data-level="56.1" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#preliminaries-16"><i class="fa fa-check"></i><b>56.1</b> Preliminaries</a></li>
<li class="chapter" data-level="56.2" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#identifying-significant-open-reading-frames"><i class="fa fa-check"></i><b>56.2</b> Identifying significant open reading frames</a></li>
<li class="chapter" data-level="56.3" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#summary-3"><i class="fa fa-check"></i><b>56.3</b> Summary</a></li>
<li class="chapter" data-level="56.4" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#links-and-further-reading-2"><i class="fa fa-check"></i><b>56.4</b> Links and Further Reading</a></li>
<li class="chapter" data-level="56.5" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#acknowledgements-2"><i class="fa fa-check"></i><b>56.5</b> Acknowledgements</a></li>
<li class="chapter" data-level="56.6" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#exercises-3"><i class="fa fa-check"></i><b>56.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="57" data-path="comparative-genomics.html"><a href="comparative-genomics.html"><i class="fa fa-check"></i><b>57</b> Comparative Genomics</a>
<ul>
<li class="chapter" data-level="57.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#preliminaries-17"><i class="fa fa-check"></i><b>57.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="57.1.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#packages-8"><i class="fa fa-check"></i><b>57.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="57.2" data-path="comparative-genomics.html"><a href="comparative-genomics.html#introduction-11"><i class="fa fa-check"></i><b>57.2</b> Introduction</a></li>
<li class="chapter" data-level="57.3" data-path="comparative-genomics.html"><a href="comparative-genomics.html#using-the-biomart-r-library-to-query-the-ensembl-database"><i class="fa fa-check"></i><b>57.3</b> Using the biomaRt R Library to Query the Ensembl Database</a></li>
<li class="chapter" data-level="57.4" data-path="comparative-genomics.html"><a href="comparative-genomics.html#comparing-the-number-of-genes-in-two-species"><i class="fa fa-check"></i><b>57.4</b> Comparing the number of genes in two species</a></li>
<li class="chapter" data-level="57.5" data-path="comparative-genomics.html"><a href="comparative-genomics.html#identifying-homologous-genes-between-two-species"><i class="fa fa-check"></i><b>57.5</b> Identifying homologous genes between two species</a></li>
</ul></li>
<li class="chapter" data-level="58" data-path="summary-4.html"><a href="summary-4.html"><i class="fa fa-check"></i><b>58</b> Summary</a></li>
<li class="chapter" data-level="59" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html"><i class="fa fa-check"></i><b>59</b> #Links and Further Reading</a>
<ul>
<li class="chapter" data-level="59.1" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#acknowledgements-3"><i class="fa fa-check"></i><b>59.1</b> Acknowledgements</a></li>
<li class="chapter" data-level="59.2" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#exercises-4"><i class="fa fa-check"></i><b>59.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="60" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html"><i class="fa fa-check"></i><b>60</b> Hidden Markov Models</a>
<ul>
<li class="chapter" data-level="60.1" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-multinomial-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>60.1</b> A multinomial model of DNA sequence evolution</a></li>
<li class="chapter" data-level="60.2" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-multinomial-model"><i class="fa fa-check"></i><b>60.2</b> Generating a DNA sequence using a multinomial model</a></li>
<li class="chapter" data-level="60.3" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>60.3</b> A Markov model of DNA sequence evolution</a></li>
<li class="chapter" data-level="60.4" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-for-a-markov-model"><i class="fa fa-check"></i><b>60.4</b> The transition matrix for a Markov model</a></li>
<li class="chapter" data-level="60.5" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-markov-model"><i class="fa fa-check"></i><b>60.5</b> Generating a DNA sequence using a Markov model</a></li>
<li class="chapter" data-level="60.6" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>60.6</b> A Hidden Markov Model of DNA sequence evolution</a></li>
<li class="chapter" data-level="60.7" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-and-emission-matrix-for-a-hmm"><i class="fa fa-check"></i><b>60.7</b> The transition matrix and emission matrix for a HMM</a></li>
<li class="chapter" data-level="60.8" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-hmm"><i class="fa fa-check"></i><b>60.8</b> Generating a DNA sequence using a HMM</a></li>
<li class="chapter" data-level="60.9" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#inferring-the-states-of-a-hmm-that-generated-a-dna-sequence"><i class="fa fa-check"></i><b>60.9</b> Inferring the states of a HMM that generated a DNA sequence</a></li>
<li class="chapter" data-level="60.10" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-protein-sequence-evolution"><i class="fa fa-check"></i><b>60.10</b> A Hidden Markov Model of protein sequence evolution</a></li>
<li class="chapter" data-level="60.11" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#summary-5"><i class="fa fa-check"></i><b>60.11</b> Summary</a></li>
<li class="chapter" data-level="60.12" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#links-and-further-reading-4"><i class="fa fa-check"></i><b>60.12</b> Links and Further Reading</a></li>
<li class="chapter" data-level="60.13" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#exercises-5"><i class="fa fa-check"></i><b>60.13</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="61" data-path="blast-summary.html"><a href="blast-summary.html"><i class="fa fa-check"></i><b>61</b> BLAST Summary</a>
<ul>
<li class="chapter" data-level="61.1" data-path="blast-summary.html"><a href="blast-summary.html#blast-overview"><i class="fa fa-check"></i><b>61.1</b> BLAST Overview</a>
<ul>
<li class="chapter" data-level="61.1.1" data-path="blast-summary.html"><a href="blast-summary.html#why-do-a-blast-search"><i class="fa fa-check"></i><b>61.1.1</b> Why do a BLAST search?</a></li>
<li class="chapter" data-level="61.1.2" data-path="blast-summary.html"><a href="blast-summary.html#what-does-blast-let-you-do"><i class="fa fa-check"></i><b>61.1.2</b> What does BLAST let you do?</a></li>
</ul></li>
<li class="chapter" data-level="61.2" data-path="blast-summary.html"><a href="blast-summary.html#blast-recipe"><i class="fa fa-check"></i><b>61.2</b> BLAST recipe</a></li>
<li class="chapter" data-level="61.3" data-path="blast-summary.html"><a href="blast-summary.html#modules-in-the-blast-workflow-and-algorithm"><i class="fa fa-check"></i><b>61.3</b> Modules in the BLAST Workflow and Algorithm</a>
<ul>
<li class="chapter" data-level="61.3.1" data-path="blast-summary.html"><a href="blast-summary.html#module-1-setup-query-sequence-protein-blast-oriented"><i class="fa fa-check"></i><b>61.3.1</b> Module 1: Setup query sequence (Protein BLAST oriented)</a></li>
<li class="chapter" data-level="61.3.2" data-path="blast-summary.html"><a href="blast-summary.html#module-2-scanning-database-for-high-scoring-words"><i class="fa fa-check"></i><b>61.3.2</b> Module 2: Scanning database for high-scoring words</a></li>
<li class="chapter" data-level="61.3.3" data-path="blast-summary.html"><a href="blast-summary.html#module-3-traceback"><i class="fa fa-check"></i><b>61.3.3</b> Module 3: Traceback</a></li>
</ul></li>
<li class="chapter" data-level="61.4" data-path="blast-summary.html"><a href="blast-summary.html#blast-alignment-statistics"><i class="fa fa-check"></i><b>61.4</b> BLAST alignment statistics</a></li>
</ul></li>
<li class="chapter" data-level="62" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html"><i class="fa fa-check"></i><b>62</b> BLAST Alignment score distribution</a>
<ul>
<li class="chapter" data-level="62.1" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#packages-9"><i class="fa fa-check"></i><b>62.1</b> Packages</a></li>
<li class="chapter" data-level="62.2" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#distributions-of-alignment-scores"><i class="fa fa-check"></i><b>62.2</b> Distributions of alignment scores</a></li>
<li class="chapter" data-level="62.3" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-data"><i class="fa fa-check"></i><b>62.3</b> Simulating data</a></li>
<li class="chapter" data-level="62.4" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#lots-of-data-in-biology-is-normal"><i class="fa fa-check"></i><b>62.4</b> Lots of data in biology is normal</a></li>
<li class="chapter" data-level="62.5" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-the-normal-distribution"><i class="fa fa-check"></i><b>62.5</b> Simulating the normal distribution</a></li>
<li class="chapter" data-level="62.6" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#the-extreme-value-distribution"><i class="fa fa-check"></i><b>62.6</b> The extreme value distribution</a></li>
<li class="chapter" data-level="62.7" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#replicating-figure-4.14"><i class="fa fa-check"></i><b>62.7</b> Replicating Figure 4.14</a></li>
<li class="chapter" data-level="62.8" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#why-do-we-care"><i class="fa fa-check"></i><b>62.8</b> Why do we care?</a></li>
</ul></li>
<li class="chapter" data-level="63" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><i class="fa fa-check"></i><b>63</b> Calculating E values, Bit scores, and P-values from BLAST output</a>
<ul>
<li class="chapter" data-level="63.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-12"><i class="fa fa-check"></i><b>63.1</b> Introduction</a></li>
<li class="chapter" data-level="63.2" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#case-study"><i class="fa fa-check"></i><b>63.2</b> Case Study</a></li>
<li class="chapter" data-level="63.3" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-statistics"><i class="fa fa-check"></i><b>63.3</b> BLAST Statistics</a></li>
<li class="chapter" data-level="63.4" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot"><i class="fa fa-check"></i><b>63.4</b> Dotplot</a>
<ul>
<li class="chapter" data-level="63.4.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot-in-r---optional-experimental-code"><i class="fa fa-check"></i><b>63.4.1</b> Dotplot in R - optional, experimental code</a></li>
</ul></li>
<li class="chapter" data-level="63.5" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-search-summary"><i class="fa fa-check"></i><b>63.5</b> BLAST “Search Summary”</a></li>
<li class="chapter" data-level="63.6" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-alignments"><i class="fa fa-check"></i><b>63.6</b> BLAST Alignments</a></li>
<li class="chapter" data-level="63.7" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#equations"><i class="fa fa-check"></i><b>63.7</b> Equations</a>
<ul>
<li class="chapter" data-level="63.7.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-value-eq-4.5-pg-142"><i class="fa fa-check"></i><b>63.7.1</b> E value (eq 4.5, pg 142)</a></li>
</ul></li>
<li class="chapter" data-level="63.8" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bit-scores-s"><i class="fa fa-check"></i><b>63.8</b> Bit Scores (S’)</a></li>
<li class="chapter" data-level="63.9" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-from-bit-scores"><i class="fa fa-check"></i><b>63.9</b> E from Bit scores</a></li>
<li class="chapter" data-level="63.10" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#p-values"><i class="fa fa-check"></i><b>63.10</b> P values</a></li>
<li class="chapter" data-level="63.11" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#repliminaries"><i class="fa fa-check"></i><b>63.11</b> Repliminaries</a></li>
<li class="chapter" data-level="63.12" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#concepts"><i class="fa fa-check"></i><b>63.12</b> Concepts</a></li>
<li class="chapter" data-level="63.13" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#functions-6"><i class="fa fa-check"></i><b>63.13</b> Functions</a></li>
<li class="chapter" data-level="63.14" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-13"><i class="fa fa-check"></i><b>63.14</b> Introduction</a></li>
<li class="chapter" data-level="63.15" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees"><i class="fa fa-check"></i><b>63.15</b> Number of rooted trees</a></li>
<li class="chapter" data-level="63.16" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees-in-r"><i class="fa fa-check"></i><b>63.16</b> Number of rooted trees in R</a>
<ul>
<li class="chapter" data-level="63.16.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#unique-topologies-versus-rotations"><i class="fa fa-check"></i><b>63.16.1</b> Unique topologies versus rotations</a></li>
</ul></li>
<li class="chapter" data-level="63.17" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bilography"><i class="fa fa-check"></i><b>63.17</b> Bilography</a></li>
</ul></li>
<li class="chapter" data-level="64" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html"><i class="fa fa-check"></i><b>64</b> UPGMA the hard way</a>
<ul>
<li class="chapter" data-level="64.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#introudction-to-upgma"><i class="fa fa-check"></i><b>64.1</b> Introudction to UPGMA</a></li>
<li class="chapter" data-level="64.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#preliminaries-18"><i class="fa fa-check"></i><b>64.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="64.2.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#r-libraries"><i class="fa fa-check"></i><b>64.2.1</b> R Libraries</a></li>
<li class="chapter" data-level="64.2.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#custom-function-well-use"><i class="fa fa-check"></i><b>64.2.2</b> Custom function we’ll use</a></li>
</ul></li>
<li class="chapter" data-level="64.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#data-matrix"><i class="fa fa-check"></i><b>64.3</b> Data matrix</a>
<ul>
<li class="chapter" data-level="64.3.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#building-the-matrix"><i class="fa fa-check"></i><b>64.3.1</b> Building the matrix</a></li>
</ul></li>
<li class="chapter" data-level="64.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#what-does-this-distance-matrix-represent"><i class="fa fa-check"></i><b>64.4</b> What does this distance matrix represent?</a>
<ul>
<li class="chapter" data-level="64.4.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-pairwise-linear-distances"><i class="fa fa-check"></i><b>64.4.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="64.4.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-2d-map-of-linear-distances"><i class="fa fa-check"></i><b>64.4.2</b> Genetic distances conceptualized as 2D map of linear distances</a></li>
</ul></li>
<li class="chapter" data-level="64.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#from-linear-distances-to-tree-distances"><i class="fa fa-check"></i><b>64.5</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="64.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#the-upgma-algorithm"><i class="fa fa-check"></i><b>64.6</b> The UPGMA algorithm</a>
<ul>
<li class="chapter" data-level="64.6.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-1-find-the-2-taxa-that-are-currently-closest-together"><i class="fa fa-check"></i><b>64.6.1</b> Step 1: find the 2 taxa that are currently closest together</a></li>
<li class="chapter" data-level="64.6.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node"><i class="fa fa-check"></i><b>64.6.2</b> Step 2: calculate the branch lengths connect the taxa to their shared node</a></li>
<li class="chapter" data-level="64.6.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-determine-if-the-algorithm-is-done"><i class="fa fa-check"></i><b>64.6.3</b> Step 3a: determine if the algorithm is done</a></li>
<li class="chapter" data-level="64.6.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u"><i class="fa fa-check"></i><b>64.6.4</b> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</a></li>
<li class="chapter" data-level="64.6.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-4-determine-the-distance-from-u-to-the-other-remaining-clusters."><i class="fa fa-check"></i><b>64.6.5</b> Step 4: Determine the distance from u to the other remaining clusters.</a></li>
<li class="chapter" data-level="64.6.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-5-create-a-new-matrix"><i class="fa fa-check"></i><b>64.6.6</b> Step 5: Create a new matrix</a></li>
<li class="chapter" data-level="64.6.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa"><i class="fa fa-check"></i><b>64.6.7</b> Key step: calculating the distance from a clade to the remaining taxa</a></li>
</ul></li>
<li class="chapter" data-level="64.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-2d-representation"><i class="fa fa-check"></i><b>64.7</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="64.8" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---round-1"><i class="fa fa-check"></i><b>64.8</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="64.8.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#determine-minimum-value-in-current-matrix"><i class="fa fa-check"></i><b>64.8.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="64.8.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade"><i class="fa fa-check"></i><b>64.8.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="64.8.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length"><i class="fa fa-check"></i><b>64.8.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="64.8.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>64.8.4</b> Calculate distance from clade to all other points</a></li>
<li class="chapter" data-level="64.8.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-matrix-with-new-distance"><i class="fa fa-check"></i><b>64.8.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="64.9" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---next-round"><i class="fa fa-check"></i><b>64.9</b> Algorithm - next round</a>
<ul>
<li class="chapter" data-level="64.9.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-1"><i class="fa fa-check"></i><b>64.9.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="64.9.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-1"><i class="fa fa-check"></i><b>64.9.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="64.9.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#recalculate-distance-from-abe-to-all-remainign-taxa"><i class="fa fa-check"></i><b>64.9.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="64.9.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-the-matrix"><i class="fa fa-check"></i><b>64.9.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="64.10" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#next-iteration"><i class="fa fa-check"></i><b>64.10</b> Next iteration</a>
<ul>
<li class="chapter" data-level="64.10.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-lengths"><i class="fa fa-check"></i><b>64.10.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="64.10.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-2"><i class="fa fa-check"></i><b>64.10.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="64.10.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-2"><i class="fa fa-check"></i><b>64.10.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="64.10.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#distance-of-c-to-abe-and-d-to-abe"><i class="fa fa-check"></i><b>64.10.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="64.11" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finish-up"><i class="fa fa-check"></i><b>64.11</b> Finish up</a>
<ul>
<li class="chapter" data-level="64.11.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-3"><i class="fa fa-check"></i><b>64.11.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="64.11.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-3"><i class="fa fa-check"></i><b>64.11.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="64.12" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finalizing-branch-lengths"><i class="fa fa-check"></i><b>64.12</b> Finalizing branch lengths</a></li>
</ul></li>
<li class="chapter" data-level="65" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html"><i class="fa fa-check"></i><b>65</b> Representing phylogenetic trees in Newick format</a>
<ul>
<li class="chapter" data-level="65.1" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#vocab-7"><i class="fa fa-check"></i><b>65.1</b> Vocab</a></li>
<li class="chapter" data-level="65.2" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#introduction-14"><i class="fa fa-check"></i><b>65.2</b> Introduction</a></li>
<li class="chapter" data-level="65.3" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-three-taxa"><i class="fa fa-check"></i><b>65.3</b> Trees with three taxa</a></li>
<li class="chapter" data-level="65.4" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-four-taxa"><i class="fa fa-check"></i><b>65.4</b> Trees with four taxa</a></li>
</ul></li>
<li class="chapter" data-level="66" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html"><i class="fa fa-check"></i><b>66</b> Plotting trees from Newick notation in R</a>
<ul>
<li class="chapter" data-level="66.1" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#preliminaries-19"><i class="fa fa-check"></i><b>66.1</b> Preliminaries</a></li>
<li class="chapter" data-level="66.2" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#note-1"><i class="fa fa-check"></i><b>66.2</b> Note</a></li>
<li class="chapter" data-level="66.3" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick"><i class="fa fa-check"></i><b>66.3</b> 3 taxa tree in Newick</a></li>
<li class="chapter" data-level="66.4" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick-1"><i class="fa fa-check"></i><b>66.4</b> 4 taxa tree in Newick</a></li>
<li class="chapter" data-level="66.5" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-in-newick-format"><i class="fa fa-check"></i><b>66.5</b> 5 taxa in Newick Format</a></li>
<li class="chapter" data-level="66.6" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#many-taxa-tree"><i class="fa fa-check"></i><b>66.6</b> Many-taxa tree</a></li>
<li class="chapter" data-level="66.7" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#newick-and-branch-lengths"><i class="fa fa-check"></i><b>66.7</b> Newick and Branch lengths</a></li>
<li class="chapter" data-level="66.8" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#additional-reading"><i class="fa fa-check"></i><b>66.8</b> Additional Reading</a></li>
</ul></li>
<li class="chapter" data-level="67" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><i class="fa fa-check"></i><b>67</b> Phylogenetic trees using the UPGMA clustering algoirthm</a>
<ul>
<li class="chapter" data-level="67.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#introduction-15"><i class="fa fa-check"></i><b>67.1</b> Introduction</a></li>
<li class="chapter" data-level="67.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#preliminaries-20"><i class="fa fa-check"></i><b>67.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="67.2.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#vocab-8"><i class="fa fa-check"></i><b>67.2.1</b> Vocab</a></li>
<li class="chapter" data-level="67.2.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-r-functions"><i class="fa fa-check"></i><b>67.2.2</b> General R functions</a></li>
<li class="chapter" data-level="67.2.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#specific-functions"><i class="fa fa-check"></i><b>67.2.3</b> Specific functions</a></li>
</ul></li>
<li class="chapter" data-level="67.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#sequence-comparisons"><i class="fa fa-check"></i><b>67.3</b> Sequence comparisons</a></li>
<li class="chapter" data-level="67.4" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrices"><i class="fa fa-check"></i><b>67.4</b> Distance matrices</a></li>
<li class="chapter" data-level="67.5" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#from-a-distance-matrix-to-phylogenetic-tree"><i class="fa fa-check"></i><b>67.5</b> From a distance matrix to phylogenetic tree</a></li>
<li class="chapter" data-level="67.6" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-raw-sequences"><i class="fa fa-check"></i><b>67.6</b> Distance matrix in R from raw sequences</a></li>
<li class="chapter" data-level="67.7" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-normal-matrix"><i class="fa fa-check"></i><b>67.7</b> Distance matrix in R from normal matrix</a></li>
<li class="chapter" data-level="67.8" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#upgma-in-r"><i class="fa fa-check"></i><b>67.8</b> UPGMA in R</a></li>
<li class="chapter" data-level="67.9" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#how-upgma-clustering-works"><i class="fa fa-check"></i><b>67.9</b> How UPGMA clustering works</a>
<ul>
<li class="chapter" data-level="67.9.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#challenge-2"><i class="fa fa-check"></i><b>67.9.1</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="67.10" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#further-information"><i class="fa fa-check"></i><b>67.10</b> Further information</a>
<ul>
<li class="chapter" data-level="67.10.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-info"><i class="fa fa-check"></i><b>67.10.1</b> General info</a></li>
<li class="chapter" data-level="67.10.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#worked-examples"><i class="fa fa-check"></i><b>67.10.2</b> Worked examples:</a></li>
<li class="chapter" data-level="67.10.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#additional-notes"><i class="fa fa-check"></i><b>67.10.3</b> Additional notes:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="68" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><i class="fa fa-check"></i><b>68</b> Tutorial: Writing functions to calculate the number of phylogenetic trees</a>
<ul>
<li class="chapter" data-level="68.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#repliminaries-1"><i class="fa fa-check"></i><b>68.1</b> Repliminaries</a>
<ul>
<li class="chapter" data-level="68.1.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-7"><i class="fa fa-check"></i><b>68.1.1</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="68.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#concepts-1"><i class="fa fa-check"></i><b>68.2</b> Concepts</a>
<ul>
<li class="chapter" data-level="68.2.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vocab-9"><i class="fa fa-check"></i><b>68.2.1</b> Vocab:</a></li>
<li class="chapter" data-level="68.2.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#packages-10"><i class="fa fa-check"></i><b>68.2.2</b> Packages:</a></li>
</ul></li>
<li class="chapter" data-level="68.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#introduction-16"><i class="fa fa-check"></i><b>68.3</b> Introduction</a></li>
<li class="chapter" data-level="68.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-1"><i class="fa fa-check"></i><b>68.4</b> Number of rooted trees</a></li>
<li class="chapter" data-level="68.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-in-r-1"><i class="fa fa-check"></i><b>68.5</b> Number of rooted trees in R</a></li>
<li class="chapter" data-level="68.6" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-in-r-1"><i class="fa fa-check"></i><b>68.6</b> Functions in R</a></li>
<li class="chapter" data-level="68.7" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-to-calculate-the-number-of-possible-phylogenetic-trees"><i class="fa fa-check"></i><b>68.7</b> Function to calculate the number of possible phylogenetic trees</a></li>
<li class="chapter" data-level="68.8" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vectorized-inputs-to-functions"><i class="fa fa-check"></i><b>68.8</b> Vectorized inputs to functions</a></li>
<li class="chapter" data-level="68.9" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-conditional-statements"><i class="fa fa-check"></i><b>68.9</b> Adding conditional statements</a></li>
<li class="chapter" data-level="68.10" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-multiple-conditional-statements"><i class="fa fa-check"></i><b>68.10</b> Adding multiple conditional statements</a></li>
<li class="chapter" data-level="68.11" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-additional-arguments"><i class="fa fa-check"></i><b>68.11</b> Adding additional arguments</a></li>
<li class="chapter" data-level="68.12" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-_un_rooted-trees"><i class="fa fa-check"></i><b>68.12</b> Assignment: _un_rooted trees</a></li>
<li class="chapter" data-level="68.13" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-1"><i class="fa fa-check"></i><b>68.13</b> Assignment part 1</a>
<ul>
<li class="chapter" data-level="68.13.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#part-1-answer"><i class="fa fa-check"></i><b>68.13.1</b> Part 1 answer</a></li>
</ul></li>
<li class="chapter" data-level="68.14" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-2"><i class="fa fa-check"></i><b>68.14</b> Assignment part 2</a>
<ul>
<li class="chapter" data-level="68.14.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#answer-1"><i class="fa fa-check"></i><b>68.14.1</b> Answer</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="69" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>69</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="69.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>69.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="69.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>69.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="69.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>69.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="69.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>69.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="69.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>69.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-3"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="69.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>69.2</b> Help!</a>
<ul>
<li class="chapter" data-level="69.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>69.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="69.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>69.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="69.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>69.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="69.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>69.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="69.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>69.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="69.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>69.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="69.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>69.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="69.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>69.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="or-loops-in-r---copy-alt" class="section level1" number="10">
<h1><span class="header-section-number">Chapter 10</span> or loops in R - copy? alt?</h1>
<!-- TODO: merge? cut? -->
<div id="introduction-3" class="section level2" number="10.1">
<h2><span class="header-section-number">10.1</span> Introduction</h2>
<p>There are several way to write for() loops in R. For this assignment, I asked you to write a loop which printed out each letter of the alphabet.</p>
<p>There are several ways to write a for loop in R. I will show several different ways. Out of habitat I will generally use the style of (for i in 1:length(x)), though this is not an optimal way to do things. I will discuss this way as well as some more sounds approaches.</p>
</div>
<div id="reminders" class="section level2" number="10.2">
<h2><span class="header-section-number">10.2</span> Reminders</h2>
<p>R has a built in object with has the data, called LETTERS.</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="or-loops-in-r-copy-alt.html#cb203-1" aria-hidden="true" tabindex="-1"></a>LETTERS</span></code></pre></div>
<pre><code>##  [1] &quot;A&quot; &quot;B&quot; &quot;C&quot; &quot;D&quot; &quot;E&quot; &quot;F&quot; &quot;G&quot; &quot;H&quot; &quot;I&quot; &quot;J&quot; &quot;K&quot; &quot;L&quot; &quot;M&quot; &quot;N&quot; &quot;O&quot; &quot;P&quot; &quot;Q&quot; &quot;R&quot; &quot;S&quot;
## [20] &quot;T&quot; &quot;U&quot; &quot;V&quot; &quot;W&quot; &quot;X&quot; &quot;Y&quot; &quot;Z&quot;</code></pre>
<p>The object LETTERS is a vector containing character data.</p>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="or-loops-in-r-copy-alt.html#cb205-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(LETTERS)</span></code></pre></div>
<pre><code>## [1] &quot;character&quot;           &quot;vector&quot;              &quot;data.frameRowLabels&quot;
## [4] &quot;SuperClassMethod&quot;</code></pre>
<p>We we access each element of a vector with its index number. To get “A”</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="or-loops-in-r-copy-alt.html#cb207-1" aria-hidden="true" tabindex="-1"></a>LETTERS[<span class="dv">1</span>]</span></code></pre></div>
<pre><code>## [1] &quot;A&quot;</code></pre>
<p>To get “Z”</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="or-loops-in-r-copy-alt.html#cb209-1" aria-hidden="true" tabindex="-1"></a>LETTERS[<span class="dv">26</span>]</span></code></pre></div>
<pre><code>## [1] &quot;Z&quot;</code></pre>
<p>If I want, I can access a given letter by assigning its index number to an object. Say I assign 26 to the letter i</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="or-loops-in-r-copy-alt.html#cb211-1" aria-hidden="true" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="dv">26</span></span></code></pre></div>
<p>I can get Z like this then</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="or-loops-in-r-copy-alt.html#cb212-1" aria-hidden="true" tabindex="-1"></a>LETTERS[i]</span></code></pre></div>
<pre><code>## [1] &quot;Z&quot;</code></pre>
</div>
<div id="for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop" class="section level2" number="10.3">
<h2><span class="header-section-number">10.3</span> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</h2>
<p>This is a bad way to do a for loop but makes the process totally transparent. There are 26 letters in the alphabet, so there are 26 elements in the vector LETTERS. Therefore, what we want R to do - print out a letter from the alphabet - need to be repeated 26 times. So our for loop needs to repeat the same action 26 times.</p>
<p>We can make an <strong>index</strong> for R like this</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="or-loops-in-r-copy-alt.html#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">26</span></span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
## [26] 26</code></pre>
<p>Similarly, we could do this</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="or-loops-in-r-copy-alt.html#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">26</span>)</span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
## [26] 26</code></pre>
<p>The colon in R is used to expand a sequence of numbers. Its a shortcut for the seq() command.</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="or-loops-in-r-copy-alt.html#cb218-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">1</span>, <span class="at">to =</span> <span class="dv">26</span>, <span class="at">by =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
## [26] 26</code></pre>
<p>Remember that index values can be used to pull out elements from a vectors. So as noted above if I want to print the letter A, which is stored in the first element of the LETTERS vector, I can do this</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="or-loops-in-r-copy-alt.html#cb220-1" aria-hidden="true" tabindex="-1"></a>LETTERS[<span class="dv">1</span>]</span></code></pre></div>
<pre><code>## [1] &quot;A&quot;</code></pre>
<p>If I want, I could store this output in a an object</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="or-loops-in-r-copy-alt.html#cb222-1" aria-hidden="true" tabindex="-1"></a>letter.A <span class="ot">&lt;-</span> LETTERS[<span class="dv">1</span>]</span>
<span id="cb222-2"><a href="or-loops-in-r-copy-alt.html#cb222-2" aria-hidden="true" tabindex="-1"></a>letter.A</span></code></pre></div>
<pre><code>## [1] &quot;A&quot;</code></pre>
<p>What a for loop can do is cycle through each index value, 1 to 26, and put it in between the bracket. It would look like this loop below;note that I added an extra step where I print out the index value.</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="or-loops-in-r-copy-alt.html#cb224-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">26</span>){</span>
<span id="cb224-2"><a href="or-loops-in-r-copy-alt.html#cb224-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i)                <span class="co">#print the index</span></span>
<span id="cb224-3"><a href="or-loops-in-r-copy-alt.html#cb224-3" aria-hidden="true" tabindex="-1"></a>  letter.i <span class="ot">&lt;-</span> LETTERS[i]  <span class="co">#get leter i</span></span>
<span id="cb224-4"><a href="or-loops-in-r-copy-alt.html#cb224-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(letter.i)         <span class="co">#print letter i</span></span>
<span id="cb224-5"><a href="or-loops-in-r-copy-alt.html#cb224-5" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] 1
## [1] &quot;A&quot;
## [1] 2
## [1] &quot;B&quot;
## [1] 3
## [1] &quot;C&quot;
## [1] 4
## [1] &quot;D&quot;
## [1] 5
## [1] &quot;E&quot;
## [1] 6
## [1] &quot;F&quot;
## [1] 7
## [1] &quot;G&quot;
## [1] 8
## [1] &quot;H&quot;
## [1] 9
## [1] &quot;I&quot;
## [1] 10
## [1] &quot;J&quot;
## [1] 11
## [1] &quot;K&quot;
## [1] 12
## [1] &quot;L&quot;
## [1] 13
## [1] &quot;M&quot;
## [1] 14
## [1] &quot;N&quot;
## [1] 15
## [1] &quot;O&quot;
## [1] 16
## [1] &quot;P&quot;
## [1] 17
## [1] &quot;Q&quot;
## [1] 18
## [1] &quot;R&quot;
## [1] 19
## [1] &quot;S&quot;
## [1] 20
## [1] &quot;T&quot;
## [1] 21
## [1] &quot;U&quot;
## [1] 22
## [1] &quot;V&quot;
## [1] 23
## [1] &quot;W&quot;
## [1] 24
## [1] &quot;X&quot;
## [1] 25
## [1] &quot;Y&quot;
## [1] 26
## [1] &quot;Z&quot;</code></pre>
<p>In this loop I used 1:26 for my index. This is <strong>hard coding</strong> the index values, which is never a good idea. But for the sake of illustration i makes it totally obvious what we’re doing.</p>
<p>As I said above 1:26 is the same as c(1:26), so this loop does the same thing</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="or-loops-in-r-copy-alt.html#cb226-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">26</span>){</span>
<span id="cb226-2"><a href="or-loops-in-r-copy-alt.html#cb226-2" aria-hidden="true" tabindex="-1"></a>  letter.i <span class="ot">&lt;-</span> LETTERS[i]  <span class="co">#get leter i</span></span>
<span id="cb226-3"><a href="or-loops-in-r-copy-alt.html#cb226-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(letter.i)         <span class="co">#print letter i</span></span>
<span id="cb226-4"><a href="or-loops-in-r-copy-alt.html#cb226-4" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>The colon : is a shortcut for the seq() command; if I wanted to be 100% transparent to someone who didn’t know R I could write this:</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="or-loops-in-r-copy-alt.html#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">1</span>, <span class="at">to =</span> <span class="dv">26</span>, <span class="at">by =</span> <span class="dv">1</span>)){</span>
<span id="cb227-2"><a href="or-loops-in-r-copy-alt.html#cb227-2" aria-hidden="true" tabindex="-1"></a>  letter.i <span class="ot">&lt;-</span> LETTERS[i]  <span class="co">#get leter i</span></span>
<span id="cb227-3"><a href="or-loops-in-r-copy-alt.html#cb227-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(letter.i)         <span class="co">#print letter i</span></span>
<span id="cb227-4"><a href="or-loops-in-r-copy-alt.html#cb227-4" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>The step of assigning the current letter indexed by i to an object letter.i isn’t actually necessary. I could simplify the code like this by <strong>nesting</strong> the LETTERS[i] statement within the print() statement. The letter indexed by i therefore gets pulled up and directly <strong>passed</strong> to print().</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="or-loops-in-r-copy-alt.html#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">26</span>){</span>
<span id="cb228-2"><a href="or-loops-in-r-copy-alt.html#cb228-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(LETTERS[i])         <span class="co">#print letter i</span></span>
<span id="cb228-3"><a href="or-loops-in-r-copy-alt.html#cb228-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Lines and spacing are arbitrary in R. When a for loop is short you may see people do this to save space.</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="or-loops-in-r-copy-alt.html#cb229-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">26</span>){ <span class="fu">print</span>(LETTERS[i]) }</span></code></pre></div>
<p>Or even this, dropping the curly braces.</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="or-loops-in-r-copy-alt.html#cb230-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">26</span>) <span class="fu">print</span>(LETTERS[i])  </span></code></pre></div>
<p>Some people value compact code; there’s no penalty in any way for using multiple lines so its generally considered best practice to split code over more lines rather than fewer.</p>
</div>
<div id="for-loop-3-the-classic-for-loop" class="section level2" number="10.4">
<h2><span class="header-section-number">10.4</span> For loop 3: the classic for loop</h2>
<p>A common way to code for loops is to use a function like length() for vectors or dim() for matrices to determine how many times a loop must repeat its action. Instead of hard-coding 1:26 to get the values 1, 2, 3,… 26 I can do this:</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="or-loops-in-r-copy-alt.html#cb231-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(LETTERS)</span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
## [26] 26</code></pre>
<p>This is useful in more complex situations where you could end up changing how many elements are in LETTERS. For this approach,the for loop looks like this</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="or-loops-in-r-copy-alt.html#cb233-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(LETTERS)){</span>
<span id="cb233-2"><a href="or-loops-in-r-copy-alt.html#cb233-2" aria-hidden="true" tabindex="-1"></a>  letter.i <span class="ot">&lt;-</span> LETTERS[i]  <span class="co">#get leter i</span></span>
<span id="cb233-3"><a href="or-loops-in-r-copy-alt.html#cb233-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(letter.i)         <span class="co">#print letter i</span></span>
<span id="cb233-4"><a href="or-loops-in-r-copy-alt.html#cb233-4" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>In this formulation, R is making an index 1 to 26 based on the length of LETTERS. Its then assigning a value to i each time it goes through the loop. Then it pulls out the appropriate letter based on i. This is basically the same as the first example, just the values of the index, 1 to 26, is determine by R and not set by hand.</p>
<p>I’ve written hundreds of loops this way an it this way works just fine. However, in more complicated code a problem can arise if someone a 0 gets its way into a loop. See <a href="https://jef.works/R-style-guide/#loops" class="uri">https://jef.works/R-style-guide/#loops</a> for a reference to this.</p>
<p>The code below is optional, but if your curious it displays a behavior which may not be desirable.</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="or-loops-in-r-copy-alt.html#cb234-1" aria-hidden="true" tabindex="-1"></a>bad.index <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb234-2"><a href="or-loops-in-r-copy-alt.html#cb234-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(bad.index)</span></code></pre></div>
<pre><code>## [1] 0</code></pre>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="or-loops-in-r-copy-alt.html#cb236-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(bad.index)){</span>
<span id="cb236-2"><a href="or-loops-in-r-copy-alt.html#cb236-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(bad.index[i])</span>
<span id="cb236-3"><a href="or-loops-in-r-copy-alt.html#cb236-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## NULL
## NULL</code></pre>
<p>This problem is most likely to arise when a for loop is embedded in a larger program and vectors like LETTERS get generated by one part of the program and get fed later into the loop.</p>
<p>For this course I will probably continue to use the general format of for(i in 1:length(x)) out of habit. Hopefully by next year I’ll be re-trained to use one of the techniques shown below.</p>
</div>
<div id="for-loop-3-the-power-user-for-loop" class="section level2" number="10.5">
<h2><span class="header-section-number">10.5</span> For loop 3: The Power-user for loop</h2>
<p>Probably the cleanest way to do a for loop in R is like this next example. A scientific survey on Twitter indicates that this how the cool kids write their loops.</p>
<p>In this version, R steps through each element of the vector LETTERS and on the fly assigns that element to the object i. Print then tells R to print the letter to the console.</p>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="or-loops-in-r-copy-alt.html#cb238-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> LETTERS){</span>
<span id="cb238-2"><a href="or-loops-in-r-copy-alt.html#cb238-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i)</span>
<span id="cb238-3"><a href="or-loops-in-r-copy-alt.html#cb238-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Note that if we leave out print() then we get blank screen. R is basically just reading the numbers to itself.</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="or-loops-in-r-copy-alt.html#cb239-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> LETTERS){</span>
<span id="cb239-2"><a href="or-loops-in-r-copy-alt.html#cb239-2" aria-hidden="true" tabindex="-1"></a> i</span>
<span id="cb239-3"><a href="or-loops-in-r-copy-alt.html#cb239-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>The use if “i” is arbitrary. We could just j instead</p>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="or-loops-in-r-copy-alt.html#cb240-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> LETTERS){</span>
<span id="cb240-2"><a href="or-loops-in-r-copy-alt.html#cb240-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(j)</span>
<span id="cb240-3"><a href="or-loops-in-r-copy-alt.html#cb240-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>To remember this format it might be useful to think about it like this: instead of using i or j, use something representative of what the loop is accessing from LETTERS. In the next version, I’ll use letter instead of i.</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="or-loops-in-r-copy-alt.html#cb241-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(letter <span class="cf">in</span> LETTERS){</span>
<span id="cb241-2"><a href="or-loops-in-r-copy-alt.html#cb241-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(letter)</span>
<span id="cb241-3"><a href="or-loops-in-r-copy-alt.html#cb241-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;A&quot;
## [1] &quot;B&quot;
## [1] &quot;C&quot;
## [1] &quot;D&quot;
## [1] &quot;E&quot;
## [1] &quot;F&quot;
## [1] &quot;G&quot;
## [1] &quot;H&quot;
## [1] &quot;I&quot;
## [1] &quot;J&quot;
## [1] &quot;K&quot;
## [1] &quot;L&quot;
## [1] &quot;M&quot;
## [1] &quot;N&quot;
## [1] &quot;O&quot;
## [1] &quot;P&quot;
## [1] &quot;Q&quot;
## [1] &quot;R&quot;
## [1] &quot;S&quot;
## [1] &quot;T&quot;
## [1] &quot;U&quot;
## [1] &quot;V&quot;
## [1] &quot;W&quot;
## [1] &quot;X&quot;
## [1] &quot;Y&quot;
## [1] &quot;Z&quot;</code></pre>
<p>This kind of reads like "for each letter in the vector LETTERS, print the current letter.</p>
<p>This general approach avoids the problem with the first two versions of the loops. Recall that I made an object called bad.index that was just NULL</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="or-loops-in-r-copy-alt.html#cb243-1" aria-hidden="true" tabindex="-1"></a>bad.index <span class="ot">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
<p>This truly is nothing; it contains no data and has a length of zero.</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="or-loops-in-r-copy-alt.html#cb244-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(bad.index)</span></code></pre></div>
<pre><code>## [1] &quot;NULL&quot;             &quot;OptionalFunction&quot; &quot;optionalMethod&quot;</code></pre>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="or-loops-in-r-copy-alt.html#cb246-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(bad.index)</span></code></pre></div>
<pre><code>## [1] &quot;NULL&quot;</code></pre>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="or-loops-in-r-copy-alt.html#cb248-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(bad.index)</span></code></pre></div>
<pre><code>## [1] 0</code></pre>
<p>If I use this in a for loop in the current style I get no output</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="or-loops-in-r-copy-alt.html#cb250-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> bad.index){</span>
<span id="cb250-2"><a href="or-loops-in-r-copy-alt.html#cb250-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i)</span>
<span id="cb250-3"><a href="or-loops-in-r-copy-alt.html#cb250-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>This is because bad.index has nothing to in it.</p>
<p>A limitation of this method is that i (or j, or letters) isn’t an index, but takes on an assigned value from LETTERS. So in my old school loop I can do this, which allows me to print 2 consecutive letters. This is done by done math on the index i.</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="or-loops-in-r-copy-alt.html#cb251-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(LETTERS)){</span>
<span id="cb251-2"><a href="or-loops-in-r-copy-alt.html#cb251-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(LETTERS[<span class="fu">c</span>(i,i<span class="sc">+</span><span class="dv">1</span>)])</span>
<span id="cb251-3"><a href="or-loops-in-r-copy-alt.html#cb251-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;B&quot;
## [1] &quot;B&quot; &quot;C&quot;
## [1] &quot;C&quot; &quot;D&quot;
## [1] &quot;D&quot; &quot;E&quot;
## [1] &quot;E&quot; &quot;F&quot;
## [1] &quot;F&quot; &quot;G&quot;
## [1] &quot;G&quot; &quot;H&quot;
## [1] &quot;H&quot; &quot;I&quot;
## [1] &quot;I&quot; &quot;J&quot;
## [1] &quot;J&quot; &quot;K&quot;
## [1] &quot;K&quot; &quot;L&quot;
## [1] &quot;L&quot; &quot;M&quot;
## [1] &quot;M&quot; &quot;N&quot;
## [1] &quot;N&quot; &quot;O&quot;
## [1] &quot;O&quot; &quot;P&quot;
## [1] &quot;P&quot; &quot;Q&quot;
## [1] &quot;Q&quot; &quot;R&quot;
## [1] &quot;R&quot; &quot;S&quot;
## [1] &quot;S&quot; &quot;T&quot;
## [1] &quot;T&quot; &quot;U&quot;
## [1] &quot;U&quot; &quot;V&quot;
## [1] &quot;V&quot; &quot;W&quot;
## [1] &quot;W&quot; &quot;X&quot;
## [1] &quot;X&quot; &quot;Y&quot;
## [1] &quot;Y&quot; &quot;Z&quot;
## [1] &quot;Z&quot; NA</code></pre>
<p>With the other approach, I can’t do math on i</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="or-loops-in-r-copy-alt.html#cb253-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> LETTERS){</span>
<span id="cb253-2"><a href="or-loops-in-r-copy-alt.html#cb253-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i)</span>
<span id="cb253-3"><a href="or-loops-in-r-copy-alt.html#cb253-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb253-4"><a href="or-loops-in-r-copy-alt.html#cb253-4" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>There’s probably a way around this, but its not apparent to me right now.</p>
</div>
<div id="for-loop-4-a-for-loop-for-all-purposes" class="section level2" number="10.6">
<h2><span class="header-section-number">10.6</span> For loop 4: A for loop for all purposes</h2>
<p>For the discerning R programming who doesn’t like for loop version 3 there is a function called seq_along()</p>
<p>seq_along() determines what is a valid index for a given object</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="or-loops-in-r-copy-alt.html#cb254-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seq_along</span>(LETTERS)</span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
## [26] 26</code></pre>
<p>If you pass it a null value it can tell</p>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="or-loops-in-r-copy-alt.html#cb256-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seq_along</span>(bad.index)</span></code></pre></div>
<pre><code>## integer(0)</code></pre>
<p>seq_along therefore avoids the bad features of 1:length(x)</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="or-loops-in-r-copy-alt.html#cb258-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(bad.index)){</span>
<span id="cb258-2"><a href="or-loops-in-r-copy-alt.html#cb258-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(bad.index[i])</span>
<span id="cb258-3"><a href="or-loops-in-r-copy-alt.html#cb258-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>A for loop with seq_along(LETTERS) looks like this</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="or-loops-in-r-copy-alt.html#cb259-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(LETTERS)){</span>
<span id="cb259-2"><a href="or-loops-in-r-copy-alt.html#cb259-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(LETTERS[i])</span>
<span id="cb259-3"><a href="or-loops-in-r-copy-alt.html#cb259-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>You can also do math on the index generated by seq_along)9</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="or-loops-in-r-copy-alt.html#cb260-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(LETTERS)){</span>
<span id="cb260-2"><a href="or-loops-in-r-copy-alt.html#cb260-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(LETTERS[<span class="fu">c</span>(i, i<span class="sc">+</span><span class="dv">1</span>)])</span>
<span id="cb260-3"><a href="or-loops-in-r-copy-alt.html#cb260-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="advanced-aside-if-else-and-next-statements" class="section level2" number="10.7">
<h2><span class="header-section-number">10.7</span> Advanced aside: if(), else and next statements</h2>
<p>The following is a more advanced bit of R programming information and can be skipped</p>
<p>Its common to but if() statements within loop to screen for certain conditions. In the code I’ve been using which prints out 2 consecutive letters, when I get to i = 26, the code LETTERS[c(i, i+1)] produces LETTERS[c(26, 26+1)] = LETTERS[c(26, 27)]. There is no letter 27 so an NA is thrown. I could avoid this using an if statement</p>
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="or-loops-in-r-copy-alt.html#cb261-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(LETTERS)){</span>
<span id="cb261-2"><a href="or-loops-in-r-copy-alt.html#cb261-2" aria-hidden="true" tabindex="-1"></a> <span class="cf">if</span>(i<span class="sc">+</span><span class="dv">1</span> <span class="sc">&gt;</span> <span class="dv">26</span>){</span>
<span id="cb261-3"><a href="or-loops-in-r-copy-alt.html#cb261-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">print</span>(LETTERS[<span class="fu">c</span>(<span class="dv">26</span>,<span class="dv">26</span>)])</span>
<span id="cb261-4"><a href="or-loops-in-r-copy-alt.html#cb261-4" aria-hidden="true" tabindex="-1"></a> } <span class="cf">else</span></span>
<span id="cb261-5"><a href="or-loops-in-r-copy-alt.html#cb261-5" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(LETTERS[<span class="fu">c</span>(i, i<span class="sc">+</span><span class="dv">1</span>)])</span>
<span id="cb261-6"><a href="or-loops-in-r-copy-alt.html#cb261-6" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;B&quot;
## [1] &quot;B&quot; &quot;C&quot;
## [1] &quot;C&quot; &quot;D&quot;
## [1] &quot;D&quot; &quot;E&quot;
## [1] &quot;E&quot; &quot;F&quot;
## [1] &quot;F&quot; &quot;G&quot;
## [1] &quot;G&quot; &quot;H&quot;
## [1] &quot;H&quot; &quot;I&quot;
## [1] &quot;I&quot; &quot;J&quot;
## [1] &quot;J&quot; &quot;K&quot;
## [1] &quot;K&quot; &quot;L&quot;
## [1] &quot;L&quot; &quot;M&quot;
## [1] &quot;M&quot; &quot;N&quot;
## [1] &quot;N&quot; &quot;O&quot;
## [1] &quot;O&quot; &quot;P&quot;
## [1] &quot;P&quot; &quot;Q&quot;
## [1] &quot;Q&quot; &quot;R&quot;
## [1] &quot;R&quot; &quot;S&quot;
## [1] &quot;S&quot; &quot;T&quot;
## [1] &quot;T&quot; &quot;U&quot;
## [1] &quot;U&quot; &quot;V&quot;
## [1] &quot;V&quot; &quot;W&quot;
## [1] &quot;W&quot; &quot;X&quot;
## [1] &quot;X&quot; &quot;Y&quot;
## [1] &quot;Y&quot; &quot;Z&quot;
## [1] &quot;Z&quot; &quot;Z&quot;</code></pre>
<p>Instead of else I could also use next.</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="or-loops-in-r-copy-alt.html#cb263-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(LETTERS)){</span>
<span id="cb263-2"><a href="or-loops-in-r-copy-alt.html#cb263-2" aria-hidden="true" tabindex="-1"></a> <span class="cf">if</span>(i<span class="sc">+</span><span class="dv">1</span> <span class="sc">&gt;</span> <span class="dv">26</span>){</span>
<span id="cb263-3"><a href="or-loops-in-r-copy-alt.html#cb263-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">print</span>(LETTERS[<span class="fu">c</span>(<span class="dv">26</span>,<span class="dv">26</span>)])</span>
<span id="cb263-4"><a href="or-loops-in-r-copy-alt.html#cb263-4" aria-hidden="true" tabindex="-1"></a>   <span class="cf">next</span></span>
<span id="cb263-5"><a href="or-loops-in-r-copy-alt.html#cb263-5" aria-hidden="true" tabindex="-1"></a> } </span>
<span id="cb263-6"><a href="or-loops-in-r-copy-alt.html#cb263-6" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(LETTERS[<span class="fu">c</span>(i, i<span class="sc">+</span><span class="dv">1</span>)])</span>
<span id="cb263-7"><a href="or-loops-in-r-copy-alt.html#cb263-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;B&quot;
## [1] &quot;B&quot; &quot;C&quot;
## [1] &quot;C&quot; &quot;D&quot;
## [1] &quot;D&quot; &quot;E&quot;
## [1] &quot;E&quot; &quot;F&quot;
## [1] &quot;F&quot; &quot;G&quot;
## [1] &quot;G&quot; &quot;H&quot;
## [1] &quot;H&quot; &quot;I&quot;
## [1] &quot;I&quot; &quot;J&quot;
## [1] &quot;J&quot; &quot;K&quot;
## [1] &quot;K&quot; &quot;L&quot;
## [1] &quot;L&quot; &quot;M&quot;
## [1] &quot;M&quot; &quot;N&quot;
## [1] &quot;N&quot; &quot;O&quot;
## [1] &quot;O&quot; &quot;P&quot;
## [1] &quot;P&quot; &quot;Q&quot;
## [1] &quot;Q&quot; &quot;R&quot;
## [1] &quot;R&quot; &quot;S&quot;
## [1] &quot;S&quot; &quot;T&quot;
## [1] &quot;T&quot; &quot;U&quot;
## [1] &quot;U&quot; &quot;V&quot;
## [1] &quot;V&quot; &quot;W&quot;
## [1] &quot;W&quot; &quot;X&quot;
## [1] &quot;X&quot; &quot;Y&quot;
## [1] &quot;Y&quot; &quot;Z&quot;
## [1] &quot;Z&quot; &quot;Z&quot;</code></pre>
</div>
<div id="on-avoiding-for-loops" class="section level2" number="10.8">
<h2><span class="header-section-number">10.8</span> On avoiding for loops</h2>
<p>The following section is for background. You will not be expected to write any of the code shown.</p>
<p>In many cases for() loops can be avoided in R. This is because R is designed to make actions on entire vectors, dataframes, and matrices easy. R also as a set of functions called apply(), and a package called purr, which allow you to avoid writing for loops directly. This relates (in part) to the concept of <strong>vectorization</strong> in R. For a deep dive in to some of these issues see <a href="https://www.noamross.net/archives/2014-04-16-vectorization-in-r-why/" class="uri">https://www.noamross.net/archives/2014-04-16-vectorization-in-r-why/</a>.</p>
<p>Here’s an example. Let’s say I had a DNA sequence stored in a vector, which each base in a seperate slot in the vector.</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="or-loops-in-r-copy-alt.html#cb265-1" aria-hidden="true" tabindex="-1"></a>a.sequence <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;T&quot;</span>,<span class="st">&quot;C&quot;</span>,<span class="st">&quot;A&quot;</span>,<span class="st">&quot;A&quot;</span>,<span class="st">&quot;A&quot;</span>,<span class="st">&quot;G&quot;</span>,<span class="st">&quot;G&quot;</span>,<span class="st">&quot;G&quot;</span>)</span></code></pre></div>
<p>What if for some reason I wanted these letters to be lower case. R has a handy function called tolower() which makes upper case tolower case</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="or-loops-in-r-copy-alt.html#cb266-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tolower</span>(<span class="st">&quot;A&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;a&quot;</code></pre>
<p>In some programming languages, to turn all of these letters to lower case you might have to do this (I’ve written this out with extra code to make it obvious what I’m doing)</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="or-loops-in-r-copy-alt.html#cb268-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(a.sequence)){</span>
<span id="cb268-2"><a href="or-loops-in-r-copy-alt.html#cb268-2" aria-hidden="true" tabindex="-1"></a>  lowercase.letter.i <span class="ot">&lt;-</span> <span class="fu">tolower</span>(a.sequence[i]) <span class="co">#make lower case</span></span>
<span id="cb268-3"><a href="or-loops-in-r-copy-alt.html#cb268-3" aria-hidden="true" tabindex="-1"></a>  a.sequence[i] <span class="ot">&lt;-</span> lowercase.letter.i          <span class="co">#overwrite old entry</span></span>
<span id="cb268-4"><a href="or-loops-in-r-copy-alt.html#cb268-4" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>In R (and perhaps some other languages) I can do this</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="or-loops-in-r-copy-alt.html#cb269-1" aria-hidden="true" tabindex="-1"></a>a.sequence <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;T&quot;</span>,<span class="st">&quot;C&quot;</span>,<span class="st">&quot;A&quot;</span>,<span class="st">&quot;A&quot;</span>,<span class="st">&quot;A&quot;</span>,<span class="st">&quot;G&quot;</span>,<span class="st">&quot;G&quot;</span>,<span class="st">&quot;G&quot;</span>)</span>
<span id="cb269-2"><a href="or-loops-in-r-copy-alt.html#cb269-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tolower</span>(a.sequence)</span></code></pre></div>
<pre><code>## [1] &quot;a&quot; &quot;t&quot; &quot;c&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;g&quot; &quot;g&quot; &quot;g&quot;</code></pre>
<p>This is a somewhat trivial example. Let’s say I have a matrix of DNA sequences, with each row a different sequence.</p>
<p>First, I’ll make up some data. Don’t worry about what this is doing</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="or-loops-in-r-copy-alt.html#cb271-1" aria-hidden="true" tabindex="-1"></a>dna <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;T&quot;</span>,<span class="st">&quot;C&quot;</span>,<span class="st">&quot;G&quot;</span>)</span>
<span id="cb271-2"><a href="or-loops-in-r-copy-alt.html#cb271-2" aria-hidden="true" tabindex="-1"></a>make_seq <span class="ot">&lt;-</span> <span class="cf">function</span>() <span class="fu">sample</span>(<span class="at">x =</span> dna,<span class="at">size =</span> <span class="dv">5</span>,<span class="at">replace =</span> T)</span>
<span id="cb271-3"><a href="or-loops-in-r-copy-alt.html#cb271-3" aria-hidden="true" tabindex="-1"></a>my.matrix <span class="ot">&lt;-</span> <span class="fu">rbind</span>(<span class="fu">make_seq</span>(),<span class="fu">make_seq</span>(),<span class="fu">make_seq</span>(),<span class="fu">make_seq</span>(),<span class="fu">make_seq</span>())</span></code></pre></div>
<p>The matrix looks like this</p>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="or-loops-in-r-copy-alt.html#cb272-1" aria-hidden="true" tabindex="-1"></a>my.matrix</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,] &quot;T&quot;  &quot;A&quot;  &quot;T&quot;  &quot;C&quot;  &quot;T&quot; 
## [2,] &quot;C&quot;  &quot;C&quot;  &quot;C&quot;  &quot;A&quot;  &quot;T&quot; 
## [3,] &quot;T&quot;  &quot;C&quot;  &quot;C&quot;  &quot;C&quot;  &quot;T&quot; 
## [4,] &quot;T&quot;  &quot;G&quot;  &quot;C&quot;  &quot;G&quot;  &quot;T&quot; 
## [5,] &quot;A&quot;  &quot;T&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;</code></pre>
<p>In some programming languages I’d have to write a loop to change this to lower case. In R I can use a fancy function called apply() to do it in one step</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="or-loops-in-r-copy-alt.html#cb274-1" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(my.matrix,<span class="dv">1</span>,tolower)</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,] &quot;t&quot;  &quot;c&quot;  &quot;t&quot;  &quot;t&quot;  &quot;a&quot; 
## [2,] &quot;a&quot;  &quot;c&quot;  &quot;c&quot;  &quot;g&quot;  &quot;t&quot; 
## [3,] &quot;t&quot;  &quot;c&quot;  &quot;c&quot;  &quot;c&quot;  &quot;c&quot; 
## [4,] &quot;c&quot;  &quot;a&quot;  &quot;c&quot;  &quot;g&quot;  &quot;t&quot; 
## [5,] &quot;t&quot;  &quot;t&quot;  &quot;t&quot;  &quot;t&quot;  &quot;t&quot;</code></pre>
</div>
<div id="challenge" class="section level2" number="10.9">
<h2><span class="header-section-number">10.9</span> Challenge</h2>
<p>Create a vector that contains the four codes for the DNA bases, A, T, C, G. Write a for() loop which prints these out.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="working-with-square-brackets.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="function-practice.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/_book60_programming-misc/prog-assignment-for_loops.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/_book60_programming-misc/prog-assignment-for_loops.Rmd",
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
