<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 56 Phylogenetic trees using the UPGMA clustering algoirthm | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 56 Phylogenetic trees using the UPGMA clustering algoirthm | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 56 Phylogenetic trees using the UPGMA clustering algoirthm | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-08-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="plotting-trees-from-newick-notation-in-r.html"/>
<link rel="next" href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.5.1/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>1</b> How to review this book</a>
<ul>
<li class="chapter" data-level="1.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>1.1</b> File format</a></li>
<li class="chapter" data-level="1.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>1.2</b> What can you do?</a></li>
<li class="chapter" data-level="1.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>1.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>1.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="1.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>1.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="1.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>1.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="1.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>1.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="test-audio-link.html"><a href="test-audio-link.html"><i class="fa fa-check"></i><b>2</b> Test audio link</a></li>
<li class="chapter" data-level="3" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>3</b> Test shiny app</a></li>
<li class="chapter" data-level="4" data-path="test-youtube-vidoe.html"><a href="test-youtube-vidoe.html"><i class="fa fa-check"></i><b>4</b> Test YouTube vidoe</a></li>
<li class="chapter" data-level="5" data-path="test-google-doc-insertion.html"><a href="test-google-doc-insertion.html"><i class="fa fa-check"></i><b>5</b> Test google doc insertion</a></li>
<li class="chapter" data-level="6" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>6</b> Datacamp test</a></li>
<li class="chapter" data-level="7" data-path="an-example-of-a-plotly-graph.html"><a href="an-example-of-a-plotly-graph.html"><i class="fa fa-check"></i><b>7</b> An example of a plotly graph</a></li>
<li class="chapter" data-level="8" data-path="directly-embedding-of-a-libretext-page.html"><a href="directly-embedding-of-a-libretext-page.html"><i class="fa fa-check"></i><b>8</b> Directly embedding of a libretext page</a></li>
<li class="chapter" data-level="9" data-path="libretext-image.html"><a href="libretext-image.html"><i class="fa fa-check"></i><b>9</b> Libretext image</a>
<ul>
<li class="chapter" data-level="9.1" data-path="libretext-image.html"><a href="libretext-image.html#prelude-to-population-and-community-ecology"><i class="fa fa-check"></i><b>9.1</b> 45.0: Prelude to Population and Community Ecology</a></li>
<li class="chapter" data-level="9.2" data-path="libretext-image.html"><a href="libretext-image.html#note"><i class="fa fa-check"></i><b>9.2</b> Note</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="nucleic-acids.html"><a href="nucleic-acids.html"><i class="fa fa-check"></i><b>10</b> Nucleic Acids</a>
<ul>
<li class="chapter" data-level="10.1" data-path="nucleic-acids.html"><a href="nucleic-acids.html#dna-and-rna"><i class="fa fa-check"></i><b>10.1</b> DNA and RNA</a></li>
<li class="chapter" data-level="10.2" data-path="nucleic-acids.html"><a href="nucleic-acids.html#dna-double-helix-structure"><i class="fa fa-check"></i><b>10.2</b> DNA Double-Helix Structure</a></li>
<li class="chapter" data-level="10.3" data-path="nucleic-acids.html"><a href="nucleic-acids.html#rna"><i class="fa fa-check"></i><b>10.3</b> RNA</a></li>
<li class="chapter" data-level="10.4" data-path="nucleic-acids.html"><a href="nucleic-acids.html#summary"><i class="fa fa-check"></i><b>10.4</b> Summary</a></li>
<li class="chapter" data-level="10.5" data-path="nucleic-acids.html"><a href="nucleic-acids.html#analysis-questions"><i class="fa fa-check"></i><b>10.5</b> Analysis Questions</a></li>
<li class="chapter" data-level="10.6" data-path="nucleic-acids.html"><a href="nucleic-acids.html#glossary"><i class="fa fa-check"></i><b>10.6</b> Glossary</a></li>
<li class="chapter" data-level="10.7" data-path="nucleic-acids.html"><a href="nucleic-acids.html#contributors-and-attributions"><i class="fa fa-check"></i><b>10.7</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="proteins.html"><a href="proteins.html"><i class="fa fa-check"></i><b>11</b> Proteins</a>
<ul>
<li class="chapter" data-level="11.0.1" data-path="proteins.html"><a href="proteins.html#skills-to-develop"><i class="fa fa-check"></i><b>11.0.1</b> Skills to develop:</a></li>
<li class="chapter" data-level="11.1" data-path="proteins.html"><a href="proteins.html#types-and-functions-of-proteins"><i class="fa fa-check"></i><b>11.1</b> Types and Functions of Proteins</a></li>
<li class="chapter" data-level="11.2" data-path="proteins.html"><a href="proteins.html#amino-acids"><i class="fa fa-check"></i><b>11.2</b> Amino Acids</a></li>
<li class="chapter" data-level="11.3" data-path="proteins.html"><a href="proteins.html#evolution-connection"><i class="fa fa-check"></i><b>11.3</b> Evolution Connection:</a></li>
<li class="chapter" data-level="11.4" data-path="proteins.html"><a href="proteins.html#protein-structure"><i class="fa fa-check"></i><b>11.4</b> Protein Structure</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="proteins.html"><a href="proteins.html#primary-structure"><i class="fa fa-check"></i><b>11.4.1</b> Primary Structure</a></li>
<li class="chapter" data-level="11.4.2" data-path="proteins.html"><a href="proteins.html#secondary-structure"><i class="fa fa-check"></i><b>11.4.2</b> Secondary Structure</a></li>
<li class="chapter" data-level="11.4.3" data-path="proteins.html"><a href="proteins.html#tertiary-structure"><i class="fa fa-check"></i><b>11.4.3</b> Tertiary Structure</a></li>
<li class="chapter" data-level="11.4.4" data-path="proteins.html"><a href="proteins.html#quaternary-structure"><i class="fa fa-check"></i><b>11.4.4</b> Quaternary Structure</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="proteins.html"><a href="proteins.html#denaturation-and-protein-folding"><i class="fa fa-check"></i><b>11.5</b> Denaturation and Protein Folding</a></li>
<li class="chapter" data-level="11.6" data-path="proteins.html"><a href="proteins.html#summary-1"><i class="fa fa-check"></i><b>11.6</b> Summary</a></li>
<li class="chapter" data-level="11.7" data-path="proteins.html"><a href="proteins.html#art-connections"><i class="fa fa-check"></i><b>11.7</b> Art Connections</a></li>
<li class="chapter" data-level="11.8" data-path="proteins.html"><a href="proteins.html#review-questions"><i class="fa fa-check"></i><b>11.8</b> Review Questions</a></li>
<li class="chapter" data-level="11.9" data-path="proteins.html"><a href="proteins.html#free-response"><i class="fa fa-check"></i><b>11.9</b> Free Response</a></li>
<li class="chapter" data-level="11.10" data-path="proteins.html"><a href="proteins.html#glossary-1"><i class="fa fa-check"></i><b>11.10</b> Glossary</a></li>
<li class="chapter" data-level="11.11" data-path="proteins.html"><a href="proteins.html#contributors-and-attributions-1"><i class="fa fa-check"></i><b>11.11</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>12</b> Downloading R</a>
<ul>
<li class="chapter" data-level="12.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>12.1</b> Preface</a></li>
<li class="chapter" data-level="12.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>12.2</b> Introduction to R</a></li>
<li class="chapter" data-level="12.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>12.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>12.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="12.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>12.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>12.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>13</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="13.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>13.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="13.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>13.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>14</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="14.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>14.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="14.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>14.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="14.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>14.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>15</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="15.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>15.1</b> Bioconductor</a></li>
<li class="chapter" data-level="15.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>15.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="15.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>15.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="15.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>15.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="15.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>15.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="15.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>15.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="15.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>15.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>16</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="16.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>16.1</b> Vocabulary</a></li>
<li class="chapter" data-level="16.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>16.2</b> R functions</a></li>
<li class="chapter" data-level="16.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>16.3</b> Interacting with R</a></li>
<li class="chapter" data-level="16.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>16.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="16.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>16.4.1</b> Vectors</a></li>
<li class="chapter" data-level="16.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>16.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="16.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>16.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="16.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>16.4.4</b> Lists</a></li>
<li class="chapter" data-level="16.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>16.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="16.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>16.5</b> Arguments</a></li>
<li class="chapter" data-level="16.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>16.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="16.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>16.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="16.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>16.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="16.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>16.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="16.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>16.9</b> Quiting R</a></li>
<li class="chapter" data-level="16.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>16.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="a-primer-for-working-with-vecotrs.html"><a href="a-primer-for-working-with-vecotrs.html"><i class="fa fa-check"></i><b>17</b> A primer for working with vecotrs</a>
<ul>
<li class="chapter" data-level="17.1" data-path="a-primer-for-working-with-vecotrs.html"><a href="a-primer-for-working-with-vecotrs.html#preface-1"><i class="fa fa-check"></i><b>17.1</b> Preface</a></li>
<li class="chapter" data-level="17.2" data-path="a-primer-for-working-with-vecotrs.html"><a href="a-primer-for-working-with-vecotrs.html#vocab"><i class="fa fa-check"></i><b>17.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>18</b> Functions</a>
<ul>
<li class="chapter" data-level="18.1" data-path="functions.html"><a href="functions.html#vectors-in-r"><i class="fa fa-check"></i><b>18.1</b> Vectors in R</a></li>
<li class="chapter" data-level="18.2" data-path="functions.html"><a href="functions.html#math-on-vectors"><i class="fa fa-check"></i><b>18.2</b> Math on vectors</a></li>
<li class="chapter" data-level="18.3" data-path="functions.html"><a href="functions.html#functions-on-vectors"><i class="fa fa-check"></i><b>18.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="18.4" data-path="functions.html"><a href="functions.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>18.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="18.5" data-path="functions.html"><a href="functions.html#subsetting-vectors"><i class="fa fa-check"></i><b>18.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="18.6" data-path="functions.html"><a href="functions.html#sequences-of-numbers"><i class="fa fa-check"></i><b>18.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="18.7" data-path="functions.html"><a href="functions.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>18.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="18.8" data-path="functions.html"><a href="functions.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>18.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>19</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="19.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-2"><i class="fa fa-check"></i><b>19.1</b> Preface</a></li>
<li class="chapter" data-level="19.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>19.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="19.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>19.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html"><i class="fa fa-check"></i><b>20</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="20.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#commands-used"><i class="fa fa-check"></i><b>20.1</b> Commands used</a></li>
<li class="chapter" data-level="20.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#r-objects"><i class="fa fa-check"></i><b>20.2</b> R Objects</a></li>
<li class="chapter" data-level="20.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>20.3</b> Differences between objects</a></li>
<li class="chapter" data-level="20.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-data"><i class="fa fa-check"></i><b>20.4</b> The Data</a></li>
<li class="chapter" data-level="20.5" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>20.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="20.5.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#is"><i class="fa fa-check"></i><b>20.5.1</b> is()</a></li>
<li class="chapter" data-level="20.5.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#length"><i class="fa fa-check"></i><b>20.5.2</b> length()</a></li>
<li class="chapter" data-level="20.5.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#str"><i class="fa fa-check"></i><b>20.5.3</b> str()</a></li>
<li class="chapter" data-level="20.5.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#c"><i class="fa fa-check"></i><b>20.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="20.6" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#debrief"><i class="fa fa-check"></i><b>20.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="fasta-files.html"><a href="fasta-files.html"><i class="fa fa-check"></i><b>21</b> FASTA Files</a>
<ul>
<li class="chapter" data-level="21.1" data-path="fasta-files.html"><a href="fasta-files.html#example-fasta-file"><i class="fa fa-check"></i><b>21.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="21.2" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>21.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="21.3" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>21.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="21.4" data-path="fasta-files.html"><a href="fasta-files.html#fastq-format"><i class="fa fa-check"></i><b>21.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>22</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="22.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-1"><i class="fa fa-check"></i><b>22.0.1</b> Functions</a></li>
<li class="chapter" data-level="22.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>22.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="22.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>22.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="22.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>22.0.4</b> File types</a></li>
<li class="chapter" data-level="22.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>22.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="22.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>22.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="22.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries"><i class="fa fa-check"></i><b>22.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="22.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>22.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="22.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>22.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="22.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>22.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="22.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>22.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="22.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>22.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="22.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>22.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="22.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>22.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="22.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>22.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>23</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="23.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-1"><i class="fa fa-check"></i><b>23.1</b> Preliminaries</a></li>
<li class="chapter" data-level="23.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>23.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="23.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>23.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="23.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>23.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="24" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>24</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="24.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-3"><i class="fa fa-check"></i><b>24.1</b> Preface</a></li>
<li class="chapter" data-level="24.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>24.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="24.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>24.3</b> Introduction</a></li>
<li class="chapter" data-level="24.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>24.4</b> Vocabulary</a></li>
<li class="chapter" data-level="24.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions-2"><i class="fa fa-check"></i><b>24.5</b> Functions</a></li>
<li class="chapter" data-level="24.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries-2"><i class="fa fa-check"></i><b>24.6</b> Preliminaries</a></li>
<li class="chapter" data-level="24.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>24.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="24.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>24.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="24.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>24.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="24.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>24.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="24.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>24.8.3</b> DNA words</a></li>
<li class="chapter" data-level="24.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary-2"><i class="fa fa-check"></i><b>24.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="24.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>24.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="24.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>24.9.1</b> License</a></li>
<li class="chapter" data-level="24.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>24.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="25" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>25</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="25.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>25.1</b> Preliminaries</a></li>
<li class="chapter" data-level="25.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>25.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="changes-to-database-entries.html"><a href="changes-to-database-entries.html"><i class="fa fa-check"></i><b>26</b> Changes to database entries</a></li>
<li class="chapter" data-level="27" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>27</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="27.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>27.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="27.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>27.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="27.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>27.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="27.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>27.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="27.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>27.4</b> Inversions</a></li>
<li class="chapter" data-level="27.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>27.5</b> Translocations</a></li>
<li class="chapter" data-level="27.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>27.6</b> Random sequence</a></li>
<li class="chapter" data-level="27.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>27.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="28" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>28</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="28.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>28.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="28.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>28.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="28.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>28.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="28.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>28.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="28.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>28.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="28.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>28.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="28.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>28.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="28.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>28.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>29</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="29.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>29.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="29.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>29.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="29.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>29.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html"><i class="fa fa-check"></i><b>30</b> Alignment by eye in R</a>
<ul>
<li class="chapter" data-level="30.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>30.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="30.1.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#packages"><i class="fa fa-check"></i><b>30.1.1</b> Packages</a></li>
<li class="chapter" data-level="30.1.2" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#data"><i class="fa fa-check"></i><b>30.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="31" data-path="alignment-in-r.html"><a href="alignment-in-r.html"><i class="fa fa-check"></i><b>31</b> Alignment in R</a>
<ul>
<li class="chapter" data-level="31.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#preliminaries-8"><i class="fa fa-check"></i><b>31.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="31.1.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#packages-1"><i class="fa fa-check"></i><b>31.1.1</b> Packages</a></li>
<li class="chapter" data-level="31.1.2" data-path="alignment-in-r.html"><a href="alignment-in-r.html#data-1"><i class="fa fa-check"></i><b>31.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="32" data-path="working-with-vectors-and-matrices.html"><a href="working-with-vectors-and-matrices.html"><i class="fa fa-check"></i><b>32</b> Working with vectors and matrices</a></li>
<li class="chapter" data-level="33" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html"><i class="fa fa-check"></i><b>33</b> Matrix vocabulary</a>
<ul>
<li class="chapter" data-level="33.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#introduction-1"><i class="fa fa-check"></i><b>33.1</b> Introduction</a></li>
<li class="chapter" data-level="33.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#preliminaries-9"><i class="fa fa-check"></i><b>33.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="33.2.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#packages-2"><i class="fa fa-check"></i><b>33.2.1</b> Packages</a></li>
<li class="chapter" data-level="33.2.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#vocab-1"><i class="fa fa-check"></i><b>33.2.2</b> Vocab</a></li>
<li class="chapter" data-level="33.2.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#functions-used"><i class="fa fa-check"></i><b>33.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="33.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#load-packages"><i class="fa fa-check"></i><b>33.3</b> Load packages</a></li>
<li class="chapter" data-level="33.4" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#matrix-structure"><i class="fa fa-check"></i><b>33.4</b> Matrix structure</a></li>
<li class="chapter" data-level="33.5" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#diagonal-of-a-matrix"><i class="fa fa-check"></i><b>33.5</b> Diagonal of a matrix</a></li>
</ul></li>
<li class="chapter" data-level="34" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html"><i class="fa fa-check"></i><b>34</b> Testing the significance of an alignment</a>
<ul>
<li class="chapter" data-level="34.1" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#calculating-the-statistical-significance-of-a-pairwise-global-alignment"><i class="fa fa-check"></i><b>34.1</b> Calculating the statistical significance of a pairwise global alignment</a></li>
<li class="chapter" data-level="34.2" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#summary-3"><i class="fa fa-check"></i><b>34.2</b> Summary</a></li>
<li class="chapter" data-level="34.3" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#links-and-further-reading-1"><i class="fa fa-check"></i><b>34.3</b> Links and Further Reading</a></li>
<li class="chapter" data-level="34.4" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#exercises-1"><i class="fa fa-check"></i><b>34.4</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>35</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="35.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>35.1</b> Prelminaries</a></li>
<li class="chapter" data-level="35.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>35.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="35.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>35.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html"><i class="fa fa-check"></i><b>36</b> Manipulating matrices and vectors: worked example</a>
<ul>
<li class="chapter" data-level="36.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#introduction-2"><i class="fa fa-check"></i><b>36.1</b> Introduction</a></li>
<li class="chapter" data-level="36.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#preliminaries-10"><i class="fa fa-check"></i><b>36.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="36.2.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-3"><i class="fa fa-check"></i><b>36.2.1</b> Packages</a></li>
<li class="chapter" data-level="36.2.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#vocab-2"><i class="fa fa-check"></i><b>36.2.2</b> Vocab</a></li>
<li class="chapter" data-level="36.2.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#functions-used-1"><i class="fa fa-check"></i><b>36.2.3</b> Functions used</a></li>
<li class="chapter" data-level="36.2.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-4"><i class="fa fa-check"></i><b>36.2.4</b> Packages</a></li>
<li class="chapter" data-level="36.2.5" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#data-2"><i class="fa fa-check"></i><b>36.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="36.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#matrix-elements-can-be-accessed-with-square-bracket-notation"><i class="fa fa-check"></i><b>36.3</b> Matrix elements can be accessed with square bracket notation</a></li>
<li class="chapter" data-level="36.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#r-object-names"><i class="fa fa-check"></i><b>36.4</b> R object names</a>
<ul>
<li class="chapter" data-level="36.4.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-vectors"><i class="fa fa-check"></i><b>36.4.1</b> Accessing items in names vectors</a></li>
<li class="chapter" data-level="36.4.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-matrices"><i class="fa fa-check"></i><b>36.4.2</b> Accessing items in names matrices</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="37" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>37</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="37.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-11"><i class="fa fa-check"></i><b>37.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="37.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages-5"><i class="fa fa-check"></i><b>37.1.1</b> Packages</a></li>
<li class="chapter" data-level="37.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-3"><i class="fa fa-check"></i><b>37.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="37.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>37.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="37.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>37.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="37.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>37.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="37.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>37.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="37.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>37.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="37.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>37.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="37.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>37.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="37.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>37.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="38" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html"><i class="fa fa-check"></i><b>38</b> The BLOSUM scoring matrix in R</a>
<ul>
<li class="chapter" data-level="38.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#introduction-3"><i class="fa fa-check"></i><b>38.1</b> Introduction</a></li>
<li class="chapter" data-level="38.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#preliminaries-12"><i class="fa fa-check"></i><b>38.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="38.2.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#packages-6"><i class="fa fa-check"></i><b>38.2.1</b> Packages</a></li>
<li class="chapter" data-level="38.2.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#vocab-3"><i class="fa fa-check"></i><b>38.2.2</b> Vocab</a></li>
<li class="chapter" data-level="38.2.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#functions-used-2"><i class="fa fa-check"></i><b>38.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="38.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#load-packages-1"><i class="fa fa-check"></i><b>38.3</b> Load packages</a></li>
<li class="chapter" data-level="38.4" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#the-blosum-matrix"><i class="fa fa-check"></i><b>38.4</b> The BLOSUM matrix</a></li>
<li class="chapter" data-level="38.5" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#visualizing-the-blosum-matrix"><i class="fa fa-check"></i><b>38.5</b> Visualizing the BLOSUM matrix</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>39</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="40" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>40</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="40.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-4"><i class="fa fa-check"></i><b>40.1</b> Introduction</a></li>
<li class="chapter" data-level="40.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>40.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="40.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-4"><i class="fa fa-check"></i><b>40.2.1</b> Vocab</a></li>
<li class="chapter" data-level="40.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>40.2.2</b> R functions</a></li>
<li class="chapter" data-level="40.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>40.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="40.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>40.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="40.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>40.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="40.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>40.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="40.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>40.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="40.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>40.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="40.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>40.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="40.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>40.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="40.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>40.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="40.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>40.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="40.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>40.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="40.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>40.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="40.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>40.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="41" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>41</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="41.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-13"><i class="fa fa-check"></i><b>41.1</b> Preliminaries</a></li>
<li class="chapter" data-level="41.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-5"><i class="fa fa-check"></i><b>41.2</b> Introduction</a></li>
<li class="chapter" data-level="41.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>41.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="41.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>41.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="42" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>42</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="42.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-14"><i class="fa fa-check"></i><b>42.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="42.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>42.1.1</b> Key functions</a></li>
<li class="chapter" data-level="42.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>42.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="42.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>42.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="42.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>42.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="42.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>42.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="42.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>42.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="42.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>42.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="43" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html"><i class="fa fa-check"></i><b>43</b> Local variation in GC content</a>
<ul>
<li class="chapter" data-level="43.0.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#note-on-the-biology-in-this-section"><i class="fa fa-check"></i><b>43.0.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="43.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#vocabulary-2"><i class="fa fa-check"></i><b>43.1</b> Vocabulary</a></li>
<li class="chapter" data-level="43.2" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#reading-sequence-data-with-rentrezentrez_fetch"><i class="fa fa-check"></i><b>43.2</b> Reading sequence data with rentrez::entrez_fetch</a></li>
<li class="chapter" data-level="43.3" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#local-variation-in-gc-content-1"><i class="fa fa-check"></i><b>43.3</b> Local variation in GC content</a></li>
<li class="chapter" data-level="43.4" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-analysis-of-gc-content"><i class="fa fa-check"></i><b>43.4</b> A sliding window analysis of GC content</a></li>
<li class="chapter" data-level="43.5" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-plot-of-gc-content"><i class="fa fa-check"></i><b>43.5</b> A sliding window plot of GC content</a>
<ul>
<li class="chapter" data-level="43.5.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#over-represented-and-under-represented-dna-words"><i class="fa fa-check"></i><b>43.5.1</b> Over-represented and under-represented DNA words</a></li>
</ul></li>
<li class="chapter" data-level="43.6" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#summary-4"><i class="fa fa-check"></i><b>43.6</b> Summary</a></li>
<li class="chapter" data-level="43.7" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#further-reading"><i class="fa fa-check"></i><b>43.7</b> Further Reading</a></li>
<li class="chapter" data-level="43.8" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#acknowledgements-1"><i class="fa fa-check"></i><b>43.8</b> Acknowledgements</a></li>
<li class="chapter" data-level="43.9" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#exercises-2"><i class="fa fa-check"></i><b>43.9</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="44" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html"><i class="fa fa-check"></i><b>44</b> Computational gene finding</a>
<ul>
<li class="chapter" data-level="44.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#preliminaries-15"><i class="fa fa-check"></i><b>44.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="44.1.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#packages-7"><i class="fa fa-check"></i><b>44.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="44.2" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#the-genetic-code"><i class="fa fa-check"></i><b>44.2</b> The genetic code</a></li>
<li class="chapter" data-level="44.3" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-start-and-stop-codons-in-a-dna-sequence"><i class="fa fa-check"></i><b>44.3</b> Finding start and stop codons in a DNA sequence</a></li>
<li class="chapter" data-level="44.4" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#reading-frames"><i class="fa fa-check"></i><b>44.4</b> Reading frames</a></li>
<li class="chapter" data-level="44.5" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-forward-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>44.5</b> Finding open reading frames on the forward strand of a DNA sequence</a></li>
<li class="chapter" data-level="44.6" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#predicting-the-protein-sequence-for-an-orf"><i class="fa fa-check"></i><b>44.6</b> Predicting the protein sequence for an ORF</a></li>
<li class="chapter" data-level="44.7" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-reverse-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>44.7</b> Finding open reading frames on the reverse strand of a DNA sequence</a></li>
<li class="chapter" data-level="44.8" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#lengths-of-open-reading-frames"><i class="fa fa-check"></i><b>44.8</b> Lengths of open reading frames</a></li>
</ul></li>
<li class="chapter" data-level="45" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html"><i class="fa fa-check"></i><b>45</b> Significance of ORFs</a>
<ul>
<li class="chapter" data-level="45.1" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#preliminaries-16"><i class="fa fa-check"></i><b>45.1</b> Preliminaries</a></li>
<li class="chapter" data-level="45.2" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#identifying-significant-open-reading-frames"><i class="fa fa-check"></i><b>45.2</b> Identifying significant open reading frames</a></li>
<li class="chapter" data-level="45.3" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#summary-5"><i class="fa fa-check"></i><b>45.3</b> Summary</a></li>
<li class="chapter" data-level="45.4" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#links-and-further-reading-2"><i class="fa fa-check"></i><b>45.4</b> Links and Further Reading</a></li>
<li class="chapter" data-level="45.5" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#acknowledgements-2"><i class="fa fa-check"></i><b>45.5</b> Acknowledgements</a></li>
<li class="chapter" data-level="45.6" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#exercises-3"><i class="fa fa-check"></i><b>45.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="46" data-path="comparative-genomics.html"><a href="comparative-genomics.html"><i class="fa fa-check"></i><b>46</b> Comparative Genomics</a>
<ul>
<li class="chapter" data-level="46.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#preliminaries-17"><i class="fa fa-check"></i><b>46.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="46.1.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#packages-8"><i class="fa fa-check"></i><b>46.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="46.2" data-path="comparative-genomics.html"><a href="comparative-genomics.html#introduction-6"><i class="fa fa-check"></i><b>46.2</b> Introduction</a></li>
<li class="chapter" data-level="46.3" data-path="comparative-genomics.html"><a href="comparative-genomics.html#using-the-biomart-r-library-to-query-the-ensembl-database"><i class="fa fa-check"></i><b>46.3</b> Using the biomaRt R Library to Query the Ensembl Database</a></li>
<li class="chapter" data-level="46.4" data-path="comparative-genomics.html"><a href="comparative-genomics.html#comparing-the-number-of-genes-in-two-species"><i class="fa fa-check"></i><b>46.4</b> Comparing the number of genes in two species</a></li>
<li class="chapter" data-level="46.5" data-path="comparative-genomics.html"><a href="comparative-genomics.html#identifying-homologous-genes-between-two-species"><i class="fa fa-check"></i><b>46.5</b> Identifying homologous genes between two species</a></li>
</ul></li>
<li class="chapter" data-level="47" data-path="summary-6.html"><a href="summary-6.html"><i class="fa fa-check"></i><b>47</b> Summary</a></li>
<li class="chapter" data-level="48" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html"><i class="fa fa-check"></i><b>48</b> #Links and Further Reading</a>
<ul>
<li class="chapter" data-level="48.1" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#acknowledgements-3"><i class="fa fa-check"></i><b>48.1</b> Acknowledgements</a></li>
<li class="chapter" data-level="48.2" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#exercises-4"><i class="fa fa-check"></i><b>48.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="49" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html"><i class="fa fa-check"></i><b>49</b> Hidden Markov Models</a>
<ul>
<li class="chapter" data-level="49.1" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-multinomial-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>49.1</b> A multinomial model of DNA sequence evolution</a></li>
<li class="chapter" data-level="49.2" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-multinomial-model"><i class="fa fa-check"></i><b>49.2</b> Generating a DNA sequence using a multinomial model</a></li>
<li class="chapter" data-level="49.3" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>49.3</b> A Markov model of DNA sequence evolution</a></li>
<li class="chapter" data-level="49.4" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-for-a-markov-model"><i class="fa fa-check"></i><b>49.4</b> The transition matrix for a Markov model</a></li>
<li class="chapter" data-level="49.5" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-markov-model"><i class="fa fa-check"></i><b>49.5</b> Generating a DNA sequence using a Markov model</a></li>
<li class="chapter" data-level="49.6" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>49.6</b> A Hidden Markov Model of DNA sequence evolution</a></li>
<li class="chapter" data-level="49.7" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-and-emission-matrix-for-a-hmm"><i class="fa fa-check"></i><b>49.7</b> The transition matrix and emission matrix for a HMM</a></li>
<li class="chapter" data-level="49.8" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-hmm"><i class="fa fa-check"></i><b>49.8</b> Generating a DNA sequence using a HMM</a></li>
<li class="chapter" data-level="49.9" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#inferring-the-states-of-a-hmm-that-generated-a-dna-sequence"><i class="fa fa-check"></i><b>49.9</b> Inferring the states of a HMM that generated a DNA sequence</a></li>
<li class="chapter" data-level="49.10" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-protein-sequence-evolution"><i class="fa fa-check"></i><b>49.10</b> A Hidden Markov Model of protein sequence evolution</a></li>
<li class="chapter" data-level="49.11" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#summary-7"><i class="fa fa-check"></i><b>49.11</b> Summary</a></li>
<li class="chapter" data-level="49.12" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#links-and-further-reading-4"><i class="fa fa-check"></i><b>49.12</b> Links and Further Reading</a></li>
<li class="chapter" data-level="49.13" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#exercises-5"><i class="fa fa-check"></i><b>49.13</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="50" data-path="blast-summary.html"><a href="blast-summary.html"><i class="fa fa-check"></i><b>50</b> BLAST Summary</a>
<ul>
<li class="chapter" data-level="50.1" data-path="blast-summary.html"><a href="blast-summary.html#blast-overview"><i class="fa fa-check"></i><b>50.1</b> BLAST Overview</a>
<ul>
<li class="chapter" data-level="50.1.1" data-path="blast-summary.html"><a href="blast-summary.html#why-do-a-blast-search"><i class="fa fa-check"></i><b>50.1.1</b> Why do a BLAST search?</a></li>
<li class="chapter" data-level="50.1.2" data-path="blast-summary.html"><a href="blast-summary.html#what-does-blast-let-you-do"><i class="fa fa-check"></i><b>50.1.2</b> What does BLAST let you do?</a></li>
</ul></li>
<li class="chapter" data-level="50.2" data-path="blast-summary.html"><a href="blast-summary.html#blast-recipe"><i class="fa fa-check"></i><b>50.2</b> BLAST recipe</a></li>
<li class="chapter" data-level="50.3" data-path="blast-summary.html"><a href="blast-summary.html#modules-in-the-blast-workflow-and-algorithm"><i class="fa fa-check"></i><b>50.3</b> Modules in the BLAST Workflow and Algorithm</a>
<ul>
<li class="chapter" data-level="50.3.1" data-path="blast-summary.html"><a href="blast-summary.html#module-1-setup-query-sequence-protein-blast-oriented"><i class="fa fa-check"></i><b>50.3.1</b> Module 1: Setup query sequence (Protein BLAST oriented)</a></li>
<li class="chapter" data-level="50.3.2" data-path="blast-summary.html"><a href="blast-summary.html#module-2-scanning-database-for-high-scoring-words"><i class="fa fa-check"></i><b>50.3.2</b> Module 2: Scanning database for high-scoring words</a></li>
<li class="chapter" data-level="50.3.3" data-path="blast-summary.html"><a href="blast-summary.html#module-3-traceback"><i class="fa fa-check"></i><b>50.3.3</b> Module 3: Traceback</a></li>
</ul></li>
<li class="chapter" data-level="50.4" data-path="blast-summary.html"><a href="blast-summary.html#blast-alignment-statistics"><i class="fa fa-check"></i><b>50.4</b> BLAST alignment statistics</a></li>
</ul></li>
<li class="chapter" data-level="51" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html"><i class="fa fa-check"></i><b>51</b> BLAST Alignment score distribution</a>
<ul>
<li class="chapter" data-level="51.1" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#packages-9"><i class="fa fa-check"></i><b>51.1</b> Packages</a></li>
<li class="chapter" data-level="51.2" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#distributions-of-alignment-scores"><i class="fa fa-check"></i><b>51.2</b> Distributions of alignment scores</a></li>
<li class="chapter" data-level="51.3" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-data"><i class="fa fa-check"></i><b>51.3</b> Simulating data</a></li>
<li class="chapter" data-level="51.4" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#lots-of-data-in-biology-is-normal"><i class="fa fa-check"></i><b>51.4</b> Lots of data in biology is normal</a></li>
<li class="chapter" data-level="51.5" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-the-normal-distribution"><i class="fa fa-check"></i><b>51.5</b> Simulating the normal distribution</a></li>
<li class="chapter" data-level="51.6" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#the-extreme-value-distribution"><i class="fa fa-check"></i><b>51.6</b> The extreme value distribution</a></li>
<li class="chapter" data-level="51.7" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#replicating-figure-4.14"><i class="fa fa-check"></i><b>51.7</b> Replicating Figure 4.14</a></li>
<li class="chapter" data-level="51.8" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#why-do-we-care"><i class="fa fa-check"></i><b>51.8</b> Why do we care?</a></li>
</ul></li>
<li class="chapter" data-level="52" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><i class="fa fa-check"></i><b>52</b> Calculating E values, Bit scores, and P-values from BLAST output</a>
<ul>
<li class="chapter" data-level="52.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-7"><i class="fa fa-check"></i><b>52.1</b> Introduction</a></li>
<li class="chapter" data-level="52.2" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#case-study"><i class="fa fa-check"></i><b>52.2</b> Case Study</a></li>
<li class="chapter" data-level="52.3" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-statistics"><i class="fa fa-check"></i><b>52.3</b> BLAST Statistics</a></li>
<li class="chapter" data-level="52.4" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot"><i class="fa fa-check"></i><b>52.4</b> Dotplot</a>
<ul>
<li class="chapter" data-level="52.4.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot-in-r---optional-experimental-code"><i class="fa fa-check"></i><b>52.4.1</b> Dotplot in R - optional, experimental code</a></li>
</ul></li>
<li class="chapter" data-level="52.5" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-search-summary"><i class="fa fa-check"></i><b>52.5</b> BLAST “Search Summary”</a></li>
<li class="chapter" data-level="52.6" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-alignments"><i class="fa fa-check"></i><b>52.6</b> BLAST Alignments</a></li>
<li class="chapter" data-level="52.7" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#equations"><i class="fa fa-check"></i><b>52.7</b> Equations</a>
<ul>
<li class="chapter" data-level="52.7.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-value-eq-4.5-pg-142"><i class="fa fa-check"></i><b>52.7.1</b> E value (eq 4.5, pg 142)</a></li>
</ul></li>
<li class="chapter" data-level="52.8" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bit-scores-s"><i class="fa fa-check"></i><b>52.8</b> Bit Scores (S’)</a></li>
<li class="chapter" data-level="52.9" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-from-bit-scores"><i class="fa fa-check"></i><b>52.9</b> E from Bit scores</a></li>
<li class="chapter" data-level="52.10" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#p-values"><i class="fa fa-check"></i><b>52.10</b> P values</a></li>
<li class="chapter" data-level="52.11" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#repliminaries"><i class="fa fa-check"></i><b>52.11</b> Repliminaries</a></li>
<li class="chapter" data-level="52.12" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#concepts"><i class="fa fa-check"></i><b>52.12</b> Concepts</a></li>
<li class="chapter" data-level="52.13" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#functions-4"><i class="fa fa-check"></i><b>52.13</b> Functions</a></li>
<li class="chapter" data-level="52.14" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-8"><i class="fa fa-check"></i><b>52.14</b> Introduction</a></li>
<li class="chapter" data-level="52.15" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees"><i class="fa fa-check"></i><b>52.15</b> Number of rooted trees</a></li>
<li class="chapter" data-level="52.16" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees-in-r"><i class="fa fa-check"></i><b>52.16</b> Number of rooted trees in R</a>
<ul>
<li class="chapter" data-level="52.16.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#unique-topologies-versus-rotations"><i class="fa fa-check"></i><b>52.16.1</b> Unique topologies versus rotations</a></li>
</ul></li>
<li class="chapter" data-level="52.17" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bilography"><i class="fa fa-check"></i><b>52.17</b> Bilography</a></li>
</ul></li>
<li class="chapter" data-level="53" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html"><i class="fa fa-check"></i><b>53</b> UPGMA the hard way</a>
<ul>
<li class="chapter" data-level="53.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#introudction-to-upgma"><i class="fa fa-check"></i><b>53.1</b> Introudction to UPGMA</a></li>
<li class="chapter" data-level="53.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#preliminaries-18"><i class="fa fa-check"></i><b>53.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="53.2.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#r-libraries"><i class="fa fa-check"></i><b>53.2.1</b> R Libraries</a></li>
<li class="chapter" data-level="53.2.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#custom-function-well-use"><i class="fa fa-check"></i><b>53.2.2</b> Custom function we’ll use</a></li>
</ul></li>
<li class="chapter" data-level="53.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#data-matrix"><i class="fa fa-check"></i><b>53.3</b> Data matrix</a>
<ul>
<li class="chapter" data-level="53.3.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#building-the-matrix"><i class="fa fa-check"></i><b>53.3.1</b> Building the matrix</a></li>
</ul></li>
<li class="chapter" data-level="53.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#what-does-this-distance-matrix-represent"><i class="fa fa-check"></i><b>53.4</b> What does this distance matrix represent?</a>
<ul>
<li class="chapter" data-level="53.4.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-pairwise-linear-distances"><i class="fa fa-check"></i><b>53.4.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="53.4.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-2d-map-of-linear-distances"><i class="fa fa-check"></i><b>53.4.2</b> Genetic distances conceptualized as 2D map of linear distances</a></li>
</ul></li>
<li class="chapter" data-level="53.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#from-linear-distances-to-tree-distances"><i class="fa fa-check"></i><b>53.5</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="53.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#the-upgma-algorithm"><i class="fa fa-check"></i><b>53.6</b> The UPGMA algorithm</a>
<ul>
<li class="chapter" data-level="53.6.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-1-find-the-2-taxa-that-are-currently-closest-together"><i class="fa fa-check"></i><b>53.6.1</b> Step 1: find the 2 taxa that are currently closest together</a></li>
<li class="chapter" data-level="53.6.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node"><i class="fa fa-check"></i><b>53.6.2</b> Step 2: calculate the branch lengths connect the taxa to their shared node</a></li>
<li class="chapter" data-level="53.6.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-determine-if-the-algorithm-is-done"><i class="fa fa-check"></i><b>53.6.3</b> Step 3a: determine if the algorithm is done</a></li>
<li class="chapter" data-level="53.6.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u"><i class="fa fa-check"></i><b>53.6.4</b> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</a></li>
<li class="chapter" data-level="53.6.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-4-determine-the-distance-from-u-to-the-other-remaining-clusters."><i class="fa fa-check"></i><b>53.6.5</b> Step 4: Determine the distance from u to the other remaining clusters.</a></li>
<li class="chapter" data-level="53.6.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-5-create-a-new-matrix"><i class="fa fa-check"></i><b>53.6.6</b> Step 5: Create a new matrix</a></li>
<li class="chapter" data-level="53.6.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa"><i class="fa fa-check"></i><b>53.6.7</b> Key step: calculating the distance from a clade to the remaining taxa</a></li>
</ul></li>
<li class="chapter" data-level="53.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-2d-representation"><i class="fa fa-check"></i><b>53.7</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="53.8" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---round-1"><i class="fa fa-check"></i><b>53.8</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="53.8.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#determine-minimum-value-in-current-matrix"><i class="fa fa-check"></i><b>53.8.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="53.8.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade"><i class="fa fa-check"></i><b>53.8.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="53.8.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length"><i class="fa fa-check"></i><b>53.8.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="53.8.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>53.8.4</b> Calculate distance from clade to all other points</a></li>
<li class="chapter" data-level="53.8.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-matrix-with-new-distance"><i class="fa fa-check"></i><b>53.8.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="53.9" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---next-round"><i class="fa fa-check"></i><b>53.9</b> Algorithm - next round</a>
<ul>
<li class="chapter" data-level="53.9.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-1"><i class="fa fa-check"></i><b>53.9.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="53.9.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-1"><i class="fa fa-check"></i><b>53.9.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="53.9.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#recalculate-distance-from-abe-to-all-remainign-taxa"><i class="fa fa-check"></i><b>53.9.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="53.9.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-the-matrix"><i class="fa fa-check"></i><b>53.9.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="53.10" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#next-iteration"><i class="fa fa-check"></i><b>53.10</b> Next iteration</a>
<ul>
<li class="chapter" data-level="53.10.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-lengths"><i class="fa fa-check"></i><b>53.10.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="53.10.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-2"><i class="fa fa-check"></i><b>53.10.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="53.10.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-2"><i class="fa fa-check"></i><b>53.10.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="53.10.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#distance-of-c-to-abe-and-d-to-abe"><i class="fa fa-check"></i><b>53.10.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="53.11" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finish-up"><i class="fa fa-check"></i><b>53.11</b> Finish up</a>
<ul>
<li class="chapter" data-level="53.11.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-3"><i class="fa fa-check"></i><b>53.11.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="53.11.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-3"><i class="fa fa-check"></i><b>53.11.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="53.12" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finalizing-branch-lengths"><i class="fa fa-check"></i><b>53.12</b> Finalizing branch lengths</a></li>
</ul></li>
<li class="chapter" data-level="54" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html"><i class="fa fa-check"></i><b>54</b> Representing phylogenetic trees in Newick format</a>
<ul>
<li class="chapter" data-level="54.1" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#vocab-5"><i class="fa fa-check"></i><b>54.1</b> Vocab</a></li>
<li class="chapter" data-level="54.2" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#introduction-9"><i class="fa fa-check"></i><b>54.2</b> Introduction</a></li>
<li class="chapter" data-level="54.3" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-three-taxa"><i class="fa fa-check"></i><b>54.3</b> Trees with three taxa</a></li>
<li class="chapter" data-level="54.4" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-four-taxa"><i class="fa fa-check"></i><b>54.4</b> Trees with four taxa</a></li>
</ul></li>
<li class="chapter" data-level="55" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html"><i class="fa fa-check"></i><b>55</b> Plotting trees from Newick notation in R</a>
<ul>
<li class="chapter" data-level="55.1" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#preliminaries-19"><i class="fa fa-check"></i><b>55.1</b> Preliminaries</a></li>
<li class="chapter" data-level="55.2" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#note-1"><i class="fa fa-check"></i><b>55.2</b> Note</a></li>
<li class="chapter" data-level="55.3" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick"><i class="fa fa-check"></i><b>55.3</b> 3 taxa tree in Newick</a></li>
<li class="chapter" data-level="55.4" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick-1"><i class="fa fa-check"></i><b>55.4</b> 4 taxa tree in Newick</a></li>
<li class="chapter" data-level="55.5" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-in-newick-format"><i class="fa fa-check"></i><b>55.5</b> 5 taxa in Newick Format</a></li>
<li class="chapter" data-level="55.6" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#many-taxa-tree"><i class="fa fa-check"></i><b>55.6</b> Many-taxa tree</a></li>
<li class="chapter" data-level="55.7" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#newick-and-branch-lengths"><i class="fa fa-check"></i><b>55.7</b> Newick and Branch lengths</a></li>
<li class="chapter" data-level="55.8" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#additional-reading"><i class="fa fa-check"></i><b>55.8</b> Additional Reading</a></li>
</ul></li>
<li class="chapter" data-level="56" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><i class="fa fa-check"></i><b>56</b> Phylogenetic trees using the UPGMA clustering algoirthm</a>
<ul>
<li class="chapter" data-level="56.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#introduction-10"><i class="fa fa-check"></i><b>56.1</b> Introduction</a></li>
<li class="chapter" data-level="56.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#preliminaries-20"><i class="fa fa-check"></i><b>56.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="56.2.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#vocab-6"><i class="fa fa-check"></i><b>56.2.1</b> Vocab</a></li>
<li class="chapter" data-level="56.2.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-r-functions"><i class="fa fa-check"></i><b>56.2.2</b> General R functions</a></li>
<li class="chapter" data-level="56.2.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#specific-functions"><i class="fa fa-check"></i><b>56.2.3</b> Specific functions</a></li>
</ul></li>
<li class="chapter" data-level="56.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#sequence-comparisons"><i class="fa fa-check"></i><b>56.3</b> Sequence comparisons</a></li>
<li class="chapter" data-level="56.4" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrices"><i class="fa fa-check"></i><b>56.4</b> Distance matrices</a></li>
<li class="chapter" data-level="56.5" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#from-a-distance-matrix-to-phylogenetic-tree"><i class="fa fa-check"></i><b>56.5</b> From a distance matrix to phylogenetic tree</a></li>
<li class="chapter" data-level="56.6" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-raw-sequences"><i class="fa fa-check"></i><b>56.6</b> Distance matrix in R from raw sequences</a></li>
<li class="chapter" data-level="56.7" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-normal-matrix"><i class="fa fa-check"></i><b>56.7</b> Distance matrix in R from normal matrix</a></li>
<li class="chapter" data-level="56.8" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#upgma-in-r"><i class="fa fa-check"></i><b>56.8</b> UPGMA in R</a></li>
<li class="chapter" data-level="56.9" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#how-upgma-clustering-works"><i class="fa fa-check"></i><b>56.9</b> How UPGMA clustering works</a>
<ul>
<li class="chapter" data-level="56.9.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#challenge"><i class="fa fa-check"></i><b>56.9.1</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="56.10" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#further-information"><i class="fa fa-check"></i><b>56.10</b> Further information</a>
<ul>
<li class="chapter" data-level="56.10.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-info"><i class="fa fa-check"></i><b>56.10.1</b> General info</a></li>
<li class="chapter" data-level="56.10.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#worked-examples"><i class="fa fa-check"></i><b>56.10.2</b> Worked examples:</a></li>
<li class="chapter" data-level="56.10.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#additional-notes"><i class="fa fa-check"></i><b>56.10.3</b> Additional notes:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="57" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><i class="fa fa-check"></i><b>57</b> Tutorial: Writing functions to calculate the number of phylogenetic trees</a>
<ul>
<li class="chapter" data-level="57.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#repliminaries-1"><i class="fa fa-check"></i><b>57.1</b> Repliminaries</a>
<ul>
<li class="chapter" data-level="57.1.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-5"><i class="fa fa-check"></i><b>57.1.1</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="57.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#concepts-1"><i class="fa fa-check"></i><b>57.2</b> Concepts</a>
<ul>
<li class="chapter" data-level="57.2.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vocab-7"><i class="fa fa-check"></i><b>57.2.1</b> Vocab:</a></li>
<li class="chapter" data-level="57.2.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#packages-10"><i class="fa fa-check"></i><b>57.2.2</b> Packages:</a></li>
</ul></li>
<li class="chapter" data-level="57.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#introduction-11"><i class="fa fa-check"></i><b>57.3</b> Introduction</a></li>
<li class="chapter" data-level="57.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-1"><i class="fa fa-check"></i><b>57.4</b> Number of rooted trees</a></li>
<li class="chapter" data-level="57.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-in-r-1"><i class="fa fa-check"></i><b>57.5</b> Number of rooted trees in R</a></li>
<li class="chapter" data-level="57.6" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-in-r"><i class="fa fa-check"></i><b>57.6</b> Functions in R</a></li>
<li class="chapter" data-level="57.7" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-to-calculate-the-number-of-possible-phylogenetic-trees"><i class="fa fa-check"></i><b>57.7</b> Function to calculate the number of possible phylogenetic trees</a></li>
<li class="chapter" data-level="57.8" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vectorized-inputs-to-functions"><i class="fa fa-check"></i><b>57.8</b> Vectorized inputs to functions</a></li>
<li class="chapter" data-level="57.9" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-conditional-statements"><i class="fa fa-check"></i><b>57.9</b> Adding conditional statements</a></li>
<li class="chapter" data-level="57.10" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-multiple-conditional-statements"><i class="fa fa-check"></i><b>57.10</b> Adding multiple conditional statements</a></li>
<li class="chapter" data-level="57.11" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-additional-arguments"><i class="fa fa-check"></i><b>57.11</b> Adding additional arguments</a></li>
<li class="chapter" data-level="57.12" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-_un_rooted-trees"><i class="fa fa-check"></i><b>57.12</b> Assignment: _un_rooted trees</a></li>
<li class="chapter" data-level="57.13" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-1"><i class="fa fa-check"></i><b>57.13</b> Assignment part 1</a>
<ul>
<li class="chapter" data-level="57.13.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#part-1-answer"><i class="fa fa-check"></i><b>57.13.1</b> Part 1 answer</a></li>
</ul></li>
<li class="chapter" data-level="57.14" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-2"><i class="fa fa-check"></i><b>57.14</b> Assignment part 2</a>
<ul>
<li class="chapter" data-level="57.14.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#answer"><i class="fa fa-check"></i><b>57.14.1</b> Answer</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="58" data-path="writing-functions.html"><a href="writing-functions.html"><i class="fa fa-check"></i><b>58</b> Writing functions</a>
<ul>
<li class="chapter" data-level="58.1" data-path="writing-functions.html"><a href="writing-functions.html#preface-4"><i class="fa fa-check"></i><b>58.1</b> Preface</a></li>
<li class="chapter" data-level="58.2" data-path="writing-functions.html"><a href="writing-functions.html#vocab-8"><i class="fa fa-check"></i><b>58.2</b> Vocab</a></li>
<li class="chapter" data-level="58.3" data-path="writing-functions.html"><a href="writing-functions.html#functions-6"><i class="fa fa-check"></i><b>58.3</b> Functions</a></li>
<li class="chapter" data-level="58.4" data-path="writing-functions.html"><a href="writing-functions.html#functions-in-r-1"><i class="fa fa-check"></i><b>58.4</b> Functions in R</a></li>
<li class="chapter" data-level="58.5" data-path="writing-functions.html"><a href="writing-functions.html#comments-in-r"><i class="fa fa-check"></i><b>58.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="59" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html"><i class="fa fa-check"></i><b>59</b> Writing functions - practice problems</a>
<ul>
<li class="chapter" data-level="" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#introduction-12"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="59.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#pure-practice"><i class="fa fa-check"></i><b>59.1</b> Pure practice</a>
<ul>
<li class="chapter" data-level="59.1.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#hello-world"><i class="fa fa-check"></i><b>59.1.1</b> Hello world!</a></li>
<li class="chapter" data-level="59.1.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______"><i class="fa fa-check"></i><b>59.1.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="59.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function"><i class="fa fa-check"></i><b>59.2</b> Practice Function 3: Make you own natural log function</a></li>
<li class="chapter" data-level="59.3" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-4-make-function-that-converts-dna-directly-to-rna"><i class="fa fa-check"></i><b>59.3</b> Practice Function 4: Make function that converts DNA directly to RNA</a></li>
<li class="chapter" data-level="59.4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#key"><i class="fa fa-check"></i><b>59.4</b> Key</a>
<ul>
<li class="chapter" data-level="59.4.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-1-hello-world"><i class="fa fa-check"></i><b>59.4.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="59.4.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______-1"><i class="fa fa-check"></i><b>59.4.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="59.5" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function-1"><i class="fa fa-check"></i><b>59.5</b> Practice Function 3: Make you own natural log function</a></li>
</ul></li>
<li class="chapter" data-level="60" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html"><i class="fa fa-check"></i><b>60</b> Simulating random amino acid sequences</a>
<ul>
<li class="chapter" data-level="60.1" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#selecting-random-letters-from-a-vector"><i class="fa fa-check"></i><b>60.1</b> Selecting random letters from a vector</a></li>
<li class="chapter" data-level="60.2" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-letters-from-the-whole-alphabet"><i class="fa fa-check"></i><b>60.2</b> Select random letters from the whole alphabet</a></li>
<li class="chapter" data-level="60.3" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-amino-acids-to-build-polypeptide"><i class="fa fa-check"></i><b>60.3</b> Select random amino acids to build polypeptide</a></li>
</ul></li>
<li class="chapter" data-level="61" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html"><i class="fa fa-check"></i><b>61</b> Writing user-friendly functions</a>
<ul>
<li class="chapter" data-level="" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#introduction-13"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="61.1" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-1"><i class="fa fa-check"></i><b>61.1</b> Version 1</a></li>
<li class="chapter" data-level="61.2" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-2"><i class="fa fa-check"></i><b>61.2</b> Version 2</a></li>
<li class="chapter" data-level="61.3" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-3"><i class="fa fa-check"></i><b>61.3</b> Version 3</a></li>
<li class="chapter" data-level="61.4" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-4-adding-conditions-and-warnigns"><i class="fa fa-check"></i><b>61.4</b> Version 4: Adding conditions and warnigns</a></li>
<li class="chapter" data-level="61.5" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-5"><i class="fa fa-check"></i><b>61.5</b> Version 5</a></li>
</ul></li>
<li class="chapter" data-level="62" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>62</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="62.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-5"><i class="fa fa-check"></i><b>62.1</b> Preface</a></li>
<li class="chapter" data-level="62.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab-9"><i class="fa fa-check"></i><b>62.2</b> Vocab</a></li>
<li class="chapter" data-level="62.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-7"><i class="fa fa-check"></i><b>62.3</b> Functions</a></li>
<li class="chapter" data-level="62.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>62.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="62.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>62.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="63" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html"><i class="fa fa-check"></i><b>63</b> More on for() loops in R</a>
<ul>
<li class="chapter" data-level="63.1" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#introduction-14"><i class="fa fa-check"></i><b>63.1</b> Introduction</a></li>
<li class="chapter" data-level="63.2" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#reminders"><i class="fa fa-check"></i><b>63.2</b> Reminders</a></li>
<li class="chapter" data-level="63.3" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop"><i class="fa fa-check"></i><b>63.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="63.4" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#for-loop-3-the-classic-for-loop"><i class="fa fa-check"></i><b>63.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="63.5" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#for-loop-3-the-power-user-for-loop"><i class="fa fa-check"></i><b>63.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="63.6" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#for-loop-4-a-for-loop-for-all-purposes"><i class="fa fa-check"></i><b>63.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="63.7" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#advanced-aside-if-else-and-next-statements"><i class="fa fa-check"></i><b>63.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="63.8" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#on-avoiding-for-loops"><i class="fa fa-check"></i><b>63.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="63.9" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#challenge-1"><i class="fa fa-check"></i><b>63.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="64" data-path="random-number-generation.html"><a href="random-number-generation.html"><i class="fa fa-check"></i><b>64</b> Random number generation</a>
<ul>
<li class="chapter" data-level="64.1" data-path="random-number-generation.html"><a href="random-number-generation.html#introduction-15"><i class="fa fa-check"></i><b>64.1</b> Introduction</a></li>
<li class="chapter" data-level="64.2" data-path="random-number-generation.html"><a href="random-number-generation.html#using-set.seed-to-make-simulations-reproducible"><i class="fa fa-check"></i><b>64.2</b> Using set.seed() to make simulations reproducible</a></li>
<li class="chapter" data-level="64.3" data-path="random-number-generation.html"><a href="random-number-generation.html#types-of-random-number-generators"><i class="fa fa-check"></i><b>64.3</b> Types of random number generators</a></li>
<li class="chapter" data-level="64.4" data-path="random-number-generation.html"><a href="random-number-generation.html#reporting-simulations"><i class="fa fa-check"></i><b>64.4</b> Reporting simulations</a></li>
<li class="chapter" data-level="64.5" data-path="random-number-generation.html"><a href="random-number-generation.html#notes"><i class="fa fa-check"></i><b>64.5</b> Notes:</a></li>
</ul></li>
<li class="chapter" data-level="65" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>65</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="65.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>65.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="65.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>65.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="65.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>65.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="65.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>65.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="65.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>65.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-3"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="65.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>65.2</b> Help!</a>
<ul>
<li class="chapter" data-level="65.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>65.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="65.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>65.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="65.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>65.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="65.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>65.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="65.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>65.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="65.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>65.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="65.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>65.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="65.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>65.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="phylogenetic-trees-using-the-upgma-clustering-algoirthm" class="section level1" number="56">
<h1><span class="header-section-number">Chapter 56</span> Phylogenetic trees using the UPGMA clustering algoirthm</h1>
<p><strong>By</strong>: Nathan Brouwer</p>
<div id="introduction-10" class="section level2" number="56.1">
<h2><span class="header-section-number">56.1</span> Introduction</h2>
<p><strong>UPGMA</strong> is a method for constructing <strong>phylogenetic trees</strong> using <strong>genetic distances</strong>. It is an old, outmoded method that is rarely ever used any more for building final versions of trees. Instead of UPGMA, the <strong>neighbor-joining</strong> (NJ) algorithm is used when a distance-based phylogeny is made. UPGMA’s relative simplicity, however, makes it a useful starting point for thinking about how to construct phylogenetic trees and code up algorithms to build them. UPGMA has also been used in the past in statistics, machine learning, transcriptomics, and community ecology as a general <strong>clustering algorithm</strong> and so is a useful starting point when understanding current algorithms in those fields.</p>
</div>
<div id="preliminaries-20" class="section level2" number="56.2">
<h2><span class="header-section-number">56.2</span> Preliminaries</h2>
<div class="sourceCode" id="cb1131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1131-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Potentially new packages</span></span>
<span id="cb1131-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1131-2" aria-hidden="true" tabindex="-1"></a><span class="do">## install.packages(&quot;phangorn&quot;)</span></span>
<span id="cb1131-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1131-3" aria-hidden="true" tabindex="-1"></a><span class="do">## install.packages(&quot;ape&quot;)</span></span>
<span id="cb1131-4"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1131-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1131-5"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1131-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb1131-6"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1131-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phangorn)</span></code></pre></div>
<div id="vocab-6" class="section level3" number="56.2.1">
<h3><span class="header-section-number">56.2.1</span> Vocab</h3>
<ul>
<li>matrix, matrices</li>
<li>diagonal of matrix</li>
<li>symmetrical matrix</li>
<li>lower triangle of matrix</li>
<li>Newick format</li>
<li>character string</li>
<li>function, argument</li>
<li>branch, edge, clade</li>
</ul>
</div>
<div id="general-r-functions" class="section level3" number="56.2.2">
<h3><span class="header-section-number">56.2.2</span> General R functions</h3>
<ul>
<li><code>c()</code></li>
<li><code>matrix()</code></li>
<li>colnames, rownames</li>
<li>cbind, rbind</li>
<li>class</li>
<li>as.dist()</li>
<li>par</li>
<li>min</li>
<li>str</li>
</ul>
</div>
<div id="specific-functions" class="section level3" number="56.2.3">
<h3><span class="header-section-number">56.2.3</span> Specific functions</h3>
<ul>
<li>phangorn::phangorn()</li>
<li>ape::read.tree</li>
<li>ape::as.DNAbin</li>
<li>ape::dist.dna()</li>
</ul>
</div>
</div>
<div id="sequence-comparisons" class="section level2" number="56.3">
<h2><span class="header-section-number">56.3</span> Sequence comparisons</h2>
<p>Let’s consider three short, hypothetical DNA sequences. We’ll put each into a <strong>vector</strong> using <code>c(...)</code>:</p>
<div class="sourceCode" id="cb1132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1132-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1132-1" aria-hidden="true" tabindex="-1"></a>human   <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;t&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;g&quot;</span>,<span class="st">&quot;a&quot;</span>,<span class="st">&quot;t&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;g&quot;</span>)</span>
<span id="cb1132-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1132-2" aria-hidden="true" tabindex="-1"></a>chimp   <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;t&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;a&quot;</span>,<span class="st">&quot;a&quot;</span>,<span class="st">&quot;t&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;a&quot;</span>)</span>
<span id="cb1132-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1132-3" aria-hidden="true" tabindex="-1"></a>gorilla <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;a&quot;</span>,<span class="st">&quot;a&quot;</span>,<span class="st">&quot;a&quot;</span>,<span class="st">&quot;a&quot;</span>,<span class="st">&quot;a&quot;</span>,<span class="st">&quot;a&quot;</span>,<span class="st">&quot;a&quot;</span>)</span></code></pre></div>
<p>The sequences are each 8 bases long.</p>
<div class="sourceCode" id="cb1133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1133-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(gorilla)</span></code></pre></div>
<pre><code>## [1] 8</code></pre>
<p>This means we’ll have 8 <strong>loci</strong>, or unique <strong>homologous positions</strong>, to consider.</p>
<p>We can examine each pair of sequences and count up the number of differences between each one. This gives us a preliminary estimate of how <strong>diverged</strong> the sequences are.<br />
Humans and chimps have 2 differences, at the 4th and 8th loci</p>
<div class="sourceCode" id="cb1135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1135-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rbind</span>(human, chimp)</span></code></pre></div>
<pre><code>##       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
## human &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;g&quot;  &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;g&quot; 
## chimp &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;a&quot;  &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;a&quot;</code></pre>
<p>Chimps and gorillas have 4 differences (2nd, 3rd, 6th, and 7th loci):</p>
<div class="sourceCode" id="cb1137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1137-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1137-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rbind</span>(chimp, gorilla)</span></code></pre></div>
<pre><code>##         [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
## chimp   &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;a&quot;  &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;a&quot; 
## gorilla &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;</code></pre>
<p>Humans and gorillas have 6 differences (all but 1st and 6th different)</p>
<div class="sourceCode" id="cb1139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1139-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rbind</span>(human, gorilla)</span></code></pre></div>
<pre><code>##         [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
## human   &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;g&quot;  &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;g&quot; 
## gorilla &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;</code></pre>
</div>
<div id="distance-matrices" class="section level2" number="56.4">
<h2><span class="header-section-number">56.4</span> Distance matrices</h2>
<p>Having counted up the number <strong>pairwise differences</strong> between each pair of sequences, we can represent this data as a genetic <strong>distance matrix</strong> like this.</p>
<p>First, make the matrix using the <code>matrix()</code> command.</p>
<div class="sourceCode" id="cb1141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1141-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1141-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make the matrix</span></span>
<span id="cb1141-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1141-2" aria-hidden="true" tabindex="-1"></a>dist_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>,</span>
<span id="cb1141-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1141-3" aria-hidden="true" tabindex="-1"></a>                            <span class="dv">2</span>, <span class="cn">NA</span>,<span class="cn">NA</span>,</span>
<span id="cb1141-4"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1141-4" aria-hidden="true" tabindex="-1"></a>                            <span class="dv">6</span>, <span class="dv">4</span>, <span class="cn">NA</span>),</span>
<span id="cb1141-5"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1141-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">nrow =</span> <span class="dv">3</span>, </span>
<span id="cb1141-6"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1141-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">byrow =</span> T)</span></code></pre></div>
<p>Now, label the rows and columns.</p>
<div class="sourceCode" id="cb1142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1142-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1142-1" aria-hidden="true" tabindex="-1"></a><span class="co"># label the matrix</span></span>
<span id="cb1142-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1142-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dist_mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;H&quot;</span>,<span class="st">&quot;C&quot;</span>,<span class="st">&quot;G&quot;</span>)</span>
<span id="cb1142-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1142-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(dist_mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;H&quot;</span>,<span class="st">&quot;C&quot;</span>,<span class="st">&quot;G&quot;</span>)</span></code></pre></div>
<p>The final matrix looks like this</p>
<div class="sourceCode" id="cb1143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1143-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1143-1" aria-hidden="true" tabindex="-1"></a><span class="co"># look at the output</span></span>
<span id="cb1143-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1143-2" aria-hidden="true" tabindex="-1"></a>dist_mat</span></code></pre></div>
<pre><code>##    H  C  G
## H NA NA NA
## C  2 NA NA
## G  6  4 NA</code></pre>
<p><strong>ASIDE: Building matrices</strong>:</p>
<p>Alternatively, we could make each row of the matrix into a vector and then stack them with <code>rbind()</code>.</p>
<div class="sourceCode" id="cb1145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1145-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make vectors</span></span>
<span id="cb1145-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-2" aria-hidden="true" tabindex="-1"></a>human.row <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">NA</span>,<span class="cn">NA</span>,<span class="cn">NA</span>)</span>
<span id="cb1145-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-3" aria-hidden="true" tabindex="-1"></a>chimp.row <span class="ot">&lt;-</span> <span class="fu">c</span>( <span class="dv">2</span>,<span class="cn">NA</span>,<span class="cn">NA</span>)</span>
<span id="cb1145-4"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-4" aria-hidden="true" tabindex="-1"></a>gorilla.row <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">6</span>,<span class="dv">4</span>,<span class="cn">NA</span>)</span>
<span id="cb1145-5"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1145-6"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-6" aria-hidden="true" tabindex="-1"></a><span class="co"># stack vectors by row using rbind()</span></span>
<span id="cb1145-7"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-7" aria-hidden="true" tabindex="-1"></a>dist_mat <span class="ot">&lt;-</span> <span class="fu">rbind</span>(human.row, </span>
<span id="cb1145-8"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-8" aria-hidden="true" tabindex="-1"></a>                  chimp.row, </span>
<span id="cb1145-9"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-9" aria-hidden="true" tabindex="-1"></a>                  gorilla.row)</span>
<span id="cb1145-10"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1145-11"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-11" aria-hidden="true" tabindex="-1"></a><span class="co"># add names</span></span>
<span id="cb1145-12"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-12" aria-hidden="true" tabindex="-1"></a>spp.names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;human&quot;</span>,<span class="st">&quot;chimp&quot;</span>,<span class="st">&quot;gorilla&quot;</span>)</span>
<span id="cb1145-13"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dist_mat) <span class="ot">&lt;-</span> spp.names</span>
<span id="cb1145-14"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(dist_mat) <span class="ot">&lt;-</span> spp.names</span>
<span id="cb1145-15"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1145-16"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-16" aria-hidden="true" tabindex="-1"></a><span class="co"># look at output</span></span>
<span id="cb1145-17"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1145-17" aria-hidden="true" tabindex="-1"></a>dist_mat</span></code></pre></div>
<pre><code>##         human chimp gorilla
## human      NA    NA      NA
## chimp       2    NA      NA
## gorilla     6     4      NA</code></pre>
<p><strong>End ASIDE</strong></p>
<p>Note that we leave the values on the <strong>diagonal</strong> of the matrix as <code>NA</code> because the genetic distance between humans and humans, or chimps and chimps, is 0 and isn’t helpful. A matrix like this is also <strong>symmetrical</strong>, where the <strong>lower triangle</strong> portion is the same as the upper portion We could build an <strong>upper triangular matrix</strong> like the one below if we wanted, but the convention is to use the lower.</p>
<div class="sourceCode" id="cb1147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1147-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-1" aria-hidden="true" tabindex="-1"></a><span class="co"># An upper triangular matrix</span></span>
<span id="cb1147-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-2" aria-hidden="true" tabindex="-1"></a><span class="do">## make data in vectors</span></span>
<span id="cb1147-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-3" aria-hidden="true" tabindex="-1"></a>human.row <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">NA</span>,<span class="dv">2</span>,<span class="dv">6</span>)</span>
<span id="cb1147-4"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-4" aria-hidden="true" tabindex="-1"></a>chimp.row <span class="ot">&lt;-</span> <span class="fu">c</span>( <span class="cn">NA</span>,<span class="cn">NA</span>,<span class="dv">4</span>)</span>
<span id="cb1147-5"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-5" aria-hidden="true" tabindex="-1"></a>gorilla.row <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">NA</span>,<span class="cn">NA</span>,<span class="cn">NA</span>)</span>
<span id="cb1147-6"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1147-7"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-7" aria-hidden="true" tabindex="-1"></a><span class="co"># stack with ribind()</span></span>
<span id="cb1147-8"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-8" aria-hidden="true" tabindex="-1"></a>dist_mat_upper <span class="ot">&lt;-</span> <span class="fu">rbind</span>(human.row, </span>
<span id="cb1147-9"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-9" aria-hidden="true" tabindex="-1"></a>                       chimp.row, </span>
<span id="cb1147-10"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-10" aria-hidden="true" tabindex="-1"></a>                       gorilla.row)</span>
<span id="cb1147-11"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1147-12"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-12" aria-hidden="true" tabindex="-1"></a><span class="co"># add names</span></span>
<span id="cb1147-13"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dist_mat_upper) <span class="ot">&lt;-</span> spp.names</span>
<span id="cb1147-14"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1147-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(dist_mat_upper) <span class="ot">&lt;-</span> spp.names</span></code></pre></div>
<p>Compare lower triangular matrix form of these data to the upper triangular for:</p>
<div class="sourceCode" id="cb1148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1148-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># lower triangular matrix:</span></span>
<span id="cb1148-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1148-2" aria-hidden="true" tabindex="-1"></a>dist_mat</span></code></pre></div>
<pre><code>##         human chimp gorilla
## human      NA    NA      NA
## chimp       2    NA      NA
## gorilla     6     4      NA</code></pre>
<div class="sourceCode" id="cb1150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1150-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1150-1" aria-hidden="true" tabindex="-1"></a><span class="co"># upper triangular matrix</span></span>
<span id="cb1150-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1150-2" aria-hidden="true" tabindex="-1"></a>dist_mat_upper</span></code></pre></div>
<pre><code>##         human chimp gorilla
## human      NA     2       6
## chimp      NA    NA       4
## gorilla    NA    NA      NA</code></pre>
<p>In both matrices the distance from humans to gorillas is 6 differences, humans to chimps is 2, and chimps to gorillas is 4. Make sure you understand why this is true.</p>
<p>Again, the lower triangular form is the form usually used. Because <code>NA</code> occurs in the entire top row and last column, people often omit it by dropping the row and column of all NAs.</p>
<div class="sourceCode" id="cb1152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1152-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1152-1" aria-hidden="true" tabindex="-1"></a>dist_mat[<span class="sc">-</span><span class="dv">1</span>, <span class="sc">-</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>##         human chimp
## chimp       2    NA
## gorilla     6     4</code></pre>
</div>
<div id="from-a-distance-matrix-to-phylogenetic-tree" class="section level2" number="56.5">
<h2><span class="header-section-number">56.5</span> From a distance matrix to phylogenetic tree</h2>
<p>These calculations imply that humans and chimps are more closely related (the sequences are made up, but this is true). Using data like this, <strong>[phylogenetic algorithms</strong> therefore cluster humans and chimps into a <strong>clade</strong>. In <strong>Newick Format</strong> this would be (H,C).</p>
<p>Since we have just three taxa, we add G outside the first clade: ((H,C),G).</p>
<p>We can plot a tree from Newick like this</p>
<div class="sourceCode" id="cb1154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1154-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1154-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define tree</span></span>
<span id="cb1154-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1154-2" aria-hidden="true" tabindex="-1"></a><span class="do">## </span><span class="al">NOTE</span><span class="do">: the string ends with a semi-colon!</span></span>
<span id="cb1154-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1154-3" aria-hidden="true" tabindex="-1"></a>str1<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;(G,(C, H) );&quot;</span>)</span>
<span id="cb1154-4"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1154-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1154-5"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1154-5" aria-hidden="true" tabindex="-1"></a><span class="do">## convert character string to tree</span></span>
<span id="cb1154-6"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1154-6" aria-hidden="true" tabindex="-1"></a>tree1<span class="ot">&lt;-</span>ape<span class="sc">::</span><span class="fu">read.tree</span>(<span class="at">text=</span>str1)</span>
<span id="cb1154-7"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1154-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1154-8"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1154-8" aria-hidden="true" tabindex="-1"></a><span class="do">## plot tree</span></span>
<span id="cb1154-9"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1154-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span>(tree1,<span class="at">edge.width=</span><span class="dv">2</span>, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-821-1.png" width="672" /></p>
<p>Newick format does not represent <strong>genetic distances</strong> and therefore the <strong>branch length</strong> of phylogenetic, only the clade structure and branching pattern of the tree. We can calculate branch lengths using the UPGMA algorithm.</p>
</div>
<div id="distance-matrix-in-r-from-raw-sequences" class="section level2" number="56.6">
<h2><span class="header-section-number">56.6</span> Distance matrix in R from raw sequences</h2>
<p>To make a distance matrix in R from the original sequences we first need to put our original sequences into a matrix with the <code>rbind()</code> command.</p>
<div class="sourceCode" id="cb1155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1155-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1155-1" aria-hidden="true" tabindex="-1"></a>my_seqs <span class="ot">&lt;-</span> <span class="fu">rbind</span>(human, </span>
<span id="cb1155-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1155-2" aria-hidden="true" tabindex="-1"></a>                 chimp, </span>
<span id="cb1155-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1155-3" aria-hidden="true" tabindex="-1"></a>                 gorilla)</span></code></pre></div>
<p>This is what it looks like</p>
<div class="sourceCode" id="cb1156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1156-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1156-1" aria-hidden="true" tabindex="-1"></a>my_seqs</span></code></pre></div>
<pre><code>##         [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
## human   &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;g&quot;  &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;g&quot; 
## chimp   &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;a&quot;  &quot;a&quot;  &quot;t&quot;  &quot;c&quot;  &quot;a&quot; 
## gorilla &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;  &quot;a&quot;</code></pre>
<p>We then need to do some fancy processing (this has to do with how the packages we’re going to use works, which is picky about formats because its optimized to accommodate real sequences which can be very very long). We’ll use the <code>as.DNAbin()</code> function.</p>
<div class="sourceCode" id="cb1158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1158-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1158-1" aria-hidden="true" tabindex="-1"></a>my_seqs_bin <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">as.DNAbin</span>(my_seqs)</span></code></pre></div>
<p>We then can calculate the <strong>distance</strong> between each sequence in terms of the number (N) of mutations. This uses the <code>dist.dna()</code> function. We set the argument <code>model = "N"</code>. This is the most basic and naive model of mutation and ignores the possibility of back mutations.</p>
<div class="sourceCode" id="cb1159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1159-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1159-1" aria-hidden="true" tabindex="-1"></a>seq_dist0 <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">dist.dna</span>(my_seqs_bin, <span class="at">model =</span> <span class="st">&quot;N&quot;</span>)</span></code></pre></div>
<p>This gives us a distance matrix</p>
<div class="sourceCode" id="cb1160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1160-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1160-1" aria-hidden="true" tabindex="-1"></a>seq_dist0</span></code></pre></div>
<pre><code>##         human chimp
## chimp       2      
## gorilla     6     4</code></pre>
<p>Note that there is a subtle difference in how R looks at the matrix we made earlier with the matrix() command and this matrix.</p>
<p>Them matrices have the same data, but slightly different formats.</p>
<div class="sourceCode" id="cb1162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1162-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1162-1" aria-hidden="true" tabindex="-1"></a>dist_mat</span></code></pre></div>
<pre><code>##         human chimp gorilla
## human      NA    NA      NA
## chimp       2    NA      NA
## gorilla     6     4      NA</code></pre>
<div class="sourceCode" id="cb1164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1164-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1164-1" aria-hidden="true" tabindex="-1"></a>seq_dist0</span></code></pre></div>
<pre><code>##         human chimp
## chimp       2      
## gorilla     6     4</code></pre>
<p>They also each have a different class</p>
<div class="sourceCode" id="cb1166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1166-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1166-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(dist_mat)</span></code></pre></div>
<pre><code>## [1] &quot;matrix&quot; &quot;array&quot;</code></pre>
<div class="sourceCode" id="cb1168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1168-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1168-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(seq_dist0)</span></code></pre></div>
<pre><code>## [1] &quot;dist&quot;</code></pre>
</div>
<div id="distance-matrix-in-r-from-normal-matrix" class="section level2" number="56.7">
<h2><span class="header-section-number">56.7</span> Distance matrix in R from normal matrix</h2>
<p>Sometimes we already have a distance matrix computed that we want to enter directly into R. For example, perhaps it was reported in a paper and we want to explore the matrix on our own. In this case we can convert our normal matrix we made initially into a properly formatted distance matrix. This requires the base R <code>as.dist()</code> function. I’ll call this object “seq_dist_from_mat” to indicate that it came “from a matrix”. Note that I’m going to give it the full 3 x 3 matrix, <code>dist_mat</code>.</p>
<div class="sourceCode" id="cb1170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1170-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1170-1" aria-hidden="true" tabindex="-1"></a><span class="co"># input matrix</span></span>
<span id="cb1170-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1170-2" aria-hidden="true" tabindex="-1"></a>dist_mat</span></code></pre></div>
<pre><code>##         human chimp gorilla
## human      NA    NA      NA
## chimp       2    NA      NA
## gorilla     6     4      NA</code></pre>
<div class="sourceCode" id="cb1172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1172-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1172-1" aria-hidden="true" tabindex="-1"></a><span class="co"># covert to distance matrix</span></span>
<span id="cb1172-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1172-2" aria-hidden="true" tabindex="-1"></a>seq_dist_from_mat <span class="ot">&lt;-</span> <span class="fu">as.dist</span>(dist_mat)</span>
<span id="cb1172-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1172-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1172-4"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1172-4" aria-hidden="true" tabindex="-1"></a><span class="co"># new format</span></span>
<span id="cb1172-5"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1172-5" aria-hidden="true" tabindex="-1"></a>seq_dist_from_mat</span></code></pre></div>
<pre><code>##         human chimp
## chimp       2      
## gorilla     6     4</code></pre>
<p>I’ll now compare the two ways of making the matrix: <code>seq_dist0</code>, which was build from the original sequences, and <code>seq_dist_from_mat</code>, which was built from the hand-entered matrix.</p>
<p>First, let’s compare their structure</p>
<div class="sourceCode" id="cb1174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1174-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1174-1" aria-hidden="true" tabindex="-1"></a>seq_dist0</span></code></pre></div>
<pre><code>##         human chimp
## chimp       2      
## gorilla     6     4</code></pre>
<div class="sourceCode" id="cb1176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1176-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1176-1" aria-hidden="true" tabindex="-1"></a>seq_dist_from_mat</span></code></pre></div>
<pre><code>##         human chimp
## chimp       2      
## gorilla     6     4</code></pre>
<p>Then check them out with is() and class()</p>
<div class="sourceCode" id="cb1178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1178-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1178-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(seq_dist0) <span class="sc">==</span> <span class="fu">is</span>(seq_dist_from_mat)</span></code></pre></div>
<pre><code>## [1] TRUE TRUE</code></pre>
<div class="sourceCode" id="cb1180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1180-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1180-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(seq_dist0) <span class="sc">==</span> <span class="fu">class</span>(seq_dist_from_mat)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Now I’ll confirm that that have the same content (Note that distance matrices are not indexed internally by rows and columns, but by the order that the values appear. The following code therefore returns only 3 responses, not 4.)</p>
<div class="sourceCode" id="cb1182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1182-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1182-1" aria-hidden="true" tabindex="-1"></a>seq_dist0 <span class="sc">==</span> seq_dist_from_mat</span></code></pre></div>
<pre><code>## [1] TRUE TRUE TRUE</code></pre>
<p>Cool, so we can make a matrix either from the raw sequences or from a hand-entered matrix.</p>
</div>
<div id="upgma-in-r" class="section level2" number="56.8">
<h2><span class="header-section-number">56.8</span> UPGMA in R</h2>
<p>Once you have a distance matrix, UPGMA is easy in R with the <code>upgma()</code> function</p>
<div class="sourceCode" id="cb1184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1184-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1184-1" aria-hidden="true" tabindex="-1"></a>seq_upgma0 <span class="ot">&lt;-</span> <span class="fu">upgma</span>(seq_dist0)</span></code></pre></div>
<p>Now plot it with <code>plot()</code></p>
<div class="sourceCode" id="cb1185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1185-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1185-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(seq_upgma0, <span class="at">main =</span> <span class="st">&quot;Original data&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-834-1.png" width="672" /></p>
<p>This doesn’t look too different than the phylogeny produced from Newick. Let’s play with the distance matrix to see what happens.</p>
<p>First, make some more copies of the distance matrix</p>
<div class="sourceCode" id="cb1186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1186-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1186-1" aria-hidden="true" tabindex="-1"></a>seq.dist1 <span class="ot">&lt;-</span> seq_dist0</span>
<span id="cb1186-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1186-2" aria-hidden="true" tabindex="-1"></a>seq.dist2 <span class="ot">&lt;-</span> seq_dist0</span>
<span id="cb1186-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1186-3" aria-hidden="true" tabindex="-1"></a>seq.dist3 <span class="ot">&lt;-</span> seq_dist0</span></code></pre></div>
<p>The relationship between humans and chimps in in the upper left hand cell of the matrix. Internally, <em>R</em> is calling this the first cell of the matrix</p>
<div class="sourceCode" id="cb1187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1187-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1187-1" aria-hidden="true" tabindex="-1"></a>seq.dist1[<span class="dv">1</span>]</span></code></pre></div>
<pre><code>## [1] 2</code></pre>
<p>Let’s increase and decrease this value</p>
<div class="sourceCode" id="cb1189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1189-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1189-1" aria-hidden="true" tabindex="-1"></a>seq.dist1[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="dv">20</span>   <span class="co">#change from 2 to 20</span></span>
<span id="cb1189-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1189-2" aria-hidden="true" tabindex="-1"></a>seq.dist2[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="fl">0.5</span>  <span class="co">#change from 2 to 0.5</span></span>
<span id="cb1189-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1189-3" aria-hidden="true" tabindex="-1"></a>seq.dist3[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="fl">0.05</span> <span class="co">#change from 2 to 0.05</span></span></code></pre></div>
<p>Now we’ll build trees with our original matrix and the news ones to compare them.</p>
<p>Note: We’ll make a grid of plots using some fancy-looking code <code>par(mfrow = c(2,2), mar = c(2,2,2,2))</code>. Don’t worry about what exactly this means; it just sets up some graphical parameters to make things look nice. I’ve also added some code to make the plot of the original data look a little different</p>
<div class="sourceCode" id="cb1190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1190-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set up graphical parameters</span></span>
<span id="cb1190-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb1190-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1190-4"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-4" aria-hidden="true" tabindex="-1"></a><span class="co"># build 1st tree and plot</span></span>
<span id="cb1190-5"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-5" aria-hidden="true" tabindex="-1"></a><span class="do">## build tree</span></span>
<span id="cb1190-6"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-6" aria-hidden="true" tabindex="-1"></a>seq_upgma0 <span class="ot">&lt;-</span> <span class="fu">upgma</span>(seq_dist0)</span>
<span id="cb1190-7"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1190-8"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-8" aria-hidden="true" tabindex="-1"></a><span class="do">## plot, in upper left-hand corner</span></span>
<span id="cb1190-9"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span>(seq_upgma0, </span>
<span id="cb1190-10"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-10" aria-hidden="true" tabindex="-1"></a>           <span class="at">main =</span> <span class="st">&quot;Original data&quot;</span>,</span>
<span id="cb1190-11"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">edge.color =</span> <span class="st">&quot;darkgreen&quot;</span>,</span>
<span id="cb1190-12"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-12" aria-hidden="true" tabindex="-1"></a>           <span class="at">edge.width =</span> <span class="dv">4</span>)</span>
<span id="cb1190-13"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1190-14"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-14" aria-hidden="true" tabindex="-1"></a><span class="co"># build 2nd tree and plot</span></span>
<span id="cb1190-15"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-15" aria-hidden="true" tabindex="-1"></a>seq.upgma1 <span class="ot">&lt;-</span> <span class="fu">upgma</span>(seq.dist1)</span>
<span id="cb1190-16"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(seq.upgma1, <span class="at">main =</span> <span class="st">&quot;Very long distance humans vs. chimps!&quot;</span>)</span>
<span id="cb1190-17"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1190-18"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-18" aria-hidden="true" tabindex="-1"></a>seq.upgma2 <span class="ot">&lt;-</span> <span class="fu">upgma</span>(seq.dist2)</span>
<span id="cb1190-19"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(seq.upgma2, <span class="at">main =</span> <span class="st">&quot;Shorter distance humans vs. chimps&quot;</span>)</span>
<span id="cb1190-20"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1190-21"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1190-22"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-22" aria-hidden="true" tabindex="-1"></a>seq.upgma3 <span class="ot">&lt;-</span> <span class="fu">upgma</span>(seq.dist3)</span>
<span id="cb1190-23"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1190-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(seq.upgma3, <span class="at">main =</span> <span class="st">&quot;Much shorter distance humans vs. chimps!&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-838-1.png" width="672" /></p>
</div>
<div id="how-upgma-clustering-works" class="section level2" number="56.9">
<h2><span class="header-section-number">56.9</span> How UPGMA clustering works</h2>
<p>UPGMA and other <strong>clustering algorithms</strong> work by finding looking at a distance matrix and finding the two things that are closet together.</p>
<p>Human and chimps are closest</p>
<div class="sourceCode" id="cb1191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1191-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1191-1" aria-hidden="true" tabindex="-1"></a>seq_dist0</span></code></pre></div>
<pre><code>##         human chimp
## chimp       2      
## gorilla     6     4</code></pre>
<p>UPGAM therefore extracts the this minimum distance. We’ll extract it by hand using the <code>min()</code> function.</p>
<div class="sourceCode" id="cb1193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1193-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1193-1" aria-hidden="true" tabindex="-1"></a>d.min <span class="ot">&lt;-</span> <span class="fu">min</span>(seq_dist0)</span></code></pre></div>
<p>When two individual taxa are grouped into a clade the branch lengths are set at 1/2 the distance between them. So, the total distance is 2, so the branch lengths will be 1.</p>
<div class="sourceCode" id="cb1194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1194-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1194-1" aria-hidden="true" tabindex="-1"></a>d.min<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<p>We can confirm that this is correct by checking what R calculated via the <code>upgma()</code> function. We can access <em>R’s</em> calculations by looking at the object we saved the UPGMA output too</p>
<div class="sourceCode" id="cb1196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1196-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1196-1" aria-hidden="true" tabindex="-1"></a>seq_upgma0</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 3 tips and 2 internal nodes.
## 
## Tip labels:
##   human, chimp, gorilla
## 
## Rooted; includes branch lengths.</code></pre>
<p>On its own this isn’t helpful. But what we can do is use the <code>str()</code> command to look under the hood of the object</p>
<div class="sourceCode" id="cb1198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1198-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1198-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(seq_upgma0)</span></code></pre></div>
<pre><code>## List of 4
##  $ edge       : int [1:4, 1:2] 5 5 4 4 1 2 3 5
##  $ edge.length: num [1:4] 1 1 2.5 1.5
##  $ tip.label  : chr [1:3] &quot;human&quot; &quot;chimp&quot; &quot;gorilla&quot;
##  $ Nnode      : int 2
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;postorder&quot;</code></pre>
<p>The output tells us there is something called <code>edge.length</code> associated with the <code>seq_upgma0</code> object, and the smallest two values are 1. This looks promising, since <strong>edge</strong> is the math term for branches in a tree.</p>
<p>Next, UPGMA builds a new matrix that replace the Human and Chimp columns with a combined Humans-chimp column. The algorithm then re-calculates the distance from this human-chimp clade to the remaining gorilla group The distance from this human-chimp group to gorillas is the average distance from humans to gorillas and chimps to gorillas.</p>
<p>The original distance from humans to gorillas was 6, and the distance from chimps to gorillas was 4, as shown by the distance matrix</p>
<div class="sourceCode" id="cb1200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1200-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1200-1" aria-hidden="true" tabindex="-1"></a>seq_dist0</span></code></pre></div>
<pre><code>##         human chimp
## chimp       2      
## gorilla     6     4</code></pre>
<p>Let’s store these values in R objects.</p>
<div class="sourceCode" id="cb1202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1202-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1202-1" aria-hidden="true" tabindex="-1"></a><span class="co">#distance humans to gorillas</span></span>
<span id="cb1202-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1202-2" aria-hidden="true" tabindex="-1"></a>Dh_g <span class="ot">&lt;-</span> <span class="dv">6</span></span>
<span id="cb1202-3"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1202-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1202-4"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1202-4" aria-hidden="true" tabindex="-1"></a><span class="co">#distance chimps to gorillas</span></span>
<span id="cb1202-5"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1202-5" aria-hidden="true" tabindex="-1"></a>Dc_g <span class="ot">&lt;-</span> <span class="dv">4</span></span></code></pre></div>
<p>Now, we’ll take the average of these two distance. The math written out in explicit detail is:</p>
<div class="sourceCode" id="cb1203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1203-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1203-1" aria-hidden="true" tabindex="-1"></a>Dhc_g <span class="ot">&lt;-</span> (Dh_g<span class="sc">*</span><span class="dv">1</span> <span class="sc">+</span> Dc_g<span class="sc">*</span><span class="dv">1</span>)<span class="sc">/</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb1203-2"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1203-2" aria-hidden="true" tabindex="-1"></a>Dhc_g</span></code></pre></div>
<pre><code>## [1] 5</code></pre>
<p>The result is total distance between our new human-chimp clade and gorillas.</p>
<div id="challenge" class="section level3" number="56.9.1">
<h3><span class="header-section-number">56.9.1</span> Challenge</h3>
<p>This new distance allows us to calculate the branch lengths for the rest of the tree. The branch length from the human-chimp (h-c) clade to the gorillas (g) is 1/2 half the distance we just calculated:</p>
<p>Dhc_g/2 = 5/2 = 2.5</p>
<p>The remaining branch length is
2.5 - 1 = 1.5.</p>
<p>See if you can figure out why this is done by sketching out the tree.</p>
<p>These values match what is in the upgma object</p>
<div class="sourceCode" id="cb1205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1205-1"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#cb1205-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(seq_upgma0)</span></code></pre></div>
<pre><code>## List of 4
##  $ edge       : int [1:4, 1:2] 5 5 4 4 1 2 3 5
##  $ edge.length: num [1:4] 1 1 2.5 1.5
##  $ tip.label  : chr [1:3] &quot;human&quot; &quot;chimp&quot; &quot;gorilla&quot;
##  $ Nnode      : int 2
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;postorder&quot;</code></pre>
</div>
</div>
<div id="further-information" class="section level2" number="56.10">
<h2><span class="header-section-number">56.10</span> Further information</h2>
<div id="general-info" class="section level3" number="56.10.1">
<h3><span class="header-section-number">56.10.1</span> General info</h3>
<p>A good, no-math overview of the general ideas of UPGMA is at this <a href="https://medium.com/@sharma.ravit/upgma-method-designing-a-phylogenetic-tree-9a708de18419">blog post</a></p>
<p><a href="https://en.wikipedia.org/wiki/UPGMA">Wikipedia</a> provides a good overview with math that isn’t too daunting (though it could be improved). <a href="https://en.wikipedia.org/wiki/UPGMA" class="uri">https://en.wikipedia.org/wiki/UPGMA</a></p>
</div>
<div id="worked-examples" class="section level3" number="56.10.2">
<h3><span class="header-section-number">56.10.2</span> Worked examples:</h3>
<p>An excellent worked example is at <a href="http://www.slimsuite.unsw.edu.au/teaching/upgma/">this site</a>. <a href="http://www.slimsuite.unsw.edu.au/teaching/upgma/" class="uri">http://www.slimsuite.unsw.edu.au/teaching/upgma/</a>
This example breaks down the math into basic steps, but there’s a lot of taxa so its still complicated.</p>
<p>A worked example showing the limits of UPGMA is <a href="https://www.icp.ucl.ac.be/~opperd/private/upgma.html">here</a>:
<a href="https://www.icp.ucl.ac.be/~opperd/private/upgma.html" class="uri">https://www.icp.ucl.ac.be/~opperd/private/upgma.html</a></p>
<p>Another, dense-looking example is <a href="http://www.nmsr.org/upgma.htm">here</a>:
<a href="http://www.nmsr.org/upgma.htm" class="uri">http://www.nmsr.org/upgma.htm</a></p>
<p>Another worked example is
<a href="https://www.mun.ca/biology/scarr/2900_UPGMA.htm" class="uri">https://www.mun.ca/biology/scarr/2900_UPGMA.htm</a></p>
</div>
<div id="additional-notes" class="section level3" number="56.10.3">
<h3><span class="header-section-number">56.10.3</span> Additional notes:</h3>
<p>Some useful but terse notes are <a href="https://www.sequentix.de/gelquest/help/upgma_method.htm">here</a>:
<a href="https://www.sequentix.de/gelquest/help/upgma_method.htm" class="uri">https://www.sequentix.de/gelquest/help/upgma_method.htm</a></p>
<p>If you want to understand what the <strong>ultrametricity</strong> assumption is, check out <a href="https://www.sequentix.de/gelquest/help/explanation_of_the_term_ultrametric.htm">this site</a> (however, I don’t get hung up on this):
<a href="https://www.sequentix.de/gelquest/help/explanation_of_the_term_ultrametric.htm" class="uri">https://www.sequentix.de/gelquest/help/explanation_of_the_term_ultrametric.htm</a></p>
<p>Explanation of the convoluted meaning of “Un-weighted” is <a href="https://www.mun.ca/biology/scarr/UPGMA_vs_WPGMA.html">here</a>
<a href="https://www.mun.ca/biology/scarr/UPGMA_vs_WPGMA.html" class="uri">https://www.mun.ca/biology/scarr/UPGMA_vs_WPGMA.html</a></p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="plotting-trees-from-newick-notation-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/666-treethinking/tree-think-05-UPGMA_in_R.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/666-treethinking/tree-think-05-UPGMA_in_R.Rmd",
"text": null
},
"download": ["lbrb.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
