<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Proteins | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Proteins | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Proteins | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-08-10" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="nucleic-acids-review.html"/>
<link rel="next" href="downloadR.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html"><i class="fa fa-check"></i><b>1</b> Nucleic Acids</a>
<ul>
<li class="chapter" data-level="1.1" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#dna-and-rna"><i class="fa fa-check"></i><b>1.1</b> DNA and RNA</a></li>
<li class="chapter" data-level="1.2" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#dna-double-helix-structure"><i class="fa fa-check"></i><b>1.2</b> DNA Double-Helix Structure</a></li>
<li class="chapter" data-level="1.3" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#rna"><i class="fa fa-check"></i><b>1.3</b> RNA</a></li>
<li class="chapter" data-level="1.4" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#summary"><i class="fa fa-check"></i><b>1.4</b> Summary</a></li>
<li class="chapter" data-level="1.5" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#analysis-questions"><i class="fa fa-check"></i><b>1.5</b> Analysis Questions</a></li>
<li class="chapter" data-level="1.6" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#glossary"><i class="fa fa-check"></i><b>1.6</b> Glossary</a></li>
<li class="chapter" data-level="1.7" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#contributors-and-attributions"><i class="fa fa-check"></i><b>1.7</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="proteins-review.html"><a href="proteins-review.html"><i class="fa fa-check"></i><b>2</b> Proteins</a>
<ul>
<li class="chapter" data-level="2.0.1" data-path="proteins-review.html"><a href="proteins-review.html#skills-to-develop"><i class="fa fa-check"></i><b>2.0.1</b> Skills to develop:</a></li>
<li class="chapter" data-level="2.1" data-path="proteins-review.html"><a href="proteins-review.html#types-and-functions-of-proteins"><i class="fa fa-check"></i><b>2.1</b> Types and Functions of Proteins</a></li>
<li class="chapter" data-level="2.2" data-path="proteins-review.html"><a href="proteins-review.html#amino-acids"><i class="fa fa-check"></i><b>2.2</b> Amino Acids</a></li>
<li class="chapter" data-level="2.3" data-path="proteins-review.html"><a href="proteins-review.html#evolution-connection"><i class="fa fa-check"></i><b>2.3</b> Evolution Connection:</a></li>
<li class="chapter" data-level="2.4" data-path="proteins-review.html"><a href="proteins-review.html#protein-structure"><i class="fa fa-check"></i><b>2.4</b> Protein Structure</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="proteins-review.html"><a href="proteins-review.html#primary-structure"><i class="fa fa-check"></i><b>2.4.1</b> Primary Structure</a></li>
<li class="chapter" data-level="2.4.2" data-path="proteins-review.html"><a href="proteins-review.html#secondary-structure"><i class="fa fa-check"></i><b>2.4.2</b> Secondary Structure</a></li>
<li class="chapter" data-level="2.4.3" data-path="proteins-review.html"><a href="proteins-review.html#tertiary-structure"><i class="fa fa-check"></i><b>2.4.3</b> Tertiary Structure</a></li>
<li class="chapter" data-level="2.4.4" data-path="proteins-review.html"><a href="proteins-review.html#quaternary-structure"><i class="fa fa-check"></i><b>2.4.4</b> Quaternary Structure</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="proteins-review.html"><a href="proteins-review.html#denaturation-and-protein-folding"><i class="fa fa-check"></i><b>2.5</b> Denaturation and Protein Folding</a></li>
<li class="chapter" data-level="2.6" data-path="proteins-review.html"><a href="proteins-review.html#summary-1"><i class="fa fa-check"></i><b>2.6</b> Summary</a></li>
<li class="chapter" data-level="2.7" data-path="proteins-review.html"><a href="proteins-review.html#art-connections"><i class="fa fa-check"></i><b>2.7</b> Art Connections</a></li>
<li class="chapter" data-level="2.8" data-path="proteins-review.html"><a href="proteins-review.html#review-questions"><i class="fa fa-check"></i><b>2.8</b> Review Questions</a></li>
<li class="chapter" data-level="2.9" data-path="proteins-review.html"><a href="proteins-review.html#free-response"><i class="fa fa-check"></i><b>2.9</b> Free Response</a></li>
<li class="chapter" data-level="2.10" data-path="proteins-review.html"><a href="proteins-review.html#glossary-1"><i class="fa fa-check"></i><b>2.10</b> Glossary</a></li>
<li class="chapter" data-level="2.11" data-path="proteins-review.html"><a href="proteins-review.html#contributors-and-attributions-1"><i class="fa fa-check"></i><b>2.11</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="downloadR.html"><a href="downloadR.html"><i class="fa fa-check"></i><b>3</b> Downloading R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="downloadR.html"><a href="downloadR.html#preface"><i class="fa fa-check"></i><b>3.1</b> Preface</a></li>
<li class="chapter" data-level="3.2" data-path="downloadR.html"><a href="downloadR.html#introduction-to-r"><i class="fa fa-check"></i><b>3.2</b> Introduction to R</a></li>
<li class="chapter" data-level="3.3" data-path="downloadR.html"><a href="downloadR.html#installing-r"><i class="fa fa-check"></i><b>3.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="downloadR.html"><a href="downloadR.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>3.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="3.3.2" data-path="downloadR.html"><a href="downloadR.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>3.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="downloadR.html"><a href="downloadR.html#starting-r"><i class="fa fa-check"></i><b>3.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html"><i class="fa fa-check"></i><b>4</b> Installing the RStudio IDE {$installRStudio}</a>
<ul>
<li class="chapter" data-level="4.1" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>4.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="4.2" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>4.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="installing-packages.html"><a href="installing-packages.html"><i class="fa fa-check"></i><b>5</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="5.1" data-path="installing-packages.html"><a href="installing-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>5.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="5.2" data-path="installing-packages.html"><a href="installing-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>5.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="5.3" data-path="installing-packages.html"><a href="installing-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>5.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>6</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="6.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>6.1</b> Bioconductor</a></li>
<li class="chapter" data-level="6.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>6.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="6.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>6.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>6.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="6.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>6.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="6.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>6.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>6.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="basic-R.html"><a href="basic-R.html"><i class="fa fa-check"></i><b>7</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="7.1" data-path="basic-R.html"><a href="basic-R.html#vocabulary"><i class="fa fa-check"></i><b>7.1</b> Vocabulary</a></li>
<li class="chapter" data-level="7.2" data-path="basic-R.html"><a href="basic-R.html#r-functions"><i class="fa fa-check"></i><b>7.2</b> R functions</a></li>
<li class="chapter" data-level="7.3" data-path="basic-R.html"><a href="basic-R.html#interacting-with-r"><i class="fa fa-check"></i><b>7.3</b> Interacting with R</a></li>
<li class="chapter" data-level="7.4" data-path="basic-R.html"><a href="basic-R.html#variables-in-r"><i class="fa fa-check"></i><b>7.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="basic-R.html"><a href="basic-R.html#vectors"><i class="fa fa-check"></i><b>7.4.1</b> Vectors</a></li>
<li class="chapter" data-level="7.4.2" data-path="basic-R.html"><a href="basic-R.html#vector-indexing"><i class="fa fa-check"></i><b>7.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="7.4.3" data-path="basic-R.html"><a href="basic-R.html#character-vectors"><i class="fa fa-check"></i><b>7.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="7.4.4" data-path="basic-R.html"><a href="basic-R.html#lists"><i class="fa fa-check"></i><b>7.4.4</b> Lists</a></li>
<li class="chapter" data-level="7.4.5" data-path="basic-R.html"><a href="basic-R.html#tables"><i class="fa fa-check"></i><b>7.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="basic-R.html"><a href="basic-R.html#arguments"><i class="fa fa-check"></i><b>7.5</b> Arguments</a></li>
<li class="chapter" data-level="7.6" data-path="basic-R.html"><a href="basic-R.html#help-files-with-help-and"><i class="fa fa-check"></i><b>7.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="7.7" data-path="basic-R.html"><a href="basic-R.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>7.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="7.8" data-path="basic-R.html"><a href="basic-R.html#more-on-functions"><i class="fa fa-check"></i><b>7.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="7.8.1" data-path="basic-R.html"><a href="basic-R.html#writing-your-own-functions"><i class="fa fa-check"></i><b>7.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="basic-R.html"><a href="basic-R.html#quiting-r"><i class="fa fa-check"></i><b>7.9</b> Quiting R</a></li>
<li class="chapter" data-level="7.10" data-path="basic-R.html"><a href="basic-R.html#links-and-further-reading"><i class="fa fa-check"></i><b>7.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="vectors-in-R.html"><a href="vectors-in-R.html"><i class="fa fa-check"></i><b>8</b> A primer for working with vectors</a>
<ul>
<li class="chapter" data-level="8.1" data-path="vectors-in-R.html"><a href="vectors-in-R.html#preface-1"><i class="fa fa-check"></i><b>8.1</b> Preface</a></li>
<li class="chapter" data-level="8.2" data-path="vectors-in-R.html"><a href="vectors-in-R.html#vocab"><i class="fa fa-check"></i><b>8.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>9</b> Functions</a>
<ul>
<li class="chapter" data-level="9.1" data-path="functions.html"><a href="functions.html#vectors-in-r"><i class="fa fa-check"></i><b>9.1</b> Vectors in R</a></li>
<li class="chapter" data-level="9.2" data-path="functions.html"><a href="functions.html#math-on-vectors"><i class="fa fa-check"></i><b>9.2</b> Math on vectors</a></li>
<li class="chapter" data-level="9.3" data-path="functions.html"><a href="functions.html#functions-on-vectors"><i class="fa fa-check"></i><b>9.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="9.4" data-path="functions.html"><a href="functions.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>9.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="9.5" data-path="functions.html"><a href="functions.html#subsetting-vectors"><i class="fa fa-check"></i><b>9.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="9.6" data-path="functions.html"><a href="functions.html#sequences-of-numbers"><i class="fa fa-check"></i><b>9.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="9.7" data-path="functions.html"><a href="functions.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>9.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="9.8" data-path="functions.html"><a href="functions.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>9.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="plotting-vectors.html"><a href="plotting-vectors.html"><i class="fa fa-check"></i><b>10</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="10.1" data-path="plotting-vectors.html"><a href="plotting-vectors.html#preface-2"><i class="fa fa-check"></i><b>10.1</b> Preface</a></li>
<li class="chapter" data-level="10.2" data-path="plotting-vectors.html"><a href="plotting-vectors.html#plotting-numeric-data"><i class="fa fa-check"></i><b>10.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="10.3" data-path="plotting-vectors.html"><a href="plotting-vectors.html#other-plotting-packages"><i class="fa fa-check"></i><b>10.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="R-objects.html"><a href="R-objects.html"><i class="fa fa-check"></i><b>11</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="11.1" data-path="R-objects.html"><a href="R-objects.html#commands-used"><i class="fa fa-check"></i><b>11.1</b> Commands used</a></li>
<li class="chapter" data-level="11.2" data-path="R-objects.html"><a href="R-objects.html#r-objects"><i class="fa fa-check"></i><b>11.2</b> R Objects</a></li>
<li class="chapter" data-level="11.3" data-path="R-objects.html"><a href="R-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>11.3</b> Differences between objects</a></li>
<li class="chapter" data-level="11.4" data-path="R-objects.html"><a href="R-objects.html#the-data"><i class="fa fa-check"></i><b>11.4</b> The Data</a></li>
<li class="chapter" data-level="11.5" data-path="R-objects.html"><a href="R-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>11.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="R-objects.html"><a href="R-objects.html#is"><i class="fa fa-check"></i><b>11.5.1</b> is()</a></li>
<li class="chapter" data-level="11.5.2" data-path="R-objects.html"><a href="R-objects.html#length"><i class="fa fa-check"></i><b>11.5.2</b> length()</a></li>
<li class="chapter" data-level="11.5.3" data-path="R-objects.html"><a href="R-objects.html#str"><i class="fa fa-check"></i><b>11.5.3</b> str()</a></li>
<li class="chapter" data-level="11.5.4" data-path="R-objects.html"><a href="R-objects.html#c"><i class="fa fa-check"></i><b>11.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="R-objects.html"><a href="R-objects.html#debrief"><i class="fa fa-check"></i><b>11.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html"><i class="fa fa-check"></i><b>12</b> Introducing FASTA Files</a>
<ul>
<li class="chapter" data-level="12.1" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html#example-fasta-file"><i class="fa fa-check"></i><b>12.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="12.2" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>12.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="12.3" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>12.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="12.4" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html#fastq-format"><i class="fa fa-check"></i><b>12.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html"><i class="fa fa-check"></i><b>13</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="13.0.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#functions-1"><i class="fa fa-check"></i><b>13.0.1</b> Functions</a></li>
<li class="chapter" data-level="13.0.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#softwarewebsites"><i class="fa fa-check"></i><b>13.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="13.0.3" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#r-vocabulary"><i class="fa fa-check"></i><b>13.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="13.0.4" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#file-types"><i class="fa fa-check"></i><b>13.0.4</b> File types</a></li>
<li class="chapter" data-level="13.0.5" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>13.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="13.0.6" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>13.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="13.0.7" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#preliminaries"><i class="fa fa-check"></i><b>13.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="13.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>13.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="13.1.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>13.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="13.1.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>13.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="13.1.3" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#fasta-file-format"><i class="fa fa-check"></i><b>13.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="13.1.4" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>13.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="13.1.5" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>13.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>13.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="13.3" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#next-steps"><i class="fa fa-check"></i><b>13.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>14</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="14.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>14.1</b> Preliminaries</a></li>
<li class="chapter" data-level="14.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>14.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>14.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="14.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>14.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>15</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="15.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-3"><i class="fa fa-check"></i><b>15.1</b> Preface</a></li>
<li class="chapter" data-level="15.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>15.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="15.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>15.3</b> Introduction</a></li>
<li class="chapter" data-level="15.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>15.4</b> Vocabulary</a></li>
<li class="chapter" data-level="15.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions-2"><i class="fa fa-check"></i><b>15.5</b> Functions</a></li>
<li class="chapter" data-level="15.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries-2"><i class="fa fa-check"></i><b>15.6</b> Preliminaries</a></li>
<li class="chapter" data-level="15.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>15.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="15.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>15.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="15.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>15.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="15.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>15.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="15.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>15.8.3</b> DNA words</a></li>
<li class="chapter" data-level="15.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary-2"><i class="fa fa-check"></i><b>15.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="15.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>15.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="15.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>15.9.1</b> License</a></li>
<li class="chapter" data-level="15.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>15.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="download-protein-seqs-R.html"><a href="download-protein-seqs-R.html"><i class="fa fa-check"></i><b>16</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="16.1" data-path="download-protein-seqs-R.html"><a href="download-protein-seqs-R.html#preliminaries-3"><i class="fa fa-check"></i><b>16.1</b> Preliminaries</a></li>
<li class="chapter" data-level="16.2" data-path="download-protein-seqs-R.html"><a href="download-protein-seqs-R.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>16.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="changes-to-database-entries.html"><a href="changes-to-database-entries.html"><i class="fa fa-check"></i><b>17</b> Changes to database entries</a></li>
<li class="chapter" data-level="18" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>18</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="18.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>18.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>18.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>18.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="18.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>18.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="18.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>18.4</b> Inversions</a></li>
<li class="chapter" data-level="18.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>18.5</b> Translocations</a></li>
<li class="chapter" data-level="18.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>18.6</b> Random sequence</a></li>
<li class="chapter" data-level="18.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>18.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>19</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="19.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>19.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="19.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>19.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="19.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>19.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="19.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>19.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>19.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="19.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>19.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>19.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="19.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>19.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>20</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="20.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>20.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="20.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>20.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="20.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>20.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html"><i class="fa fa-check"></i><b>21</b> Alignment by eye in R</a>
<ul>
<li class="chapter" data-level="21.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>21.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="21.1.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#packages"><i class="fa fa-check"></i><b>21.1.1</b> Packages</a></li>
<li class="chapter" data-level="21.1.2" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#data"><i class="fa fa-check"></i><b>21.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="22" data-path="alignment-in-r.html"><a href="alignment-in-r.html"><i class="fa fa-check"></i><b>22</b> Alignment in R</a>
<ul>
<li class="chapter" data-level="22.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#preliminaries-8"><i class="fa fa-check"></i><b>22.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="22.1.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#packages-1"><i class="fa fa-check"></i><b>22.1.1</b> Packages</a></li>
<li class="chapter" data-level="22.1.2" data-path="alignment-in-r.html"><a href="alignment-in-r.html#data-1"><i class="fa fa-check"></i><b>22.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="working-with-vectors-and-matrices.html"><a href="working-with-vectors-and-matrices.html"><i class="fa fa-check"></i><b>23</b> Working with vectors and matrices</a></li>
<li class="chapter" data-level="24" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html"><i class="fa fa-check"></i><b>24</b> Matrix vocabulary</a>
<ul>
<li class="chapter" data-level="24.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#introduction-1"><i class="fa fa-check"></i><b>24.1</b> Introduction</a></li>
<li class="chapter" data-level="24.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#preliminaries-9"><i class="fa fa-check"></i><b>24.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="24.2.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#packages-2"><i class="fa fa-check"></i><b>24.2.1</b> Packages</a></li>
<li class="chapter" data-level="24.2.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#vocab-1"><i class="fa fa-check"></i><b>24.2.2</b> Vocab</a></li>
<li class="chapter" data-level="24.2.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#functions-used"><i class="fa fa-check"></i><b>24.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="24.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#load-packages"><i class="fa fa-check"></i><b>24.3</b> Load packages</a></li>
<li class="chapter" data-level="24.4" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#matrix-structure"><i class="fa fa-check"></i><b>24.4</b> Matrix structure</a></li>
<li class="chapter" data-level="24.5" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#diagonal-of-a-matrix"><i class="fa fa-check"></i><b>24.5</b> Diagonal of a matrix</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html"><i class="fa fa-check"></i><b>25</b> Testing the significance of an alignment</a>
<ul>
<li class="chapter" data-level="25.1" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#calculating-the-statistical-significance-of-a-pairwise-global-alignment"><i class="fa fa-check"></i><b>25.1</b> Calculating the statistical significance of a pairwise global alignment</a></li>
<li class="chapter" data-level="25.2" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#summary-3"><i class="fa fa-check"></i><b>25.2</b> Summary</a></li>
<li class="chapter" data-level="25.3" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#links-and-further-reading-1"><i class="fa fa-check"></i><b>25.3</b> Links and Further Reading</a></li>
<li class="chapter" data-level="25.4" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#exercises-1"><i class="fa fa-check"></i><b>25.4</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>26</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="26.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>26.1</b> Prelminaries</a></li>
<li class="chapter" data-level="26.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>26.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="26.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>26.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html"><i class="fa fa-check"></i><b>27</b> Manipulating matrices and vectors: worked example</a>
<ul>
<li class="chapter" data-level="27.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#introduction-2"><i class="fa fa-check"></i><b>27.1</b> Introduction</a></li>
<li class="chapter" data-level="27.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#preliminaries-10"><i class="fa fa-check"></i><b>27.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="27.2.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-3"><i class="fa fa-check"></i><b>27.2.1</b> Packages</a></li>
<li class="chapter" data-level="27.2.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#vocab-2"><i class="fa fa-check"></i><b>27.2.2</b> Vocab</a></li>
<li class="chapter" data-level="27.2.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#functions-used-1"><i class="fa fa-check"></i><b>27.2.3</b> Functions used</a></li>
<li class="chapter" data-level="27.2.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-4"><i class="fa fa-check"></i><b>27.2.4</b> Packages</a></li>
<li class="chapter" data-level="27.2.5" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#data-2"><i class="fa fa-check"></i><b>27.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="27.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#matrix-elements-can-be-accessed-with-square-bracket-notation"><i class="fa fa-check"></i><b>27.3</b> Matrix elements can be accessed with square bracket notation</a></li>
<li class="chapter" data-level="27.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#r-object-names"><i class="fa fa-check"></i><b>27.4</b> R object names</a>
<ul>
<li class="chapter" data-level="27.4.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-vectors"><i class="fa fa-check"></i><b>27.4.1</b> Accessing items in names vectors</a></li>
<li class="chapter" data-level="27.4.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-matrices"><i class="fa fa-check"></i><b>27.4.2</b> Accessing items in names matrices</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="28" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>28</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="28.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-11"><i class="fa fa-check"></i><b>28.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="28.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages-5"><i class="fa fa-check"></i><b>28.1.1</b> Packages</a></li>
<li class="chapter" data-level="28.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-3"><i class="fa fa-check"></i><b>28.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="28.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>28.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="28.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>28.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="28.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>28.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="28.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>28.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="28.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>28.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="28.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>28.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="28.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>28.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="28.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>28.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html"><i class="fa fa-check"></i><b>29</b> The BLOSUM scoring matrix in R</a>
<ul>
<li class="chapter" data-level="29.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#introduction-3"><i class="fa fa-check"></i><b>29.1</b> Introduction</a></li>
<li class="chapter" data-level="29.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#preliminaries-12"><i class="fa fa-check"></i><b>29.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="29.2.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#packages-6"><i class="fa fa-check"></i><b>29.2.1</b> Packages</a></li>
<li class="chapter" data-level="29.2.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#vocab-3"><i class="fa fa-check"></i><b>29.2.2</b> Vocab</a></li>
<li class="chapter" data-level="29.2.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#functions-used-2"><i class="fa fa-check"></i><b>29.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="29.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#load-packages-1"><i class="fa fa-check"></i><b>29.3</b> Load packages</a></li>
<li class="chapter" data-level="29.4" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#the-blosum-matrix"><i class="fa fa-check"></i><b>29.4</b> The BLOSUM matrix</a></li>
<li class="chapter" data-level="29.5" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#visualizing-the-blosum-matrix"><i class="fa fa-check"></i><b>29.5</b> Visualizing the BLOSUM matrix</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>30</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="31" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>31</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="31.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-4"><i class="fa fa-check"></i><b>31.1</b> Introduction</a></li>
<li class="chapter" data-level="31.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>31.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="31.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-4"><i class="fa fa-check"></i><b>31.2.1</b> Vocab</a></li>
<li class="chapter" data-level="31.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>31.2.2</b> R functions</a></li>
<li class="chapter" data-level="31.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>31.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="31.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>31.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="31.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>31.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="31.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>31.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="31.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>31.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="31.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>31.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="31.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>31.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="31.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>31.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="31.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>31.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="31.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>31.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="31.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>31.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="31.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>31.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="31.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>31.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="32" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>32</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="32.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-13"><i class="fa fa-check"></i><b>32.1</b> Preliminaries</a></li>
<li class="chapter" data-level="32.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-5"><i class="fa fa-check"></i><b>32.2</b> Introduction</a></li>
<li class="chapter" data-level="32.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>32.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="32.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>32.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="33" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>33</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="33.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-14"><i class="fa fa-check"></i><b>33.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="33.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>33.1.1</b> Key functions</a></li>
<li class="chapter" data-level="33.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>33.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="33.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>33.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="33.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>33.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="33.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>33.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="33.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>33.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="33.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>33.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="34" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html"><i class="fa fa-check"></i><b>34</b> Local variation in GC content</a>
<ul>
<li class="chapter" data-level="34.0.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#note-on-the-biology-in-this-section"><i class="fa fa-check"></i><b>34.0.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="34.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#vocabulary-2"><i class="fa fa-check"></i><b>34.1</b> Vocabulary</a></li>
<li class="chapter" data-level="34.2" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#reading-sequence-data-with-rentrezentrez_fetch"><i class="fa fa-check"></i><b>34.2</b> Reading sequence data with rentrez::entrez_fetch</a></li>
<li class="chapter" data-level="34.3" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#local-variation-in-gc-content-1"><i class="fa fa-check"></i><b>34.3</b> Local variation in GC content</a></li>
<li class="chapter" data-level="34.4" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-analysis-of-gc-content"><i class="fa fa-check"></i><b>34.4</b> A sliding window analysis of GC content</a></li>
<li class="chapter" data-level="34.5" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-plot-of-gc-content"><i class="fa fa-check"></i><b>34.5</b> A sliding window plot of GC content</a>
<ul>
<li class="chapter" data-level="34.5.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#over-represented-and-under-represented-dna-words"><i class="fa fa-check"></i><b>34.5.1</b> Over-represented and under-represented DNA words</a></li>
</ul></li>
<li class="chapter" data-level="34.6" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#summary-4"><i class="fa fa-check"></i><b>34.6</b> Summary</a></li>
<li class="chapter" data-level="34.7" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#further-reading"><i class="fa fa-check"></i><b>34.7</b> Further Reading</a></li>
<li class="chapter" data-level="34.8" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#acknowledgements-1"><i class="fa fa-check"></i><b>34.8</b> Acknowledgements</a></li>
<li class="chapter" data-level="34.9" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#exercises-2"><i class="fa fa-check"></i><b>34.9</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html"><i class="fa fa-check"></i><b>35</b> Computational gene finding</a>
<ul>
<li class="chapter" data-level="35.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#preliminaries-15"><i class="fa fa-check"></i><b>35.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="35.1.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#packages-7"><i class="fa fa-check"></i><b>35.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="35.2" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#the-genetic-code"><i class="fa fa-check"></i><b>35.2</b> The genetic code</a></li>
<li class="chapter" data-level="35.3" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-start-and-stop-codons-in-a-dna-sequence"><i class="fa fa-check"></i><b>35.3</b> Finding start and stop codons in a DNA sequence</a></li>
<li class="chapter" data-level="35.4" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#reading-frames"><i class="fa fa-check"></i><b>35.4</b> Reading frames</a></li>
<li class="chapter" data-level="35.5" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-forward-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>35.5</b> Finding open reading frames on the forward strand of a DNA sequence</a></li>
<li class="chapter" data-level="35.6" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#predicting-the-protein-sequence-for-an-orf"><i class="fa fa-check"></i><b>35.6</b> Predicting the protein sequence for an ORF</a></li>
<li class="chapter" data-level="35.7" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-reverse-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>35.7</b> Finding open reading frames on the reverse strand of a DNA sequence</a></li>
<li class="chapter" data-level="35.8" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#lengths-of-open-reading-frames"><i class="fa fa-check"></i><b>35.8</b> Lengths of open reading frames</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html"><i class="fa fa-check"></i><b>36</b> Significance of ORFs</a>
<ul>
<li class="chapter" data-level="36.1" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#preliminaries-16"><i class="fa fa-check"></i><b>36.1</b> Preliminaries</a></li>
<li class="chapter" data-level="36.2" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#identifying-significant-open-reading-frames"><i class="fa fa-check"></i><b>36.2</b> Identifying significant open reading frames</a></li>
<li class="chapter" data-level="36.3" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#summary-5"><i class="fa fa-check"></i><b>36.3</b> Summary</a></li>
<li class="chapter" data-level="36.4" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#links-and-further-reading-2"><i class="fa fa-check"></i><b>36.4</b> Links and Further Reading</a></li>
<li class="chapter" data-level="36.5" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#acknowledgements-2"><i class="fa fa-check"></i><b>36.5</b> Acknowledgements</a></li>
<li class="chapter" data-level="36.6" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#exercises-3"><i class="fa fa-check"></i><b>36.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="comparative-genomics.html"><a href="comparative-genomics.html"><i class="fa fa-check"></i><b>37</b> Comparative Genomics</a>
<ul>
<li class="chapter" data-level="37.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#preliminaries-17"><i class="fa fa-check"></i><b>37.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="37.1.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#packages-8"><i class="fa fa-check"></i><b>37.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="37.2" data-path="comparative-genomics.html"><a href="comparative-genomics.html#introduction-6"><i class="fa fa-check"></i><b>37.2</b> Introduction</a></li>
<li class="chapter" data-level="37.3" data-path="comparative-genomics.html"><a href="comparative-genomics.html#using-the-biomart-r-library-to-query-the-ensembl-database"><i class="fa fa-check"></i><b>37.3</b> Using the biomaRt R Library to Query the Ensembl Database</a></li>
<li class="chapter" data-level="37.4" data-path="comparative-genomics.html"><a href="comparative-genomics.html#comparing-the-number-of-genes-in-two-species"><i class="fa fa-check"></i><b>37.4</b> Comparing the number of genes in two species</a></li>
<li class="chapter" data-level="37.5" data-path="comparative-genomics.html"><a href="comparative-genomics.html#identifying-homologous-genes-between-two-species"><i class="fa fa-check"></i><b>37.5</b> Identifying homologous genes between two species</a></li>
</ul></li>
<li class="chapter" data-level="38" data-path="summary-6.html"><a href="summary-6.html"><i class="fa fa-check"></i><b>38</b> Summary</a></li>
<li class="chapter" data-level="39" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html"><i class="fa fa-check"></i><b>39</b> #Links and Further Reading</a>
<ul>
<li class="chapter" data-level="39.1" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#acknowledgements-3"><i class="fa fa-check"></i><b>39.1</b> Acknowledgements</a></li>
<li class="chapter" data-level="39.2" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#exercises-4"><i class="fa fa-check"></i><b>39.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="40" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html"><i class="fa fa-check"></i><b>40</b> Hidden Markov Models</a>
<ul>
<li class="chapter" data-level="40.1" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-multinomial-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>40.1</b> A multinomial model of DNA sequence evolution</a></li>
<li class="chapter" data-level="40.2" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-multinomial-model"><i class="fa fa-check"></i><b>40.2</b> Generating a DNA sequence using a multinomial model</a></li>
<li class="chapter" data-level="40.3" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>40.3</b> A Markov model of DNA sequence evolution</a></li>
<li class="chapter" data-level="40.4" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-for-a-markov-model"><i class="fa fa-check"></i><b>40.4</b> The transition matrix for a Markov model</a></li>
<li class="chapter" data-level="40.5" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-markov-model"><i class="fa fa-check"></i><b>40.5</b> Generating a DNA sequence using a Markov model</a></li>
<li class="chapter" data-level="40.6" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>40.6</b> A Hidden Markov Model of DNA sequence evolution</a></li>
<li class="chapter" data-level="40.7" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-and-emission-matrix-for-a-hmm"><i class="fa fa-check"></i><b>40.7</b> The transition matrix and emission matrix for a HMM</a></li>
<li class="chapter" data-level="40.8" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-hmm"><i class="fa fa-check"></i><b>40.8</b> Generating a DNA sequence using a HMM</a></li>
<li class="chapter" data-level="40.9" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#inferring-the-states-of-a-hmm-that-generated-a-dna-sequence"><i class="fa fa-check"></i><b>40.9</b> Inferring the states of a HMM that generated a DNA sequence</a></li>
<li class="chapter" data-level="40.10" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-protein-sequence-evolution"><i class="fa fa-check"></i><b>40.10</b> A Hidden Markov Model of protein sequence evolution</a></li>
<li class="chapter" data-level="40.11" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#summary-7"><i class="fa fa-check"></i><b>40.11</b> Summary</a></li>
<li class="chapter" data-level="40.12" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#links-and-further-reading-4"><i class="fa fa-check"></i><b>40.12</b> Links and Further Reading</a></li>
<li class="chapter" data-level="40.13" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#exercises-5"><i class="fa fa-check"></i><b>40.13</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="41" data-path="blast-summary.html"><a href="blast-summary.html"><i class="fa fa-check"></i><b>41</b> BLAST Summary</a>
<ul>
<li class="chapter" data-level="41.1" data-path="blast-summary.html"><a href="blast-summary.html#blast-overview"><i class="fa fa-check"></i><b>41.1</b> BLAST Overview</a>
<ul>
<li class="chapter" data-level="41.1.1" data-path="blast-summary.html"><a href="blast-summary.html#why-do-a-blast-search"><i class="fa fa-check"></i><b>41.1.1</b> Why do a BLAST search?</a></li>
<li class="chapter" data-level="41.1.2" data-path="blast-summary.html"><a href="blast-summary.html#what-does-blast-let-you-do"><i class="fa fa-check"></i><b>41.1.2</b> What does BLAST let you do?</a></li>
</ul></li>
<li class="chapter" data-level="41.2" data-path="blast-summary.html"><a href="blast-summary.html#blast-recipe"><i class="fa fa-check"></i><b>41.2</b> BLAST recipe</a></li>
<li class="chapter" data-level="41.3" data-path="blast-summary.html"><a href="blast-summary.html#modules-in-the-blast-workflow-and-algorithm"><i class="fa fa-check"></i><b>41.3</b> Modules in the BLAST Workflow and Algorithm</a>
<ul>
<li class="chapter" data-level="41.3.1" data-path="blast-summary.html"><a href="blast-summary.html#module-1-setup-query-sequence-protein-blast-oriented"><i class="fa fa-check"></i><b>41.3.1</b> Module 1: Setup query sequence (Protein BLAST oriented)</a></li>
<li class="chapter" data-level="41.3.2" data-path="blast-summary.html"><a href="blast-summary.html#module-2-scanning-database-for-high-scoring-words"><i class="fa fa-check"></i><b>41.3.2</b> Module 2: Scanning database for high-scoring words</a></li>
<li class="chapter" data-level="41.3.3" data-path="blast-summary.html"><a href="blast-summary.html#module-3-traceback"><i class="fa fa-check"></i><b>41.3.3</b> Module 3: Traceback</a></li>
</ul></li>
<li class="chapter" data-level="41.4" data-path="blast-summary.html"><a href="blast-summary.html#blast-alignment-statistics"><i class="fa fa-check"></i><b>41.4</b> BLAST alignment statistics</a></li>
</ul></li>
<li class="chapter" data-level="42" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html"><i class="fa fa-check"></i><b>42</b> BLAST Alignment score distribution</a>
<ul>
<li class="chapter" data-level="42.1" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#packages-9"><i class="fa fa-check"></i><b>42.1</b> Packages</a></li>
<li class="chapter" data-level="42.2" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#distributions-of-alignment-scores"><i class="fa fa-check"></i><b>42.2</b> Distributions of alignment scores</a></li>
<li class="chapter" data-level="42.3" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-data"><i class="fa fa-check"></i><b>42.3</b> Simulating data</a></li>
<li class="chapter" data-level="42.4" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#lots-of-data-in-biology-is-normal"><i class="fa fa-check"></i><b>42.4</b> Lots of data in biology is normal</a></li>
<li class="chapter" data-level="42.5" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-the-normal-distribution"><i class="fa fa-check"></i><b>42.5</b> Simulating the normal distribution</a></li>
<li class="chapter" data-level="42.6" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#the-extreme-value-distribution"><i class="fa fa-check"></i><b>42.6</b> The extreme value distribution</a></li>
<li class="chapter" data-level="42.7" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#replicating-figure-4.14"><i class="fa fa-check"></i><b>42.7</b> Replicating Figure 4.14</a></li>
<li class="chapter" data-level="42.8" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#why-do-we-care"><i class="fa fa-check"></i><b>42.8</b> Why do we care?</a></li>
</ul></li>
<li class="chapter" data-level="43" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><i class="fa fa-check"></i><b>43</b> Calculating E values, Bit scores, and P-values from BLAST output</a>
<ul>
<li class="chapter" data-level="43.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-7"><i class="fa fa-check"></i><b>43.1</b> Introduction</a></li>
<li class="chapter" data-level="43.2" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#case-study"><i class="fa fa-check"></i><b>43.2</b> Case Study</a></li>
<li class="chapter" data-level="43.3" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-statistics"><i class="fa fa-check"></i><b>43.3</b> BLAST Statistics</a></li>
<li class="chapter" data-level="43.4" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot"><i class="fa fa-check"></i><b>43.4</b> Dotplot</a>
<ul>
<li class="chapter" data-level="43.4.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot-in-r---optional-experimental-code"><i class="fa fa-check"></i><b>43.4.1</b> Dotplot in R - optional, experimental code</a></li>
</ul></li>
<li class="chapter" data-level="43.5" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-search-summary"><i class="fa fa-check"></i><b>43.5</b> BLAST “Search Summary”</a></li>
<li class="chapter" data-level="43.6" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-alignments"><i class="fa fa-check"></i><b>43.6</b> BLAST Alignments</a></li>
<li class="chapter" data-level="43.7" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#equations"><i class="fa fa-check"></i><b>43.7</b> Equations</a>
<ul>
<li class="chapter" data-level="43.7.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-value-eq-4.5-pg-142"><i class="fa fa-check"></i><b>43.7.1</b> E value (eq 4.5, pg 142)</a></li>
</ul></li>
<li class="chapter" data-level="43.8" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bit-scores-s"><i class="fa fa-check"></i><b>43.8</b> Bit Scores (S’)</a></li>
<li class="chapter" data-level="43.9" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-from-bit-scores"><i class="fa fa-check"></i><b>43.9</b> E from Bit scores</a></li>
<li class="chapter" data-level="43.10" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#p-values"><i class="fa fa-check"></i><b>43.10</b> P values</a></li>
<li class="chapter" data-level="43.11" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#repliminaries"><i class="fa fa-check"></i><b>43.11</b> Repliminaries</a></li>
<li class="chapter" data-level="43.12" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#concepts"><i class="fa fa-check"></i><b>43.12</b> Concepts</a></li>
<li class="chapter" data-level="43.13" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#functions-4"><i class="fa fa-check"></i><b>43.13</b> Functions</a></li>
<li class="chapter" data-level="43.14" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-8"><i class="fa fa-check"></i><b>43.14</b> Introduction</a></li>
<li class="chapter" data-level="43.15" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees"><i class="fa fa-check"></i><b>43.15</b> Number of rooted trees</a></li>
<li class="chapter" data-level="43.16" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees-in-r"><i class="fa fa-check"></i><b>43.16</b> Number of rooted trees in R</a>
<ul>
<li class="chapter" data-level="43.16.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#unique-topologies-versus-rotations"><i class="fa fa-check"></i><b>43.16.1</b> Unique topologies versus rotations</a></li>
</ul></li>
<li class="chapter" data-level="43.17" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bilography"><i class="fa fa-check"></i><b>43.17</b> Bilography</a></li>
</ul></li>
<li class="chapter" data-level="44" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html"><i class="fa fa-check"></i><b>44</b> UPGMA the hard way</a>
<ul>
<li class="chapter" data-level="44.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#introudction-to-upgma"><i class="fa fa-check"></i><b>44.1</b> Introudction to UPGMA</a></li>
<li class="chapter" data-level="44.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#preliminaries-18"><i class="fa fa-check"></i><b>44.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="44.2.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#r-libraries"><i class="fa fa-check"></i><b>44.2.1</b> R Libraries</a></li>
<li class="chapter" data-level="44.2.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#custom-function-well-use"><i class="fa fa-check"></i><b>44.2.2</b> Custom function we’ll use</a></li>
</ul></li>
<li class="chapter" data-level="44.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#data-matrix"><i class="fa fa-check"></i><b>44.3</b> Data matrix</a>
<ul>
<li class="chapter" data-level="44.3.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#building-the-matrix"><i class="fa fa-check"></i><b>44.3.1</b> Building the matrix</a></li>
</ul></li>
<li class="chapter" data-level="44.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#what-does-this-distance-matrix-represent"><i class="fa fa-check"></i><b>44.4</b> What does this distance matrix represent?</a>
<ul>
<li class="chapter" data-level="44.4.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-pairwise-linear-distances"><i class="fa fa-check"></i><b>44.4.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="44.4.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-2d-map-of-linear-distances"><i class="fa fa-check"></i><b>44.4.2</b> Genetic distances conceptualized as 2D map of linear distances</a></li>
</ul></li>
<li class="chapter" data-level="44.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#from-linear-distances-to-tree-distances"><i class="fa fa-check"></i><b>44.5</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="44.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#the-upgma-algorithm"><i class="fa fa-check"></i><b>44.6</b> The UPGMA algorithm</a>
<ul>
<li class="chapter" data-level="44.6.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-1-find-the-2-taxa-that-are-currently-closest-together"><i class="fa fa-check"></i><b>44.6.1</b> Step 1: find the 2 taxa that are currently closest together</a></li>
<li class="chapter" data-level="44.6.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node"><i class="fa fa-check"></i><b>44.6.2</b> Step 2: calculate the branch lengths connect the taxa to their shared node</a></li>
<li class="chapter" data-level="44.6.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-determine-if-the-algorithm-is-done"><i class="fa fa-check"></i><b>44.6.3</b> Step 3a: determine if the algorithm is done</a></li>
<li class="chapter" data-level="44.6.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u"><i class="fa fa-check"></i><b>44.6.4</b> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</a></li>
<li class="chapter" data-level="44.6.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-4-determine-the-distance-from-u-to-the-other-remaining-clusters."><i class="fa fa-check"></i><b>44.6.5</b> Step 4: Determine the distance from u to the other remaining clusters.</a></li>
<li class="chapter" data-level="44.6.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-5-create-a-new-matrix"><i class="fa fa-check"></i><b>44.6.6</b> Step 5: Create a new matrix</a></li>
<li class="chapter" data-level="44.6.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa"><i class="fa fa-check"></i><b>44.6.7</b> Key step: calculating the distance from a clade to the remaining taxa</a></li>
</ul></li>
<li class="chapter" data-level="44.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-2d-representation"><i class="fa fa-check"></i><b>44.7</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="44.8" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---round-1"><i class="fa fa-check"></i><b>44.8</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="44.8.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#determine-minimum-value-in-current-matrix"><i class="fa fa-check"></i><b>44.8.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="44.8.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade"><i class="fa fa-check"></i><b>44.8.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="44.8.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length"><i class="fa fa-check"></i><b>44.8.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="44.8.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>44.8.4</b> Calculate distance from clade to all other points</a></li>
<li class="chapter" data-level="44.8.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-matrix-with-new-distance"><i class="fa fa-check"></i><b>44.8.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="44.9" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---next-round"><i class="fa fa-check"></i><b>44.9</b> Algorithm - next round</a>
<ul>
<li class="chapter" data-level="44.9.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-1"><i class="fa fa-check"></i><b>44.9.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="44.9.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-1"><i class="fa fa-check"></i><b>44.9.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="44.9.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#recalculate-distance-from-abe-to-all-remainign-taxa"><i class="fa fa-check"></i><b>44.9.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="44.9.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-the-matrix"><i class="fa fa-check"></i><b>44.9.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="44.10" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#next-iteration"><i class="fa fa-check"></i><b>44.10</b> Next iteration</a>
<ul>
<li class="chapter" data-level="44.10.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-lengths"><i class="fa fa-check"></i><b>44.10.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="44.10.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-2"><i class="fa fa-check"></i><b>44.10.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="44.10.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-2"><i class="fa fa-check"></i><b>44.10.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="44.10.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#distance-of-c-to-abe-and-d-to-abe"><i class="fa fa-check"></i><b>44.10.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="44.11" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finish-up"><i class="fa fa-check"></i><b>44.11</b> Finish up</a>
<ul>
<li class="chapter" data-level="44.11.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-3"><i class="fa fa-check"></i><b>44.11.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="44.11.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-3"><i class="fa fa-check"></i><b>44.11.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="44.12" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finalizing-branch-lengths"><i class="fa fa-check"></i><b>44.12</b> Finalizing branch lengths</a></li>
</ul></li>
<li class="chapter" data-level="45" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html"><i class="fa fa-check"></i><b>45</b> Representing phylogenetic trees in Newick format</a>
<ul>
<li class="chapter" data-level="45.1" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#vocab-5"><i class="fa fa-check"></i><b>45.1</b> Vocab</a></li>
<li class="chapter" data-level="45.2" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#introduction-9"><i class="fa fa-check"></i><b>45.2</b> Introduction</a></li>
<li class="chapter" data-level="45.3" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-three-taxa"><i class="fa fa-check"></i><b>45.3</b> Trees with three taxa</a></li>
<li class="chapter" data-level="45.4" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-four-taxa"><i class="fa fa-check"></i><b>45.4</b> Trees with four taxa</a></li>
</ul></li>
<li class="chapter" data-level="46" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html"><i class="fa fa-check"></i><b>46</b> Plotting trees from Newick notation in R</a>
<ul>
<li class="chapter" data-level="46.1" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#preliminaries-19"><i class="fa fa-check"></i><b>46.1</b> Preliminaries</a></li>
<li class="chapter" data-level="46.2" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#note"><i class="fa fa-check"></i><b>46.2</b> Note</a></li>
<li class="chapter" data-level="46.3" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick"><i class="fa fa-check"></i><b>46.3</b> 3 taxa tree in Newick</a></li>
<li class="chapter" data-level="46.4" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick-1"><i class="fa fa-check"></i><b>46.4</b> 4 taxa tree in Newick</a></li>
<li class="chapter" data-level="46.5" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-in-newick-format"><i class="fa fa-check"></i><b>46.5</b> 5 taxa in Newick Format</a></li>
<li class="chapter" data-level="46.6" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#many-taxa-tree"><i class="fa fa-check"></i><b>46.6</b> Many-taxa tree</a></li>
<li class="chapter" data-level="46.7" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#newick-and-branch-lengths"><i class="fa fa-check"></i><b>46.7</b> Newick and Branch lengths</a></li>
<li class="chapter" data-level="46.8" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#additional-reading"><i class="fa fa-check"></i><b>46.8</b> Additional Reading</a></li>
</ul></li>
<li class="chapter" data-level="47" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><i class="fa fa-check"></i><b>47</b> Phylogenetic trees using the UPGMA clustering algoirthm</a>
<ul>
<li class="chapter" data-level="47.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#introduction-10"><i class="fa fa-check"></i><b>47.1</b> Introduction</a></li>
<li class="chapter" data-level="47.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#preliminaries-20"><i class="fa fa-check"></i><b>47.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="47.2.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#vocab-6"><i class="fa fa-check"></i><b>47.2.1</b> Vocab</a></li>
<li class="chapter" data-level="47.2.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-r-functions"><i class="fa fa-check"></i><b>47.2.2</b> General R functions</a></li>
<li class="chapter" data-level="47.2.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#specific-functions"><i class="fa fa-check"></i><b>47.2.3</b> Specific functions</a></li>
</ul></li>
<li class="chapter" data-level="47.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#sequence-comparisons"><i class="fa fa-check"></i><b>47.3</b> Sequence comparisons</a></li>
<li class="chapter" data-level="47.4" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrices"><i class="fa fa-check"></i><b>47.4</b> Distance matrices</a></li>
<li class="chapter" data-level="47.5" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#from-a-distance-matrix-to-phylogenetic-tree"><i class="fa fa-check"></i><b>47.5</b> From a distance matrix to phylogenetic tree</a></li>
<li class="chapter" data-level="47.6" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-raw-sequences"><i class="fa fa-check"></i><b>47.6</b> Distance matrix in R from raw sequences</a></li>
<li class="chapter" data-level="47.7" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-normal-matrix"><i class="fa fa-check"></i><b>47.7</b> Distance matrix in R from normal matrix</a></li>
<li class="chapter" data-level="47.8" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#upgma-in-r"><i class="fa fa-check"></i><b>47.8</b> UPGMA in R</a></li>
<li class="chapter" data-level="47.9" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#how-upgma-clustering-works"><i class="fa fa-check"></i><b>47.9</b> How UPGMA clustering works</a>
<ul>
<li class="chapter" data-level="47.9.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#challenge"><i class="fa fa-check"></i><b>47.9.1</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="47.10" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#further-information"><i class="fa fa-check"></i><b>47.10</b> Further information</a>
<ul>
<li class="chapter" data-level="47.10.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-info"><i class="fa fa-check"></i><b>47.10.1</b> General info</a></li>
<li class="chapter" data-level="47.10.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#worked-examples"><i class="fa fa-check"></i><b>47.10.2</b> Worked examples:</a></li>
<li class="chapter" data-level="47.10.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#additional-notes"><i class="fa fa-check"></i><b>47.10.3</b> Additional notes:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="48" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><i class="fa fa-check"></i><b>48</b> Tutorial: Writing functions to calculate the number of phylogenetic trees</a>
<ul>
<li class="chapter" data-level="48.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#repliminaries-1"><i class="fa fa-check"></i><b>48.1</b> Repliminaries</a>
<ul>
<li class="chapter" data-level="48.1.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-5"><i class="fa fa-check"></i><b>48.1.1</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="48.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#concepts-1"><i class="fa fa-check"></i><b>48.2</b> Concepts</a>
<ul>
<li class="chapter" data-level="48.2.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vocab-7"><i class="fa fa-check"></i><b>48.2.1</b> Vocab:</a></li>
<li class="chapter" data-level="48.2.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#packages-10"><i class="fa fa-check"></i><b>48.2.2</b> Packages:</a></li>
</ul></li>
<li class="chapter" data-level="48.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#introduction-11"><i class="fa fa-check"></i><b>48.3</b> Introduction</a></li>
<li class="chapter" data-level="48.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-1"><i class="fa fa-check"></i><b>48.4</b> Number of rooted trees</a></li>
<li class="chapter" data-level="48.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-in-r-1"><i class="fa fa-check"></i><b>48.5</b> Number of rooted trees in R</a></li>
<li class="chapter" data-level="48.6" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-in-r"><i class="fa fa-check"></i><b>48.6</b> Functions in R</a></li>
<li class="chapter" data-level="48.7" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-to-calculate-the-number-of-possible-phylogenetic-trees"><i class="fa fa-check"></i><b>48.7</b> Function to calculate the number of possible phylogenetic trees</a></li>
<li class="chapter" data-level="48.8" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vectorized-inputs-to-functions"><i class="fa fa-check"></i><b>48.8</b> Vectorized inputs to functions</a></li>
<li class="chapter" data-level="48.9" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-conditional-statements"><i class="fa fa-check"></i><b>48.9</b> Adding conditional statements</a></li>
<li class="chapter" data-level="48.10" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-multiple-conditional-statements"><i class="fa fa-check"></i><b>48.10</b> Adding multiple conditional statements</a></li>
<li class="chapter" data-level="48.11" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-additional-arguments"><i class="fa fa-check"></i><b>48.11</b> Adding additional arguments</a></li>
<li class="chapter" data-level="48.12" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-_un_rooted-trees"><i class="fa fa-check"></i><b>48.12</b> Assignment: _un_rooted trees</a></li>
<li class="chapter" data-level="48.13" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-1"><i class="fa fa-check"></i><b>48.13</b> Assignment part 1</a>
<ul>
<li class="chapter" data-level="48.13.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#part-1-answer"><i class="fa fa-check"></i><b>48.13.1</b> Part 1 answer</a></li>
</ul></li>
<li class="chapter" data-level="48.14" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-2"><i class="fa fa-check"></i><b>48.14</b> Assignment part 2</a>
<ul>
<li class="chapter" data-level="48.14.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#answer"><i class="fa fa-check"></i><b>48.14.1</b> Answer</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="49" data-path="writing-functions.html"><a href="writing-functions.html"><i class="fa fa-check"></i><b>49</b> Writing functions</a>
<ul>
<li class="chapter" data-level="49.1" data-path="writing-functions.html"><a href="writing-functions.html#preface-4"><i class="fa fa-check"></i><b>49.1</b> Preface</a></li>
<li class="chapter" data-level="49.2" data-path="writing-functions.html"><a href="writing-functions.html#vocab-8"><i class="fa fa-check"></i><b>49.2</b> Vocab</a></li>
<li class="chapter" data-level="49.3" data-path="writing-functions.html"><a href="writing-functions.html#functions-6"><i class="fa fa-check"></i><b>49.3</b> Functions</a></li>
<li class="chapter" data-level="49.4" data-path="writing-functions.html"><a href="writing-functions.html#functions-in-r-1"><i class="fa fa-check"></i><b>49.4</b> Functions in R</a></li>
<li class="chapter" data-level="49.5" data-path="writing-functions.html"><a href="writing-functions.html#comments-in-r"><i class="fa fa-check"></i><b>49.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="50" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html"><i class="fa fa-check"></i><b>50</b> Writing functions - practice problems</a>
<ul>
<li class="chapter" data-level="" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#introduction-12"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="50.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#pure-practice"><i class="fa fa-check"></i><b>50.1</b> Pure practice</a>
<ul>
<li class="chapter" data-level="50.1.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#hello-world"><i class="fa fa-check"></i><b>50.1.1</b> Hello world!</a></li>
<li class="chapter" data-level="50.1.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______"><i class="fa fa-check"></i><b>50.1.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="50.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function"><i class="fa fa-check"></i><b>50.2</b> Practice Function 3: Make you own natural log function</a></li>
<li class="chapter" data-level="50.3" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-4-make-function-that-converts-dna-directly-to-rna"><i class="fa fa-check"></i><b>50.3</b> Practice Function 4: Make function that converts DNA directly to RNA</a></li>
<li class="chapter" data-level="50.4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#key"><i class="fa fa-check"></i><b>50.4</b> Key</a>
<ul>
<li class="chapter" data-level="50.4.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-1-hello-world"><i class="fa fa-check"></i><b>50.4.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="50.4.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______-1"><i class="fa fa-check"></i><b>50.4.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="50.5" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function-1"><i class="fa fa-check"></i><b>50.5</b> Practice Function 3: Make you own natural log function</a></li>
</ul></li>
<li class="chapter" data-level="51" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html"><i class="fa fa-check"></i><b>51</b> Simulating random amino acid sequences</a>
<ul>
<li class="chapter" data-level="51.1" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#selecting-random-letters-from-a-vector"><i class="fa fa-check"></i><b>51.1</b> Selecting random letters from a vector</a></li>
<li class="chapter" data-level="51.2" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-letters-from-the-whole-alphabet"><i class="fa fa-check"></i><b>51.2</b> Select random letters from the whole alphabet</a></li>
<li class="chapter" data-level="51.3" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-amino-acids-to-build-polypeptide"><i class="fa fa-check"></i><b>51.3</b> Select random amino acids to build polypeptide</a></li>
</ul></li>
<li class="chapter" data-level="52" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html"><i class="fa fa-check"></i><b>52</b> Writing user-friendly functions</a>
<ul>
<li class="chapter" data-level="" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#introduction-13"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="52.1" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-1"><i class="fa fa-check"></i><b>52.1</b> Version 1</a></li>
<li class="chapter" data-level="52.2" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-2"><i class="fa fa-check"></i><b>52.2</b> Version 2</a></li>
<li class="chapter" data-level="52.3" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-3"><i class="fa fa-check"></i><b>52.3</b> Version 3</a></li>
<li class="chapter" data-level="52.4" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-4-adding-conditions-and-warnigns"><i class="fa fa-check"></i><b>52.4</b> Version 4: Adding conditions and warnigns</a></li>
<li class="chapter" data-level="52.5" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-5"><i class="fa fa-check"></i><b>52.5</b> Version 5</a></li>
</ul></li>
<li class="chapter" data-level="53" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>53</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="53.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-5"><i class="fa fa-check"></i><b>53.1</b> Preface</a></li>
<li class="chapter" data-level="53.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab-9"><i class="fa fa-check"></i><b>53.2</b> Vocab</a></li>
<li class="chapter" data-level="53.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-7"><i class="fa fa-check"></i><b>53.3</b> Functions</a></li>
<li class="chapter" data-level="53.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>53.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="53.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>53.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="54" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html"><i class="fa fa-check"></i><b>54</b> More on for() loops in R</a>
<ul>
<li class="chapter" data-level="54.1" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#introduction-14"><i class="fa fa-check"></i><b>54.1</b> Introduction</a></li>
<li class="chapter" data-level="54.2" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#reminders"><i class="fa fa-check"></i><b>54.2</b> Reminders</a></li>
<li class="chapter" data-level="54.3" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#for-loop-1-the-hard-coded-for-loop-dont-actually-do-this"><i class="fa fa-check"></i><b>54.3</b> For loop 1: The hard-coded for loop (don’t actually do this!)</a></li>
<li class="chapter" data-level="54.4" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#for-loop-3-the-classic-for-loop"><i class="fa fa-check"></i><b>54.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="54.5" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#for-loop-3-the-power-user-for-loop"><i class="fa fa-check"></i><b>54.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="54.6" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#for-loop-4-a-for-loop-for-all-purposes"><i class="fa fa-check"></i><b>54.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="54.7" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#advanced-aside-if-else-and-next-statements"><i class="fa fa-check"></i><b>54.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="54.8" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#on-avoiding-for-loops"><i class="fa fa-check"></i><b>54.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="54.9" data-path="more-on-for-loops-in-r.html"><a href="more-on-for-loops-in-r.html#challenge-1"><i class="fa fa-check"></i><b>54.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="55" data-path="random-number-generation.html"><a href="random-number-generation.html"><i class="fa fa-check"></i><b>55</b> Random number generation</a>
<ul>
<li class="chapter" data-level="55.1" data-path="random-number-generation.html"><a href="random-number-generation.html#introduction-15"><i class="fa fa-check"></i><b>55.1</b> Introduction</a></li>
<li class="chapter" data-level="55.2" data-path="random-number-generation.html"><a href="random-number-generation.html#using-set.seed-to-make-simulations-reproducible"><i class="fa fa-check"></i><b>55.2</b> Using set.seed() to make simulations reproducible</a></li>
<li class="chapter" data-level="55.3" data-path="random-number-generation.html"><a href="random-number-generation.html#types-of-random-number-generators"><i class="fa fa-check"></i><b>55.3</b> Types of random number generators</a></li>
<li class="chapter" data-level="55.4" data-path="random-number-generation.html"><a href="random-number-generation.html#reporting-simulations"><i class="fa fa-check"></i><b>55.4</b> Reporting simulations</a></li>
<li class="chapter" data-level="55.5" data-path="random-number-generation.html"><a href="random-number-generation.html#notes"><i class="fa fa-check"></i><b>55.5</b> Notes:</a></li>
</ul></li>
<li class="chapter" data-level="56" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>56</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="56.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>56.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="56.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>56.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="56.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>56.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="56.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>56.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="56.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>56.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-3"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="56.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>56.2</b> Help!</a>
<ul>
<li class="chapter" data-level="56.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>56.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="56.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>56.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="56.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>56.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="56.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>56.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="56.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>56.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="56.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>56.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="56.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>56.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="56.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>56.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="proteins-review" class="section level1" number="2">
<h1><span class="header-section-number">Chapter 2</span> Proteins</h1>
<p><strong>Authors</strong>: OpenStax / LibreText. Formatted in RMarkdown by Nathan Brouwer under the Creative Commons Attribution License 4.0 license.</p>
<p>This chapter was adapted from <a href="">LibreText General Biology</a>, Chapter 3, Section 3.4: <a href="https://bio.libretexts.org/Bookshelves/Introductory_and_General_Biology/Book%3A_General_Biology_(OpenStax)/1%3A_The_Chemistry_of_Life/3%3A_Biological_Macromolecules/3.4%3A_Proteins">Proteins</a>. The LibreText book is based on <a href="https://openstax.org/details/books/biology-2e">OpenStax Biology 2nd edition</a>, Chapter 3, Section 3.4: <a href="https://openstax.org/books/biology-2e/pages/3-4-proteins">Nucleic Acids</a>. A full list of authors is found under the <strong>Contributors and Attributions</strong> section at the end of this document.</p>
<div id="skills-to-develop" class="section level3" number="2.0.1">
<h3><span class="header-section-number">2.0.1</span> Skills to develop:</h3>
<ol style="list-style-type: decimal">
<li>Describe the functions proteins perform in the cell and in tissues</li>
<li>Discuss the relationship between amino acids and proteins</li>
<li>Explain the four levels of protein organization</li>
<li>Describe the ways in which protein shape and function are linked</li>
</ol>
<p>Proteins are one of the most abundant organic molecules in living systems and have the most diverse range of functions of all <strong>macromolecules</strong>. Proteins may be structural, regulatory, contractile, or protective; they may serve in transport, storage, or membranes; or they may be toxins or enzymes. Each cell in a living system may contain thousands of proteins, each with a unique function. Their structures, like their functions, vary greatly. They are all, however, <strong>polymers</strong> of <strong>amino acids</strong>, arranged in a linear sequence.</p>
</div>
<div id="types-and-functions-of-proteins" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Types and Functions of Proteins</h2>
<p><strong>Enzymes</strong>, which are produced by living cells, are <strong>catalysts</strong> in biochemical reactions (like digestion) and are usually complex or conjugated proteins. Each enzyme is specific for the substrate (a reactant that binds to an enzyme) it acts on. The enzyme may help in breakdown, rearrangement, or synthesis reactions. Enzymes that break down their substrates are called <strong>catabolic enzymes</strong>, enzymes that build more complex molecules from their substrates are called <strong>anabolic enzymes</strong>, and enzymes that affect the rate of reaction are called <strong>catalytic enzymes</strong>.
<!-- TODO: is this terminology standard?  -->
It should be noted that all enzymes increase the rate of reaction and, therefore, are considered to be organic catalysts. An example of an enzyme is salivary amylase, which
hydrolyzes its substrate amylose, a component of starch.</p>
<p><strong>Hormones</strong> are chemical-signaling molecules, usually small proteins or steroids, secreted by endocrine cells that act to control or regulate specific physiological processes, including growth, development, metabolism, and reproduction. For example, insulin is a <strong>protein hormone</strong> that helps to regulate the blood glucose level.</p>
<!-- TODO: Fix table -->
<!-- The primary types and functions of proteins are listed in Table 3.4.1. -->
<!-- **Table** 3.4.1***:*** Protein Types and Functions -->
<!-- Digestive Enzymes Amylase, lipase, pepsin, trypsin Help in digestion -->
<!-- of food by catabolizing -->
<!-- nutrients into monomeric units -->
<!-- Structural Actin, tubulin, keratin Construct different structures, -->
<!-- like the cytoskeleton -->
<!-- Defense Immunoglobulins Protect the body from foreign pathogens -->
<!-- Storage Legume storage proteins, egg white (albumin) Provide -->
<!-- nourishment in early development of -->
<!-- the embryo and the seedling -->
<p>Proteins have different shapes and molecular weights; some proteins are <strong>globular</strong> in shape whereas others are fibrous. For example, hemoglobin is a globular protein, but collagen, found in our skin, is a fibrous protein. Protein shape is critical to its function, and this shape is maintained by many different types of chemical bonds. Changes in temperature, pH, and exposure to chemicals may lead to permanent changes in the shape of the protein, leading to loss of function, known as <strong>denaturation</strong>. All proteins are made up of different arrangements of the same 20 types of <strong>amino acids</strong>.</p>
<!-- shoutout to non-cannonical AAs? -->
</div>
<div id="amino-acids" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Amino Acids</h2>
<p>Amino acids are the <strong>monomers</strong> that make up proteins. Each amino acid has the same fundamental structure, which consists of a central carbon atom, also known as the alpha (<em>α</em>) carbon, bonded to an amino group (NH<sub>2</sub>), a carboxyl group (COOH), and to a hydrogen atom. Every amino acid also has another atom or group of atoms bonded to the central atom known as the R group (Figure 3.4.1).</p>
<div class="figure"><span id="fig:unnamed-chunk-7"></span>
<img src="images/protimage1.jpeg" alt="Amino acids have a central asymmetric carbon to which an amino group, a carboxyl group, a hydrogen atom, and a side chain (R group) are attached." width="258" />
<p class="caption">
Figure 2.1: Amino acids have a central asymmetric carbon to which an amino group, a carboxyl group, a hydrogen atom, and a side chain (R group) are attached.
</p>
</div>
<p>The name “amino acid” is derived from the fact that they contain both an amino group and carboxyl-acid-group in their basic structure. For each amino acid, the R group (or side chain) is different (Figure 3.4.2).
<!-- As mentioned, there are 20 amino acids present in proteins. Ten of these are considered essential amino acids in humans because the human body cannot produce them and they are obtained from the diet.  Just as some fatty acids are essential to a diet, some amino acids are necessary as well. They are
known as essential amino acids, and in humans they include isoleucine, leucine, and cysteine. Essential amino acids refer to those necessary for construction of proteins in the body, although not produced by the body; which amino acids are essential varies from organism to organism.--></p>
<div class="figure"><span id="fig:unnamed-chunk-8"></span>
<img src="images/protimage2.jpeg" alt="There are 20 common amino acids commonly found in proteins, each with a different R group (variant group) that determines its chemical nature." width="60%" />
<p class="caption">
Figure 2.2: There are 20 common amino acids commonly found in proteins, each with a different R group (variant group) that determines its chemical nature.
</p>
</div>
<!-- TODO:awk transition -->
<p>Which categories of amino acid would you expect to find on the surface
of a soluble protein, and which would you expect to find in the
interior? What distribution of amino acids would you expect to find in
a protein embedded in a lipid bilayer?</p>
<p>The chemical nature of the side chain determines the nature of the amino acid (that is, whether it is <strong>acidic</strong>, <strong>basic</strong>, <strong>polar</strong>, or <strong>nonpolar</strong>). For example, the amino acid glycine has a hydrogen atom as the R group. Amino acids such as valine, methionine, and alanine are nonpolar or hydrophobic, while amino acids such as serine, threonine, and cysteine are polar and have hydrophilic side chains. The side chains of lysine and arginine are positively charged, and therefore these amino acids are also known as basic amino acids. Proline has an R group that is linked to the amino group, forming a ring-like structure. Proline is an exception to the standard structure of an animo acid since its amino group is not separate from the side chain
<!-- (Figure 3.4.2). --></p>
<p>Amino acids are represented by a single upper case letter or a three-letter abbreviation. For example, valine is known by the letter V or the three-letter code val.</p>
<p>The sequence and the number of amino acids ultimately determine a protein’s shape, size, and function. Each amino acid is attached to another amino acid by a covalent bond, known as a <strong>peptide bond</strong>, which is formed by a dehydration reaction. The carboxyl group of one amino acid and the amino group of the incoming amino acid combine, releasing a molecule of water. The resulting bond is the peptide bond
<!-- (Figure 3.4.3). --></p>
<div class="figure"><span id="fig:unnamed-chunk-9"></span>
<img src="images/protimage3.jpeg" alt="Peptide bond formation is a dehydration synthesis reaction. The carboxyl group of one amino acid is linked to the amino group of the incoming amino acid. In the process, a molecule of water is released." width="210" />
<p class="caption">
Figure 2.3: Peptide bond formation is a dehydration synthesis reaction. The carboxyl group of one amino acid is linked to the amino group of the incoming amino acid. In the process, a molecule of water is released.
</p>
</div>
<p>The products formed by such linkages are called <strong>peptides</strong>. As more amino acids join to this growing chain, the resulting chain is known as a <strong>polypeptide</strong>. Each polypeptide has a free amino group at one end. This end is called the <strong>N-terminal</strong>, or the amino terminal, and the other end has a free carboxyl group, also known as the C- or carboxyl
terminal. The terms polypeptide and protein are sometimes used interchangeably, a polypeptide is technically a polymer of amino acids, whereas the term protein is more formally used for a polypeptide or polypeptides that have combined together, often have bound non-peptide prosthetic groups, have a distinct shape, and have a unique function.
<!-- is this a standard definition? --></p>
<p>After <strong>protein synthesis (translation)</strong>, most proteins are modified. These are known as <strong>post-translational modifications</strong>. They may undergo <strong>cleavage</strong>, <strong>phosphorylation</strong> (addition of a phosphate group), or may require the addition of other chemical groups. Only after these modifications is the protein completely functional.</p>
</div>
<div id="evolution-connection" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Evolution Connection:</h2>
<hr />
<p><strong>The Evolutionary Significance of Cytochrome</strong>: Cytochrome c is an
important component of the electron transport chain, a part of
cellular respiration, and it is normally found in the cellular
organelle, the mitochondrion. This protein has a heme prosthetic
group, and the central ion of the heme gets alternately reduced and
oxidized during electron transfer. Because this essential protein’s
role in producing cellular energy is crucial, it has changed very
little over millions of years. Protein sequencing has shown that there
is a considerable amount of cytochrome c amino acid sequence <strong>homology</strong>
among different species; in other words, evolutionary kinship can be
assessed by measuring the similarities or differences among various
species’ DNA or protein sequences.</p>
<p>Scientists have determined that human cytochrome c contains 104 amino
acids. For each cytochrome c molecule from different organisms that
has been sequenced to date, 37 of these amino acids appear in the same
position in all samples of cytochrome c. This indicates that there may
have been a <strong>common ancestor</strong>. On comparing the human and chimpanzee
protein sequences, no sequence difference was found. When human and
rhesus monkey sequences were compared, the single difference found was
in one amino acid. In another comparison, human to yeast sequencing
shows a difference in the 44th position.</p>
<hr />
</div>
<div id="protein-structure" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Protein Structure</h2>
<p>The shape of a protein is critical to its function. For example, an enzyme can bind to a specific substrate at a site known as the <strong>active site</strong>. If this active site is altered because of local changes or changes in overall protein structure, the enzyme may be unable to bind to the substrate. To understand how the protein gets its final shape or conformation, we need to understand the four levels of protein structure: primary, secondary, tertiary, and quaternary.</p>
<div id="primary-structure" class="section level3" number="2.4.1">
<h3><span class="header-section-number">2.4.1</span> Primary Structure</h3>
<p>The unique sequence of amino acids in a polypeptide chain is its
primary structure. For example, the pancreatic hormone insulin has two
polypeptide chains, A and B, and they are linked together by disulfide
bonds. The N terminal amino acid of the A chain is glycine, whereas
the C terminal amino acid is asparagine (Figure 3.4.4). The sequences
of amino acids in the A and B chains are unique to insulin.</p>
<div class="figure"><span id="fig:unnamed-chunk-10"></span>
<img src="images/protimage4.jpeg" alt="Bovine serum insulin is a protein hormone made of two peptide chains, A (21 amino acids long) and B (30 amino acids long). In each chain, primary structure is indicated by three-letter abbreviations that represent the names of the amino acids in the order they are present. The amino acid cysteine (cys) has a sulfhydryl (SH) group as a side chain. Two sulfhydryl groups can react in the presence of oxygen to form a disulfide (S-S) bond. Two disulfide bonds connect the A and B chains together, and a third helps the A chain fold into the correct shape. Note that all disulfide bonds are the same length, but are drawn different sizes for clarity." width="40%" />
<p class="caption">
Figure 2.4: Bovine serum insulin is a protein hormone made of two peptide chains, A (21 amino acids long) and B (30 amino acids long). In each chain, primary structure is indicated by three-letter abbreviations that represent the names of the amino acids in the order they are present. The amino acid cysteine (cys) has a sulfhydryl (SH) group as a side chain. Two sulfhydryl groups can react in the presence of oxygen to form a disulfide (S-S) bond. Two disulfide bonds connect the A and B chains together, and a third helps the A chain fold into the correct shape. Note that all disulfide bonds are the same length, but are drawn different sizes for clarity.
</p>
</div>
<p>The unique sequence for every protein is ultimately determined by the
gene encoding the protein. A change in nucleotide sequence of the
gene’s coding region may lead to a different amino acid being added to
the growing polypeptide chain, causing a change in protein structure
and function. In sickle cell anemia, the hemoglobin <em>β</em> chain (a small
portion of which is shown in Figure 3.4.5) has a single amino acid
substitution, causing a change in protein structure and function.
Specifically, the amino acid glutamic acid is substituted by valine in
the beta chain. What is most remarkable to consider is that a
hemoglobin molecule is made up of two alpha chains and two beta chains
that each consist of about 150 amino acids. The molecule, therefore,
has about 600 amino acids. The structural difference between a normal
hemoglobin molecule and a sickle cell molecule—which dramatically
decreases life expectancy—is a single amino acid of the 600. What is
even more remarkable is that those 600 amino acids are encoded by
three nucleotides each, and the mutation is caused by a single base
change (point mutation), 1 in 1800 bases.</p>
<div class="figure"><span id="fig:unnamed-chunk-11"></span>
<img src="images/protimage5.jpeg" alt="The beta chain of hemoglobin is 147 residues in length, yet a single amino acid substitution leads to sickle cell anemia. In normal hemoglobin, the amino acid at position seven is glutamate. In sickle cell hemoglobin, this glutamate is replaced by a valine." width="40%" />
<p class="caption">
Figure 2.5: The beta chain of hemoglobin is 147 residues in length, yet a single amino acid substitution leads to sickle cell anemia. In normal hemoglobin, the amino acid at position seven is glutamate. In sickle cell hemoglobin, this glutamate is replaced by a valine.
</p>
</div>
<p>Because of this change of one amino acid in the chain, hemoglobin
molecules form long fibers that distort the biconcave, or disc-shaped,
red blood cells and assume a crescent or “sickle” shape, which clogs
arteries (Figure 3.4.6). This can lead to myriad serious health
problems such as breathlessness, dizziness, headaches, and abdominal
pain for those affected by this disease.</p>
<div class="figure"><span id="fig:unnamed-chunk-12"></span>
<img src="images/protimage6.jpeg" alt="In this blood smear, visualized at 535x magnification using bright field microscopy, sickle cells are crescent shaped, while normal cells are disc-shaped. (credit: modification of work by Ed Uthman; scale-bar data from Matt Russell)" width="60%" />
<p class="caption">
Figure 2.6: In this blood smear, visualized at 535x magnification using bright field microscopy, sickle cells are crescent shaped, while normal cells are disc-shaped. (credit: modification of work by Ed Uthman; scale-bar data from Matt Russell)
</p>
</div>
</div>
<div id="secondary-structure" class="section level3" number="2.4.2">
<h3><span class="header-section-number">2.4.2</span> Secondary Structure</h3>
<p>The local folding of the polypeptide in some regions gives rise to the
secondary structure of the protein. The most common are the <em>α</em>-helix
and <em>β</em>-pleated sheet structures (Figure 3.4.7). Both structures are
the <em>α</em>-helix structure—the helix held in shape by hydrogen bonds.
The hydrogen bonds form between the oxygen atom in the carbonyl group
in one amino acid and another amino acid that is four amino acids
farther along the chain.</p>
<div class="figure"><span id="fig:unnamed-chunk-13"></span>
<img src="images/protimage7.jpeg" alt="The α-helix and β-pleated sheet are secondary structures of proteins that form because of hydrogen bonding between carbonyl and amino groups in the peptide backbone. Certain amino acids have a propensity to form an α-helix, while others have a propensity to form a β-pleated sheet." width="40%" />
<p class="caption">
Figure 2.7: The α-helix and β-pleated sheet are secondary structures of proteins that form because of hydrogen bonding between carbonyl and amino groups in the peptide backbone. Certain amino acids have a propensity to form an α-helix, while others have a propensity to form a β-pleated sheet.
</p>
</div>
<p>Every helical turn in an alpha helix has 3.6 amino acid residues. The
R groups (the variant groups) of the polypeptide protrude out from the
<em>α</em>-helix chain. In the <em>β</em>-pleated sheet, the “pleats” are formed by
hydrogen bonding between atoms on the backbone of the polypeptide
chain. The R groups are attached to the carbons and extend above and
below the folds of the pleat. The pleated segments align parallel or
antiparallel to each other, and hydrogen bonds form between the
partially positive nitrogen atom in the amino group and the partially
negative oxygen atom in the carbonyl group of the peptide backbone.
The <em>α</em>-helix and <em>β</em>-pleated sheet structures are found in most
globular and fibrous proteins and they play an important structural
role.</p>
</div>
<div id="tertiary-structure" class="section level3" number="2.4.3">
<h3><span class="header-section-number">2.4.3</span> Tertiary Structure</h3>
<p>The unique three-dimensional structure of a polypeptide is its
tertiary structure (Figure 3.4.8). This structure is in part due to
chemical interactions at work on the polypeptide chain. Primarily, the
interactions among R groups creates the complex three- dimensional
tertiary structure of a protein. The nature of the R groups found in
the amino acids involved can counteract the formation of the hydrogen
bonds described for standard secondary structures. For example, R
groups with like charges are repelled by each other and those with
unlike charges are attracted to each other (ionic bonds). When protein
folding takes place, the hydrophobic R groups of nonpolar amino acids
lay in the interior of the protein, whereas the hydrophilic R groups
lay on the outside. The former types of interactions are also known as
hydrophobic interactions. Interaction between cysteine side chains
forms disulfide linkages in the presence of oxygen, the only covalent
bond forming during protein folding.</p>
<div class="figure"><span id="fig:unnamed-chunk-14"></span>
<img src="images/protimage8.jpeg" alt="The tertiary structure of proteins is determined by a variety of chemical interactions. These include hydrophobic interactions, ionic bonding, hydrogen bonding and disulfide linkages." width="40%" />
<p class="caption">
Figure 2.8: The tertiary structure of proteins is determined by a variety of chemical interactions. These include hydrophobic interactions, ionic bonding, hydrogen bonding and disulfide linkages.
</p>
</div>
<p>All of these interactions, weak and strong, determine the final
three-dimensional shape of the protein. When a protein loses its
three-dimensional shape, it may no longer be functional.</p>
</div>
<div id="quaternary-structure" class="section level3" number="2.4.4">
<h3><span class="header-section-number">2.4.4</span> Quaternary Structure</h3>
<p>In nature, some proteins are formed from several polypeptides, also
known as subunits, and the interaction of these subunits forms the
quaternary structure. Weak interactions between the subunits help to
stabilize the overall structure. For example, insulin (a globular
protein) has a combination of hydrogen bonds and disulfide bonds that
cause it to be mostly clumped into a ball shape. Insulin starts out as
a single polypeptide and loses some internal sequences in the presence
of post-translational modification after the formation of the
disulfide linkages that hold the remaining chains together. Silk (a
fibrous protein), however, has a <em>β</em>-pleated sheet structure that is
the result of hydrogen bonding between different chains.</p>
<p>The four levels of protein structure (primary, secondary, tertiary,
and quaternary) are illustrated in Figure 3.4.9.</p>
<div class="figure"><span id="fig:unnamed-chunk-15"></span>
<img src="images/protimage9.jpeg" alt="The four levels of protein structure can be observed in these illustrations. (credit: modification of work by National Human Genome Research Institute)" width="398" />
<p class="caption">
Figure 2.9: The four levels of protein structure can be observed in these illustrations. (credit: modification of work by National Human Genome Research Institute)
</p>
</div>
</div>
</div>
<div id="denaturation-and-protein-folding" class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span> Denaturation and Protein Folding</h2>
<p>Each protein has its own unique sequence and shape that are held
together by chemical interactions. If the protein is subject to
changes in temperature, pH, or exposure to chemicals, the protein
structure may change, losing its shape without losing its primary
sequence in what is known as denaturation. Denaturation is often
reversible because the primary structure of the polypeptide is
conserved in the process if the denaturing agent is removed, allowing
the protein to resume its function. Sometimes denaturation is
irreversible, leading to loss of function. One example of irreversible
protein denaturation is when an egg is fried. The albumin protein in
the liquid egg white is denatured when placed in a hot pan. Not all
proteins are denatured at high temperatures; for instance, bacteria
that survive in hot springs have proteins that function at
temperatures close to boiling. The stomach is also very acidic, has a
low pH, and denatures proteins as part of the digestion process;
however, the digestive enzymes of the stomach retain their activity
under these conditions.</p>
<p>Protein folding is critical to its function. It was originally thought
that the proteins themselves were responsible for the folding process.
Only recently was it found that often they receive assistance in the
folding process from protein helpers known as chaperones (or
chaperonins) that associate with the target protein during the folding
process. They act by preventing aggregation of polypeptides that make
up the complete protein structure, and they disassociate from the
protein once the target protein is folded.</p>
</div>
<div id="summary-1" class="section level2" number="2.6">
<h2><span class="header-section-number">2.6</span> Summary</h2>
<p>Proteins are a class of macromolecules that perform a diverse range of
functions for the cell. They help in metabolism by providing
structural support and by acting as enzymes, carriers, or hormones.
The building blocks of proteins (monomers) are amino acids. Each amino
acid has a central carbon that is linked to an amino group, a carboxyl
group, a hydrogen atom, and an</p>
<p>R group or side chain. There are 20 commonly occurring amino acids,
each of which differs in the R group. Each amino acid is linked to its
neighbors by a peptide bond. A long chain of amino acids is known as a
polypeptide.</p>
<p>Proteins are organized at four levels: primary, secondary, tertiary,
and (optional) quaternary. The primary structure is the unique
sequence of amino acids. The local folding of the polypeptide to form
structures such as the <em>α</em> helix and <em>β</em>-pleated sheet constitutes the
secondary structure. The overall three-dimensional structure is the
tertiary structure. When two or more polypeptides combine to form the
complete protein structure, the configuration is known as the
quaternary structure of a protein. Protein shape and function are
intricately linked; any change in shape caused by changes in
temperature or pH may lead to protein denaturation and a loss in
function.</p>
</div>
<div id="art-connections" class="section level2" number="2.7">
<h2><span class="header-section-number">2.7</span> Art Connections</h2>
<ul>
<li><p>Which categories of amino acid would you expect to find on
the surface of a soluble protein, and which would you expect to find
in the interior? What distribution of amino acids would you expect to
find in a protein embedded in a lipid bilayer?</p></li>
<li><p>Polar and charged amino acid residues (the remainder after
peptide bond formation) are more likely to be found on the surface of
soluble proteins where they can interact with water, and nonpolar
(e.g., amino acid side chains) are more likely to be found in the
interior where they are sequestered from water. In membrane proteins,
nonpolar and hydrophobic amino acid side chains associate with the
hydrophobic tails of phospholipids, while polar and charged amino acid
side chains interact with the polar head groups or with the aqueous
solution. However, there are exceptions. Sometimes, positively and
negatively charged amino acid side chains interact with one another in
the interior of a protein, and polar or charged amino acid side chains
that interact with a ligand can be found in the ligand binding pocket.</p></li>
</ul>
</div>
<div id="review-questions" class="section level2" number="2.8">
<h2><span class="header-section-number">2.8</span> Review Questions</h2>
<p>The monomers that make up proteins are called
1. nucleotides
1. disaccharides
1. amino acids
1. chaperones</p>
<p>Answer: C</p>
<p>The <em>α</em> helix and the <em>β</em>-pleated sheet are part of which protein
structure?</p>
<ol style="list-style-type: decimal">
<li>primary</li>
<li>secondary</li>
<li>tertiary</li>
<li>quaternary</li>
</ol>
<p>Answer: B</p>
</div>
<div id="free-response" class="section level2" number="2.9">
<h2><span class="header-section-number">2.9</span> Free Response</h2>
<p><strong>Explain what happens if even one amino acid is substituted for another
in a polypeptide chain. Provide a specific example.</strong></p>
<p>A change in gene sequence can lead to a different amino acid being
added to a polypeptide chain instead of the normal one. This causes a
change in protein structure and function. For example, in sickle cell
anemia, the hemoglobin <em>β</em> chain has a single amino acid
substitution—the amino acid glutamic acid in position six is
substituted by valine. Because of this change, hemoglobin molecules
form aggregates, and the disc-shaped red blood cells assume a crescent
shape, which results in serious health problems.</p>
<p><strong>Describe the differences in the four protein structures.</strong></p>
<p>The sequence and number of amino acids in a polypeptide chain is its
primary structure. The local folding of the polypeptide in some
regions is the secondary structure of the protein. The
three-dimensional structure of a polypeptide is known as its tertiary
structure, created in part by chemical interactions such as hydrogen
bonds between polar side chains, van der Waals interactions, disulfide
linkages, and hydrophobic interactions. Some proteins are formed from
multiple polypeptides, also known as subunits, and the interaction of
these subunits forms the quaternary structure.</p>
</div>
<div id="glossary-1" class="section level2" number="2.10">
<h2><span class="header-section-number">2.10</span> Glossary</h2>
<p><strong>alpha-helix structure (<em>α</em>-helix)</strong>: type of secondary structure of proteins formed by folding of the polypeptide into a helix shape with hydrogen bonds stabilizing the
structure<br />
<strong>amino acid</strong>: monomer of a protein; has a central carbon or alpha carbon to which an
amino group, a carboxyl group, a hydrogen, and an R group or side
chain is attached; the R group is different for all 20 amino acids<br />
<strong>beta-pleated sheet (<em>β</em>-pleated)</strong>: secondary structure found in proteins in which “pleats” are formed by hydrogen bonding between atoms on the backbone of the polypeptide chain<br />
<strong>chaperone</strong>: (also, chaperonin) protein that helps nascent protein in the folding process <strong>denaturation</strong>: loss of shape in a protein as a result of changes in temperature, pH, or exposure to chemicals<br />
<strong>enzyme</strong>: catalyst in a biochemical reaction that is usually a complex or conjugated protein<br />
<strong>hormone</strong>: chemical signaling molecule, usually protein or steroid, secreted by endocrine cells that act to control or regulate specific physiological processes
<strong>peptide bond</strong>: bond formed between two amino acids by a dehydration reaction<br />
<strong>polypeptide</strong>: long chain of amino acids linked by peptide bonds<br />
<strong>primary structure</strong>: linear sequence of amino acids in a protein<br />
<strong>protein</strong>: biological macromolecule composed of one or more chains of amino acids<br />
<strong>quaternary structure</strong>: association of discrete polypeptide subunits in a protein<br />
<strong>secondary structure</strong>: regular structure formed by proteins by intramolecular hydrogen bonding between the oxygen atom of one amino acid residue and the hydrogen attached to the nitrogen atom of another amino acid residue
<strong>tertiary structure</strong>: three-dimensional conformation of a protein, including interactions between secondary structural elements; formed from interactions between amino acid side chains</p>
</div>
<div id="contributors-and-attributions-1" class="section level2" number="2.11">
<h2><span class="header-section-number">2.11</span> Contributors and Attributions</h2>
<p>Connie Rye (East Mississippi Community College), Robert Wise
(University of Wisconsin, Oshkosh), Vladimir Jurukovski (Suffolk
County Community College), Jean DeSaix (University of North Carolina
at Chapel Hill), Jung Choi (Georgia Institute of Technology), Yael
Avissar (Rhode Island College) among other contributing authors.
Original content by OpenStax (CC BY 4.0; Download for free at
<a href="http://cnx.org/contents/185cbf87-c72e-48f5-b51e-f14f21b5eabd%409.87">http://cnx.org/contents/185cbf87-c72...f21b5eabd@9.87</a>).</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="nucleic-acids-review.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="downloadR.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/000-biology_review/00h-proteins-LibreText.rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/000-biology_review/00h-proteins-LibreText.rmd",
"text": null
},
"download": ["lbrb.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
