<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 16 Sequence dotplots in R | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 16 Sequence dotplots in R | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 16 Sequence dotplots in R | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="downloading-protein-sequences-in-r.html"/>
<link rel="next" href="global-proteins-aligments-in-r.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>1</b> How to review this book</a>
<ul>
<li class="chapter" data-level="1.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>1.1</b> File format</a></li>
<li class="chapter" data-level="1.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>1.2</b> What can you do?</a></li>
<li class="chapter" data-level="1.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>1.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>1.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="1.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>1.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="1.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>1.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="1.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>1.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>2</b> Downloading R</a>
<ul>
<li class="chapter" data-level="2.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>2.1</b> Preface</a></li>
<li class="chapter" data-level="2.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>2.2</b> Introduction to R</a></li>
<li class="chapter" data-level="2.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>2.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>2.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="2.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>2.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>2.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>3</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="3.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>3.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="3.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>3.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>4</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="4.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>4.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="4.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>4.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="4.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>4.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>5</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="5.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>5.1</b> Bioconductor</a></li>
<li class="chapter" data-level="5.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>5.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="5.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>5.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>5.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="5.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>5.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="5.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>5.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>5.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>6</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="6.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>6.1</b> Vocabulary</a></li>
<li class="chapter" data-level="6.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>6.2</b> R functions</a></li>
<li class="chapter" data-level="6.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>6.3</b> Interacting with R</a></li>
<li class="chapter" data-level="6.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>6.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>6.4.1</b> Vectors</a></li>
<li class="chapter" data-level="6.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>6.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="6.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>6.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="6.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>6.4.4</b> Lists</a></li>
<li class="chapter" data-level="6.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>6.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>6.5</b> Arguments</a></li>
<li class="chapter" data-level="6.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>6.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="6.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>6.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="6.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>6.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="6.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>6.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="6.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>6.9</b> Quiting R</a></li>
<li class="chapter" data-level="6.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>6.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>7</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="7.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-1"><i class="fa fa-check"></i><b>7.1</b> Preface</a></li>
<li class="chapter" data-level="7.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>7.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="7.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>7.3</b> Introduction</a></li>
<li class="chapter" data-level="7.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>7.4</b> Vocabulary</a></li>
<li class="chapter" data-level="7.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions"><i class="fa fa-check"></i><b>7.5</b> Functions</a></li>
<li class="chapter" data-level="7.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>7.6</b> Preliminaries</a></li>
<li class="chapter" data-level="7.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>7.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="7.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>7.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="7.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>7.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="7.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>7.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="7.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>7.8.3</b> DNA words</a></li>
<li class="chapter" data-level="7.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>7.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>7.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="7.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>7.9.1</b> License</a></li>
<li class="chapter" data-level="7.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>7.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>8</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="8.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-2"><i class="fa fa-check"></i><b>8.1</b> Preface</a></li>
<li class="chapter" data-level="8.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>8.2</b> Vocab</a></li>
<li class="chapter" data-level="8.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-1"><i class="fa fa-check"></i><b>8.3</b> Functions</a></li>
<li class="chapter" data-level="8.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>8.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="8.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>8.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>9</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="9.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-3"><i class="fa fa-check"></i><b>9.1</b> Preface</a></li>
<li class="chapter" data-level="9.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>9.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="functions-2.html"><a href="functions-2.html"><i class="fa fa-check"></i><b>10</b> Functions</a>
<ul>
<li class="chapter" data-level="10.1" data-path="functions-2.html"><a href="functions-2.html#vectors-in-r"><i class="fa fa-check"></i><b>10.1</b> Vectors in R</a></li>
<li class="chapter" data-level="10.2" data-path="functions-2.html"><a href="functions-2.html#math-on-vectors"><i class="fa fa-check"></i><b>10.2</b> Math on vectors</a></li>
<li class="chapter" data-level="10.3" data-path="functions-2.html"><a href="functions-2.html#functions-on-vectors"><i class="fa fa-check"></i><b>10.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="10.4" data-path="functions-2.html"><a href="functions-2.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>10.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="10.5" data-path="functions-2.html"><a href="functions-2.html#subsetting-vectors"><i class="fa fa-check"></i><b>10.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="10.6" data-path="functions-2.html"><a href="functions-2.html#sequences-of-numbers"><i class="fa fa-check"></i><b>10.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="10.7" data-path="functions-2.html"><a href="functions-2.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>10.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="10.8" data-path="functions-2.html"><a href="functions-2.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>10.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>11</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-4"><i class="fa fa-check"></i><b>11.1</b> Preface</a></li>
<li class="chapter" data-level="11.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>11.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="11.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>11.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>12</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="12.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-5"><i class="fa fa-check"></i><b>12.1</b> Preface</a></li>
<li class="chapter" data-level="12.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>12.2</b> Vocab</a></li>
<li class="chapter" data-level="12.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-3"><i class="fa fa-check"></i><b>12.3</b> Functions</a></li>
<li class="chapter" data-level="12.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>12.4</b> Functions in R</a></li>
<li class="chapter" data-level="12.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>12.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>13</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="13.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-4"><i class="fa fa-check"></i><b>13.0.1</b> Functions</a></li>
<li class="chapter" data-level="13.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>13.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="13.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>13.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="13.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>13.0.4</b> File types</a></li>
<li class="chapter" data-level="13.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>13.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="13.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>13.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="13.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>13.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="13.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>13.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="13.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>13.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="13.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>13.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="13.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>13.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="13.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>13.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="13.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>13.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>13.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="13.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>13.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>14</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="14.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-2"><i class="fa fa-check"></i><b>14.1</b> Preliminaries</a></li>
<li class="chapter" data-level="14.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>14.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>14.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="14.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>14.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>15</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="15.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>15.1</b> Preliminaries</a></li>
<li class="chapter" data-level="15.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>15.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>16</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="16.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>16.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>16.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>16.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="16.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>16.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="16.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>16.4</b> Inversions</a></li>
<li class="chapter" data-level="16.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>16.5</b> Translocations</a></li>
<li class="chapter" data-level="16.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>16.6</b> Random sequence</a></li>
<li class="chapter" data-level="16.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>16.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>17</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="17.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>17.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="17.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>17.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="17.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>17.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="17.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>17.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="17.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>17.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="17.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>17.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>17.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="17.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>17.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>18</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="18.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>18.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>18.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>18.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>19</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="19.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>19.1</b> Prelminaries</a></li>
<li class="chapter" data-level="19.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>19.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="19.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>19.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="19.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>19.4</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>19.4.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="19.4.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>19.4.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="19.4.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>19.4.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="19.4.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>19.4.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-2"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="19.5" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>19.5</b> Help!</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>19.5.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="19.5.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>19.5.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="19.5.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>19.5.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="19.5.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>19.5.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>19.6</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="19.6.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>19.6.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="19.6.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>19.6.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="19.7" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>19.7</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="sequence-dotplots-in-r" class="section level1" number="16">
<h1><span class="header-section-number">Chapter 16</span> Sequence dotplots in <em>R</em></h1>
<p><strong>By</strong>: Avril Coghlan.</p>
<p><strong>Adapted, edited and expanded</strong>: Nathan Brouwer under the Creative Commons 3.0 Attribution License <a href="https://creativecommons.org/licenses/by/3.0/">(CC BY 3.0)</a>.</p>
<p><strong>NOTE: I’ve added some new material that is rather terse and lacks explication.</strong></p>
<!-- stronger motivation / explanation -->
<!-- good sources -->
<!-- https://omicstutorials.com/interpreting-dot-plot-bioinformatics-with-an-example/ -->
<!-- http://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/650/Examples_interpretations_dot_plots.html -->
<div id="preliminaries-4" class="section level2" number="16.1">
<h2><span class="header-section-number">16.1</span> Preliminaries</h2>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="sequence-dotplots-in-r.html#cb256-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compbio4all)</span></code></pre></div>
<div id="download-sequences" class="section level3" number="16.1.1">
<h3><span class="header-section-number">16.1.1</span> Download sequences</h3>
<p>As we did in the previous lesson on dotplots, we’ll look at two sequences.</p>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="sequence-dotplots-in-r.html#cb257-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sequence 1: Q9CD83</span></span>
<span id="cb257-2"><a href="sequence-dotplots-in-r.html#cb257-2" aria-hidden="true" tabindex="-1"></a>leprae_fasta <span class="ot">&lt;-</span> rentrez<span class="sc">::</span><span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>,</span>
<span id="cb257-3"><a href="sequence-dotplots-in-r.html#cb257-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">id =</span> <span class="st">&quot;Q9CD83&quot;</span>,</span>
<span id="cb257-4"><a href="sequence-dotplots-in-r.html#cb257-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb257-5"><a href="sequence-dotplots-in-r.html#cb257-5" aria-hidden="true" tabindex="-1"></a><span class="co"># sequence 2: OIN17619.1</span></span>
<span id="cb257-6"><a href="sequence-dotplots-in-r.html#cb257-6" aria-hidden="true" tabindex="-1"></a>ulcerans_fasta <span class="ot">&lt;-</span> rentrez<span class="sc">::</span><span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>,</span>
<span id="cb257-7"><a href="sequence-dotplots-in-r.html#cb257-7" aria-hidden="true" tabindex="-1"></a>                         <span class="at">id =</span> <span class="st">&quot;OIN17619.1&quot;</span>,</span>
<span id="cb257-8"><a href="sequence-dotplots-in-r.html#cb257-8" aria-hidden="true" tabindex="-1"></a>                         <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb257-9"><a href="sequence-dotplots-in-r.html#cb257-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-10"><a href="sequence-dotplots-in-r.html#cb257-10" aria-hidden="true" tabindex="-1"></a>leprae_vector   <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(leprae_fasta)</span>
<span id="cb257-11"><a href="sequence-dotplots-in-r.html#cb257-11" aria-hidden="true" tabindex="-1"></a>ulcerans_vector <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(ulcerans_fasta)</span></code></pre></div>
</div>
</div>
<div id="visualzing-two-identical-sequences" class="section level2" number="16.2">
<h2><span class="header-section-number">16.2</span> Visualzing two identical sequences</h2>
<p>To help build our intuition about dotplots we’ll first look at some artificial examples. First, we’ll see what happens when we make a dotplot comparing the alphabet versus itself. The build-in <code>LETTERS</code> object in R contains the alphabet from A to Z. This is a sequence with no repeats.</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="sequence-dotplots-in-r.html#cb258-1" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">dotPlot</span>(LETTERS, </span>
<span id="cb258-2"><a href="sequence-dotplots-in-r.html#cb258-2" aria-hidden="true" tabindex="-1"></a>                LETTERS)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-166-1.png" width="672" /></p>
<p>What we get is a perfect diagonal line.</p>
</div>
<div id="visualizing-repeats" class="section level2" number="16.3">
<h2><span class="header-section-number">16.3</span> Visualizing repeats</h2>
<p>Now lets’ make a sequence where the alphabet gets repeats twice</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="sequence-dotplots-in-r.html#cb259-1" aria-hidden="true" tabindex="-1"></a>LETTERS.<span class="fl">2.</span>times <span class="ot">&lt;-</span> <span class="fu">c</span>(LETTERS,LETTERS)</span>
<span id="cb259-2"><a href="sequence-dotplots-in-r.html#cb259-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb259-3"><a href="sequence-dotplots-in-r.html#cb259-3" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">dotPlot</span>(LETTERS.<span class="fl">2.</span>times, </span>
<span id="cb259-4"><a href="sequence-dotplots-in-r.html#cb259-4" aria-hidden="true" tabindex="-1"></a>                LETTERS.<span class="fl">2.</span>times)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-167-1.png" width="672" /></p>
<p>Note the diagonal lines.</p>
<p>Now 3 repeats</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="sequence-dotplots-in-r.html#cb260-1" aria-hidden="true" tabindex="-1"></a>LETTERS.<span class="fl">3.</span>times <span class="ot">&lt;-</span> <span class="fu">c</span>(LETTERS,LETTERS,LETTERS)</span>
<span id="cb260-2"><a href="sequence-dotplots-in-r.html#cb260-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-3"><a href="sequence-dotplots-in-r.html#cb260-3" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">dotPlot</span>(LETTERS.<span class="fl">3.</span>times, </span>
<span id="cb260-4"><a href="sequence-dotplots-in-r.html#cb260-4" aria-hidden="true" tabindex="-1"></a>                LETTERS.<span class="fl">3.</span>times)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-168-1.png" width="672" /></p>
<p>Here’s another example of repeats.</p>
<p>Create sequence with repeats:</p>
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="sequence-dotplots-in-r.html#cb261-1" aria-hidden="true" tabindex="-1"></a>seq.repeat <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;C&quot;</span>,<span class="st">&quot;D&quot;</span>,<span class="st">&quot;E&quot;</span>,<span class="st">&quot;F&quot;</span>,<span class="st">&quot;G&quot;</span>,<span class="st">&quot;H&quot;</span>,<span class="st">&quot;I&quot;</span>)</span>
<span id="cb261-2"><a href="sequence-dotplots-in-r.html#cb261-2" aria-hidden="true" tabindex="-1"></a>seq1 <span class="ot">&lt;-</span> <span class="fu">rep</span>(seq.repeat,<span class="dv">3</span>)</span></code></pre></div>
<p>Make the dotplot:</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="sequence-dotplots-in-r.html#cb262-1" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">dotPlot</span>(seq1, </span>
<span id="cb262-2"><a href="sequence-dotplots-in-r.html#cb262-2" aria-hidden="true" tabindex="-1"></a>                seq1)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-170-1.png" width="672" /></p>
</div>
<div id="inversions" class="section level2" number="16.4">
<h2><span class="header-section-number">16.4</span> Inversions</h2>
<p>See if you can figure out what’s going on here.</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="sequence-dotplots-in-r.html#cb263-1" aria-hidden="true" tabindex="-1"></a>LETTERS.<span class="fl">3.</span>times.with.invert <span class="ot">&lt;-</span> <span class="fu">c</span>(LETTERS,<span class="fu">rev</span>(LETTERS),LETTERS)</span>
<span id="cb263-2"><a href="sequence-dotplots-in-r.html#cb263-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb263-3"><a href="sequence-dotplots-in-r.html#cb263-3" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">dotPlot</span>(LETTERS.<span class="fl">3.</span>times, </span>
<span id="cb263-4"><a href="sequence-dotplots-in-r.html#cb263-4" aria-hidden="true" tabindex="-1"></a>                LETTERS.<span class="fl">3.</span>times.with.invert)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-171-1.png" width="672" /></p>
</div>
<div id="translocations" class="section level2" number="16.5">
<h2><span class="header-section-number">16.5</span> Translocations</h2>
<p>See if you can figure out what’s going on here.</p>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb264-1"><a href="sequence-dotplots-in-r.html#cb264-1" aria-hidden="true" tabindex="-1"></a>seg1 <span class="ot">&lt;-</span> LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span>
<span id="cb264-2"><a href="sequence-dotplots-in-r.html#cb264-2" aria-hidden="true" tabindex="-1"></a>seg2 <span class="ot">&lt;-</span> LETTERS[<span class="dv">9</span><span class="sc">:</span><span class="dv">18</span>]</span>
<span id="cb264-3"><a href="sequence-dotplots-in-r.html#cb264-3" aria-hidden="true" tabindex="-1"></a>seg3 <span class="ot">&lt;-</span> LETTERS[<span class="dv">18</span><span class="sc">:</span><span class="dv">26</span>]</span>
<span id="cb264-4"><a href="sequence-dotplots-in-r.html#cb264-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb264-5"><a href="sequence-dotplots-in-r.html#cb264-5" aria-hidden="true" tabindex="-1"></a>LETTERS.with.transloc <span class="ot">&lt;-</span> <span class="fu">c</span>(seg1,seg3,seg2)</span>
<span id="cb264-6"><a href="sequence-dotplots-in-r.html#cb264-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb264-7"><a href="sequence-dotplots-in-r.html#cb264-7" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">dotPlot</span>(LETTERS, </span>
<span id="cb264-8"><a href="sequence-dotplots-in-r.html#cb264-8" aria-hidden="true" tabindex="-1"></a>                LETTERS.with.transloc)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-172-1.png" width="672" /></p>
</div>
<div id="random-sequence" class="section level2" number="16.6">
<h2><span class="header-section-number">16.6</span> Random sequence</h2>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="sequence-dotplots-in-r.html#cb265-1" aria-hidden="true" tabindex="-1"></a>letters.rand1 <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x =</span> LETTERS, <span class="at">size =</span> <span class="dv">26</span>, <span class="at">replace =</span> F)</span>
<span id="cb265-2"><a href="sequence-dotplots-in-r.html#cb265-2" aria-hidden="true" tabindex="-1"></a>letters.rand2 <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x =</span> LETTERS, <span class="at">size =</span> <span class="dv">26</span>, <span class="at">replace =</span> F)</span>
<span id="cb265-3"><a href="sequence-dotplots-in-r.html#cb265-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb265-4"><a href="sequence-dotplots-in-r.html#cb265-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb265-5"><a href="sequence-dotplots-in-r.html#cb265-5" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">dotPlot</span>(letters.rand1, </span>
<span id="cb265-6"><a href="sequence-dotplots-in-r.html#cb265-6" aria-hidden="true" tabindex="-1"></a>                letters.rand2)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-173-1.png" width="672" /></p>
</div>
<div id="comparing-two-real-sequences-using-a-dotplot" class="section level2" number="16.7">
<h2><span class="header-section-number">16.7</span> Comparing two real sequences using a dotplot</h2>
<p>As a first step in comparing two protein, RNA or DNA sequences, it is a good idea to make a <strong>dotplot</strong>. A <strong>dotplot</strong> is a graphical method that allows the comparison of two protein or DNA sequences and identify regions of close similarity between them. A dotplot is essentially a two-dimensional matrix (like a grid), which has the sequences of the proteins being compared along the vertical and horizontal axes.</p>
<p>In order to make a simple dotplot to represent of the similarity between two sequences, individual cells in the matrix can be shaded black if residues are identical, so that matching sequence segments appear as runs of diagonal lines across the matrix. Identical proteins will have a line exactly on the main diagonal of the dotplot, that spans across the whole matrix.</p>
<p>For proteins that are not identical, but share regions of similarity, the dotplot will have shorter lines that may be on the <strong>main diagonal</strong>, or off the main diagonal of the matrix. In essence, a dotplot will reveal if there are any regions that are clearly very similar in two protein (or DNA) sequences.</p>
<p>We can create a dotplot for two sequences using the <code>dotPlot()</code> function in the <code>seqinr</code> package.</p>
<p>First, let’s look at a dotplot created using only a single sequence. You’d never do this in practice, but it will give you a sense of what dotplots are doing.</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="sequence-dotplots-in-r.html#cb266-1" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">dotPlot</span>(leprae_vector, </span>
<span id="cb266-2"><a href="sequence-dotplots-in-r.html#cb266-2" aria-hidden="true" tabindex="-1"></a>                leprae_vector)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-174-1.png" width="672" /></p>
<p>These two sequences are identical, so we have a very distinct diagonal line. But there’s also other</p>
<p>Now we’ll make a real dotplot of the <a href="https://en.wikipedia.org/wiki/Chorismate_lyase">chorismate lyase</a> proteins from two closely related species, <em>Mycobacterium leprae</em> and <em>Mycobacterium ulcerans</em>.</p>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="sequence-dotplots-in-r.html#cb267-1" aria-hidden="true" tabindex="-1"></a>seqinr<span class="sc">::</span><span class="fu">dotPlot</span>(leprae_vector, </span>
<span id="cb267-2"><a href="sequence-dotplots-in-r.html#cb267-2" aria-hidden="true" tabindex="-1"></a>                ulcerans_vector)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-175-1.png" width="672" /></p>
<p>In the dotplot above, the <em>M. leprae</em> sequence is plotted along the x-axis (horizontal axis), and the <em>M. ulcerans</em> sequence is plotted along the y-axis (vertical axis). The dotplot displays a dot at points where there is an identical amino acid in the two sequences.</p>
<p>For example, if amino acid 53 in the <em>M. leprae</em> sequence is the same amino acid (eg. “W”) as amino acid 70 in the M. ulcerans sequence, then the dotplot will show a dot the position in the plot where x =50 and y =53.</p>
<p>In this case you can see a lot of dots along a diagonal line, which indicates that the two protein sequences contain many identical amino acids at the same (or very similar) positions along their lengths. This is what you would expect, because we know that these two proteins are <strong>homologs</strong> (related proteins) because they share a close evolutionary history.</p>
<!-- TO DO:  Add the same gene from more distance relatives -->

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="downloading-protein-sequences-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="global-proteins-aligments-in-r.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/15-dotplots.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/15-dotplots.Rmd",
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
