<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 54 UPGMA the hard way | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 54 UPGMA the hard way | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 54 UPGMA the hard way | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"/>
<link rel="next" href="representing-phylogenetic-trees-in-newick-format.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.5.1/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>1</b> How to review this book</a>
<ul>
<li class="chapter" data-level="1.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>1.1</b> File format</a></li>
<li class="chapter" data-level="1.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>1.2</b> What can you do?</a></li>
<li class="chapter" data-level="1.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>1.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>1.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="1.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>1.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="1.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>1.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="1.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>1.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="test-audio-link.html"><a href="test-audio-link.html"><i class="fa fa-check"></i><b>2</b> Test audio link</a></li>
<li class="chapter" data-level="3" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>3</b> Test shiny app</a></li>
<li class="chapter" data-level="4" data-path="test-youtube-vidoe.html"><a href="test-youtube-vidoe.html"><i class="fa fa-check"></i><b>4</b> Test YouTube vidoe</a></li>
<li class="chapter" data-level="5" data-path="test-google-doc-insertion.html"><a href="test-google-doc-insertion.html"><i class="fa fa-check"></i><b>5</b> Test google doc insertion</a></li>
<li class="chapter" data-level="6" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>6</b> Datacamp test</a></li>
<li class="chapter" data-level="7" data-path="an-example-of-a-plotly-graph.html"><a href="an-example-of-a-plotly-graph.html"><i class="fa fa-check"></i><b>7</b> An example of a plotly graph</a></li>
<li class="chapter" data-level="8" data-path="directly-embedding-of-a-libretext-page.html"><a href="directly-embedding-of-a-libretext-page.html"><i class="fa fa-check"></i><b>8</b> Directly embedding of a libretext page</a></li>
<li class="chapter" data-level="9" data-path="libretext-image.html"><a href="libretext-image.html"><i class="fa fa-check"></i><b>9</b> Libretext image</a>
<ul>
<li class="chapter" data-level="9.1" data-path="libretext-image.html"><a href="libretext-image.html#prelude-to-population-and-community-ecology"><i class="fa fa-check"></i><b>9.1</b> 45.0: Prelude to Population and Community Ecology</a></li>
<li class="chapter" data-level="9.2" data-path="libretext-image.html"><a href="libretext-image.html#note"><i class="fa fa-check"></i><b>9.2</b> Note</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="nucleic-acids.html"><a href="nucleic-acids.html"><i class="fa fa-check"></i><b>10</b> Nucleic Acids</a>
<ul>
<li class="chapter" data-level="10.1" data-path="nucleic-acids.html"><a href="nucleic-acids.html#dna-and-rna"><i class="fa fa-check"></i><b>10.1</b> DNA and RNA</a></li>
<li class="chapter" data-level="10.2" data-path="nucleic-acids.html"><a href="nucleic-acids.html#dna-double-helix-structure"><i class="fa fa-check"></i><b>10.2</b> DNA Double-Helix Structure</a></li>
<li class="chapter" data-level="10.3" data-path="nucleic-acids.html"><a href="nucleic-acids.html#rna"><i class="fa fa-check"></i><b>10.3</b> RNA</a></li>
<li class="chapter" data-level="10.4" data-path="nucleic-acids.html"><a href="nucleic-acids.html#summary"><i class="fa fa-check"></i><b>10.4</b> Summary</a></li>
<li class="chapter" data-level="10.5" data-path="nucleic-acids.html"><a href="nucleic-acids.html#analysis-questions"><i class="fa fa-check"></i><b>10.5</b> Analysis Questions</a></li>
<li class="chapter" data-level="10.6" data-path="nucleic-acids.html"><a href="nucleic-acids.html#glossary"><i class="fa fa-check"></i><b>10.6</b> Glossary</a></li>
<li class="chapter" data-level="10.7" data-path="nucleic-acids.html"><a href="nucleic-acids.html#contributors-and-attributions"><i class="fa fa-check"></i><b>10.7</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>11</b> Downloading R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="downloading-r.html"><a href="downloading-r.html#preface"><i class="fa fa-check"></i><b>11.1</b> Preface</a></li>
<li class="chapter" data-level="11.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>11.2</b> Introduction to R</a></li>
<li class="chapter" data-level="11.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>11.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>11.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="11.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>11.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>11.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>12</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="12.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>12.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="12.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>12.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>13</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="13.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>13.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="13.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>13.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="13.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>13.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>14</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="14.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>14.1</b> Bioconductor</a></li>
<li class="chapter" data-level="14.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>14.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="14.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>14.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>14.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="14.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>14.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="14.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>14.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>14.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>15</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="15.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary"><i class="fa fa-check"></i><b>15.1</b> Vocabulary</a></li>
<li class="chapter" data-level="15.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>15.2</b> R functions</a></li>
<li class="chapter" data-level="15.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>15.3</b> Interacting with R</a></li>
<li class="chapter" data-level="15.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>15.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="15.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>15.4.1</b> Vectors</a></li>
<li class="chapter" data-level="15.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>15.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="15.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>15.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="15.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>15.4.4</b> Lists</a></li>
<li class="chapter" data-level="15.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>15.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="15.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>15.5</b> Arguments</a></li>
<li class="chapter" data-level="15.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>15.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="15.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>15.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="15.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>15.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="15.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>15.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="15.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>15.9</b> Quiting R</a></li>
<li class="chapter" data-level="15.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>15.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>16</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="16.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-1"><i class="fa fa-check"></i><b>16.1</b> Preface</a></li>
<li class="chapter" data-level="16.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>16.2</b> Vocab</a></li>
<li class="chapter" data-level="16.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions"><i class="fa fa-check"></i><b>16.3</b> Functions</a></li>
<li class="chapter" data-level="16.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>16.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="16.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>16.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>17</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="17.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-2"><i class="fa fa-check"></i><b>17.1</b> Preface</a></li>
<li class="chapter" data-level="17.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>17.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="functions-1.html"><a href="functions-1.html"><i class="fa fa-check"></i><b>18</b> Functions</a>
<ul>
<li class="chapter" data-level="18.1" data-path="functions-1.html"><a href="functions-1.html#vectors-in-r"><i class="fa fa-check"></i><b>18.1</b> Vectors in R</a></li>
<li class="chapter" data-level="18.2" data-path="functions-1.html"><a href="functions-1.html#math-on-vectors"><i class="fa fa-check"></i><b>18.2</b> Math on vectors</a></li>
<li class="chapter" data-level="18.3" data-path="functions-1.html"><a href="functions-1.html#functions-on-vectors"><i class="fa fa-check"></i><b>18.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="18.4" data-path="functions-1.html"><a href="functions-1.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>18.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="18.5" data-path="functions-1.html"><a href="functions-1.html#subsetting-vectors"><i class="fa fa-check"></i><b>18.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="18.6" data-path="functions-1.html"><a href="functions-1.html#sequences-of-numbers"><i class="fa fa-check"></i><b>18.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="18.7" data-path="functions-1.html"><a href="functions-1.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>18.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="18.8" data-path="functions-1.html"><a href="functions-1.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>18.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>19</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="19.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-3"><i class="fa fa-check"></i><b>19.1</b> Preface</a></li>
<li class="chapter" data-level="19.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>19.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="19.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>19.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>20</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="20.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-4"><i class="fa fa-check"></i><b>20.1</b> Preface</a></li>
<li class="chapter" data-level="20.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>20.2</b> Vocab</a></li>
<li class="chapter" data-level="20.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-2"><i class="fa fa-check"></i><b>20.3</b> Functions</a></li>
<li class="chapter" data-level="20.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>20.4</b> Functions in R</a></li>
<li class="chapter" data-level="20.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>20.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html"><i class="fa fa-check"></i><b>21</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="21.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#commands-used"><i class="fa fa-check"></i><b>21.1</b> Commands used</a></li>
<li class="chapter" data-level="21.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#r-objects"><i class="fa fa-check"></i><b>21.2</b> R Objects</a></li>
<li class="chapter" data-level="21.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>21.3</b> Differences between objects</a></li>
<li class="chapter" data-level="21.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-data"><i class="fa fa-check"></i><b>21.4</b> The Data</a></li>
<li class="chapter" data-level="21.5" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>21.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="21.5.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#is"><i class="fa fa-check"></i><b>21.5.1</b> is()</a></li>
<li class="chapter" data-level="21.5.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#length"><i class="fa fa-check"></i><b>21.5.2</b> length()</a></li>
<li class="chapter" data-level="21.5.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#str"><i class="fa fa-check"></i><b>21.5.3</b> str()</a></li>
<li class="chapter" data-level="21.5.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#c"><i class="fa fa-check"></i><b>21.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="21.6" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#debrief"><i class="fa fa-check"></i><b>21.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="fasta-files.html"><a href="fasta-files.html"><i class="fa fa-check"></i><b>22</b> FASTA Files</a>
<ul>
<li class="chapter" data-level="22.1" data-path="fasta-files.html"><a href="fasta-files.html#example-fasta-file"><i class="fa fa-check"></i><b>22.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="22.2" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>22.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="22.3" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>22.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="22.4" data-path="fasta-files.html"><a href="fasta-files.html#fastq-format"><i class="fa fa-check"></i><b>22.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>23</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="23.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-3"><i class="fa fa-check"></i><b>23.0.1</b> Functions</a></li>
<li class="chapter" data-level="23.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>23.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="23.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>23.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="23.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>23.0.4</b> File types</a></li>
<li class="chapter" data-level="23.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>23.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="23.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>23.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="23.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries"><i class="fa fa-check"></i><b>23.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="23.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>23.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="23.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>23.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="23.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>23.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="23.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>23.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="23.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>23.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="23.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>23.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="23.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>23.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="23.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>23.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>24</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="24.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-1"><i class="fa fa-check"></i><b>24.1</b> Preliminaries</a></li>
<li class="chapter" data-level="24.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>24.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="24.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>24.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="24.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>24.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="25" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>25</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="25.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-5"><i class="fa fa-check"></i><b>25.1</b> Preface</a></li>
<li class="chapter" data-level="25.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>25.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="25.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>25.3</b> Introduction</a></li>
<li class="chapter" data-level="25.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>25.4</b> Vocabulary</a></li>
<li class="chapter" data-level="25.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions-4"><i class="fa fa-check"></i><b>25.5</b> Functions</a></li>
<li class="chapter" data-level="25.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries-2"><i class="fa fa-check"></i><b>25.6</b> Preliminaries</a></li>
<li class="chapter" data-level="25.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>25.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="25.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>25.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="25.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>25.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="25.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>25.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="25.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>25.8.3</b> DNA words</a></li>
<li class="chapter" data-level="25.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary-1"><i class="fa fa-check"></i><b>25.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="25.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>25.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="25.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>25.9.1</b> License</a></li>
<li class="chapter" data-level="25.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>25.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="26" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>26</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="26.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>26.1</b> Preliminaries</a></li>
<li class="chapter" data-level="26.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>26.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="changes-to-database-entries.html"><a href="changes-to-database-entries.html"><i class="fa fa-check"></i><b>27</b> Changes to database entries</a></li>
<li class="chapter" data-level="28" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>28</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="28.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>28.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="28.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>28.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="28.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>28.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="28.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>28.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="28.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>28.4</b> Inversions</a></li>
<li class="chapter" data-level="28.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>28.5</b> Translocations</a></li>
<li class="chapter" data-level="28.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>28.6</b> Random sequence</a></li>
<li class="chapter" data-level="28.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>28.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>29</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="29.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>29.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="29.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>29.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="29.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>29.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="29.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>29.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="29.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>29.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="29.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>29.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="29.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>29.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="29.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>29.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>30</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="30.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>30.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="30.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>30.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="30.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>30.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="31" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html"><i class="fa fa-check"></i><b>31</b> Alignment by eye in R</a>
<ul>
<li class="chapter" data-level="31.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>31.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="31.1.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#packages"><i class="fa fa-check"></i><b>31.1.1</b> Packages</a></li>
<li class="chapter" data-level="31.1.2" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#data"><i class="fa fa-check"></i><b>31.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="32" data-path="alignment-in-r.html"><a href="alignment-in-r.html"><i class="fa fa-check"></i><b>32</b> Alignment in R</a>
<ul>
<li class="chapter" data-level="32.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#preliminaries-8"><i class="fa fa-check"></i><b>32.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="32.1.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#packages-1"><i class="fa fa-check"></i><b>32.1.1</b> Packages</a></li>
<li class="chapter" data-level="32.1.2" data-path="alignment-in-r.html"><a href="alignment-in-r.html#data-1"><i class="fa fa-check"></i><b>32.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="33" data-path="working-with-vectors-and-matrices.html"><a href="working-with-vectors-and-matrices.html"><i class="fa fa-check"></i><b>33</b> Working with vectors and matrices</a></li>
<li class="chapter" data-level="34" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html"><i class="fa fa-check"></i><b>34</b> Matrix vocabulary</a>
<ul>
<li class="chapter" data-level="34.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#introduction-1"><i class="fa fa-check"></i><b>34.1</b> Introduction</a></li>
<li class="chapter" data-level="34.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#preliminaries-9"><i class="fa fa-check"></i><b>34.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="34.2.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#packages-2"><i class="fa fa-check"></i><b>34.2.1</b> Packages</a></li>
<li class="chapter" data-level="34.2.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#vocab-3"><i class="fa fa-check"></i><b>34.2.2</b> Vocab</a></li>
<li class="chapter" data-level="34.2.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#functions-used"><i class="fa fa-check"></i><b>34.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="34.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#load-packages"><i class="fa fa-check"></i><b>34.3</b> Load packages</a></li>
<li class="chapter" data-level="34.4" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#matrix-structure"><i class="fa fa-check"></i><b>34.4</b> Matrix structure</a></li>
<li class="chapter" data-level="34.5" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#diagonal-of-a-matrix"><i class="fa fa-check"></i><b>34.5</b> Diagonal of a matrix</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html"><i class="fa fa-check"></i><b>35</b> Testing the significance of an alignment</a>
<ul>
<li class="chapter" data-level="35.1" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#calculating-the-statistical-significance-of-a-pairwise-global-alignment"><i class="fa fa-check"></i><b>35.1</b> Calculating the statistical significance of a pairwise global alignment</a></li>
<li class="chapter" data-level="35.2" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#summary-2"><i class="fa fa-check"></i><b>35.2</b> Summary</a></li>
<li class="chapter" data-level="35.3" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#links-and-further-reading-1"><i class="fa fa-check"></i><b>35.3</b> Links and Further Reading</a></li>
<li class="chapter" data-level="35.4" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#exercises-1"><i class="fa fa-check"></i><b>35.4</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>36</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="36.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>36.1</b> Prelminaries</a></li>
<li class="chapter" data-level="36.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>36.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="36.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>36.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html"><i class="fa fa-check"></i><b>37</b> Manipulating matrices and vectors: worked example</a>
<ul>
<li class="chapter" data-level="37.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#introduction-2"><i class="fa fa-check"></i><b>37.1</b> Introduction</a></li>
<li class="chapter" data-level="37.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#preliminaries-10"><i class="fa fa-check"></i><b>37.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="37.2.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-3"><i class="fa fa-check"></i><b>37.2.1</b> Packages</a></li>
<li class="chapter" data-level="37.2.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#vocab-4"><i class="fa fa-check"></i><b>37.2.2</b> Vocab</a></li>
<li class="chapter" data-level="37.2.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#functions-used-1"><i class="fa fa-check"></i><b>37.2.3</b> Functions used</a></li>
<li class="chapter" data-level="37.2.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-4"><i class="fa fa-check"></i><b>37.2.4</b> Packages</a></li>
<li class="chapter" data-level="37.2.5" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#data-2"><i class="fa fa-check"></i><b>37.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="37.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#matrix-elements-can-be-accessed-with-square-bracket-notation"><i class="fa fa-check"></i><b>37.3</b> Matrix elements can be accessed with square bracket notation</a></li>
<li class="chapter" data-level="37.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#r-object-names"><i class="fa fa-check"></i><b>37.4</b> R object names</a>
<ul>
<li class="chapter" data-level="37.4.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-vectors"><i class="fa fa-check"></i><b>37.4.1</b> Accessing items in names vectors</a></li>
<li class="chapter" data-level="37.4.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-matrices"><i class="fa fa-check"></i><b>37.4.2</b> Accessing items in names matrices</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="38" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>38</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="38.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-11"><i class="fa fa-check"></i><b>38.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="38.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages-5"><i class="fa fa-check"></i><b>38.1.1</b> Packages</a></li>
<li class="chapter" data-level="38.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-5"><i class="fa fa-check"></i><b>38.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="38.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>38.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="38.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>38.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="38.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>38.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="38.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>38.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="38.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>38.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="38.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>38.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="38.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>38.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="38.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>38.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html"><i class="fa fa-check"></i><b>39</b> The BLOSUM scoring matrix in R</a>
<ul>
<li class="chapter" data-level="39.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#introduction-3"><i class="fa fa-check"></i><b>39.1</b> Introduction</a></li>
<li class="chapter" data-level="39.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#preliminaries-12"><i class="fa fa-check"></i><b>39.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="39.2.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#packages-6"><i class="fa fa-check"></i><b>39.2.1</b> Packages</a></li>
<li class="chapter" data-level="39.2.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#vocab-5"><i class="fa fa-check"></i><b>39.2.2</b> Vocab</a></li>
<li class="chapter" data-level="39.2.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#functions-used-2"><i class="fa fa-check"></i><b>39.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="39.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#load-packages-1"><i class="fa fa-check"></i><b>39.3</b> Load packages</a></li>
<li class="chapter" data-level="39.4" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#the-blosum-matrix"><i class="fa fa-check"></i><b>39.4</b> The BLOSUM matrix</a></li>
<li class="chapter" data-level="39.5" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#visualizing-the-blosum-matrix"><i class="fa fa-check"></i><b>39.5</b> Visualizing the BLOSUM matrix</a></li>
</ul></li>
<li class="chapter" data-level="40" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>40</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="41" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>41</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="41.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-4"><i class="fa fa-check"></i><b>41.1</b> Introduction</a></li>
<li class="chapter" data-level="41.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>41.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="41.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-6"><i class="fa fa-check"></i><b>41.2.1</b> Vocab</a></li>
<li class="chapter" data-level="41.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>41.2.2</b> R functions</a></li>
<li class="chapter" data-level="41.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>41.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="41.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>41.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="41.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>41.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="41.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>41.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="41.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>41.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="41.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>41.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="41.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>41.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="41.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>41.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="41.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>41.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="41.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>41.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="41.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>41.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="41.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>41.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="41.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>41.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="42" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>42</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="42.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-13"><i class="fa fa-check"></i><b>42.1</b> Preliminaries</a></li>
<li class="chapter" data-level="42.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-5"><i class="fa fa-check"></i><b>42.2</b> Introduction</a></li>
<li class="chapter" data-level="42.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>42.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="42.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>42.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="43" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>43</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="43.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-14"><i class="fa fa-check"></i><b>43.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="43.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>43.1.1</b> Key functions</a></li>
<li class="chapter" data-level="43.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>43.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="43.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>43.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="43.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>43.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="43.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>43.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="43.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>43.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="43.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>43.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="44" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html"><i class="fa fa-check"></i><b>44</b> Local variation in GC content</a>
<ul>
<li class="chapter" data-level="44.0.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#note-on-the-biology-in-this-section"><i class="fa fa-check"></i><b>44.0.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="44.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#vocabulary-2"><i class="fa fa-check"></i><b>44.1</b> Vocabulary</a></li>
<li class="chapter" data-level="44.2" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#reading-sequence-data-with-rentrezentrez_fetch"><i class="fa fa-check"></i><b>44.2</b> Reading sequence data with rentrez::entrez_fetch</a></li>
<li class="chapter" data-level="44.3" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#local-variation-in-gc-content-1"><i class="fa fa-check"></i><b>44.3</b> Local variation in GC content</a></li>
<li class="chapter" data-level="44.4" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-analysis-of-gc-content"><i class="fa fa-check"></i><b>44.4</b> A sliding window analysis of GC content</a></li>
<li class="chapter" data-level="44.5" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-plot-of-gc-content"><i class="fa fa-check"></i><b>44.5</b> A sliding window plot of GC content</a>
<ul>
<li class="chapter" data-level="44.5.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#over-represented-and-under-represented-dna-words"><i class="fa fa-check"></i><b>44.5.1</b> Over-represented and under-represented DNA words</a></li>
</ul></li>
<li class="chapter" data-level="44.6" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#summary-3"><i class="fa fa-check"></i><b>44.6</b> Summary</a></li>
<li class="chapter" data-level="44.7" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#further-reading"><i class="fa fa-check"></i><b>44.7</b> Further Reading</a></li>
<li class="chapter" data-level="44.8" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#acknowledgements-1"><i class="fa fa-check"></i><b>44.8</b> Acknowledgements</a></li>
<li class="chapter" data-level="44.9" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#exercises-2"><i class="fa fa-check"></i><b>44.9</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="45" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html"><i class="fa fa-check"></i><b>45</b> Computational gene finding</a>
<ul>
<li class="chapter" data-level="45.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#preliminaries-15"><i class="fa fa-check"></i><b>45.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="45.1.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#packages-7"><i class="fa fa-check"></i><b>45.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="45.2" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#the-genetic-code"><i class="fa fa-check"></i><b>45.2</b> The genetic code</a></li>
<li class="chapter" data-level="45.3" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-start-and-stop-codons-in-a-dna-sequence"><i class="fa fa-check"></i><b>45.3</b> Finding start and stop codons in a DNA sequence</a></li>
<li class="chapter" data-level="45.4" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#reading-frames"><i class="fa fa-check"></i><b>45.4</b> Reading frames</a></li>
<li class="chapter" data-level="45.5" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-forward-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>45.5</b> Finding open reading frames on the forward strand of a DNA sequence</a></li>
<li class="chapter" data-level="45.6" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#predicting-the-protein-sequence-for-an-orf"><i class="fa fa-check"></i><b>45.6</b> Predicting the protein sequence for an ORF</a></li>
<li class="chapter" data-level="45.7" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-reverse-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>45.7</b> Finding open reading frames on the reverse strand of a DNA sequence</a></li>
<li class="chapter" data-level="45.8" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#lengths-of-open-reading-frames"><i class="fa fa-check"></i><b>45.8</b> Lengths of open reading frames</a></li>
</ul></li>
<li class="chapter" data-level="46" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html"><i class="fa fa-check"></i><b>46</b> Significance of ORFs</a>
<ul>
<li class="chapter" data-level="46.1" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#preliminaries-16"><i class="fa fa-check"></i><b>46.1</b> Preliminaries</a></li>
<li class="chapter" data-level="46.2" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#identifying-significant-open-reading-frames"><i class="fa fa-check"></i><b>46.2</b> Identifying significant open reading frames</a></li>
<li class="chapter" data-level="46.3" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#summary-4"><i class="fa fa-check"></i><b>46.3</b> Summary</a></li>
<li class="chapter" data-level="46.4" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#links-and-further-reading-2"><i class="fa fa-check"></i><b>46.4</b> Links and Further Reading</a></li>
<li class="chapter" data-level="46.5" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#acknowledgements-2"><i class="fa fa-check"></i><b>46.5</b> Acknowledgements</a></li>
<li class="chapter" data-level="46.6" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#exercises-3"><i class="fa fa-check"></i><b>46.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="47" data-path="comparative-genomics.html"><a href="comparative-genomics.html"><i class="fa fa-check"></i><b>47</b> Comparative Genomics</a>
<ul>
<li class="chapter" data-level="47.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#preliminaries-17"><i class="fa fa-check"></i><b>47.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="47.1.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#packages-8"><i class="fa fa-check"></i><b>47.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="47.2" data-path="comparative-genomics.html"><a href="comparative-genomics.html#introduction-6"><i class="fa fa-check"></i><b>47.2</b> Introduction</a></li>
<li class="chapter" data-level="47.3" data-path="comparative-genomics.html"><a href="comparative-genomics.html#using-the-biomart-r-library-to-query-the-ensembl-database"><i class="fa fa-check"></i><b>47.3</b> Using the biomaRt R Library to Query the Ensembl Database</a></li>
<li class="chapter" data-level="47.4" data-path="comparative-genomics.html"><a href="comparative-genomics.html#comparing-the-number-of-genes-in-two-species"><i class="fa fa-check"></i><b>47.4</b> Comparing the number of genes in two species</a></li>
<li class="chapter" data-level="47.5" data-path="comparative-genomics.html"><a href="comparative-genomics.html#identifying-homologous-genes-between-two-species"><i class="fa fa-check"></i><b>47.5</b> Identifying homologous genes between two species</a></li>
</ul></li>
<li class="chapter" data-level="48" data-path="summary-5.html"><a href="summary-5.html"><i class="fa fa-check"></i><b>48</b> Summary</a></li>
<li class="chapter" data-level="49" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html"><i class="fa fa-check"></i><b>49</b> #Links and Further Reading</a>
<ul>
<li class="chapter" data-level="49.1" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#acknowledgements-3"><i class="fa fa-check"></i><b>49.1</b> Acknowledgements</a></li>
<li class="chapter" data-level="49.2" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#exercises-4"><i class="fa fa-check"></i><b>49.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="50" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html"><i class="fa fa-check"></i><b>50</b> Hidden Markov Models</a>
<ul>
<li class="chapter" data-level="50.1" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-multinomial-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>50.1</b> A multinomial model of DNA sequence evolution</a></li>
<li class="chapter" data-level="50.2" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-multinomial-model"><i class="fa fa-check"></i><b>50.2</b> Generating a DNA sequence using a multinomial model</a></li>
<li class="chapter" data-level="50.3" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>50.3</b> A Markov model of DNA sequence evolution</a></li>
<li class="chapter" data-level="50.4" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-for-a-markov-model"><i class="fa fa-check"></i><b>50.4</b> The transition matrix for a Markov model</a></li>
<li class="chapter" data-level="50.5" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-markov-model"><i class="fa fa-check"></i><b>50.5</b> Generating a DNA sequence using a Markov model</a></li>
<li class="chapter" data-level="50.6" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>50.6</b> A Hidden Markov Model of DNA sequence evolution</a></li>
<li class="chapter" data-level="50.7" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-and-emission-matrix-for-a-hmm"><i class="fa fa-check"></i><b>50.7</b> The transition matrix and emission matrix for a HMM</a></li>
<li class="chapter" data-level="50.8" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-hmm"><i class="fa fa-check"></i><b>50.8</b> Generating a DNA sequence using a HMM</a></li>
<li class="chapter" data-level="50.9" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#inferring-the-states-of-a-hmm-that-generated-a-dna-sequence"><i class="fa fa-check"></i><b>50.9</b> Inferring the states of a HMM that generated a DNA sequence</a></li>
<li class="chapter" data-level="50.10" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-protein-sequence-evolution"><i class="fa fa-check"></i><b>50.10</b> A Hidden Markov Model of protein sequence evolution</a></li>
<li class="chapter" data-level="50.11" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#summary-6"><i class="fa fa-check"></i><b>50.11</b> Summary</a></li>
<li class="chapter" data-level="50.12" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#links-and-further-reading-4"><i class="fa fa-check"></i><b>50.12</b> Links and Further Reading</a></li>
<li class="chapter" data-level="50.13" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#exercises-5"><i class="fa fa-check"></i><b>50.13</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="51" data-path="blast-summary.html"><a href="blast-summary.html"><i class="fa fa-check"></i><b>51</b> BLAST Summary</a>
<ul>
<li class="chapter" data-level="51.1" data-path="blast-summary.html"><a href="blast-summary.html#blast-overview"><i class="fa fa-check"></i><b>51.1</b> BLAST Overview</a>
<ul>
<li class="chapter" data-level="51.1.1" data-path="blast-summary.html"><a href="blast-summary.html#why-do-a-blast-search"><i class="fa fa-check"></i><b>51.1.1</b> Why do a BLAST search?</a></li>
<li class="chapter" data-level="51.1.2" data-path="blast-summary.html"><a href="blast-summary.html#what-does-blast-let-you-do"><i class="fa fa-check"></i><b>51.1.2</b> What does BLAST let you do?</a></li>
</ul></li>
<li class="chapter" data-level="51.2" data-path="blast-summary.html"><a href="blast-summary.html#blast-recipe"><i class="fa fa-check"></i><b>51.2</b> BLAST recipe</a></li>
<li class="chapter" data-level="51.3" data-path="blast-summary.html"><a href="blast-summary.html#modules-in-the-blast-workflow-and-algorithm"><i class="fa fa-check"></i><b>51.3</b> Modules in the BLAST Workflow and Algorithm</a>
<ul>
<li class="chapter" data-level="51.3.1" data-path="blast-summary.html"><a href="blast-summary.html#module-1-setup-query-sequence-protein-blast-oriented"><i class="fa fa-check"></i><b>51.3.1</b> Module 1: Setup query sequence (Protein BLAST oriented)</a></li>
<li class="chapter" data-level="51.3.2" data-path="blast-summary.html"><a href="blast-summary.html#module-2-scanning-database-for-high-scoring-words"><i class="fa fa-check"></i><b>51.3.2</b> Module 2: Scanning database for high-scoring words</a></li>
<li class="chapter" data-level="51.3.3" data-path="blast-summary.html"><a href="blast-summary.html#module-3-traceback"><i class="fa fa-check"></i><b>51.3.3</b> Module 3: Traceback</a></li>
</ul></li>
<li class="chapter" data-level="51.4" data-path="blast-summary.html"><a href="blast-summary.html#blast-alignment-statistics"><i class="fa fa-check"></i><b>51.4</b> BLAST alignment statistics</a></li>
</ul></li>
<li class="chapter" data-level="52" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html"><i class="fa fa-check"></i><b>52</b> BLAST Alignment score distribution</a>
<ul>
<li class="chapter" data-level="52.1" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#packages-9"><i class="fa fa-check"></i><b>52.1</b> Packages</a></li>
<li class="chapter" data-level="52.2" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#distributions-of-alignment-scores"><i class="fa fa-check"></i><b>52.2</b> Distributions of alignment scores</a></li>
<li class="chapter" data-level="52.3" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-data"><i class="fa fa-check"></i><b>52.3</b> Simulating data</a></li>
<li class="chapter" data-level="52.4" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#lots-of-data-in-biology-is-normal"><i class="fa fa-check"></i><b>52.4</b> Lots of data in biology is normal</a></li>
<li class="chapter" data-level="52.5" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-the-normal-distribution"><i class="fa fa-check"></i><b>52.5</b> Simulating the normal distribution</a></li>
<li class="chapter" data-level="52.6" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#the-extreme-value-distribution"><i class="fa fa-check"></i><b>52.6</b> The extreme value distribution</a></li>
<li class="chapter" data-level="52.7" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#replicating-figure-4.14"><i class="fa fa-check"></i><b>52.7</b> Replicating Figure 4.14</a></li>
<li class="chapter" data-level="52.8" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#why-do-we-care"><i class="fa fa-check"></i><b>52.8</b> Why do we care?</a></li>
</ul></li>
<li class="chapter" data-level="53" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><i class="fa fa-check"></i><b>53</b> Calculating E values, Bit scores, and P-values from BLAST output</a>
<ul>
<li class="chapter" data-level="53.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-7"><i class="fa fa-check"></i><b>53.1</b> Introduction</a></li>
<li class="chapter" data-level="53.2" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#case-study"><i class="fa fa-check"></i><b>53.2</b> Case Study</a></li>
<li class="chapter" data-level="53.3" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-statistics"><i class="fa fa-check"></i><b>53.3</b> BLAST Statistics</a></li>
<li class="chapter" data-level="53.4" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot"><i class="fa fa-check"></i><b>53.4</b> Dotplot</a>
<ul>
<li class="chapter" data-level="53.4.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot-in-r---optional-experimental-code"><i class="fa fa-check"></i><b>53.4.1</b> Dotplot in R - optional, experimental code</a></li>
</ul></li>
<li class="chapter" data-level="53.5" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-search-summary"><i class="fa fa-check"></i><b>53.5</b> BLAST “Search Summary”</a></li>
<li class="chapter" data-level="53.6" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-alignments"><i class="fa fa-check"></i><b>53.6</b> BLAST Alignments</a></li>
<li class="chapter" data-level="53.7" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#equations"><i class="fa fa-check"></i><b>53.7</b> Equations</a>
<ul>
<li class="chapter" data-level="53.7.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-value-eq-4.5-pg-142"><i class="fa fa-check"></i><b>53.7.1</b> E value (eq 4.5, pg 142)</a></li>
</ul></li>
<li class="chapter" data-level="53.8" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bit-scores-s"><i class="fa fa-check"></i><b>53.8</b> Bit Scores (S’)</a></li>
<li class="chapter" data-level="53.9" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-from-bit-scores"><i class="fa fa-check"></i><b>53.9</b> E from Bit scores</a></li>
<li class="chapter" data-level="53.10" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#p-values"><i class="fa fa-check"></i><b>53.10</b> P values</a></li>
<li class="chapter" data-level="53.11" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#repliminaries"><i class="fa fa-check"></i><b>53.11</b> Repliminaries</a></li>
<li class="chapter" data-level="53.12" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#concepts"><i class="fa fa-check"></i><b>53.12</b> Concepts</a></li>
<li class="chapter" data-level="53.13" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#functions-6"><i class="fa fa-check"></i><b>53.13</b> Functions</a></li>
<li class="chapter" data-level="53.14" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-8"><i class="fa fa-check"></i><b>53.14</b> Introduction</a></li>
<li class="chapter" data-level="53.15" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees"><i class="fa fa-check"></i><b>53.15</b> Number of rooted trees</a></li>
<li class="chapter" data-level="53.16" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees-in-r"><i class="fa fa-check"></i><b>53.16</b> Number of rooted trees in R</a>
<ul>
<li class="chapter" data-level="53.16.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#unique-topologies-versus-rotations"><i class="fa fa-check"></i><b>53.16.1</b> Unique topologies versus rotations</a></li>
</ul></li>
<li class="chapter" data-level="53.17" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bilography"><i class="fa fa-check"></i><b>53.17</b> Bilography</a></li>
</ul></li>
<li class="chapter" data-level="54" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html"><i class="fa fa-check"></i><b>54</b> UPGMA the hard way</a>
<ul>
<li class="chapter" data-level="54.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#introudction-to-upgma"><i class="fa fa-check"></i><b>54.1</b> Introudction to UPGMA</a></li>
<li class="chapter" data-level="54.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#preliminaries-18"><i class="fa fa-check"></i><b>54.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="54.2.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#r-libraries"><i class="fa fa-check"></i><b>54.2.1</b> R Libraries</a></li>
<li class="chapter" data-level="54.2.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#custom-function-well-use"><i class="fa fa-check"></i><b>54.2.2</b> Custom function we’ll use</a></li>
</ul></li>
<li class="chapter" data-level="54.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#data-matrix"><i class="fa fa-check"></i><b>54.3</b> Data matrix</a>
<ul>
<li class="chapter" data-level="54.3.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#building-the-matrix"><i class="fa fa-check"></i><b>54.3.1</b> Building the matrix</a></li>
</ul></li>
<li class="chapter" data-level="54.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#what-does-this-distance-matrix-represent"><i class="fa fa-check"></i><b>54.4</b> What does this distance matrix represent?</a>
<ul>
<li class="chapter" data-level="54.4.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-pairwise-linear-distances"><i class="fa fa-check"></i><b>54.4.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="54.4.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-2d-map-of-linear-distances"><i class="fa fa-check"></i><b>54.4.2</b> Genetic distances conceptualized as 2D map of linear distances</a></li>
</ul></li>
<li class="chapter" data-level="54.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#from-linear-distances-to-tree-distances"><i class="fa fa-check"></i><b>54.5</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="54.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#the-upgma-algorithm"><i class="fa fa-check"></i><b>54.6</b> The UPGMA algorithm</a>
<ul>
<li class="chapter" data-level="54.6.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-1-find-the-2-taxa-that-are-currently-closest-together"><i class="fa fa-check"></i><b>54.6.1</b> Step 1: find the 2 taxa that are currently closest together</a></li>
<li class="chapter" data-level="54.6.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node"><i class="fa fa-check"></i><b>54.6.2</b> Step 2: calculate the branch lengths connect the taxa to their shared node</a></li>
<li class="chapter" data-level="54.6.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-determine-if-the-algorithm-is-done"><i class="fa fa-check"></i><b>54.6.3</b> Step 3a: determine if the algorithm is done</a></li>
<li class="chapter" data-level="54.6.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u"><i class="fa fa-check"></i><b>54.6.4</b> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</a></li>
<li class="chapter" data-level="54.6.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-4-determine-the-distance-from-u-to-the-other-remaining-clusters."><i class="fa fa-check"></i><b>54.6.5</b> Step 4: Determine the distance from u to the other remaining clusters.</a></li>
<li class="chapter" data-level="54.6.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-5-create-a-new-matrix"><i class="fa fa-check"></i><b>54.6.6</b> Step 5: Create a new matrix</a></li>
<li class="chapter" data-level="54.6.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa"><i class="fa fa-check"></i><b>54.6.7</b> Key step: calculating the distance from a clade to the remaining taxa</a></li>
</ul></li>
<li class="chapter" data-level="54.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-2d-representation"><i class="fa fa-check"></i><b>54.7</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="54.8" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---round-1"><i class="fa fa-check"></i><b>54.8</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="54.8.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#determine-minimum-value-in-current-matrix"><i class="fa fa-check"></i><b>54.8.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="54.8.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade"><i class="fa fa-check"></i><b>54.8.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.8.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length"><i class="fa fa-check"></i><b>54.8.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="54.8.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>54.8.4</b> Calculate distance from clade to all other points</a></li>
<li class="chapter" data-level="54.8.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-matrix-with-new-distance"><i class="fa fa-check"></i><b>54.8.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="54.9" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---next-round"><i class="fa fa-check"></i><b>54.9</b> Algorithm - next round</a>
<ul>
<li class="chapter" data-level="54.9.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-1"><i class="fa fa-check"></i><b>54.9.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.9.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-1"><i class="fa fa-check"></i><b>54.9.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="54.9.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#recalculate-distance-from-abe-to-all-remainign-taxa"><i class="fa fa-check"></i><b>54.9.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="54.9.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-the-matrix"><i class="fa fa-check"></i><b>54.9.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="54.10" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#next-iteration"><i class="fa fa-check"></i><b>54.10</b> Next iteration</a>
<ul>
<li class="chapter" data-level="54.10.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-lengths"><i class="fa fa-check"></i><b>54.10.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="54.10.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-2"><i class="fa fa-check"></i><b>54.10.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.10.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-2"><i class="fa fa-check"></i><b>54.10.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="54.10.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#distance-of-c-to-abe-and-d-to-abe"><i class="fa fa-check"></i><b>54.10.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="54.11" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finish-up"><i class="fa fa-check"></i><b>54.11</b> Finish up</a>
<ul>
<li class="chapter" data-level="54.11.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-3"><i class="fa fa-check"></i><b>54.11.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="54.11.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-3"><i class="fa fa-check"></i><b>54.11.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="54.12" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finalizing-branch-lengths"><i class="fa fa-check"></i><b>54.12</b> Finalizing branch lengths</a></li>
</ul></li>
<li class="chapter" data-level="55" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html"><i class="fa fa-check"></i><b>55</b> Representing phylogenetic trees in Newick format</a>
<ul>
<li class="chapter" data-level="55.1" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#vocab-7"><i class="fa fa-check"></i><b>55.1</b> Vocab</a></li>
<li class="chapter" data-level="55.2" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#introduction-9"><i class="fa fa-check"></i><b>55.2</b> Introduction</a></li>
<li class="chapter" data-level="55.3" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-three-taxa"><i class="fa fa-check"></i><b>55.3</b> Trees with three taxa</a></li>
<li class="chapter" data-level="55.4" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-four-taxa"><i class="fa fa-check"></i><b>55.4</b> Trees with four taxa</a></li>
</ul></li>
<li class="chapter" data-level="56" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html"><i class="fa fa-check"></i><b>56</b> Plotting trees from Newick notation in R</a>
<ul>
<li class="chapter" data-level="56.1" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#preliminaries-19"><i class="fa fa-check"></i><b>56.1</b> Preliminaries</a></li>
<li class="chapter" data-level="56.2" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#note-1"><i class="fa fa-check"></i><b>56.2</b> Note</a></li>
<li class="chapter" data-level="56.3" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick"><i class="fa fa-check"></i><b>56.3</b> 3 taxa tree in Newick</a></li>
<li class="chapter" data-level="56.4" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick-1"><i class="fa fa-check"></i><b>56.4</b> 4 taxa tree in Newick</a></li>
<li class="chapter" data-level="56.5" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-in-newick-format"><i class="fa fa-check"></i><b>56.5</b> 5 taxa in Newick Format</a></li>
<li class="chapter" data-level="56.6" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#many-taxa-tree"><i class="fa fa-check"></i><b>56.6</b> Many-taxa tree</a></li>
<li class="chapter" data-level="56.7" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#newick-and-branch-lengths"><i class="fa fa-check"></i><b>56.7</b> Newick and Branch lengths</a></li>
<li class="chapter" data-level="56.8" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#additional-reading"><i class="fa fa-check"></i><b>56.8</b> Additional Reading</a></li>
</ul></li>
<li class="chapter" data-level="57" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><i class="fa fa-check"></i><b>57</b> Phylogenetic trees using the UPGMA clustering algoirthm</a>
<ul>
<li class="chapter" data-level="57.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#introduction-10"><i class="fa fa-check"></i><b>57.1</b> Introduction</a></li>
<li class="chapter" data-level="57.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#preliminaries-20"><i class="fa fa-check"></i><b>57.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="57.2.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#vocab-8"><i class="fa fa-check"></i><b>57.2.1</b> Vocab</a></li>
<li class="chapter" data-level="57.2.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-r-functions"><i class="fa fa-check"></i><b>57.2.2</b> General R functions</a></li>
<li class="chapter" data-level="57.2.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#specific-functions"><i class="fa fa-check"></i><b>57.2.3</b> Specific functions</a></li>
</ul></li>
<li class="chapter" data-level="57.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#sequence-comparisons"><i class="fa fa-check"></i><b>57.3</b> Sequence comparisons</a></li>
<li class="chapter" data-level="57.4" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrices"><i class="fa fa-check"></i><b>57.4</b> Distance matrices</a></li>
<li class="chapter" data-level="57.5" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#from-a-distance-matrix-to-phylogenetic-tree"><i class="fa fa-check"></i><b>57.5</b> From a distance matrix to phylogenetic tree</a></li>
<li class="chapter" data-level="57.6" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-raw-sequences"><i class="fa fa-check"></i><b>57.6</b> Distance matrix in R from raw sequences</a></li>
<li class="chapter" data-level="57.7" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-normal-matrix"><i class="fa fa-check"></i><b>57.7</b> Distance matrix in R from normal matrix</a></li>
<li class="chapter" data-level="57.8" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#upgma-in-r"><i class="fa fa-check"></i><b>57.8</b> UPGMA in R</a></li>
<li class="chapter" data-level="57.9" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#how-upgma-clustering-works"><i class="fa fa-check"></i><b>57.9</b> How UPGMA clustering works</a>
<ul>
<li class="chapter" data-level="57.9.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#challenge"><i class="fa fa-check"></i><b>57.9.1</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="57.10" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#further-information"><i class="fa fa-check"></i><b>57.10</b> Further information</a>
<ul>
<li class="chapter" data-level="57.10.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-info"><i class="fa fa-check"></i><b>57.10.1</b> General info</a></li>
<li class="chapter" data-level="57.10.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#worked-examples"><i class="fa fa-check"></i><b>57.10.2</b> Worked examples:</a></li>
<li class="chapter" data-level="57.10.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#additional-notes"><i class="fa fa-check"></i><b>57.10.3</b> Additional notes:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="58" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><i class="fa fa-check"></i><b>58</b> Tutorial: Writing functions to calculate the number of phylogenetic trees</a>
<ul>
<li class="chapter" data-level="58.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#repliminaries-1"><i class="fa fa-check"></i><b>58.1</b> Repliminaries</a>
<ul>
<li class="chapter" data-level="58.1.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-7"><i class="fa fa-check"></i><b>58.1.1</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="58.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#concepts-1"><i class="fa fa-check"></i><b>58.2</b> Concepts</a>
<ul>
<li class="chapter" data-level="58.2.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vocab-9"><i class="fa fa-check"></i><b>58.2.1</b> Vocab:</a></li>
<li class="chapter" data-level="58.2.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#packages-10"><i class="fa fa-check"></i><b>58.2.2</b> Packages:</a></li>
</ul></li>
<li class="chapter" data-level="58.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#introduction-11"><i class="fa fa-check"></i><b>58.3</b> Introduction</a></li>
<li class="chapter" data-level="58.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-1"><i class="fa fa-check"></i><b>58.4</b> Number of rooted trees</a></li>
<li class="chapter" data-level="58.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-in-r-1"><i class="fa fa-check"></i><b>58.5</b> Number of rooted trees in R</a></li>
<li class="chapter" data-level="58.6" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-in-r-1"><i class="fa fa-check"></i><b>58.6</b> Functions in R</a></li>
<li class="chapter" data-level="58.7" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-to-calculate-the-number-of-possible-phylogenetic-trees"><i class="fa fa-check"></i><b>58.7</b> Function to calculate the number of possible phylogenetic trees</a></li>
<li class="chapter" data-level="58.8" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vectorized-inputs-to-functions"><i class="fa fa-check"></i><b>58.8</b> Vectorized inputs to functions</a></li>
<li class="chapter" data-level="58.9" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-conditional-statements"><i class="fa fa-check"></i><b>58.9</b> Adding conditional statements</a></li>
<li class="chapter" data-level="58.10" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-multiple-conditional-statements"><i class="fa fa-check"></i><b>58.10</b> Adding multiple conditional statements</a></li>
<li class="chapter" data-level="58.11" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-additional-arguments"><i class="fa fa-check"></i><b>58.11</b> Adding additional arguments</a></li>
<li class="chapter" data-level="58.12" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-_un_rooted-trees"><i class="fa fa-check"></i><b>58.12</b> Assignment: _un_rooted trees</a></li>
<li class="chapter" data-level="58.13" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-1"><i class="fa fa-check"></i><b>58.13</b> Assignment part 1</a>
<ul>
<li class="chapter" data-level="58.13.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#part-1-answer"><i class="fa fa-check"></i><b>58.13.1</b> Part 1 answer</a></li>
</ul></li>
<li class="chapter" data-level="58.14" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-2"><i class="fa fa-check"></i><b>58.14</b> Assignment part 2</a>
<ul>
<li class="chapter" data-level="58.14.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#answer"><i class="fa fa-check"></i><b>58.14.1</b> Answer</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="59" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html"><i class="fa fa-check"></i><b>59</b> Writing functions - practice problems</a>
<ul>
<li class="chapter" data-level="59.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#introduction-12"><i class="fa fa-check"></i><b>59.1</b> Introduction</a></li>
<li class="chapter" data-level="59.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#part-1-pure-practice"><i class="fa fa-check"></i><b>59.2</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="59.2.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-1-hello-world"><i class="fa fa-check"></i><b>59.2.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="59.2.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______"><i class="fa fa-check"></i><b>59.2.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="59.3" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function"><i class="fa fa-check"></i><b>59.3</b> Practice Function 3: Make you own natural log function</a></li>
<li class="chapter" data-level="59.4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-4-make-function-that-converts-dna-directly-to-rna"><i class="fa fa-check"></i><b>59.4</b> Practice Function 4: Make function that converts DNA directly to RNA</a></li>
</ul></li>
<li class="chapter" data-level="60" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html"><i class="fa fa-check"></i><b>60</b> Writing user-friendly functions</a>
<ul>
<li class="chapter" data-level="60.1" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-1"><i class="fa fa-check"></i><b>60.1</b> Version 1</a></li>
<li class="chapter" data-level="60.2" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-2"><i class="fa fa-check"></i><b>60.2</b> Version 2</a></li>
<li class="chapter" data-level="60.3" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-3"><i class="fa fa-check"></i><b>60.3</b> Version 3</a></li>
<li class="chapter" data-level="60.4" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-4-adding-conditions-and-warnigns"><i class="fa fa-check"></i><b>60.4</b> Version 4: Adding conditions and warnigns</a></li>
<li class="chapter" data-level="60.5" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-5"><i class="fa fa-check"></i><b>60.5</b> Version 5</a></li>
</ul></li>
<li class="chapter" data-level="61" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html"><i class="fa fa-check"></i><b>61</b> Function practice exercises</a>
<ul>
<li class="chapter" data-level="61.1" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#part-1-pure-practice-1"><i class="fa fa-check"></i><b>61.1</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="61.1.1" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#practice-function-1-hello-world-1"><i class="fa fa-check"></i><b>61.1.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="61.1.2" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#practice-function-2-hello-_______-1"><i class="fa fa-check"></i><b>61.1.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="61.2" data-path="function-practice-exercises.html"><a href="function-practice-exercises.html#practice-function-3-make-you-own-natural-log-function-1"><i class="fa fa-check"></i><b>61.2</b> Practice Function 3: Make you own natural log function</a></li>
</ul></li>
<li class="chapter" data-level="62" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html"><i class="fa fa-check"></i><b>62</b> for() loops in R</a>
<ul>
<li class="chapter" data-level="62.1" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#introduction-13"><i class="fa fa-check"></i><b>62.1</b> Introduction</a></li>
<li class="chapter" data-level="62.2" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#reminders"><i class="fa fa-check"></i><b>62.2</b> Reminders</a></li>
<li class="chapter" data-level="62.3" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop"><i class="fa fa-check"></i><b>62.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="62.4" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-classic-for-loop"><i class="fa fa-check"></i><b>62.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="62.5" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-power-user-for-loop"><i class="fa fa-check"></i><b>62.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="62.6" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-4-a-for-loop-for-all-purposes"><i class="fa fa-check"></i><b>62.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="62.7" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#advanced-aside-if-else-and-next-statements"><i class="fa fa-check"></i><b>62.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="62.8" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#on-avoiding-for-loops"><i class="fa fa-check"></i><b>62.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="62.9" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#challenge-1"><i class="fa fa-check"></i><b>62.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="63" data-path="random-number-generation.html"><a href="random-number-generation.html"><i class="fa fa-check"></i><b>63</b> Random number generation</a>
<ul>
<li class="chapter" data-level="63.1" data-path="random-number-generation.html"><a href="random-number-generation.html#introduction-14"><i class="fa fa-check"></i><b>63.1</b> Introduction</a></li>
<li class="chapter" data-level="63.2" data-path="random-number-generation.html"><a href="random-number-generation.html#using-set.seed-to-make-simulations-reproducible"><i class="fa fa-check"></i><b>63.2</b> Using set.seed() to make simulations reproducible</a></li>
<li class="chapter" data-level="63.3" data-path="random-number-generation.html"><a href="random-number-generation.html#types-of-random-number-generators"><i class="fa fa-check"></i><b>63.3</b> Types of random number generators</a></li>
<li class="chapter" data-level="63.4" data-path="random-number-generation.html"><a href="random-number-generation.html#reporting-simulations"><i class="fa fa-check"></i><b>63.4</b> Reporting simulations</a></li>
<li class="chapter" data-level="63.5" data-path="random-number-generation.html"><a href="random-number-generation.html#notes"><i class="fa fa-check"></i><b>63.5</b> Notes:</a></li>
</ul></li>
<li class="chapter" data-level="64" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html"><i class="fa fa-check"></i><b>64</b> Randomization practice question</a>
<ul>
<li class="chapter" data-level="64.1" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#practice-question"><i class="fa fa-check"></i><b>64.1</b> Practice question:</a></li>
<li class="chapter" data-level="64.2" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#your-answer"><i class="fa fa-check"></i><b>64.2</b> Your answer:</a></li>
<li class="chapter" data-level="64.3" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-0"><i class="fa fa-check"></i><b>64.3</b> Hint 0</a></li>
<li class="chapter" data-level="64.4" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-1"><i class="fa fa-check"></i><b>64.4</b> Hint 1:</a></li>
<li class="chapter" data-level="64.5" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-2-follow-up-to-hint-0"><i class="fa fa-check"></i><b>64.5</b> Hint 2 (follow up to Hint 0)</a></li>
<li class="chapter" data-level="64.6" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#hint-3"><i class="fa fa-check"></i><b>64.6</b> Hint 3:</a></li>
<li class="chapter" data-level="64.7" data-path="randomization-practice-question.html"><a href="randomization-practice-question.html#answer-1"><i class="fa fa-check"></i><b>64.7</b> Answer</a></li>
</ul></li>
<li class="chapter" data-level="65" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html"><i class="fa fa-check"></i><b>65</b> Simulating random amino acid sequences</a>
<ul>
<li class="chapter" data-level="65.1" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#selecting-random-letters-from-a-vector"><i class="fa fa-check"></i><b>65.1</b> Selecting random letters from a vector</a></li>
<li class="chapter" data-level="65.2" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-letters-from-the-whole-alphabet"><i class="fa fa-check"></i><b>65.2</b> Select random letters from the whole alphabet</a></li>
<li class="chapter" data-level="65.3" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-amino-acids-to-build-polypeptide"><i class="fa fa-check"></i><b>65.3</b> Select random amino acids to build polypeptide</a></li>
</ul></li>
<li class="chapter" data-level="66" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>66</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="66.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>66.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="66.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>66.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="66.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>66.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="66.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>66.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="66.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>66.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-3"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="66.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>66.2</b> Help!</a>
<ul>
<li class="chapter" data-level="66.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>66.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="66.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>66.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="66.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>66.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="66.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>66.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="66.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>66.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="66.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>66.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="66.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>66.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="66.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>66.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="upgma-the-hard-way" class="section level1" number="54">
<h1><span class="header-section-number">Chapter 54</span> UPGMA the hard way</h1>
<div class="sourceCode" id="cb1052"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1052-1"><a href="upgma-the-hard-way.html#cb1052-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compbio4all)</span></code></pre></div>
<div id="introudction-to-upgma" class="section level2" number="54.1">
<h2><span class="header-section-number">54.1</span> Introudction to UPGMA</h2>
<ul>
<li>UPGMA is a basic <strong>clustering algorithm</strong> which can be used for building phylognetic tree from <strong>distance matrices</strong>.<br />
</li>
<li>It still has some applications because it is fast, but it has mostly beenn replaced by Neighbor Joining and Minimum Evolution methods.</li>
<li>UPGMA is still useful for teaching the basic steps involved in data clustering and building trees.</li>
<li>UPGMA is a particular variant of a general class of algorithms. Another varient is WPGMA.</li>
</ul>
</div>
<div id="preliminaries-18" class="section level2" number="54.2">
<h2><span class="header-section-number">54.2</span> Preliminaries</h2>
<div id="r-libraries" class="section level3" number="54.2.1">
<h3><span class="header-section-number">54.2.1</span> R Libraries</h3>
<p>The phangorn package contains the functions</p>
<ul>
<li>upgma()</li>
<li>wpgma()</li>
</ul>
<p>The ape package has the function nj() for neighbor joinging, which is useful for comparison to UPGMA.</p>
<div id="phylogenetic-libraries" class="section level4" number="54.2.1.1">
<h4><span class="header-section-number">54.2.1.1</span> Phylogenetic libraries</h4>
<div class="sourceCode" id="cb1053"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1053-1"><a href="upgma-the-hard-way.html#cb1053-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages(&quot;ape&quot;)</span></span>
<span id="cb1053-2"><a href="upgma-the-hard-way.html#cb1053-2" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages(&quot;phangorn&quot;)</span></span>
<span id="cb1053-3"><a href="upgma-the-hard-way.html#cb1053-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb1053-4"><a href="upgma-the-hard-way.html#cb1053-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phangorn)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;phangorn&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:igraph&#39;:
## 
##     diversity</code></pre>
</div>
<div id="plotting-libraries" class="section level4" number="54.2.1.2">
<h4><span class="header-section-number">54.2.1.2</span> Plotting libraries</h4>
<div class="sourceCode" id="cb1056"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1056-1"><a href="upgma-the-hard-way.html#cb1056-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages(&quot;plotrix&quot;)</span></span>
<span id="cb1056-2"><a href="upgma-the-hard-way.html#cb1056-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plotrix)</span></code></pre></div>
</div>
</div>
<div id="custom-function-well-use" class="section level3" number="54.2.2">
<h3><span class="header-section-number">54.2.2</span> Custom function we’ll use</h3>
<p>To help us understand clustering and UPGMA I’ve written a cusome function.</p>
<p>This function, plot_dist_as_nmds(), will make a 2-dimensional (2D) representation of the distance matrices we’ll be working with. This is only an approximation and will occassionally result in distorions.</p>
<p>What this function does is take a <strong>distance matix</strong> and attempt to plot the data as if they were in 2-dimensional space. If we were giving it the distance between building on campus it would plot a reasonable map. Since we’ll be plotting data that isn’t physical distance but genetic distances this will not be a real evolutionary tree. The intent here is to help you build your intuition about what a distane matrix is.</p>
<div class="sourceCode" id="cb1057"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1057-1"><a href="upgma-the-hard-way.html#cb1057-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Load ALL of this code</span></span>
<span id="cb1057-2"><a href="upgma-the-hard-way.html#cb1057-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1057-3"><a href="upgma-the-hard-way.html#cb1057-3" aria-hidden="true" tabindex="-1"></a><span class="co"># </span><span class="al">TODO</span><span class="co">: ylim and xlim are current hard coded</span></span>
<span id="cb1057-4"><a href="upgma-the-hard-way.html#cb1057-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1057-5"><a href="upgma-the-hard-way.html#cb1057-5" aria-hidden="true" tabindex="-1"></a>plot_dist_as_nmds <span class="ot">&lt;-</span> <span class="cf">function</span>(mat, <span class="at">x.lim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">20</span>,<span class="dv">20</span>), <span class="at">y.lim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">20</span>,<span class="dv">20</span>)){</span>
<span id="cb1057-6"><a href="upgma-the-hard-way.html#cb1057-6" aria-hidden="true" tabindex="-1"></a>  matno0 <span class="ot">&lt;-</span> mat</span>
<span id="cb1057-7"><a href="upgma-the-hard-way.html#cb1057-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1057-8"><a href="upgma-the-hard-way.html#cb1057-8" aria-hidden="true" tabindex="-1"></a>matno0[<span class="fu">is.na</span>(matno0)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb1057-9"><a href="upgma-the-hard-way.html#cb1057-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1057-10"><a href="upgma-the-hard-way.html#cb1057-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1057-11"><a href="upgma-the-hard-way.html#cb1057-11" aria-hidden="true" tabindex="-1"></a>d2  <span class="ot">&lt;-</span> stats<span class="sc">::</span><span class="fu">cmdscale</span>(matno0, <span class="at">add =</span>T)</span>
<span id="cb1057-12"><a href="upgma-the-hard-way.html#cb1057-12" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> d2<span class="sc">$</span>points[,<span class="dv">1</span>]</span>
<span id="cb1057-13"><a href="upgma-the-hard-way.html#cb1057-13" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> d2<span class="sc">$</span>points[,<span class="dv">2</span>]</span>
<span id="cb1057-14"><a href="upgma-the-hard-way.html#cb1057-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x,y, <span class="at">main =</span> <span class="st">&quot;APPROXIMATE 2D distances&quot;</span>,</span>
<span id="cb1057-15"><a href="upgma-the-hard-way.html#cb1057-15" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlim =</span> x.lim, </span>
<span id="cb1057-16"><a href="upgma-the-hard-way.html#cb1057-16" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylim =</span> y.lim)</span>
<span id="cb1057-17"><a href="upgma-the-hard-way.html#cb1057-17" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(x, y, <span class="at">labels =</span> <span class="fu">row.names</span>(d2<span class="sc">$</span>points), <span class="at">cex=</span><span class="dv">3</span>)</span>
<span id="cb1057-18"><a href="upgma-the-hard-way.html#cb1057-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1057-19"><a href="upgma-the-hard-way.html#cb1057-19" aria-hidden="true" tabindex="-1"></a><span class="fu">return</span>(d2)</span>
<span id="cb1057-20"><a href="upgma-the-hard-way.html#cb1057-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1057-21"><a href="upgma-the-hard-way.html#cb1057-21" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
</div>
<div id="data-matrix" class="section level2" number="54.3">
<h2><span class="header-section-number">54.3</span> Data matrix</h2>
<p>We’ll use data from a study that built a tree using rRNA from bacteria.</p>
<p>The data and a worked example are available at <a href="https://en.wikipedia.org/wiki/UPGMA#First_step" class="uri">https://en.wikipedia.org/wiki/UPGMA#First_step</a></p>
<p>This particular matrix is set up as the number of bases that are different between each sequence. (In the next unit we’ll discuss how they didn’t use just a simple count of the differences).</p>
<div id="building-the-matrix" class="section level3" number="54.3.1">
<h3><span class="header-section-number">54.3.1</span> Building the matrix</h3>
<p>One way to make the data is to use the rbind() individual vectors as rows</p>
<div class="sourceCode" id="cb1058"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1058-1"><a href="upgma-the-hard-way.html#cb1058-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The data</span></span>
<span id="cb1058-2"><a href="upgma-the-hard-way.html#cb1058-2" aria-hidden="true" tabindex="-1"></a><span class="co">#      a   b   c   d   e</span></span>
<span id="cb1058-3"><a href="upgma-the-hard-way.html#cb1058-3" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>,    <span class="dv">17</span>, <span class="dv">21</span>, <span class="dv">31</span>, <span class="dv">23</span>)</span>
<span id="cb1058-4"><a href="upgma-the-hard-way.html#cb1058-4" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">17</span>, <span class="dv">0</span>,    <span class="dv">30</span>, <span class="dv">34</span>, <span class="dv">21</span>)</span>
<span id="cb1058-5"><a href="upgma-the-hard-way.html#cb1058-5" aria-hidden="true" tabindex="-1"></a>c   <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">30</span>, <span class="dv">0</span>,  <span class="dv">28</span>, <span class="dv">39</span>)</span>
<span id="cb1058-6"><a href="upgma-the-hard-way.html#cb1058-6" aria-hidden="true" tabindex="-1"></a>d   <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">31</span>, <span class="dv">34</span>, <span class="dv">28</span>,    <span class="dv">0</span>, <span class="dv">43</span>)</span>
<span id="cb1058-7"><a href="upgma-the-hard-way.html#cb1058-7" aria-hidden="true" tabindex="-1"></a>e   <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">23</span>, <span class="dv">21</span>, <span class="dv">39</span>, <span class="dv">43</span>,    <span class="dv">0</span>)</span>
<span id="cb1058-8"><a href="upgma-the-hard-way.html#cb1058-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1058-9"><a href="upgma-the-hard-way.html#cb1058-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Bind into a matrix</span></span>
<span id="cb1058-10"><a href="upgma-the-hard-way.html#cb1058-10" aria-hidden="true" tabindex="-1"></a>dmat1 <span class="ot">&lt;-</span> <span class="fu">rbind</span>(a,b,c,d,e)</span></code></pre></div>
<p>rbind() takes each vector (a through e) and glues them together into a matrix</p>
<div class="sourceCode" id="cb1059"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1059-1"><a href="upgma-the-hard-way.html#cb1059-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(dmat1)</span></code></pre></div>
<pre><code>## [1] &quot;matrix&quot;           &quot;array&quot;            &quot;mMatrix&quot;          &quot;structure&quot;       
## [5] &quot;vector&quot;           &quot;vector_OR_Vector&quot; &quot;vector_OR_factor&quot;</code></pre>
<p>As a challenge, can you take these data and make these data into a matrix using the matrix() function?</p>
<p>The finished matrix looks like this</p>
<div class="sourceCode" id="cb1061"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1061-1"><a href="upgma-the-hard-way.html#cb1061-1" aria-hidden="true" tabindex="-1"></a>dmat1</span></code></pre></div>
<pre><code>##   [,1] [,2] [,3] [,4] [,5]
## a    0   17   21   31   23
## b   17    0   30   34   21
## c   21   30    0   28   39
## d   31   34   28    0   43
## e   23   21   39   43    0</code></pre>
<p>Add names using colnames()</p>
<div class="sourceCode" id="cb1063"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1063-1"><a href="upgma-the-hard-way.html#cb1063-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dmat1) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;b&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;d&quot;</span>,<span class="st">&quot;e&quot;</span>)</span></code></pre></div>
<p>The matrix is <strong>symmetrical</strong>.</p>
<p><strong>Challenge questions:</strong> How would we interpret the diagonal? Why is it not 1? Why is this matrix symmetrical?</p>
<p>Its easier if you make redundant elements NA because we don’t need them.</p>
<p>We can access the <strong>diagonal</strong> of the matrix directly and set it to NA using the diag() function</p>
<div class="sourceCode" id="cb1064"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1064-1"><a href="upgma-the-hard-way.html#cb1064-1" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(dmat1) <span class="ot">&lt;-</span> <span class="cn">NA</span></span></code></pre></div>
<p>Now the matrix looks like this</p>
<div class="sourceCode" id="cb1065"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1065-1"><a href="upgma-the-hard-way.html#cb1065-1" aria-hidden="true" tabindex="-1"></a>dmat1</span></code></pre></div>
<pre><code>##    a  b  c  d  e
## a NA 17 21 31 23
## b 17 NA 30 34 21
## c 21 30 NA 28 39
## d 31 34 28 NA 43
## e 23 21 39 43 NA</code></pre>
<p>We only need the lower left-hand part of the matrix. We can set the upper right-hand part of the matrix to NA like this:</p>
<div class="sourceCode" id="cb1067"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1067-1"><a href="upgma-the-hard-way.html#cb1067-1" aria-hidden="true" tabindex="-1"></a>dmat1[<span class="fu">upper.tri</span>(dmat1)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span></code></pre></div>
<p>(don’t worry about exactly what’s going on here)</p>
<p>Matrix now looks like this</p>
<div class="sourceCode" id="cb1068"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1068-1"><a href="upgma-the-hard-way.html#cb1068-1" aria-hidden="true" tabindex="-1"></a>dmat1</span></code></pre></div>
<pre><code>##    a  b  c  d  e
## a NA NA NA NA NA
## b 17 NA NA NA NA
## c 21 30 NA NA NA
## d 31 34 28 NA NA
## e 23 21 39 43 NA</code></pre>
</div>
</div>
<div id="what-does-this-distance-matrix-represent" class="section level2" number="54.4">
<h2><span class="header-section-number">54.4</span> What does this distance matrix represent?</h2>
<div id="genetic-distances-conceptualized-as-pairwise-linear-distances" class="section level3" number="54.4.1">
<h3><span class="header-section-number">54.4.1</span> Genetic distances conceptualized as pairwise linear distances</h3>
<ul>
<li>Each element of the matrix is a <strong>pairwise distance.</strong><br />
</li>
<li>These distances were computed using <strong>pairwise alignments</strong> of each sequence.</li>
<li>In this case, it is the number of nucleotides that are different between two sequences.</li>
<li>As is standard, these alginments ignored any <strong>indels.</strong></li>
<li>There are 5 sequences, and if each is compared against the other then there are 10 total <strong>distances.</strong><br />
</li>
<li>Mathematically these are called <strong>Hamming distances</strong>.</li>
</ul>
<p>Each distance is the distance between two species, and it can be conceptualized on its own as a linear distance.</p>
<p>You could visualize a single <strong>pairwise distance</strong> like a linear distance like this:</p>
<p>(note: you don’t need to understand all of this code, but you do need to understand how it relates to the distance amtrix)</p>
<div class="sourceCode" id="cb1070"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1070-1"><a href="upgma-the-hard-way.html#cb1070-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb1070-2"><a href="upgma-the-hard-way.html#cb1070-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fu">max</span>(dmat1,<span class="at">na.rm =</span> T)), </span>
<span id="cb1070-3"><a href="upgma-the-hard-way.html#cb1070-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">12</span>), </span>
<span id="cb1070-4"><a href="upgma-the-hard-way.html#cb1070-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">type =</span> <span class="st">&quot;l&quot;</span>,</span>
<span id="cb1070-5"><a href="upgma-the-hard-way.html#cb1070-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="fu">c</span>(<span class="st">&quot;Distance (number of mutations)&quot;</span>),</span>
<span id="cb1070-6"><a href="upgma-the-hard-way.html#cb1070-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">&quot;Pairwise comparison&quot;</span>,</span>
<span id="cb1070-7"><a href="upgma-the-hard-way.html#cb1070-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">&quot;Linear distance a vs. b&quot;</span>)</span>
<span id="cb1070-8"><a href="upgma-the-hard-way.html#cb1070-8" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">pch =</span> <span class="st">&quot;a&quot;</span>)</span>
<span id="cb1070-9"><a href="upgma-the-hard-way.html#cb1070-9" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(dmat1[<span class="dv">2</span>,<span class="dv">1</span>], <span class="dv">1</span>, <span class="at">pch =</span> <span class="st">&quot;b&quot;</span>)</span>
<span id="cb1070-10"><a href="upgma-the-hard-way.html#cb1070-10" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="dv">0</span>, <span class="at">y0 =</span> <span class="dv">1</span>, </span>
<span id="cb1070-11"><a href="upgma-the-hard-way.html#cb1070-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> dmat1[<span class="dv">2</span>,<span class="dv">1</span>],<span class="at">y1 =</span> <span class="dv">1</span>)</span>
<span id="cb1070-12"><a href="upgma-the-hard-way.html#cb1070-12" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="fl">1.5</span>, <span class="at">col =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-715-1.png" width="672" /></p>
<p>We can look at all of the distance Note that each line is <em>independent</em> and cannot be compared (the y axis has no meaning). This is NOT a 2D map!</p>
<div class="sourceCode" id="cb1071"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1071-1"><a href="upgma-the-hard-way.html#cb1071-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This code is complex and the goal is to assess the plot, not understand the code</span></span>
<span id="cb1071-2"><a href="upgma-the-hard-way.html#cb1071-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fu">max</span>(dmat1,<span class="at">na.rm =</span> T)), </span>
<span id="cb1071-3"><a href="upgma-the-hard-way.html#cb1071-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">12</span>), </span>
<span id="cb1071-4"><a href="upgma-the-hard-way.html#cb1071-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">type =</span> <span class="st">&quot;l&quot;</span>,</span>
<span id="cb1071-5"><a href="upgma-the-hard-way.html#cb1071-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="fu">c</span>(<span class="st">&quot;Distance (number of mutations)&quot;</span>),</span>
<span id="cb1071-6"><a href="upgma-the-hard-way.html#cb1071-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">&quot;Pairwise comparison&quot;</span>)</span>
<span id="cb1071-7"><a href="upgma-the-hard-way.html#cb1071-7" aria-hidden="true" tabindex="-1"></a>pair.k <span class="ot">&lt;-</span><span class="dv">0</span></span>
<span id="cb1071-8"><a href="upgma-the-hard-way.html#cb1071-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dmat1)){</span>
<span id="cb1071-9"><a href="upgma-the-hard-way.html#cb1071-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dmat1)){</span>
<span id="cb1071-10"><a href="upgma-the-hard-way.html#cb1071-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1071-11"><a href="upgma-the-hard-way.html#cb1071-11" aria-hidden="true" tabindex="-1"></a>dist.ij <span class="ot">&lt;-</span> dmat1[i,j] </span>
<span id="cb1071-12"><a href="upgma-the-hard-way.html#cb1071-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.na</span>(dist.ij) <span class="sc">==</span> <span class="cn">FALSE</span>){</span>
<span id="cb1071-13"><a href="upgma-the-hard-way.html#cb1071-13" aria-hidden="true" tabindex="-1"></a>  pair.k <span class="ot">&lt;-</span> pair.k<span class="sc">+</span><span class="dv">1</span>    </span>
<span id="cb1071-14"><a href="upgma-the-hard-way.html#cb1071-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1071-15"><a href="upgma-the-hard-way.html#cb1071-15" aria-hidden="true" tabindex="-1"></a>       </span>
<span id="cb1071-16"><a href="upgma-the-hard-way.html#cb1071-16" aria-hidden="true" tabindex="-1"></a>spp.i <span class="ot">&lt;-</span> <span class="fu">rownames</span>(dmat1)[i]</span>
<span id="cb1071-17"><a href="upgma-the-hard-way.html#cb1071-17" aria-hidden="true" tabindex="-1"></a>spp.j <span class="ot">&lt;-</span> <span class="fu">colnames</span>(dmat1)[j]</span>
<span id="cb1071-18"><a href="upgma-the-hard-way.html#cb1071-18" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="sc">-</span><span class="fl">0.5</span>,               pair.k, <span class="at">pch =</span> spp.j, <span class="at">col =</span> <span class="fu">which</span>(LETTERS <span class="sc">%in%</span> <span class="fu">toupper</span>(spp.j)), <span class="at">cex =</span> <span class="fl">1.75</span>)</span>
<span id="cb1071-19"><a href="upgma-the-hard-way.html#cb1071-19" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(dist.ij<span class="fl">+0.75</span>,       pair.k, <span class="at">pch =</span>spp.i, <span class="at">col =</span> <span class="fu">which</span>(LETTERS <span class="sc">%in%</span> <span class="fu">toupper</span>(spp.i)), <span class="at">cex =</span> <span class="fl">1.75</span>)</span>
<span id="cb1071-20"><a href="upgma-the-hard-way.html#cb1071-20" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="dv">0</span>, <span class="at">y0 =</span> pair.k, </span>
<span id="cb1071-21"><a href="upgma-the-hard-way.html#cb1071-21" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> dist.ij, <span class="at">y1 =</span> pair.k)</span>
<span id="cb1071-22"><a href="upgma-the-hard-way.html#cb1071-22" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> pair.k<span class="fl">+0.5</span>, <span class="at">col =</span> <span class="dv">2</span>)</span>
<span id="cb1071-23"><a href="upgma-the-hard-way.html#cb1071-23" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1071-24"><a href="upgma-the-hard-way.html#cb1071-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1071-25"><a href="upgma-the-hard-way.html#cb1071-25" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1071-26"><a href="upgma-the-hard-way.html#cb1071-26" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-716-1.png" width="672" /></p>
<ul>
<li>We can see that the distance between a and b is the smallest, at 17 amino acid changes.</li>
<li>UPGMA, WPGMA and related algorithms all begin by identifying the two taxa that are closest together.</li>
<li>These 2 taxa form a clade (cluster). THe algorithm then proceeds by finding the next taxa (eg, c, d or e) which is closest to that clade of a and b</li>
<li>We can summarize this first clad using <strong>Newick notation</strong> as: (a, b)</li>
</ul>
</div>
<div id="genetic-distances-conceptualized-as-2d-map-of-linear-distances" class="section level3" number="54.4.2">
<h3><span class="header-section-number">54.4.2</span> Genetic distances conceptualized as 2D map of linear distances</h3>
<ul>
<li>We can extend the analogy of linear distance to two dimensions.<br />
</li>
<li>With some fancy math we can make a 2D plot where the distances in 2D space are <em>similar</em> to the distances in the matrix.<br />
</li>
<li>This is NOT exact and should not be taken literally. (See Pevsner Chapter 7 for a similar use of a 2D map)</li>
<li>This is just a tool for teaching and not something done as part of an actual analysis.</li>
</ul>
<div class="sourceCode" id="cb1072"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1072-1"><a href="upgma-the-hard-way.html#cb1072-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb1072-2"><a href="upgma-the-hard-way.html#cb1072-2" aria-hidden="true" tabindex="-1"></a>dmat1no0 <span class="ot">&lt;-</span> dmat1</span>
<span id="cb1072-3"><a href="upgma-the-hard-way.html#cb1072-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1072-4"><a href="upgma-the-hard-way.html#cb1072-4" aria-hidden="true" tabindex="-1"></a>dmat1no0[<span class="fu">is.na</span>(dmat1no0)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb1072-5"><a href="upgma-the-hard-way.html#cb1072-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1072-6"><a href="upgma-the-hard-way.html#cb1072-6" aria-hidden="true" tabindex="-1"></a>d2  <span class="ot">&lt;-</span> <span class="fu">cmdscale</span>(dmat1no0, <span class="at">add =</span>T)</span>
<span id="cb1072-7"><a href="upgma-the-hard-way.html#cb1072-7" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> d2<span class="sc">$</span>points[,<span class="dv">1</span>]</span>
<span id="cb1072-8"><a href="upgma-the-hard-way.html#cb1072-8" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> d2<span class="sc">$</span>points[,<span class="dv">2</span>]</span>
<span id="cb1072-9"><a href="upgma-the-hard-way.html#cb1072-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x,y, <span class="at">main =</span> <span class="st">&quot;APPROXIMATE 2D distances&quot;</span>,<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">20</span>,<span class="dv">20</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">20</span>,<span class="dv">20</span>))</span>
<span id="cb1072-10"><a href="upgma-the-hard-way.html#cb1072-10" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(x, y, <span class="at">labels =</span> <span class="fu">row.names</span>(d2<span class="sc">$</span>points), <span class="at">cex=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-717-1.png" width="672" /></p>
<ul>
<li>In this plot, things appear to have been distorted a bit.</li>
<li>a and b are close to each other as expect,</li>
<li>but b looks closest than e (to me).<br />
</li>
<li>It does show how a and b are close, and d and e are further apart.</li>
</ul>
</div>
</div>
<div id="from-linear-distances-to-tree-distances" class="section level2" number="54.5">
<h2><span class="header-section-number">54.5</span> From linear distances to “tree distances”</h2>
<p><strong>Genetic distances</strong> aren’t actually linear - we are assuming a “tree shape” to represent the divergence of two species genetically (and possibly geographically).</p>
<p>The simplest methods of building trees assume that all branch lengths terminate at the same distance from the root, and the total mutational distance traversed by each species is the same. (This is the assumption or constraint related to <strong>ultrametricity</strong>).</p>
<p>The distance between a and b is 17. We assume that both species a and species b have undergone the same number of mutations since they split, so the branch lengths for each are 17/2 = 8.5. (You obviously can’t have 8.5 mutations, but this is how the math works).</p>
<p>The following plots shows how the linear distnace of 17 between a and b (bottom) gets split into two branches, each of 8.5.</p>
<div class="sourceCode" id="cb1073"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1073-1"><a href="upgma-the-hard-way.html#cb1073-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This code is dense; run it and focus on the plot</span></span>
<span id="cb1073-2"><a href="upgma-the-hard-way.html#cb1073-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb1073-3"><a href="upgma-the-hard-way.html#cb1073-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fu">max</span>(dmat1,<span class="at">na.rm =</span> T)), </span>
<span id="cb1073-4"><a href="upgma-the-hard-way.html#cb1073-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">12</span>), </span>
<span id="cb1073-5"><a href="upgma-the-hard-way.html#cb1073-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">type =</span> <span class="st">&quot;l&quot;</span>,</span>
<span id="cb1073-6"><a href="upgma-the-hard-way.html#cb1073-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="fu">c</span>(<span class="st">&quot;Distance (number of mutations)&quot;</span>),</span>
<span id="cb1073-7"><a href="upgma-the-hard-way.html#cb1073-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">&quot;Pairwise comparison&quot;</span>)</span>
<span id="cb1073-8"><a href="upgma-the-hard-way.html#cb1073-8" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">pch =</span> <span class="st">&quot;a&quot;</span>)</span>
<span id="cb1073-9"><a href="upgma-the-hard-way.html#cb1073-9" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(dmat1[<span class="dv">2</span>,<span class="dv">1</span>], <span class="dv">1</span>, <span class="at">pch =</span> <span class="st">&quot;b&quot;</span>)</span>
<span id="cb1073-10"><a href="upgma-the-hard-way.html#cb1073-10" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="dv">0</span>, <span class="at">y0 =</span> <span class="dv">1</span>, </span>
<span id="cb1073-11"><a href="upgma-the-hard-way.html#cb1073-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> dmat1[<span class="dv">2</span>,<span class="dv">1</span>],<span class="at">y1 =</span> <span class="dv">1</span>)</span>
<span id="cb1073-12"><a href="upgma-the-hard-way.html#cb1073-12" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="fl">1.5</span>, <span class="at">col =</span> <span class="dv">2</span>)</span>
<span id="cb1073-13"><a href="upgma-the-hard-way.html#cb1073-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1073-14"><a href="upgma-the-hard-way.html#cb1073-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1073-15"><a href="upgma-the-hard-way.html#cb1073-15" aria-hidden="true" tabindex="-1"></a>branch.length <span class="ot">&lt;-</span> dmat1[<span class="dv">2</span>,<span class="dv">1</span>]<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb1073-16"><a href="upgma-the-hard-way.html#cb1073-16" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="dv">0</span>, <span class="dv">4</span>)</span>
<span id="cb1073-17"><a href="upgma-the-hard-way.html#cb1073-17" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(branch.length<span class="fl">+0.75</span>, <span class="dv">4</span>, <span class="at">pch =</span> <span class="st">&quot;a&quot;</span>)</span>
<span id="cb1073-18"><a href="upgma-the-hard-way.html#cb1073-18" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="dv">0</span>, <span class="dv">6</span>)</span>
<span id="cb1073-19"><a href="upgma-the-hard-way.html#cb1073-19" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(branch.length<span class="fl">+0.75</span>, <span class="dv">6</span>, <span class="at">pch =</span> <span class="st">&quot;b&quot;</span>)</span>
<span id="cb1073-20"><a href="upgma-the-hard-way.html#cb1073-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1073-21"><a href="upgma-the-hard-way.html#cb1073-21" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="dv">0</span>, <span class="at">y0 =</span> <span class="dv">4</span>, </span>
<span id="cb1073-22"><a href="upgma-the-hard-way.html#cb1073-22" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> branch.length,<span class="at">y1 =</span> <span class="dv">4</span>)</span>
<span id="cb1073-23"><a href="upgma-the-hard-way.html#cb1073-23" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="dv">0</span>, <span class="at">y0 =</span> <span class="dv">6</span>, </span>
<span id="cb1073-24"><a href="upgma-the-hard-way.html#cb1073-24" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> branch.length,<span class="at">y1 =</span> <span class="dv">6</span>)</span>
<span id="cb1073-25"><a href="upgma-the-hard-way.html#cb1073-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1073-26"><a href="upgma-the-hard-way.html#cb1073-26" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="dv">0</span>, <span class="at">y0 =</span> <span class="dv">4</span>, </span>
<span id="cb1073-27"><a href="upgma-the-hard-way.html#cb1073-27" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> <span class="dv">0</span>, <span class="at">y1 =</span> <span class="dv">6</span>, <span class="at">col =</span> <span class="dv">1</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb1073-28"><a href="upgma-the-hard-way.html#cb1073-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1073-29"><a href="upgma-the-hard-way.html#cb1073-29" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">8</span>, <span class="at">col =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-718-1.png" width="672" /></p>
<p>Again, in <strong>Newick notation</strong>, we’d write (a, b) to indicate that a and b are a clade.</p>
</div>
<div id="the-upgma-algorithm" class="section level2" number="54.6">
<h2><span class="header-section-number">54.6</span> The UPGMA algorithm</h2>
<p><em>The UPGMA algorithm is described by Swofford et al (1996?, page 486) as below.<br />
</em> The word “cluster” means “taxa” in some cases and “clade” in others.<br />
* This is a bit tricky at first but once you get a sense for the whole algorithm it will make more sense.
* To keep track of everything we’ll use subscripts i and j, represented in plain text as e.g. d.ij for the distance (d) in row i and column j.</p>
<div id="step-1-find-the-2-taxa-that-are-currently-closest-together" class="section level3" number="54.6.1">
<h3><span class="header-section-number">54.6.1</span> Step 1: find the 2 taxa that are currently closest together</h3>
<ol style="list-style-type: decimal">
<li>“Given a matrix of <strong>pairwise distances</strong>, find the clusters (taxa) i and j such the d.ij is the minmum value in the table.” (d.ij is the distance between taxa i and taxa j)</li>
</ol>
<p>In our example, d.ij = d.ba = 17, with i = b and j = a.</p>
<div class="sourceCode" id="cb1074"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1074-1"><a href="upgma-the-hard-way.html#cb1074-1" aria-hidden="true" tabindex="-1"></a>dmat1[<span class="st">&quot;b&quot;</span>,<span class="st">&quot;a&quot;</span>]</span></code></pre></div>
<pre><code>## [1] 17</code></pre>
</div>
<div id="step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node" class="section level3" number="54.6.2">
<h3><span class="header-section-number">54.6.2</span> Step 2: calculate the branch lengths connect the taxa to their shared node</h3>
<ol start="2" style="list-style-type: decimal">
<li>“Define the depth of the branching between i and j (l.ij) to be d.ij/2” (“depth of branching” = <strong>branch length</strong>; so i is attached to a branch of d.ij/2 and j is attached to a separate branch of d.ij/2)</li>
</ol>
<p>d.ab/2 = 8.5</p>
</div>
<div id="step-3a-determine-if-the-algorithm-is-done" class="section level3" number="54.6.3">
<h3><span class="header-section-number">54.6.3</span> Step 3a: determine if the algorithm is done</h3>
<ol start="3" style="list-style-type: decimal">
<li>“If i and j were the last two clusters, the tree is complete”</li>
</ol>
</div>
<div id="step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u" class="section level3" number="54.6.4">
<h3><span class="header-section-number">54.6.4</span> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</h3>
<ol start="3" style="list-style-type: decimal">
<li>If there is more to do, “create a new cluster called u”. This new cluster is a clade of a and b will represent the combined features of a and b.</li>
</ol>
<p>What this means is that now that we’ve mapped taxa a and b into a clade we need to update our matrix to determine how far all the remaining taca (c, d and e) are from that clade.</p>
</div>
<div id="step-4-determine-the-distance-from-u-to-the-other-remaining-clusters." class="section level3" number="54.6.5">
<h3><span class="header-section-number">54.6.5</span> Step 4: Determine the distance from u to the other remaining clusters.</h3>
<ol start="4" style="list-style-type: decimal">
<li>“Define the distance from [the new cluster] u to each other [remaining] cluster (k, with k==i or j) [here cluster = clade in the first part of the sentence and taxa in the other] to be an <strong>average</strong> of the distnace d.ki and d.jk”. That is, calculate the distance from the new clade u to all the remaining taxa, which we’ll generically call “k”. The new distance will be called d.ku, for “distance from taxa k to clade u.”</li>
</ol>
<p>The exact calculation of the average can take different forms. This will be unpacked below.</p>
</div>
<div id="step-5-create-a-new-matrix" class="section level3" number="54.6.6">
<h3><span class="header-section-number">54.6.6</span> Step 5: Create a new matrix</h3>
<ol start="5" style="list-style-type: decimal">
<li>“Go back to step 1 with one less cluster; cluster i and j have been eliminated, and cluster u has been added.” The distance matrix is now shrunk with all distance related to i and j removed and replaced by information about u.</li>
</ol>
</div>
<div id="key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa" class="section level3" number="54.6.7">
<h3><span class="header-section-number">54.6.7</span> Key step: calculating the distance from a clade to the remaining taxa</h3>
<p>For step 4, UPGMA uses the equation below to calculate distances. This will require some notation:</p>
<ul>
<li>Ti = the number of taxa in cluster i. This is 1 if the “cluster” is a single taxa (eg. 1 species), or it can be &gt;1 if a previous iteration has created a clade.<br />
</li>
<li>Tj = the number of taxa in cluster j.</li>
<li>d.ku = the distance from k to u. k and u can be individual taxa OR clades created by the algorithm.</li>
</ul>
<p>The key quantity d.ku is calculated as:</p>
<p>d.ku = (Ti x d.ki + Tj x d.kj)/(Ti + Tj)</p>
<p>(You should know this equation AND you should be able to write R code to calcualte it)</p>
<p>You can think of this as a <strong>weighted mean</strong> where the <strong>weights</strong> are the number of taxa in each clade. Its a very annoying that the “U” in UPGMA means “un-weighted” even though there is, from one perspective weighting, but this is how the math people say we should think about it (the un-weighted part has to do with some other aspect of the math).</p>
<p>In the above outline of the algorithm, if i and j are the first two taxa grouped then the equation would be</p>
<p>d.ku = (Ti x d.ki + Tj x d.kj)/(Ti + Tj)
Ti = 1, Tj = 1, so
d.ku = (1 x d.ki + 1 x d.kj)/(1 + 1)
d.ku = (d.ki + d.kj)/2</p>
</div>
</div>
<div id="calculate-2d-representation" class="section level2" number="54.7">
<h2><span class="header-section-number">54.7</span> Calculate 2D representation</h2>
<p>We can think about this using a 2D representation.</p>
<p>First, plot the 2D map.</p>
<div class="sourceCode" id="cb1076"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1076-1"><a href="upgma-the-hard-way.html#cb1076-1" aria-hidden="true" tabindex="-1"></a>dmat1.out<span class="ot">&lt;-</span> <span class="fu">plot_dist_as_nmds</span>(dmat1)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-720-1.png" width="672" /></p>
<p>a and b are closest together, so we group them</p>
<p>I’ll show this in the code below; you don’t need to understand exactly how this code works.</p>
<div class="sourceCode" id="cb1077"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1077-1"><a href="upgma-the-hard-way.html#cb1077-1" aria-hidden="true" tabindex="-1"></a>dmat1.out<span class="ot">&lt;-</span> <span class="fu">plot_dist_as_nmds</span>(dmat1)</span>
<span id="cb1077-2"><a href="upgma-the-hard-way.html#cb1077-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1077-3"><a href="upgma-the-hard-way.html#cb1077-3" aria-hidden="true" tabindex="-1"></a><span class="co">#get x y coord of a and b</span></span>
<span id="cb1077-4"><a href="upgma-the-hard-way.html#cb1077-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1077-5"><a href="upgma-the-hard-way.html#cb1077-5" aria-hidden="true" tabindex="-1"></a>a.<span class="ot">&lt;-</span>dmat1.out<span class="sc">$</span>points[<span class="st">&quot;a&quot;</span>,]</span>
<span id="cb1077-6"><a href="upgma-the-hard-way.html#cb1077-6" aria-hidden="true" tabindex="-1"></a>b.<span class="ot">&lt;-</span>dmat1.out<span class="sc">$</span>points[<span class="st">&quot;b&quot;</span>,]</span>
<span id="cb1077-7"><a href="upgma-the-hard-way.html#cb1077-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1077-8"><a href="upgma-the-hard-way.html#cb1077-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_dist_as_nmds</span>(dmat1)</span></code></pre></div>
<pre><code>## $points
##         [,1]       [,2]
## a   6.606653  4.9400937
## b  12.421702 -6.5267890
## c  -9.532025 12.9665271
## d -17.839642 -9.0739515
## e  20.738283  0.4897976
## 
## $eig
## NULL
## 
## $x
## NULL
## 
## $ac
## [1] -7.934712e-15
## 
## $GOF
## [1] 0.6755273 0.9367305</code></pre>
<div class="sourceCode" id="cb1079"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1079-1"><a href="upgma-the-hard-way.html#cb1079-1" aria-hidden="true" tabindex="-1"></a><span class="fu">draw.ellipse</span>(<span class="at">x =</span> <span class="fu">mean</span>(<span class="fu">c</span>(a.[<span class="dv">1</span>],b.[<span class="dv">1</span>])), </span>
<span id="cb1079-2"><a href="upgma-the-hard-way.html#cb1079-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> <span class="fu">mean</span>(<span class="fu">c</span>(a.[<span class="dv">2</span>],b.[<span class="dv">2</span>])),</span>
<span id="cb1079-3"><a href="upgma-the-hard-way.html#cb1079-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">a =</span><span class="dv">3</span>, </span>
<span id="cb1079-4"><a href="upgma-the-hard-way.html#cb1079-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">b =</span> <span class="dv">10</span>, </span>
<span id="cb1079-5"><a href="upgma-the-hard-way.html#cb1079-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">angle =</span> <span class="dv">30</span></span>
<span id="cb1079-6"><a href="upgma-the-hard-way.html#cb1079-6" aria-hidden="true" tabindex="-1"></a>             )</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-721-1.png" width="672" /></p>
<p>Once we have created a cluster (clade) we can consider the distance from the members of clade to all the remaining points. For example, here’s the distnace from a and b to c</p>
<div class="sourceCode" id="cb1080"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1080-1"><a href="upgma-the-hard-way.html#cb1080-1" aria-hidden="true" tabindex="-1"></a><span class="co"># distance from ab to c</span></span>
<span id="cb1080-2"><a href="upgma-the-hard-way.html#cb1080-2" aria-hidden="true" tabindex="-1"></a>c.<span class="ot">&lt;-</span>dmat1.out<span class="sc">$</span>points[<span class="st">&quot;c&quot;</span>,]</span>
<span id="cb1080-3"><a href="upgma-the-hard-way.html#cb1080-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1080-4"><a href="upgma-the-hard-way.html#cb1080-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_dist_as_nmds</span>(dmat1)</span></code></pre></div>
<pre><code>## $points
##         [,1]       [,2]
## a   6.606653  4.9400937
## b  12.421702 -6.5267890
## c  -9.532025 12.9665271
## d -17.839642 -9.0739515
## e  20.738283  0.4897976
## 
## $eig
## NULL
## 
## $x
## NULL
## 
## $ac
## [1] -7.934712e-15
## 
## $GOF
## [1] 0.6755273 0.9367305</code></pre>
<div class="sourceCode" id="cb1082"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1082-1"><a href="upgma-the-hard-way.html#cb1082-1" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> a.[<span class="dv">1</span>],<span class="at">x1 =</span> c.[<span class="dv">1</span>],</span>
<span id="cb1082-2"><a href="upgma-the-hard-way.html#cb1082-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> a.[<span class="dv">2</span>],<span class="at">y1 =</span> c.[<span class="dv">2</span>])</span>
<span id="cb1082-3"><a href="upgma-the-hard-way.html#cb1082-3" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> b.[<span class="dv">1</span>],<span class="at">x1 =</span> c.[<span class="dv">1</span>],</span>
<span id="cb1082-4"><a href="upgma-the-hard-way.html#cb1082-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> b.[<span class="dv">2</span>],<span class="at">y1 =</span> c.[<span class="dv">2</span>])                                    </span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-722-1.png" width="672" /></p>
<p>In UPGMA, we first define a clade. We then need to see which taxa is closest to that clade. Visually it looks like this</p>
<div class="sourceCode" id="cb1083"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1083-1"><a href="upgma-the-hard-way.html#cb1083-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb1083-2"><a href="upgma-the-hard-way.html#cb1083-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1083-3"><a href="upgma-the-hard-way.html#cb1083-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ab to c</span></span>
<span id="cb1083-4"><a href="upgma-the-hard-way.html#cb1083-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_dist_as_nmds</span>(dmat1)</span></code></pre></div>
<pre><code>## $points
##         [,1]       [,2]
## a   6.606653  4.9400937
## b  12.421702 -6.5267890
## c  -9.532025 12.9665271
## d -17.839642 -9.0739515
## e  20.738283  0.4897976
## 
## $eig
## NULL
## 
## $x
## NULL
## 
## $ac
## [1] -7.934712e-15
## 
## $GOF
## [1] 0.6755273 0.9367305</code></pre>
<div class="sourceCode" id="cb1085"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1085-1"><a href="upgma-the-hard-way.html#cb1085-1" aria-hidden="true" tabindex="-1"></a><span class="co">#get x y coord of a and b</span></span>
<span id="cb1085-2"><a href="upgma-the-hard-way.html#cb1085-2" aria-hidden="true" tabindex="-1"></a>a.<span class="ot">&lt;-</span>dmat1.out<span class="sc">$</span>points[<span class="st">&quot;a&quot;</span>,]</span>
<span id="cb1085-3"><a href="upgma-the-hard-way.html#cb1085-3" aria-hidden="true" tabindex="-1"></a>b.<span class="ot">&lt;-</span>dmat1.out<span class="sc">$</span>points[<span class="st">&quot;b&quot;</span>,]</span>
<span id="cb1085-4"><a href="upgma-the-hard-way.html#cb1085-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1085-5"><a href="upgma-the-hard-way.html#cb1085-5" aria-hidden="true" tabindex="-1"></a><span class="fu">draw.ellipse</span>(<span class="at">x =</span> <span class="fu">mean</span>(<span class="fu">c</span>(a.[<span class="dv">1</span>],b.[<span class="dv">1</span>])), </span>
<span id="cb1085-6"><a href="upgma-the-hard-way.html#cb1085-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> <span class="fu">mean</span>(<span class="fu">c</span>(a.[<span class="dv">2</span>],b.[<span class="dv">2</span>])),</span>
<span id="cb1085-7"><a href="upgma-the-hard-way.html#cb1085-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">a =</span><span class="dv">3</span>, </span>
<span id="cb1085-8"><a href="upgma-the-hard-way.html#cb1085-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">b =</span> <span class="dv">10</span>, </span>
<span id="cb1085-9"><a href="upgma-the-hard-way.html#cb1085-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">angle =</span> <span class="dv">30</span>)</span>
<span id="cb1085-10"><a href="upgma-the-hard-way.html#cb1085-10" aria-hidden="true" tabindex="-1"></a>c.<span class="ot">&lt;-</span>d2<span class="sc">$</span>points[<span class="st">&quot;c&quot;</span>,]</span>
<span id="cb1085-11"><a href="upgma-the-hard-way.html#cb1085-11" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> a.[<span class="dv">1</span>],<span class="at">x1 =</span> c.[<span class="dv">1</span>],</span>
<span id="cb1085-12"><a href="upgma-the-hard-way.html#cb1085-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> a.[<span class="dv">2</span>],<span class="at">y1 =</span> c.[<span class="dv">2</span>])</span>
<span id="cb1085-13"><a href="upgma-the-hard-way.html#cb1085-13" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> b.[<span class="dv">1</span>],<span class="at">x1 =</span> c.[<span class="dv">1</span>],</span>
<span id="cb1085-14"><a href="upgma-the-hard-way.html#cb1085-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> b.[<span class="dv">2</span>],<span class="at">y1 =</span> c.[<span class="dv">2</span>])  </span>
<span id="cb1085-15"><a href="upgma-the-hard-way.html#cb1085-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1085-16"><a href="upgma-the-hard-way.html#cb1085-16" aria-hidden="true" tabindex="-1"></a><span class="co"># ab to d</span></span>
<span id="cb1085-17"><a href="upgma-the-hard-way.html#cb1085-17" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_dist_as_nmds</span>(dmat1)</span></code></pre></div>
<pre><code>## $points
##         [,1]       [,2]
## a   6.606653  4.9400937
## b  12.421702 -6.5267890
## c  -9.532025 12.9665271
## d -17.839642 -9.0739515
## e  20.738283  0.4897976
## 
## $eig
## NULL
## 
## $x
## NULL
## 
## $ac
## [1] -7.934712e-15
## 
## $GOF
## [1] 0.6755273 0.9367305</code></pre>
<div class="sourceCode" id="cb1087"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1087-1"><a href="upgma-the-hard-way.html#cb1087-1" aria-hidden="true" tabindex="-1"></a><span class="fu">draw.ellipse</span>(<span class="at">x =</span> <span class="fu">mean</span>(<span class="fu">c</span>(a.[<span class="dv">1</span>],b.[<span class="dv">1</span>])), </span>
<span id="cb1087-2"><a href="upgma-the-hard-way.html#cb1087-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> <span class="fu">mean</span>(<span class="fu">c</span>(a.[<span class="dv">2</span>],b.[<span class="dv">2</span>])),</span>
<span id="cb1087-3"><a href="upgma-the-hard-way.html#cb1087-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">a =</span><span class="dv">3</span>, </span>
<span id="cb1087-4"><a href="upgma-the-hard-way.html#cb1087-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">b =</span> <span class="dv">10</span>, </span>
<span id="cb1087-5"><a href="upgma-the-hard-way.html#cb1087-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">angle =</span> <span class="dv">30</span>)</span>
<span id="cb1087-6"><a href="upgma-the-hard-way.html#cb1087-6" aria-hidden="true" tabindex="-1"></a>d.<span class="ot">&lt;-</span>d2<span class="sc">$</span>points[<span class="st">&quot;d&quot;</span>,]</span>
<span id="cb1087-7"><a href="upgma-the-hard-way.html#cb1087-7" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> a.[<span class="dv">1</span>],<span class="at">x1 =</span> d.[<span class="dv">1</span>],</span>
<span id="cb1087-8"><a href="upgma-the-hard-way.html#cb1087-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> a.[<span class="dv">2</span>],<span class="at">y1 =</span> d.[<span class="dv">2</span>])</span>
<span id="cb1087-9"><a href="upgma-the-hard-way.html#cb1087-9" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> b.[<span class="dv">1</span>],<span class="at">x1 =</span> d.[<span class="dv">1</span>],</span>
<span id="cb1087-10"><a href="upgma-the-hard-way.html#cb1087-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> b.[<span class="dv">2</span>],<span class="at">y1 =</span> d.[<span class="dv">2</span>])  </span>
<span id="cb1087-11"><a href="upgma-the-hard-way.html#cb1087-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1087-12"><a href="upgma-the-hard-way.html#cb1087-12" aria-hidden="true" tabindex="-1"></a><span class="co"># ab to e</span></span>
<span id="cb1087-13"><a href="upgma-the-hard-way.html#cb1087-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_dist_as_nmds</span>(dmat1)</span></code></pre></div>
<pre><code>## $points
##         [,1]       [,2]
## a   6.606653  4.9400937
## b  12.421702 -6.5267890
## c  -9.532025 12.9665271
## d -17.839642 -9.0739515
## e  20.738283  0.4897976
## 
## $eig
## NULL
## 
## $x
## NULL
## 
## $ac
## [1] -7.934712e-15
## 
## $GOF
## [1] 0.6755273 0.9367305</code></pre>
<div class="sourceCode" id="cb1089"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1089-1"><a href="upgma-the-hard-way.html#cb1089-1" aria-hidden="true" tabindex="-1"></a><span class="fu">draw.ellipse</span>(<span class="at">x =</span> <span class="fu">mean</span>(<span class="fu">c</span>(a.[<span class="dv">1</span>],b.[<span class="dv">1</span>])), </span>
<span id="cb1089-2"><a href="upgma-the-hard-way.html#cb1089-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> <span class="fu">mean</span>(<span class="fu">c</span>(a.[<span class="dv">2</span>],b.[<span class="dv">2</span>])),</span>
<span id="cb1089-3"><a href="upgma-the-hard-way.html#cb1089-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">a =</span><span class="dv">3</span>, </span>
<span id="cb1089-4"><a href="upgma-the-hard-way.html#cb1089-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">b =</span> <span class="dv">10</span>, </span>
<span id="cb1089-5"><a href="upgma-the-hard-way.html#cb1089-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">angle =</span> <span class="dv">30</span>)</span>
<span id="cb1089-6"><a href="upgma-the-hard-way.html#cb1089-6" aria-hidden="true" tabindex="-1"></a>e.<span class="ot">&lt;-</span>d2<span class="sc">$</span>points[<span class="st">&quot;e&quot;</span>,]</span>
<span id="cb1089-7"><a href="upgma-the-hard-way.html#cb1089-7" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> a.[<span class="dv">1</span>],<span class="at">x1 =</span> e.[<span class="dv">1</span>],</span>
<span id="cb1089-8"><a href="upgma-the-hard-way.html#cb1089-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> a.[<span class="dv">2</span>],<span class="at">y1 =</span> e.[<span class="dv">2</span>])</span>
<span id="cb1089-9"><a href="upgma-the-hard-way.html#cb1089-9" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> b.[<span class="dv">1</span>],<span class="at">x1 =</span> e.[<span class="dv">1</span>],</span>
<span id="cb1089-10"><a href="upgma-the-hard-way.html#cb1089-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> b.[<span class="dv">2</span>],<span class="at">y1 =</span> e.[<span class="dv">2</span>]) </span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-723-1.png" width="672" /></p>
</div>
<div id="algorithm---round-1" class="section level2" number="54.8">
<h2><span class="header-section-number">54.8</span> Algorithm - round 1</h2>
<p>We will now implement the algorithm.</p>
<p>NOTE: I am not using a consistent system of nomeclature - I am mixing my own conventions with hose of Swofford et al. Sorry :(</p>
<p>First we’ll need a dataframe to hold tings</p>
<ul>
<li>it = iteration</li>
<li>clust.i, clust.j = the two taxa or clades being combined</li>
<li>dist.ij = value from distnace matrix between the taxa</li>
<li>clust.u = name of the new combined cluster (clade)</li>
<li>branch.l.ij = branch lengths</li>
</ul>
<p>A matrix to hold things as I work</p>
<div class="sourceCode" id="cb1090"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1090-1"><a href="upgma-the-hard-way.html#cb1090-1" aria-hidden="true" tabindex="-1"></a>n.taxa <span class="ot">&lt;-</span> <span class="fu">nrow</span>(dmat1)</span>
<span id="cb1090-2"><a href="upgma-the-hard-way.html#cb1090-2" aria-hidden="true" tabindex="-1"></a>UPGMA_output <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">it =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span>n.taxa),</span>
<span id="cb1090-3"><a href="upgma-the-hard-way.html#cb1090-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">clust.i =</span> <span class="cn">NA</span>,</span>
<span id="cb1090-4"><a href="upgma-the-hard-way.html#cb1090-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">clust.j =</span> <span class="cn">NA</span>,</span>
<span id="cb1090-5"><a href="upgma-the-hard-way.html#cb1090-5" aria-hidden="true" tabindex="-1"></a>                           <span class="at">dist.ij =</span> <span class="cn">NA</span>,</span>
<span id="cb1090-6"><a href="upgma-the-hard-way.html#cb1090-6" aria-hidden="true" tabindex="-1"></a>                           <span class="at">clust.u =</span> <span class="cn">NA</span>,</span>
<span id="cb1090-7"><a href="upgma-the-hard-way.html#cb1090-7" aria-hidden="true" tabindex="-1"></a>                           <span class="at">branch.l.ij =</span> <span class="cn">NA</span>)</span></code></pre></div>
<div id="determine-minimum-value-in-current-matrix" class="section level3" number="54.8.1">
<h3><span class="header-section-number">54.8.1</span> Determine minimum value in current matrix</h3>
<p>The minimum distance using the min() function</p>
<div class="sourceCode" id="cb1091"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1091-1"><a href="upgma-the-hard-way.html#cb1091-1" aria-hidden="true" tabindex="-1"></a>dist.min.i <span class="ot">&lt;-</span> <span class="fu">min</span>(dmat1, <span class="at">na.rm =</span> T)</span></code></pre></div>
<p>The index value of the minimum, using which()</p>
<div class="sourceCode" id="cb1092"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1092-1"><a href="upgma-the-hard-way.html#cb1092-1" aria-hidden="true" tabindex="-1"></a>index.min.i <span class="ot">&lt;-</span> <span class="fu">which</span>(dmat1 <span class="sc">==</span> dist.min.i, <span class="at">arr.ind =</span>T)</span>
<span id="cb1092-2"><a href="upgma-the-hard-way.html#cb1092-2" aria-hidden="true" tabindex="-1"></a>index.row.i <span class="ot">&lt;-</span> index.min.i[<span class="dv">1</span>]</span>
<span id="cb1092-3"><a href="upgma-the-hard-way.html#cb1092-3" aria-hidden="true" tabindex="-1"></a>index.col.i <span class="ot">&lt;-</span> index.min.i[<span class="dv">2</span>]</span></code></pre></div>
<p>We can get the names of our taxa using rownames() and colnames()</p>
<div class="sourceCode" id="cb1093"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1093-1"><a href="upgma-the-hard-way.html#cb1093-1" aria-hidden="true" tabindex="-1"></a>cluster.i2 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(dmat1)[index.row.i]</span>
<span id="cb1093-2"><a href="upgma-the-hard-way.html#cb1093-2" aria-hidden="true" tabindex="-1"></a>cluster.i1 <span class="ot">&lt;-</span> <span class="fu">colnames</span>(dmat1)[index.col.i]</span></code></pre></div>
<p>Add output to dataframe</p>
<div class="sourceCode" id="cb1094"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1094-1"><a href="upgma-the-hard-way.html#cb1094-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the clusters we&#39;re working iwth</span></span>
<span id="cb1094-2"><a href="upgma-the-hard-way.html#cb1094-2" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">1</span>,<span class="st">&quot;clust.i&quot;</span>] <span class="ot">&lt;-</span> cluster.i1</span>
<span id="cb1094-3"><a href="upgma-the-hard-way.html#cb1094-3" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">1</span>,<span class="st">&quot;clust.j&quot;</span>] <span class="ot">&lt;-</span> cluster.i2</span>
<span id="cb1094-4"><a href="upgma-the-hard-way.html#cb1094-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1094-5"><a href="upgma-the-hard-way.html#cb1094-5" aria-hidden="true" tabindex="-1"></a><span class="co"># the distance between them</span></span>
<span id="cb1094-6"><a href="upgma-the-hard-way.html#cb1094-6" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">1</span>,<span class="st">&quot;dist.ij&quot;</span>] <span class="ot">&lt;-</span> dist.min.i</span></code></pre></div>
</div>
<div id="combine-species-into-clade" class="section level3" number="54.8.2">
<h3><span class="header-section-number">54.8.2</span> Combine species into clade</h3>
<p>Combine the two taxa names into a new name. This can be done a couple ways.</p>
<div class="sourceCode" id="cb1095"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1095-1"><a href="upgma-the-hard-way.html#cb1095-1" aria-hidden="true" tabindex="-1"></a><span class="co"># name of our clade</span></span>
<span id="cb1095-2"><a href="upgma-the-hard-way.html#cb1095-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1095-3"><a href="upgma-the-hard-way.html#cb1095-3" aria-hidden="true" tabindex="-1"></a><span class="do">## hard-code &quot;ab&quot;</span></span>
<span id="cb1095-4"><a href="upgma-the-hard-way.html#cb1095-4" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="st">&quot;ab&quot;</span></span>
<span id="cb1095-5"><a href="upgma-the-hard-way.html#cb1095-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1095-6"><a href="upgma-the-hard-way.html#cb1095-6" aria-hidden="true" tabindex="-1"></a><span class="do">## use paste() to generate name</span></span>
<span id="cb1095-7"><a href="upgma-the-hard-way.html#cb1095-7" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;b&quot;</span>,<span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1095-8"><a href="upgma-the-hard-way.html#cb1095-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1095-9"><a href="upgma-the-hard-way.html#cb1095-9" aria-hidden="true" tabindex="-1"></a><span class="do">## general, re-usable code using paste() and object names</span></span>
<span id="cb1095-10"><a href="upgma-the-hard-way.html#cb1095-10" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="fu">paste</span>(cluster.i1,cluster.i2, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>We’ll add this to the dataframe</p>
<div class="sourceCode" id="cb1096"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1096-1"><a href="upgma-the-hard-way.html#cb1096-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">1</span>,<span class="st">&quot;clust.u&quot;</span>] <span class="ot">&lt;-</span> clade.i</span></code></pre></div>
<p>Look at things so far</p>
<div class="sourceCode" id="cb1097"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1097-1"><a href="upgma-the-hard-way.html#cb1097-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output</span></code></pre></div>
<pre><code>##   it clust.i clust.j dist.ij clust.u branch.l.ij
## 1  1       a       b      17      ab          NA
## 2  2    &lt;NA&gt;    &lt;NA&gt;      NA    &lt;NA&gt;          NA
## 3  3    &lt;NA&gt;    &lt;NA&gt;      NA    &lt;NA&gt;          NA
## 4  4    &lt;NA&gt;    &lt;NA&gt;      NA    &lt;NA&gt;          NA
## 5  5    &lt;NA&gt;    &lt;NA&gt;      NA    &lt;NA&gt;          NA</code></pre>
</div>
<div id="calculate-branch-length" class="section level3" number="54.8.3">
<h3><span class="header-section-number">54.8.3</span> Calculate branch length</h3>
<p>Branch length is the distance between the two taxa: distance/2 or d.ij/2</p>
<p>We can access the distance in our first row of the matrix like this</p>
<div class="sourceCode" id="cb1099"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1099-1"><a href="upgma-the-hard-way.html#cb1099-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">1</span>,<span class="st">&quot;dist.ij&quot;</span>]</span></code></pre></div>
<pre><code>## [1] 17</code></pre>
<p>We can divided the distance betyween a and b by 2 to get our branch length like this</p>
<div class="sourceCode" id="cb1101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1101-1"><a href="upgma-the-hard-way.html#cb1101-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">1</span>,<span class="st">&quot;dist.ij&quot;</span>]<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
<pre><code>## [1] 8.5</code></pre>
<p>Now store the distance d in the dataframe</p>
<div class="sourceCode" id="cb1103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1103-1"><a href="upgma-the-hard-way.html#cb1103-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">1</span>,<span class="st">&quot;branch.l.ij&quot;</span>] <span class="ot">&lt;-</span> UPGMA_output[<span class="dv">1</span>,<span class="st">&quot;dist.ij&quot;</span>]<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
<p>We’ve now completed our first round of calculations</p>
<div class="sourceCode" id="cb1104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1104-1"><a href="upgma-the-hard-way.html#cb1104-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output</span></code></pre></div>
<pre><code>##   it clust.i clust.j dist.ij clust.u branch.l.ij
## 1  1       a       b      17      ab         8.5
## 2  2    &lt;NA&gt;    &lt;NA&gt;      NA    &lt;NA&gt;          NA
## 3  3    &lt;NA&gt;    &lt;NA&gt;      NA    &lt;NA&gt;          NA
## 4  4    &lt;NA&gt;    &lt;NA&gt;      NA    &lt;NA&gt;          NA
## 5  5    &lt;NA&gt;    &lt;NA&gt;      NA    &lt;NA&gt;          NA</code></pre>
</div>
<div id="calculate-distance-from-clade-to-all-other-points" class="section level3" number="54.8.4">
<h3><span class="header-section-number">54.8.4</span> Calculate distance from clade to all other points</h3>
<p>Distance (ab) to all other points</p>
<ul>
<li>ab to c</li>
<li>ab to d</li>
<li>ab to e</li>
</ul>
<div id="ab-to-c" class="section level4" number="54.8.4.1">
<h4><span class="header-section-number">54.8.4.1</span> ab to c</h4>
<p>This is calculated at the average of the distance</p>
<ul>
<li>from a to c, and</li>
<li>from b to c.</li>
</ul>
<p>In the Wikipedia article the math is framed like this</p>
<p>D(ab to c) = [D(a to c) + D(b to c)] / 2
D(ab to c) = [ 21 + 30 ] / 2
D(ab to c) = [ 21 + 30 ] / 2
D(ab to c) = 25.5</p>
<p>Dab_c = [Da_c + Db_c] / 2
Dab_c = 25.5</p>
<p>The math in the wikipedia article is more accurately written out as this, which is more similar to Swofford’s notation.</p>
<p>D(ab to c) = [T.i<em>D(a to c) + T.j</em>D(b to c)] / 2
D(ab to c) = [1<em>D(a to c) + 1</em>D(b to c)] / (1+1)
D(ab to c) = [ 1<em>21 + 1</em>30 ] / (1+1)
D(ab to c) = [ 1<em>21 + 1</em>30 ] / (1+1)
D(ab to c) = 25.5</p>
<p>Dab_c = [1<em>Da_c + 1</em>Db_c] / (1+1)
Dab_c = 25.5</p>
<p>Using Swofford et al’s notation we’d do this:</p>
<p>d.ku = (Ti x d.ki + Tj x d.kj)/(Ti + Tj)</p>
<p>I’ll just switch the subscripts to be consistent with the Wikipedia notation</p>
<p>d.uk = (Ti x d.ik + Tj x d.jk)/(Ti + Tj)</p>
<p>where</p>
<p>u = our new cluster ab
k = the remaining clusters; we’ll start with c
Ti, Tj = the number of clusters the constitute the clusters that were just combined. This starts out at 1.
Ti = Ta = 1
Tj = Tb = 1
d.ik = the distance from cluster i to the reamining clusters; we’ll start with c
d.ik = d.ac = 21
d.jk = d.bc = 30</p>
<p>d.ku = (Ti x d.ik + Tj x d.jk)/(Ti + Tj)
d.ku = (1 x 21 + 1 x 30 )/( 1 + 1 )
d.ku = ( 21 + 30 )/(2)
d.ku = 51/2
d.ku = 25.5</p>
<p>The Swofford notation is more generic and useful because in the next iteration we’ll have a cluster of ab which will have two taxa in it.</p>
<div class="sourceCode" id="cb1106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1106-1"><a href="upgma-the-hard-way.html#cb1106-1" aria-hidden="true" tabindex="-1"></a>Ta <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb1106-2"><a href="upgma-the-hard-way.html#cb1106-2" aria-hidden="true" tabindex="-1"></a>Tb <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb1106-3"><a href="upgma-the-hard-way.html#cb1106-3" aria-hidden="true" tabindex="-1"></a>Da_c <span class="ot">&lt;-</span> <span class="dv">21</span></span>
<span id="cb1106-4"><a href="upgma-the-hard-way.html#cb1106-4" aria-hidden="true" tabindex="-1"></a>Db_c <span class="ot">&lt;-</span> <span class="dv">30</span></span>
<span id="cb1106-5"><a href="upgma-the-hard-way.html#cb1106-5" aria-hidden="true" tabindex="-1"></a>Dab_c <span class="ot">&lt;-</span> (Ta<span class="sc">*</span>Da_c<span class="sc">+</span>Tb<span class="sc">*</span>Db_c)<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
</div>
<div id="ab-to-d" class="section level4" number="54.8.4.2">
<h4><span class="header-section-number">54.8.4.2</span> ab to d</h4>
<p>We now continue for the other distances. I’ll stick to the wikipedia notation for now.</p>
<p>D(ab to d) = [D(a to d) + D(b to d)] / 2
D(ab to d) = [ 31 + 34 ] / 2
D(ab to d) = [ 31 + 34 ] / 2
D(ab to d) = 32.5</p>
<p>Dab_d = [Da_d + Db_d] / 2
Dab_d = 32.5</p>
<div class="sourceCode" id="cb1107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1107-1"><a href="upgma-the-hard-way.html#cb1107-1" aria-hidden="true" tabindex="-1"></a>Da_d <span class="ot">&lt;-</span> <span class="dv">31</span></span>
<span id="cb1107-2"><a href="upgma-the-hard-way.html#cb1107-2" aria-hidden="true" tabindex="-1"></a>Db_d <span class="ot">&lt;-</span> <span class="dv">34</span></span>
<span id="cb1107-3"><a href="upgma-the-hard-way.html#cb1107-3" aria-hidden="true" tabindex="-1"></a>Dab_d <span class="ot">&lt;-</span> (Da_d<span class="sc">+</span>Db_d)<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
</div>
<div id="ab-to-e" class="section level4" number="54.8.4.3">
<h4><span class="header-section-number">54.8.4.3</span> ab to e</h4>
<p>D(ab to d) = [D(a to d) + D(b to d)] / 2
D(ab to d) = [ 23 + 21 ] / 2
D(ab to d) = [ 23 + 21 ] / 2
D(ab to d) = Dab_e</p>
<p>Dab_d = [Da_d + Db_d] / 2
Dab_d = Dab_e</p>
<div class="sourceCode" id="cb1108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1108-1"><a href="upgma-the-hard-way.html#cb1108-1" aria-hidden="true" tabindex="-1"></a>Da_e <span class="ot">&lt;-</span> <span class="dv">23</span></span>
<span id="cb1108-2"><a href="upgma-the-hard-way.html#cb1108-2" aria-hidden="true" tabindex="-1"></a>Db_e <span class="ot">&lt;-</span> <span class="dv">21</span></span>
<span id="cb1108-3"><a href="upgma-the-hard-way.html#cb1108-3" aria-hidden="true" tabindex="-1"></a>Dab_e <span class="ot">&lt;-</span> (Da_e<span class="sc">+</span>Db_e)<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
</div>
</div>
<div id="update-matrix-with-new-distance" class="section level3" number="54.8.5">
<h3><span class="header-section-number">54.8.5</span> Update matrix with new distance</h3>
<p>Old matrix has distances from a and b to all other points: we want to replace these. distances among c, d, and e are still the same values.</p>
<div class="sourceCode" id="cb1109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1109-1"><a href="upgma-the-hard-way.html#cb1109-1" aria-hidden="true" tabindex="-1"></a>dmat1</span></code></pre></div>
<pre><code>##    a  b  c  d  e
## a NA NA NA NA NA
## b 17 NA NA NA NA
## c 21 30 NA NA NA
## d 31 34 28 NA NA
## e 23 21 39 43 NA</code></pre>
<div id="illustration" class="section level4" number="54.8.5.1">
<h4><span class="header-section-number">54.8.5.1</span> Illustration</h4>
<p>Values in parenthesese will be REPLACED</p>
<div class="sourceCode" id="cb1111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1111-1"><a href="upgma-the-hard-way.html#cb1111-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>)</span>
<span id="cb1111-2"><a href="upgma-the-hard-way.html#cb1111-2" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;(17)&quot;</span>, <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>)</span>
<span id="cb1111-3"><a href="upgma-the-hard-way.html#cb1111-3" aria-hidden="true" tabindex="-1"></a>c   <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;(21)&quot;</span>, <span class="st">&quot;(30)&quot;</span>, <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>)</span>
<span id="cb1111-4"><a href="upgma-the-hard-way.html#cb1111-4" aria-hidden="true" tabindex="-1"></a>d   <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;(31)&quot;</span>, <span class="st">&quot;(34)&quot;</span>, <span class="st">&quot;28&quot;</span>, <span class="cn">NA</span>, <span class="cn">NA</span>)</span>
<span id="cb1111-5"><a href="upgma-the-hard-way.html#cb1111-5" aria-hidden="true" tabindex="-1"></a>e   <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;(23)&quot;</span>, <span class="st">&quot;(21)&quot;</span>, <span class="st">&quot;39&quot;</span>, <span class="st">&quot;43&quot;</span>,    <span class="cn">NA</span>)</span>
<span id="cb1111-6"><a href="upgma-the-hard-way.html#cb1111-6" aria-hidden="true" tabindex="-1"></a>dmat1.alt <span class="ot">&lt;-</span> <span class="fu">rbind</span>(a,b,c,d,e)</span>
<span id="cb1111-7"><a href="upgma-the-hard-way.html#cb1111-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dmat1.alt) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;b&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;d&quot;</span>,<span class="st">&quot;e&quot;</span>)</span>
<span id="cb1111-8"><a href="upgma-the-hard-way.html#cb1111-8" aria-hidden="true" tabindex="-1"></a>dmat1.alt</span></code></pre></div>
<pre><code>##   a      b      c    d    e 
## a NA     NA     NA   NA   NA
## b &quot;(17)&quot; NA     NA   NA   NA
## c &quot;(21)&quot; &quot;(30)&quot; NA   NA   NA
## d &quot;(31)&quot; &quot;(34)&quot; &quot;28&quot; NA   NA
## e &quot;(23)&quot; &quot;(21)&quot; &quot;39&quot; &quot;43&quot; NA</code></pre>
<p>Create smaller matrix with clade ab</p>
<div class="sourceCode" id="cb1113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1113-1"><a href="upgma-the-hard-way.html#cb1113-1" aria-hidden="true" tabindex="-1"></a><span class="co">#       ab      c   d   e</span></span>
<span id="cb1113-2"><a href="upgma-the-hard-way.html#cb1113-2" aria-hidden="true" tabindex="-1"></a>ab  <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">NA</span>,    <span class="cn">NA</span>,  <span class="cn">NA</span>, <span class="cn">NA</span>)</span>
<span id="cb1113-3"><a href="upgma-the-hard-way.html#cb1113-3" aria-hidden="true" tabindex="-1"></a>c     <span class="ot">&lt;-</span> <span class="fu">c</span>(Dab_c, <span class="cn">NA</span>,  <span class="cn">NA</span>, <span class="cn">NA</span>)</span>
<span id="cb1113-4"><a href="upgma-the-hard-way.html#cb1113-4" aria-hidden="true" tabindex="-1"></a>d     <span class="ot">&lt;-</span> <span class="fu">c</span>(Dab_d, <span class="dv">28</span>,    <span class="cn">NA</span>, <span class="cn">NA</span>)</span>
<span id="cb1113-5"><a href="upgma-the-hard-way.html#cb1113-5" aria-hidden="true" tabindex="-1"></a>e     <span class="ot">&lt;-</span> <span class="fu">c</span>(Dab_e, <span class="dv">39</span>,  <span class="dv">43</span>, <span class="cn">NA</span>)</span>
<span id="cb1113-6"><a href="upgma-the-hard-way.html#cb1113-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1113-7"><a href="upgma-the-hard-way.html#cb1113-7" aria-hidden="true" tabindex="-1"></a>dmat2 <span class="ot">&lt;-</span> <span class="fu">rbind</span>(ab,c,d,e)</span>
<span id="cb1113-8"><a href="upgma-the-hard-way.html#cb1113-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dmat2) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;ab&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;d&quot;</span>,<span class="st">&quot;e&quot;</span>)</span></code></pre></div>
</div>
<div id="estimate-2-d" class="section level4" number="54.8.5.2">
<h4><span class="header-section-number">54.8.5.2</span> Estimate 2 D</h4>
<p>Estimate of 2D representation of the new matrix.</p>
<div class="sourceCode" id="cb1114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1114-1"><a href="upgma-the-hard-way.html#cb1114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_dist_as_nmds</span>(dmat2)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-742-1.png" width="672" /></p>
<pre><code>## $points
##          [,1]       [,2]
## ab  13.453890 -3.0373250
## c   -6.600828 13.4680501
## d  -16.726270 -8.5744114
## e   19.948871 -0.6844397
## 
## $eig
## NULL
## 
## $x
## NULL
## 
## $ac
## [1] -4.570935e-15
## 
## $GOF
## [1] 0.7573625 1.0000000</code></pre>
</div>
</div>
</div>
<div id="algorithm---next-round" class="section level2" number="54.9">
<h2><span class="header-section-number">54.9</span> Algorithm - next round</h2>
<p>We’ve combined a and b and calculated the disnace of this clade (a,b) to the remaining taxa. What is now the minimum distance?</p>
<p>The minimum distance of the current matrix</p>
<div class="sourceCode" id="cb1116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1116-1"><a href="upgma-the-hard-way.html#cb1116-1" aria-hidden="true" tabindex="-1"></a>dist.min.i <span class="ot">&lt;-</span> <span class="fu">min</span>(dmat2, <span class="at">na.rm =</span> T)</span></code></pre></div>
<p>The index value of the current minimum</p>
<div class="sourceCode" id="cb1117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1117-1"><a href="upgma-the-hard-way.html#cb1117-1" aria-hidden="true" tabindex="-1"></a>index.min.i <span class="ot">&lt;-</span> <span class="fu">which</span>(dmat2 <span class="sc">==</span> dist.min.i, <span class="at">arr.ind =</span>T)</span>
<span id="cb1117-2"><a href="upgma-the-hard-way.html#cb1117-2" aria-hidden="true" tabindex="-1"></a>index.row.i <span class="ot">&lt;-</span> index.min.i[<span class="dv">1</span>]</span>
<span id="cb1117-3"><a href="upgma-the-hard-way.html#cb1117-3" aria-hidden="true" tabindex="-1"></a>index.col.i <span class="ot">&lt;-</span> index.min.i[<span class="dv">2</span>]</span></code></pre></div>
<p>We can get the names of our taxa using the</p>
<div class="sourceCode" id="cb1118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1118-1"><a href="upgma-the-hard-way.html#cb1118-1" aria-hidden="true" tabindex="-1"></a>cluster.i2 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(dmat2)[index.row.i]</span>
<span id="cb1118-2"><a href="upgma-the-hard-way.html#cb1118-2" aria-hidden="true" tabindex="-1"></a>cluster.i1 <span class="ot">&lt;-</span> <span class="fu">colnames</span>(dmat2)[index.col.i]</span></code></pre></div>
<p>Add output to dataframe</p>
<div class="sourceCode" id="cb1119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1119-1"><a href="upgma-the-hard-way.html#cb1119-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">2</span>,<span class="st">&quot;clust.i&quot;</span>] <span class="ot">&lt;-</span> cluster.i1</span>
<span id="cb1119-2"><a href="upgma-the-hard-way.html#cb1119-2" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">2</span>,<span class="st">&quot;clust.j&quot;</span>] <span class="ot">&lt;-</span> cluster.i2</span>
<span id="cb1119-3"><a href="upgma-the-hard-way.html#cb1119-3" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">2</span>,<span class="st">&quot;dist.ij&quot;</span>] <span class="ot">&lt;-</span> dist.min.i</span></code></pre></div>
<div id="combine-species-into-clade-1" class="section level3" number="54.9.1">
<h3><span class="header-section-number">54.9.1</span> Combine species into clade</h3>
<p>Combine the two taxa names into a new name. This can be done a couple ways.</p>
<div class="sourceCode" id="cb1120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1120-1"><a href="upgma-the-hard-way.html#cb1120-1" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="st">&quot;abe&quot;</span></span>
<span id="cb1120-2"><a href="upgma-the-hard-way.html#cb1120-2" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;ab&quot;</span>,<span class="st">&quot;e&quot;</span>,<span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1120-3"><a href="upgma-the-hard-way.html#cb1120-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1120-4"><a href="upgma-the-hard-way.html#cb1120-4" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="fu">paste</span>(cluster.i1,cluster.i2, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>We’ll add this to the dataframe</p>
<div class="sourceCode" id="cb1121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1121-1"><a href="upgma-the-hard-way.html#cb1121-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">2</span>,<span class="st">&quot;clust.u&quot;</span>] <span class="ot">&lt;-</span> clade.i</span></code></pre></div>
</div>
<div id="calculate-branch-length-1" class="section level3" number="54.9.2">
<h3><span class="header-section-number">54.9.2</span> Calculate branch length</h3>
<p>Branch length is distance/2 or d.ij/2 = d.ab.e = 22/2</p>
<div class="sourceCode" id="cb1122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1122-1"><a href="upgma-the-hard-way.html#cb1122-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">2</span>,<span class="st">&quot;branch.l.ij&quot;</span>] <span class="ot">&lt;-</span> UPGMA_output[<span class="dv">2</span>,<span class="st">&quot;dist.ij&quot;</span>]<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
</div>
<div id="recalculate-distance-from-abe-to-all-remainign-taxa" class="section level3" number="54.9.3">
<h3><span class="header-section-number">54.9.3</span> Recalculate distance from abe to all remainign taxa</h3>
<p>In the current matrix, the smallest distance is between ab and e</p>
<p>We therefore want to form a clade between ab and e (abe), the measure the distance from this clade to allthe other species.</p>
<p>Visualize this as an approximate 2D situation</p>
<div class="sourceCode" id="cb1123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1123-1"><a href="upgma-the-hard-way.html#cb1123-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb1123-2"><a href="upgma-the-hard-way.html#cb1123-2" aria-hidden="true" tabindex="-1"></a>dmat2.out <span class="ot">&lt;-</span> <span class="fu">plot_dist_as_nmds</span>(dmat2)</span>
<span id="cb1123-3"><a href="upgma-the-hard-way.html#cb1123-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1123-4"><a href="upgma-the-hard-way.html#cb1123-4" aria-hidden="true" tabindex="-1"></a><span class="co">#get x y coord of ab and e</span></span>
<span id="cb1123-5"><a href="upgma-the-hard-way.html#cb1123-5" aria-hidden="true" tabindex="-1"></a>ab.<span class="ot">&lt;-</span>dmat2.out<span class="sc">$</span>points[<span class="st">&quot;ab&quot;</span>,]</span>
<span id="cb1123-6"><a href="upgma-the-hard-way.html#cb1123-6" aria-hidden="true" tabindex="-1"></a>e.<span class="ot">&lt;-</span>dmat2.out<span class="sc">$</span>points[<span class="st">&quot;e&quot;</span>,]</span>
<span id="cb1123-7"><a href="upgma-the-hard-way.html#cb1123-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1123-8"><a href="upgma-the-hard-way.html#cb1123-8" aria-hidden="true" tabindex="-1"></a><span class="fu">draw.ellipse</span>(<span class="at">x =</span> <span class="fu">mean</span>(<span class="fu">c</span>(ab.[<span class="dv">1</span>],e.[<span class="dv">1</span>])), </span>
<span id="cb1123-9"><a href="upgma-the-hard-way.html#cb1123-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> <span class="fu">mean</span>(<span class="fu">c</span>(ab.[<span class="dv">2</span>],e.[<span class="dv">2</span>])),</span>
<span id="cb1123-10"><a href="upgma-the-hard-way.html#cb1123-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">a =</span><span class="dv">7</span>, </span>
<span id="cb1123-11"><a href="upgma-the-hard-way.html#cb1123-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">b =</span> <span class="dv">7</span>, </span>
<span id="cb1123-12"><a href="upgma-the-hard-way.html#cb1123-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">angle =</span> <span class="dv">0</span>)</span>
<span id="cb1123-13"><a href="upgma-the-hard-way.html#cb1123-13" aria-hidden="true" tabindex="-1"></a>c.<span class="ot">&lt;-</span>dmat2.out<span class="sc">$</span>points[<span class="st">&quot;c&quot;</span>,]</span>
<span id="cb1123-14"><a href="upgma-the-hard-way.html#cb1123-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1123-15"><a href="upgma-the-hard-way.html#cb1123-15" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> ab.[<span class="dv">1</span>],<span class="at">x1 =</span> c.[<span class="dv">1</span>],</span>
<span id="cb1123-16"><a href="upgma-the-hard-way.html#cb1123-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> ab.[<span class="dv">2</span>],<span class="at">y1 =</span> c.[<span class="dv">2</span>])</span>
<span id="cb1123-17"><a href="upgma-the-hard-way.html#cb1123-17" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> e.[<span class="dv">1</span>],<span class="at">x1 =</span> c.[<span class="dv">1</span>],</span>
<span id="cb1123-18"><a href="upgma-the-hard-way.html#cb1123-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> e.[<span class="dv">2</span>],<span class="at">y1 =</span> c.[<span class="dv">2</span>]) </span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-750-1.png" width="672" /></p>
<div id="abe-to-c" class="section level4" number="54.9.3.1">
<h4><span class="header-section-number">54.9.3.1</span> abe to c</h4>
<p>NOTE: calculations are “weighted” in proportion number of species in clade.
So D(ab to c) is multiplied by two because ab is 2 species</p>
<p>The denominators is 3 because of the weights used in the numerator.</p>
<p>The notation in the Wikipedia article is:</p>
<p>D(abe to c) = [D(ab to c)<em>2 + D(e to c)</em>1] / (2+1)
D(abe to c) = [ 25.5<em>2 + 39</em>1 ] / (2+1)
D(abe to c) = [ 25.5<em>2 + 39</em>1 ] / (2+1)
D(abe to c) = 32.5</p>
<p>Dabe_c = [Dab_c<em>2 + De_c</em>1] / (2+1)
Dabe_c = 32.5</p>
<div class="sourceCode" id="cb1124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1124-1"><a href="upgma-the-hard-way.html#cb1124-1" aria-hidden="true" tabindex="-1"></a>Dab_c <span class="ot">&lt;-</span> <span class="fl">25.5</span></span>
<span id="cb1124-2"><a href="upgma-the-hard-way.html#cb1124-2" aria-hidden="true" tabindex="-1"></a>De_c  <span class="ot">&lt;-</span> <span class="dv">39</span></span>
<span id="cb1124-3"><a href="upgma-the-hard-way.html#cb1124-3" aria-hidden="true" tabindex="-1"></a>Dabe_c <span class="ot">&lt;-</span> (Dab_c<span class="sc">*</span><span class="dv">2</span> <span class="sc">+</span> De_c<span class="sc">*</span><span class="dv">1</span>)<span class="sc">/</span>(<span class="dv">2</span><span class="sc">+</span><span class="dv">1</span>)</span></code></pre></div>
</div>
<div id="abe-to-c-1" class="section level4" number="54.9.3.2">
<h4><span class="header-section-number">54.9.3.2</span> abe to c</h4>
<p>For abe to d</p>
<div class="sourceCode" id="cb1125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1125-1"><a href="upgma-the-hard-way.html#cb1125-1" aria-hidden="true" tabindex="-1"></a>Dab_d <span class="ot">&lt;-</span> <span class="fl">32.5</span></span>
<span id="cb1125-2"><a href="upgma-the-hard-way.html#cb1125-2" aria-hidden="true" tabindex="-1"></a>De_d  <span class="ot">&lt;-</span> <span class="dv">43</span></span>
<span id="cb1125-3"><a href="upgma-the-hard-way.html#cb1125-3" aria-hidden="true" tabindex="-1"></a>Dabe_d <span class="ot">&lt;-</span> (Dab_d<span class="sc">*</span><span class="dv">2</span> <span class="sc">+</span> De_d<span class="sc">*</span><span class="dv">1</span>)<span class="sc">/</span>(<span class="dv">2</span><span class="sc">+</span><span class="dv">1</span>)</span></code></pre></div>
</div>
</div>
<div id="update-the-matrix" class="section level3" number="54.9.4">
<h3><span class="header-section-number">54.9.4</span> Update the matrix</h3>
<div class="sourceCode" id="cb1126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1126-1"><a href="upgma-the-hard-way.html#cb1126-1" aria-hidden="true" tabindex="-1"></a>abe <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">NA</span>,     <span class="cn">NA</span>,  <span class="cn">NA</span>)</span>
<span id="cb1126-2"><a href="upgma-the-hard-way.html#cb1126-2" aria-hidden="true" tabindex="-1"></a>c      <span class="ot">&lt;-</span> <span class="fu">c</span>(Dabe_c, <span class="cn">NA</span>,  <span class="cn">NA</span>)</span>
<span id="cb1126-3"><a href="upgma-the-hard-way.html#cb1126-3" aria-hidden="true" tabindex="-1"></a>d      <span class="ot">&lt;-</span> <span class="fu">c</span>(Dabe_d, <span class="dv">28</span>,  <span class="cn">NA</span>)</span>
<span id="cb1126-4"><a href="upgma-the-hard-way.html#cb1126-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1126-5"><a href="upgma-the-hard-way.html#cb1126-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1126-6"><a href="upgma-the-hard-way.html#cb1126-6" aria-hidden="true" tabindex="-1"></a>dmat3 <span class="ot">&lt;-</span> <span class="fu">rbind</span>(abe,c,d)</span>
<span id="cb1126-7"><a href="upgma-the-hard-way.html#cb1126-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dmat3) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;abe&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;d&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="next-iteration" class="section level2" number="54.10">
<h2><span class="header-section-number">54.10</span> Next iteration</h2>
<div id="estimate-2-d-1" class="section level4" number="54.10.0.1">
<h4><span class="header-section-number">54.10.0.1</span> Estimate 2 D</h4>
<p>a, b, e are now combined into a clade</p>
<div class="sourceCode" id="cb1127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1127-1"><a href="upgma-the-hard-way.html#cb1127-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb1127-2"><a href="upgma-the-hard-way.html#cb1127-2" aria-hidden="true" tabindex="-1"></a>dmat3.out <span class="ot">&lt;-</span> <span class="fu">plot_dist_as_nmds</span>(dmat3)</span>
<span id="cb1127-3"><a href="upgma-the-hard-way.html#cb1127-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1127-4"><a href="upgma-the-hard-way.html#cb1127-4" aria-hidden="true" tabindex="-1"></a><span class="co">#get x y coord of ab and e</span></span>
<span id="cb1127-5"><a href="upgma-the-hard-way.html#cb1127-5" aria-hidden="true" tabindex="-1"></a>c. <span class="ot">&lt;-</span>dmat3.out<span class="sc">$</span>points[<span class="st">&quot;c&quot;</span>,]</span>
<span id="cb1127-6"><a href="upgma-the-hard-way.html#cb1127-6" aria-hidden="true" tabindex="-1"></a>d.<span class="ot">&lt;-</span>dmat3.out<span class="sc">$</span>points[<span class="st">&quot;d&quot;</span>,]</span>
<span id="cb1127-7"><a href="upgma-the-hard-way.html#cb1127-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1127-8"><a href="upgma-the-hard-way.html#cb1127-8" aria-hidden="true" tabindex="-1"></a><span class="fu">draw.ellipse</span>(<span class="at">x =</span> <span class="fu">mean</span>(<span class="fu">c</span>(c.[<span class="dv">1</span>],d.[<span class="dv">1</span>])), </span>
<span id="cb1127-9"><a href="upgma-the-hard-way.html#cb1127-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> <span class="fu">mean</span>(<span class="fu">c</span>(c.[<span class="dv">2</span>],d.[<span class="dv">2</span>])),</span>
<span id="cb1127-10"><a href="upgma-the-hard-way.html#cb1127-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">a =</span><span class="dv">14</span>, </span>
<span id="cb1127-11"><a href="upgma-the-hard-way.html#cb1127-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">b =</span> <span class="dv">4</span>, </span>
<span id="cb1127-12"><a href="upgma-the-hard-way.html#cb1127-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">angle =</span> <span class="dv">105</span>)</span>
<span id="cb1127-13"><a href="upgma-the-hard-way.html#cb1127-13" aria-hidden="true" tabindex="-1"></a>abe.<span class="ot">&lt;-</span>dmat3.out<span class="sc">$</span>points[<span class="st">&quot;abe&quot;</span>,]</span>
<span id="cb1127-14"><a href="upgma-the-hard-way.html#cb1127-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1127-15"><a href="upgma-the-hard-way.html#cb1127-15" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> c.[<span class="dv">1</span>],<span class="at">x1 =</span> abe.[<span class="dv">1</span>],</span>
<span id="cb1127-16"><a href="upgma-the-hard-way.html#cb1127-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> c.[<span class="dv">2</span>],<span class="at">y1 =</span> abe.[<span class="dv">2</span>])</span>
<span id="cb1127-17"><a href="upgma-the-hard-way.html#cb1127-17" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> d.[<span class="dv">1</span>],<span class="at">x1 =</span> abe.[<span class="dv">1</span>],</span>
<span id="cb1127-18"><a href="upgma-the-hard-way.html#cb1127-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">y0 =</span> d.[<span class="dv">2</span>],<span class="at">y1 =</span> abe.[<span class="dv">2</span>]) </span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-754-1.png" width="672" /></p>
</div>
<div id="calculate-branch-lengths" class="section level3" number="54.10.1">
<h3><span class="header-section-number">54.10.1</span> Calculate branch lengths</h3>
<p>The minimum distance of the current matrix</p>
<div class="sourceCode" id="cb1128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1128-1"><a href="upgma-the-hard-way.html#cb1128-1" aria-hidden="true" tabindex="-1"></a>dist.min.i <span class="ot">&lt;-</span> <span class="fu">min</span>(dmat3, <span class="at">na.rm =</span> T)</span></code></pre></div>
<p>The index value of the current minimum</p>
<div class="sourceCode" id="cb1129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1129-1"><a href="upgma-the-hard-way.html#cb1129-1" aria-hidden="true" tabindex="-1"></a>index.min.i <span class="ot">&lt;-</span> <span class="fu">which</span>(dmat3 <span class="sc">==</span> dist.min.i, <span class="at">arr.ind =</span>T)</span>
<span id="cb1129-2"><a href="upgma-the-hard-way.html#cb1129-2" aria-hidden="true" tabindex="-1"></a>index.row.i <span class="ot">&lt;-</span> index.min.i[<span class="dv">1</span>]</span>
<span id="cb1129-3"><a href="upgma-the-hard-way.html#cb1129-3" aria-hidden="true" tabindex="-1"></a>index.col.i <span class="ot">&lt;-</span> index.min.i[<span class="dv">2</span>]</span></code></pre></div>
<p>We can get the names of our taxa using the</p>
<div class="sourceCode" id="cb1130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1130-1"><a href="upgma-the-hard-way.html#cb1130-1" aria-hidden="true" tabindex="-1"></a>cluster.i2 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(dmat3)[index.row.i]</span>
<span id="cb1130-2"><a href="upgma-the-hard-way.html#cb1130-2" aria-hidden="true" tabindex="-1"></a>cluster.i1 <span class="ot">&lt;-</span> <span class="fu">colnames</span>(dmat3)[index.col.i]</span></code></pre></div>
<p>Add output to dataframe</p>
<div class="sourceCode" id="cb1131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1131-1"><a href="upgma-the-hard-way.html#cb1131-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">3</span>,<span class="st">&quot;clust.i&quot;</span>] <span class="ot">&lt;-</span> cluster.i1</span>
<span id="cb1131-2"><a href="upgma-the-hard-way.html#cb1131-2" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">3</span>,<span class="st">&quot;clust.j&quot;</span>] <span class="ot">&lt;-</span> cluster.i2</span>
<span id="cb1131-3"><a href="upgma-the-hard-way.html#cb1131-3" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">3</span>,<span class="st">&quot;dist.ij&quot;</span>] <span class="ot">&lt;-</span> dist.min.i</span></code></pre></div>
</div>
<div id="combine-species-into-clade-2" class="section level3" number="54.10.2">
<h3><span class="header-section-number">54.10.2</span> Combine species into clade</h3>
<p>Combine the two taxa names into a new name. This can be done a couple ways.</p>
<div class="sourceCode" id="cb1132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1132-1"><a href="upgma-the-hard-way.html#cb1132-1" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="st">&quot;cd&quot;</span></span>
<span id="cb1132-2"><a href="upgma-the-hard-way.html#cb1132-2" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;c&quot;</span>,<span class="st">&quot;d&quot;</span>,<span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1132-3"><a href="upgma-the-hard-way.html#cb1132-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1132-4"><a href="upgma-the-hard-way.html#cb1132-4" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="fu">paste</span>(cluster.i1,cluster.i2, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>We’ll add this to the dataframe</p>
<div class="sourceCode" id="cb1133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1133-1"><a href="upgma-the-hard-way.html#cb1133-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">3</span>,<span class="st">&quot;clust.u&quot;</span>] <span class="ot">&lt;-</span> clade.i</span></code></pre></div>
</div>
<div id="calculate-branch-length-2" class="section level3" number="54.10.3">
<h3><span class="header-section-number">54.10.3</span> Calculate branch length</h3>
<p>Branch length is distance/2 or d.ij/2 = d.d.e = 28/2</p>
<div class="sourceCode" id="cb1134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1134-1"><a href="upgma-the-hard-way.html#cb1134-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">3</span>,<span class="st">&quot;branch.l.ij&quot;</span>] <span class="ot">&lt;-</span> UPGMA_output[<span class="dv">3</span>,<span class="st">&quot;dist.ij&quot;</span>]<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
</div>
<div id="distance-of-c-to-abe-and-d-to-abe" class="section level3" number="54.10.4">
<h3><span class="header-section-number">54.10.4</span> Distance of c to abe and d to abe</h3>
<div class="sourceCode" id="cb1135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1135-1"><a href="upgma-the-hard-way.html#cb1135-1" aria-hidden="true" tabindex="-1"></a>Dc_abe <span class="ot">&lt;-</span> <span class="dv">30</span></span>
<span id="cb1135-2"><a href="upgma-the-hard-way.html#cb1135-2" aria-hidden="true" tabindex="-1"></a>Dd_abe <span class="ot">&lt;-</span> <span class="dv">36</span></span>
<span id="cb1135-3"><a href="upgma-the-hard-way.html#cb1135-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1135-4"><a href="upgma-the-hard-way.html#cb1135-4" aria-hidden="true" tabindex="-1"></a><span class="co"># abe is composted of 3 </span></span>
<span id="cb1135-5"><a href="upgma-the-hard-way.html#cb1135-5" aria-hidden="true" tabindex="-1"></a>Ddc_abe <span class="ot">&lt;-</span> (Dc_abe<span class="sc">*</span><span class="dv">3</span> <span class="sc">+</span> Dd_abe<span class="sc">*</span><span class="dv">3</span>)<span class="sc">/</span>(<span class="dv">3</span><span class="sc">+</span><span class="dv">3</span>)</span>
<span id="cb1135-6"><a href="upgma-the-hard-way.html#cb1135-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1135-7"><a href="upgma-the-hard-way.html#cb1135-7" aria-hidden="true" tabindex="-1"></a><span class="co"># factor out 3</span></span>
<span id="cb1135-8"><a href="upgma-the-hard-way.html#cb1135-8" aria-hidden="true" tabindex="-1"></a>Ddc_abe <span class="ot">&lt;-</span> <span class="dv">3</span><span class="sc">*</span>(Dc_abe<span class="sc">*</span><span class="dv">1</span> <span class="sc">+</span> Dd_abe<span class="sc">*</span><span class="dv">1</span>)<span class="sc">/</span><span class="dv">6</span></span>
<span id="cb1135-9"><a href="upgma-the-hard-way.html#cb1135-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1135-10"><a href="upgma-the-hard-way.html#cb1135-10" aria-hidden="true" tabindex="-1"></a><span class="co"># simplify</span></span>
<span id="cb1135-11"><a href="upgma-the-hard-way.html#cb1135-11" aria-hidden="true" tabindex="-1"></a>Ddc_abe <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">*</span>(Dc_abe<span class="sc">*</span><span class="dv">1</span> <span class="sc">+</span> Dd_abe<span class="sc">*</span><span class="dv">1</span>)<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb1135-12"><a href="upgma-the-hard-way.html#cb1135-12" aria-hidden="true" tabindex="-1"></a>Ddc_abe <span class="ot">&lt;-</span> (Dc_abe<span class="sc">*</span><span class="dv">1</span> <span class="sc">+</span> Dd_abe)<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
<p>Create matrix</p>
<div class="sourceCode" id="cb1136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1136-1"><a href="upgma-the-hard-way.html#cb1136-1" aria-hidden="true" tabindex="-1"></a>abe <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">NA</span>,     <span class="cn">NA</span>)</span>
<span id="cb1136-2"><a href="upgma-the-hard-way.html#cb1136-2" aria-hidden="true" tabindex="-1"></a>cd   <span class="ot">&lt;-</span> <span class="fu">c</span>(Ddc_abe, <span class="cn">NA</span>)</span>
<span id="cb1136-3"><a href="upgma-the-hard-way.html#cb1136-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1136-4"><a href="upgma-the-hard-way.html#cb1136-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1136-5"><a href="upgma-the-hard-way.html#cb1136-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1136-6"><a href="upgma-the-hard-way.html#cb1136-6" aria-hidden="true" tabindex="-1"></a>dmat4 <span class="ot">&lt;-</span> <span class="fu">rbind</span>(abe,cd)</span>
<span id="cb1136-7"><a href="upgma-the-hard-way.html#cb1136-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dmat4) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;abe&quot;</span>,<span class="st">&quot;cd&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="finish-up" class="section level2" number="54.11">
<h2><span class="header-section-number">54.11</span> Finish up</h2>
<p>The final entry of the matrix is 33. This means that the distnace from the clade abe to the clade de is 33, with branch length of 33/2 = 16.5</p>
<p>The minimum distance</p>
<div class="sourceCode" id="cb1137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1137-1"><a href="upgma-the-hard-way.html#cb1137-1" aria-hidden="true" tabindex="-1"></a>dist.min.i <span class="ot">&lt;-</span> <span class="fu">min</span>(dmat4, <span class="at">na.rm =</span> T)</span></code></pre></div>
<p>The index value of the minimum</p>
<div class="sourceCode" id="cb1138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1138-1"><a href="upgma-the-hard-way.html#cb1138-1" aria-hidden="true" tabindex="-1"></a>index.min.i <span class="ot">&lt;-</span> <span class="fu">which</span>(dmat4 <span class="sc">==</span> dist.min.i, <span class="at">arr.ind =</span>T)</span>
<span id="cb1138-2"><a href="upgma-the-hard-way.html#cb1138-2" aria-hidden="true" tabindex="-1"></a>index.row.i <span class="ot">&lt;-</span> index.min.i[<span class="dv">1</span>]</span>
<span id="cb1138-3"><a href="upgma-the-hard-way.html#cb1138-3" aria-hidden="true" tabindex="-1"></a>index.col.i <span class="ot">&lt;-</span> index.min.i[<span class="dv">2</span>]</span></code></pre></div>
<p>We can get the names of our taxa using the</p>
<div class="sourceCode" id="cb1139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1139-1"><a href="upgma-the-hard-way.html#cb1139-1" aria-hidden="true" tabindex="-1"></a>cluster.i2 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(dmat4)[index.row.i]</span>
<span id="cb1139-2"><a href="upgma-the-hard-way.html#cb1139-2" aria-hidden="true" tabindex="-1"></a>cluster.i1 <span class="ot">&lt;-</span> <span class="fu">colnames</span>(dmat4)[index.col.i]</span></code></pre></div>
<p>Add output to dataframe</p>
<div class="sourceCode" id="cb1140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1140-1"><a href="upgma-the-hard-way.html#cb1140-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">4</span>,<span class="st">&quot;clust.i&quot;</span>] <span class="ot">&lt;-</span> cluster.i1</span>
<span id="cb1140-2"><a href="upgma-the-hard-way.html#cb1140-2" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">4</span>,<span class="st">&quot;clust.j&quot;</span>] <span class="ot">&lt;-</span> cluster.i2</span>
<span id="cb1140-3"><a href="upgma-the-hard-way.html#cb1140-3" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">4</span>,<span class="st">&quot;dist.ij&quot;</span>] <span class="ot">&lt;-</span> dist.min.i</span></code></pre></div>
<div id="combine-species-into-clade-3" class="section level3" number="54.11.1">
<h3><span class="header-section-number">54.11.1</span> Combine species into clade</h3>
<p>Combine the two taxa names into a new name. This can be done a couple ways.</p>
<div class="sourceCode" id="cb1141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1141-1"><a href="upgma-the-hard-way.html#cb1141-1" aria-hidden="true" tabindex="-1"></a>clade.i <span class="ot">&lt;-</span> <span class="fu">paste</span>(cluster.i1,cluster.i2, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>We’ll add this to the dataframe</p>
<div class="sourceCode" id="cb1142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1142-1"><a href="upgma-the-hard-way.html#cb1142-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">4</span>,<span class="st">&quot;clust.u&quot;</span>] <span class="ot">&lt;-</span> clade.i</span></code></pre></div>
</div>
<div id="calculate-branch-length-3" class="section level3" number="54.11.2">
<h3><span class="header-section-number">54.11.2</span> Calculate branch length</h3>
<p>Branch length is distance/2 or d.ij/2</p>
<div class="sourceCode" id="cb1143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1143-1"><a href="upgma-the-hard-way.html#cb1143-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output[<span class="dv">4</span>,<span class="st">&quot;branch.l.ij&quot;</span>] <span class="ot">&lt;-</span> UPGMA_output[<span class="dv">4</span>,<span class="st">&quot;dist.ij&quot;</span>]<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
</div>
</div>
<div id="finalizing-branch-lengths" class="section level2" number="54.12">
<h2><span class="header-section-number">54.12</span> Finalizing branch lengths</h2>
<p>We now have a final tree structure. In Newick format it would be</p>
<div class="sourceCode" id="cb1144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1144-1"><a href="upgma-the-hard-way.html#cb1144-1" aria-hidden="true" tabindex="-1"></a>full.tree <span class="ot">&lt;-</span> <span class="st">&quot;((c,d),(e,(b, a)));&quot;</span> <span class="co"># semi colon!</span></span>
<span id="cb1144-2"><a href="upgma-the-hard-way.html#cb1144-2" aria-hidden="true" tabindex="-1"></a>full.tree <span class="ot">&lt;-</span><span class="fu">read.tree</span>(<span class="at">text=</span>full.tree)</span>
<span id="cb1144-3"><a href="upgma-the-hard-way.html#cb1144-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;3-taxa tree&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-771-1.png" width="672" /></p>
<p>While we’ve done our clustering we’ve calcualte branch lengths and put them in the column branch.l.ij of the dataframe UPGMA_output</p>
<div class="sourceCode" id="cb1145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1145-1"><a href="upgma-the-hard-way.html#cb1145-1" aria-hidden="true" tabindex="-1"></a>UPGMA_output</span></code></pre></div>
<pre><code>##   it clust.i clust.j dist.ij clust.u branch.l.ij
## 1  1       a       b      17      ab         8.5
## 2  2      ab       e      22     abe        11.0
## 3  3       c       d      28      cd        14.0
## 4  4     abe      cd      33   abecd        16.5
## 5  5    &lt;NA&gt;    &lt;NA&gt;      NA    &lt;NA&gt;          NA</code></pre>
<p>The following code adds them to the tree.</p>
<p>First, the brnach lenghts from a to b</p>
<div class="sourceCode" id="cb1147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1147-1"><a href="upgma-the-hard-way.html#cb1147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1147-2"><a href="upgma-the-hard-way.html#cb1147-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a-b branch lengths= 8.5</span></span>
<span id="cb1147-3"><a href="upgma-the-hard-way.html#cb1147-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">5.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1147-4"><a href="upgma-the-hard-way.html#cb1147-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-773-1.png" width="672" /></p>
<p>now a to b, and ab to e</p>
<div class="sourceCode" id="cb1148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1148-1"><a href="upgma-the-hard-way.html#cb1148-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1148-2"><a href="upgma-the-hard-way.html#cb1148-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a-b branch lengths= 8.5</span></span>
<span id="cb1148-3"><a href="upgma-the-hard-way.html#cb1148-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">5.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1148-4"><a href="upgma-the-hard-way.html#cb1148-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1148-5"><a href="upgma-the-hard-way.html#cb1148-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ab-e branch lengths= 1</span></span>
<span id="cb1148-6"><a href="upgma-the-hard-way.html#cb1148-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">3.1</span>,<span class="at">labels =</span> <span class="dv">11</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-774-1.png" width="672" /></p>
<p>There’s a branch between the clade ab and the node that connects with e. What’s its length? FOr a simple ultrametric tree calcualted with UPGMA we can calculate this by subtraction</p>
<div class="sourceCode" id="cb1149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1149-1"><a href="upgma-the-hard-way.html#cb1149-1" aria-hidden="true" tabindex="-1"></a><span class="dv">11</span><span class="fl">-8.5</span></span></code></pre></div>
<pre><code>## [1] 2.5</code></pre>
<p>We’ll added this short branch of 2.5. Note this things ARE NOT to scale.</p>
<div class="sourceCode" id="cb1151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1151-1"><a href="upgma-the-hard-way.html#cb1151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1151-2"><a href="upgma-the-hard-way.html#cb1151-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a-b branch lengths= 8.5</span></span>
<span id="cb1151-3"><a href="upgma-the-hard-way.html#cb1151-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">5.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1151-4"><a href="upgma-the-hard-way.html#cb1151-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1151-5"><a href="upgma-the-hard-way.html#cb1151-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ab-e branch lengths= 1</span></span>
<span id="cb1151-6"><a href="upgma-the-hard-way.html#cb1151-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">3.1</span>,<span class="at">labels =</span> <span class="dv">11</span>)</span>
<span id="cb1151-7"><a href="upgma-the-hard-way.html#cb1151-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1151-8"><a href="upgma-the-hard-way.html#cb1151-8" aria-hidden="true" tabindex="-1"></a><span class="co"># short branch</span></span>
<span id="cb1151-9"><a href="upgma-the-hard-way.html#cb1151-9" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">2.5</span>,<span class="at">y =</span> <span class="fl">4.6</span>,<span class="at">labels =</span> <span class="fl">2.5</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-776-1.png" width="672" /></p>
<p>To visualize exactly what is going on here its useful to add this</p>
<div class="sourceCode" id="cb1152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1152-1"><a href="upgma-the-hard-way.html#cb1152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb1152-2"><a href="upgma-the-hard-way.html#cb1152-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1152-3"><a href="upgma-the-hard-way.html#cb1152-3" aria-hidden="true" tabindex="-1"></a><span class="co"># a-b branch lengths= 8.5</span></span>
<span id="cb1152-4"><a href="upgma-the-hard-way.html#cb1152-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.9</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1152-5"><a href="upgma-the-hard-way.html#cb1152-5" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1152-6"><a href="upgma-the-hard-way.html#cb1152-6" aria-hidden="true" tabindex="-1"></a><span class="co"># ab-e branch lengths= 1</span></span>
<span id="cb1152-7"><a href="upgma-the-hard-way.html#cb1152-7" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">3.1</span>,<span class="at">labels =</span> <span class="dv">11</span>)</span>
<span id="cb1152-8"><a href="upgma-the-hard-way.html#cb1152-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1152-9"><a href="upgma-the-hard-way.html#cb1152-9" aria-hidden="true" tabindex="-1"></a><span class="co"># short branch</span></span>
<span id="cb1152-10"><a href="upgma-the-hard-way.html#cb1152-10" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">2.5</span>,<span class="at">y =</span> <span class="fl">4.6</span>,<span class="at">labels =</span> <span class="fl">2.5</span>)</span>
<span id="cb1152-11"><a href="upgma-the-hard-way.html#cb1152-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1152-12"><a href="upgma-the-hard-way.html#cb1152-12" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(<span class="at">x0 =</span> <span class="dv">2</span>,<span class="at">x1 =</span> <span class="dv">4</span>,<span class="at">y0 =</span><span class="fl">5.15</span>,<span class="at">y1=</span><span class="fl">5.15</span>, <span class="at">code =</span><span class="dv">3</span>,<span class="at">angle =</span> <span class="dv">90</span>,<span class="at">length =</span><span class="fl">0.1</span>, <span class="at">lty =</span><span class="dv">2</span>, <span class="at">col =</span> <span class="dv">2</span>)</span>
<span id="cb1152-13"><a href="upgma-the-hard-way.html#cb1152-13" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">2.5</span>,<span class="at">y =</span> <span class="dv">5</span>,<span class="at">labels =</span> <span class="st">&quot;2.5+8.5=11&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-777-1.png" width="672" /></p>
<p>We can now add the d-e branch of 14 (note again - not to scale!)</p>
<div class="sourceCode" id="cb1153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1153-1"><a href="upgma-the-hard-way.html#cb1153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1153-2"><a href="upgma-the-hard-way.html#cb1153-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a-b branch lengths= 8.5</span></span>
<span id="cb1153-3"><a href="upgma-the-hard-way.html#cb1153-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">5.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1153-4"><a href="upgma-the-hard-way.html#cb1153-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1153-5"><a href="upgma-the-hard-way.html#cb1153-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ab-e branch lengths= 1</span></span>
<span id="cb1153-6"><a href="upgma-the-hard-way.html#cb1153-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">3.1</span>,<span class="at">labels =</span> <span class="dv">11</span>)</span>
<span id="cb1153-7"><a href="upgma-the-hard-way.html#cb1153-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1153-8"><a href="upgma-the-hard-way.html#cb1153-8" aria-hidden="true" tabindex="-1"></a><span class="co"># short branch</span></span>
<span id="cb1153-9"><a href="upgma-the-hard-way.html#cb1153-9" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">2.5</span>,<span class="at">y =</span> <span class="fl">4.6</span>,<span class="at">labels =</span> <span class="fl">2.5</span>)</span>
<span id="cb1153-10"><a href="upgma-the-hard-way.html#cb1153-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1153-11"><a href="upgma-the-hard-way.html#cb1153-11" aria-hidden="true" tabindex="-1"></a><span class="co"># d-e branch</span></span>
<span id="cb1153-12"><a href="upgma-the-hard-way.html#cb1153-12" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">1.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1153-13"><a href="upgma-the-hard-way.html#cb1153-13" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">2.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-778-1.png" width="672" /></p>
<p>The distance from abe to de is 16.5 We can visualize this as</p>
<div class="sourceCode" id="cb1154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1154-1"><a href="upgma-the-hard-way.html#cb1154-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1154-2"><a href="upgma-the-hard-way.html#cb1154-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a-b branch lengths= 8.5</span></span>
<span id="cb1154-3"><a href="upgma-the-hard-way.html#cb1154-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">5.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1154-4"><a href="upgma-the-hard-way.html#cb1154-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1154-5"><a href="upgma-the-hard-way.html#cb1154-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ab-e branch lengths= 1</span></span>
<span id="cb1154-6"><a href="upgma-the-hard-way.html#cb1154-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">3.1</span>,<span class="at">labels =</span> <span class="dv">11</span>)</span>
<span id="cb1154-7"><a href="upgma-the-hard-way.html#cb1154-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1154-8"><a href="upgma-the-hard-way.html#cb1154-8" aria-hidden="true" tabindex="-1"></a><span class="co"># short branch</span></span>
<span id="cb1154-9"><a href="upgma-the-hard-way.html#cb1154-9" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">2.5</span>,<span class="at">y =</span> <span class="fl">4.6</span>,<span class="at">labels =</span> <span class="fl">2.5</span>)</span>
<span id="cb1154-10"><a href="upgma-the-hard-way.html#cb1154-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1154-11"><a href="upgma-the-hard-way.html#cb1154-11" aria-hidden="true" tabindex="-1"></a><span class="co"># d-e branch</span></span>
<span id="cb1154-12"><a href="upgma-the-hard-way.html#cb1154-12" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">1.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1154-13"><a href="upgma-the-hard-way.html#cb1154-13" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">2.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1154-14"><a href="upgma-the-hard-way.html#cb1154-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1154-15"><a href="upgma-the-hard-way.html#cb1154-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1154-16"><a href="upgma-the-hard-way.html#cb1154-16" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(<span class="at">x0 =</span> <span class="dv">0</span>,<span class="at">x1 =</span> <span class="dv">4</span>,<span class="at">y0 =</span><span class="fl">0.85</span>,<span class="at">y1=</span><span class="fl">0.85</span>, <span class="at">code =</span><span class="dv">3</span>,<span class="at">angle =</span> <span class="dv">90</span>,<span class="at">length =</span><span class="fl">0.1</span>, <span class="at">lty =</span><span class="dv">2</span>, <span class="at">col =</span> <span class="dv">2</span>)</span>
<span id="cb1154-17"><a href="upgma-the-hard-way.html#cb1154-17" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">1</span>,<span class="at">y =</span> <span class="dv">1</span>,<span class="at">labels =</span> <span class="fl">16.5</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-779-1.png" width="672" /></p>
<p>We can calcualte the distance of the remaining short branch from de to the root as 16.5-14 = 2.5. When we plot this again, it won’t be to scale</p>
<div class="sourceCode" id="cb1155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1155-1"><a href="upgma-the-hard-way.html#cb1155-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1155-2"><a href="upgma-the-hard-way.html#cb1155-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a-b branch lengths= 8.5</span></span>
<span id="cb1155-3"><a href="upgma-the-hard-way.html#cb1155-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">5.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1155-4"><a href="upgma-the-hard-way.html#cb1155-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1155-5"><a href="upgma-the-hard-way.html#cb1155-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ab-e branch lengths= 1</span></span>
<span id="cb1155-6"><a href="upgma-the-hard-way.html#cb1155-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">3.1</span>,<span class="at">labels =</span> <span class="dv">11</span>)</span>
<span id="cb1155-7"><a href="upgma-the-hard-way.html#cb1155-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1155-8"><a href="upgma-the-hard-way.html#cb1155-8" aria-hidden="true" tabindex="-1"></a><span class="co"># short branch</span></span>
<span id="cb1155-9"><a href="upgma-the-hard-way.html#cb1155-9" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">2.5</span>,<span class="at">y =</span> <span class="fl">4.6</span>,<span class="at">labels =</span> <span class="fl">2.5</span>)</span>
<span id="cb1155-10"><a href="upgma-the-hard-way.html#cb1155-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1155-11"><a href="upgma-the-hard-way.html#cb1155-11" aria-hidden="true" tabindex="-1"></a><span class="co"># d-e branch</span></span>
<span id="cb1155-12"><a href="upgma-the-hard-way.html#cb1155-12" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">1.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1155-13"><a href="upgma-the-hard-way.html#cb1155-13" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">2.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1155-14"><a href="upgma-the-hard-way.html#cb1155-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1155-15"><a href="upgma-the-hard-way.html#cb1155-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1155-16"><a href="upgma-the-hard-way.html#cb1155-16" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(<span class="at">x0 =</span> <span class="dv">0</span>,<span class="at">x1 =</span> <span class="dv">4</span>,<span class="at">y0 =</span><span class="fl">0.85</span>,<span class="at">y1=</span><span class="fl">0.85</span>, <span class="at">code =</span><span class="dv">3</span>,<span class="at">angle =</span> <span class="dv">90</span>,<span class="at">length =</span><span class="fl">0.1</span>, <span class="at">lty =</span><span class="dv">2</span>, <span class="at">col =</span> <span class="dv">2</span>)</span>
<span id="cb1155-17"><a href="upgma-the-hard-way.html#cb1155-17" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">1</span>,<span class="at">y =</span> <span class="dv">1</span>,<span class="at">labels =</span> <span class="st">&quot;16.5 - 14 = 2.5&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-780-1.png" width="672" /></p>
<p>We can calculate the remaining short branch from abe to the root as
16.5-8.5-2.5
=16.5-11
=5.5</p>
<div class="sourceCode" id="cb1156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1156-1"><a href="upgma-the-hard-way.html#cb1156-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1156-2"><a href="upgma-the-hard-way.html#cb1156-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a-b branch lengths= 8.5</span></span>
<span id="cb1156-3"><a href="upgma-the-hard-way.html#cb1156-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">5.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1156-4"><a href="upgma-the-hard-way.html#cb1156-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1156-5"><a href="upgma-the-hard-way.html#cb1156-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ab-e branch lengths= 1</span></span>
<span id="cb1156-6"><a href="upgma-the-hard-way.html#cb1156-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">3.1</span>,<span class="at">labels =</span> <span class="dv">11</span>)</span>
<span id="cb1156-7"><a href="upgma-the-hard-way.html#cb1156-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1156-8"><a href="upgma-the-hard-way.html#cb1156-8" aria-hidden="true" tabindex="-1"></a><span class="co"># short branch</span></span>
<span id="cb1156-9"><a href="upgma-the-hard-way.html#cb1156-9" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">2.5</span>,<span class="at">y =</span> <span class="fl">4.6</span>,<span class="at">labels =</span> <span class="fl">2.5</span>)</span>
<span id="cb1156-10"><a href="upgma-the-hard-way.html#cb1156-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1156-11"><a href="upgma-the-hard-way.html#cb1156-11" aria-hidden="true" tabindex="-1"></a><span class="co"># d-e branch</span></span>
<span id="cb1156-12"><a href="upgma-the-hard-way.html#cb1156-12" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">1.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1156-13"><a href="upgma-the-hard-way.html#cb1156-13" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">2.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1156-14"><a href="upgma-the-hard-way.html#cb1156-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1156-15"><a href="upgma-the-hard-way.html#cb1156-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1156-16"><a href="upgma-the-hard-way.html#cb1156-16" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(<span class="at">x0 =</span> <span class="dv">0</span>,<span class="at">x1 =</span> <span class="dv">4</span>,<span class="at">y0 =</span><span class="fl">0.85</span>,<span class="at">y1=</span><span class="fl">0.85</span>, <span class="at">code =</span><span class="dv">3</span>,<span class="at">angle =</span> <span class="dv">90</span>,<span class="at">length =</span><span class="fl">0.1</span>, <span class="at">lty =</span><span class="dv">2</span>, <span class="at">col =</span> <span class="dv">2</span>)</span>
<span id="cb1156-17"><a href="upgma-the-hard-way.html#cb1156-17" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">1</span>,<span class="at">y =</span> <span class="dv">1</span>,<span class="at">labels =</span> <span class="st">&quot;16.5 - 14 = 2.5&quot;</span>)</span>
<span id="cb1156-18"><a href="upgma-the-hard-way.html#cb1156-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1156-19"><a href="upgma-the-hard-way.html#cb1156-19" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(<span class="at">x0 =</span> <span class="dv">0</span>,<span class="at">x1 =</span> <span class="dv">4</span>,<span class="at">y0 =</span><span class="fl">5.1</span>,<span class="at">y1=</span><span class="fl">5.1</span>, <span class="at">code =</span><span class="dv">3</span>,<span class="at">angle =</span> <span class="dv">90</span>,<span class="at">length =</span><span class="fl">0.1</span>, <span class="at">lty =</span><span class="dv">2</span>, <span class="at">col =</span> <span class="dv">2</span>)</span>
<span id="cb1156-20"><a href="upgma-the-hard-way.html#cb1156-20" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">1</span>,<span class="at">y =</span> <span class="dv">5</span>,<span class="at">labels =</span> <span class="st">&quot;16.5 - 8.5 - 2.5 = 5.5&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-781-1.png" width="672" /></p>
<p>So the final tree has branch lengths (not to scale!) of</p>
<div class="sourceCode" id="cb1157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1157-1"><a href="upgma-the-hard-way.html#cb1157-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1157-2"><a href="upgma-the-hard-way.html#cb1157-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a-b branch lengths= 8.5</span></span>
<span id="cb1157-3"><a href="upgma-the-hard-way.html#cb1157-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">5.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1157-4"><a href="upgma-the-hard-way.html#cb1157-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1157-5"><a href="upgma-the-hard-way.html#cb1157-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ab-e branch lengths= 1</span></span>
<span id="cb1157-6"><a href="upgma-the-hard-way.html#cb1157-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">3.1</span>,<span class="at">labels =</span> <span class="dv">11</span>)</span>
<span id="cb1157-7"><a href="upgma-the-hard-way.html#cb1157-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1157-8"><a href="upgma-the-hard-way.html#cb1157-8" aria-hidden="true" tabindex="-1"></a><span class="co"># short branch</span></span>
<span id="cb1157-9"><a href="upgma-the-hard-way.html#cb1157-9" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">2.5</span>,<span class="at">y =</span> <span class="fl">4.6</span>,<span class="at">labels =</span> <span class="fl">2.5</span>)</span>
<span id="cb1157-10"><a href="upgma-the-hard-way.html#cb1157-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1157-11"><a href="upgma-the-hard-way.html#cb1157-11" aria-hidden="true" tabindex="-1"></a><span class="co"># d-e branch</span></span>
<span id="cb1157-12"><a href="upgma-the-hard-way.html#cb1157-12" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">1.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1157-13"><a href="upgma-the-hard-way.html#cb1157-13" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">3.5</span>,<span class="at">y =</span> <span class="fl">2.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1157-14"><a href="upgma-the-hard-way.html#cb1157-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1157-15"><a href="upgma-the-hard-way.html#cb1157-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1157-16"><a href="upgma-the-hard-way.html#cb1157-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1157-17"><a href="upgma-the-hard-way.html#cb1157-17" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">1</span>,<span class="at">y =</span> <span class="fl">1.6</span>,<span class="at">labels =</span> <span class="st">&quot;2.5&quot;</span>)</span>
<span id="cb1157-18"><a href="upgma-the-hard-way.html#cb1157-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1157-19"><a href="upgma-the-hard-way.html#cb1157-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1157-20"><a href="upgma-the-hard-way.html#cb1157-20" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">1</span>,<span class="at">y =</span> <span class="dv">4</span>,<span class="at">labels =</span> <span class="st">&quot;5.5&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-782-1.png" width="672" /></p>
<p>Most of the time we don’t include branch lengths in Newick format, but there is a way to do it. You don’t need to know how to do this - this is just for illustration</p>
<div class="sourceCode" id="cb1158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1158-1"><a href="upgma-the-hard-way.html#cb1158-1" aria-hidden="true" tabindex="-1"></a>full.tree <span class="ot">&lt;-</span> <span class="st">&quot;((c:14,d:14):2.5,(e:11,(b:8.5, a:8.5):2.5):5.5);&quot;</span> <span class="co"># semi colon!</span></span>
<span id="cb1158-2"><a href="upgma-the-hard-way.html#cb1158-2" aria-hidden="true" tabindex="-1"></a>full.tree <span class="ot">&lt;-</span><span class="fu">read.tree</span>(<span class="at">text=</span>full.tree)</span>
<span id="cb1158-3"><a href="upgma-the-hard-way.html#cb1158-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(full.tree, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb1158-4"><a href="upgma-the-hard-way.html#cb1158-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1158-5"><a href="upgma-the-hard-way.html#cb1158-5" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">10</span>,<span class="at">y =</span> <span class="fl">5.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1158-6"><a href="upgma-the-hard-way.html#cb1158-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">10</span>,<span class="at">y =</span> <span class="fl">4.1</span>,<span class="at">labels =</span> <span class="fl">8.5</span>)</span>
<span id="cb1158-7"><a href="upgma-the-hard-way.html#cb1158-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ab-e branch lengths= 1</span></span>
<span id="cb1158-8"><a href="upgma-the-hard-way.html#cb1158-8" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">10</span>,<span class="at">y =</span> <span class="fl">3.1</span>,<span class="at">labels =</span> <span class="dv">11</span>)</span>
<span id="cb1158-9"><a href="upgma-the-hard-way.html#cb1158-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1158-10"><a href="upgma-the-hard-way.html#cb1158-10" aria-hidden="true" tabindex="-1"></a><span class="co"># short branch</span></span>
<span id="cb1158-11"><a href="upgma-the-hard-way.html#cb1158-11" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">7</span>,<span class="at">y =</span> <span class="fl">4.6</span>,<span class="at">labels =</span> <span class="fl">2.5</span>)</span>
<span id="cb1158-12"><a href="upgma-the-hard-way.html#cb1158-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1158-13"><a href="upgma-the-hard-way.html#cb1158-13" aria-hidden="true" tabindex="-1"></a><span class="co"># d-e branch</span></span>
<span id="cb1158-14"><a href="upgma-the-hard-way.html#cb1158-14" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">10</span>,<span class="at">y =</span> <span class="fl">1.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1158-15"><a href="upgma-the-hard-way.html#cb1158-15" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">10</span>,<span class="at">y =</span> <span class="fl">2.1</span>,<span class="at">labels =</span> <span class="dv">14</span>)</span>
<span id="cb1158-16"><a href="upgma-the-hard-way.html#cb1158-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1158-17"><a href="upgma-the-hard-way.html#cb1158-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1158-18"><a href="upgma-the-hard-way.html#cb1158-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1158-19"><a href="upgma-the-hard-way.html#cb1158-19" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">1.5</span>,<span class="at">y =</span> <span class="fl">1.6</span>,<span class="at">labels =</span> <span class="st">&quot;2.5&quot;</span>)</span>
<span id="cb1158-20"><a href="upgma-the-hard-way.html#cb1158-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1158-21"><a href="upgma-the-hard-way.html#cb1158-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1158-22"><a href="upgma-the-hard-way.html#cb1158-22" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">1.5</span>,<span class="at">y =</span> <span class="fl">3.9</span>,<span class="at">labels =</span> <span class="st">&quot;5.5&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-783-1.png" width="672" /></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="representing-phylogenetic-trees-in-newick-format.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/666-treethinking/tree-think_UPGMA_wikidat_compare_to_ordination_calcs_by_hand-UPDATED.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/666-treethinking/tree-think_UPGMA_wikidat_compare_to_ordination_calcs_by_hand-UPDATED.Rmd",
"text": null
},
"download": ["lbrb.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
