<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 51 “Worked example: Building a phylogeny in R” | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 51 “Worked example: Building a phylogeny in R” | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 51 “Worked example: Building a phylogeny in R” | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-07-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="a-complete-bioinformatics-workflow-in-r.html"/>
<link rel="next" href="calculating-genetic-distances-between-sequences.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.5.1/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>1</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="1.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>1.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="1.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction"><i class="fa fa-check"></i><b>1.3</b> Introduction</a></li>
<li class="chapter" data-level="1.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary"><i class="fa fa-check"></i><b>1.4</b> Vocabulary</a></li>
<li class="chapter" data-level="1.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions"><i class="fa fa-check"></i><b>1.5</b> Functions</a></li>
<li class="chapter" data-level="1.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries"><i class="fa fa-check"></i><b>1.6</b> Preliminaries</a></li>
<li class="chapter" data-level="1.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>1.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="1.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>1.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="1.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>1.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="1.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>1.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="1.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>1.8.3</b> DNA words</a></li>
<li class="chapter" data-level="1.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary"><i class="fa fa-check"></i><b>1.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="1.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>1.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="1.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>1.9.1</b> License</a></li>
<li class="chapter" data-level="1.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>1.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="working-with-vectors-and-matrices.html"><a href="working-with-vectors-and-matrices.html"><i class="fa fa-check"></i><b>2</b> Working with vectors and matrices</a></li>
<li class="chapter" data-level="3" data-path="a-note-on-changes-to-database-entries.html"><a href="a-note-on-changes-to-database-entries.html"><i class="fa fa-check"></i><b>3</b> A note on changes to database entries</a></li>
<li class="chapter" data-level="4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html"><i class="fa fa-check"></i><b>4</b> Writing functions - practice problems</a>
<ul>
<li class="chapter" data-level="4.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#introduction-1"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#part-1-pure-practice"><i class="fa fa-check"></i><b>4.2</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-1-hello-world"><i class="fa fa-check"></i><b>4.2.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="4.2.2" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-2-hello-_______"><i class="fa fa-check"></i><b>4.2.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-3-make-you-own-natural-log-function"><i class="fa fa-check"></i><b>4.3</b> Practice Function 3: Make you own natural log function</a></li>
<li class="chapter" data-level="4.4" data-path="writing-functions-practice-problems.html"><a href="writing-functions-practice-problems.html#practice-function-4-make-function-that-converts-dna-directly-to-rna"><i class="fa fa-check"></i><b>4.4</b> Practice Function 4: Make function that converts DNA directly to RNA</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html"><i class="fa fa-check"></i><b>5</b> Computational Biology Portfolio: Practice Working with Strings</a>
<ul>
<li class="chapter" data-level="5.1" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#introduction-2"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#purpose"><i class="fa fa-check"></i><b>5.2</b> Purpose</a></li>
<li class="chapter" data-level="5.3" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#gonder-et-al-2007-data"><i class="fa fa-check"></i><b>5.3</b> Gonder et al 2007 data</a></li>
<li class="chapter" data-level="5.4" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#gonder-et-al-2007-new-sequences"><i class="fa fa-check"></i><b>5.4</b> Gonder et al 2007 new sequences</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#creating-the-codes"><i class="fa fa-check"></i><b>5.4.1</b> Creating the codes</a></li>
<li class="chapter" data-level="5.4.2" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#selecting-a-code-for-demonstration"><i class="fa fa-check"></i><b>5.4.2</b> Selecting a code for demonstration</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#searching-genbank"><i class="fa fa-check"></i><b>5.5</b> Searching Genbank</a></li>
<li class="chapter" data-level="5.6" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#gonder-et-al-2007-previously-published-sequences"><i class="fa fa-check"></i><b>5.6</b> Gonder et al 2007 previously published sequences</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#creating-the-codes-1"><i class="fa fa-check"></i><b>5.6.1</b> Creating the codes</a></li>
<li class="chapter" data-level="5.6.2" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#selecting-a-code-for-demonstration-1"><i class="fa fa-check"></i><b>5.6.2</b> Selecting a code for demonstration</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="computational-biology-portfolio-practice-working-with-strings.html"><a href="computational-biology-portfolio-practice-working-with-strings.html#searching-genbank-1"><i class="fa fa-check"></i><b>5.7</b> Searching Genbank</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html"><i class="fa fa-check"></i><b>6</b> Writing user-friendly functions</a>
<ul>
<li class="chapter" data-level="6.1" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-1"><i class="fa fa-check"></i><b>6.1</b> Version 1</a></li>
<li class="chapter" data-level="6.2" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-2"><i class="fa fa-check"></i><b>6.2</b> Version 2</a></li>
<li class="chapter" data-level="6.3" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-3"><i class="fa fa-check"></i><b>6.3</b> Version 3</a></li>
<li class="chapter" data-level="6.4" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-4-adding-conditions-and-warnigns"><i class="fa fa-check"></i><b>6.4</b> Version 4: Adding conditions and warnigns</a></li>
<li class="chapter" data-level="6.5" data-path="writing-user-friendly-functions.html"><a href="writing-user-friendly-functions.html#version-5"><i class="fa fa-check"></i><b>6.5</b> Version 5</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html"><i class="fa fa-check"></i><b>7</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="7.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#commands-used"><i class="fa fa-check"></i><b>7.1</b> Commands used</a></li>
<li class="chapter" data-level="7.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#r-objects"><i class="fa fa-check"></i><b>7.2</b> R Objects</a></li>
<li class="chapter" data-level="7.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>7.3</b> Differences between objects</a></li>
<li class="chapter" data-level="7.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-data"><i class="fa fa-check"></i><b>7.4</b> The Data</a></li>
<li class="chapter" data-level="7.5" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>7.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#is"><i class="fa fa-check"></i><b>7.5.1</b> is()</a></li>
<li class="chapter" data-level="7.5.2" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#length"><i class="fa fa-check"></i><b>7.5.2</b> length()</a></li>
<li class="chapter" data-level="7.5.3" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#str"><i class="fa fa-check"></i><b>7.5.3</b> str()</a></li>
<li class="chapter" data-level="7.5.4" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#c"><i class="fa fa-check"></i><b>7.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#you-can-save-statistical-output-as-an-object"><i class="fa fa-check"></i><b>7.6</b> You can save statistical output as an object</a></li>
<li class="chapter" data-level="7.7" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#how-does-r-save-statistical-output"><i class="fa fa-check"></i><b>7.7</b> How does R save statistical output</a></li>
<li class="chapter" data-level="7.8" data-path="intro-to-r-objects.html"><a href="intro-to-r-objects.html#lists-in-r"><i class="fa fa-check"></i><b>7.8</b> Lists in R</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="an-aside-about-t-tests.html"><a href="an-aside-about-t-tests.html"><i class="fa fa-check"></i><b>8</b> An aside about t-tests</a></li>
<li class="chapter" data-level="9" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html"><i class="fa fa-check"></i><b>9</b> Working with square brackets</a>
<ul>
<li class="chapter" data-level="9.1" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#square-brackets-for-real"><i class="fa fa-check"></i><b>9.1</b> Square brackets for real</a></li>
<li class="chapter" data-level="9.2" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#difference-between-means-using-square-brackets"><i class="fa fa-check"></i><b>9.2</b> Difference between means using square brackets</a></li>
<li class="chapter" data-level="9.3" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#upshot-plotting-output-of-an-r-statistical-test"><i class="fa fa-check"></i><b>9.3</b> Upshot: plotting output of an R statistical test</a></li>
<li class="chapter" data-level="9.4" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#the-not-bad-way-typing-directly-into-r"><i class="fa fa-check"></i><b>9.4</b> The “not bad way”: typing directly into R</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#not-bad-way-with-a-bar-plot"><i class="fa fa-check"></i><b>9.4.1</b> “NOt bad way” with a bar plot</a></li>
<li class="chapter" data-level="9.4.2" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#not-bad-way-with-using-a-single-dot"><i class="fa fa-check"></i><b>9.4.2</b> “NOt bad way” with using a single dot</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#the-best-way"><i class="fa fa-check"></i><b>9.5</b> The best way</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#a-final-fancy-touch"><i class="fa fa-check"></i><b>9.5.1</b> A final fancy touch</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="working-with-square-brackets.html"><a href="working-with-square-brackets.html#debrief"><i class="fa fa-check"></i><b>9.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html"><i class="fa fa-check"></i><b>10</b> or loops in R - copy? alt?</a>
<ul>
<li class="chapter" data-level="10.1" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#introduction-3"><i class="fa fa-check"></i><b>10.1</b> Introduction</a></li>
<li class="chapter" data-level="10.2" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#reminders"><i class="fa fa-check"></i><b>10.2</b> Reminders</a></li>
<li class="chapter" data-level="10.3" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop"><i class="fa fa-check"></i><b>10.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="10.4" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-3-the-classic-for-loop"><i class="fa fa-check"></i><b>10.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="10.5" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-3-the-power-user-for-loop"><i class="fa fa-check"></i><b>10.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="10.6" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#for-loop-4-a-for-loop-for-all-purposes"><i class="fa fa-check"></i><b>10.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="10.7" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#advanced-aside-if-else-and-next-statements"><i class="fa fa-check"></i><b>10.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="10.8" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#on-avoiding-for-loops"><i class="fa fa-check"></i><b>10.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="10.9" data-path="or-loops-in-r-copy-alt.html"><a href="or-loops-in-r-copy-alt.html#challenge"><i class="fa fa-check"></i><b>10.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="function-practice.html"><a href="function-practice.html"><i class="fa fa-check"></i><b>11</b> Function practice</a>
<ul>
<li class="chapter" data-level="11.1" data-path="function-practice.html"><a href="function-practice.html#part-1-pure-practice-1"><i class="fa fa-check"></i><b>11.1</b> Part 1: Pure practice</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="function-practice.html"><a href="function-practice.html#practice-function-1-hello-world-1"><i class="fa fa-check"></i><b>11.1.1</b> Practice Function 1: Hello world!</a></li>
<li class="chapter" data-level="11.1.2" data-path="function-practice.html"><a href="function-practice.html#practice-function-2-hello-_______-1"><i class="fa fa-check"></i><b>11.1.2</b> Practice Function 2: Hello _______ !</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="function-practice.html"><a href="function-practice.html#practice-function-3-make-you-own-natural-log-function-1"><i class="fa fa-check"></i><b>11.2</b> Practice Function 3: Make you own natural log function</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html"><i class="fa fa-check"></i><b>12</b> for() loops in R</a>
<ul>
<li class="chapter" data-level="12.1" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#introduction-4"><i class="fa fa-check"></i><b>12.1</b> Introduction</a></li>
<li class="chapter" data-level="12.2" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#reminders-1"><i class="fa fa-check"></i><b>12.2</b> Reminders</a></li>
<li class="chapter" data-level="12.3" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-1-the-hard-coded-for-loop-aka-the-dont-actually-do-this-for-loop-1"><i class="fa fa-check"></i><b>12.3</b> For loop 1: The hard-coded for loop (aka, the don’t actually do this for loop)</a></li>
<li class="chapter" data-level="12.4" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-classic-for-loop-1"><i class="fa fa-check"></i><b>12.4</b> For loop 3: the classic for loop</a></li>
<li class="chapter" data-level="12.5" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-3-the-power-user-for-loop-1"><i class="fa fa-check"></i><b>12.5</b> For loop 3: The Power-user for loop</a></li>
<li class="chapter" data-level="12.6" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#for-loop-4-a-for-loop-for-all-purposes-1"><i class="fa fa-check"></i><b>12.6</b> For loop 4: A for loop for all purposes</a></li>
<li class="chapter" data-level="12.7" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#advanced-aside-if-else-and-next-statements-1"><i class="fa fa-check"></i><b>12.7</b> Advanced aside: if(), else and next statements</a></li>
<li class="chapter" data-level="12.8" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#on-avoiding-for-loops-1"><i class="fa fa-check"></i><b>12.8</b> On avoiding for loops</a></li>
<li class="chapter" data-level="12.9" data-path="for-loops-in-r.html"><a href="for-loops-in-r.html#challenge-1"><i class="fa fa-check"></i><b>12.9</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="random-number-generation.html"><a href="random-number-generation.html"><i class="fa fa-check"></i><b>13</b> Random number generation</a>
<ul>
<li class="chapter" data-level="13.1" data-path="random-number-generation.html"><a href="random-number-generation.html#introduction-5"><i class="fa fa-check"></i><b>13.1</b> Introduction</a></li>
<li class="chapter" data-level="13.2" data-path="random-number-generation.html"><a href="random-number-generation.html#using-set.seed-to-make-simulations-reproducible"><i class="fa fa-check"></i><b>13.2</b> Using set.seed() to make simulations reproducible</a></li>
<li class="chapter" data-level="13.3" data-path="random-number-generation.html"><a href="random-number-generation.html#types-of-random-number-generators"><i class="fa fa-check"></i><b>13.3</b> Types of random number generators</a></li>
<li class="chapter" data-level="13.4" data-path="random-number-generation.html"><a href="random-number-generation.html#reporting-simulations"><i class="fa fa-check"></i><b>13.4</b> Reporting simulations</a></li>
<li class="chapter" data-level="13.5" data-path="random-number-generation.html"><a href="random-number-generation.html#notes"><i class="fa fa-check"></i><b>13.5</b> Notes:</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html"><i class="fa fa-check"></i><b>14</b> Sample test question-Randomization</a>
<ul>
<li class="chapter" data-level="14.1" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#sample-test-question"><i class="fa fa-check"></i><b>14.1</b> Sample test question:</a></li>
<li class="chapter" data-level="14.2" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#your-answer"><i class="fa fa-check"></i><b>14.2</b> Your answer:</a></li>
<li class="chapter" data-level="14.3" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-0"><i class="fa fa-check"></i><b>14.3</b> Hint 0</a></li>
<li class="chapter" data-level="14.4" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-1"><i class="fa fa-check"></i><b>14.4</b> Hint 1:</a></li>
<li class="chapter" data-level="14.5" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-2-follow-up-to-hint-0"><i class="fa fa-check"></i><b>14.5</b> Hint 2 (follow up to Hint 0)</a></li>
<li class="chapter" data-level="14.6" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#hint-3"><i class="fa fa-check"></i><b>14.6</b> Hint 3:</a></li>
<li class="chapter" data-level="14.7" data-path="sample-test-question-randomization.html"><a href="sample-test-question-randomization.html#answer"><i class="fa fa-check"></i><b>14.7</b> Answer</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html"><i class="fa fa-check"></i><b>15</b> Simulating random amino acid sequences</a>
<ul>
<li class="chapter" data-level="15.1" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#selecting-random-letters-from-a-vector"><i class="fa fa-check"></i><b>15.1</b> Selecting random letters from a vector</a></li>
<li class="chapter" data-level="15.2" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-letters-from-the-whole-alphabet"><i class="fa fa-check"></i><b>15.2</b> Select random letters from the whole alphabet</a></li>
<li class="chapter" data-level="15.3" data-path="simulating-random-amino-acid-sequences.html"><a href="simulating-random-amino-acid-sequences.html#select-random-amino-acids-to-build-polypeptide"><i class="fa fa-check"></i><b>15.3</b> Select random amino acids to build polypeptide</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html"><i class="fa fa-check"></i><b>16</b> How to review this book</a>
<ul>
<li class="chapter" data-level="16.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#file-format"><i class="fa fa-check"></i><b>16.1</b> File format</a></li>
<li class="chapter" data-level="16.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#what-can-you-do"><i class="fa fa-check"></i><b>16.2</b> What can you do?</a></li>
<li class="chapter" data-level="16.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#adding-comments"><i class="fa fa-check"></i><b>16.3</b> Adding comments</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#html-tagged-comments"><i class="fa fa-check"></i><b>16.3.1</b> HTML-tagged comments</a></li>
<li class="chapter" data-level="16.3.2" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#code-chunk-comments"><i class="fa fa-check"></i><b>16.3.2</b> Code chunk comments</a></li>
<li class="chapter" data-level="16.3.3" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#keys-to-exercise"><i class="fa fa-check"></i><b>16.3.3</b> Keys to exercise</a></li>
<li class="chapter" data-level="16.3.4" data-path="how-to-review-this-book.html"><a href="how-to-review-this-book.html#problem---fix-this-code"><i class="fa fa-check"></i><b>16.3.4</b> Problem - fix this code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="test-audio-link.html"><a href="test-audio-link.html"><i class="fa fa-check"></i><b>17</b> Test audio link</a></li>
<li class="chapter" data-level="18" data-path="test-shiny-app.html"><a href="test-shiny-app.html"><i class="fa fa-check"></i><b>18</b> Test shiny app</a></li>
<li class="chapter" data-level="19" data-path="test-youtube-vidoe.html"><a href="test-youtube-vidoe.html"><i class="fa fa-check"></i><b>19</b> Test YouTube vidoe</a></li>
<li class="chapter" data-level="20" data-path="test-google-doc-insertion.html"><a href="test-google-doc-insertion.html"><i class="fa fa-check"></i><b>20</b> Test google doc insertion</a></li>
<li class="chapter" data-level="21" data-path="datacamp-test.html"><a href="datacamp-test.html"><i class="fa fa-check"></i><b>21</b> Datacamp test</a></li>
<li class="chapter" data-level="22" data-path="an-example-of-a-plotly-graph.html"><a href="an-example-of-a-plotly-graph.html"><i class="fa fa-check"></i><b>22</b> An example of a plotly graph</a></li>
<li class="chapter" data-level="23" data-path="directly-embedding-of-a-libretext-page.html"><a href="directly-embedding-of-a-libretext-page.html"><i class="fa fa-check"></i><b>23</b> Directly embedding of a libretext page</a></li>
<li class="chapter" data-level="24" data-path="libretext-image.html"><a href="libretext-image.html"><i class="fa fa-check"></i><b>24</b> Libretext image</a>
<ul>
<li class="chapter" data-level="24.1" data-path="libretext-image.html"><a href="libretext-image.html#prelude-to-population-and-community-ecology"><i class="fa fa-check"></i><b>24.1</b> 45.0: Prelude to Population and Community Ecology</a></li>
<li class="chapter" data-level="24.2" data-path="libretext-image.html"><a href="libretext-image.html#note"><i class="fa fa-check"></i><b>24.2</b> Note</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="downloading-r.html"><a href="downloading-r.html"><i class="fa fa-check"></i><b>25</b> Downloading R</a>
<ul>
<li class="chapter" data-level="25.1" data-path="downloading-r.html"><a href="downloading-r.html#preface-1"><i class="fa fa-check"></i><b>25.1</b> Preface</a></li>
<li class="chapter" data-level="25.2" data-path="downloading-r.html"><a href="downloading-r.html#introduction-to-r"><i class="fa fa-check"></i><b>25.2</b> Introduction to R</a></li>
<li class="chapter" data-level="25.3" data-path="downloading-r.html"><a href="downloading-r.html#installing-r"><i class="fa fa-check"></i><b>25.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="25.3.1" data-path="downloading-r.html"><a href="downloading-r.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>25.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="25.3.2" data-path="downloading-r.html"><a href="downloading-r.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>25.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="25.4" data-path="downloading-r.html"><a href="downloading-r.html#starting-r"><i class="fa fa-check"></i><b>25.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html"><i class="fa fa-check"></i><b>26</b> Installing the RStudio IDE</a>
<ul>
<li class="chapter" data-level="26.1" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>26.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="26.2" data-path="installing-the-rstudio-ide.html"><a href="installing-the-rstudio-ide.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>26.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>27</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="27.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>27.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="27.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>27.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="27.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>27.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="28" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>28</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="28.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>28.1</b> Bioconductor</a></li>
<li class="chapter" data-level="28.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>28.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="28.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>28.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="28.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>28.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="28.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>28.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="28.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>28.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="28.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>28.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html"><i class="fa fa-check"></i><b>29</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="29.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vocabulary-1"><i class="fa fa-check"></i><b>29.1</b> Vocabulary</a></li>
<li class="chapter" data-level="29.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#r-functions"><i class="fa fa-check"></i><b>29.2</b> R functions</a></li>
<li class="chapter" data-level="29.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#interacting-with-r"><i class="fa fa-check"></i><b>29.3</b> Interacting with R</a></li>
<li class="chapter" data-level="29.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#variables-in-r"><i class="fa fa-check"></i><b>29.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="29.4.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>29.4.1</b> Vectors</a></li>
<li class="chapter" data-level="29.4.2" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#vector-indexing"><i class="fa fa-check"></i><b>29.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="29.4.3" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#character-vectors"><i class="fa fa-check"></i><b>29.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="29.4.4" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#lists"><i class="fa fa-check"></i><b>29.4.4</b> Lists</a></li>
<li class="chapter" data-level="29.4.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#tables"><i class="fa fa-check"></i><b>29.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="29.5" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#arguments"><i class="fa fa-check"></i><b>29.5</b> Arguments</a></li>
<li class="chapter" data-level="29.6" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#help-files-with-help-and"><i class="fa fa-check"></i><b>29.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="29.7" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>29.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="29.8" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#more-on-functions"><i class="fa fa-check"></i><b>29.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="29.8.1" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#writing-your-own-functions"><i class="fa fa-check"></i><b>29.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="29.9" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#quiting-r"><i class="fa fa-check"></i><b>29.9</b> Quiting R</a></li>
<li class="chapter" data-level="29.10" data-path="a-brief-introduction-to-r.html"><a href="a-brief-introduction-to-r.html#links-and-further-reading"><i class="fa fa-check"></i><b>29.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html"><i class="fa fa-check"></i><b>30</b> Programming in R: for loops</a>
<ul>
<li class="chapter" data-level="30.1" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#preface-2"><i class="fa fa-check"></i><b>30.1</b> Preface</a></li>
<li class="chapter" data-level="30.2" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#vocab"><i class="fa fa-check"></i><b>30.2</b> Vocab</a></li>
<li class="chapter" data-level="30.3" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#functions-1"><i class="fa fa-check"></i><b>30.3</b> Functions</a></li>
<li class="chapter" data-level="30.4" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#basic-for-loops-in-in-r"><i class="fa fa-check"></i><b>30.4</b> Basic for loops in in R</a></li>
<li class="chapter" data-level="30.5" data-path="programming-in-r-for-loops.html"><a href="programming-in-r-for-loops.html#challenge-complicated-vectors-of-values"><i class="fa fa-check"></i><b>30.5</b> Challenge: complicated vectors of values</a></li>
</ul></li>
<li class="chapter" data-level="31" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html"><i class="fa fa-check"></i><b>31</b> Mini tutorial: Vectors in R</a>
<ul>
<li class="chapter" data-level="31.1" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#preface-3"><i class="fa fa-check"></i><b>31.1</b> Preface</a></li>
<li class="chapter" data-level="31.2" data-path="mini-tutorial-vectors-in-r.html"><a href="mini-tutorial-vectors-in-r.html#vocab-1"><i class="fa fa-check"></i><b>31.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="32" data-path="functions-2.html"><a href="functions-2.html"><i class="fa fa-check"></i><b>32</b> Functions</a>
<ul>
<li class="chapter" data-level="32.1" data-path="functions-2.html"><a href="functions-2.html#vectors-in-r"><i class="fa fa-check"></i><b>32.1</b> Vectors in R</a></li>
<li class="chapter" data-level="32.2" data-path="functions-2.html"><a href="functions-2.html#math-on-vectors"><i class="fa fa-check"></i><b>32.2</b> Math on vectors</a></li>
<li class="chapter" data-level="32.3" data-path="functions-2.html"><a href="functions-2.html#functions-on-vectors"><i class="fa fa-check"></i><b>32.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="32.4" data-path="functions-2.html"><a href="functions-2.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>32.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="32.5" data-path="functions-2.html"><a href="functions-2.html#subsetting-vectors"><i class="fa fa-check"></i><b>32.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="32.6" data-path="functions-2.html"><a href="functions-2.html#sequences-of-numbers"><i class="fa fa-check"></i><b>32.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="32.7" data-path="functions-2.html"><a href="functions-2.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>32.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="32.8" data-path="functions-2.html"><a href="functions-2.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>32.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="33" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html"><i class="fa fa-check"></i><b>33</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="33.1" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#preface-4"><i class="fa fa-check"></i><b>33.1</b> Preface</a></li>
<li class="chapter" data-level="33.2" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#plotting-numeric-data"><i class="fa fa-check"></i><b>33.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="33.3" data-path="plotting-vectors-in-base-r.html"><a href="plotting-vectors-in-base-r.html#other-plotting-packages"><i class="fa fa-check"></i><b>33.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="34" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html"><i class="fa fa-check"></i><b>34</b> Programming in R: functions</a>
<ul>
<li class="chapter" data-level="34.1" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#preface-5"><i class="fa fa-check"></i><b>34.1</b> Preface</a></li>
<li class="chapter" data-level="34.2" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#vocab-2"><i class="fa fa-check"></i><b>34.2</b> Vocab</a></li>
<li class="chapter" data-level="34.3" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-3"><i class="fa fa-check"></i><b>34.3</b> Functions</a></li>
<li class="chapter" data-level="34.4" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#functions-in-r"><i class="fa fa-check"></i><b>34.4</b> Functions in R</a></li>
<li class="chapter" data-level="34.5" data-path="programming-in-r-functions.html"><a href="programming-in-r-functions.html#comments-in-r"><i class="fa fa-check"></i><b>34.5</b> Comments in R</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="fasta-files.html"><a href="fasta-files.html"><i class="fa fa-check"></i><b>35</b> FASTA Files</a>
<ul>
<li class="chapter" data-level="35.1" data-path="fasta-files.html"><a href="fasta-files.html#example-fasta-file"><i class="fa fa-check"></i><b>35.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="35.2" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>35.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="35.3" data-path="fasta-files.html"><a href="fasta-files.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>35.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="35.4" data-path="fasta-files.html"><a href="fasta-files.html#fastq-format"><i class="fa fa-check"></i><b>35.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>36</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="36.0.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#functions-4"><i class="fa fa-check"></i><b>36.0.1</b> Functions</a></li>
<li class="chapter" data-level="36.0.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#softwarewebsites"><i class="fa fa-check"></i><b>36.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="36.0.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-vocabulary"><i class="fa fa-check"></i><b>36.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="36.0.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#file-types"><i class="fa fa-check"></i><b>36.0.4</b> File types</a></li>
<li class="chapter" data-level="36.0.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>36.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="36.0.6" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>36.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="36.0.7" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>36.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="36.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>36.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="36.1.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>36.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="36.1.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>36.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="36.1.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#fasta-file-format"><i class="fa fa-check"></i><b>36.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="36.1.4" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>36.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="36.1.5" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>36.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="36.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>36.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="36.3" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#next-steps"><i class="fa fa-check"></i><b>36.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html"><i class="fa fa-check"></i><b>37</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="37.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#preliminaries-2"><i class="fa fa-check"></i><b>37.1</b> Preliminaries</a></li>
<li class="chapter" data-level="37.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>37.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="37.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>37.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="37.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r-1.html"><a href="downloading-dna-sequences-as-fasta-files-in-r-1.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>37.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="38" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html"><i class="fa fa-check"></i><b>38</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="38.1" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#preliminaries-3"><i class="fa fa-check"></i><b>38.1</b> Preliminaries</a></li>
<li class="chapter" data-level="38.2" data-path="downloading-protein-sequences-in-r.html"><a href="downloading-protein-sequences-in-r.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>38.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html"><i class="fa fa-check"></i><b>39</b> Sequence dotplots in <em>R</em></a>
<ul>
<li class="chapter" data-level="39.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#preliminaries-4"><i class="fa fa-check"></i><b>39.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="39.1.1" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#download-sequences"><i class="fa fa-check"></i><b>39.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="39.2" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualzing-two-identical-sequences"><i class="fa fa-check"></i><b>39.2</b> Visualzing two identical sequences</a></li>
<li class="chapter" data-level="39.3" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#visualizing-repeats"><i class="fa fa-check"></i><b>39.3</b> Visualizing repeats</a></li>
<li class="chapter" data-level="39.4" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#inversions"><i class="fa fa-check"></i><b>39.4</b> Inversions</a></li>
<li class="chapter" data-level="39.5" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#translocations"><i class="fa fa-check"></i><b>39.5</b> Translocations</a></li>
<li class="chapter" data-level="39.6" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#random-sequence"><i class="fa fa-check"></i><b>39.6</b> Random sequence</a></li>
<li class="chapter" data-level="39.7" data-path="sequence-dotplots-in-r.html"><a href="sequence-dotplots-in-r.html#comparing-two-real-sequences-using-a-dotplot"><i class="fa fa-check"></i><b>39.7</b> Comparing two real sequences using a dotplot</a></li>
</ul></li>
<li class="chapter" data-level="40" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html"><i class="fa fa-check"></i><b>40</b> Global proteins aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="40.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#preliminaries-5"><i class="fa fa-check"></i><b>40.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="40.1.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#download-sequences-1"><i class="fa fa-check"></i><b>40.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="40.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-dna-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>40.2</b> Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm</a></li>
<li class="chapter" data-level="40.3" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#pairwise-global-alignment-of-protein-sequences-using-the-needleman-wunsch-algorithm"><i class="fa fa-check"></i><b>40.3</b> Pairwise global alignment of protein sequences using the Needleman-Wunsch algorithm</a>
<ul>
<li class="chapter" data-level="40.3.1" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#protein-score-matrices"><i class="fa fa-check"></i><b>40.3.1</b> Protein score matrices</a></li>
<li class="chapter" data-level="40.3.2" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#example-protein-alignment"><i class="fa fa-check"></i><b>40.3.2</b> Example protein alignment</a></li>
</ul></li>
<li class="chapter" data-level="40.4" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#aligning-uniprot-sequences"><i class="fa fa-check"></i><b>40.4</b> Aligning UniProt sequences</a></li>
<li class="chapter" data-level="40.5" data-path="global-proteins-aligments-in-r.html"><a href="global-proteins-aligments-in-r.html#viewing-a-long-pairwise-alignment"><i class="fa fa-check"></i><b>40.5</b> Viewing a long pairwise alignment</a></li>
</ul></li>
<li class="chapter" data-level="41" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html"><i class="fa fa-check"></i><b>41</b> Local protein aligments in <em>R</em></a>
<ul>
<li class="chapter" data-level="41.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#preliminaries-6"><i class="fa fa-check"></i><b>41.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="41.1.1" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#download-sequences-2"><i class="fa fa-check"></i><b>41.1.1</b> Download sequences</a></li>
</ul></li>
<li class="chapter" data-level="41.2" data-path="local-protein-aligments-in-r.html"><a href="local-protein-aligments-in-r.html#pairwise-local-alignment-of-protein-sequences-using-the-smith-waterman-algorithm"><i class="fa fa-check"></i><b>41.2</b> Pairwise local alignment of protein sequences using the Smith-Waterman algorithm</a></li>
</ul></li>
<li class="chapter" data-level="42" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html"><i class="fa fa-check"></i><b>42</b> Alignment by eye in R</a>
<ul>
<li class="chapter" data-level="42.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#preliminaries-7"><i class="fa fa-check"></i><b>42.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="42.1.1" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#packages"><i class="fa fa-check"></i><b>42.1.1</b> Packages</a></li>
<li class="chapter" data-level="42.1.2" data-path="alignment-by-eye-in-r.html"><a href="alignment-by-eye-in-r.html#data"><i class="fa fa-check"></i><b>42.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="43" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html"><i class="fa fa-check"></i><b>43</b> Matrix vocabulary</a>
<ul>
<li class="chapter" data-level="43.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#introduction-6"><i class="fa fa-check"></i><b>43.1</b> Introduction</a></li>
<li class="chapter" data-level="43.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#preliminaries-8"><i class="fa fa-check"></i><b>43.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="43.2.1" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#packages-1"><i class="fa fa-check"></i><b>43.2.1</b> Packages</a></li>
<li class="chapter" data-level="43.2.2" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#vocab-3"><i class="fa fa-check"></i><b>43.2.2</b> Vocab</a></li>
<li class="chapter" data-level="43.2.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#functions-used"><i class="fa fa-check"></i><b>43.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="43.3" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#load-packages"><i class="fa fa-check"></i><b>43.3</b> Load packages</a></li>
<li class="chapter" data-level="43.4" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#matrix-structure"><i class="fa fa-check"></i><b>43.4</b> Matrix structure</a></li>
<li class="chapter" data-level="43.5" data-path="matrix-vocabulary.html"><a href="matrix-vocabulary.html#diagonal-of-a-matrix"><i class="fa fa-check"></i><b>43.5</b> Diagonal of a matrix</a></li>
</ul></li>
<li class="chapter" data-level="44" data-path="alignment-in-r.html"><a href="alignment-in-r.html"><i class="fa fa-check"></i><b>44</b> Alignment in R</a>
<ul>
<li class="chapter" data-level="44.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#preliminaries-9"><i class="fa fa-check"></i><b>44.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="44.1.1" data-path="alignment-in-r.html"><a href="alignment-in-r.html#packages-2"><i class="fa fa-check"></i><b>44.1.1</b> Packages</a></li>
<li class="chapter" data-level="44.1.2" data-path="alignment-in-r.html"><a href="alignment-in-r.html#data-1"><i class="fa fa-check"></i><b>44.1.2</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="45" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html"><i class="fa fa-check"></i><b>45</b> Testing the significance of an alignment</a>
<ul>
<li class="chapter" data-level="45.1" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#calculating-the-statistical-significance-of-a-pairwise-global-alignment"><i class="fa fa-check"></i><b>45.1</b> Calculating the statistical significance of a pairwise global alignment</a></li>
<li class="chapter" data-level="45.2" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#summary-1"><i class="fa fa-check"></i><b>45.2</b> Summary</a></li>
<li class="chapter" data-level="45.3" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#links-and-further-reading-1"><i class="fa fa-check"></i><b>45.3</b> Links and Further Reading</a></li>
<li class="chapter" data-level="45.4" data-path="testing-the-significance-of-an-alignment.html"><a href="testing-the-significance-of-an-alignment.html#exercises-1"><i class="fa fa-check"></i><b>45.4</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="46" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html"><i class="fa fa-check"></i><b>46</b> Retrieving multiple sequences in R</a>
<ul>
<li class="chapter" data-level="46.1" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#prelminaries"><i class="fa fa-check"></i><b>46.1</b> Prelminaries</a></li>
<li class="chapter" data-level="46.2" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#retrieving-a-set-of-sequences-from-uniprot"><i class="fa fa-check"></i><b>46.2</b> Retrieving a set of sequences from UniProt</a></li>
<li class="chapter" data-level="46.3" data-path="retrieving-multiple-sequences-in-r.html"><a href="retrieving-multiple-sequences-in-r.html#downloading-sequences-in-bulk"><i class="fa fa-check"></i><b>46.3</b> Downloading sequences in bulk</a></li>
</ul></li>
<li class="chapter" data-level="47" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html"><i class="fa fa-check"></i><b>47</b> Manipulating matrices and vectors: worked example</a>
<ul>
<li class="chapter" data-level="47.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#introduction-7"><i class="fa fa-check"></i><b>47.1</b> Introduction</a></li>
<li class="chapter" data-level="47.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#preliminaries-10"><i class="fa fa-check"></i><b>47.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="47.2.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-3"><i class="fa fa-check"></i><b>47.2.1</b> Packages</a></li>
<li class="chapter" data-level="47.2.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#vocab-4"><i class="fa fa-check"></i><b>47.2.2</b> Vocab</a></li>
<li class="chapter" data-level="47.2.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#functions-used-1"><i class="fa fa-check"></i><b>47.2.3</b> Functions used</a></li>
<li class="chapter" data-level="47.2.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#packages-4"><i class="fa fa-check"></i><b>47.2.4</b> Packages</a></li>
<li class="chapter" data-level="47.2.5" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#data-2"><i class="fa fa-check"></i><b>47.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="47.3" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#matrix-elements-can-be-accessed-with-square-bracket-notation"><i class="fa fa-check"></i><b>47.3</b> Matrix elements can be accessed with square bracket notation</a></li>
<li class="chapter" data-level="47.4" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#r-object-names"><i class="fa fa-check"></i><b>47.4</b> R object names</a>
<ul>
<li class="chapter" data-level="47.4.1" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-vectors"><i class="fa fa-check"></i><b>47.4.1</b> Accessing items in names vectors</a></li>
<li class="chapter" data-level="47.4.2" data-path="manipulating-matrices-and-vectors-worked-example.html"><a href="manipulating-matrices-and-vectors-worked-example.html#accessing-items-in-names-matrices"><i class="fa fa-check"></i><b>47.4.2</b> Accessing items in names matrices</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="48" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html"><i class="fa fa-check"></i><b>48</b> Multiple sequence alignment in R</a>
<ul>
<li class="chapter" data-level="48.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#preliminaries-11"><i class="fa fa-check"></i><b>48.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="48.1.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#packages-5"><i class="fa fa-check"></i><b>48.1.1</b> Packages</a></li>
<li class="chapter" data-level="48.1.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#functions-5"><i class="fa fa-check"></i><b>48.1.2</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="48.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>48.2</b> Multiple sequence alignment (MSA)</a></li>
<li class="chapter" data-level="48.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#make-msa-with-msa"><i class="fa fa-check"></i><b>48.3</b> Make MSA with msa()</a>
<ul>
<li class="chapter" data-level="48.3.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#data-preparation"><i class="fa fa-check"></i><b>48.3.1</b> Data preparation</a></li>
</ul></li>
<li class="chapter" data-level="48.4" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-your-msa"><i class="fa fa-check"></i><b>48.4</b> Viewing your MSA</a>
<ul>
<li class="chapter" data-level="48.4.1" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#viewing-a-long-multiple-alignment-in-the-r-console."><i class="fa fa-check"></i><b>48.4.1</b> Viewing a long multiple alignment in the R console.</a></li>
<li class="chapter" data-level="48.4.2" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#visualizing-alignments-as-an-r-plot"><i class="fa fa-check"></i><b>48.4.2</b> Visualizing alignments as an R plot</a></li>
<li class="chapter" data-level="48.4.3" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#optional-print-msa-to-pdf"><i class="fa fa-check"></i><b>48.4.3</b> OPTIONAL: Print MSA to PDF</a></li>
</ul></li>
<li class="chapter" data-level="48.5" data-path="multiple-sequence-alignment-in-r.html"><a href="multiple-sequence-alignment-in-r.html#discarding-very-poorly-conserved-regions-from-an-alignment"><i class="fa fa-check"></i><b>48.5</b> Discarding very poorly conserved regions from an alignment</a></li>
</ul></li>
<li class="chapter" data-level="49" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html"><i class="fa fa-check"></i><b>49</b> The BLOSUM scoring matrix in R</a>
<ul>
<li class="chapter" data-level="49.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#introduction-8"><i class="fa fa-check"></i><b>49.1</b> Introduction</a></li>
<li class="chapter" data-level="49.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#preliminaries-12"><i class="fa fa-check"></i><b>49.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="49.2.1" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#packages-6"><i class="fa fa-check"></i><b>49.2.1</b> Packages</a></li>
<li class="chapter" data-level="49.2.2" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#vocab-5"><i class="fa fa-check"></i><b>49.2.2</b> Vocab</a></li>
<li class="chapter" data-level="49.2.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#functions-used-2"><i class="fa fa-check"></i><b>49.2.3</b> Functions used</a></li>
</ul></li>
<li class="chapter" data-level="49.3" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#load-packages-1"><i class="fa fa-check"></i><b>49.3</b> Load packages</a></li>
<li class="chapter" data-level="49.4" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#the-blosum-matrix"><i class="fa fa-check"></i><b>49.4</b> The BLOSUM matrix</a></li>
<li class="chapter" data-level="49.5" data-path="the-blosum-scoring-matrix-in-r.html"><a href="the-blosum-scoring-matrix-in-r.html#visualizing-the-blosum-matrix"><i class="fa fa-check"></i><b>49.5</b> Visualizing the BLOSUM matrix</a></li>
</ul></li>
<li class="chapter" data-level="50" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>50</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="51" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>51</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="51.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-9"><i class="fa fa-check"></i><b>51.1</b> Introduction</a></li>
<li class="chapter" data-level="51.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>51.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="51.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-6"><i class="fa fa-check"></i><b>51.2.1</b> Vocab</a></li>
<li class="chapter" data-level="51.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>51.2.2</b> R functions</a></li>
<li class="chapter" data-level="51.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>51.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="51.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>51.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="51.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>51.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="51.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>51.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="51.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>51.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="51.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>51.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="51.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>51.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="51.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>51.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="51.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>51.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="51.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>51.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="51.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>51.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="51.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>51.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="51.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>51.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="52" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html"><i class="fa fa-check"></i><b>52</b> Calculating genetic distances between sequences</a>
<ul>
<li class="chapter" data-level="52.1" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#preliminaries-13"><i class="fa fa-check"></i><b>52.1</b> Preliminaries</a></li>
<li class="chapter" data-level="52.2" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#introduction-10"><i class="fa fa-check"></i><b>52.2</b> Introduction</a></li>
<li class="chapter" data-level="52.3" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculating-genetic-distances-between-dnamrna-sequences"><i class="fa fa-check"></i><b>52.3</b> Calculating genetic distances between DNA/mRNA sequences</a></li>
<li class="chapter" data-level="52.4" data-path="calculating-genetic-distances-between-sequences.html"><a href="calculating-genetic-distances-between-sequences.html#calculationg-genetic-distance"><i class="fa fa-check"></i><b>52.4</b> Calculationg genetic distance</a></li>
</ul></li>
<li class="chapter" data-level="53" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html"><i class="fa fa-check"></i><b>53</b> Unrooted neighbor-joining phylogenetic trees</a>
<ul>
<li class="chapter" data-level="53.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#preliminaries-14"><i class="fa fa-check"></i><b>53.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="53.1.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-functions"><i class="fa fa-check"></i><b>53.1.1</b> Key functions</a></li>
<li class="chapter" data-level="53.1.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#key-vocab"><i class="fa fa-check"></i><b>53.1.2</b> Key vocab</a></li>
</ul></li>
<li class="chapter" data-level="53.2" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#building-an-unrooted-phylogenetic-tree-for-protein-sequences"><i class="fa fa-check"></i><b>53.2</b> Building an unrooted phylogenetic tree for protein sequences</a>
<ul>
<li class="chapter" data-level="53.2.1" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#build-tree-with-unrooted_nj_tree"><i class="fa fa-check"></i><b>53.2.1</b> Build tree with <code>unrooted_NJ_tree()</code></a></li>
</ul></li>
<li class="chapter" data-level="53.3" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#boostrap-values-indicate-support-for-clades"><i class="fa fa-check"></i><b>53.3</b> Boostrap values indicate support for clades</a></li>
<li class="chapter" data-level="53.4" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#branch-lengths-indicate-divergence-between-sequences"><i class="fa fa-check"></i><b>53.4</b> Branch lengths indicate divergence between sequences</a></li>
<li class="chapter" data-level="53.5" data-path="unrooted-neighbor-joining-phylogenetic-trees.html"><a href="unrooted-neighbor-joining-phylogenetic-trees.html#unrooted-trees-lack-an-outgroup"><i class="fa fa-check"></i><b>53.5</b> Unrooted trees lack an outgroup</a></li>
</ul></li>
<li class="chapter" data-level="54" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html"><i class="fa fa-check"></i><b>54</b> Local variation in GC content</a>
<ul>
<li class="chapter" data-level="54.0.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#note-on-the-biology-in-this-section"><i class="fa fa-check"></i><b>54.0.1</b> Note on the biology in this section</a></li>
<li class="chapter" data-level="54.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#vocabulary-2"><i class="fa fa-check"></i><b>54.1</b> Vocabulary</a></li>
<li class="chapter" data-level="54.2" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#reading-sequence-data-with-rentrezentrez_fetch"><i class="fa fa-check"></i><b>54.2</b> Reading sequence data with rentrez::entrez_fetch</a></li>
<li class="chapter" data-level="54.3" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#local-variation-in-gc-content-1"><i class="fa fa-check"></i><b>54.3</b> Local variation in GC content</a></li>
<li class="chapter" data-level="54.4" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-analysis-of-gc-content"><i class="fa fa-check"></i><b>54.4</b> A sliding window analysis of GC content</a></li>
<li class="chapter" data-level="54.5" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#a-sliding-window-plot-of-gc-content"><i class="fa fa-check"></i><b>54.5</b> A sliding window plot of GC content</a>
<ul>
<li class="chapter" data-level="54.5.1" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#over-represented-and-under-represented-dna-words"><i class="fa fa-check"></i><b>54.5.1</b> Over-represented and under-represented DNA words</a></li>
</ul></li>
<li class="chapter" data-level="54.6" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#summary-2"><i class="fa fa-check"></i><b>54.6</b> Summary</a></li>
<li class="chapter" data-level="54.7" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#further-reading"><i class="fa fa-check"></i><b>54.7</b> Further Reading</a></li>
<li class="chapter" data-level="54.8" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#acknowledgements-1"><i class="fa fa-check"></i><b>54.8</b> Acknowledgements</a></li>
<li class="chapter" data-level="54.9" data-path="local-variation-in-gc-content.html"><a href="local-variation-in-gc-content.html#exercises-2"><i class="fa fa-check"></i><b>54.9</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="55" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html"><i class="fa fa-check"></i><b>55</b> Computational gene finding</a>
<ul>
<li class="chapter" data-level="55.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#preliminaries-15"><i class="fa fa-check"></i><b>55.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="55.1.1" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#packages-7"><i class="fa fa-check"></i><b>55.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="55.2" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#the-genetic-code"><i class="fa fa-check"></i><b>55.2</b> The genetic code</a></li>
<li class="chapter" data-level="55.3" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-start-and-stop-codons-in-a-dna-sequence"><i class="fa fa-check"></i><b>55.3</b> Finding start and stop codons in a DNA sequence</a></li>
<li class="chapter" data-level="55.4" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#reading-frames"><i class="fa fa-check"></i><b>55.4</b> Reading frames</a></li>
<li class="chapter" data-level="55.5" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-forward-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>55.5</b> Finding open reading frames on the forward strand of a DNA sequence</a></li>
<li class="chapter" data-level="55.6" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#predicting-the-protein-sequence-for-an-orf"><i class="fa fa-check"></i><b>55.6</b> Predicting the protein sequence for an ORF</a></li>
<li class="chapter" data-level="55.7" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#finding-open-reading-frames-on-the-reverse-strand-of-a-dna-sequence"><i class="fa fa-check"></i><b>55.7</b> Finding open reading frames on the reverse strand of a DNA sequence</a></li>
<li class="chapter" data-level="55.8" data-path="computational-gene-finding.html"><a href="computational-gene-finding.html#lengths-of-open-reading-frames"><i class="fa fa-check"></i><b>55.8</b> Lengths of open reading frames</a></li>
</ul></li>
<li class="chapter" data-level="56" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html"><i class="fa fa-check"></i><b>56</b> Significance of ORFs</a>
<ul>
<li class="chapter" data-level="56.1" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#preliminaries-16"><i class="fa fa-check"></i><b>56.1</b> Preliminaries</a></li>
<li class="chapter" data-level="56.2" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#identifying-significant-open-reading-frames"><i class="fa fa-check"></i><b>56.2</b> Identifying significant open reading frames</a></li>
<li class="chapter" data-level="56.3" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#summary-3"><i class="fa fa-check"></i><b>56.3</b> Summary</a></li>
<li class="chapter" data-level="56.4" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#links-and-further-reading-2"><i class="fa fa-check"></i><b>56.4</b> Links and Further Reading</a></li>
<li class="chapter" data-level="56.5" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#acknowledgements-2"><i class="fa fa-check"></i><b>56.5</b> Acknowledgements</a></li>
<li class="chapter" data-level="56.6" data-path="significance-of-orfs.html"><a href="significance-of-orfs.html#exercises-3"><i class="fa fa-check"></i><b>56.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="57" data-path="comparative-genomics.html"><a href="comparative-genomics.html"><i class="fa fa-check"></i><b>57</b> Comparative Genomics</a>
<ul>
<li class="chapter" data-level="57.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#preliminaries-17"><i class="fa fa-check"></i><b>57.1</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="57.1.1" data-path="comparative-genomics.html"><a href="comparative-genomics.html#packages-8"><i class="fa fa-check"></i><b>57.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="57.2" data-path="comparative-genomics.html"><a href="comparative-genomics.html#introduction-11"><i class="fa fa-check"></i><b>57.2</b> Introduction</a></li>
<li class="chapter" data-level="57.3" data-path="comparative-genomics.html"><a href="comparative-genomics.html#using-the-biomart-r-library-to-query-the-ensembl-database"><i class="fa fa-check"></i><b>57.3</b> Using the biomaRt R Library to Query the Ensembl Database</a></li>
<li class="chapter" data-level="57.4" data-path="comparative-genomics.html"><a href="comparative-genomics.html#comparing-the-number-of-genes-in-two-species"><i class="fa fa-check"></i><b>57.4</b> Comparing the number of genes in two species</a></li>
<li class="chapter" data-level="57.5" data-path="comparative-genomics.html"><a href="comparative-genomics.html#identifying-homologous-genes-between-two-species"><i class="fa fa-check"></i><b>57.5</b> Identifying homologous genes between two species</a></li>
</ul></li>
<li class="chapter" data-level="58" data-path="summary-4.html"><a href="summary-4.html"><i class="fa fa-check"></i><b>58</b> Summary</a></li>
<li class="chapter" data-level="59" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html"><i class="fa fa-check"></i><b>59</b> #Links and Further Reading</a>
<ul>
<li class="chapter" data-level="59.1" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#acknowledgements-3"><i class="fa fa-check"></i><b>59.1</b> Acknowledgements</a></li>
<li class="chapter" data-level="59.2" data-path="links-and-further-reading-3.html"><a href="links-and-further-reading-3.html#exercises-4"><i class="fa fa-check"></i><b>59.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="60" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html"><i class="fa fa-check"></i><b>60</b> Hidden Markov Models</a>
<ul>
<li class="chapter" data-level="60.1" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-multinomial-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>60.1</b> A multinomial model of DNA sequence evolution</a></li>
<li class="chapter" data-level="60.2" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-multinomial-model"><i class="fa fa-check"></i><b>60.2</b> Generating a DNA sequence using a multinomial model</a></li>
<li class="chapter" data-level="60.3" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>60.3</b> A Markov model of DNA sequence evolution</a></li>
<li class="chapter" data-level="60.4" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-for-a-markov-model"><i class="fa fa-check"></i><b>60.4</b> The transition matrix for a Markov model</a></li>
<li class="chapter" data-level="60.5" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-markov-model"><i class="fa fa-check"></i><b>60.5</b> Generating a DNA sequence using a Markov model</a></li>
<li class="chapter" data-level="60.6" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-dna-sequence-evolution"><i class="fa fa-check"></i><b>60.6</b> A Hidden Markov Model of DNA sequence evolution</a></li>
<li class="chapter" data-level="60.7" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#the-transition-matrix-and-emission-matrix-for-a-hmm"><i class="fa fa-check"></i><b>60.7</b> The transition matrix and emission matrix for a HMM</a></li>
<li class="chapter" data-level="60.8" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#generating-a-dna-sequence-using-a-hmm"><i class="fa fa-check"></i><b>60.8</b> Generating a DNA sequence using a HMM</a></li>
<li class="chapter" data-level="60.9" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#inferring-the-states-of-a-hmm-that-generated-a-dna-sequence"><i class="fa fa-check"></i><b>60.9</b> Inferring the states of a HMM that generated a DNA sequence</a></li>
<li class="chapter" data-level="60.10" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#a-hidden-markov-model-of-protein-sequence-evolution"><i class="fa fa-check"></i><b>60.10</b> A Hidden Markov Model of protein sequence evolution</a></li>
<li class="chapter" data-level="60.11" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#summary-5"><i class="fa fa-check"></i><b>60.11</b> Summary</a></li>
<li class="chapter" data-level="60.12" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#links-and-further-reading-4"><i class="fa fa-check"></i><b>60.12</b> Links and Further Reading</a></li>
<li class="chapter" data-level="60.13" data-path="hidden-markov-models.html"><a href="hidden-markov-models.html#exercises-5"><i class="fa fa-check"></i><b>60.13</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="61" data-path="blast-summary.html"><a href="blast-summary.html"><i class="fa fa-check"></i><b>61</b> BLAST Summary</a>
<ul>
<li class="chapter" data-level="61.1" data-path="blast-summary.html"><a href="blast-summary.html#blast-overview"><i class="fa fa-check"></i><b>61.1</b> BLAST Overview</a>
<ul>
<li class="chapter" data-level="61.1.1" data-path="blast-summary.html"><a href="blast-summary.html#why-do-a-blast-search"><i class="fa fa-check"></i><b>61.1.1</b> Why do a BLAST search?</a></li>
<li class="chapter" data-level="61.1.2" data-path="blast-summary.html"><a href="blast-summary.html#what-does-blast-let-you-do"><i class="fa fa-check"></i><b>61.1.2</b> What does BLAST let you do?</a></li>
</ul></li>
<li class="chapter" data-level="61.2" data-path="blast-summary.html"><a href="blast-summary.html#blast-recipe"><i class="fa fa-check"></i><b>61.2</b> BLAST recipe</a></li>
<li class="chapter" data-level="61.3" data-path="blast-summary.html"><a href="blast-summary.html#modules-in-the-blast-workflow-and-algorithm"><i class="fa fa-check"></i><b>61.3</b> Modules in the BLAST Workflow and Algorithm</a>
<ul>
<li class="chapter" data-level="61.3.1" data-path="blast-summary.html"><a href="blast-summary.html#module-1-setup-query-sequence-protein-blast-oriented"><i class="fa fa-check"></i><b>61.3.1</b> Module 1: Setup query sequence (Protein BLAST oriented)</a></li>
<li class="chapter" data-level="61.3.2" data-path="blast-summary.html"><a href="blast-summary.html#module-2-scanning-database-for-high-scoring-words"><i class="fa fa-check"></i><b>61.3.2</b> Module 2: Scanning database for high-scoring words</a></li>
<li class="chapter" data-level="61.3.3" data-path="blast-summary.html"><a href="blast-summary.html#module-3-traceback"><i class="fa fa-check"></i><b>61.3.3</b> Module 3: Traceback</a></li>
</ul></li>
<li class="chapter" data-level="61.4" data-path="blast-summary.html"><a href="blast-summary.html#blast-alignment-statistics"><i class="fa fa-check"></i><b>61.4</b> BLAST alignment statistics</a></li>
</ul></li>
<li class="chapter" data-level="62" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html"><i class="fa fa-check"></i><b>62</b> BLAST Alignment score distribution</a>
<ul>
<li class="chapter" data-level="62.1" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#packages-9"><i class="fa fa-check"></i><b>62.1</b> Packages</a></li>
<li class="chapter" data-level="62.2" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#distributions-of-alignment-scores"><i class="fa fa-check"></i><b>62.2</b> Distributions of alignment scores</a></li>
<li class="chapter" data-level="62.3" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-data"><i class="fa fa-check"></i><b>62.3</b> Simulating data</a></li>
<li class="chapter" data-level="62.4" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#lots-of-data-in-biology-is-normal"><i class="fa fa-check"></i><b>62.4</b> Lots of data in biology is normal</a></li>
<li class="chapter" data-level="62.5" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#simulating-the-normal-distribution"><i class="fa fa-check"></i><b>62.5</b> Simulating the normal distribution</a></li>
<li class="chapter" data-level="62.6" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#the-extreme-value-distribution"><i class="fa fa-check"></i><b>62.6</b> The extreme value distribution</a></li>
<li class="chapter" data-level="62.7" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#replicating-figure-4.14"><i class="fa fa-check"></i><b>62.7</b> Replicating Figure 4.14</a></li>
<li class="chapter" data-level="62.8" data-path="blast-alignment-score-distribution.html"><a href="blast-alignment-score-distribution.html#why-do-we-care"><i class="fa fa-check"></i><b>62.8</b> Why do we care?</a></li>
</ul></li>
<li class="chapter" data-level="63" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><i class="fa fa-check"></i><b>63</b> Calculating E values, Bit scores, and P-values from BLAST output</a>
<ul>
<li class="chapter" data-level="63.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-12"><i class="fa fa-check"></i><b>63.1</b> Introduction</a></li>
<li class="chapter" data-level="63.2" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#case-study"><i class="fa fa-check"></i><b>63.2</b> Case Study</a></li>
<li class="chapter" data-level="63.3" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-statistics"><i class="fa fa-check"></i><b>63.3</b> BLAST Statistics</a></li>
<li class="chapter" data-level="63.4" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot"><i class="fa fa-check"></i><b>63.4</b> Dotplot</a>
<ul>
<li class="chapter" data-level="63.4.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#dotplot-in-r---optional-experimental-code"><i class="fa fa-check"></i><b>63.4.1</b> Dotplot in R - optional, experimental code</a></li>
</ul></li>
<li class="chapter" data-level="63.5" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-search-summary"><i class="fa fa-check"></i><b>63.5</b> BLAST “Search Summary”</a></li>
<li class="chapter" data-level="63.6" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#blast-alignments"><i class="fa fa-check"></i><b>63.6</b> BLAST Alignments</a></li>
<li class="chapter" data-level="63.7" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#equations"><i class="fa fa-check"></i><b>63.7</b> Equations</a>
<ul>
<li class="chapter" data-level="63.7.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-value-eq-4.5-pg-142"><i class="fa fa-check"></i><b>63.7.1</b> E value (eq 4.5, pg 142)</a></li>
</ul></li>
<li class="chapter" data-level="63.8" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bit-scores-s"><i class="fa fa-check"></i><b>63.8</b> Bit Scores (S’)</a></li>
<li class="chapter" data-level="63.9" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#e-from-bit-scores"><i class="fa fa-check"></i><b>63.9</b> E from Bit scores</a></li>
<li class="chapter" data-level="63.10" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#p-values"><i class="fa fa-check"></i><b>63.10</b> P values</a></li>
<li class="chapter" data-level="63.11" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#repliminaries"><i class="fa fa-check"></i><b>63.11</b> Repliminaries</a></li>
<li class="chapter" data-level="63.12" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#concepts"><i class="fa fa-check"></i><b>63.12</b> Concepts</a></li>
<li class="chapter" data-level="63.13" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#functions-6"><i class="fa fa-check"></i><b>63.13</b> Functions</a></li>
<li class="chapter" data-level="63.14" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#introduction-13"><i class="fa fa-check"></i><b>63.14</b> Introduction</a></li>
<li class="chapter" data-level="63.15" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees"><i class="fa fa-check"></i><b>63.15</b> Number of rooted trees</a></li>
<li class="chapter" data-level="63.16" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#number-of-rooted-trees-in-r"><i class="fa fa-check"></i><b>63.16</b> Number of rooted trees in R</a>
<ul>
<li class="chapter" data-level="63.16.1" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#unique-topologies-versus-rotations"><i class="fa fa-check"></i><b>63.16.1</b> Unique topologies versus rotations</a></li>
</ul></li>
<li class="chapter" data-level="63.17" data-path="calculating-e-values-bit-scores-and-p-values-from-blast-output.html"><a href="calculating-e-values-bit-scores-and-p-values-from-blast-output.html#bilography"><i class="fa fa-check"></i><b>63.17</b> Bilography</a></li>
</ul></li>
<li class="chapter" data-level="64" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html"><i class="fa fa-check"></i><b>64</b> UPGMA the hard way</a>
<ul>
<li class="chapter" data-level="64.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#introudction-to-upgma"><i class="fa fa-check"></i><b>64.1</b> Introudction to UPGMA</a></li>
<li class="chapter" data-level="64.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#preliminaries-18"><i class="fa fa-check"></i><b>64.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="64.2.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#r-libraries"><i class="fa fa-check"></i><b>64.2.1</b> R Libraries</a></li>
<li class="chapter" data-level="64.2.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#custom-function-well-use"><i class="fa fa-check"></i><b>64.2.2</b> Custom function we’ll use</a></li>
</ul></li>
<li class="chapter" data-level="64.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#data-matrix"><i class="fa fa-check"></i><b>64.3</b> Data matrix</a>
<ul>
<li class="chapter" data-level="64.3.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#building-the-matrix"><i class="fa fa-check"></i><b>64.3.1</b> Building the matrix</a></li>
</ul></li>
<li class="chapter" data-level="64.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#what-does-this-distance-matrix-represent"><i class="fa fa-check"></i><b>64.4</b> What does this distance matrix represent?</a>
<ul>
<li class="chapter" data-level="64.4.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-pairwise-linear-distances"><i class="fa fa-check"></i><b>64.4.1</b> Genetic distances conceptualized as pairwise linear distances</a></li>
<li class="chapter" data-level="64.4.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#genetic-distances-conceptualized-as-2d-map-of-linear-distances"><i class="fa fa-check"></i><b>64.4.2</b> Genetic distances conceptualized as 2D map of linear distances</a></li>
</ul></li>
<li class="chapter" data-level="64.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#from-linear-distances-to-tree-distances"><i class="fa fa-check"></i><b>64.5</b> From linear distances to “tree distances”</a></li>
<li class="chapter" data-level="64.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#the-upgma-algorithm"><i class="fa fa-check"></i><b>64.6</b> The UPGMA algorithm</a>
<ul>
<li class="chapter" data-level="64.6.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-1-find-the-2-taxa-that-are-currently-closest-together"><i class="fa fa-check"></i><b>64.6.1</b> Step 1: find the 2 taxa that are currently closest together</a></li>
<li class="chapter" data-level="64.6.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-2-calculate-the-branch-lengths-connect-the-taxa-to-their-shared-node"><i class="fa fa-check"></i><b>64.6.2</b> Step 2: calculate the branch lengths connect the taxa to their shared node</a></li>
<li class="chapter" data-level="64.6.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-determine-if-the-algorithm-is-done"><i class="fa fa-check"></i><b>64.6.3</b> Step 3a: determine if the algorithm is done</a></li>
<li class="chapter" data-level="64.6.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-3a-continue-on-if-needed-and-combine-the-previous-cluster-into-cluster-u"><i class="fa fa-check"></i><b>64.6.4</b> Step 3a: Continue on if needed and combine the previous cluster into cluster “u”</a></li>
<li class="chapter" data-level="64.6.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-4-determine-the-distance-from-u-to-the-other-remaining-clusters."><i class="fa fa-check"></i><b>64.6.5</b> Step 4: Determine the distance from u to the other remaining clusters.</a></li>
<li class="chapter" data-level="64.6.6" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#step-5-create-a-new-matrix"><i class="fa fa-check"></i><b>64.6.6</b> Step 5: Create a new matrix</a></li>
<li class="chapter" data-level="64.6.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#key-step-calculating-the-distance-from-a-clade-to-the-remaining-taxa"><i class="fa fa-check"></i><b>64.6.7</b> Key step: calculating the distance from a clade to the remaining taxa</a></li>
</ul></li>
<li class="chapter" data-level="64.7" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-2d-representation"><i class="fa fa-check"></i><b>64.7</b> Calculate 2D representation</a></li>
<li class="chapter" data-level="64.8" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---round-1"><i class="fa fa-check"></i><b>64.8</b> Algorithm - round 1</a>
<ul>
<li class="chapter" data-level="64.8.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#determine-minimum-value-in-current-matrix"><i class="fa fa-check"></i><b>64.8.1</b> Determine minimum value in current matrix</a></li>
<li class="chapter" data-level="64.8.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade"><i class="fa fa-check"></i><b>64.8.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="64.8.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length"><i class="fa fa-check"></i><b>64.8.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="64.8.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-distance-from-clade-to-all-other-points"><i class="fa fa-check"></i><b>64.8.4</b> Calculate distance from clade to all other points</a></li>
<li class="chapter" data-level="64.8.5" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-matrix-with-new-distance"><i class="fa fa-check"></i><b>64.8.5</b> Update matrix with new distance</a></li>
</ul></li>
<li class="chapter" data-level="64.9" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#algorithm---next-round"><i class="fa fa-check"></i><b>64.9</b> Algorithm - next round</a>
<ul>
<li class="chapter" data-level="64.9.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-1"><i class="fa fa-check"></i><b>64.9.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="64.9.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-1"><i class="fa fa-check"></i><b>64.9.2</b> Calculate branch length</a></li>
<li class="chapter" data-level="64.9.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#recalculate-distance-from-abe-to-all-remainign-taxa"><i class="fa fa-check"></i><b>64.9.3</b> Recalculate distance from abe to all remainign taxa</a></li>
<li class="chapter" data-level="64.9.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#update-the-matrix"><i class="fa fa-check"></i><b>64.9.4</b> Update the matrix</a></li>
</ul></li>
<li class="chapter" data-level="64.10" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#next-iteration"><i class="fa fa-check"></i><b>64.10</b> Next iteration</a>
<ul>
<li class="chapter" data-level="64.10.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-lengths"><i class="fa fa-check"></i><b>64.10.1</b> Calculate branch lengths</a></li>
<li class="chapter" data-level="64.10.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-2"><i class="fa fa-check"></i><b>64.10.2</b> Combine species into clade</a></li>
<li class="chapter" data-level="64.10.3" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-2"><i class="fa fa-check"></i><b>64.10.3</b> Calculate branch length</a></li>
<li class="chapter" data-level="64.10.4" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#distance-of-c-to-abe-and-d-to-abe"><i class="fa fa-check"></i><b>64.10.4</b> Distance of c to abe and d to abe</a></li>
</ul></li>
<li class="chapter" data-level="64.11" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finish-up"><i class="fa fa-check"></i><b>64.11</b> Finish up</a>
<ul>
<li class="chapter" data-level="64.11.1" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#combine-species-into-clade-3"><i class="fa fa-check"></i><b>64.11.1</b> Combine species into clade</a></li>
<li class="chapter" data-level="64.11.2" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#calculate-branch-length-3"><i class="fa fa-check"></i><b>64.11.2</b> Calculate branch length</a></li>
</ul></li>
<li class="chapter" data-level="64.12" data-path="upgma-the-hard-way.html"><a href="upgma-the-hard-way.html#finalizing-branch-lengths"><i class="fa fa-check"></i><b>64.12</b> Finalizing branch lengths</a></li>
</ul></li>
<li class="chapter" data-level="65" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html"><i class="fa fa-check"></i><b>65</b> Representing phylogenetic trees in Newick format</a>
<ul>
<li class="chapter" data-level="65.1" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#vocab-7"><i class="fa fa-check"></i><b>65.1</b> Vocab</a></li>
<li class="chapter" data-level="65.2" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#introduction-14"><i class="fa fa-check"></i><b>65.2</b> Introduction</a></li>
<li class="chapter" data-level="65.3" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-three-taxa"><i class="fa fa-check"></i><b>65.3</b> Trees with three taxa</a></li>
<li class="chapter" data-level="65.4" data-path="representing-phylogenetic-trees-in-newick-format.html"><a href="representing-phylogenetic-trees-in-newick-format.html#trees-with-four-taxa"><i class="fa fa-check"></i><b>65.4</b> Trees with four taxa</a></li>
</ul></li>
<li class="chapter" data-level="66" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html"><i class="fa fa-check"></i><b>66</b> Plotting trees from Newick notation in R</a>
<ul>
<li class="chapter" data-level="66.1" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#preliminaries-19"><i class="fa fa-check"></i><b>66.1</b> Preliminaries</a></li>
<li class="chapter" data-level="66.2" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#note-1"><i class="fa fa-check"></i><b>66.2</b> Note</a></li>
<li class="chapter" data-level="66.3" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick"><i class="fa fa-check"></i><b>66.3</b> 3 taxa tree in Newick</a></li>
<li class="chapter" data-level="66.4" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-tree-in-newick-1"><i class="fa fa-check"></i><b>66.4</b> 4 taxa tree in Newick</a></li>
<li class="chapter" data-level="66.5" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#taxa-in-newick-format"><i class="fa fa-check"></i><b>66.5</b> 5 taxa in Newick Format</a></li>
<li class="chapter" data-level="66.6" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#many-taxa-tree"><i class="fa fa-check"></i><b>66.6</b> Many-taxa tree</a></li>
<li class="chapter" data-level="66.7" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#newick-and-branch-lengths"><i class="fa fa-check"></i><b>66.7</b> Newick and Branch lengths</a></li>
<li class="chapter" data-level="66.8" data-path="plotting-trees-from-newick-notation-in-r.html"><a href="plotting-trees-from-newick-notation-in-r.html#additional-reading"><i class="fa fa-check"></i><b>66.8</b> Additional Reading</a></li>
</ul></li>
<li class="chapter" data-level="67" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><i class="fa fa-check"></i><b>67</b> Phylogenetic trees using the UPGMA clustering algoirthm</a>
<ul>
<li class="chapter" data-level="67.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#introduction-15"><i class="fa fa-check"></i><b>67.1</b> Introduction</a></li>
<li class="chapter" data-level="67.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#preliminaries-20"><i class="fa fa-check"></i><b>67.2</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="67.2.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#vocab-8"><i class="fa fa-check"></i><b>67.2.1</b> Vocab</a></li>
<li class="chapter" data-level="67.2.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-r-functions"><i class="fa fa-check"></i><b>67.2.2</b> General R functions</a></li>
<li class="chapter" data-level="67.2.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#specific-functions"><i class="fa fa-check"></i><b>67.2.3</b> Specific functions</a></li>
</ul></li>
<li class="chapter" data-level="67.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#sequence-comparisons"><i class="fa fa-check"></i><b>67.3</b> Sequence comparisons</a></li>
<li class="chapter" data-level="67.4" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrices"><i class="fa fa-check"></i><b>67.4</b> Distance matrices</a></li>
<li class="chapter" data-level="67.5" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#from-a-distance-matrix-to-phylogenetic-tree"><i class="fa fa-check"></i><b>67.5</b> From a distance matrix to phylogenetic tree</a></li>
<li class="chapter" data-level="67.6" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-raw-sequences"><i class="fa fa-check"></i><b>67.6</b> Distance matrix in R from raw sequences</a></li>
<li class="chapter" data-level="67.7" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#distance-matrix-in-r-from-normal-matrix"><i class="fa fa-check"></i><b>67.7</b> Distance matrix in R from normal matrix</a></li>
<li class="chapter" data-level="67.8" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#upgma-in-r"><i class="fa fa-check"></i><b>67.8</b> UPGMA in R</a></li>
<li class="chapter" data-level="67.9" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#how-upgma-clustering-works"><i class="fa fa-check"></i><b>67.9</b> How UPGMA clustering works</a>
<ul>
<li class="chapter" data-level="67.9.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#challenge-2"><i class="fa fa-check"></i><b>67.9.1</b> Challenge</a></li>
</ul></li>
<li class="chapter" data-level="67.10" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#further-information"><i class="fa fa-check"></i><b>67.10</b> Further information</a>
<ul>
<li class="chapter" data-level="67.10.1" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#general-info"><i class="fa fa-check"></i><b>67.10.1</b> General info</a></li>
<li class="chapter" data-level="67.10.2" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#worked-examples"><i class="fa fa-check"></i><b>67.10.2</b> Worked examples:</a></li>
<li class="chapter" data-level="67.10.3" data-path="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html"><a href="phylogenetic-trees-using-the-upgma-clustering-algoirthm.html#additional-notes"><i class="fa fa-check"></i><b>67.10.3</b> Additional notes:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="68" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><i class="fa fa-check"></i><b>68</b> Tutorial: Writing functions to calculate the number of phylogenetic trees</a>
<ul>
<li class="chapter" data-level="68.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#repliminaries-1"><i class="fa fa-check"></i><b>68.1</b> Repliminaries</a>
<ul>
<li class="chapter" data-level="68.1.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-7"><i class="fa fa-check"></i><b>68.1.1</b> Functions</a></li>
</ul></li>
<li class="chapter" data-level="68.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#concepts-1"><i class="fa fa-check"></i><b>68.2</b> Concepts</a>
<ul>
<li class="chapter" data-level="68.2.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vocab-9"><i class="fa fa-check"></i><b>68.2.1</b> Vocab:</a></li>
<li class="chapter" data-level="68.2.2" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#packages-10"><i class="fa fa-check"></i><b>68.2.2</b> Packages:</a></li>
</ul></li>
<li class="chapter" data-level="68.3" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#introduction-16"><i class="fa fa-check"></i><b>68.3</b> Introduction</a></li>
<li class="chapter" data-level="68.4" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-1"><i class="fa fa-check"></i><b>68.4</b> Number of rooted trees</a></li>
<li class="chapter" data-level="68.5" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#number-of-rooted-trees-in-r-1"><i class="fa fa-check"></i><b>68.5</b> Number of rooted trees in R</a></li>
<li class="chapter" data-level="68.6" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#functions-in-r-1"><i class="fa fa-check"></i><b>68.6</b> Functions in R</a></li>
<li class="chapter" data-level="68.7" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#function-to-calculate-the-number-of-possible-phylogenetic-trees"><i class="fa fa-check"></i><b>68.7</b> Function to calculate the number of possible phylogenetic trees</a></li>
<li class="chapter" data-level="68.8" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#vectorized-inputs-to-functions"><i class="fa fa-check"></i><b>68.8</b> Vectorized inputs to functions</a></li>
<li class="chapter" data-level="68.9" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-conditional-statements"><i class="fa fa-check"></i><b>68.9</b> Adding conditional statements</a></li>
<li class="chapter" data-level="68.10" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-multiple-conditional-statements"><i class="fa fa-check"></i><b>68.10</b> Adding multiple conditional statements</a></li>
<li class="chapter" data-level="68.11" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#adding-additional-arguments"><i class="fa fa-check"></i><b>68.11</b> Adding additional arguments</a></li>
<li class="chapter" data-level="68.12" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-_un_rooted-trees"><i class="fa fa-check"></i><b>68.12</b> Assignment: _un_rooted trees</a></li>
<li class="chapter" data-level="68.13" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-1"><i class="fa fa-check"></i><b>68.13</b> Assignment part 1</a>
<ul>
<li class="chapter" data-level="68.13.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#part-1-answer"><i class="fa fa-check"></i><b>68.13.1</b> Part 1 answer</a></li>
</ul></li>
<li class="chapter" data-level="68.14" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#assignment-part-2"><i class="fa fa-check"></i><b>68.14</b> Assignment part 2</a>
<ul>
<li class="chapter" data-level="68.14.1" data-path="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html"><a href="tutorial-writing-functions-to-calculate-the-number-of-phylogenetic-trees.html#answer-1"><i class="fa fa-check"></i><b>68.14.1</b> Answer</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="69" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>69</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="69.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>69.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="69.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>69.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="69.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>69.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="69.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>69.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="69.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>69.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-3"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="69.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>69.2</b> Help!</a>
<ul>
<li class="chapter" data-level="69.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>69.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="69.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>69.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="69.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>69.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="69.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>69.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="69.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>69.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="69.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>69.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="69.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>69.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="69.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>69.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="worked-example-building-a-phylogeny-in-r" class="section level1" number="51">
<h1><span class="header-section-number">Chapter 51</span> “Worked example: Building a phylogeny in R”</h1>
<div id="introduction-9" class="section level2" number="51.1">
<h2><span class="header-section-number">51.1</span> Introduction</h2>
<p>Phylogenies play an important role in computational biology and bioinformatics. Phylogenetics itself is an obligatly computational field that only began rapid growth when computational power allowed the many algorithms it relies on to be done rapidly. Phylogeneies of species, genes and proteins are used to address many biological issues, including</p>
<ul>
<li>Patterns of protein evolution</li>
<li>Origin and evolution of phenotypic traits</li>
<li>Origin and progression of epidemics</li>
<li>Origin of evolution of diseases (eg, zooenoses)</li>
<li>Prediction of protein function from its sequence</li>
<li>… and many more</li>
</ul>
<p>The actual building of a phylogeny is a computationally intensive task; moreover, there are many bioinformatics and computational tasks the precede the construction of a phylogeny:</p>
<ul>
<li>genome sequencing and assembly</li>
<li>computational gene prediction and annotation</li>
<li>database searching and results screening</li>
<li>pairwise sequence alignment</li>
<li>data organization and cleaning</li>
<li>multiple sequence alignment</li>
<li>evaluation and validation of alignment accuracy</li>
</ul>
<p>Once all of these steps have been carried out, the building of a phylogeny involves</p>
<ul>
<li>picking a model of sequence evolution or other description of evolution</li>
<li>picking a statistical approach to tree construction</li>
<li>evaluationg uncertainty in the final tree</li>
</ul>
<p>In this chapter we will work through many of these steps. In most cases we will pick the easiest or fastest option; in later chapters we will unpack the various options. This chapter is written as an interactive R sessions. You can follow along by opening the .Rmd file of the chapter or typing the appropriate commands into your own script. I assume that all the necessary packages have been installed and they only need to be loaded into R using the <em>library()</em> command.</p>
<p>This lesson walks you through and entire workflow for a bioinformatics, including</p>
<!-- check all of this -->
<ol style="list-style-type: decimal">
<li>obtaining FASTA sequences</li>
<li>cleaning sequences</li>
<li>creating alignments</li>
<li>creating distance a distance matrix</li>
<li>building a phylogenetic tree</li>
</ol>
<p>We’ll examine the Shroom family of genes, which produces Shroom proteins essential for tissue formation in many multicellular eukaryotes, including neural tube formation in vertebrates. We’ll examine shroom in severall very different organism, including humans, mice and sea urchins. There is more than one type of shroom in vertebrates, and we’ll also look at two different Shroom genes: shroom 1 and shroom 2.</p>
<p>This lesson draws on skills from previous sections of the book, but is written to act as a independent summary of these activities. There is therefore review of key aspects of R and bioinformatics throughout it.</p>
</div>
<div id="software-preliminaires" class="section level2" number="51.2">
<h2><span class="header-section-number">51.2</span> Software Preliminaires</h2>
<div id="vocab-6" class="section level3" number="51.2.1">
<h3><span class="header-section-number">51.2.1</span> Vocab</h3>
<ul>
<li>arguement</li>
<li>function</li>
<li>list</li>
<li>named list</li>
<li>vector</li>
<li>named vector</li>
<li>for() loop</li>
<li>R console</li>
</ul>
</div>
<div id="r-functions-1" class="section level3" number="51.2.2">
<h3><span class="header-section-number">51.2.2</span> R functions</h3>
<ul>
<li>library()</li>
<li>round()</li>
<li>plot()</li>
<li>mtext()</li>
<li>nchar()</li>
<li>rentrez::entrez_fetch()</li>
<li>combio4all::entrez_fetch_list()</li>
<li>combio4aal::print_msa() (Coghlan 2011)</li>
<li>Biostrings::AAStringSet()</li>
<li>msa::msa()</li>
<li>msa::msaConvert()</li>
<li>msa::msaPrettyPrint()</li>
<li>seqinr::dist.alignment()</li>
<li>ape::nj()</li>
</ul>
<p>A few things need to be done to get started with our R session.</p>
</div>
<div id="download-necessary-packages" class="section level3" number="51.2.3">
<h3><span class="header-section-number">51.2.3</span> Download necessary packages</h3>
<p>Many <em>R</em> sessions begin by downloading necessary software packages to augment <em>R’s</em> functionality.</p>
<p>If you don’t have them already, you’ll need the following packages from CRAN:</p>
<ol style="list-style-type: decimal">
<li><code>ape</code></li>
<li><code>seqinr</code></li>
<li><code>rentrez</code></li>
<li><code>devtools</code></li>
</ol>
<p>The CRAN packages can be loaded with <code>install.packages()</code>.</p>
<p>You’ll also need these packages from Bioconductor:</p>
<ol style="list-style-type: decimal">
<li><code>msa</code></li>
<li><code>Biostrings</code></li>
</ol>
<p>For install packages from Bioconductor, see the chapter at the beginning of this book on this process.</p>
<p>Finally, you’ll need this package from GitHub</p>
<ol style="list-style-type: decimal">
<li><code>compbio4all</code></li>
</ol>
<p>To install packages from GitHub you can use the code <code>devtools::install_github("brouwern/combio4all")</code></p>
</div>
<div id="load-packages-into-memory" class="section level3" number="51.2.4">
<h3><span class="header-section-number">51.2.4</span> Load packages into memory</h3>
<p>We noew need to load up all our bioinformatics and phylogenetics software into R. This is done with the <code>library()</code> command.</p>
<p>To run this code just clock on the sideways green triangel all the way to the right of the code.</p>
<p>NOTE: You’ll likely see some red code appear on your screen. No worries, totally normal!</p>
<div class="sourceCode" id="cb987"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb987-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-1" aria-hidden="true" tabindex="-1"></a><span class="co"># github packages</span></span>
<span id="cb987-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compbio4all)</span>
<span id="cb987-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb987-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-4" aria-hidden="true" tabindex="-1"></a><span class="co"># CRAN packages</span></span>
<span id="cb987-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rentrez)</span>
<span id="cb987-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(seqinr)</span>
<span id="cb987-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb987-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb987-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Bioconductor packages</span></span>
<span id="cb987-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(msa)</span>
<span id="cb987-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb987-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span></code></pre></div>
</div>
</div>
<div id="downloading-macro-molecular-sequences" class="section level2" number="51.3">
<h2><span class="header-section-number">51.3</span> Downloading macro-molecular sequences</h2>
<p>We’re going to explore some sequences. First we need to download them. To do this we’ll use a function, <code>entrez_fretch()</code>, which accesses the <a href="ncbi.nlm.nih.gov/search/"><strong>Entrez</strong></a> system of database (ncbi.nlm.nih.gov/search/). This function is from the <code>rentrez</code> package, which stands for “R-Entrez.”</p>
<p>We need to tell <code>entrez_fetch()</code> several things</p>
<ol style="list-style-type: decimal">
<li><code>db = ...</code> the type of entrez database.<br />
</li>
<li><code>id = ...</code> the <strong>accession</strong> (ID) number of the sequence</li>
<li><code>rettype = ...</code> file type what we want the function to return.</li>
</ol>
<p>Formally, these things are called <strong>arguements</strong> by <em>R</em>.</p>
<p>We’ll use these settings:</p>
<ol style="list-style-type: decimal">
<li><code>db = "protein"</code> to access the Entrez database of protein sequences<br />
</li>
<li><code>rettype = "fasta"</code>, which is a standard file format for nucleic acid and protein sequences</li>
</ol>
<p>We’ll set <code>id = ...</code> to sequences whose <strong>accession numbers</strong> are:</p>
<ol style="list-style-type: decimal">
<li>NP_065910: Human shroom 3</li>
<li>AAF13269: Mouse shroom 3a</li>
<li>CAA58534: Human shroom 2</li>
<li>XP_783573: Sea urchin shroom</li>
</ol>
<p>There are two highly conserved regions of shroom 3
1. ASD 1: aa 884 to aa 1062 in hShroom3
1. ASD 2: aa 1671 to aa 1955 in hShroom3</p>
<p>Normally we’d have to download these sequences by hand through pointing and clicking on GeneBank records on the NCBI website. In <em>R</em> we can do it automatically; this might take a second.</p>
<p>All the code needed is this:</p>
<div class="sourceCode" id="cb988"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb988-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb988-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Human shroom 3 (H. sapiens)</span></span>
<span id="cb988-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb988-2" aria-hidden="true" tabindex="-1"></a>hShroom3 <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb988-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb988-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;NP_065910&quot;</span>, </span>
<span id="cb988-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb988-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>The output is in FASTA format; we’ll use the <code>cat()</code> to do a little formating for us:</p>
<div class="sourceCode" id="cb989"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb989-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb989-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(hShroom3)</span></code></pre></div>
<pre><code>## &gt;NP_065910.3 protein Shroom3 [Homo sapiens]
## MMRTTEDFHKPSATLNSNTATKGRYIYLEAFLEGGAPWGFTLKGGLEHGEPLIISKVEEGGKADTLSSKL
## QAGDEVVHINEVTLSSSRKEAVSLVKGSYKTLRLVVRRDVCTDPGHADTGASNFVSPEHLTSGPQHRKAA
## WSGGVKLRLKHRRSEPAGRPHSWHTTKSGEKQPDASMMQISQGMIGPPWHQSYHSSSSTSDLSNYDHAYL
## RRSPDQCSSQGSMESLEPSGAYPPCHLSPAKSTGSIDQLSHFHNKRDSAYSSFSTSSSILEYPHPGISGR
## ERSGSMDNTSARGGLLEGMRQADIRYVKTVYDTRRGVSAEYEVNSSALLLQGREARASANGQGYDKWSNI
## PRGKGVPPPSWSQQCPSSLETATDNLPPKVGAPLPPARSDSYAAFRHRERPSSWSSLDQKRLCRPQANSL
## GSLKSPFIEEQLHTVLEKSPENSPPVKPKHNYTQKAQPGQPLLPTSIYPVPSLEPHFAQVPQPSVSSNGM
## LYPALAKESGYIAPQGACNKMATIDENGNQNGSGRPGFAFCQPLEHDLLSPVEKKPEATAKYVPSKVHFC
## SVPENEEDASLKRHLTPPQGNSPHSNERKSTHSNKPSSHPHSLKCPQAQAWQAGEDKRSSRLSEPWEGDF
## QEDHNANLWRRLEREGLGQSLSGNFGKTKSAFSSLQNIPESLRRHSSLELGRGTQEGYPGGRPTCAVNTK
## AEDPGRKAAPDLGSHLDRQVSYPRPEGRTGASASFNSTDPSPEEPPAPSHPHTSSLGRRGPGPGSASALQ
## GFQYGKPHCSVLEKVSKFEQREQGSQRPSVGGSGFGHNYRPHRTVSTSSTSGNDFEETKAHIRFSESAEP
## LGNGEQHFKNGELKLEEASRQPCGQQLSGGASDSGRGPQRPDARLLRSQSTFQLSSEPEREPEWRDRPGS
## PESPLLDAPFSRAYRNSIKDAQSRVLGATSFRRRDLELGAPVASRSWRPRPSSAHVGLRSPEASASASPH
## TPRERHSVTPAEGDLARPVPPAARRGARRRLTPEQKKRSYSEPEKMNEVGIVEEAEPAPLGPQRNGMRFP
## ESSVADRRRLFERDGKACSTLSLSGPELKQFQQSALADYIQRKTGKRPTSAAGCSLQEPGPLRERAQSAY
## LQPGPAALEGSGLASASSLSSLREPSLQPRREATLLPATVAETQQAPRDRSSSFAGGRRLGERRRGDLLS
## GANGGTRGTQRGDETPREPSSWGARAGKSMSAEDLLERSDVLAGPVHVRSRSSPATADKRQDVLLGQDSG
## FGLVKDPCYLAGPGSRSLSCSERGQEEMLPLFHHLTPRWGGSGCKAIGDSSVPSECPGTLDHQRQASRTP
## CPRPPLAGTQGLVTDTRAAPLTPIGTPLPSAIPSGYCSQDGQTGRQPLPPYTPAMMHRSNGHTLTQPPGP
## RGCEGDGPEHGVEEGTRKRVSLPQWPPPSRAKWAHAAREDSLPEESSAPDFANLKHYQKQQSLPSLCSTS
## DPDTPLGAPSTPGRISLRISESVLRDSPPPHEDYEDEVFVRDPHPKATSSPTFEPLPPPPPPPPSQETPV
## YSMDDFPPPPPHTVCEAQLDSEDPEGPRPSFNKLSKVTIARERHMPGAAHVVGSQTLASRLQTSIKGSEA
## ESTPPSFMSVHAQLAGSLGGQPAPIQTQSLSHDPVSGTQGLEKKVSPDPQKSSEDIRTEALAKEIVHQDK
## SLADILDPDSRLKTTMDLMEGLFPRDVNLLKENSVKRKAIQRTVSSSGCEGKRNEDKEAVSMLVNCPAYY
## SVSAPKAELLNKIKEMPAEVNEEEEQADVNEKKAELIGSLTHKLETLQEAKGSLLTDIKLNNALGEEVEA
## LISELCKPNEFDKYRMFIGDLDKVVNLLLSLSGRLARVENVLSGLGEDASNEERSSLYEKRKILAGQHED
## ARELKENLDRRERVVLGILANYLSEEQLQDYQHFVKMKSTLLIEQRKLDDKIKLGQEQVKCLLESLPSDF
## IPKAGALALPPNLTSEPIPAGGCTFSGIFPTLTSPL</code></pre>
<p>Note the initial <code>&gt;</code>, then the header line of <code>NP_065910.3 protein Shroom3 [Homo sapiens]</code>. After that is the amino acid sequence. The underlying data also includes the <strong>newline character</strong> <code>\n</code> to designate where each line of amino acids stops.</p>
<p>We can get the rest of the data by just chaing the <code>id = ...</code> argument:</p>
<div class="sourceCode" id="cb991"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb991-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mouse shroom 3a (M. musculus)</span></span>
<span id="cb991-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-2" aria-hidden="true" tabindex="-1"></a>mShroom3a <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb991-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;AAF13269&quot;</span>, </span>
<span id="cb991-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb991-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb991-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Human shroom 2 (H. sapiens)</span></span>
<span id="cb991-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-7" aria-hidden="true" tabindex="-1"></a>hShroom2 <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb991-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-8" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;CAA58534&quot;</span>, </span>
<span id="cb991-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb991-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb991-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb991-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Sea-urchin shroom</span></span>
<span id="cb991-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-13" aria-hidden="true" tabindex="-1"></a>sShroom <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb991-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-14" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;XP_783573&quot;</span>, </span>
<span id="cb991-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb991-15" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>I’m going to check about how long each of these sequences is - each should have an at least slightly different length. If any are identical, I might have repeated an accession name or re-used an object name. The function <code>nchar()</code> counts of the number of characters in an <em>R</em> object.</p>
<div class="sourceCode" id="cb992"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb992-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb992-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(hShroom3)</span></code></pre></div>
<pre><code>## [1] 2070</code></pre>
<div class="sourceCode" id="cb994"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb994-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb994-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(mShroom3a)</span></code></pre></div>
<pre><code>## [1] 2083</code></pre>
<div class="sourceCode" id="cb996"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb996-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb996-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(sShroom)</span></code></pre></div>
<pre><code>## [1] 1758</code></pre>
<div class="sourceCode" id="cb998"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb998-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb998-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(hShroom2)</span></code></pre></div>
<pre><code>## [1] 1673</code></pre>
</div>
<div id="prepping-macromolecular-sequences" class="section level2" number="51.4">
<h2><span class="header-section-number">51.4</span> Prepping macromolecular sequences</h2>
<blockquote>
<p>“90% of data analysis is data cleaning” (-Just about every data analyst and data scientist on twitter)</p>
</blockquote>
<p>We have our sequences, but the current format isn’t directly usable for us yet because there are several thigns that aren’t sequence information</p>
<ol style="list-style-type: decimal">
<li>metadata (the header)</li>
<li>page formatting information (the newline character)</li>
</ol>
<p>We can remove this non-sequence information using a function I wrote called <code>fasta_cleaner()</code>, which is in the <code>compbio4all</code> package. The function uses <strong>regular expressions</strong> to remove the info we don’t need.</p>
<p><strong>ASIDE</strong>: If we run the name of the command with out any quotation marks we can see the code:</p>
<div class="sourceCode" id="cb1000"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1000-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1000-1" aria-hidden="true" tabindex="-1"></a>fasta_cleaner</span></code></pre></div>
<pre><code>## function(fasta_object, parse = TRUE){
## 
##   fasta_object &lt;- sub(&quot;^(&gt;)(.*?)(\\n)(.*)(\\n\\n)&quot;,&quot;\\4&quot;,fasta_object)
##   fasta_object &lt;- gsub(&quot;\n&quot;, &quot;&quot;, fasta_object)
## 
##   if(parse == TRUE){
##     fasta_object &lt;- stringr::str_split(fasta_object,
##                                        pattern = &quot;&quot;,
##                                        simplify = FALSE)
##   }
## 
##   return(fasta_object[[1]])
## }
## &lt;bytecode: 0x7fa045a5bfb0&gt;
## &lt;environment: namespace:compbio4all&gt;</code></pre>
<p><strong>End ASIDE</strong></p>
<p>Now use the function to clean our sequences; we won’t worry about what <code>pare = ...</code> is for.</p>
<div class="sourceCode" id="cb1002"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1002-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1002-1" aria-hidden="true" tabindex="-1"></a>hShroom3  <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(hShroom3,  <span class="at">parse =</span> F)</span>
<span id="cb1002-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1002-2" aria-hidden="true" tabindex="-1"></a>mShroom3a <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(mShroom3a, <span class="at">parse =</span> F)</span>
<span id="cb1002-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1002-3" aria-hidden="true" tabindex="-1"></a>hShroom2  <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(hShroom2,  <span class="at">parse =</span> F)</span>
<span id="cb1002-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb1002-4" aria-hidden="true" tabindex="-1"></a>sShroom   <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(sShroom,   <span class="at">parse =</span> F)</span></code></pre></div>
<p>Now let’s take a peek at what our sequences look like:</p>
<div class="sourceCode" id="cb1003"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1003-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1003-1" aria-hidden="true" tabindex="-1"></a>hShroom3</span></code></pre></div>
<pre><code>## [1] &quot;MMRTTEDFHKPSATLNSNTATKGRYIYLEAFLEGGAPWGFTLKGGLEHGEPLIISKVEEGGKADTLSSKLQAGDEVVHINEVTLSSSRKEAVSLVKGSYKTLRLVVRRDVCTDPGHADTGASNFVSPEHLTSGPQHRKAAWSGGVKLRLKHRRSEPAGRPHSWHTTKSGEKQPDASMMQISQGMIGPPWHQSYHSSSSTSDLSNYDHAYLRRSPDQCSSQGSMESLEPSGAYPPCHLSPAKSTGSIDQLSHFHNKRDSAYSSFSTSSSILEYPHPGISGRERSGSMDNTSARGGLLEGMRQADIRYVKTVYDTRRGVSAEYEVNSSALLLQGREARASANGQGYDKWSNIPRGKGVPPPSWSQQCPSSLETATDNLPPKVGAPLPPARSDSYAAFRHRERPSSWSSLDQKRLCRPQANSLGSLKSPFIEEQLHTVLEKSPENSPPVKPKHNYTQKAQPGQPLLPTSIYPVPSLEPHFAQVPQPSVSSNGMLYPALAKESGYIAPQGACNKMATIDENGNQNGSGRPGFAFCQPLEHDLLSPVEKKPEATAKYVPSKVHFCSVPENEEDASLKRHLTPPQGNSPHSNERKSTHSNKPSSHPHSLKCPQAQAWQAGEDKRSSRLSEPWEGDFQEDHNANLWRRLEREGLGQSLSGNFGKTKSAFSSLQNIPESLRRHSSLELGRGTQEGYPGGRPTCAVNTKAEDPGRKAAPDLGSHLDRQVSYPRPEGRTGASASFNSTDPSPEEPPAPSHPHTSSLGRRGPGPGSASALQGFQYGKPHCSVLEKVSKFEQREQGSQRPSVGGSGFGHNYRPHRTVSTSSTSGNDFEETKAHIRFSESAEPLGNGEQHFKNGELKLEEASRQPCGQQLSGGASDSGRGPQRPDARLLRSQSTFQLSSEPEREPEWRDRPGSPESPLLDAPFSRAYRNSIKDAQSRVLGATSFRRRDLELGAPVASRSWRPRPSSAHVGLRSPEASASASPHTPRERHSVTPAEGDLARPVPPAARRGARRRLTPEQKKRSYSEPEKMNEVGIVEEAEPAPLGPQRNGMRFPESSVADRRRLFERDGKACSTLSLSGPELKQFQQSALADYIQRKTGKRPTSAAGCSLQEPGPLRERAQSAYLQPGPAALEGSGLASASSLSSLREPSLQPRREATLLPATVAETQQAPRDRSSSFAGGRRLGERRRGDLLSGANGGTRGTQRGDETPREPSSWGARAGKSMSAEDLLERSDVLAGPVHVRSRSSPATADKRQDVLLGQDSGFGLVKDPCYLAGPGSRSLSCSERGQEEMLPLFHHLTPRWGGSGCKAIGDSSVPSECPGTLDHQRQASRTPCPRPPLAGTQGLVTDTRAAPLTPIGTPLPSAIPSGYCSQDGQTGRQPLPPYTPAMMHRSNGHTLTQPPGPRGCEGDGPEHGVEEGTRKRVSLPQWPPPSRAKWAHAAREDSLPEESSAPDFANLKHYQKQQSLPSLCSTSDPDTPLGAPSTPGRISLRISESVLRDSPPPHEDYEDEVFVRDPHPKATSSPTFEPLPPPPPPPPSQETPVYSMDDFPPPPPHTVCEAQLDSEDPEGPRPSFNKLSKVTIARERHMPGAAHVVGSQTLASRLQTSIKGSEAESTPPSFMSVHAQLAGSLGGQPAPIQTQSLSHDPVSGTQGLEKKVSPDPQKSSEDIRTEALAKEIVHQDKSLADILDPDSRLKTTMDLMEGLFPRDVNLLKENSVKRKAIQRTVSSSGCEGKRNEDKEAVSMLVNCPAYYSVSAPKAELLNKIKEMPAEVNEEEEQADVNEKKAELIGSLTHKLETLQEAKGSLLTDIKLNNALGEEVEALISELCKPNEFDKYRMFIGDLDKVVNLLLSLSGRLARVENVLSGLGEDASNEERSSLYEKRKILAGQHEDARELKENLDRRERVVLGILANYLSEEQLQDYQHFVKMKSTLLIEQRKLDDKIKLGQEQVKCLLESLPSDFIPKAGALALPPNLTSEPIPAGGCTFSGIFPTLTSPL&quot;</code></pre>
</div>
<div id="aligning-sequences" class="section level2" number="51.5">
<h2><span class="header-section-number">51.5</span> Aligning sequences</h2>
<p>We can do a <a href="https://tinyurl.com/y4du73zq"><strong>global alignment</strong></a> of one sequence against another using the <code>pairwiseAlignment()</code> function from the <strong>Bioconductor</strong> package <code>Biostrings</code> (note that capital “B” in <code>Biostrings</code>; most <em>R</em> package names are all lower case, but not this one).</p>
<p>Let’s align human versus mouse shroom:</p>
<div class="sourceCode" id="cb1005"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1005-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1005-1" aria-hidden="true" tabindex="-1"></a>align.h3.vs.m3a <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">pairwiseAlignment</span>(</span>
<span id="cb1005-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1005-2" aria-hidden="true" tabindex="-1"></a>                  hShroom3,</span>
<span id="cb1005-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1005-3" aria-hidden="true" tabindex="-1"></a>                  mShroom3a)</span></code></pre></div>
<p>We can peek at the alignment</p>
<div class="sourceCode" id="cb1006"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1006-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1006-1" aria-hidden="true" tabindex="-1"></a>align.h3.vs.m3a</span></code></pre></div>
<pre><code>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: MMRTTEDFHKPSATLN-SNTATKGRYIYLEAFLE...KAGALALPPNLTSEPIPAGGCTFSGIFPTLTSPL
## subject: MK-TPENLEEPSATPNPSRTPTE-RFVYLEALLE...KAGAISLPPALTGHATPGGTSVFGGVFPTLTSPL
## score: 2189.934</code></pre>
<p>The <strong>score</strong> tells us how closely they are aligned; higher scores mean the sequences are more similar. Its hard to interpret the number on its own so we can get the <strong>percent sequence identity (PID)</strong> using the <code>pid()</code> function.</p>
<div class="sourceCode" id="cb1008"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1008-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1008-1" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">pid</span>(align.h3.vs.m3a)</span></code></pre></div>
<pre><code>## [1] 70.56511</code></pre>
<p>So, <em>shroom3</em> from humans and <em>shroom3</em> from mice are ~71% similar (at least using this particular method of alignment, and there are many ways to do this!)</p>
<p>What about human shroom 3 and sea-urchin shroom?</p>
<div class="sourceCode" id="cb1010"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1010-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1010-1" aria-hidden="true" tabindex="-1"></a>align.h3.vs.h2 <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">pairwiseAlignment</span>(</span>
<span id="cb1010-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1010-2" aria-hidden="true" tabindex="-1"></a>                  hShroom3,</span>
<span id="cb1010-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1010-3" aria-hidden="true" tabindex="-1"></a>                  hShroom2)</span></code></pre></div>
<p>First check out the score using <code>score()</code>, which accesses it directly without all the other information.</p>
<div class="sourceCode" id="cb1011"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1011-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1011-1" aria-hidden="true" tabindex="-1"></a><span class="fu">score</span>(align.h3.vs.h2)</span></code></pre></div>
<pre><code>## [1] -5673.853</code></pre>
<p>Now the percent sequence alignment with <code>pid()</code>:</p>
<div class="sourceCode" id="cb1013"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1013-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1013-1" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">pid</span>(align.h3.vs.h2)</span></code></pre></div>
<pre><code>## [1] 33.83277</code></pre>
<p>So Human shroom 3 and Mouse shroom 3 are 71% identical, but Human shroom 3 and human shroom 2 are only 34% similar? How does it work out evolutionary that a human and mouse gene are more similar than a human and a human gene? What are the evolutionary relationships among these genes within the shroom gene family?</p>
</div>
<div id="the-shroom-family-of-genes" class="section level2" number="51.6">
<h2><span class="header-section-number">51.6</span> The shroom family of genes</h2>
<p>I’ve copied a table from a published paper which has accession numbers for 15 different Shroom genes.</p>
<div class="sourceCode" id="cb1015"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1015-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-1" aria-hidden="true" tabindex="-1"></a>shroom_table <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;CAA78718&quot;</span> , <span class="st">&quot;X. laevis Apx&quot;</span> ,         <span class="st">&quot;xShroom1&quot;</span>,</span>
<span id="cb1015-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-2" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;NP_597713&quot;</span> , <span class="st">&quot;H. sapiens APXL2&quot;</span> ,     <span class="st">&quot;hShroom1&quot;</span>,</span>
<span id="cb1015-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-3" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;CAA58534&quot;</span> , <span class="st">&quot;H. sapiens APXL&quot;</span>,        <span class="st">&quot;hShroom2&quot;</span>,</span>
<span id="cb1015-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-4" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;ABD19518&quot;</span> , <span class="st">&quot;M. musculus Apxl&quot;</span> ,      <span class="st">&quot;mShroom2&quot;</span>,</span>
<span id="cb1015-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-5" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;AAF13269&quot;</span> , <span class="st">&quot;M. musculus ShroomL&quot;</span> ,   <span class="st">&quot;mShroom3a&quot;</span>,</span>
<span id="cb1015-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-6" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;AAF13270&quot;</span> , <span class="st">&quot;M. musculus ShroomS&quot;</span> ,   <span class="st">&quot;mShroom3b&quot;</span>,</span>
<span id="cb1015-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-7" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;NP_065910&quot;</span>, <span class="st">&quot;H. sapiens Shroom&quot;</span> ,     <span class="st">&quot;hShroom3&quot;</span>,</span>
<span id="cb1015-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-8" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;ABD59319&quot;</span> , <span class="st">&quot;X. laevis Shroom-like&quot;</span>,  <span class="st">&quot;xShroom3&quot;</span>,</span>
<span id="cb1015-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-9" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;NP_065768&quot;</span>, <span class="st">&quot;H. sapiens KIAA1202&quot;</span> ,   <span class="st">&quot;hShroom4a&quot;</span>,</span>
<span id="cb1015-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-10" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;AAK95579&quot;</span> , <span class="st">&quot;H. sapiens SHAP-A&quot;</span> ,     <span class="st">&quot;hShroom4b&quot;</span>,</span>
<span id="cb1015-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-11" aria-hidden="true" tabindex="-1"></a>            <span class="co">#&quot;DQ435686&quot; , &quot;M. musculus KIAA1202&quot; ,  &quot;mShroom4&quot;,</span></span>
<span id="cb1015-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-12" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;ABA81834&quot;</span> , <span class="st">&quot;D. melanogaster Shroom&quot;</span>, <span class="st">&quot;dmShroom&quot;</span>,</span>
<span id="cb1015-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-13" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;EAA12598&quot;</span> , <span class="st">&quot;A. gambiae Shroom&quot;</span>,      <span class="st">&quot;agShroom&quot;</span>,</span>
<span id="cb1015-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-14" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;XP_392427&quot;</span> , <span class="st">&quot;A. mellifera Shroom&quot;</span> ,  <span class="st">&quot;amShroom&quot;</span>,</span>
<span id="cb1015-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb1015-15" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;XP_783573&quot;</span> , <span class="st">&quot;S. purpuratus Shroom&quot;</span> , <span class="st">&quot;spShroom&quot;</span>) <span class="co">#sea urchin</span></span></code></pre></div>
<p>I’ll do a bit of formatting; you can ignore these details if you want</p>
<div class="sourceCode" id="cb1016"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1016-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to matrix</span></span>
<span id="cb1016-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-2" aria-hidden="true" tabindex="-1"></a>shroom_table_matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(shroom_table,</span>
<span id="cb1016-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">byrow =</span> T,</span>
<span id="cb1016-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">nrow =</span> <span class="dv">14</span>)</span>
<span id="cb1016-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-5" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to dataframe</span></span>
<span id="cb1016-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-6" aria-hidden="true" tabindex="-1"></a>shroom_table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(shroom_table_matrix, </span>
<span id="cb1016-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-7" aria-hidden="true" tabindex="-1"></a>                     <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb1016-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1016-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-9" aria-hidden="true" tabindex="-1"></a><span class="co"># name columns</span></span>
<span id="cb1016-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(shroom_table) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;accession&quot;</span>, <span class="st">&quot;name.orig&quot;</span>,<span class="st">&quot;name.new&quot;</span>)</span>
<span id="cb1016-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1016-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Create simplified species names</span></span>
<span id="cb1016-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-13" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp <span class="ot">&lt;-</span> <span class="st">&quot;Homo&quot;</span></span>
<span id="cb1016-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-14" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;laevis&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;Xenopus&quot;</span></span>
<span id="cb1016-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-15" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;musculus&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;Mus&quot;</span></span>
<span id="cb1016-16"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-16" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;melanogaster&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;Drosophila&quot;</span></span>
<span id="cb1016-17"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-17" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;gambiae&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;mosquito&quot;</span></span>
<span id="cb1016-18"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-18" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;mellifera&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;bee&quot;</span></span>
<span id="cb1016-19"><a href="worked-example-building-a-phylogeny-in-r.html#cb1016-19" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;purpuratus&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;sea urchin&quot;</span></span></code></pre></div>
<p>Take a look:</p>
<div class="sourceCode" id="cb1017"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1017-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1017-1" aria-hidden="true" tabindex="-1"></a>shroom_table</span></code></pre></div>
<pre><code>##    accession              name.orig  name.new        spp
## 1   CAA78718          X. laevis Apx  xShroom1    Xenopus
## 2  NP_597713       H. sapiens APXL2  hShroom1       Homo
## 3   CAA58534        H. sapiens APXL  hShroom2       Homo
## 4   ABD19518       M. musculus Apxl  mShroom2        Mus
## 5   AAF13269    M. musculus ShroomL mShroom3a        Mus
## 6   AAF13270    M. musculus ShroomS mShroom3b        Mus
## 7  NP_065910      H. sapiens Shroom  hShroom3       Homo
## 8   ABD59319  X. laevis Shroom-like  xShroom3    Xenopus
## 9  NP_065768    H. sapiens KIAA1202 hShroom4a       Homo
## 10  AAK95579      H. sapiens SHAP-A hShroom4b       Homo
## 11  ABA81834 D. melanogaster Shroom  dmShroom Drosophila
## 12  EAA12598      A. gambiae Shroom  agShroom   mosquito
## 13 XP_392427    A. mellifera Shroom  amShroom        bee
## 14 XP_783573   S. purpuratus Shroom  spShroom sea urchin</code></pre>
</div>
<div id="downloading-multiple-sequences" class="section level2" number="51.7">
<h2><span class="header-section-number">51.7</span> Downloading multiple sequences</h2>
<p>Instead of getting one sequence at a time we can download several by accessing the “accession” column from the table</p>
<div class="sourceCode" id="cb1019"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1019-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1019-1" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>accession</span></code></pre></div>
<pre><code>##  [1] &quot;CAA78718&quot;  &quot;NP_597713&quot; &quot;CAA58534&quot;  &quot;ABD19518&quot;  &quot;AAF13269&quot;  &quot;AAF13270&quot; 
##  [7] &quot;NP_065910&quot; &quot;ABD59319&quot;  &quot;NP_065768&quot; &quot;AAK95579&quot;  &quot;ABA81834&quot;  &quot;EAA12598&quot; 
## [13] &quot;XP_392427&quot; &quot;XP_783573&quot;</code></pre>
<p>We can give this whole set of accessions to <code>entrez_fetch()</code>:</p>
<div class="sourceCode" id="cb1021"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1021-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1021-1" aria-hidden="true" tabindex="-1"></a>shrooms <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb1021-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1021-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> shroom_table<span class="sc">$</span>accession, </span>
<span id="cb1021-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1021-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>We can look at what we got here with <code>cat()</code> (I won’t display this because it is very long!)</p>
<div class="sourceCode" id="cb1022"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1022-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1022-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(shrooms)</span></code></pre></div>
<p>The current format of these data is a single, long set of data. This is a standard way to store, share and transmit FASTA files, but in <em>R</em> we’ll need a slightly different format.</p>
<p>We’ll download all of the sequences again, this time using a function from <code>compbio4all</code> called <code>entrez_fetch_list()</code> which is a <strong>wrapper</strong> function I wrote to put the output of <code>entrez_fetch()</code> into an <em>R</em> data format called a <strong>list</strong>.</p>
<div class="sourceCode" id="cb1023"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1023-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1023-1" aria-hidden="true" tabindex="-1"></a>shrooms_list <span class="ot">&lt;-</span> <span class="fu">entrez_fetch_list</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb1023-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1023-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> shroom_table<span class="sc">$</span>accession, </span>
<span id="cb1023-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1023-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>Now we have a list which as 14 <strong>elements</strong>, one for each sequence in our table.</p>
<div class="sourceCode" id="cb1024"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1024-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1024-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(shrooms_list)</span></code></pre></div>
<pre><code>## [1] 14</code></pre>
<p>We now need to clean up each one fo these sequences. We can do that using a simple <code>for()</code> loop:</p>
<div class="sourceCode" id="cb1026"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1026-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1026-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(shrooms_list)){</span>
<span id="cb1026-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1026-2" aria-hidden="true" tabindex="-1"></a>  shrooms_list[[i]] <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(shrooms_list[[i]], <span class="at">parse =</span> F)</span>
<span id="cb1026-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1026-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Second to last step: we need to take each one of our sequences from our list and put it into a <strong>vector</strong>, in particular a <strong>named vector</strong></p>
<div class="sourceCode" id="cb1027"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1027-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a vector to store output</span></span>
<span id="cb1027-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-2" aria-hidden="true" tabindex="-1"></a>shrooms_vector <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">length</span>(shrooms_list))</span>
<span id="cb1027-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1027-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-4" aria-hidden="true" tabindex="-1"></a><span class="co"># run the loop</span></span>
<span id="cb1027-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(shrooms_vector)){</span>
<span id="cb1027-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-6" aria-hidden="true" tabindex="-1"></a>  shrooms_vector[i] <span class="ot">&lt;-</span> shrooms_list[[i]]</span>
<span id="cb1027-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1027-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1027-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-9" aria-hidden="true" tabindex="-1"></a><span class="co"># name the vector</span></span>
<span id="cb1027-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb1027-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(shrooms_vector) <span class="ot">&lt;-</span> <span class="fu">names</span>(shrooms_list)</span></code></pre></div>
<p>Now the final step: we need to convert our named vector to a <strong>string set</strong> using <code>Biostrings::AAStringSet()</code>. Note the <code>_ss</code> tag at the end of the object we’re assigning the output to, which designates this as a string set.</p>
<div class="sourceCode" id="cb1028"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1028-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1028-1" aria-hidden="true" tabindex="-1"></a>shrooms_vector_ss <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">AAStringSet</span>(shrooms_vector)</span></code></pre></div>
</div>
<div id="multiple-sequence-alignment" class="section level2" number="51.8">
<h2><span class="header-section-number">51.8</span> Multiple sequence alignment</h2>
<p>We must <strong>align</strong> all of the sequences we downloaded and use that <strong>alignment</strong> to build a <strong>phylogenetic tree</strong>. This will tell us how the different genes, both within and between species, are likely to be related.</p>
<div id="building-an-multiple-sequence-alignment-msa" class="section level3" number="51.8.1">
<h3><span class="header-section-number">51.8.1</span> Building an Multiple Sequence Alignment (MSA)</h3>
<p>We’ll use the software <code>msa,</code> which implements the <strong>ClustalW</strong> multiple sequence alignment algorithm. Normally we’d have to download the ClustalW program and either point-and-click our way through it or use the <strong>command line</strong>*, but these folks wrote up the algorithm in R so we can do this with a line of R code. This will take a second or two.</p>
<div class="sourceCode" id="cb1029"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1029-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1029-1" aria-hidden="true" tabindex="-1"></a>shrooms_align <span class="ot">&lt;-</span> <span class="fu">msa</span>(shrooms_vector_ss,</span>
<span id="cb1029-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1029-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">method =</span> <span class="st">&quot;ClustalW&quot;</span>)</span></code></pre></div>
<pre><code>## use default substitution matrix</code></pre>
</div>
<div id="viewing-an-msa" class="section level3" number="51.8.2">
<h3><span class="header-section-number">51.8.2</span> Viewing an MSA</h3>
<p>Once we build an MSA we need to visualize it.</p>
<div id="viewing-an-msa-in-r" class="section level4" number="51.8.2.1">
<h4><span class="header-section-number">51.8.2.1</span> Viewing an MSA in R</h4>
<p>We can look at the output from <code>msa()</code>, but its not very helpful</p>
<div class="sourceCode" id="cb1031"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1031-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1031-1" aria-hidden="true" tabindex="-1"></a>shrooms_align</span></code></pre></div>
<pre><code>## CLUSTAL 2.1  
## 
## Call:
##    msa(shrooms_vector_ss, method = &quot;ClustalW&quot;)
## 
## MsaAAMultipleAlignment with 14 rows and 2252 columns
##      aln                                                   names
##  [1] -------------------------...------------------------- NP_065768
##  [2] -------------------------...------------------------- AAK95579
##  [3] -------------------------...SVFGGVFPTLTSPL----------- AAF13269
##  [4] -------------------------...SVFGGVFPTLTSPL----------- AAF13270
##  [5] -------------------------...CTFSGIFPTLTSPL----------- NP_065910
##  [6] -------------------------...NKS--LPPPLTSSL----------- ABD59319
##  [7] -------------------------...------------------------- CAA58534
##  [8] -------------------------...------------------------- ABD19518
##  [9] -------------------------...LT----------------------- NP_597713
## [10] -------------------------...------------------------- CAA78718
## [11] -------------------------...------------------------- EAA12598
## [12] -------------------------...------------------------- ABA81834
## [13] MTELQPSPPGYRVQDEAPGPPSCPP...------------------------- XP_392427
## [14] -------------------------...AATSSSSNGIGGPEQLNSNATSSYC XP_783573
##  Con -------------------------...------------------------- Consensus</code></pre>
<p>A function called <code>print_msa()</code> (Coghlan 2011) which I’ve put into<code>combio4all</code> can give us more informative output by printing out the actual alignment into the R console.</p>
<p>To use <code>print_msa()</code> We need to make a few minor tweaks though first. These are behind the scenes changes so don’t worry about the details right now. We’ll change the name to <code>shrooms_align_seqinr</code> to indcate that one of our changes is putting this into a format defined by the bioinformatics package <code>seqinr</code>.</p>
<div class="sourceCode" id="cb1033"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1033-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1033-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(shrooms_align) <span class="ot">&lt;-</span> <span class="st">&quot;AAMultipleAlignment&quot;</span></span>
<span id="cb1033-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1033-2" aria-hidden="true" tabindex="-1"></a>shrooms_align_seqinr <span class="ot">&lt;-</span> <span class="fu">msaConvert</span>(shrooms_align, <span class="at">type =</span> <span class="st">&quot;seqinr::alignment&quot;</span>)</span></code></pre></div>
<p>I won’t display the output from <code>shrooms_align_seqinr</code> because its very long; we have 14 shroom genes, and shroom happens to be a rather long gene.
<!-- TODO: set up for displaying only part of the alignment, subsetting it etc. --></p>
<div class="sourceCode" id="cb1034"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1034-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1034-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print_msa</span>(<span class="at">alignment =</span> shrooms_align_seqinr, </span>
<span id="cb1034-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1034-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">chunksize =</span> <span class="dv">60</span>)</span></code></pre></div>
</div>
<div id="displaying-an-msa-as-an-r-plot" class="section level4" number="51.8.2.2">
<h4><span class="header-section-number">51.8.2.2</span> Displaying an MSA as an R plot</h4>
<p>I’m going to just show about 100 amino acids near the end of the alignment, where there is the most overlap across all of the sequences. This is set with the <code>start = ...</code> and <code>end = ...</code> arguments. Note that we’re using the <code>shrooms_align</code> object.</p>
<div class="sourceCode" id="cb1035"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1035-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1035-1" aria-hidden="true" tabindex="-1"></a>ggmsa<span class="sc">::</span><span class="fu">ggmsa</span>(shrooms_align,   <span class="co"># shrooms_align, NOT shrooms_align_seqinr</span></span>
<span id="cb1035-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1035-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">start =</span> <span class="dv">2000</span>, </span>
<span id="cb1035-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1035-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">end =</span> <span class="dv">2100</span>) </span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-670-1.png" width="672" /></p>
</div>
<div id="saving-an-msa-as-pdf" class="section level4" number="51.8.2.3">
<h4><span class="header-section-number">51.8.2.3</span> Saving an MSA as PDF</h4>
<p>We can take a look at the alignment in PDF format if we want. I this case I’m going to just show about 100 amino acids near the end of the alignment, where there is the most overlap across all of the sequences. This is set with the <code>y = c(...)</code> argument.</p>
<div class="sourceCode" id="cb1036"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1036-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1036-1" aria-hidden="true" tabindex="-1"></a><span class="fu">msaPrettyPrint</span>(shrooms_align,             <span class="co"># alignment</span></span>
<span id="cb1036-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1036-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">file =</span> <span class="st">&quot;shroom_msa.pdf&quot;</span>,   <span class="co"># file name</span></span>
<span id="cb1036-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1036-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">y=</span><span class="fu">c</span>(<span class="dv">2000</span>, <span class="dv">2100</span>),           <span class="co"># range</span></span>
<span id="cb1036-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb1036-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">askForOverwrite=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>You can see where R is saving thigns by running <code>getwd()</code></p>
<div class="sourceCode" id="cb1037"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1037-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1037-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code></pre></div>
<pre><code>## [1] &quot;/Users/nlb24/google_backup_sync_nlb24/lbrb&quot;</code></pre>
<p>On a Mac you can usually find the file by searching in Finder for the file name, which I set to be “shroom_msa.pdf” using the <code>file = ...</code> arguement above.</p>
</div>
</div>
</div>
<div id="genetic-distance." class="section level2" number="51.9">
<h2><span class="header-section-number">51.9</span> Genetic distance.</h2>
<p>Next need to first get an estimate of how similar each sequences is. The more amino acids that are identical to each other, the more similar.</p>
<p>Instead of similarity, we usually work in terms of <em>difference</em> or <strong>genetic distance</strong> (a.k.a. <strong>evolutionary distance</strong>). This is done with the <code>dist.alignment()</code> function.</p>
<div class="sourceCode" id="cb1039"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1039-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1039-1" aria-hidden="true" tabindex="-1"></a>shrooms_dist <span class="ot">&lt;-</span> seqinr<span class="sc">::</span><span class="fu">dist.alignment</span>(shrooms_align_seqinr, </span>
<span id="cb1039-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1039-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">matrix =</span> <span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p>We’ve made a matrix using <code>dist.alignment()</code>; let’s round it off so its easier to look at using the <code>round()</code> function.</p>
<div class="sourceCode" id="cb1040"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1040-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1040-1" aria-hidden="true" tabindex="-1"></a>shrooms_dist_rounded <span class="ot">&lt;-</span> <span class="fu">round</span>(shrooms_dist,</span>
<span id="cb1040-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1040-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">digits =</span> <span class="dv">3</span>)</span></code></pre></div>
<p>Now let’s look at it</p>
<div class="sourceCode" id="cb1041"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1041-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1041-1" aria-hidden="true" tabindex="-1"></a>shrooms_dist_rounded</span></code></pre></div>
<pre><code>##           NP_065768 AAK95579 AAF13269 AAF13270 NP_065910 ABD59319 CAA58534
## AAK95579      0.000                                                       
## AAF13269      0.884    0.917                                              
## AAF13270      0.897    0.917    0.000                                     
## NP_065910     0.878    0.912    0.533    0.536                            
## ABD59319      0.893    0.921    0.783    0.783     0.782                  
## CAA58534      0.872    0.908    0.838    0.849     0.840    0.864         
## ABD19518      0.866    0.912    0.834    0.846     0.838    0.855    0.548
## NP_597713     0.916    0.939    0.903    0.903     0.902    0.904    0.896
## CAA78718      0.925    0.955    0.896    0.895     0.893    0.893    0.898
## EAA12598      0.914    0.947    0.899    0.899     0.902    0.897    0.891
## ABA81834      0.938    0.943    0.935    0.934     0.936    0.940    0.935
## XP_392427     0.936    0.963    0.935    0.934     0.938    0.941    0.938
## XP_783573     0.940    0.958    0.942    0.939     0.942    0.935    0.942
##           ABD19518 NP_597713 CAA78718 EAA12598 ABA81834 XP_392427
## AAK95579                                                         
## AAF13269                                                         
## AAF13270                                                         
## NP_065910                                                        
## ABD59319                                                         
## CAA58534                                                         
## ABD19518                                                         
## NP_597713    0.900                                               
## CAA78718     0.891     0.919                                     
## EAA12598     0.896     0.920    0.922                            
## ABA81834     0.935     0.932    0.946    0.882                   
## XP_392427    0.934     0.927    0.947    0.878    0.923          
## XP_783573    0.946     0.947    0.941    0.925    0.954     0.943</code></pre>
</div>
<div id="phylognetic-trees-finally" class="section level2" number="51.10">
<h2><span class="header-section-number">51.10</span> Phylognetic trees (finally!)</h2>
<p>We got our sequence, built multiple sequence alignment, and calcualted the genetic distance between sequences. Now we are - finally - ready to build a phylogenetic tree.</p>
<p>First, we let R figure out the structure of the tree. There are <strong>MANY</strong> ways to build phylogenetic trees. We’ll use a common one used for exploring sequences called <strong>neighbor joining</strong> algorithm via the function <code>nj()</code>. Neighbor joining uses genetic distances to cluster sequences into <strong>clades</strong>.</p>
<div class="sourceCode" id="cb1043"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1043-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1043-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">nj</span>(shrooms_dist)</span></code></pre></div>
<div id="plotting-phylogenetic-trees" class="section level3" number="51.10.1">
<h3><span class="header-section-number">51.10.1</span> Plotting phylogenetic trees</h3>
<p>Now we’ll make a quick plot of our tree using <code>plot()</code> (and add a little label using a function called <code>mtext()</code>).</p>
<div class="sourceCode" id="cb1044"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1044-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1044-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot tree</span></span>
<span id="cb1044-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1044-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span> (tree, <span class="at">main=</span><span class="st">&quot;Phylogenetic Tree&quot;</span>, </span>
<span id="cb1044-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1044-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">type =</span> <span class="st">&quot;unrooted&quot;</span>, </span>
<span id="cb1044-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb1044-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">use.edge.length =</span> F)</span>
<span id="cb1044-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb1044-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1044-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb1044-6" aria-hidden="true" tabindex="-1"></a><span class="co"># add label</span></span>
<span id="cb1044-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb1044-7" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">text =</span> <span class="st">&quot;Shroom family gene tree - unrooted, no branch lengths&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-677-1.png" width="672" /></p>
<p>This is an **unrooted tree*. For the sake of plotting we’ve also ignored the evolutionary distance between the sequences.</p>
<p>To make a rooted tree we remove <code>type = "unrooted</code>.</p>
<div class="sourceCode" id="cb1045"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1045-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1045-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot tree</span></span>
<span id="cb1045-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1045-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span> (tree, <span class="at">main=</span><span class="st">&quot;Phylogenetic Tree&quot;</span>, </span>
<span id="cb1045-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1045-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">use.edge.length =</span> F)</span>
<span id="cb1045-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb1045-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1045-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb1045-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add label</span></span>
<span id="cb1045-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb1045-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">text =</span> <span class="st">&quot;Shroom family gene tree - rooted, no branch lenths&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-678-1.png" width="672" /></p>
<p>We can include information about branch length by setting <code>use.edge.length = ...</code> to <code>T</code>.</p>
<div class="sourceCode" id="cb1046"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1046-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1046-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot tree</span></span>
<span id="cb1046-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1046-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span> (tree, <span class="at">main=</span><span class="st">&quot;Phylogenetic Tree&quot;</span>, </span>
<span id="cb1046-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1046-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">use.edge.length =</span> T)</span>
<span id="cb1046-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb1046-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1046-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb1046-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add label</span></span>
<span id="cb1046-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb1046-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">text =</span> <span class="st">&quot;Shroom family gene tree - rooted, with branch lenths&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-679-1.png" width="672" /></p>
<p>Some of the branches are now very short, but most are very long, indicating that these genes have been evolvin independently for many millions of years.</p>
<p>Let’s make a fancier plot. Don’t worry about all the steps; I’ve added some more code to add some annotations on the right-hand side to help us see what’s going on.</p>
<!-- TODO: how does this render? -->
<div class="sourceCode" id="cb1047"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1047-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree, <span class="at">main=</span><span class="st">&quot;Phylogenetic Tree&quot;</span>)</span>
<span id="cb1047-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">text =</span> <span class="st">&quot;Shroom family gene tree&quot;</span>)</span>
<span id="cb1047-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-4" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fl">0.551</span></span>
<span id="cb1047-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-5" aria-hidden="true" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> <span class="fl">0.6</span></span>
<span id="cb1047-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-7" aria-hidden="true" tabindex="-1"></a><span class="co"># label Shrm 3</span></span>
<span id="cb1047-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-8" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> x, <span class="at">y0 =</span> <span class="dv">1</span>, </span>
<span id="cb1047-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> x, <span class="at">y1 =</span> <span class="dv">4</span>,</span>
<span id="cb1047-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb1047-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-11" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> x<span class="sc">*</span><span class="fl">1.01</span>, <span class="at">y =</span> <span class="fl">2.5</span>, <span class="st">&quot;Shrm 3&quot;</span>,<span class="at">adj =</span> <span class="dv">0</span>)</span>
<span id="cb1047-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-13" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> x, <span class="at">y0 =</span> <span class="dv">5</span>, </span>
<span id="cb1047-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> x, <span class="at">y1 =</span> <span class="dv">6</span>,</span>
<span id="cb1047-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb1047-16"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-16" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> x<span class="sc">*</span><span class="fl">1.01</span>, <span class="at">y =</span> <span class="fl">5.5</span>, <span class="st">&quot;Shrm 2&quot;</span>,<span class="at">adj =</span> <span class="dv">0</span>)</span>
<span id="cb1047-17"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-18"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-18" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> x, <span class="at">y0 =</span> <span class="dv">7</span>, </span>
<span id="cb1047-19"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> x, <span class="at">y1 =</span> <span class="dv">9</span>,</span>
<span id="cb1047-20"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-20" aria-hidden="true" tabindex="-1"></a>         <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb1047-21"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-21" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> x<span class="sc">*</span><span class="fl">1.01</span>, <span class="at">y =</span> <span class="dv">8</span>, <span class="st">&quot;Shrm 1&quot;</span>,<span class="at">adj =</span> <span class="dv">0</span>)</span>
<span id="cb1047-22"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-23"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-23" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> x, <span class="at">y0 =</span> <span class="dv">10</span>, </span>
<span id="cb1047-24"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-24" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> x, <span class="at">y1 =</span> <span class="dv">13</span>,</span>
<span id="cb1047-25"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-25" aria-hidden="true" tabindex="-1"></a>         <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb1047-26"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-26" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> x<span class="sc">*</span><span class="fl">1.01</span>, <span class="at">y =</span> <span class="dv">12</span>, <span class="st">&quot;Shrm ?&quot;</span>,<span class="at">adj =</span> <span class="dv">0</span>)</span>
<span id="cb1047-27"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-28"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-29"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-29" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> x, <span class="at">y0 =</span> <span class="dv">14</span>, </span>
<span id="cb1047-30"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-30" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> x, <span class="at">y1 =</span> <span class="dv">15</span>,</span>
<span id="cb1047-31"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-31" aria-hidden="true" tabindex="-1"></a>         <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb1047-32"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-32" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> x<span class="sc">*</span><span class="fl">1.01</span>, <span class="at">y =</span> <span class="fl">14.5</span>, <span class="st">&quot;Shrm 4&quot;</span>,<span class="at">adj =</span> <span class="dv">0</span>)</span>
<span id="cb1047-33"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-34"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-35"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-35" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> x2, <span class="at">y0 =</span> <span class="dv">1</span>, </span>
<span id="cb1047-36"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-36" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> x2, <span class="at">y1 =</span> <span class="dv">6</span>,</span>
<span id="cb1047-37"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-37" aria-hidden="true" tabindex="-1"></a>         <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb1047-38"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-39"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-39" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> x2, <span class="at">y0 =</span> <span class="dv">7</span>, </span>
<span id="cb1047-40"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-40" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> x2, <span class="at">y1 =</span> <span class="dv">9</span>,</span>
<span id="cb1047-41"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-41" aria-hidden="true" tabindex="-1"></a>         <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb1047-42"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1047-43"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-43" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> x2, <span class="at">y0 =</span> <span class="dv">10</span>, </span>
<span id="cb1047-44"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-44" aria-hidden="true" tabindex="-1"></a>         <span class="at">x1 =</span> x2, <span class="at">y1 =</span> <span class="dv">15</span>,</span>
<span id="cb1047-45"><a href="worked-example-building-a-phylogeny-in-r.html#cb1047-45" aria-hidden="true" tabindex="-1"></a>         <span class="at">lwd=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-680-1.png" width="576" /></p>
<!-- Assignment - redo with DNA; how does PID compare -->
<!-- compare pariwise PID to MSA similarity -->

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="a-complete-bioinformatics-workflow-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="calculating-genetic-distances-between-sequences.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/005-MSA/23-MSA-walkthrough-shroom.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/005-MSA/23-MSA-walkthrough-shroom.Rmd",
"text": null
},
"download": ["lbrb.pdf", "lbrb.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
